// @sharpee/world-model/traits/moveable-scenery/moveableSceneryTrait.js
"use strict";
/**
 * Moveable scenery trait implementation
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.MoveableSceneryTrait = void 0;
const trait_types_1 = require("../trait-types");
/**
 * Moveable scenery trait for large pushable/pullable objects
 *
 * Objects with this trait should also have PUSHABLE and/or PULLABLE
 * traits to define how they can be moved. This trait adds properties
 * specific to large moveable scenery objects.
 */
class MoveableSceneryTrait {
    static type = trait_types_1.TraitType.MOVEABLE_SCENERY;
    type = trait_types_1.TraitType.MOVEABLE_SCENERY;
    // MoveableSceneryData properties
    weightClass;
    revealsWhenMoved;
    reveals;
    blocksExits;
    blockedExits;
    moved;
    originalRoom;
    moveSound;
    requiresMultiplePeople;
    peopleRequired;
    constructor(data = {}) {
        this.weightClass = data.weightClass ?? 'heavy';
        this.revealsWhenMoved = data.revealsWhenMoved ?? false;
        this.reveals = data.reveals;
        this.blocksExits = data.blocksExits ?? false;
        this.blockedExits = data.blockedExits;
        this.moved = data.moved ?? false;
        this.originalRoom = data.originalRoom;
        this.moveSound = data.moveSound;
        this.requiresMultiplePeople = data.requiresMultiplePeople ?? false;
        this.peopleRequired = data.peopleRequired;
    }
}
exports.MoveableSceneryTrait = MoveableSceneryTrait;
