// @sharpee/stdlib/actions/standard/inserting/inserting-data.js
"use strict";
/**
 * Data builder for inserting action
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.insertedDataConfig = exports.buildInsertedData = void 0;
const snapshot_utils_1 = require("../../base/snapshot-utils");
const buildInsertedData = (context) => {
    const item = context.command.directObject?.entity;
    const container = context.command.indirectObject?.entity;
    if (!item || !container) {
        return { item: 'nothing', container: 'nowhere' };
    }
    return {
        item: item.name,
        itemSnapshot: (0, snapshot_utils_1.captureEntitySnapshot)(item, context.world, true),
        container: container.name,
        containerSnapshot: (0, snapshot_utils_1.captureEntitySnapshot)(container, context.world, true),
        itemId: item.id,
        containerId: container.id
    };
};
exports.buildInsertedData = buildInsertedData;
exports.insertedDataConfig = {
    builder: exports.buildInsertedData,
    protectedFields: ['item', 'container', 'itemId', 'containerId']
};
