// @sharpee/world-model/traits/pullable/pullableTrait.js
"use strict";
/**
 * Pullable trait implementation
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PullableTrait = void 0;
const trait_types_1 = require("../trait-types");
/**
 * Pullable trait for objects that can be pulled
 *
 * This trait contains only data - all logic for pulling
 * is handled by the pulling action.
 */
class PullableTrait {
    static type = trait_types_1.TraitType.PULLABLE;
    type = trait_types_1.TraitType.PULLABLE;
    // PullableData properties
    pullType;
    activates;
    linkedTo;
    pullSound;
    requiresStrength;
    repeatable;
    state;
    pullCount;
    maxPulls;
    detachesOnPull;
    effects;
    constructor(data = {}) {
        this.pullType = data.pullType ?? 'lever';
        this.activates = data.activates;
        this.linkedTo = data.linkedTo;
        this.pullSound = data.pullSound;
        this.requiresStrength = data.requiresStrength;
        this.repeatable = data.repeatable ?? true;
        this.state = data.state ?? 'default';
        this.pullCount = data.pullCount ?? 0;
        this.maxPulls = data.maxPulls;
        this.detachesOnPull = data.detachesOnPull ?? false;
        this.effects = data.effects;
    }
}
exports.PullableTrait = PullableTrait;
