// @sharpee/world-model/traits/container/container-utils.js
"use strict";
// packages/world-model/src/traits/container/container-utils.ts
Object.defineProperty(exports, "__esModule", { value: true });
exports.isContainerCapable = isContainerCapable;
exports.canContain = canContain;
exports.getContainerTrait = getContainerTrait;
exports.hasContainerProperties = hasContainerProperties;
const trait_types_1 = require("../trait-types");
/**
 * Check if a trait provides container functionality
 */
function isContainerCapable(trait) {
    return (trait.type === trait_types_1.TraitType.CONTAINER ||
        trait.type === trait_types_1.TraitType.ROOM ||
        trait.type === trait_types_1.TraitType.ACTOR);
}
/**
 * Check if an entity can contain other entities
 */
function canContain(entity) {
    // Check by entity type first (rooms and actors are containers by default)
    if (entity.type === 'room' || entity.type === 'actor') {
        return true;
    }
    // Then check for explicit container traits
    return (entity.hasTrait(trait_types_1.TraitType.CONTAINER) ||
        entity.hasTrait(trait_types_1.TraitType.ROOM) ||
        entity.hasTrait(trait_types_1.TraitType.ACTOR) ||
        entity.hasTrait(trait_types_1.TraitType.SUPPORTER) || // Supporters can also hold things
        entity.hasTrait(trait_types_1.TraitType.ENTERABLE) ||
        entity.hasTrait(trait_types_1.TraitType.VEHICLE));
}
/**
 * Get the container trait from an entity, regardless of which trait provides it
 */
function getContainerTrait(entity) {
    // Check for explicit container trait first
    const container = entity.getTrait(trait_types_1.TraitType.CONTAINER);
    if (container && isContainerCapable(container)) {
        return container;
    }
    // Check for room trait
    const room = entity.getTrait(trait_types_1.TraitType.ROOM);
    if (room && isContainerCapable(room)) {
        return room;
    }
    // Check for actor trait
    const actor = entity.getTrait(trait_types_1.TraitType.ACTOR);
    if (actor && isContainerCapable(actor)) {
        return actor;
    }
    // For entities without traits, provide default container capabilities based on type
    if (entity.type === 'room' || entity.type === 'actor') {
        // Return a default container capability
        return {
            type: entity.type === 'room' ? trait_types_1.TraitType.ROOM : trait_types_1.TraitType.ACTOR,
            isTransparent: entity.type === 'room', // Rooms are transparent by default
            enterable: entity.type === 'room', // Only rooms are enterable
            capacity: undefined // No capacity limits by default
        };
    }
    return undefined;
}
/**
 * Type guard to check if a trait has container properties
 */
function hasContainerProperties(trait) {
    return (typeof trait === 'object' &&
        trait !== null &&
        typeof trait.isTransparent === 'boolean' &&
        typeof trait.enterable === 'boolean');
}
