// @sharpee/world-model/traits/container/containerBehavior.js
"use strict";
// packages/world-model/src/traits/container/containerBehavior.ts
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContainerBehavior = void 0;
const behavior_1 = require("../../behaviors/behavior");
const trait_types_1 = require("../trait-types");
const identityBehavior_1 = require("../identity/identityBehavior");
const wearableBehavior_1 = require("../wearable/wearableBehavior");
/**
 * Behavior for entities that can contain other entities.
 *
 * Handles capacity checks, type restrictions, and container state.
 */
class ContainerBehavior extends behavior_1.Behavior {
    static requiredTraits = [trait_types_1.TraitType.CONTAINER];
    /**
     * Check if a container can accept an item
     * @returns true if the item can be accepted
     */
    static canAccept(container, item, world) {
        const trait = ContainerBehavior.require(container, trait_types_1.TraitType.CONTAINER);
        // Check if container is accessible (must be open if openable)
        if (container.has(trait_types_1.TraitType.OPENABLE)) {
            const openable = container.get(trait_types_1.TraitType.OPENABLE);
            if (openable && !openable.isOpen) {
                return false;
            }
        }
        // Check type restrictions
        if (trait.allowedTypes && trait.allowedTypes.length > 0) {
            const itemType = item.type || 'object';
            if (!trait.allowedTypes.includes(itemType)) {
                return false;
            }
        }
        if (trait.excludedTypes && trait.excludedTypes.length > 0) {
            const itemType = item.type || 'object';
            if (trait.excludedTypes.includes(itemType)) {
                return false;
            }
        }
        // Check enterable restrictions for actors
        if (item.has(trait_types_1.TraitType.ACTOR) && !trait.enterable) {
            return false;
        }
        // Check capacity constraints
        if (trait.capacity) {
            const result = this.checkCapacity(container, item, world);
            if (!result)
                return false;
        }
        return true;
    }
    /**
     * Check capacity constraints
     */
    static checkCapacity(container, item, world) {
        const trait = ContainerBehavior.require(container, trait_types_1.TraitType.CONTAINER);
        const capacity = trait.capacity;
        // If no capacity defined, allow any items
        if (!capacity) {
            return true;
        }
        // Check item count limit (excluding worn items for actors)
        if (capacity.maxItems !== undefined) {
            const contents = world.getContents(container.id);
            const currentCount = contents.filter(item => !wearableBehavior_1.WearableBehavior.isWorn(item)).length;
            if (currentCount >= capacity.maxItems) {
                return false;
            }
        }
        // Check weight limit
        if (capacity.maxWeight !== undefined) {
            const currentWeight = this.getTotalWeight(container, world);
            const itemWeight = identityBehavior_1.IdentityBehavior.getWeight(item);
            if (currentWeight + itemWeight > capacity.maxWeight) {
                return false;
            }
        }
        // Check volume limit
        if (capacity.maxVolume !== undefined) {
            const currentVolume = this.getTotalVolume(container, world);
            const itemVolume = identityBehavior_1.IdentityBehavior.getVolume(item);
            if (currentVolume + itemVolume > capacity.maxVolume) {
                return false;
            }
        }
        return true;
    }
    /**
     * Get total weight of contents (not recursive - container bears the weight)
     */
    static getTotalWeight(container, world) {
        const contents = world.getContents(container.id);
        return contents.reduce((sum, item) => {
            return sum + identityBehavior_1.IdentityBehavior.getWeight(item);
        }, 0);
    }
    /**
     * Get total volume of contents
     */
    static getTotalVolume(container, world) {
        const contents = world.getContents(container.id);
        return contents.reduce((sum, item) => {
            return sum + identityBehavior_1.IdentityBehavior.getVolume(item);
        }, 0);
    }
    /**
     * Get remaining capacity
     */
    static getRemainingCapacity(container, world) {
        const trait = ContainerBehavior.require(container, trait_types_1.TraitType.CONTAINER);
        const result = {};
        if (trait.capacity?.maxItems !== undefined) {
            const current = world.getContents(container.id).length;
            result.items = trait.capacity.maxItems - current;
        }
        if (trait.capacity?.maxWeight !== undefined) {
            result.weight = trait.capacity.maxWeight - this.getTotalWeight(container, world);
        }
        if (trait.capacity?.maxVolume !== undefined) {
            result.volume = trait.capacity.maxVolume - this.getTotalVolume(container, world);
        }
        return result;
    }
    /**
     * Check if the container is transparent (can see contents when closed)
     */
    static isTransparent(container) {
        const trait = ContainerBehavior.require(container, trait_types_1.TraitType.CONTAINER);
        return trait.isTransparent || false;
    }
    /**
     * Check if the container is enterable by actors
     */
    static isEnterable(container) {
        const trait = ContainerBehavior.require(container, trait_types_1.TraitType.CONTAINER);
        return trait.enterable || false;
    }
    /**
     * Get allowed types for this container
     */
    static getAllowedTypes(container) {
        const trait = ContainerBehavior.require(container, trait_types_1.TraitType.CONTAINER);
        return trait.allowedTypes;
    }
    /**
     * Get excluded types for this container
     */
    static getExcludedTypes(container) {
        const trait = ContainerBehavior.require(container, trait_types_1.TraitType.CONTAINER);
        return trait.excludedTypes;
    }
    /**
     * Add an item to a container
     * @param container The container to add to
     * @param item The item to add
     * @param world World query interface for getting current state
     * @returns Result describing what happened
     */
    static addItem(container, item, world) {
        const trait = ContainerBehavior.require(container, trait_types_1.TraitType.CONTAINER);
        // Check if container is accessible (must be open if openable)
        if (container.has(trait_types_1.TraitType.OPENABLE)) {
            const openable = container.get(trait_types_1.TraitType.OPENABLE);
            if (openable && !openable.isOpen) {
                return {
                    success: false,
                    containerClosed: true,
                    stateChanged: false
                };
            }
        }
        // Check if item is already in this container
        const currentLocation = world.getLocation(item.id);
        if (currentLocation === container.id) {
            return {
                success: false,
                alreadyContains: true,
                stateChanged: false
            };
        }
        // Check type restrictions
        if (trait.allowedTypes && trait.allowedTypes.length > 0) {
            const itemType = item.type || 'object';
            if (!trait.allowedTypes.includes(itemType)) {
                return {
                    success: false,
                    wrongType: true,
                    stateChanged: false
                };
            }
        }
        if (trait.excludedTypes && trait.excludedTypes.length > 0) {
            const itemType = item.type || 'object';
            if (trait.excludedTypes.includes(itemType)) {
                return {
                    success: false,
                    wrongType: true,
                    stateChanged: false
                };
            }
        }
        // Check capacity constraints
        if (!ContainerBehavior.checkCapacity(container, item, world)) {
            return {
                success: false,
                containerFull: true,
                stateChanged: false
            };
        }
        // All checks passed - this would be where state mutation happens
        // In our event-driven system, the actual mutation happens via events
        // The behavior just validates and returns success
        return {
            success: true,
            stateChanged: true
        };
    }
    /**
     * Remove an item from a container
     * @param container The container to remove from
     * @param item The item to remove
     * @param world World query interface for getting current state
     * @returns Result describing what happened
     */
    static removeItem(container, item, world) {
        // Check if container is accessible (must be open if openable)
        if (container.has(trait_types_1.TraitType.OPENABLE)) {
            const openable = container.get(trait_types_1.TraitType.OPENABLE);
            if (openable && !openable.isOpen) {
                return {
                    success: false,
                    containerClosed: true,
                    stateChanged: false
                };
            }
        }
        // Check if item is actually in this container
        const currentLocation = world.getLocation(item.id);
        if (currentLocation !== container.id) {
            return {
                success: false,
                notContained: true,
                stateChanged: false
            };
        }
        // All checks passed - removal is valid
        return {
            success: true,
            stateChanged: true
        };
    }
}
exports.ContainerBehavior = ContainerBehavior;
