// @sharpee/world-model/traits/container/containerTrait.js
"use strict";
// packages/world-model/src/traits/container/containerTrait.ts
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContainerTrait = void 0;
const trait_types_1 = require("../trait-types");
/**
 * Container trait allows entities to hold other entities inside them.
 *
 * This is a pure data structure - all validation and logic
 * should be handled by ContainerBehavior.
 */
class ContainerTrait {
    static type = trait_types_1.TraitType.CONTAINER;
    type = trait_types_1.TraitType.CONTAINER;
    /** Capacity constraints for the container */
    capacity;
    /** Whether contents are visible when the container is closed */
    isTransparent;
    /** Whether actors can enter this container */
    enterable;
    /** Only these entity types can be placed in the container */
    allowedTypes;
    /** These entity types cannot be placed in the container */
    excludedTypes;
    /** Whether this container holds a liquid */
    containsLiquid;
    /** Type of liquid in the container (e.g., 'water', 'ale') */
    liquidType;
    /** Amount of liquid remaining (arbitrary units) */
    liquidAmount;
    constructor(data) {
        // Set defaults
        this.isTransparent = false;
        this.enterable = false;
        // Override with provided data
        if (data) {
            if (data.capacity !== undefined)
                this.capacity = data.capacity;
            if (data.isTransparent !== undefined)
                this.isTransparent = data.isTransparent;
            if (data.enterable !== undefined)
                this.enterable = data.enterable;
            if (data.allowedTypes !== undefined)
                this.allowedTypes = data.allowedTypes;
            if (data.excludedTypes !== undefined)
                this.excludedTypes = data.excludedTypes;
            if (data.containsLiquid !== undefined)
                this.containsLiquid = data.containsLiquid;
            if (data.liquidType !== undefined)
                this.liquidType = data.liquidType;
            if (data.liquidAmount !== undefined)
                this.liquidAmount = data.liquidAmount;
        }
    }
}
exports.ContainerTrait = ContainerTrait;
