// @sharpee/world-model/traits/attached/attachedTrait.js
"use strict";
/**
 * Attached trait implementation
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.AttachedTrait = void 0;
const trait_types_1 = require("../trait-types");
/**
 * Attached trait for objects that are attached, fastened, or connected to something
 *
 * This trait contains only data - all logic for detachment
 * is handled by the pulling action.
 */
class AttachedTrait {
    static type = trait_types_1.TraitType.ATTACHED;
    type = trait_types_1.TraitType.ATTACHED;
    // AttachedData properties
    attachedTo;
    attachmentType;
    detachable;
    detachForce;
    loose;
    detachSound;
    onDetach;
    constructor(data = {}) {
        this.attachedTo = data.attachedTo;
        this.attachmentType = data.attachmentType ?? 'stuck';
        this.detachable = data.detachable ?? false;
        this.detachForce = data.detachForce;
        this.loose = data.loose ?? false;
        this.detachSound = data.detachSound;
        this.onDetach = data.onDetach;
    }
}
exports.AttachedTrait = AttachedTrait;
