// @sharpee/stdlib/actions/standard/examining/examining.js
"use strict";
/**
 * Examining action - looks at objects in detail
 *
 * This is a read-only action that provides detailed information about objects.
 * It validates visibility but doesn't change state.
 *
 * Uses four-phase pattern:
 * 1. validate: Check target exists and is visible
 * 2. execute: No mutations (read-only action)
 * 3. report: Generate success events with complete entity snapshot
 * 4. blocked: Generate error events when validation fails
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.examiningAction = void 0;
const constants_1 = require("../../constants");
const types_1 = require("../../../scope/types");
const emit_illustrations_1 = require("../../helpers/emit-illustrations");
const data_builder_types_1 = require("../../data-builder-types");
// Import our data builder
const examining_data_1 = require("./examining-data");
const examining_messages_1 = require("./examining-messages");
exports.examiningAction = {
    id: constants_1.IFActions.EXAMINING,
    // Default scope requirements for this action's slots
    defaultScope: {
        target: types_1.ScopeLevel.VISIBLE
    },
    requiredMessages: [
        'no_target',
        'not_visible',
        'examined',
        'examined_self',
        'examined_container',
        'examined_supporter',
        'examined_readable',
        'examined_switchable',
        'examined_wearable',
        'examined_door',
        'nothing_special',
        'description',
        'brief_description'
    ],
    validate(context) {
        const actor = context.player;
        const noun = context.command.directObject?.entity;
        // Validate we have a target
        if (!noun) {
            return {
                valid: false,
                error: examining_messages_1.ExaminingMessages.NO_TARGET
            };
        }
        // Check scope - must be able to see the target (unless examining yourself)
        if (noun.id !== actor.id) {
            const scopeCheck = context.requireScope(noun, types_1.ScopeLevel.VISIBLE);
            if (!scopeCheck.ok) {
                return scopeCheck.error;
            }
        }
        // Valid - all event data will be built in report()
        return { valid: true };
    },
    execute(context) {
        // No mutations - examining is a read-only action
    },
    report(context) {
        // report() is only called on success - validation passed
        const noun = context.command.directObject.entity;
        // Use data builder to create event data
        const eventData = (0, data_builder_types_1.buildEventData)(examining_data_1.examiningDataConfig, context);
        // Get message parameters from the data
        const { messageId, params, contentsMessage } = (0, examining_data_1.buildExaminingMessageParams)(eventData, noun);
        // Build events array - emit domain event with messageId for text rendering
        const examinedEvent = context.event('if.event.examined', {
            messageId: `${context.action.id}.${messageId}`,
            params,
            ...eventData
        });
        const events = [examinedEvent];
        // Emit illustration events for the examined entity (ADR-124)
        events.push(...(0, emit_illustrations_1.emitIllustrations)(noun, 'on-examine', examinedEvent.id, context));
        // Add contents message for containers/supporters with visible items
        if (contentsMessage) {
            events.push(context.event('if.event.examined', {
                messageId: `${context.action.id}.${contentsMessage.messageId}`,
                params: contentsMessage.params,
                isContentsMessage: true
            }));
        }
        return events;
    },
    blocked(context, result) {
        // blocked() is called when validation fails
        const noun = context.command.directObject?.entity;
        return [context.event('if.event.examined', {
                blocked: true,
                messageId: `${context.action.id}.${result.error}`,
                params: { target: noun?.name, ...result.params },
                reason: result.error,
                targetId: noun?.id,
                targetName: noun?.name
            })];
    },
    group: "observation",
    metadata: {
        requiresDirectObject: true,
        requiresIndirectObject: false,
        directObjectScope: types_1.ScopeLevel.VISIBLE
    }
};
