// @sharpee/stdlib/scope/types.js
"use strict";
/**
 * Core types for the scope system
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.WitnessLevel = exports.SenseType = exports.ScopeLevelStrings = exports.ScopeLevel = void 0;
/**
 * Levels of scope indicating how an entity can be perceived.
 *
 * Ordered from least accessible (UNAWARE=0) to most accessible (CARRIED=4).
 * Higher values imply all lower levels are also satisfied.
 * Use numeric comparisons: if (entityScope >= requiredScope) { ... }
 *
 * For actions that check scope, use ActionContext.requireScope() which
 * returns appropriate error messages for scope failures.
 */
var ScopeLevel;
(function (ScopeLevel) {
    /** Entity not known to player at all */
    ScopeLevel[ScopeLevel["UNAWARE"] = 0] = "UNAWARE";
    /** Player knows entity exists (can think about, ask about, remember) */
    ScopeLevel[ScopeLevel["AWARE"] = 1] = "AWARE";
    /** Can see with eyes (examine, look at, read) */
    ScopeLevel[ScopeLevel["VISIBLE"] = 2] = "VISIBLE";
    /** Can physically touch/manipulate (take, push, open, touch) */
    ScopeLevel[ScopeLevel["REACHABLE"] = 3] = "REACHABLE";
    /** In actor's inventory - always accessible (drop, eat, wear, insert) */
    ScopeLevel[ScopeLevel["CARRIED"] = 4] = "CARRIED";
})(ScopeLevel || (exports.ScopeLevel = ScopeLevel = {}));
/**
 * @deprecated Use ScopeLevel instead. These string values are for backwards compatibility.
 */
exports.ScopeLevelStrings = {
    CARRIED: 'carried',
    REACHABLE: 'reachable',
    VISIBLE: 'visible',
    AUDIBLE: 'audible',
    DETECTABLE: 'detectable',
    OUT_OF_SCOPE: 'out_of_scope'
};
/**
 * Types of sensory perception
 */
var SenseType;
(function (SenseType) {
    /** Visual perception - blocked by opaque barriers, needs light */
    SenseType["SIGHT"] = "sight";
    /** Auditory perception - travels through some barriers */
    SenseType["HEARING"] = "hearing";
    /** Olfactory perception - needs air path */
    SenseType["SMELL"] = "smell";
    /** Tactile perception - requires physical contact */
    SenseType["TOUCH"] = "touch";
    /** Supernatural/psychic perception - game-specific */
    SenseType["VIBE"] = "vibe";
})(SenseType || (exports.SenseType = SenseType = {}));
/**
 * Level of detail in witnessing
 */
var WitnessLevel;
(function (WitnessLevel) {
    /** Saw/heard/sensed everything clearly */
    WitnessLevel["FULL"] = "full";
    /** Saw/heard/sensed some but not all details */
    WitnessLevel["PARTIAL"] = "partial";
    /** Caught a glimpse or hint */
    WitnessLevel["PERIPHERAL"] = "peripheral";
    /** Deduced from evidence rather than direct perception */
    WitnessLevel["INFERRED"] = "inferred";
})(WitnessLevel || (exports.WitnessLevel = WitnessLevel = {}));
