// @sharpee/stdlib/npc/index.js
"use strict";
/**
 * NPC Module Exports (ADR-070)
 *
 * Provides NPC behavior system for autonomous characters.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createPatrolBehavior = exports.createFollowerBehavior = exports.createWandererBehavior = exports.passiveBehavior = exports.guardBehavior = exports.DECAY_RATE_TURNS = exports.enterLucidityWindow = exports.processLucidityDecay = exports.DefaultStateTransitions = exports.injectHallucinations = exports.filterPerception = exports.observeEvent = exports.clearNpcCombatResolver = exports.registerNpcCombatResolver = exports.createNpcService = exports.NpcService = void 0;
// Types
__exportStar(require("./types"), exports);
// Message IDs
__exportStar(require("./npc-messages"), exports);
__exportStar(require("./character-messages"), exports);
// Service
var npc_service_1 = require("./npc-service");
Object.defineProperty(exports, "NpcService", { enumerable: true, get: function () { return npc_service_1.NpcService; } });
Object.defineProperty(exports, "createNpcService", { enumerable: true, get: function () { return npc_service_1.createNpcService; } });
Object.defineProperty(exports, "registerNpcCombatResolver", { enumerable: true, get: function () { return npc_service_1.registerNpcCombatResolver; } });
Object.defineProperty(exports, "clearNpcCombatResolver", { enumerable: true, get: function () { return npc_service_1.clearNpcCombatResolver; } });
// Character model observation and lucidity (ADR-141)
var character_observer_1 = require("./character-observer");
Object.defineProperty(exports, "observeEvent", { enumerable: true, get: function () { return character_observer_1.observeEvent; } });
Object.defineProperty(exports, "filterPerception", { enumerable: true, get: function () { return character_observer_1.filterPerception; } });
Object.defineProperty(exports, "injectHallucinations", { enumerable: true, get: function () { return character_observer_1.injectHallucinations; } });
Object.defineProperty(exports, "DefaultStateTransitions", { enumerable: true, get: function () { return character_observer_1.DefaultStateTransitions; } });
var lucidity_decay_1 = require("./lucidity-decay");
Object.defineProperty(exports, "processLucidityDecay", { enumerable: true, get: function () { return lucidity_decay_1.processLucidityDecay; } });
Object.defineProperty(exports, "enterLucidityWindow", { enumerable: true, get: function () { return lucidity_decay_1.enterLucidityWindow; } });
Object.defineProperty(exports, "DECAY_RATE_TURNS", { enumerable: true, get: function () { return lucidity_decay_1.DECAY_RATE_TURNS; } });
// Standard behaviors
var behaviors_1 = require("./behaviors");
Object.defineProperty(exports, "guardBehavior", { enumerable: true, get: function () { return behaviors_1.guardBehavior; } });
Object.defineProperty(exports, "passiveBehavior", { enumerable: true, get: function () { return behaviors_1.passiveBehavior; } });
Object.defineProperty(exports, "createWandererBehavior", { enumerable: true, get: function () { return behaviors_1.createWandererBehavior; } });
Object.defineProperty(exports, "createFollowerBehavior", { enumerable: true, get: function () { return behaviors_1.createFollowerBehavior; } });
Object.defineProperty(exports, "createPatrolBehavior", { enumerable: true, get: function () { return behaviors_1.createPatrolBehavior; } });
