// @sharpee/stdlib/npc/lucidity-decay.js
"use strict";
/**
 * Lucidity decay processing (ADR-141)
 *
 * End-of-turn processing for NPC lucidity windows.
 * When an NPC is in a lucid window with no sustaining trigger active,
 * lucidity decays and eventually returns to baseline.
 *
 * Public interface: processLucidityDecay(), DECAY_RATE_TURNS.
 * Owner context: stdlib / npc
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DECAY_RATE_TURNS = void 0;
exports.processLucidityDecay = processLucidityDecay;
exports.enterLucidityWindow = enterLucidityWindow;
const world_model_1 = require("../../world-model/index.js");
const character_messages_1 = require("./character-messages");
// ---------------------------------------------------------------------------
// Decay rate mapping
// ---------------------------------------------------------------------------
/**
 * Maps decay rate words to number of turns before baseline is restored.
 * These are the window durations when no sustaining trigger is active.
 */
exports.DECAY_RATE_TURNS = {
    slow: 8,
    moderate: 4,
    fast: 2,
};
// ---------------------------------------------------------------------------
// Event helper
// ---------------------------------------------------------------------------
let decayEventCounter = 0;
function createDecayEvent(type, npcId, data) {
    return {
        id: `decay_${++decayEventCounter}_${Date.now()}`,
        type,
        timestamp: Date.now(),
        entities: { actor: npcId },
        data,
    };
}
// ---------------------------------------------------------------------------
// Main decay function
// ---------------------------------------------------------------------------
/**
 * Process end-of-turn lucidity decay for a single NPC.
 *
 * If the NPC has a CharacterModelTrait with an active lucidity window,
 * decrements the window counter. When it reaches zero, the cognitive
 * profile returns to baseline and a LUCIDITY_BASELINE_RESTORED event
 * is emitted.
 *
 * If no lucidity config or no active window, returns empty array.
 *
 * @param npc - The NPC entity
 * @param world - The world model (unused in current impl, reserved for future)
 * @param turn - Current turn number (unused in current impl, reserved for future)
 * @returns Array of events emitted (baseline restored, or empty)
 */
function processLucidityDecay(npc, world, turn) {
    const trait = npc.get(world_model_1.TraitType.CHARACTER_MODEL);
    if (!trait)
        return [];
    if (!trait.lucidityConfig)
        return [];
    // No active window
    if (trait.lucidityWindowTurns <= 0)
        return [];
    const previousState = trait.currentLucidityState;
    const expired = trait.decayLucidity();
    if (expired) {
        return [
            createDecayEvent(character_messages_1.CharacterMessages.LUCIDITY_BASELINE_RESTORED, npc.id, { from: previousState, to: trait.currentLucidityState }),
        ];
    }
    return [];
}
/**
 * Initialize a lucidity window with the appropriate turn count
 * based on the NPC's configured decay rate.
 *
 * Call this when entering a lucidity state via a trigger, so the
 * window has the correct duration based on decayRate.
 *
 * @param trait - The NPC's CharacterModelTrait
 * @param targetState - The lucidity state to enter
 */
function enterLucidityWindow(trait, targetState) {
    if (!trait.lucidityConfig) {
        trait.enterLucidityState(targetState);
        return;
    }
    const turns = exports.DECAY_RATE_TURNS[trait.lucidityConfig.decayRate];
    trait.enterLucidityState(targetState, turns);
}
