// @sharpee/stdlib/npc/npc-service.js
"use strict";
/**
 * NPC Service (ADR-070)
 *
 * Manages NPC behaviors, executes NPC actions, and handles the NPC turn phase.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.NpcService = void 0;
exports.registerNpcCombatResolver = registerNpcCombatResolver;
exports.clearNpcCombatResolver = clearNpcCombatResolver;
exports.createNpcService = createNpcService;
const world_model_1 = require("../../world-model/index.js");
const npc_messages_1 = require("./npc-messages");
const lucidity_decay_1 = require("./lucidity-decay");
/**
 * Module-level NPC combat resolver. Set via registerNpcCombatResolver().
 * Uses globalThis to share across module boundaries (same pattern as interceptor registry).
 */
const NPC_COMBAT_RESOLVER_KEY = '__sharpee_npc_combat_resolver__';
function getNpcCombatResolver() {
    return globalThis[NPC_COMBAT_RESOLVER_KEY];
}
/**
 * Register an NPC combat resolver.
 *
 * Call this in your story's initializeWorld() to provide combat resolution
 * for NPC attack actions. Without a resolver, NPC attacks produce bare events.
 *
 * @param resolver - Function that resolves NPC→target combat and returns events
 */
function registerNpcCombatResolver(resolver) {
    globalThis[NPC_COMBAT_RESOLVER_KEY] = resolver;
}
/**
 * Clear the NPC combat resolver. Used for testing cleanup.
 */
function clearNpcCombatResolver() {
    delete globalThis[NPC_COMBAT_RESOLVER_KEY];
}
/**
 * Create an event ID
 */
function createEventId(prefix) {
    return `${prefix}_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
}
/**
 * Create a semantic event
 */
function createEvent(type, data, npcId) {
    return {
        id: createEventId('npc'),
        type,
        timestamp: Date.now(),
        entities: npcId ? { actor: npcId } : {},
        data,
    };
}
/**
 * NPC Service implementation
 */
class NpcService {
    behaviors = new Map();
    tickPhases = [];
    registerBehavior(behavior) {
        this.behaviors.set(behavior.id, behavior);
    }
    removeBehavior(id) {
        this.behaviors.delete(id);
    }
    getBehavior(id) {
        return this.behaviors.get(id);
    }
    /**
     * Register a tick phase handler (ADR-142/144/145/146).
     * Phases run in registration order after behavior onTurn processing.
     *
     * @param name - Phase name for debugging
     * @param handler - Function called with active NPCs and tick context
     */
    registerTickPhase(name, handler) {
        this.tickPhases.push({ name, handler });
    }
    /**
     * Execute the NPC turn phase
     */
    tick(context) {
        const events = [];
        const { world, turn, random, playerLocation, playerId } = context;
        // Find all NPCs that can act
        const npcs = this.getActiveNpcs(world);
        // Process each NPC
        for (const npc of npcs) {
            const npcTrait = npc.get(world_model_1.TraitType.NPC);
            // Direct property access — canAct getter doesn't survive loadJSON()
            if (!npcTrait || !npcTrait.isAlive || !npcTrait.isConscious)
                continue;
            const behavior = this.getBehaviorForNpc(npc);
            if (!behavior)
                continue;
            const npcLocation = world.getLocation(npc.id) || '';
            const npcContext = this.createNpcContext(npc, world, random, turn, playerLocation, npcLocation);
            // Call onTurn hook
            const actions = behavior.onTurn(npcContext);
            const actionEvents = this.executeActions(npc, actions, world, random);
            events.push(...actionEvents);
            // Process lucidity decay for NPCs with character model (ADR-141)
            if (npc.has(world_model_1.TraitType.CHARACTER_MODEL)) {
                const decayEvents = (0, lucidity_decay_1.processLucidityDecay)(npc, world, turn);
                events.push(...decayEvents);
            }
        }
        // Registered tick phases (ADR-142/144/145/146)
        for (const phase of this.tickPhases) {
            const phaseEvents = phase.handler(npcs, context);
            events.push(...phaseEvents);
        }
        return events;
    }
    /**
     * Notify NPCs that player entered a room
     */
    onPlayerEnters(world, roomId, random, turn) {
        const events = [];
        const playerId = world.getPlayer()?.id || '';
        // Get NPCs in the room
        const entities = world.getContents(roomId);
        const npcs = entities.filter((e) => e.has(world_model_1.TraitType.NPC));
        for (const npc of npcs) {
            const npcTrait = npc.get(world_model_1.TraitType.NPC);
            if (!npcTrait || !npcTrait.isAlive || !npcTrait.isConscious)
                continue;
            const behavior = this.getBehaviorForNpc(npc);
            if (!behavior?.onPlayerEnters)
                continue;
            const npcContext = this.createNpcContext(npc, world, random, turn, roomId, roomId);
            const actions = behavior.onPlayerEnters(npcContext);
            const actionEvents = this.executeActions(npc, actions, world, random);
            events.push(...actionEvents);
        }
        return events;
    }
    /**
     * Notify NPCs that player left a room
     */
    onPlayerLeaves(world, roomId, random, turn) {
        const events = [];
        const playerId = world.getPlayer()?.id || '';
        const playerLocation = world.getLocation(playerId) || '';
        // Get NPCs in the old room
        const entities = world.getContents(roomId);
        const npcs = entities.filter((e) => e.has(world_model_1.TraitType.NPC));
        for (const npc of npcs) {
            const npcTrait = npc.get(world_model_1.TraitType.NPC);
            if (!npcTrait || !npcTrait.isAlive || !npcTrait.isConscious)
                continue;
            const behavior = this.getBehaviorForNpc(npc);
            if (!behavior?.onPlayerLeaves)
                continue;
            const npcContext = this.createNpcContext(npc, world, random, turn, playerLocation, roomId);
            const actions = behavior.onPlayerLeaves(npcContext);
            const actionEvents = this.executeActions(npc, actions, world, random);
            events.push(...actionEvents);
        }
        return events;
    }
    /**
     * Handle player speaking to an NPC
     */
    onPlayerSpeaks(world, npcId, words, random, turn) {
        const npc = world.getEntity(npcId);
        if (!npc)
            return [];
        const npcTrait = npc.get(world_model_1.TraitType.NPC);
        if (!npcTrait || !npcTrait.isAlive || !npcTrait.isConscious)
            return [];
        const behavior = this.getBehaviorForNpc(npc);
        if (!behavior?.onSpokenTo) {
            // Default response if no handler
            return [
                createEvent('npc.spoke', {
                    npc: npcId,
                    messageId: npc_messages_1.NpcMessages.NPC_NO_RESPONSE,
                    data: { npcName: npc.name },
                }, npcId),
            ];
        }
        const playerId = world.getPlayer()?.id || '';
        const playerLocation = world.getLocation(playerId) || '';
        const npcLocation = world.getLocation(npcId) || '';
        const npcContext = this.createNpcContext(npc, world, random, turn, playerLocation, npcLocation);
        const actions = behavior.onSpokenTo(npcContext, words);
        return this.executeActions(npc, actions, world, random);
    }
    /**
     * Handle player attacking an NPC
     */
    onNpcAttacked(world, npcId, attackerId, random, turn) {
        const npc = world.getEntity(npcId);
        if (!npc)
            return [];
        const npcTrait = npc.get(world_model_1.TraitType.NPC);
        if (!npcTrait || !npcTrait.isAlive || !npcTrait.isConscious)
            return [];
        const behavior = this.getBehaviorForNpc(npc);
        if (!behavior?.onAttacked)
            return [];
        const attacker = world.getEntity(attackerId);
        if (!attacker)
            return [];
        const playerLocation = world.getLocation(attackerId) || '';
        const npcLocation = world.getLocation(npcId) || '';
        const npcContext = this.createNpcContext(npc, world, random, turn, playerLocation, npcLocation);
        const actions = behavior.onAttacked(npcContext, attacker);
        return this.executeActions(npc, actions, world, random);
    }
    // ==================== Private Helpers ====================
    getActiveNpcs(world) {
        const allEntities = world.getAllEntities();
        return allEntities.filter((e) => {
            if (!e.has(world_model_1.TraitType.NPC))
                return false;
            const npcTrait = e.get(world_model_1.TraitType.NPC);
            // Use direct property access — canAct getter doesn't survive loadJSON() deserialization
            return npcTrait.isAlive && npcTrait.isConscious;
        });
    }
    getBehaviorForNpc(npc) {
        const npcTrait = npc.get(world_model_1.TraitType.NPC);
        if (!npcTrait.behaviorId)
            return undefined;
        return this.behaviors.get(npcTrait.behaviorId);
    }
    createNpcContext(npc, world, random, turn, playerLocation, npcLocation) {
        const npcInventory = world.getContents(npc.id);
        return {
            npc,
            world,
            random,
            turnCount: turn,
            playerLocation,
            npcLocation,
            npcInventory,
            playerVisible: npcLocation === playerLocation,
            getEntitiesInRoom: () => world.getContents(npcLocation),
            getAvailableExits: () => this.getExitsFromRoom(world, npcLocation, npc),
        };
    }
    getExitsFromRoom(world, roomId, npc) {
        const room = world.getEntity(roomId);
        if (!room)
            return [];
        const roomTrait = room.get(world_model_1.RoomTrait);
        if (!roomTrait?.exits)
            return [];
        const npcTrait = npc.get(world_model_1.TraitType.NPC);
        const exits = [];
        for (const [direction, exit] of Object.entries(roomTrait.exits)) {
            const exitData = exit;
            if (exitData.destination) {
                // Inline canEnterRoom() — method doesn't survive loadJSON() deserialization
                const dest = exitData.destination;
                if (!npcTrait.canMove)
                    continue;
                if (npcTrait.forbiddenRooms?.includes(dest))
                    continue;
                if (npcTrait.allowedRooms && !npcTrait.allowedRooms.includes(dest))
                    continue;
                exits.push({
                    direction: direction,
                    destination: dest,
                });
            }
        }
        return exits;
    }
    executeActions(npc, actions, world, random) {
        const events = [];
        for (const action of actions) {
            const actionEvents = this.executeAction(npc, action, world, random);
            events.push(...actionEvents);
        }
        return events;
    }
    executeAction(npc, action, world, random) {
        switch (action.type) {
            case 'move':
                return this.executeMove(npc, action.direction, world);
            case 'moveTo':
                return this.executeMoveTo(npc, action.roomId, world);
            case 'take':
                return this.executeTake(npc, action.target, world);
            case 'drop':
                return this.executeDrop(npc, action.target, world);
            case 'attack':
                return this.executeAttack(npc, action.target, world, random);
            case 'speak':
                return [
                    createEvent('npc.spoke', {
                        npc: npc.id,
                        messageId: action.messageId,
                        data: action.data,
                        params: action.data,
                    }, npc.id),
                ];
            case 'emote':
                return [
                    createEvent('npc.emoted', {
                        npc: npc.id,
                        messageId: action.messageId,
                        data: action.data,
                        params: action.data,
                    }, npc.id),
                ];
            case 'wait':
                return []; // No events for waiting
            case 'custom':
                return action.handler();
            default:
                return [];
        }
    }
    executeMove(npc, direction, world) {
        const currentLocation = world.getLocation(npc.id);
        if (!currentLocation)
            return [];
        const currentRoom = world.getEntity(currentLocation);
        if (!currentRoom)
            return [];
        const roomTrait = currentRoom.get(world_model_1.RoomTrait);
        if (!roomTrait?.exits?.[direction])
            return [];
        const destination = roomTrait.exits[direction].destination;
        if (!destination)
            return [];
        // Move the NPC
        world.moveEntity(npc.id, destination);
        return [
            createEvent('npc.moved', {
                npc: npc.id,
                from: currentLocation,
                to: destination,
                direction,
            }, npc.id),
        ];
    }
    executeMoveTo(npc, roomId, world) {
        const currentLocation = world.getLocation(npc.id);
        if (!currentLocation)
            return [];
        // Move the NPC
        world.moveEntity(npc.id, roomId);
        return [
            createEvent('npc.moved', {
                npc: npc.id,
                from: currentLocation,
                to: roomId,
            }, npc.id),
        ];
    }
    executeTake(npc, targetId, world) {
        const target = world.getEntity(targetId);
        if (!target)
            return [];
        // Move item to NPC's inventory
        world.moveEntity(targetId, npc.id);
        return [
            createEvent('npc.took', {
                npc: npc.id,
                target: targetId,
            }, npc.id),
        ];
    }
    executeDrop(npc, targetId, world) {
        const target = world.getEntity(targetId);
        if (!target)
            return [];
        const npcLocation = world.getLocation(npc.id);
        if (!npcLocation)
            return [];
        // Move item to NPC's location
        world.moveEntity(targetId, npcLocation);
        return [
            createEvent('npc.dropped', {
                npc: npc.id,
                target: targetId,
            }, npc.id),
        ];
    }
    executeAttack(npc, targetId, world, random) {
        const target = world.getEntity(targetId);
        if (!target)
            return [];
        // Delegate to registered combat resolver if available
        const resolver = getNpcCombatResolver();
        if (resolver) {
            return resolver(npc, target, world, random);
        }
        // No resolver registered — emit bare attack event (no combat resolution)
        return [
            createEvent('npc.attacked', {
                npc: npc.id,
                target: targetId,
            }, npc.id),
        ];
    }
}
exports.NpcService = NpcService;
/**
 * Create a new NPC Service instance
 */
function createNpcService() {
    return new NpcService();
}
