// @sharpee/world-model/traits/supporter/supporterBehavior.js
"use strict";
// packages/world-model/src/traits/supporter/supporterBehavior.ts
Object.defineProperty(exports, "__esModule", { value: true });
exports.SupporterBehavior = void 0;
const behavior_1 = require("../../behaviors/behavior");
const trait_types_1 = require("../trait-types");
const identityBehavior_1 = require("../identity/identityBehavior");
/**
 * Behavior for entities that can support other entities on their surface.
 *
 * Handles capacity checks, type restrictions, and weight calculations.
 */
class SupporterBehavior extends behavior_1.Behavior {
    static requiredTraits = [trait_types_1.TraitType.SUPPORTER];
    /**
     * Check if a supporter can accept an item
     * @returns true if the item can be accepted
     */
    static canAccept(supporter, item, world) {
        const trait = SupporterBehavior.require(supporter, trait_types_1.TraitType.SUPPORTER);
        // Check type restrictions
        if (trait.allowedTypes && trait.allowedTypes.length > 0) {
            const itemType = item.type || 'object';
            if (!trait.allowedTypes.includes(itemType)) {
                return false;
            }
        }
        if (trait.excludedTypes && trait.excludedTypes.length > 0) {
            const itemType = item.type || 'object';
            if (trait.excludedTypes.includes(itemType)) {
                return false;
            }
        }
        // Check enterable restrictions for actors
        if (item.has(trait_types_1.TraitType.ACTOR) && !trait.enterable) {
            return false;
        }
        // Check capacity constraints
        if (trait.capacity) {
            const result = this.checkCapacity(supporter, item, world);
            if (!result)
                return false;
        }
        return true;
    }
    /**
     * Check capacity constraints
     */
    static checkCapacity(supporter, item, world) {
        const trait = SupporterBehavior.require(supporter, trait_types_1.TraitType.SUPPORTER);
        const capacity = trait.capacity;
        // Check item count limit
        if (capacity.maxItems !== undefined) {
            const currentCount = world.getContents(supporter.id).length;
            if (currentCount >= capacity.maxItems) {
                return false;
            }
        }
        // Check weight limit
        if (capacity.maxWeight !== undefined) {
            const currentWeight = this.getTotalWeight(supporter, world);
            const itemWeight = this.getItemTotalWeight(item, world);
            if (currentWeight + itemWeight > capacity.maxWeight) {
                return false;
            }
        }
        return true;
    }
    /**
     * Get total weight on supporter
     * For supporters, we count the total weight of all items including their contents
     */
    static getTotalWeight(supporter, world) {
        const contents = world.getContents(supporter.id);
        return contents.reduce((sum, item) => {
            // For supporters, we count total weight including nested contents
            return sum + this.getItemTotalWeight(item, world);
        }, 0);
    }
    /**
     * Get total weight of an item including its contents
     */
    static getItemTotalWeight(item, world) {
        let weight = identityBehavior_1.IdentityBehavior.getWeight(item);
        // Add weight of contents if this is a container or supporter
        if (item.has(trait_types_1.TraitType.CONTAINER) || item.has(trait_types_1.TraitType.SUPPORTER)) {
            const contents = world.getContents(item.id);
            for (const subItem of contents) {
                weight += this.getItemTotalWeight(subItem, world);
            }
        }
        return weight;
    }
    /**
     * Get remaining capacity
     */
    static getRemainingCapacity(supporter, world) {
        const trait = SupporterBehavior.require(supporter, trait_types_1.TraitType.SUPPORTER);
        const result = {};
        if (trait.capacity?.maxItems !== undefined) {
            const current = world.getContents(supporter.id).length;
            result.items = trait.capacity.maxItems - current;
        }
        if (trait.capacity?.maxWeight !== undefined) {
            result.weight = trait.capacity.maxWeight - this.getTotalWeight(supporter, world);
        }
        return result;
    }
    /**
     * Check if the supporter is enterable by actors
     */
    static isEnterable(supporter) {
        const trait = SupporterBehavior.require(supporter, trait_types_1.TraitType.SUPPORTER);
        return trait.enterable || false;
    }
    /**
     * Get allowed types for this supporter
     */
    static getAllowedTypes(supporter) {
        const trait = SupporterBehavior.require(supporter, trait_types_1.TraitType.SUPPORTER);
        return trait.allowedTypes;
    }
    /**
     * Get excluded types for this supporter
     */
    static getExcludedTypes(supporter) {
        const trait = SupporterBehavior.require(supporter, trait_types_1.TraitType.SUPPORTER);
        return trait.excludedTypes;
    }
    /**
     * Add an item to a supporter
     * @param supporter The supporter receiving the item
     * @param item The item being added
     * @param world World query interface
     * @returns Result describing what happened
     */
    static addItem(supporter, item, world) {
        // Check if item is already on supporter
        const currentLocation = world.getLocation(item.id);
        if (currentLocation === supporter.id) {
            return {
                success: false,
                alreadyThere: true,
                stateChanged: false
            };
        }
        // Check if supporter can accept the item
        if (!SupporterBehavior.canAccept(supporter, item, world)) {
            // Determine specific failure reason
            const trait = SupporterBehavior.require(supporter, trait_types_1.TraitType.SUPPORTER);
            // Check type restrictions
            if (trait.allowedTypes && trait.allowedTypes.length > 0) {
                const itemType = item.type || 'object';
                if (!trait.allowedTypes.includes(itemType)) {
                    return {
                        success: false,
                        wrongType: true,
                        stateChanged: false
                    };
                }
            }
            if (trait.excludedTypes && trait.excludedTypes.length > 0) {
                const itemType = item.type || 'object';
                if (trait.excludedTypes.includes(itemType)) {
                    return {
                        success: false,
                        wrongType: true,
                        stateChanged: false
                    };
                }
            }
            // Must be a capacity issue
            return {
                success: false,
                noSpace: true,
                stateChanged: false
            };
        }
        // All checks passed - adding is valid
        return {
            success: true,
            stateChanged: true
        };
    }
    /**
     * Remove an item from a supporter
     * @param supporter The supporter losing the item
     * @param item The item being removed
     * @param world World query interface
     * @returns Result describing what happened
     */
    static removeItem(supporter, item, world) {
        // Check if item is actually on the supporter
        const currentLocation = world.getLocation(item.id);
        if (currentLocation !== supporter.id) {
            return {
                success: false,
                notThere: true,
                stateChanged: false
            };
        }
        // All checks passed - removal is valid
        return {
            success: true,
            stateChanged: true
        };
    }
}
exports.SupporterBehavior = SupporterBehavior;
