// @sharpee/world-model/traits/npc/npcTrait.js
"use strict";
/**
 * NPC Trait (ADR-070)
 *
 * Marks an entity as a Non-Player Character that can act autonomously.
 * NPCs participate in the turn cycle and can have behaviors.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.NpcTrait = void 0;
/**
 * NPC trait indicates an entity is an autonomous character.
 *
 * NPCs can:
 * - Act during the NPC turn phase
 * - Move between rooms (if canMove is true)
 * - Respond to player actions (speak, attack, observe)
 * - Maintain state across turns
 *
 * The actual behavior logic is defined in NpcBehavior implementations
 * registered with the NpcService.
 */
class NpcTrait {
    static type = 'npc';
    type = 'npc';
    // State properties
    isAlive;
    isConscious;
    isHostile;
    // Movement properties
    canMove;
    allowedRooms;
    forbiddenRooms;
    // Behavior reference
    behaviorId;
    // Interaction state
    conversationState;
    knowledge;
    goals;
    // Custom properties
    customProperties;
    constructor(data = {}) {
        // Set defaults and merge with provided data
        this.isAlive = data.isAlive ?? true;
        this.isConscious = data.isConscious ?? true;
        this.isHostile = data.isHostile ?? false;
        this.canMove = data.canMove ?? false;
        this.allowedRooms = data.allowedRooms;
        this.forbiddenRooms = data.forbiddenRooms;
        this.behaviorId = data.behaviorId;
        this.conversationState = data.conversationState;
        this.knowledge = data.knowledge;
        this.goals = data.goals;
        this.customProperties = data.customProperties;
    }
    /**
     * Check if NPC can act (alive and conscious)
     */
    get canAct() {
        return this.isAlive && this.isConscious;
    }
    /**
     * Check if NPC can enter a specific room
     */
    canEnterRoom(roomId) {
        if (!this.canMove)
            return false;
        // Check forbidden rooms first
        if (this.forbiddenRooms?.includes(roomId)) {
            return false;
        }
        // If allowed rooms is defined, room must be in list
        if (this.allowedRooms && !this.allowedRooms.includes(roomId)) {
            return false;
        }
        return true;
    }
    /**
     * Make this NPC hostile
     */
    makeHostile() {
        this.isHostile = true;
    }
    /**
     * Make this NPC non-hostile
     */
    makePassive() {
        this.isHostile = false;
    }
    /**
     * Knock out this NPC (unconscious but alive)
     */
    knockOut() {
        this.isConscious = false;
    }
    /**
     * Wake up this NPC
     */
    wakeUp() {
        if (this.isAlive) {
            this.isConscious = true;
        }
    }
    /**
     * Kill this NPC
     */
    kill() {
        this.isAlive = false;
        this.isConscious = false;
    }
    /**
     * Revive this NPC
     */
    revive() {
        this.isAlive = true;
        this.isConscious = true;
    }
    /**
     * Set a knowledge item
     */
    setKnowledge(key, value) {
        if (!this.knowledge) {
            this.knowledge = {};
        }
        this.knowledge[key] = value;
    }
    /**
     * Get a knowledge item
     */
    getKnowledge(key) {
        return this.knowledge?.[key];
    }
    /**
     * Check if NPC knows something
     */
    knows(key) {
        return this.knowledge?.[key] !== undefined;
    }
    /**
     * Add a goal
     */
    addGoal(goal) {
        if (!this.goals) {
            this.goals = [];
        }
        if (!this.goals.includes(goal)) {
            this.goals.push(goal);
        }
    }
    /**
     * Remove a goal
     */
    removeGoal(goal) {
        if (this.goals) {
            this.goals = this.goals.filter((g) => g !== goal);
        }
    }
    /**
     * Check if NPC has a goal
     */
    hasGoal(goal) {
        return this.goals?.includes(goal) ?? false;
    }
    /**
     * Set a custom property
     */
    setCustomProperty(key, value) {
        if (!this.customProperties) {
            this.customProperties = {};
        }
        this.customProperties[key] = value;
    }
    /**
     * Get a custom property
     */
    getCustomProperty(key) {
        return this.customProperties?.[key];
    }
}
exports.NpcTrait = NpcTrait;
