// @sharpee/core/rules/compatibility.js
"use strict";
/**
 * Compatibility adapter for the old RuleSystem interface
 * This allows existing code to work with the new simple rule system
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.RuleSystemAdapter = void 0;
exports.createRuleSystem = createRuleSystem;
const simple_rule_system_1 = require("./simple-rule-system");
/**
 * Adapter that makes the new SimpleRuleSystem compatible with old interfaces
 */
class RuleSystemAdapter {
    simpleRuleSystem;
    constructor(simpleRuleSystem) {
        this.simpleRuleSystem = simpleRuleSystem;
    }
    processEvent(event, context) {
        // Create a RuleWorld adapter from the GameContext
        const world = createRuleWorldFromGameContext(context);
        // Process through the simple rule system
        const result = this.simpleRuleSystem.processEvent(event, world);
        // Convert back to old format
        return {
            prevented: result.prevent || false,
            preventMessage: result.message,
            events: result.events || [],
            context: context // Return original context for now
        };
    }
}
exports.RuleSystemAdapter = RuleSystemAdapter;
/**
 * Create a RuleWorld from a GameContext
 */
function createRuleWorldFromGameContext(context) {
    return {
        getEntity: (id) => context.getEntity(id),
        updateEntity: (id, changes) => {
            // For now, just update the entity directly
            const entity = context.getEntity(id);
            if (entity) {
                if (entity.attributes) {
                    Object.assign(entity.attributes, changes);
                }
                else {
                    Object.assign(entity, changes);
                }
            }
        },
        getPlayer: () => context.player,
        getCurrentLocation: () => context.currentLocation
    };
}
/**
 * Create a compatible rule system
 */
function createRuleSystem() {
    const simpleRuleSystem = new simple_rule_system_1.SimpleRuleSystemImpl();
    return new RuleSystemAdapter(simpleRuleSystem);
}
