// @sharpee/core/rules/simple-rule-system.js
"use strict";
/**
 * Simple Rule System Implementation
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SimpleRuleSystemImpl = void 0;
exports.createSimpleRuleSystem = createSimpleRuleSystem;
const event_system_1 = require("../events/event-system");
const standard_events_1 = require("../events/standard-events");
/**
 * Simple rule system implementation
 */
class SimpleRuleSystemImpl {
    rules = new Map();
    /**
     * Add a rule to the system
     */
    addRule(rule) {
        this.rules.set(rule.id, rule);
    }
    /**
     * Remove a rule from the system
     */
    removeRule(ruleId) {
        this.rules.delete(ruleId);
    }
    /**
     * Get all rules
     */
    getRules() {
        return Array.from(this.rules.values());
    }
    /**
     * Process an event through all matching rules
     */
    processEvent(event, world) {
        // Find matching rules
        const matchingRules = this.getMatchingRules(event);
        // Sort by priority (higher first)
        matchingRules.sort((a, b) => (b.priority || 0) - (a.priority || 0));
        let prevented = false;
        let message;
        const allEvents = [];
        const allChanges = [];
        // Process each matching rule
        for (const rule of matchingRules) {
            // Check condition if present
            if (rule.condition && !rule.condition(event, world)) {
                continue;
            }
            // Execute the rule
            let result;
            try {
                result = rule.action(event, world);
            }
            catch (error) {
                console.error(`Error executing rule ${rule.id}:`, error);
                continue; // Skip this rule and continue with others
            }
            // Handle prevention (first rule to prevent wins)
            if (result.prevent && !prevented) {
                prevented = true;
                message = result.message;
                // Create a narrative event for the prevent message
                if (result.message) {
                    allEvents.push((0, event_system_1.createEvent)(standard_events_1.StandardEventTypes.NARRATIVE, { message: result.message }, {
                        actor: event.entities.actor,
                        location: event.entities.location
                    }));
                }
                // When prevented, stop processing more rules
                break;
            }
            // Collect events
            if (result.events) {
                allEvents.push(...result.events);
            }
            // Handle message (create narrative event)
            if (result.message && !result.prevent) {
                // Keep the message for the result
                if (!message) {
                    message = result.message;
                }
                allEvents.push((0, event_system_1.createEvent)(standard_events_1.StandardEventTypes.NARRATIVE, { message: result.message }, {
                    actor: event.entities.actor,
                    location: event.entities.location
                }));
            }
            // Collect entity changes
            if (result.changes) {
                allChanges.push(...result.changes);
            }
        }
        // Apply entity changes to the world
        for (const change of allChanges) {
            const entity = world.getEntity(change.entityId);
            if (entity) {
                world.updateEntity(change.entityId, {
                    [change.attribute]: change.value
                });
            }
        }
        return {
            prevent: prevented,
            message,
            events: allEvents,
            changes: allChanges
        };
    }
    /**
     * Find rules that match the given event
     */
    getMatchingRules(event) {
        return Array.from(this.rules.values()).filter(rule => {
            // Handle wildcards
            if (rule.eventType === '*')
                return true;
            // Handle category wildcards (e.g., 'item:*')
            if (rule.eventType.endsWith(':*')) {
                const category = rule.eventType.split(':')[0];
                return event.type.startsWith(`${category}:`);
            }
            // Exact match
            return rule.eventType === event.type;
        });
    }
}
exports.SimpleRuleSystemImpl = SimpleRuleSystemImpl;
/**
 * Create a new simple rule system
 */
function createSimpleRuleSystem() {
    return new SimpleRuleSystemImpl();
}
