// @sharpee/stdlib/actions/standard/reading/reading.js
"use strict";
/**
 * Reading action - handles reading text from readable entities
 *
 * Uses four-phase pattern:
 * 1. validate: Check target exists and is readable
 * 2. execute: Mark as read, compute text (mutation)
 * 3. blocked: Generate events when validation fails
 * 4. report: Generate success events
 *
 * Supports implicit take for portable readable items (books, notes).
 * Scenery items (inscriptions, signs) can be read without taking.
 *
 * @module
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.reading = void 0;
const world_model_1 = require("../../../../world-model/index.js");
const reading_events_1 = require("./reading-events");
function getReadingSharedData(context) {
    return context.sharedData;
}
/**
 * Reading action implementation
 *
 * Handles:
 * - Reading books, notes, signs, inscriptions
 * - Multi-page books
 * - Language/ability requirements
 * - Tracking what has been read
 * - Implicit take for portable items
 */
const types_1 = require("../../../scope/types");
exports.reading = {
    id: 'if.action.reading',
    // Default scope requirements for this action's slots
    defaultScope: {
        target: types_1.ScopeLevel.REACHABLE // REACHABLE allows implicit take
    },
    // ADR-104: Implicit inference requirements
    targetRequirements: {
        trait: world_model_1.TraitType.READABLE,
        description: 'readable'
    },
    // Reading typically requires holding the item (books, notes, etc.)
    // Inscriptions/signs don't need taking - they're scenery (handled in validate)
    requiresHolding: true,
    validate(context) {
        const { directObject } = context.command;
        // Require a direct object
        if (!directObject?.entity) {
            return {
                valid: false,
                error: 'what_to_read'
            };
        }
        const target = directObject.entity;
        // Check scope - must be able to see the target
        const scopeCheck = context.requireScope(target, types_1.ScopeLevel.VISIBLE);
        if (!scopeCheck.ok) {
            return scopeCheck.error;
        }
        // Check if entity has readable trait
        const readable = target.getTrait(world_model_1.ReadableTrait);
        if (!readable) {
            return {
                valid: false,
                error: 'not_readable',
                params: { item: String(target.attributes.name || 'that') }
            };
        }
        // Check if currently readable
        if (readable.isReadable === false) {
            return {
                valid: false,
                error: 'cannot_read_now',
                params: {
                    item: String(target.attributes.name || 'that'),
                    reason: readable.cannotReadMessage || 'cannot_read_now'
                }
            };
        }
        // Implicit take for portable items (not scenery)
        // Scenery items (inscriptions, signs on walls) can be read without taking
        if (!target.has(world_model_1.TraitType.SCENERY)) {
            const carryCheck = context.requireCarriedOrImplicitTake(target);
            if (!carryCheck.ok) {
                return carryCheck.error;
            }
        }
        // Check ability requirements
        if (readable.requiresAbility) {
            // If ability is required but player doesn't have it
            // TODO: Check if player has the required ability
            // For now, we'll assume they do if requiredAbility is set
        }
        return { valid: true };
    },
    execute(context) {
        const { directObject } = context.command;
        const target = directObject.entity;
        const readable = target.getTrait(world_model_1.ReadableTrait);
        const sharedData = getReadingSharedData(context);
        // Mark as read (world mutation)
        readable.hasBeenRead = true;
        // Build event data
        const eventData = {
            targetId: target.id,
            targetName: String(target.attributes.name || 'something'),
            text: readable.text,
            readableType: readable.readableType || 'text',
            hasBeenRead: true
        };
        // Handle multi-page items
        if (readable.pageContent && readable.currentPage) {
            eventData.currentPage = readable.currentPage;
            eventData.totalPages = readable.pages || readable.pageContent.length;
            eventData.text = readable.pageContent[readable.currentPage - 1];
        }
        // Create the reading event
        const readEvent = (0, reading_events_1.createReadingEvent)(eventData);
        // Determine message based on type
        let messageId = 'read_text';
        if (readable.readableType === 'book') {
            messageId = readable.pageContent ? 'read_book_page' : 'read_book';
        }
        else if (readable.readableType === 'sign') {
            messageId = 'read_sign';
        }
        else if (readable.readableType === 'inscription') {
            messageId = 'read_inscription';
        }
        const params = {
            item: String(target.attributes.name || 'something'),
            text: eventData.text
        };
        if (eventData.currentPage) {
            params.currentPage = eventData.currentPage;
            params.totalPages = eventData.totalPages;
        }
        // Store in sharedData for report phase
        sharedData.readEvent = readEvent;
        sharedData.messageId = messageId;
        sharedData.params = params;
    },
    blocked(context, result) {
        const target = context.command.directObject?.entity;
        return [context.event('if.event.read', {
                blocked: true,
                messageId: `${context.action.id}.${result.error}`,
                params: { item: target?.name, ...result.params },
                reason: result.error,
                targetId: target?.id,
                targetName: target?.name
            })];
    },
    report(context) {
        const sharedData = getReadingSharedData(context);
        const events = [];
        // Prepend any implicit take events (from requireCarriedOrImplicitTake)
        if (context.sharedData.implicitTakeEvents) {
            events.push(...context.sharedData.implicitTakeEvents);
        }
        // Emit read event with messageId for text rendering
        const target = context.command.directObject?.entity;
        events.push(context.event('if.event.read', {
            messageId: `${context.action.id}.${sharedData.messageId || 'read_text'}`,
            params: sharedData.params || {},
            targetId: target?.id,
            targetName: String(target?.attributes.name || 'something'),
            text: sharedData.params?.text
        }));
        return events;
    }
};
