// @sharpee/parser-en-us/slot-consumers/index.js
"use strict";
/**
 * @file Slot Consumer Registry
 * @description Registry for slot consumption strategies (ADR-088)
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SlotConsumerRegistry = exports.VocabularySlotConsumer = exports.TypedSlotConsumer = exports.TextSlotConsumer = exports.EntitySlotConsumer = exports.isPatternDelimiter = exports.getNextPatternToken = void 0;
exports.createDefaultRegistry = createDefaultRegistry;
const entity_slot_consumer_1 = require("./entity-slot-consumer");
const text_slot_consumer_1 = require("./text-slot-consumer");
const typed_slot_consumer_1 = require("./typed-slot-consumer");
const vocabulary_slot_consumer_1 = require("./vocabulary-slot-consumer");
var slot_consumer_1 = require("./slot-consumer");
Object.defineProperty(exports, "getNextPatternToken", { enumerable: true, get: function () { return slot_consumer_1.getNextPatternToken; } });
Object.defineProperty(exports, "isPatternDelimiter", { enumerable: true, get: function () { return slot_consumer_1.isPatternDelimiter; } });
var entity_slot_consumer_2 = require("./entity-slot-consumer");
Object.defineProperty(exports, "EntitySlotConsumer", { enumerable: true, get: function () { return entity_slot_consumer_2.EntitySlotConsumer; } });
var text_slot_consumer_2 = require("./text-slot-consumer");
Object.defineProperty(exports, "TextSlotConsumer", { enumerable: true, get: function () { return text_slot_consumer_2.TextSlotConsumer; } });
var typed_slot_consumer_2 = require("./typed-slot-consumer");
Object.defineProperty(exports, "TypedSlotConsumer", { enumerable: true, get: function () { return typed_slot_consumer_2.TypedSlotConsumer; } });
var vocabulary_slot_consumer_2 = require("./vocabulary-slot-consumer");
Object.defineProperty(exports, "VocabularySlotConsumer", { enumerable: true, get: function () { return vocabulary_slot_consumer_2.VocabularySlotConsumer; } });
/**
 * Registry for slot consumers
 * Maps slot types to their consumer implementations
 */
class SlotConsumerRegistry {
    consumers = new Map();
    /**
     * Register a consumer for its declared slot types
     */
    register(consumer) {
        for (const type of consumer.slotTypes) {
            this.consumers.set(type, consumer);
        }
    }
    /**
     * Check if a consumer is registered for a slot type
     */
    hasConsumer(slotType) {
        return this.consumers.has(slotType);
    }
    /**
     * Consume tokens for a slot type
     * @param ctx The consumption context
     * @returns SlotMatch if successful, null if no match
     * @throws Error if no consumer registered for the slot type
     */
    consume(ctx) {
        const consumer = this.consumers.get(ctx.slotType);
        if (!consumer) {
            throw new Error(`No consumer registered for slot type: ${ctx.slotType}`);
        }
        return consumer.consume(ctx);
    }
    /**
     * Get all registered slot types
     */
    getRegisteredTypes() {
        return Array.from(this.consumers.keys());
    }
}
exports.SlotConsumerRegistry = SlotConsumerRegistry;
/**
 * Create a default registry with all standard consumers
 * This will be populated as consumers are extracted
 */
function createDefaultRegistry() {
    const registry = new SlotConsumerRegistry();
    // ADR-088: All slot consumers extracted
    registry.register(new entity_slot_consumer_1.EntitySlotConsumer());
    registry.register(new text_slot_consumer_1.TextSlotConsumer());
    registry.register(new typed_slot_consumer_1.TypedSlotConsumer());
    registry.register(new vocabulary_slot_consumer_1.VocabularySlotConsumer());
    return registry;
}
