// @sharpee/parser-en-us/slot-consumers/text-slot-consumer.js
"use strict";
/**
 * @file Text Slot Consumer
 * @description Handles text capture slots (ADR-088)
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TextSlotConsumer = void 0;
const if_domain_1 = require("../../if-domain/index.js");
const slot_consumer_1 = require("./slot-consumer");
/**
 * Consumer for text slots (TEXT, TEXT_GREEDY, QUOTED_TEXT, TOPIC)
 * Captures raw text without entity resolution
 */
class TextSlotConsumer {
    slotTypes = [
        if_domain_1.SlotType.TEXT,
        if_domain_1.SlotType.TEXT_GREEDY,
        if_domain_1.SlotType.QUOTED_TEXT,
        if_domain_1.SlotType.TOPIC
    ];
    consume(ctx) {
        const { slotType } = ctx;
        switch (slotType) {
            case if_domain_1.SlotType.TEXT:
                return this.consumeText(ctx);
            case if_domain_1.SlotType.TEXT_GREEDY:
                return this.consumeGreedyText(ctx);
            case if_domain_1.SlotType.QUOTED_TEXT:
                return this.consumeQuotedText(ctx);
            case if_domain_1.SlotType.TOPIC:
                return this.consumeTopic(ctx);
            default:
                return null;
        }
    }
    /**
     * Consume a single token as raw text (no entity resolution)
     */
    consumeText(ctx) {
        const { tokens, startIndex, slotType } = ctx;
        if (startIndex >= tokens.length) {
            return null;
        }
        const token = tokens[startIndex];
        return {
            tokens: [startIndex],
            text: token.word,
            confidence: 1.0,
            slotType
        };
    }
    /**
     * Consume tokens until next pattern element or end (greedy text)
     */
    consumeGreedyText(ctx) {
        const { tokens, startIndex, slotType } = ctx;
        const nextPatternToken = (0, slot_consumer_1.getNextPatternToken)(ctx);
        const consumedIndices = [];
        const consumedWords = [];
        for (let i = startIndex; i < tokens.length; i++) {
            const token = tokens[i];
            // Check if this token matches the next pattern element (delimiter)
            if ((0, slot_consumer_1.isPatternDelimiter)(token, nextPatternToken)) {
                break; // Stop - this token belongs to next pattern element
            }
            consumedIndices.push(i);
            consumedWords.push(token.word);
        }
        if (consumedIndices.length === 0) {
            return null;
        }
        return {
            tokens: consumedIndices,
            text: consumedWords.join(' '),
            confidence: 1.0,
            slotType
        };
    }
    /**
     * Consume a quoted text slot
     * Matches text enclosed in double quotes
     */
    consumeQuotedText(ctx) {
        const { tokens, startIndex } = ctx;
        if (startIndex >= tokens.length) {
            return null;
        }
        const token = tokens[startIndex];
        const word = token.word;
        // Check if token starts with quote
        if (!word.startsWith('"')) {
            return null;
        }
        // Single token quoted text: "hello"
        if (word.endsWith('"') && word.length > 2) {
            return {
                tokens: [startIndex],
                text: word.slice(1, -1), // Remove quotes
                confidence: 1.0,
                slotType: if_domain_1.SlotType.QUOTED_TEXT
            };
        }
        // Multi-token quoted text: "hello world"
        // Consume tokens until closing quote
        const consumedIndices = [startIndex];
        const consumedWords = [word.slice(1)]; // Remove opening quote
        for (let i = startIndex + 1; i < tokens.length; i++) {
            const t = tokens[i];
            consumedIndices.push(i);
            if (t.word.endsWith('"')) {
                consumedWords.push(t.word.slice(0, -1)); // Remove closing quote
                return {
                    tokens: consumedIndices,
                    text: consumedWords.join(' '),
                    confidence: 1.0,
                    slotType: if_domain_1.SlotType.QUOTED_TEXT
                };
            }
            consumedWords.push(t.word);
        }
        // No closing quote found
        return null;
    }
    /**
     * Consume a topic slot
     * Consumes one or more words until next pattern element
     */
    consumeTopic(ctx) {
        const { tokens, startIndex } = ctx;
        if (startIndex >= tokens.length) {
            return null;
        }
        const nextPatternToken = (0, slot_consumer_1.getNextPatternToken)(ctx);
        const consumedIndices = [];
        const consumedWords = [];
        for (let i = startIndex; i < tokens.length; i++) {
            const token = tokens[i];
            // Check for pattern delimiter
            if ((0, slot_consumer_1.isPatternDelimiter)(token, nextPatternToken)) {
                break;
            }
            consumedIndices.push(i);
            consumedWords.push(token.word);
        }
        if (consumedIndices.length === 0) {
            return null;
        }
        return {
            tokens: consumedIndices,
            text: consumedWords.join(' '),
            confidence: 1.0,
            slotType: if_domain_1.SlotType.TOPIC
        };
    }
}
exports.TextSlotConsumer = TextSlotConsumer;
