// @sharpee/parser-en-us/slot-consumers/typed-slot-consumer.js
"use strict";
/**
 * @file Typed Slot Consumer
 * @description Handles typed value slots (ADR-088, ADR-082)
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypedSlotConsumer = void 0;
const if_domain_1 = require("../../if-domain/index.js");
const lang_en_us_1 = require("../../lang-en-us/index.js");
/**
 * Consumer for typed value slots (NUMBER, ORDINAL, TIME, DIRECTION)
 * Validates input matches expected format/vocabulary
 */
class TypedSlotConsumer {
    slotTypes = [
        if_domain_1.SlotType.NUMBER,
        if_domain_1.SlotType.ORDINAL,
        if_domain_1.SlotType.TIME,
        if_domain_1.SlotType.DIRECTION
    ];
    consume(ctx) {
        const { slotType } = ctx;
        switch (slotType) {
            case if_domain_1.SlotType.NUMBER:
                return this.consumeNumber(ctx);
            case if_domain_1.SlotType.ORDINAL:
                return this.consumeOrdinal(ctx);
            case if_domain_1.SlotType.TIME:
                return this.consumeTime(ctx);
            case if_domain_1.SlotType.DIRECTION:
                return this.consumeDirection(ctx);
            default:
                return null;
        }
    }
    /**
     * Consume a number slot (integer)
     * Matches digits (1, 29, 100) or words (one, twenty)
     */
    consumeNumber(ctx) {
        const { tokens, startIndex } = ctx;
        if (startIndex >= tokens.length) {
            return null;
        }
        const token = tokens[startIndex];
        const normalized = token.normalized;
        // Check word form (using imported vocabulary from lang-en-us)
        if (normalized in lang_en_us_1.cardinalNumbers) {
            return {
                tokens: [startIndex],
                text: token.word,
                confidence: 1.0,
                slotType: if_domain_1.SlotType.NUMBER
            };
        }
        // Check digit form
        if (/^\d+$/.test(normalized)) {
            return {
                tokens: [startIndex],
                text: token.word,
                confidence: 1.0,
                slotType: if_domain_1.SlotType.NUMBER
            };
        }
        return null;
    }
    /**
     * Consume an ordinal slot
     * Matches ordinal words (first, second) or suffixed numbers (1st, 2nd)
     */
    consumeOrdinal(ctx) {
        const { tokens, startIndex } = ctx;
        if (startIndex >= tokens.length) {
            return null;
        }
        const token = tokens[startIndex];
        const normalized = token.normalized;
        // Check word form (using imported vocabulary from lang-en-us)
        if (normalized in lang_en_us_1.ordinalNumbers) {
            return {
                tokens: [startIndex],
                text: token.word,
                confidence: 1.0,
                slotType: if_domain_1.SlotType.ORDINAL
            };
        }
        // Check suffixed number form (1st, 2nd, 3rd, 4th, etc.)
        const ordinalMatch = normalized.match(/^(\d+)(st|nd|rd|th)$/);
        if (ordinalMatch) {
            return {
                tokens: [startIndex],
                text: token.word,
                confidence: 1.0,
                slotType: if_domain_1.SlotType.ORDINAL
            };
        }
        return null;
    }
    /**
     * Consume a time slot
     * Matches HH:MM format (10:40, 6:00)
     */
    consumeTime(ctx) {
        const { tokens, startIndex } = ctx;
        if (startIndex >= tokens.length) {
            return null;
        }
        const token = tokens[startIndex];
        const word = token.word;
        // Match HH:MM format
        const timeMatch = word.match(/^(\d{1,2}):(\d{2})$/);
        if (timeMatch) {
            const hours = parseInt(timeMatch[1], 10);
            const minutes = parseInt(timeMatch[2], 10);
            // Validate time range
            if (hours >= 0 && hours <= 23 && minutes >= 0 && minutes <= 59) {
                return {
                    tokens: [startIndex],
                    text: word,
                    confidence: 1.0,
                    slotType: if_domain_1.SlotType.TIME
                };
            }
        }
        return null;
    }
    /**
     * Consume a direction slot
     * Matches built-in direction vocabulary (n, north, up, etc.)
     */
    consumeDirection(ctx) {
        const { tokens, startIndex } = ctx;
        if (startIndex >= tokens.length) {
            return null;
        }
        const token = tokens[startIndex];
        const normalized = token.normalized;
        // Check direction vocabulary (using imported vocabulary from lang-en-us)
        if (normalized in lang_en_us_1.directionMap) {
            return {
                tokens: [startIndex],
                text: token.word,
                confidence: 1.0,
                slotType: if_domain_1.SlotType.DIRECTION
            };
        }
        return null;
    }
}
exports.TypedSlotConsumer = TypedSlotConsumer;
