// @sharpee/stdlib/actions/standard/putting/putting-data.js
"use strict";
/**
 * Data builder for putting action
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.putDataConfig = exports.buildPutData = void 0;
const snapshot_utils_1 = require("../../base/snapshot-utils");
const buildPutData = (context) => {
    const item = context.command.directObject?.entity;
    const target = context.command.indirectObject?.entity;
    if (!item || !target) {
        return { item: 'nothing', target: 'nowhere' };
    }
    return {
        item: item.name,
        itemSnapshot: (0, snapshot_utils_1.captureEntitySnapshot)(item, context.world, true),
        target: target.name,
        targetSnapshot: (0, snapshot_utils_1.captureEntitySnapshot)(target, context.world, true),
        itemId: item.id,
        targetId: target.id,
        preposition: context.command.parsed.extras?.preposition || 'on'
    };
};
exports.buildPutData = buildPutData;
exports.putDataConfig = {
    builder: exports.buildPutData,
    protectedFields: ['item', 'target', 'itemId', 'targetId']
};
