// @sharpee/stdlib/actions/standard/removing/removing.js
"use strict";
/**
 * Removing action - remove objects from containers or surfaces
 *
 * This action handles taking objects from containers or supporters.
 * It's essentially a targeted form of taking.
 *
 * Uses four-phase pattern:
 * 1. validate: Check item and source exist and are accessible
 * 2. execute: Remove from source and take to inventory
 * 3. blocked: Generate error events when validation fails
 * 4. report: Generate success events with removed data
 *
 * Supports multi-object commands:
 * - "remove all from box" - removes all items from container
 * - "remove all but X from box" - removes all except specified items
 * - "remove X and Y from box" - removes multiple specified items
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.removingAction = void 0;
const snapshot_utils_1 = require("../../base/snapshot-utils");
const world_model_1 = require("../../../../world-model/index.js");
const constants_1 = require("../../constants");
const scope_1 = require("../../../scope");
const removing_messages_1 = require("./removing-messages");
// Import types
const removing_types_1 = require("./removing-types");
// Import multi-object helpers
const multi_object_handler_1 = require("../../../helpers/multi-object-handler");
// ============================================================================
// Helper Functions (standalone to avoid `this` issues in object literal)
// ============================================================================
/**
 * Check if an entity is portable (can be taken)
 */
function isPortable(entity) {
    if (entity.has(world_model_1.TraitType.ROOM))
        return false;
    if (entity.has(world_model_1.TraitType.SCENERY))
        return false;
    if (entity.has(world_model_1.TraitType.ACTOR))
        return false;
    return true;
}
/**
 * Get entities from source container/supporter for "remove all from X"
 */
function getSourceContents(context, source) {
    const contents = context.world.getContents(source.id);
    return contents.filter(e => isPortable(e));
}
/**
 * Validate removing a single entity from source
 */
function validateSingleEntity(context, item, source) {
    const actor = context.player;
    // Check if player already has the item
    if (world_model_1.ActorBehavior.isHolding(actor, item.id, context.world)) {
        return {
            valid: false,
            error: removing_messages_1.RemovingMessages.ALREADY_HAVE,
            params: { item: item.name }
        };
    }
    // Check if the item is actually in/on the source
    const itemLocation = context.world.getLocation(item.id);
    if (!itemLocation || itemLocation !== source.id) {
        if (source.has(world_model_1.TraitType.CONTAINER)) {
            return {
                valid: false,
                error: removing_messages_1.RemovingMessages.NOT_IN_CONTAINER,
                params: {
                    item: item.name,
                    container: source.name
                }
            };
        }
        else if (source.has(world_model_1.TraitType.SUPPORTER)) {
            return {
                valid: false,
                error: removing_messages_1.RemovingMessages.NOT_ON_SURFACE,
                params: {
                    item: item.name,
                    surface: source.name
                }
            };
        }
        else {
            return {
                valid: false,
                error: removing_messages_1.RemovingMessages.NOT_IN_CONTAINER,
                params: {
                    item: item.name,
                    container: source.name
                }
            };
        }
    }
    // Container-specific checks using behavior
    if (source.has(world_model_1.TraitType.CONTAINER)) {
        // Check if container is open
        if (source.has(world_model_1.TraitType.OPENABLE) && !world_model_1.OpenableBehavior.isOpen(source)) {
            return {
                valid: false,
                error: removing_messages_1.RemovingMessages.CONTAINER_CLOSED,
                params: { container: source.name }
            };
        }
    }
    // Use ActorBehavior to validate taking capacity
    if (!world_model_1.ActorBehavior.canTakeItem(actor, item, context.world)) {
        return {
            valid: false,
            error: removing_messages_1.RemovingMessages.CANNOT_TAKE,
            params: { item: item.name }
        };
    }
    return { valid: true };
}
/**
 * Validate a multi-object command (remove all from box, remove X and Y from box)
 */
function validateMultiObject(context) {
    const source = context.command.indirectObject?.entity;
    // Must have a source
    if (!source) {
        return { valid: false, error: removing_messages_1.RemovingMessages.NO_SOURCE };
    }
    // Container must be open if it's openable
    if (source.has(world_model_1.TraitType.CONTAINER) && source.has(world_model_1.TraitType.OPENABLE) && !world_model_1.OpenableBehavior.isOpen(source)) {
        return {
            valid: false,
            error: removing_messages_1.RemovingMessages.CONTAINER_CLOSED,
            params: { container: source.name }
        };
    }
    // Get items from source - for "remove all from X" we need contents of X
    const directObject = context.command.parsed.structure.directObject;
    let items;
    if (directObject?.isAll) {
        // "remove all from X" - get all contents of source
        items = getSourceContents(context, source);
        // Apply exclusions from "all but X"
        const excludedNames = (0, multi_object_handler_1.getExcludedNames)(context);
        if (excludedNames.length > 0) {
            items = items.filter(e => {
                const name = e.name.toLowerCase();
                return !excludedNames.includes(name);
            });
        }
    }
    else if (directObject?.isList && directObject.items) {
        // "remove X and Y from Z" - resolve each item from source contents
        const sourceContents = getSourceContents(context, source);
        items = [];
        for (const np of directObject.items) {
            const searchTerm = (np.head || np.text).toLowerCase();
            const match = sourceContents.find(e => e.name.toLowerCase() === searchTerm);
            if (match) {
                items.push(match);
            }
        }
    }
    else {
        items = [];
    }
    if (items.length === 0) {
        return { valid: false, error: removing_messages_1.RemovingMessages.NOTHING_TO_REMOVE };
    }
    // Validate each entity, store results
    const results = items.map(item => {
        const validation = validateSingleEntity(context, item, source);
        return {
            entity: item,
            success: validation.valid,
            error: validation.valid ? undefined : validation.error,
            errorParams: validation.valid ? undefined : validation.params
        };
    });
    // Store results for execute/report phases
    const sharedData = (0, removing_types_1.getRemovingSharedData)(context);
    sharedData.multiObjectResults = results;
    // Valid if at least one can be removed
    const anySuccess = results.some(r => r.success);
    if (!anySuccess) {
        // All failed - return the first error
        return { valid: false, error: results[0].error, params: results[0].errorParams };
    }
    return { valid: true };
}
/**
 * Execute removing a single entity from source
 */
function executeSingleEntity(context, item, source, result) {
    const actor = context.player;
    // First validate removal from source using appropriate behavior
    let removeResult = null;
    if (source.has(world_model_1.TraitType.CONTAINER)) {
        removeResult = world_model_1.ContainerBehavior.removeItem(source, item, context.world);
    }
    else if (source.has(world_model_1.TraitType.SUPPORTER)) {
        removeResult = world_model_1.SupporterBehavior.removeItem(source, item, context.world);
    }
    result.removeResult = removeResult;
    // Validate taking using ActorBehavior
    const takeResult = world_model_1.ActorBehavior.takeItem(actor, item, context.world);
    result.takeResult = takeResult;
    // Perform the actual move - this is the critical mutation!
    context.world.moveEntity(item.id, actor.id);
}
/**
 * Generate success events for removing a single entity
 *
 * Uses simplified event pattern (ADR-097): domain event carries messageId directly.
 */
function reportSingleSuccess(context, item, source, result, events) {
    const actor = context.player;
    // Build message params and determine message
    const params = {
        item: item.name
    };
    let messageKey;
    if (source.has(world_model_1.TraitType.CONTAINER)) {
        params.container = source.name;
        messageKey = removing_messages_1.RemovingMessages.REMOVED_FROM;
    }
    else {
        params.surface = source.name;
        messageKey = removing_messages_1.RemovingMessages.REMOVED_FROM_SURFACE;
    }
    // Emit domain event with messageId (simplified pattern - ADR-097)
    // Using if.event.taken since removing is semantically a form of taking
    events.push(context.event('if.event.taken', {
        // Rendering data (messageId + params for text-service)
        messageId: `${context.action.id}.${messageKey}`,
        params,
        // Domain data (for event sourcing / handlers)
        item: item.name,
        itemId: item.id,
        actorId: actor.id,
        fromLocation: source.id,
        container: source.name,
        fromContainer: source.has(world_model_1.TraitType.CONTAINER),
        fromSupporter: source.has(world_model_1.TraitType.SUPPORTER) && !source.has(world_model_1.TraitType.CONTAINER),
        itemSnapshot: (0, snapshot_utils_1.captureEntitySnapshot)(item, context.world, true),
        actorSnapshot: (0, snapshot_utils_1.captureEntitySnapshot)(actor, context.world, false),
        sourceSnapshot: (0, snapshot_utils_1.captureEntitySnapshot)(source, context.world, source.has(world_model_1.TraitType.ROOM))
    }));
}
/**
 * Generate blocked event for a single entity that couldn't be removed
 *
 * Uses simplified event pattern (ADR-097): domain event carries messageId directly.
 */
function reportSingleBlocked(context, item, source, error, errorParams, events) {
    events.push(context.event('if.event.remove_blocked', {
        // Rendering data
        messageId: `${context.action.id}.${error}`,
        params: { ...errorParams, item: item.name, source: source.name },
        // Domain data
        itemId: item.id,
        itemName: item.name,
        sourceId: source.id,
        sourceName: source.name,
        reason: error
    }));
}
// ============================================================================
// Action Definition
// ============================================================================
exports.removingAction = {
    id: constants_1.IFActions.REMOVING,
    // Default scope requirements for this action's slots
    defaultScope: {
        item: scope_1.ScopeLevel.REACHABLE,
        source: scope_1.ScopeLevel.REACHABLE
    },
    requiredMessages: [
        'no_target',
        'no_source',
        'not_in_container',
        'not_on_surface',
        'container_closed',
        'removed_from',
        'removed_from_surface',
        'cant_reach',
        'already_have',
        'nothing_to_remove'
    ],
    group: 'object_manipulation',
    metadata: {
        requiresDirectObject: true,
        requiresIndirectObject: true,
        directObjectScope: scope_1.ScopeLevel.REACHABLE
    },
    validate(context) {
        // Check for multi-object command
        if ((0, multi_object_handler_1.isMultiObjectCommand)(context)) {
            return validateMultiObject(context);
        }
        // Single object validation
        const item = context.command.directObject?.entity;
        const source = context.command.indirectObject?.entity;
        // Validate we have an item
        if (!item) {
            return {
                valid: false,
                error: removing_messages_1.RemovingMessages.NO_TARGET,
                params: {}
            };
        }
        // Validate we have a source
        if (!source) {
            return {
                valid: false,
                error: removing_messages_1.RemovingMessages.NO_SOURCE,
                params: { item: item.name }
            };
        }
        return validateSingleEntity(context, item, source);
    },
    execute(context) {
        const sharedData = (0, removing_types_1.getRemovingSharedData)(context);
        const source = context.command.indirectObject.entity;
        // Check for multi-object command
        if (sharedData.multiObjectResults) {
            // Execute for each successful validation
            for (const result of sharedData.multiObjectResults) {
                if (result.success) {
                    executeSingleEntity(context, result.entity, source, result);
                }
            }
            return;
        }
        // Single object execution
        const actor = context.player;
        const item = context.command.directObject.entity;
        // First validate removal from source using appropriate behavior
        let removeResult = null;
        if (source.has(world_model_1.TraitType.CONTAINER)) {
            removeResult = world_model_1.ContainerBehavior.removeItem(source, item, context.world);
        }
        else if (source.has(world_model_1.TraitType.SUPPORTER)) {
            removeResult = world_model_1.SupporterBehavior.removeItem(source, item, context.world);
        }
        // Store results for report phase using sharedData
        sharedData.removeResult = removeResult;
        // Validate taking using ActorBehavior
        const takeResult = world_model_1.ActorBehavior.takeItem(actor, item, context.world);
        sharedData.takeResult = takeResult;
        // Perform the actual move - this is the critical mutation!
        context.world.moveEntity(item.id, actor.id);
    },
    /**
     * Report events after successful removing
     * Only called on success path - validation passed
     */
    report(context) {
        const sharedData = (0, removing_types_1.getRemovingSharedData)(context);
        const source = context.command.indirectObject.entity;
        // Check for multi-object command
        if (sharedData.multiObjectResults) {
            const events = [];
            // Generate events for each item (success and failure)
            for (const result of sharedData.multiObjectResults) {
                if (result.success) {
                    reportSingleSuccess(context, result.entity, source, result, events);
                }
                else {
                    reportSingleBlocked(context, result.entity, source, result.error, result.errorParams, events);
                }
            }
            return events;
        }
        // Single object report
        const actor = context.player;
        const item = context.command.directObject.entity;
        // Build message params and determine message
        const params = {
            item: item.name
        };
        let messageKey;
        if (source.has(world_model_1.TraitType.CONTAINER)) {
            params.container = source.name;
            messageKey = removing_messages_1.RemovingMessages.REMOVED_FROM;
        }
        else {
            params.surface = source.name;
            messageKey = removing_messages_1.RemovingMessages.REMOVED_FROM_SURFACE;
        }
        // Emit domain event with messageId (simplified pattern - ADR-097)
        return [
            context.event('if.event.taken', {
                // Rendering data (messageId + params for text-service)
                messageId: `${context.action.id}.${messageKey}`,
                params,
                // Domain data (for event sourcing / handlers)
                item: item.name,
                itemId: item.id,
                actorId: actor.id,
                fromLocation: source.id,
                container: source.name,
                fromContainer: source.has(world_model_1.TraitType.CONTAINER),
                fromSupporter: source.has(world_model_1.TraitType.SUPPORTER) && !source.has(world_model_1.TraitType.CONTAINER),
                itemSnapshot: (0, snapshot_utils_1.captureEntitySnapshot)(item, context.world, true),
                actorSnapshot: (0, snapshot_utils_1.captureEntitySnapshot)(actor, context.world, false),
                sourceSnapshot: (0, snapshot_utils_1.captureEntitySnapshot)(source, context.world, source.has(world_model_1.TraitType.ROOM))
            })
        ];
    },
    /**
     * Generate events when validation fails
     *
     * Uses simplified event pattern (ADR-097): domain event carries messageId directly.
     */
    blocked(context, result) {
        const item = context.command.directObject?.entity;
        const source = context.command.indirectObject?.entity;
        return [context.event('if.event.remove_blocked', {
                // Rendering data
                messageId: `${context.action.id}.${result.error}`,
                params: {
                    ...result.params,
                    item: item?.name,
                    source: source?.name
                },
                // Domain data
                itemId: item?.id,
                itemName: item?.name,
                sourceId: source?.id,
                sourceName: source?.name,
                reason: result.error
            })];
    }
};
