// @sharpee/stdlib/actions/standard/removing/removing-data.js
"use strict";
/**
 * Data builder for removing action
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.removedDataConfig = exports.buildRemovedData = void 0;
const snapshot_utils_1 = require("../../base/snapshot-utils");
const buildRemovedData = (context) => {
    const item = context.command.directObject?.entity;
    const source = context.command.indirectObject?.entity;
    const actor = context.player;
    if (!item) {
        return { item: 'nothing' };
    }
    return {
        item: item.name,
        itemSnapshot: (0, snapshot_utils_1.captureEntitySnapshot)(item, context.world, true),
        actorSnapshot: (0, snapshot_utils_1.captureEntitySnapshot)(actor, context.world, false),
        sourceSnapshot: source ? (0, snapshot_utils_1.captureEntitySnapshot)(source, context.world, true) : undefined,
        itemId: item.id,
        fromLocation: source?.name || 'somewhere'
    };
};
exports.buildRemovedData = buildRemovedData;
exports.removedDataConfig = {
    builder: exports.buildRemovedData,
    protectedFields: ['item', 'itemId']
};
