// @sharpee/world-model/traits/pushable/pushableTrait.js
"use strict";
/**
 * Pushable trait implementation
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PushableTrait = void 0;
const trait_types_1 = require("../trait-types");
/**
 * Pushable trait for objects that can be pushed
 *
 * This trait contains only data - all logic for pushing
 * is handled by the pushing action.
 */
class PushableTrait {
    static type = trait_types_1.TraitType.PUSHABLE;
    type = trait_types_1.TraitType.PUSHABLE;
    // PushableData properties
    pushType;
    revealsPassage;
    pushSound;
    requiresStrength;
    repeatable;
    state;
    pushCount;
    maxPushes;
    pushDirection;
    activates;
    effects;
    constructor(data = {}) {
        this.pushType = data.pushType ?? 'button';
        this.revealsPassage = data.revealsPassage;
        this.pushSound = data.pushSound;
        this.requiresStrength = data.requiresStrength;
        this.repeatable = data.repeatable ?? true;
        this.state = data.state ?? 'default';
        this.pushCount = data.pushCount ?? 0;
        this.maxPushes = data.maxPushes;
        this.pushDirection = data.pushDirection;
        this.activates = data.activates;
        this.effects = data.effects;
    }
}
exports.PushableTrait = PushableTrait;
