// @sharpee/world-model/examples/event-handler-registration.js
"use strict";
// Example of registering event handlers for world-model
// This would typically be done in stdlib or a new event processor package
Object.defineProperty(exports, "__esModule", { value: true });
exports.registerStandardEventHandlers = registerStandardEventHandlers;
exports.registerStandardEventValidators = registerStandardEventValidators;
exports.registerStandardEventPreviewers = registerStandardEventPreviewers;
const constants_1 = require("../constants");
const trait_types_1 = require("../traits/trait-types");
const openableTrait_1 = require("../traits/openable/openableTrait");
const lockableTrait_1 = require("../traits/lockable/lockableTrait");
/**
 * Register standard IF event handlers with the world model
 */
function registerStandardEventHandlers(world) {
    // Handler for TAKEN event - moves entity to actor
    world.registerEventHandler(constants_1.IFEvents.TAKEN, (event, world) => {
        const { actor, target } = event.entities;
        if (actor && target) {
            world.moveEntity(target, actor);
        }
    });
    // Handler for DROPPED event - moves entity to actor's location
    world.registerEventHandler(constants_1.IFEvents.DROPPED, (event, world) => {
        const { actor, target } = event.entities;
        if (actor && target) {
            const actorLocation = world.getLocation(actor);
            if (actorLocation) {
                world.moveEntity(target, actorLocation);
            }
        }
    });
    // Handler for OPENED event - sets openable trait's isOpen to true
    world.registerEventHandler(constants_1.IFEvents.OPENED, (event, world) => {
        const { target } = event.entities;
        if (target) {
            const entity = world.getEntity(target);
            if (entity && entity.has(trait_types_1.TraitType.OPENABLE)) {
                const openable = entity.getTrait(openableTrait_1.OpenableTrait);
                if (openable) {
                    openable.isOpen = true;
                }
            }
        }
    });
    // Handler for CLOSED event - sets openable trait's isOpen to false
    world.registerEventHandler(constants_1.IFEvents.CLOSED, (event, world) => {
        const { target } = event.entities;
        if (target) {
            const entity = world.getEntity(target);
            if (entity && entity.has(trait_types_1.TraitType.OPENABLE)) {
                const openable = entity.getTrait(openableTrait_1.OpenableTrait);
                if (openable) {
                    openable.isOpen = false;
                }
            }
        }
    });
    // Handler for LOCKED event
    world.registerEventHandler(constants_1.IFEvents.LOCKED, (event, world) => {
        const { target } = event.entities;
        if (target) {
            const entity = world.getEntity(target);
            if (entity && entity.has(trait_types_1.TraitType.LOCKABLE)) {
                const lockable = entity.getTrait(lockableTrait_1.LockableTrait);
                if (lockable) {
                    lockable.isLocked = true;
                }
            }
        }
    });
    // Handler for UNLOCKED event
    world.registerEventHandler(constants_1.IFEvents.UNLOCKED, (event, world) => {
        const { target } = event.entities;
        if (target) {
            const entity = world.getEntity(target);
            if (entity && entity.has(trait_types_1.TraitType.LOCKABLE)) {
                const lockable = entity.getTrait(lockableTrait_1.LockableTrait);
                if (lockable) {
                    lockable.isLocked = false;
                }
            }
        }
    });
    // Handler for EATEN event - removes entity from world
    world.registerEventHandler(constants_1.IFEvents.EATEN, (event, world) => {
        const { target } = event.entities;
        if (target) {
            world.removeEntity(target);
        }
    });
    // Add more handlers as needed...
}
/**
 * Register event validators
 */
function registerStandardEventValidators(world) {
    // Validator for TAKEN event
    world.registerEventValidator(constants_1.IFEvents.TAKEN, (event, world) => {
        const { actor, target } = event.entities;
        if (!actor || !target)
            return false;
        // Check entities exist
        if (!world.hasEntity(actor) || !world.hasEntity(target))
            return false;
        // Check target isn't already held by actor
        if (world.getLocation(target) === actor)
            return false;
        // Check actor can contain items
        const actorEntity = world.getEntity(actor);
        if (!actorEntity?.has(trait_types_1.TraitType.CONTAINER) && !actorEntity?.has(trait_types_1.TraitType.ACTOR))
            return false;
        return true;
    });
    // Validator for DROPPED event
    world.registerEventValidator(constants_1.IFEvents.DROPPED, (event, world) => {
        const { actor, target } = event.entities;
        if (!actor || !target)
            return false;
        // Check target is held by actor
        if (world.getLocation(target) !== actor)
            return false;
        // Check actor has a location
        if (!world.getLocation(actor))
            return false;
        return true;
    });
    // Add more validators as needed...
}
/**
 * Register event previewers
 */
function registerStandardEventPreviewers(world) {
    // Previewer for TAKEN event
    world.registerEventPreviewer(constants_1.IFEvents.TAKEN, (event, world) => {
        const { actor, target } = event.entities;
        if (!actor || !target)
            return [];
        return [{
                type: 'move',
                entityId: target,
                oldValue: world.getLocation(target),
                newValue: actor,
                details: { event: 'taken' }
            }];
    });
    // Previewer for EATEN event
    world.registerEventPreviewer(constants_1.IFEvents.EATEN, (event, world) => {
        const { target } = event.entities;
        if (!target)
            return [];
        return [{
                type: 'delete',
                entityId: target,
                details: { event: 'eaten' }
            }];
    });
    // Add more previewers as needed...
}
