// @sharpee/world-model/traits/destructible/destructibleTrait.js
"use strict";
/**
 * Destructible trait for entities that require multiple hits or specific tools to destroy
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DestructibleTrait = void 0;
/**
 * Destructible trait indicates an entity can be destroyed with multiple hits.
 *
 * This trait contains only data - all destruction logic
 * is in DestructibleBehavior. Messages are stored here to be
 * injected into events during the report phase.
 */
class DestructibleTrait {
    static type = 'destructible';
    type = 'destructible';
    // DestructibleData properties
    hitPoints;
    maxHitPoints;
    requiresWeapon;
    requiresType;
    transformTo;
    revealExit;
    damageMessage;
    destroyMessage;
    invulnerable;
    armor;
    constructor(data = {}) {
        // Set defaults and merge with provided data
        this.hitPoints = data.hitPoints ?? data.maxHitPoints ?? 3;
        this.maxHitPoints = data.maxHitPoints ?? 3;
        this.requiresWeapon = data.requiresWeapon ?? false;
        this.requiresType = data.requiresType;
        this.transformTo = data.transformTo;
        this.revealExit = data.revealExit;
        this.damageMessage = data.damageMessage;
        this.destroyMessage = data.destroyMessage;
        this.invulnerable = data.invulnerable ?? false;
        this.armor = data.armor ?? 0;
    }
}
exports.DestructibleTrait = DestructibleTrait;
