// @sharpee/stdlib/actions/standard/saving/saving.js
"use strict";
/**
 * Saving action - save game state
 *
 * This is a meta action that triggers game save functionality.
 * It emits a platform event that the engine will process after turn completion.
 *
 * Four-phase pattern:
 * 1. validate: Check if save is allowed, validate save name
 * 2. execute: Analyze save context, store in sharedData (no world mutations)
 * 3. blocked: Handle validation failures
 * 4. report: Emit platform event and notifications
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.savingAction = void 0;
const core_1 = require("../../../../core/index.js");
const constants_1 = require("../../constants");
/**
 * Analyze save context from game state
 */
function analyzeSaveContext(context, saveName) {
    // Get game state
    const sharedData = context.world.getCapability('sharedData') || {};
    const score = sharedData.score || 0;
    const moves = sharedData.moves || 0;
    const turnCount = sharedData.turnCount || 0;
    // Check if this is a quick save or auto save
    const isQuickSave = saveName === 'quicksave' || context.command.parsed.extras?.quick;
    const isAutoSave = context.command.parsed.extras?.auto;
    // Build save context
    const saveContext = {
        saveName: saveName !== 'default' ? saveName : undefined,
        autosave: isAutoSave,
        timestamp: Date.now(),
        metadata: {
            score,
            moves,
            turnCount,
            quickSave: isQuickSave
        }
    };
    // Build event data
    const eventData = {
        saveName,
        timestamp: saveContext.timestamp,
        metadata: {
            score,
            moves,
            turnCount,
            quickSave: isQuickSave
        }
    };
    return {
        saveName,
        isQuickSave,
        isAutoSave,
        score,
        moves,
        turnCount,
        saveContext,
        eventData
    };
}
exports.savingAction = {
    id: constants_1.IFActions.SAVING,
    requiredMessages: [
        'game_saved',
        'game_saved_as',
        'save_successful',
        'save_slot',
        'overwrite_save',
        'save_details',
        'quick_save',
        'auto_save',
        'save_failed',
        'no_save_slots',
        'invalid_save_name',
        'save_not_allowed',
        'save_in_progress',
        'confirm_overwrite',
        'save_reminder',
        'saved_locally',
        'saved_to_cloud',
        'save_exported'
    ],
    validate(context) {
        // Extract save slot or name if provided
        const saveName = context.command.parsed.extras?.name ||
            context.command.parsed.extras?.slot ||
            context.command.indirectObject?.parsed?.text ||
            'default';
        // Get game state info
        const sharedData = context.world.getCapability('sharedData') || {};
        // Check save restrictions
        const saveRestrictions = sharedData.saveRestrictions || {};
        if (saveRestrictions.disabled) {
            return {
                valid: false,
                error: 'save_not_allowed'
            };
        }
        if (saveRestrictions.inProgress) {
            return {
                valid: false,
                error: 'save_in_progress'
            };
        }
        // Check if save name is valid
        if (saveName.length > 50 || /[<>:"/\\|?*]/.test(saveName)) {
            return {
                valid: false,
                error: 'invalid_save_name'
            };
        }
        return { valid: true };
    },
    execute(context) {
        // Extract save slot or name if provided
        const saveName = context.command.parsed.extras?.name ||
            context.command.parsed.extras?.slot ||
            context.command.directObject?.parsed?.text ||
            'default';
        // Analyze save context and store in sharedData
        const data = analyzeSaveContext(context, saveName);
        Object.assign(context.sharedData, data);
    },
    blocked(context, result) {
        return [context.event('if.event.save_blocked', {
                messageId: `if.action.saving.${result.error}`,
                params: result.params || {},
                blocked: true,
                reason: result.error
            })];
    },
    report(context) {
        const events = [];
        const data = context.sharedData;
        // Emit platform save requested event
        // The engine will handle this after turn completion
        const platformEvent = (0, core_1.createSaveRequestedEvent)(data.saveContext);
        events.push(platformEvent);
        // Emit the save requested event
        events.push(context.event('if.event.save_requested', data.eventData));
        return events;
    },
    group: "meta",
    metadata: {
        requiresDirectObject: false,
        requiresIndirectObject: false
    }
};
