// @sharpee/world-model/traits/wearable/wearableTrait.js
"use strict";
// packages/world-model/src/traits/wearable/wearableTrait.ts
Object.defineProperty(exports, "__esModule", { value: true });
exports.WearableTrait = void 0;
const trait_types_1 = require("../trait-types");
/**
 * Wearable trait indicates an entity can be worn by the player.
 * This is the base trait for all wearable items (rings, necklaces, clothing, etc.)
 *
 * This trait contains only data - all logic for wearing/removing
 * is in WearableBehavior.
 */
class WearableTrait {
    static type = trait_types_1.TraitType.WEARABLE;
    type = trait_types_1.TraitType.WEARABLE;
    // WearableData properties - internal storage
    worn = false;
    wornBy;
    slot;
    layer;
    wearMessage;
    removeMessage;
    wearableOver;
    blocksSlots;
    weight;
    bulk;
    canRemove;
    bodyPart;
    // Public accessor
    get isWorn() {
        return this.worn;
    }
    set isWorn(value) {
        this.worn = value;
    }
    constructor(data = {}) {
        // Set defaults and merge with provided data
        this.worn = data.isWorn ?? false;
        this.wornBy = data.wornBy;
        this.slot = data.slot ?? 'clothing';
        this.layer = data.layer ?? 1;
        this.wearMessage = data.wearMessage;
        this.removeMessage = data.removeMessage;
        this.wearableOver = data.wearableOver ?? true;
        this.blocksSlots = data.blocksSlots ?? [];
        this.weight = data.weight ?? 1;
        this.bulk = data.bulk ?? 1;
        this.canRemove = data.canRemove ?? true;
        this.bodyPart = data.bodyPart ?? 'torso';
    }
}
exports.WearableTrait = WearableTrait;
