// @sharpee/world-model/traits/wearable/wearableBehavior.js
"use strict";
// packages/world-model/src/traits/wearable/wearableBehavior.ts
Object.defineProperty(exports, "__esModule", { value: true });
exports.WearableBehavior = void 0;
const trait_types_1 = require("../trait-types");
/**
 * Behavior for wearable entities.
 *
 * Handles the logic for wearing and removing wearable items.
 */
class WearableBehavior {
    static requiredTraits = [trait_types_1.TraitType.WEARABLE];
    /**
     * Check if an item can be worn by an actor
     */
    static canWear(item, actor) {
        const wearable = item.get(trait_types_1.TraitType.WEARABLE);
        if (!wearable)
            return false;
        // Can't wear if already worn
        if (wearable.worn)
            return false;
        // TODO: Check if actor has conflicting items in same slot
        // TODO: Check if slots are blocked
        return true;
    }
    /**
     * Check if an item can be removed by an actor
     */
    static canRemove(item, actor) {
        const wearable = item.get(trait_types_1.TraitType.WEARABLE);
        if (!wearable)
            return false;
        // Must be worn to remove
        if (!wearable.worn)
            return false;
        // Must be worn by this actor
        if (wearable.wornBy !== actor.id)
            return false;
        // TODO: Check if other items prevent removal
        return true;
    }
    /**
     * Wear an item
     * @returns Result object describing what happened
     */
    static wear(item, actor) {
        const wearable = item.get(trait_types_1.TraitType.WEARABLE);
        if (!wearable)
            return { success: false };
        if (wearable.worn) {
            if (wearable.wornBy === actor.id) {
                return {
                    success: false,
                    alreadyWorn: true
                };
            }
            else {
                return {
                    success: false,
                    wornByOther: wearable.wornBy
                };
            }
        }
        // TODO: Check slot conflicts
        // Wear it
        wearable.worn = true;
        wearable.wornBy = actor.id;
        return {
            success: true,
            slot: wearable.slot,
            layer: wearable.layer,
            wearMessage: wearable.wearMessage
        };
    }
    /**
     * Remove (take off) a worn item
     * @returns Result object describing what happened
     */
    static remove(item, actor) {
        const wearable = item.get(trait_types_1.TraitType.WEARABLE);
        if (!wearable)
            return { success: false };
        if (!wearable.worn) {
            return {
                success: false,
                notWorn: true
            };
        }
        if (wearable.wornBy !== actor.id) {
            return {
                success: false,
                wornByOther: wearable.wornBy
            };
        }
        // TODO: Check if other items prevent removal
        // Remove it
        wearable.worn = false;
        wearable.wornBy = undefined;
        return {
            success: true,
            removeMessage: wearable.removeMessage
        };
    }
    /**
     * Check if item is currently worn
     */
    static isWorn(item) {
        const wearable = item.get(trait_types_1.TraitType.WEARABLE);
        return wearable ? wearable.worn : false;
    }
    /**
     * Check who is wearing the item
     */
    static getWearer(item) {
        const wearable = item.get(trait_types_1.TraitType.WEARABLE);
        return wearable?.wornBy;
    }
    /**
     * Get the slot this item occupies
     */
    static getSlot(item) {
        const wearable = item.get(trait_types_1.TraitType.WEARABLE);
        return wearable ? wearable.slot : 'clothing';
    }
    /**
     * Get the layer for sorting
     */
    static getLayer(item) {
        const wearable = item.get(trait_types_1.TraitType.WEARABLE);
        return wearable ? wearable.layer : 1;
    }
    /**
     * Check if this item blocks any slots
     */
    static getBlockedSlots(item) {
        const wearable = item.get(trait_types_1.TraitType.WEARABLE);
        return wearable ? [...wearable.blocksSlots] : [];
    }
}
exports.WearableBehavior = WearableBehavior;
