// @sharpee/stdlib/services/scoring/scoring-service.js
"use strict";
/**
 * ScoringService - Event-based scoring service (ADR-085)
 *
 * Features:
 * - Centralized scoring definitions (sourceId → points, reasonMessageId, category)
 * - Deduplication via scored sources Set
 * - Message IDs for all text (i18n-ready via TextService)
 * - Optional: stories can disable scoring entirely
 *
 * @see ADR-085 Event-Based Scoring System
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ScoringService = exports.DEFAULT_RANKS = void 0;
const world_model_1 = require("../../../world-model/index.js");
/**
 * Default rank definitions with message IDs
 */
exports.DEFAULT_RANKS = [
    { threshold: 0, messageId: 'if.rank.beginner', name: 'Beginner' },
    { threshold: 50, messageId: 'if.rank.novice', name: 'Novice' },
    { threshold: 100, messageId: 'if.rank.amateur', name: 'Amateur' },
    { threshold: 200, messageId: 'if.rank.experienced', name: 'Experienced' },
    { threshold: 350, messageId: 'if.rank.expert', name: 'Expert' },
    { threshold: 500, messageId: 'if.rank.master', name: 'Master' },
];
/**
 * Default scoring service implementation (ADR-085)
 *
 * Reads and writes scoring data from the SCORING capability.
 * Stories can extend this class to add custom scoring logic.
 */
class ScoringService {
    world;
    config;
    definitions = new Map();
    enabled;
    constructor(world, config = { maxScore: 350 }) {
        this.world = world;
        this.config = config;
        this.enabled = config.enabled ?? true;
        // Initialize capability data if needed
        const scoring = this.getScoringData();
        if (scoring) {
            if (scoring.maxScore === undefined) {
                scoring.maxScore = config.maxScore;
            }
            if (scoring.scoredSources === undefined) {
                scoring.scoredSources = [];
            }
        }
        // Load definitions from config
        if (config.definitions) {
            for (const [sourceId, def] of Object.entries(config.definitions)) {
                this.definitions.set(sourceId, def);
            }
        }
    }
    /**
     * Check if scoring is enabled
     */
    isEnabled() {
        return this.enabled;
    }
    /**
     * Get the current score
     */
    getScore() {
        const scoring = this.getScoringData();
        return scoring?.scoreValue ?? 0;
    }
    /**
     * Get the maximum possible score
     */
    getMaxScore() {
        const scoring = this.getScoringData();
        return scoring?.maxScore ?? this.config.maxScore;
    }
    /**
     * Set the maximum score (for dynamic max like Dungeo's hidden points)
     */
    setMaxScore(max) {
        const scoring = this.getScoringData();
        if (scoring) {
            scoring.maxScore = max;
        }
    }
    /**
     * Get the current move count
     */
    getMoves() {
        const scoring = this.getScoringData();
        return scoring?.moves ?? 0;
    }
    /**
     * Increment the move counter
     */
    incrementMoves() {
        const scoring = this.getScoringData();
        if (!scoring)
            return;
        scoring.moves = (scoring.moves || 0) + 1;
    }
    /**
     * Get rank message ID based on current score
     */
    getRankMessageId() {
        const score = this.getScore();
        const ranks = this.config.ranks ?? exports.DEFAULT_RANKS;
        let currentRank = ranks[0]?.messageId ?? 'if.rank.beginner';
        for (const rank of ranks) {
            if (score >= rank.threshold) {
                currentRank = rank.messageId;
            }
        }
        return currentRank;
    }
    /**
     * Legacy: Get rank name string (deprecated, use getRankMessageId)
     */
    getRank() {
        const score = this.getScore();
        const ranks = this.config.ranks ?? exports.DEFAULT_RANKS;
        let currentRank = ranks[0]?.name ?? 'Beginner';
        for (const rank of ranks) {
            if (score >= rank.threshold) {
                currentRank = rank.name ?? rank.messageId;
            }
        }
        return currentRank;
    }
    /**
     * Check if a source has already been scored
     */
    hasScored(sourceId) {
        const scoring = this.getScoringData();
        return scoring?.scoredSources?.includes(sourceId) ?? false;
    }
    /**
     * Process a score gain by sourceId
     * Returns false if already scored or no definition found
     */
    scorePoints(sourceId) {
        if (!this.enabled)
            return false;
        // Check for duplicate scoring
        if (this.hasScored(sourceId)) {
            return false;
        }
        // Look up definition
        const definition = this.definitions.get(sourceId);
        if (!definition) {
            // No definition found - log warning but don't fail
            console.warn(`ScoringService: No definition found for sourceId "${sourceId}"`);
            return false;
        }
        const scoring = this.getScoringData();
        if (!scoring)
            return false;
        // Mark as scored
        if (!scoring.scoredSources) {
            scoring.scoredSources = [];
        }
        scoring.scoredSources.push(sourceId);
        // Add points
        scoring.scoreValue = (scoring.scoreValue || 0) + definition.points;
        // Add to history
        if (!scoring.scoreHistory) {
            scoring.scoreHistory = [];
        }
        scoring.scoreHistory.push({
            points: definition.points,
            sourceId,
            reasonMessageId: definition.reasonMessageId,
            timestamp: Date.now(),
        });
        return true;
    }
    /**
     * Process a score loss by sourceId (or override with explicit points)
     */
    losePoints(sourceId, overridePoints) {
        if (!this.enabled)
            return;
        const scoring = this.getScoringData();
        if (!scoring)
            return;
        // Determine points to lose
        let points = overridePoints;
        let reasonMessageId = 'if.score.lost';
        if (points === undefined) {
            const definition = this.definitions.get(sourceId);
            if (definition) {
                points = definition.points;
                reasonMessageId = definition.reasonMessageId;
            }
            else {
                console.warn(`ScoringService: No definition found for sourceId "${sourceId}" and no overridePoints provided`);
                return;
            }
        }
        // Subtract points (can go negative)
        scoring.scoreValue = (scoring.scoreValue || 0) - points;
        // Add to history (negative points)
        if (!scoring.scoreHistory) {
            scoring.scoreHistory = [];
        }
        scoring.scoreHistory.push({
            points: -points,
            sourceId,
            reasonMessageId,
            timestamp: Date.now(),
        });
    }
    /**
     * Get scoring definition for a sourceId
     */
    getDefinition(sourceId) {
        return this.definitions.get(sourceId);
    }
    /**
     * Register or update a scoring definition
     */
    registerDefinition(sourceId, definition) {
        this.definitions.set(sourceId, definition);
    }
    /**
     * Get all scored source IDs (for save/restore)
     */
    getScoredSources() {
        const scoring = this.getScoringData();
        return scoring?.scoredSources ?? [];
    }
    /**
     * Restore scored sources (for save/restore)
     */
    restoreScoredSources(sources) {
        const scoring = this.getScoringData();
        if (scoring) {
            scoring.scoredSources = [...sources];
        }
    }
    /**
     * Get the score history
     */
    getHistory() {
        const scoring = this.getScoringData();
        return scoring?.scoreHistory ?? [];
    }
    /**
     * Legacy: Add points directly (deprecated, use scorePoints with sourceId)
     */
    addPoints(points, reason) {
        const scoring = this.getScoringData();
        if (!scoring)
            return;
        scoring.scoreValue = (scoring.scoreValue || 0) + points;
        // Add to history if reason provided
        if (reason) {
            if (!scoring.scoreHistory) {
                scoring.scoreHistory = [];
            }
            scoring.scoreHistory.push({
                points,
                sourceId: 'legacy',
                reasonMessageId: reason,
                timestamp: Date.now(),
            });
        }
    }
    /**
     * Get the scoring capability data
     */
    getScoringData() {
        return this.world.getCapability(world_model_1.StandardCapabilities.SCORING) ?? null;
    }
}
exports.ScoringService = ScoringService;
