// @sharpee/text-service/handlers/game.js
"use strict";
/**
 * Game Event Handler
 *
 * Handles game lifecycle events like game.started to produce
 * text output (e.g., opening banner).
 *
 * @see ADR-096 Text Service Architecture
 * @see ADR-097 IGameEvent Deprecation (now uses ISemanticEvent with typed data)
 * @see ISSUE-028 Opening banner hardcoded in browser-entry.ts
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.handleGameStarted = handleGameStarted;
const assemble_js_1 = require("../stages/assemble.js");
/**
 * Handle game.started event to produce the opening banner.
 *
 * The event data (GameStartedData) contains:
 * - story.title: Game title
 * - story.author: Author name(s)
 * - story.version: Game version
 * - story.id: Story identifier
 * - gameState: 'running'
 * - session: { startTime, turns, moves }
 *
 * The language provider supplies the banner template via 'game.started.banner'.
 * Stories can customize by overriding this message in extendLanguage().
 */
function handleGameStarted(event, context) {
    // Event data is now directly in `data` field (ISemanticEvent standard)
    const data = event.data;
    const story = data?.story;
    if (!story) {
        return [];
    }
    // Build params for template substitution
    // engineVersion and clientVersion come from event data (set by engine from world's versionInfo)
    const engineVersion = data?.engineVersion || 'unknown';
    const clientVersion = data?.clientVersion || 'N/A';
    // Format buildDate for display (show date only, not full ISO timestamp)
    const buildDate = story.buildDate
        ? new Date(story.buildDate).toISOString().split('T')[0]
        : '';
    const params = {
        title: story.title || 'Unknown',
        author: story.author || 'Unknown',
        version: story.version || '1.0.0',
        id: story.id || 'unknown',
        engineVersion: engineVersion,
        clientVersion: clientVersion,
        buildDate: buildDate,
    };
    // Look up banner message via language provider
    const message = context.languageProvider?.getMessage('game.started.banner', params);
    // If no template registered or returned the ID (not found), skip output
    if (!message || message === 'game.started.banner') {
        return [];
    }
    return [(0, assemble_js_1.createBlock)('game.banner', message)];
}
