// @sharpee/if-domain/grammar/grammar-builder.js
"use strict";
/**
 * @file Grammar Builder Interfaces
 * @description Language-agnostic interfaces for defining grammar rules
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SlotType = void 0;
/**
 * Slot types for grammar patterns
 * Controls how the parser handles slot matching
 */
var SlotType;
(function (SlotType) {
    /** Default: resolve slot text to an entity via vocabulary lookup */
    SlotType["ENTITY"] = "entity";
    /** Capture raw text without entity resolution (single token) */
    SlotType["TEXT"] = "text";
    /** Capture raw text until next pattern element or end (greedy) */
    SlotType["TEXT_GREEDY"] = "text_greedy";
    /** Resolve entity but mark as instrument for the action */
    SlotType["INSTRUMENT"] = "instrument";
    // ADR-082: Built-in typed value slots
    /** Match cardinal/ordinal direction (n, s, e, w, ne, up, down, etc.) */
    SlotType["DIRECTION"] = "direction";
    /** Match integer (digits or words: 1, 29, one, twenty) */
    SlotType["NUMBER"] = "number";
    /** Match ordinal (1st, first, 2nd, second, etc.) */
    SlotType["ORDINAL"] = "ordinal";
    /** Match time expression (10:40, 6:00) */
    SlotType["TIME"] = "time";
    /** Match manner adverb (carefully, quickly, forcefully) - feeds intention.manner */
    SlotType["MANNER"] = "manner";
    // ADR-082: Text variant slots
    /** Match text enclosed in double quotes */
    SlotType["QUOTED_TEXT"] = "quoted_text";
    /** Match conversation topic (one or more words) */
    SlotType["TOPIC"] = "topic";
    // ADR-082: Category-based vocabulary slots
    /** Match word from a story-defined vocabulary category */
    SlotType["VOCABULARY"] = "vocabulary";
    // Deprecated: Use VOCABULARY with named categories instead
    /** @deprecated Use VOCABULARY with .fromVocabulary() instead */
    SlotType["ADJECTIVE"] = "adjective";
    /** @deprecated Use VOCABULARY with .fromVocabulary() instead */
    SlotType["NOUN"] = "noun";
})(SlotType || (exports.SlotType = SlotType = {}));
