// @sharpee/if-domain/grammar/grammar-engine.js
"use strict";
/**
 * @file Grammar Engine Base
 * @description Abstract base class for grammar matching engines
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.GrammarEngine = void 0;
const grammar_builder_1 = require("./grammar-builder");
/**
 * Abstract base class for grammar engines
 * Language-specific implementations provide concrete matching logic
 */
class GrammarEngine {
    rules = [];
    rulesByAction = new Map();
    compiler;
    constructor(compiler) {
        this.compiler = compiler;
    }
    /**
     * Add a grammar rule
     */
    addRule(rule) {
        // Compile the pattern if not already compiled
        if (!rule.compiledPattern) {
            rule.compiledPattern = this.compiler.compile(rule.pattern);
        }
        this.rules.push(rule);
        // Index by action for faster lookup
        const actionRules = this.rulesByAction.get(rule.action) || [];
        actionRules.push(rule);
        this.rulesByAction.set(rule.action, actionRules);
        // Sort by priority
        this.sortRules();
    }
    /**
     * Add multiple rules
     */
    addRules(rules) {
        rules.forEach(rule => this.addRule(rule));
    }
    /**
     * Get the best match from a set of tokens
     */
    getBestMatch(tokens, context, options) {
        const matches = this.findMatches(tokens, context, options);
        return matches.length > 0 ? matches[0] : null;
    }
    /**
     * Clear all rules
     */
    clear() {
        this.rules = [];
        this.rulesByAction.clear();
    }
    /**
     * Get all rules
     */
    getRules() {
        return [...this.rules];
    }
    /**
     * Get rules for a specific action
     */
    getRulesForAction(action) {
        return this.rulesByAction.get(action) || [];
    }
    /**
     * Sort rules by priority (descending)
     */
    sortRules() {
        this.rules.sort((a, b) => b.priority - a.priority);
        // Also sort within each action group
        this.rulesByAction.forEach((rules) => {
            rules.sort((a, b) => b.priority - a.priority);
        });
    }
    /**
     * Create a grammar builder connected to this engine
     */
    createBuilder() {
        const engine = this;
        return {
            define(pattern) {
                const rule = {
                    id: `rule_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`,
                    pattern,
                    slots: new Map(),
                    priority: 100 // Default priority
                };
                const builder = {
                    hasTrait(slot, traitType) {
                        const slotConstraint = rule.slots.get(slot) || { name: slot, constraints: [], traitFilters: [] };
                        if (!slotConstraint.traitFilters)
                            slotConstraint.traitFilters = [];
                        slotConstraint.traitFilters.push(traitType);
                        rule.slots.set(slot, slotConstraint);
                        return builder;
                    },
                    where(slot, constraint) {
                        const slotConstraint = rule.slots.get(slot) || { name: slot, constraints: [] };
                        slotConstraint.constraints.push(constraint);
                        rule.slots.set(slot, slotConstraint);
                        return builder;
                    },
                    text(slot) {
                        const slotConstraint = rule.slots.get(slot) || { name: slot, constraints: [] };
                        slotConstraint.slotType = grammar_builder_1.SlotType.TEXT;
                        rule.slots.set(slot, slotConstraint);
                        return builder;
                    },
                    instrument(slot) {
                        const slotConstraint = rule.slots.get(slot) || { name: slot, constraints: [] };
                        slotConstraint.slotType = grammar_builder_1.SlotType.INSTRUMENT;
                        rule.slots.set(slot, slotConstraint);
                        return builder;
                    },
                    mapsTo(action) {
                        rule.action = action;
                        return builder;
                    },
                    withPriority(priority) {
                        rule.priority = priority;
                        return builder;
                    },
                    withSemanticVerbs(verbs) {
                        if (!rule.semantics)
                            rule.semantics = {};
                        rule.semantics.verbs = verbs;
                        return builder;
                    },
                    withSemanticPrepositions(prepositions) {
                        if (!rule.semantics)
                            rule.semantics = {};
                        rule.semantics.prepositions = prepositions;
                        return builder;
                    },
                    withSemanticDirections(directions) {
                        if (!rule.semantics)
                            rule.semantics = {};
                        rule.semantics.directions = directions;
                        return builder;
                    },
                    withDefaultSemantics(defaults) {
                        rule.defaultSemantics = defaults;
                        return builder;
                    },
                    // ADR-082: Typed Value Slots
                    number(slot) {
                        const slotConstraint = rule.slots.get(slot) || { name: slot, constraints: [] };
                        slotConstraint.slotType = grammar_builder_1.SlotType.NUMBER;
                        rule.slots.set(slot, slotConstraint);
                        return builder;
                    },
                    ordinal(slot) {
                        const slotConstraint = rule.slots.get(slot) || { name: slot, constraints: [] };
                        slotConstraint.slotType = grammar_builder_1.SlotType.ORDINAL;
                        rule.slots.set(slot, slotConstraint);
                        return builder;
                    },
                    time(slot) {
                        const slotConstraint = rule.slots.get(slot) || { name: slot, constraints: [] };
                        slotConstraint.slotType = grammar_builder_1.SlotType.TIME;
                        rule.slots.set(slot, slotConstraint);
                        return builder;
                    },
                    // ADR-082: Built-in Vocabulary Slots
                    direction(slot) {
                        const slotConstraint = rule.slots.get(slot) || { name: slot, constraints: [] };
                        slotConstraint.slotType = grammar_builder_1.SlotType.DIRECTION;
                        rule.slots.set(slot, slotConstraint);
                        return builder;
                    },
                    manner(slot) {
                        const slotConstraint = rule.slots.get(slot) || { name: slot, constraints: [] };
                        slotConstraint.slotType = grammar_builder_1.SlotType.MANNER;
                        rule.slots.set(slot, slotConstraint);
                        return builder;
                    },
                    // ADR-082: Category-Based Vocabulary Slots
                    fromVocabulary(slot, category) {
                        const slotConstraint = rule.slots.get(slot) || { name: slot, constraints: [] };
                        slotConstraint.slotType = grammar_builder_1.SlotType.VOCABULARY;
                        slotConstraint.vocabularyCategory = category;
                        rule.slots.set(slot, slotConstraint);
                        return builder;
                    },
                    // Deprecated methods - use fromVocabulary() instead
                    adjective(slot) {
                        const slotConstraint = rule.slots.get(slot) || { name: slot, constraints: [] };
                        slotConstraint.slotType = grammar_builder_1.SlotType.ADJECTIVE;
                        rule.slots.set(slot, slotConstraint);
                        return builder;
                    },
                    noun(slot) {
                        const slotConstraint = rule.slots.get(slot) || { name: slot, constraints: [] };
                        slotConstraint.slotType = grammar_builder_1.SlotType.NOUN;
                        rule.slots.set(slot, slotConstraint);
                        return builder;
                    },
                    // ADR-082: Text Variant Slots
                    quotedText(slot) {
                        const slotConstraint = rule.slots.get(slot) || { name: slot, constraints: [] };
                        slotConstraint.slotType = grammar_builder_1.SlotType.QUOTED_TEXT;
                        rule.slots.set(slot, slotConstraint);
                        return builder;
                    },
                    topic(slot) {
                        const slotConstraint = rule.slots.get(slot) || { name: slot, constraints: [] };
                        slotConstraint.slotType = grammar_builder_1.SlotType.TOPIC;
                        rule.slots.set(slot, slotConstraint);
                        return builder;
                    },
                    build() {
                        if (!rule.action) {
                            throw new Error('Grammar rule must have an action (use .mapsTo())');
                        }
                        engine.addRule(rule);
                        return rule;
                    }
                };
                return builder;
            },
            forAction(actionId) {
                // Accumulated configuration
                const verbList = [];
                const patternList = [];
                const directionMap = {};
                const slotConstraints = new Map();
                const slotTraitFilters = new Map();
                const slotTypes = new Map();
                let priority = 100;
                let defaultSemantics;
                const actionBuilder = {
                    verbs(verbs) {
                        verbList.push(...verbs);
                        return actionBuilder;
                    },
                    pattern(pattern) {
                        patternList.push(pattern);
                        return actionBuilder;
                    },
                    patterns(patterns) {
                        patternList.push(...patterns);
                        return actionBuilder;
                    },
                    directions(map) {
                        Object.assign(directionMap, map);
                        return actionBuilder;
                    },
                    hasTrait(slot, traitType) {
                        const existing = slotTraitFilters.get(slot) || [];
                        existing.push(traitType);
                        slotTraitFilters.set(slot, existing);
                        return actionBuilder;
                    },
                    where(slot, constraint) {
                        const existing = slotConstraints.get(slot) || [];
                        existing.push({ constraint });
                        slotConstraints.set(slot, existing);
                        return actionBuilder;
                    },
                    withPriority(p) {
                        priority = p;
                        return actionBuilder;
                    },
                    withDefaultSemantics(defaults) {
                        defaultSemantics = defaults;
                        return actionBuilder;
                    },
                    slotType(slot, type) {
                        slotTypes.set(slot, type);
                        return actionBuilder;
                    },
                    build() {
                        // Generate verb × pattern combinations
                        if (verbList.length > 0 && patternList.length > 0) {
                            for (const verb of verbList) {
                                for (const patternTemplate of patternList) {
                                    // Combine verb with pattern template
                                    const fullPattern = patternTemplate.trim()
                                        ? `${verb} ${patternTemplate}`
                                        : verb;
                                    // Create the rule using the existing define() API
                                    let builder = engine.createBuilder().define(fullPattern);
                                    builder = builder.mapsTo(actionId);
                                    builder = builder.withPriority(priority);
                                    if (defaultSemantics) {
                                        builder = builder.withDefaultSemantics(defaultSemantics);
                                    }
                                    // Apply slot constraints
                                    for (const [slot, constraints] of slotConstraints) {
                                        for (const { constraint } of constraints) {
                                            builder = builder.where(slot, constraint);
                                        }
                                    }
                                    // Apply trait filters
                                    for (const [slot, traits] of slotTraitFilters) {
                                        for (const traitType of traits) {
                                            builder = builder.hasTrait(slot, traitType);
                                        }
                                    }
                                    // Apply slot types
                                    for (const [slot, type] of slotTypes) {
                                        switch (type) {
                                            case grammar_builder_1.SlotType.TEXT:
                                                builder = builder.text(slot);
                                                break;
                                            case grammar_builder_1.SlotType.INSTRUMENT:
                                                builder = builder.instrument(slot);
                                                break;
                                            case grammar_builder_1.SlotType.NUMBER:
                                                builder = builder.number(slot);
                                                break;
                                            case grammar_builder_1.SlotType.ORDINAL:
                                                builder = builder.ordinal(slot);
                                                break;
                                            case grammar_builder_1.SlotType.TIME:
                                                builder = builder.time(slot);
                                                break;
                                            case grammar_builder_1.SlotType.DIRECTION:
                                                builder = builder.direction(slot);
                                                break;
                                            case grammar_builder_1.SlotType.MANNER:
                                                builder = builder.manner(slot);
                                                break;
                                        }
                                    }
                                    builder.build();
                                }
                            }
                        }
                        // Generate standalone verb patterns (verb only, no template)
                        if (verbList.length > 0 && patternList.length === 0) {
                            for (const verb of verbList) {
                                let builder = engine.createBuilder().define(verb);
                                builder = builder.mapsTo(actionId);
                                builder = builder.withPriority(priority);
                                if (defaultSemantics) {
                                    builder = builder.withDefaultSemantics(defaultSemantics);
                                }
                                builder.build();
                            }
                        }
                        // Generate direction patterns
                        for (const [canonical, aliases] of Object.entries(directionMap)) {
                            for (const alias of aliases) {
                                let builder = engine.createBuilder().define(alias);
                                builder = builder.mapsTo(actionId);
                                // Lower priority for abbreviations (single character)
                                builder = builder.withPriority(alias.length === 1 ? 90 : priority);
                                builder = builder.withDefaultSemantics({ direction: canonical });
                                builder.build();
                            }
                        }
                    }
                };
                return actionBuilder;
            },
            getRules() {
                return engine.getRules();
            },
            clear() {
                engine.clear();
            }
        };
    }
}
exports.GrammarEngine = GrammarEngine;
