// @sharpee/if-domain/grammar/scope-builder.js
"use strict";
/**
 * @file Scope Builder Implementation
 * @description Concrete implementation of the scope builder interface
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ScopeBuilderImpl = void 0;
exports.scope = scope;
/**
 * Concrete scope builder implementation
 */
class ScopeBuilderImpl {
    constraint = {
        base: 'all',
        filters: [],
        traitFilters: [],
        explicitEntities: [],
        includeRules: []
    };
    visible() {
        this.constraint.base = 'visible';
        return this;
    }
    touchable() {
        this.constraint.base = 'touchable';
        return this;
    }
    carried() {
        this.constraint.base = 'carried';
        return this;
    }
    nearby() {
        this.constraint.base = 'nearby';
        return this;
    }
    matching(constraint) {
        this.constraint.filters.push(constraint);
        return this;
    }
    kind(kind) {
        this.constraint.filters.push({ kind });
        return this;
    }
    orExplicitly(entityIds) {
        this.constraint.explicitEntities.push(...entityIds);
        return this;
    }
    orRule(ruleId) {
        this.constraint.includeRules.push(ruleId);
        return this;
    }
    hasTrait(traitType) {
        this.constraint.traitFilters.push(traitType);
        return this;
    }
    build() {
        return {
            ...this.constraint,
            filters: [...this.constraint.filters],
            traitFilters: [...this.constraint.traitFilters],
            explicitEntities: [...this.constraint.explicitEntities],
            includeRules: [...this.constraint.includeRules]
        };
    }
}
exports.ScopeBuilderImpl = ScopeBuilderImpl;
/**
 * Create a new scope builder
 */
function scope() {
    return new ScopeBuilderImpl();
}
