// @sharpee/if-domain/grammar/vocabulary-provider.js
"use strict";
/**
 * @file Grammar Vocabulary Provider
 * @description Context-aware vocabulary system for grammar pattern matching (ADR-082)
 *
 * This is distinct from vocabulary-contracts/VocabularyProvider which is for
 * providing vocabulary entries to VocabularyRegistry. This system is specifically
 * for context-aware vocabulary categories used in grammar slot matching.
 *
 * Vocabulary categories are registered with optional context predicates.
 * The parser only considers a pattern if its vocabulary is active in the current context.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.GrammarVocabularyProvider = void 0;
/**
 * Default implementation of IGrammarVocabularyProvider.
 *
 * Stores vocabulary categories with their words and context predicates.
 * All word matching is case-insensitive.
 */
class GrammarVocabularyProvider {
    categories = new Map();
    define(category, config) {
        if (this.categories.has(category)) {
            throw new Error(`Vocabulary category '${category}' already exists. Use extend() to add words.`);
        }
        // Normalize words to lowercase
        const normalizedWords = new Set(config.words.map(w => w.toLowerCase()));
        this.categories.set(category, {
            words: normalizedWords,
            when: config.when
        });
    }
    extend(category, words) {
        const existing = this.categories.get(category);
        if (!existing) {
            throw new Error(`Vocabulary category '${category}' does not exist. Use define() first.`);
        }
        // Add normalized words
        for (const word of words) {
            existing.words.add(word.toLowerCase());
        }
    }
    match(category, word, ctx) {
        const config = this.categories.get(category);
        if (!config) {
            return false;
        }
        // Check context predicate
        if (config.when && !config.when(ctx)) {
            return false;
        }
        // Check word (case-insensitive)
        return config.words.has(word.toLowerCase());
    }
    getWords(category) {
        const config = this.categories.get(category);
        if (!config) {
            return new Set();
        }
        // Return a copy to prevent external modification
        return new Set(config.words);
    }
    isActive(category, ctx) {
        const config = this.categories.get(category);
        if (!config) {
            return false;
        }
        // No predicate means always active
        if (!config.when) {
            return true;
        }
        return config.when(ctx);
    }
    hasCategory(category) {
        return this.categories.has(category);
    }
    getCategories() {
        return Array.from(this.categories.keys());
    }
    removeCategory(category) {
        return this.categories.delete(category);
    }
    clear() {
        this.categories.clear();
    }
}
exports.GrammarVocabularyProvider = GrammarVocabularyProvider;
