// @sharpee/character/apply.js
"use strict";
/**
 * Apply compiled character data to an entity (ADR-141, 144, 145, 146)
 *
 * Convenience function that creates a CharacterModelTrait from compiled
 * builder output, registers custom predicates, and adds the trait to
 * the entity. Returns the trait plus any compiled configuration
 * (propagation, goals, movement, influence) for the NPC service.
 *
 * Public interface: applyCharacter, AppliedCharacter.
 * Owner context: @sharpee/character
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.applyCharacter = applyCharacter;
const world_model_1 = require("../world-model/index.js");
/**
 * Apply a compiled character to an entity.
 *
 * Creates the CharacterModelTrait, registers custom predicates,
 * and adds the trait to the entity. Returns the trait plus any
 * compiled behavior configuration for the NPC service.
 *
 * @param entity - The NPC entity to apply the character model to
 * @param compiled - Output of CharacterBuilder.compile()
 * @returns The trait and compiled behavior configuration
 */
function applyCharacter(entity, compiled) {
    const trait = new world_model_1.CharacterModelTrait(compiled.traitData);
    // Register custom predicates
    for (const [name, fn] of compiled.customPredicates) {
        trait.registerPredicate(name, fn);
    }
    // Add trait to entity
    entity.add(trait);
    return {
        trait,
        propagationProfile: compiled.propagationProfile,
        goalDefs: compiled.goalDefs,
        movementProfile: compiled.movementProfile,
        influenceDefs: compiled.influenceDefs,
        resistanceDefs: compiled.resistanceDefs,
    };
}
