// @sharpee/world-model/traits/switchable/switchableTrait.js
"use strict";
// packages/world-model/src/traits/switchable/switchableTrait.ts
Object.defineProperty(exports, "__esModule", { value: true });
exports.SwitchableTrait = void 0;
const trait_types_1 = require("../trait-types");
/**
 * Switchable trait for entities that can be turned on and off.
 * Used for lights, machines, devices, etc.
 *
 * This trait contains only data - all switching logic
 * is in SwitchableBehavior.
 */
class SwitchableTrait {
    static type = trait_types_1.TraitType.SWITCHABLE;
    type = trait_types_1.TraitType.SWITCHABLE;
    // SwitchableData properties
    isOn;
    startsOn;
    powerConsumption;
    requiresPower;
    hasPower;
    onMessage;
    offMessage;
    alreadyOnMessage;
    alreadyOffMessage;
    noPowerMessage;
    onSound;
    offSound;
    runningSound;
    autoOffTime;
    autoOffCounter;
    onDescription;
    offDescription;
    constructor(data = {}) {
        // Set defaults and merge with provided data
        this.startsOn = data.startsOn ?? false;
        this.requiresPower = data.requiresPower ?? false;
        this.hasPower = data.hasPower ?? true;
        this.isOn = data.isOn ?? (this.startsOn && this.hasPower);
        this.powerConsumption = data.powerConsumption ?? 1;
        this.onMessage = data.onMessage;
        this.offMessage = data.offMessage;
        this.alreadyOnMessage = data.alreadyOnMessage;
        this.alreadyOffMessage = data.alreadyOffMessage;
        this.noPowerMessage = data.noPowerMessage;
        this.onSound = data.onSound;
        this.offSound = data.offSound;
        this.runningSound = data.runningSound;
        this.autoOffTime = data.autoOffTime ?? 0;
        this.autoOffCounter = data.autoOffCounter ?? (this.isOn && this.autoOffTime > 0 ? this.autoOffTime : 0);
        this.onDescription = data.onDescription;
        this.offDescription = data.offDescription;
    }
}
exports.SwitchableTrait = SwitchableTrait;
