// @sharpee/stdlib/events/event-utils.js
"use strict";
/**
 * Utilities for creating standardized event data
 *
 * These are helpers that actions can use if they want to follow
 * common patterns, but they're entirely optional.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.createTargetData = createTargetData;
exports.createObjectData = createObjectData;
exports.createContainerData = createContainerData;
exports.createLocationData = createLocationData;
exports.entitiesToIds = entitiesToIds;
exports.entitiesToNames = entitiesToNames;
exports.createMovementData = createMovementData;
/**
 * Create standard target data from an entity
 */
function createTargetData(entity) {
    return {
        targetId: entity.id,
        targetName: entity.name
    };
}
/**
 * Create standard object data from an entity
 */
function createObjectData(entity) {
    return {
        targetId: entity.id,
        targetName: entity.name,
        objectId: entity.id,
        objectName: entity.name
    };
}
/**
 * Create standard container data from an entity
 */
function createContainerData(entity) {
    return {
        containerId: entity.id,
        containerName: entity.name
    };
}
/**
 * Create standard location data from an entity
 */
function createLocationData(entity) {
    return {
        locationId: entity.id,
        locationName: entity.name
    };
}
/**
 * Helper to convert entity array to ID array
 */
function entitiesToIds(entities) {
    return entities.map(e => e.id);
}
/**
 * Helper to convert entity array to name array
 */
function entitiesToNames(entities) {
    return entities.map(e => e.name);
}
/**
 * Create movement data from two locations
 */
function createMovementData(from, to) {
    return {
        fromLocationId: from.id,
        fromLocationName: from.name,
        toLocationId: to.id,
        toLocationName: to.name
    };
}
