// @sharpee/stdlib/events/helpers.js
"use strict";
/**
 * Helper utilities for common event handler patterns
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.createToggleHandler = createToggleHandler;
exports.createOpenHandler = createOpenHandler;
exports.createRevealHandler = createRevealHandler;
exports.createMessageHandler = createMessageHandler;
exports.composeHandlers = composeHandlers;
exports.createOnceHandler = createOnceHandler;
exports.createAfterHandler = createAfterHandler;
exports.createConditionalHandler = createConditionalHandler;
/**
 * Create a handler that toggles a switchable entity
 */
function createToggleHandler(entity) {
    return (event, _world) => {
        // Toggle the switch state
        const switchable = entity.get('SWITCHABLE');
        if (switchable && 'isOn' in switchable) {
            switchable.isOn = !switchable.isOn;
            // Return event indicating the state change
            return [{
                    id: `${Date.now()}-toggled`,
                    type: 'if.event.toggled',
                    timestamp: Date.now(),
                    data: {
                        entity: entity.id,
                        newState: switchable.isOn
                    },
                    entities: { [entity.id]: entity }
                }];
        }
        return undefined;
    };
}
/**
 * Create a handler that opens something when triggered
 */
function createOpenHandler(entity, target) {
    return (event, _world) => {
        // Return event to open the target
        return [{
                id: `${Date.now()}-open-triggered`,
                type: 'if.event.open_triggered',
                timestamp: Date.now(),
                data: {
                    trigger: entity.id,
                    target: target
                },
                entities: { [entity.id]: entity }
            }];
    };
}
/**
 * Create a handler that reveals a hidden passage
 */
function createRevealHandler(entity, passage) {
    return (event, _world) => {
        // Return event to reveal the passage
        return [{
                id: `${Date.now()}-passage-revealed`,
                type: 'if.event.passage_revealed',
                timestamp: Date.now(),
                data: {
                    trigger: entity.id,
                    passage: passage
                },
                entities: { [entity.id]: entity }
            }];
    };
}
/**
 * Create a handler that displays a message
 */
function createMessageHandler(message, params) {
    return (event, _world) => {
        return [{
                id: `${Date.now()}-custom-message`,
                type: 'action.message',
                timestamp: Date.now(),
                data: {
                    messageId: 'custom',
                    message: message,
                    params: params || {}
                },
                entities: {}
            }];
    };
}
/**
 * Compose multiple handlers into one
 */
function composeHandlers(...handlers) {
    return (event, world) => {
        const results = [];
        for (const handler of handlers) {
            const result = handler(event, world);
            if (result && Array.isArray(result)) {
                results.push(...result);
            }
        }
        return results.length > 0 ? results : undefined;
    };
}
/**
 * Create a handler that only fires once
 */
function createOnceHandler(handler) {
    let fired = false;
    return (event, world) => {
        if (!fired) {
            fired = true;
            return handler(event, world);
        }
    };
}
/**
 * Create a handler that only fires after N times
 */
function createAfterHandler(count, handler) {
    let callCount = 0;
    return (event, world) => {
        callCount++;
        if (callCount >= count) {
            return handler(event, world);
        }
    };
}
/**
 * Create a handler with a condition
 */
function createConditionalHandler(condition, handler) {
    return (event, world) => {
        if (condition(event)) {
            return handler(event, world);
        }
    };
}
