// @sharpee/stdlib/actions/standard/giving/giving.js
"use strict";
/**
 * Giving action - give objects to NPCs or other actors
 *
 * This action handles transferring objects from the player to NPCs.
 * NPCs may accept or refuse items based on their state.
 *
 * Uses four-phase pattern:
 * 1. validate: Check if item can be given to recipient
 * 2. execute: Transfer item, store result in sharedData
 * 3. blocked: Generate events when validation fails
 * 4. report: Generate success events from sharedData
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.givingAction = void 0;
const types_1 = require("../../../scope/types");
const world_model_1 = require("../../../../world-model/index.js");
const constants_1 = require("../../constants");
function getGivingSharedData(context) {
    return context.sharedData;
}
function effectsToEvents(effects, context) {
    return effects.map(effect => context.event(effect.type, effect.payload));
}
exports.givingAction = {
    id: constants_1.IFActions.GIVING,
    // Default scope requirements for this action's slots
    defaultScope: {
        item: types_1.ScopeLevel.REACHABLE, // REACHABLE allows implicit take
        recipient: types_1.ScopeLevel.REACHABLE
    },
    requiredMessages: [
        'no_item',
        'no_recipient',
        'not_holding',
        'recipient_not_visible',
        'recipient_not_reachable',
        'not_actor',
        'self',
        'inventory_full',
        'too_heavy',
        'not_interested',
        'refuses',
        'given',
        'accepts',
        'gratefully_accepts',
        'reluctantly_accepts'
    ],
    metadata: {
        requiresDirectObject: true,
        requiresIndirectObject: true,
        directObjectScope: types_1.ScopeLevel.REACHABLE, // REACHABLE allows implicit take
        indirectObjectScope: types_1.ScopeLevel.REACHABLE
    },
    validate(context) {
        const actor = context.player;
        const item = context.command.directObject?.entity;
        const recipient = context.command.indirectObject?.entity;
        // Validate we have both item and recipient
        if (!item) {
            return {
                valid: false,
                error: 'no_item',
                params: {}
            };
        }
        if (!recipient) {
            return {
                valid: false,
                error: 'no_recipient',
                params: {}
            };
        }
        // Item must be carried (or implicitly takeable)
        // This enables "give apple to bob" when apple is on the ground
        const carryCheck = context.requireCarriedOrImplicitTake(item);
        if (!carryCheck.ok) {
            return carryCheck.error;
        }
        // Check if recipient is an actor (can receive items)
        if (!recipient.has(world_model_1.TraitType.ACTOR)) {
            return {
                valid: false,
                error: 'not_actor',
                params: { recipient: recipient.name }
            };
        }
        // Prevent giving to self
        if (recipient.id === actor.id) {
            return {
                valid: false,
                error: 'self',
                params: { item: item.name }
            };
        }
        // Check inventory capacity
        const recipientActor = recipient.getTrait(world_model_1.ActorTrait);
        if (recipientActor) {
            const limit = recipientActor.capacity ?? recipientActor['inventoryLimit'];
            if (limit) {
                const recipientInventory = context.world.getContents(recipient.id);
                // Check item count
                if (limit.maxItems !== undefined && recipientInventory.length >= limit.maxItems) {
                    return {
                        valid: false,
                        error: 'inventory_full',
                        params: { recipient: recipient.name }
                    };
                }
                // Check weight
                if (limit.maxWeight !== undefined) {
                    const currentWeight = recipientInventory.reduce((sum, e) => {
                        return sum + world_model_1.IdentityBehavior.getWeight(e);
                    }, 0);
                    const itemWeight = world_model_1.IdentityBehavior.getWeight(item);
                    if (currentWeight + itemWeight > limit.maxWeight) {
                        return {
                            valid: false,
                            error: 'too_heavy',
                            params: { item: item.name, recipient: recipient.name }
                        };
                    }
                }
            }
        }
        // Check for preferences (stored in actor's customProperties)
        const preferences = (recipientActor?.customProperties?.['preferences'] ?? recipientActor?.['preferences']);
        if (preferences) {
            const itemName = item.name.toLowerCase();
            if (preferences.refuses && Array.isArray(preferences.refuses)) {
                for (const refuse of preferences.refuses) {
                    if (itemName.includes(refuse.toLowerCase())) {
                        return {
                            valid: false,
                            error: 'not_interested',
                            params: { item: item.name, recipient: recipient.name }
                        };
                    }
                }
            }
        }
        return { valid: true };
    },
    execute(context) {
        const item = context.command.directObject.entity;
        const recipient = context.command.indirectObject.entity;
        const sharedData = getGivingSharedData(context);
        // Store basic info
        sharedData.itemId = item.id;
        sharedData.itemName = item.name;
        sharedData.recipientId = recipient.id;
        sharedData.recipientName = recipient.name;
        // Check if recipient has a capability behavior for giving (ADR-090)
        const capTrait = (0, world_model_1.findTraitWithCapability)(recipient, constants_1.IFActions.GIVING);
        if (capTrait) {
            const behavior = (0, world_model_1.getBehaviorForCapability)(capTrait, constants_1.IFActions.GIVING);
            if (behavior) {
                sharedData.capabilityBehavior = behavior;
                sharedData.capabilityTrait = capTrait;
                behavior.execute(recipient, context.world, context.player.id, context.sharedData);
                return;
            }
        }
        // Standard: transfer item from actor to recipient
        context.world.moveEntity(item.id, recipient.id);
        // Determine acceptance type based on preferences
        let acceptanceType = 'normal';
        const recipientActor = recipient.getTrait(world_model_1.ActorTrait);
        const preferences = (recipientActor?.customProperties?.['preferences'] ?? recipientActor?.['preferences']);
        if (preferences) {
            const itemName = item.name.toLowerCase();
            if (preferences.likes && Array.isArray(preferences.likes)) {
                for (const like of preferences.likes) {
                    if (itemName.includes(like.toLowerCase())) {
                        acceptanceType = 'grateful';
                        break;
                    }
                }
            }
            if (acceptanceType === 'normal' && preferences.dislikes && Array.isArray(preferences.dislikes)) {
                for (const dislike of preferences.dislikes) {
                    if (itemName.includes(dislike.toLowerCase())) {
                        acceptanceType = 'reluctant';
                        break;
                    }
                }
            }
        }
        sharedData.acceptanceType = acceptanceType;
        // Determine success message based on acceptance type
        switch (acceptanceType) {
            case 'grateful':
                sharedData.messageId = 'gratefully_accepts';
                break;
            case 'reluctant':
                sharedData.messageId = 'reluctantly_accepts';
                break;
            default:
                sharedData.messageId = 'given';
        }
        sharedData.params = {
            item: item.name,
            recipient: recipient.name
        };
    },
    blocked(context, result) {
        const item = context.command.directObject?.entity;
        const recipient = context.command.indirectObject?.entity;
        return [context.event('if.event.give_blocked', {
                blocked: true,
                messageId: `${context.action.id}.${result.error}`,
                params: {
                    ...result.params,
                    item: item?.name,
                    recipient: recipient?.name
                },
                reason: result.error,
                itemId: item?.id,
                itemName: item?.name,
                recipientId: recipient?.id,
                recipientName: recipient?.name
            })];
    },
    report(context) {
        const sharedData = getGivingSharedData(context);
        const events = [];
        // Prepend any implicit take events (from requireCarriedOrImplicitTake)
        if (context.sharedData.implicitTakeEvents) {
            events.push(...context.sharedData.implicitTakeEvents);
        }
        // If capability behavior handled the give, delegate report to it
        if (sharedData.capabilityBehavior) {
            const recipient = context.command.indirectObject?.entity;
            if (recipient) {
                const effects = sharedData.capabilityBehavior.report(recipient, context.world, context.player.id, context.sharedData);
                events.push(...effectsToEvents(effects, context));
            }
            return events;
        }
        // Standard: emit given event with messageId for text rendering
        events.push(context.event('if.event.given', {
            messageId: `${context.action.id}.${sharedData.messageId}`,
            params: sharedData.params,
            item: sharedData.itemId,
            itemName: sharedData.itemName,
            recipient: sharedData.recipientId,
            recipientName: sharedData.recipientName,
            accepted: true
        }));
        return events;
    },
    group: "social"
};
