// @sharpee/if-domain/vocabulary-contracts/vocabulary-types.js
"use strict";
/**
 * Vocabulary types and interfaces for the parser
 *
 * The vocabulary system is extensible at multiple levels:
 * - Base vocabulary from language packages
 * - Extension vocabulary from extensions
 * - Story-specific vocabulary overrides
 * - Entity-specific vocabulary from traits
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.GrammarPatterns = exports.PartOfSpeech = void 0;
/**
 * Part of speech categories
 */
exports.PartOfSpeech = {
    VERB: 'verb',
    NOUN: 'noun',
    ADJECTIVE: 'adjective',
    PREPOSITION: 'preposition',
    ARTICLE: 'article',
    PRONOUN: 'pronoun',
    DIRECTION: 'direction',
    SPECIAL: 'special',
    DETERMINER: 'determiner',
    CONJUNCTION: 'conjunction',
    NUMBER: 'number'
};
/**
 * Common grammar patterns
 */
exports.GrammarPatterns = Object.freeze({
    VERB_ONLY: Object.freeze({
        name: 'verb_only',
        elements: Object.freeze(['VERB']),
        example: 'look'
    }),
    VERB_NOUN: Object.freeze({
        name: 'verb_noun',
        elements: Object.freeze(['VERB', 'NOUN']),
        example: 'take sword'
    }),
    VERB_NOUN_PREP_NOUN: Object.freeze({
        name: 'verb_noun_prep_noun',
        elements: Object.freeze(['VERB', 'NOUN', 'PREP', 'NOUN']),
        example: 'put sword in chest'
    }),
    VERB_PREP_NOUN: Object.freeze({
        name: 'verb_prep_noun',
        elements: Object.freeze(['VERB', 'PREP', 'NOUN']),
        example: 'look at painting'
    }),
    VERB_DIRECTION: Object.freeze({
        name: 'verb_direction',
        elements: Object.freeze(['VERB', 'DIRECTION']),
        example: 'go north'
    }),
    DIRECTION_ONLY: Object.freeze({
        name: 'direction_only',
        elements: Object.freeze(['DIRECTION']),
        example: 'north'
    })
});
