// @sharpee/if-domain/vocabulary-contracts/vocabulary-adapters.js
"use strict";
/**
 * Vocabulary adapters for converting language-specific vocabulary
 * to the standard vocabulary types used by the parser
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.adaptVerbVocabulary = adaptVerbVocabulary;
exports.adaptDirectionVocabulary = adaptDirectionVocabulary;
exports.adaptSpecialVocabulary = adaptSpecialVocabulary;
/**
 * Adapt verb vocabulary from language provider format
 */
function adaptVerbVocabulary(languageProvider) {
    // If provider has direct getVerbs method, use it
    if (languageProvider.getVerbs) {
        return languageProvider.getVerbs();
    }
    // Otherwise try to build from mappings
    const verbs = [];
    const verbMappings = languageProvider.getVerbMappings?.() || {};
    for (const [actionId, verbList] of Object.entries(verbMappings)) {
        if (Array.isArray(verbList)) {
            verbs.push({
                actionId,
                verbs: verbList,
                pattern: languageProvider.getVerbPattern?.(actionId)
            });
        }
    }
    return verbs;
}
/**
 * Adapt direction vocabulary from language provider format
 */
function adaptDirectionVocabulary(languageProvider) {
    // If provider has direct getDirections method, use it
    if (languageProvider.getDirections) {
        return languageProvider.getDirections();
    }
    // Otherwise try to build from mappings
    const directions = [];
    const directionMappings = languageProvider.getDirectionMappings?.() || {};
    for (const [direction, words] of Object.entries(directionMappings)) {
        if (Array.isArray(words)) {
            // Separate full words from abbreviations (1-2 character words)
            const fullWords = words.filter(w => w.length > 2);
            const abbreviations = words.filter(w => w.length <= 2);
            directions.push({
                direction,
                words: fullWords.length > 0 ? fullWords : words,
                abbreviations: abbreviations.length > 0 ? abbreviations : undefined
            });
        }
    }
    return directions;
}
/**
 * Adapt special vocabulary from language provider format
 */
function adaptSpecialVocabulary(languageProvider) {
    // If provider has direct getSpecialVocabulary method, use it
    if (languageProvider.getSpecialVocabulary) {
        return languageProvider.getSpecialVocabulary();
    }
    // Otherwise build from individual methods
    return {
        pronouns: languageProvider.getPronouns?.() || ['it', 'them'],
        allWords: languageProvider.getAllWords?.() || ['all', 'everything'],
        exceptWords: languageProvider.getExceptWords?.() || ['except', 'but'],
        articles: languageProvider.getArticles?.() || ['a', 'an', 'the']
    };
}
