// @sharpee/stdlib/actions/standard/taking_off/taking-off.js
"use strict";
/**
 * Taking off action - remove worn clothing or equipment
 *
 * This action handles removing items that are currently worn.
 * It validates layering rules and provides appropriate feedback.
 *
 * Uses four-phase pattern:
 * 1. validate: Check if item is worn and can be removed
 * 2. execute: Call WearableBehavior.remove(), store result in sharedData
 * 3. blocked: Generate events when validation fails
 * 4. report: Generate success events from sharedData
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.takingOffAction = void 0;
const world_model_1 = require("../../../../world-model/index.js");
const constants_1 = require("../../constants");
const scope_1 = require("../../../scope");
const wearable_shared_1 = require("../wearable-shared");
function getTakingOffSharedData(context) {
    return context.sharedData;
}
exports.takingOffAction = {
    id: constants_1.IFActions.TAKING_OFF,
    // Default scope requirements for this action's slots
    defaultScope: {
        item: scope_1.ScopeLevel.CARRIED // Must be wearing the item (which implies possession)
    },
    requiredMessages: [
        'no_target',
        'not_wearing',
        'removed',
        'cant_remove',
        'prevents_removal'
    ],
    group: 'wearable_manipulation',
    validate(context) {
        const actor = context.player;
        const item = context.command.directObject?.entity;
        if (!item) {
            return { valid: false, error: 'no_target' };
        }
        if (!item.has(world_model_1.TraitType.WEARABLE)) {
            return { valid: false, error: 'not_wearing' };
        }
        if (!world_model_1.WearableBehavior.canRemove(item, actor)) {
            const wearable = item.get(world_model_1.TraitType.WEARABLE);
            if (!wearable.worn) {
                return { valid: false, error: 'not_wearing' };
            }
            if (wearable.wornBy !== actor.id) {
                return { valid: false, error: 'not_wearing' };
            }
            return { valid: false, error: 'cant_remove' };
        }
        return { valid: true };
    },
    execute(context) {
        const actor = context.player;
        const item = context.command.directObject.entity;
        const sharedData = getTakingOffSharedData(context);
        // Store basic info
        sharedData.itemId = item.id;
        sharedData.itemName = item.name;
        // Analyze the wearable context
        const wearableContext = (0, wearable_shared_1.analyzeWearableContext)(context, item, actor);
        const { wearableTrait } = wearableContext;
        // Store body part and layer
        sharedData.bodyPart = wearableTrait.bodyPart;
        sharedData.layer = wearableTrait.layer;
        // Check for layering conflicts using shared helper
        const blockingItem = (0, wearable_shared_1.checkRemovalBlockers)(wearableContext);
        if (blockingItem) {
            sharedData.failed = true;
            sharedData.errorMessageId = 'prevents_removal';
            sharedData.errorReason = 'prevents_removal';
            sharedData.errorParams = { blocking: blockingItem.name };
            return;
        }
        // Check if removing this would cause problems (e.g., cursed items)
        if ((0, wearable_shared_1.hasRemovalRestrictions)(wearableTrait)) {
            sharedData.failed = true;
            sharedData.errorMessageId = 'cant_remove';
            sharedData.errorReason = 'cant_remove';
            sharedData.errorParams = { item: item.name };
            return;
        }
        // Delegate state change to behavior
        const result = world_model_1.WearableBehavior.remove(item, actor);
        // Handle failure cases (defensive checks)
        if (!result.success) {
            sharedData.failed = true;
            if (result.notWorn) {
                sharedData.errorMessageId = 'not_wearing';
                sharedData.errorReason = 'not_wearing';
                sharedData.errorParams = { item: item.name };
            }
            else if (result.wornByOther) {
                sharedData.errorMessageId = 'not_wearing';
                sharedData.errorReason = 'worn_by_other';
                sharedData.errorParams = { item: item.name, wornBy: result.wornByOther };
            }
            else {
                sharedData.errorMessageId = 'cant_remove';
                sharedData.errorReason = 'cant_remove';
                sharedData.errorParams = { item: item.name };
            }
            return;
        }
        // Success
        sharedData.failed = false;
        sharedData.params = (0, wearable_shared_1.buildWearableEventParams)(item, wearableTrait);
        sharedData.messageId = 'removed';
    },
    /**
     * Generate events when validation fails
     *
     * Uses simplified event pattern (ADR-097): domain event carries messageId directly.
     */
    blocked(context, result) {
        const item = context.command.directObject?.entity;
        return [context.event('if.event.take_off_blocked', {
                // Rendering data
                messageId: `${context.action.id}.${result.error}`,
                params: result.params || {},
                // Domain data
                itemId: item?.id,
                itemName: item?.name,
                reason: result.error
            })];
    },
    /**
     * Report phase - generates events after successful execution
     *
     * Uses simplified event pattern (ADR-097): domain event carries messageId directly.
     */
    report(context) {
        const sharedData = getTakingOffSharedData(context);
        // Handle behavior failures (safety net - should be rare after validation)
        if (sharedData.failed) {
            return [context.event('if.event.take_off_blocked', {
                    messageId: `${context.action.id}.${sharedData.errorMessageId}`,
                    params: sharedData.errorParams || {},
                    itemId: sharedData.itemId,
                    itemName: sharedData.itemName,
                    reason: sharedData.errorReason
                })];
        }
        // Emit domain event with messageId (simplified pattern - ADR-097)
        return [
            context.event('if.event.removed', {
                // Rendering data (messageId + params for text-service)
                messageId: `${context.action.id}.${sharedData.messageId}`,
                params: sharedData.params,
                // Domain data (for event sourcing / handlers)
                itemId: sharedData.itemId,
                itemName: sharedData.itemName,
                actorId: context.player.id,
                bodyPart: sharedData.bodyPart,
                layer: sharedData.layer
            })
        ];
    },
    metadata: {
        requiresDirectObject: true,
        requiresIndirectObject: false,
        directObjectScope: scope_1.ScopeLevel.CARRIED
    }
};
