// @sharpee/core/constants/attributes.js
"use strict";
// packages/core/src/constants/attributes.ts
Object.defineProperty(exports, "__esModule", { value: true });
exports.CORE_ATTRIBUTE_TYPES = exports.CoreAttributeType = exports.CoreAttributes = void 0;
/**
 * Core entity attributes - generic attributes that any entity might have
 * Game-specific attributes should be defined in their respective packages
 */
var CoreAttributes;
(function (CoreAttributes) {
    // Identity attributes
    CoreAttributes["ID"] = "id";
    CoreAttributes["TYPE"] = "type";
    CoreAttributes["NAME"] = "name";
    // Metadata attributes
    CoreAttributes["CREATED_AT"] = "created_at";
    CoreAttributes["UPDATED_AT"] = "updated_at";
    CoreAttributes["VERSION"] = "version";
    // State attributes
    CoreAttributes["ACTIVE"] = "active";
    CoreAttributes["VISIBLE"] = "visible";
    // Extension attributes
    CoreAttributes["EXTENSIONS"] = "extensions";
    CoreAttributes["METADATA"] = "metadata";
})(CoreAttributes || (exports.CoreAttributes = CoreAttributes = {}));
/**
 * Core attribute types for validation
 */
var CoreAttributeType;
(function (CoreAttributeType) {
    CoreAttributeType["STRING"] = "string";
    CoreAttributeType["NUMBER"] = "number";
    CoreAttributeType["BOOLEAN"] = "boolean";
    CoreAttributeType["OBJECT"] = "object";
    CoreAttributeType["ARRAY"] = "array";
    CoreAttributeType["DATE"] = "date";
})(CoreAttributeType || (exports.CoreAttributeType = CoreAttributeType = {}));
/**
 * Core attribute definitions with their expected types
 */
exports.CORE_ATTRIBUTE_TYPES = {
    [CoreAttributes.ID]: CoreAttributeType.STRING,
    [CoreAttributes.TYPE]: CoreAttributeType.STRING,
    [CoreAttributes.NAME]: CoreAttributeType.STRING,
    [CoreAttributes.CREATED_AT]: CoreAttributeType.DATE,
    [CoreAttributes.UPDATED_AT]: CoreAttributeType.DATE,
    [CoreAttributes.VERSION]: CoreAttributeType.NUMBER,
    [CoreAttributes.ACTIVE]: CoreAttributeType.BOOLEAN,
    [CoreAttributes.VISIBLE]: CoreAttributeType.BOOLEAN,
    [CoreAttributes.EXTENSIONS]: CoreAttributeType.ARRAY,
    [CoreAttributes.METADATA]: CoreAttributeType.OBJECT
};
