// @sharpee/core/constants/index.js
"use strict";
/**
 * Core constants for Sharpee
 *
 * These are generic, system-level constants that don't assume any particular
 * type of narrative or game. IF-specific constants are in @sharpee/stdlib.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CoreConstants = void 0;
// Export generic constants only
__exportStar(require("./entity-types"), exports);
__exportStar(require("./relationships"), exports);
__exportStar(require("./attributes"), exports);
__exportStar(require("./core-events"), exports);
// Convenience export
const entity_types_1 = require("./entity-types");
const relationships_1 = require("./relationships");
const attributes_1 = require("./attributes");
const core_events_1 = require("./core-events");
exports.CoreConstants = {
    EntityTypes: entity_types_1.CoreEntityType,
    Relationships: relationships_1.CoreRelationshipType,
    Attributes: attributes_1.CoreAttributes,
    Events: core_events_1.CoreEvents,
    EventCategories: core_events_1.CoreEventCategory
};
