// @sharpee/core/events/event-factory.js
"use strict";
/**
 * Event Factory - Type-safe semantic event creation
 *
 * @see ADR-082 for design rationale
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.createTypedEvent = createTypedEvent;
exports.createMessageEvent = createMessageEvent;
exports.createEmptyEvent = createEmptyEvent;
exports.resetEventCounter = resetEventCounter;
let eventCounter = 0;
/**
 * Create a type-safe semantic event.
 *
 * The data parameter is strictly typed based on the event type,
 * providing compile-time verification of event data.
 *
 * @example
 * ```typescript
 * // TypeScript enforces correct data shape
 * const event = createTypedEvent('query.invalid', {
 *   message: 'Invalid input',
 *   hint: 'Please try again'
 * });
 *
 * // This would be a compile error:
 * // createTypedEvent('query.invalid', { wrongField: true });
 * ```
 */
function createTypedEvent(type, data, options = {}) {
    return {
        id: options.id ?? `evt-${Date.now()}-${++eventCounter}`,
        type,
        timestamp: Date.now(),
        data,
        entities: options.entities ?? {},
        tags: options.tags,
        priority: options.priority,
        narrate: options.narrate
    };
}
/**
 * Create a message event (convenience helper).
 *
 * @example
 * ```typescript
 * const event = createMessageEvent('success', 'item_taken', {
 *   item: 'brass lantern'
 * });
 * ```
 */
function createMessageEvent(variant, messageId, params, options) {
    const type = `message.${variant}`;
    return createTypedEvent(type, { messageId, params }, options);
}
/**
 * Create an empty event (for events with no data).
 *
 * @example
 * ```typescript
 * const event = createEmptyEvent('platform.quit_cancelled');
 * ```
 */
function createEmptyEvent(type, options) {
    return createTypedEvent(type, {}, options);
}
/**
 * Reset the event counter (useful for testing).
 */
function resetEventCounter() {
    eventCounter = 0;
}
