// @sharpee/core/events/event-source.js
"use strict";
/**
 * Generic event source infrastructure for the narrative engine.
 * This provides a simple pub/sub mechanism for any type of event.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SimpleEventSource = void 0;
exports.createEventSource = createEventSource;
/**
 * Simple synchronous implementation of EventSource
 */
class SimpleEventSource {
    handlers = [];
    emit(event) {
        // Create a copy to avoid issues if handlers modify the array
        const currentHandlers = [...this.handlers];
        currentHandlers.forEach(handler => {
            try {
                handler(event);
            }
            catch (error) {
                // Prevent one handler's error from affecting others
                console.error('Error in event handler:', error);
            }
        });
    }
    subscribe(handler) {
        this.handlers.push(handler);
        // Return unsubscribe function
        return () => {
            const index = this.handlers.indexOf(handler);
            if (index > -1) {
                this.handlers.splice(index, 1);
            }
        };
    }
    /**
     * Get the current number of subscribers
     */
    get subscriberCount() {
        return this.handlers.length;
    }
    /**
     * Remove all subscribers
     */
    clear() {
        this.handlers = [];
    }
}
exports.SimpleEventSource = SimpleEventSource;
/**
 * Create a new event source for a specific event type
 */
function createEventSource() {
    return new SimpleEventSource();
}
