// @sharpee/core/events/event-system.js
"use strict";
// packages/core/src/events/event-system.ts
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventSourceImpl = void 0;
exports.createEvent = createEvent;
exports.createEventSource = createEventSource;
const semantic_event_source_1 = require("./semantic-event-source");
/**
 * Create a new semantic event
 */
function createEvent(type, data, entities, metadata) {
    // Extract values from metadata
    const tags = metadata?.tags || [];
    const priority = metadata?.priority ?? 0;
    const narrate = metadata?.narrate ?? true;
    // Clean metadata for storage (remove known properties)
    const cleanMetadata = metadata ? { ...metadata } : undefined;
    if (cleanMetadata) {
        delete cleanMetadata.tags;
        delete cleanMetadata.priority;
        delete cleanMetadata.narrate;
    }
    const event = {
        id: generateEventId(),
        type,
        timestamp: Date.now(),
        entities: entities || {},
        data: data || {},
        tags,
        priority,
        narrate
    };
    // Add metadata if provided (excluding extracted properties)
    if (cleanMetadata && Object.keys(cleanMetadata).length > 0) {
        event.metadata = cleanMetadata;
    }
    return event;
}
// Event source implementation moved to semantic-event-source.ts
// Re-export for backwards compatibility
var semantic_event_source_2 = require("./semantic-event-source");
Object.defineProperty(exports, "EventSourceImpl", { enumerable: true, get: function () { return semantic_event_source_2.SemanticEventSourceImpl; } });
// Counter for ensuring unique IDs even with same timestamp
let eventCounter = 0;
/**
 * Generate a unique ID for an event
 */
function generateEventId() {
    // Use the expected format: evt_timestamp_random
    // Increment counter and reset if it gets too large
    eventCounter = (eventCounter + 1) % 10000;
    // Use both random and counter for better uniqueness
    const random = Math.floor(Math.random() * 1000);
    const unique = random + (eventCounter * 1000);
    return `evt_${Date.now()}_${unique}`;
}
/**
 * Create a new event source
 * @deprecated Use createSemanticEventSource from './semantic-event-source'
 */
function createEventSource() {
    return (0, semantic_event_source_1.createSemanticEventSource)();
}
