// @sharpee/core/events/game-events.js
"use strict";
/**
 * Game lifecycle events for tracking game state transitions
 * These events mark important milestones in a game session
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.GameEventType = void 0;
exports.isGameEvent = isGameEvent;
exports.isGameStartSequenceEvent = isGameStartSequenceEvent;
exports.isGameEndSequenceEvent = isGameEndSequenceEvent;
exports.createGameEvent = createGameEvent;
exports.createGameInitializingEvent = createGameInitializingEvent;
exports.createGameInitializedEvent = createGameInitializedEvent;
exports.createStoryLoadingEvent = createStoryLoadingEvent;
exports.createStoryLoadedEvent = createStoryLoadedEvent;
exports.createGameStartingEvent = createGameStartingEvent;
exports.createGameStartedEvent = createGameStartedEvent;
exports.createGameEndingEvent = createGameEndingEvent;
exports.createGameEndedEvent = createGameEndedEvent;
exports.createGameWonEvent = createGameWonEvent;
exports.createGameLostEvent = createGameLostEvent;
exports.createGameQuitEvent = createGameQuitEvent;
exports.createGameAbortedEvent = createGameAbortedEvent;
exports.createPcSwitchedEvent = createPcSwitchedEvent;
exports.isGameStartedEvent = isGameStartedEvent;
exports.isGameEndedEvent = isGameEndedEvent;
exports.isGameWonEvent = isGameWonEvent;
exports.isGameLostEvent = isGameLostEvent;
exports.isGameQuitEvent = isGameQuitEvent;
exports.isGameAbortedEvent = isGameAbortedEvent;
// =============================================================================
// Game Event Type Constants
// =============================================================================
/**
 * Game event types for lifecycle transitions
 */
exports.GameEventType = {
    // Game start events
    GAME_INITIALIZING: 'game.initializing', // Engine is setting up
    GAME_INITIALIZED: 'game.initialized', // Engine ready, before story
    STORY_LOADING: 'game.story_loading', // Story is being loaded
    STORY_LOADED: 'game.story_loaded', // Story loaded successfully
    GAME_STARTING: 'game.starting', // Game is about to start
    GAME_STARTED: 'game.started', // Game has started, ready for input
    // Game end events  
    GAME_ENDING: 'game.ending', // Game is about to end
    GAME_ENDED: 'game.ended', // Game has ended normally
    GAME_WON: 'game.won', // Player achieved victory
    GAME_LOST: 'game.lost', // Player was defeated
    GAME_QUIT: 'game.quit', // Player quit the game
    GAME_ABORTED: 'game.aborted', // Game ended abnormally
    // Session events
    SESSION_SAVING: 'game.session_saving', // Save in progress
    SESSION_SAVED: 'game.session_saved', // Save completed
    SESSION_RESTORING: 'game.session_restoring', // Restore in progress  
    SESSION_RESTORED: 'game.session_restored', // Restore completed
    // Player character events (ADR-132)
    PC_SWITCHED: 'game.pc_switched', // Player character changed
    // Error events
    INITIALIZATION_FAILED: 'game.initialization_failed',
    STORY_LOAD_FAILED: 'game.story_load_failed',
    FATAL_ERROR: 'game.fatal_error'
};
/**
 * Check if an event is a game lifecycle event (any GameEventType).
 * Use specific type guards (isGameStartedEvent, etc.) for typed data access.
 */
function isGameEvent(event) {
    return Object.values(exports.GameEventType).includes(event.type);
}
/**
 * Check if an event is part of the game start sequence.
 */
function isGameStartSequenceEvent(event) {
    return (event.type === exports.GameEventType.GAME_INITIALIZING ||
        event.type === exports.GameEventType.GAME_INITIALIZED ||
        event.type === exports.GameEventType.STORY_LOADING ||
        event.type === exports.GameEventType.STORY_LOADED ||
        event.type === exports.GameEventType.GAME_STARTING ||
        event.type === exports.GameEventType.GAME_STARTED);
}
/**
 * Check if an event is part of the game end sequence.
 */
function isGameEndSequenceEvent(event) {
    return (event.type === exports.GameEventType.GAME_ENDING ||
        event.type === exports.GameEventType.GAME_ENDED ||
        event.type === exports.GameEventType.GAME_WON ||
        event.type === exports.GameEventType.GAME_LOST ||
        event.type === exports.GameEventType.GAME_QUIT ||
        event.type === exports.GameEventType.GAME_ABORTED);
}
// =============================================================================
// Helper Functions to Create Game Events
// =============================================================================
/**
 * Generate a unique event ID
 */
function generateEventId() {
    return `game-${Date.now()}-${Math.random().toString(36).substr(2, 9)}`;
}
/**
 * Create a generic game event with typed data.
 * Use specific creators (createGameStartedEvent, etc.) when possible for type safety.
 */
function createGameEvent(type, data) {
    return {
        id: generateEventId(),
        type,
        timestamp: Date.now(),
        entities: {},
        data: data || {}
    };
}
// Specific event creators - all return ISemanticEvent with typed data
function createGameInitializingEvent() {
    return createGameEvent(exports.GameEventType.GAME_INITIALIZING, {
        gameState: 'initializing'
    });
}
function createGameInitializedEvent() {
    return createGameEvent(exports.GameEventType.GAME_INITIALIZED, {
        gameState: 'ready'
    });
}
function createStoryLoadingEvent(storyId) {
    return createGameEvent(exports.GameEventType.STORY_LOADING, {
        story: { id: storyId }
    });
}
function createStoryLoadedEvent(story) {
    return createGameEvent(exports.GameEventType.STORY_LOADED, {
        story,
        gameState: 'ready'
    });
}
function createGameStartingEvent(story) {
    return createGameEvent(exports.GameEventType.GAME_STARTING, {
        story,
        gameState: 'ready'
    });
}
function createGameStartedEvent(story, startTime, engineVersion, clientVersion) {
    return createGameEvent(exports.GameEventType.GAME_STARTED, {
        story,
        engineVersion,
        clientVersion,
        gameState: 'running',
        session: {
            startTime: startTime || Date.now(),
            turns: 0,
            moves: 0
        }
    });
}
function createGameEndingEvent(reason, session) {
    return createGameEvent(exports.GameEventType.GAME_ENDING, {
        gameState: 'ending',
        session,
        ending: { reason }
    });
}
function createGameEndedEvent(endingType, session, ending) {
    return createGameEvent(exports.GameEventType.GAME_ENDED, {
        gameState: 'ended',
        session: {
            ...session,
            endTime: Date.now()
        },
        ending: {
            ...ending,
            type: endingType
        }
    });
}
function createGameWonEvent(session, ending) {
    return createGameEvent(exports.GameEventType.GAME_WON, {
        gameState: 'ended',
        session: {
            ...session,
            endTime: Date.now()
        },
        ending: {
            ...ending,
            type: 'victory'
        }
    });
}
function createGameLostEvent(reason, session) {
    return createGameEvent(exports.GameEventType.GAME_LOST, {
        gameState: 'ended',
        session: {
            ...session,
            endTime: Date.now()
        },
        ending: {
            type: 'defeat',
            reason
        }
    });
}
function createGameQuitEvent(session) {
    return createGameEvent(exports.GameEventType.GAME_QUIT, {
        gameState: 'ended',
        session: {
            ...session,
            endTime: Date.now()
        },
        ending: {
            type: 'quit',
            reason: 'Player quit'
        }
    });
}
function createGameAbortedEvent(error, session) {
    return createGameEvent(exports.GameEventType.GAME_ABORTED, {
        gameState: 'ended',
        session: {
            ...session,
            endTime: Date.now()
        },
        ending: {
            type: 'abort',
            reason: error
        },
        error: {
            message: error
        }
    });
}
function createPcSwitchedEvent(previousPlayerId, newPlayerId) {
    return createGameEvent(exports.GameEventType.PC_SWITCHED, {
        previousPlayerId,
        newPlayerId
    });
}
// =============================================================================
// Type Guards for Game Events
// =============================================================================
/** Type guard to check if an event has specific data type */
function isGameStartedEvent(event) {
    return event.type === exports.GameEventType.GAME_STARTED;
}
function isGameEndedEvent(event) {
    return event.type === exports.GameEventType.GAME_ENDED;
}
function isGameWonEvent(event) {
    return event.type === exports.GameEventType.GAME_WON;
}
function isGameLostEvent(event) {
    return event.type === exports.GameEventType.GAME_LOST;
}
function isGameQuitEvent(event) {
    return event.type === exports.GameEventType.GAME_QUIT;
}
function isGameAbortedEvent(event) {
    return event.type === exports.GameEventType.GAME_ABORTED;
}
