// @sharpee/core/events/index.js
"use strict";
// packages/core/src/events/index.ts
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createEventSource = exports.EventSourceImpl = exports.createEvent = exports.createSemanticEventSource = exports.SemanticEventSourceImpl = exports.createGenericEventSource = exports.SimpleEventSource = void 0;
/**
 * Core event system
 *
 * Provides a generic event bus and text processing system
 * without knowledge of specific event types
 */
// Export types first
__exportStar(require("./types"), exports);
__exportStar(require("./standard-events"), exports);
__exportStar(require("./system-event"), exports);
__exportStar(require("./platform-events"), exports);
__exportStar(require("./game-events"), exports);
// Export typed event system (ADR-082)
__exportStar(require("./event-registry"), exports);
__exportStar(require("./typed-event"), exports);
__exportStar(require("./event-factory"), exports);
__exportStar(require("./event-helpers"), exports);
// Export event sources with specific names to avoid conflicts
var event_source_1 = require("./event-source");
Object.defineProperty(exports, "SimpleEventSource", { enumerable: true, get: function () { return event_source_1.SimpleEventSource; } });
Object.defineProperty(exports, "createGenericEventSource", { enumerable: true, get: function () { return event_source_1.createEventSource; } });
var semantic_event_source_1 = require("./semantic-event-source");
Object.defineProperty(exports, "SemanticEventSourceImpl", { enumerable: true, get: function () { return semantic_event_source_1.SemanticEventSourceImpl; } });
Object.defineProperty(exports, "createSemanticEventSource", { enumerable: true, get: function () { return semantic_event_source_1.createSemanticEventSource; } });
// Export from event-system
var event_system_1 = require("./event-system");
Object.defineProperty(exports, "createEvent", { enumerable: true, get: function () { return event_system_1.createEvent; } });
Object.defineProperty(exports, "EventSourceImpl", { enumerable: true, get: function () { return event_system_1.EventSourceImpl; } });
Object.defineProperty(exports, "createEventSource", { enumerable: true, get: function () { return event_system_1.createEventSource; } });
// message-resolver moved to stdlib (uses GameContext)
// enhanced-text-processor moved to stdlib (uses GameContext) 
// message-builder moved to stdlib (uses GameContext)
