// @sharpee/core/events/platform-events.js
"use strict";
/**
 * Platform events for operations that require client/host intervention
 * These events are processed after turn completion but before text service
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PlatformEventType = void 0;
exports.isPlatformEvent = isPlatformEvent;
exports.isPlatformRequestEvent = isPlatformRequestEvent;
exports.isPlatformCompletionEvent = isPlatformCompletionEvent;
exports.createPlatformEvent = createPlatformEvent;
exports.createSaveRequestedEvent = createSaveRequestedEvent;
exports.createRestoreRequestedEvent = createRestoreRequestedEvent;
exports.createQuitRequestedEvent = createQuitRequestedEvent;
exports.createRestartRequestedEvent = createRestartRequestedEvent;
exports.createSaveCompletedEvent = createSaveCompletedEvent;
exports.createRestoreCompletedEvent = createRestoreCompletedEvent;
exports.createQuitConfirmedEvent = createQuitConfirmedEvent;
exports.createQuitCancelledEvent = createQuitCancelledEvent;
exports.createRestartCompletedEvent = createRestartCompletedEvent;
exports.createUndoRequestedEvent = createUndoRequestedEvent;
exports.createUndoCompletedEvent = createUndoCompletedEvent;
exports.createAgainRequestedEvent = createAgainRequestedEvent;
exports.createAgainFailedEvent = createAgainFailedEvent;
/**
 * Platform event types for save, restore, quit, and restart operations
 */
exports.PlatformEventType = {
    // Request events - emitted by actions
    SAVE_REQUESTED: 'platform.save_requested',
    RESTORE_REQUESTED: 'platform.restore_requested',
    QUIT_REQUESTED: 'platform.quit_requested',
    RESTART_REQUESTED: 'platform.restart_requested',
    UNDO_REQUESTED: 'platform.undo_requested',
    AGAIN_REQUESTED: 'platform.again_requested',
    // Completion events - emitted by engine after processing
    SAVE_COMPLETED: 'platform.save_completed',
    RESTORE_COMPLETED: 'platform.restore_completed',
    QUIT_CONFIRMED: 'platform.quit_confirmed',
    RESTART_COMPLETED: 'platform.restart_completed',
    UNDO_COMPLETED: 'platform.undo_completed',
    // Error events
    SAVE_FAILED: 'platform.save_failed',
    RESTORE_FAILED: 'platform.restore_failed',
    QUIT_CANCELLED: 'platform.quit_cancelled',
    RESTART_CANCELLED: 'platform.restart_cancelled',
    UNDO_FAILED: 'platform.undo_failed',
    AGAIN_FAILED: 'platform.again_failed'
};
/**
 * Type guards for platform events
 */
function isPlatformEvent(event) {
    return 'requiresClientAction' in event && event.requiresClientAction === true;
}
function isPlatformRequestEvent(event) {
    return isPlatformEvent(event) && (event.type === exports.PlatformEventType.SAVE_REQUESTED ||
        event.type === exports.PlatformEventType.RESTORE_REQUESTED ||
        event.type === exports.PlatformEventType.QUIT_REQUESTED ||
        event.type === exports.PlatformEventType.RESTART_REQUESTED ||
        event.type === exports.PlatformEventType.UNDO_REQUESTED ||
        event.type === exports.PlatformEventType.AGAIN_REQUESTED);
}
function isPlatformCompletionEvent(event) {
    return isPlatformEvent(event) && (event.type === exports.PlatformEventType.SAVE_COMPLETED ||
        event.type === exports.PlatformEventType.RESTORE_COMPLETED ||
        event.type === exports.PlatformEventType.QUIT_CONFIRMED ||
        event.type === exports.PlatformEventType.RESTART_COMPLETED ||
        event.type === exports.PlatformEventType.UNDO_COMPLETED ||
        event.type === exports.PlatformEventType.SAVE_FAILED ||
        event.type === exports.PlatformEventType.RESTORE_FAILED ||
        event.type === exports.PlatformEventType.QUIT_CANCELLED ||
        event.type === exports.PlatformEventType.RESTART_CANCELLED ||
        event.type === exports.PlatformEventType.UNDO_FAILED ||
        event.type === exports.PlatformEventType.AGAIN_FAILED);
}
/**
 * Helper functions to create platform events
 */
function createPlatformEvent(type, context, additionalPayload) {
    return {
        id: `platform-${Date.now()}-${Math.random().toString(36).substr(2, 9)}`,
        type,
        timestamp: Date.now(),
        requiresClientAction: true,
        entities: {},
        payload: {
            context,
            ...additionalPayload
        }
    };
}
function createSaveRequestedEvent(context) {
    return createPlatformEvent(exports.PlatformEventType.SAVE_REQUESTED, context);
}
function createRestoreRequestedEvent(context) {
    return createPlatformEvent(exports.PlatformEventType.RESTORE_REQUESTED, context);
}
function createQuitRequestedEvent(context) {
    return createPlatformEvent(exports.PlatformEventType.QUIT_REQUESTED, context);
}
function createRestartRequestedEvent(context) {
    return createPlatformEvent(exports.PlatformEventType.RESTART_REQUESTED, context);
}
/**
 * Helper functions for completion events
 */
function createSaveCompletedEvent(success, error) {
    return createPlatformEvent(success ? exports.PlatformEventType.SAVE_COMPLETED : exports.PlatformEventType.SAVE_FAILED, undefined, { success, error });
}
function createRestoreCompletedEvent(success, error) {
    return createPlatformEvent(success ? exports.PlatformEventType.RESTORE_COMPLETED : exports.PlatformEventType.RESTORE_FAILED, undefined, { success, error });
}
function createQuitConfirmedEvent() {
    return createPlatformEvent(exports.PlatformEventType.QUIT_CONFIRMED, undefined, { success: true });
}
function createQuitCancelledEvent() {
    return createPlatformEvent(exports.PlatformEventType.QUIT_CANCELLED, undefined, { success: false });
}
function createRestartCompletedEvent(success) {
    return createPlatformEvent(success ? exports.PlatformEventType.RESTART_COMPLETED : exports.PlatformEventType.RESTART_CANCELLED, undefined, { success });
}
function createUndoRequestedEvent(context) {
    return createPlatformEvent(exports.PlatformEventType.UNDO_REQUESTED, context);
}
function createUndoCompletedEvent(success, restoredToTurn, error) {
    return createPlatformEvent(success ? exports.PlatformEventType.UNDO_COMPLETED : exports.PlatformEventType.UNDO_FAILED, undefined, { success, restoredToTurn, error });
}
function createAgainRequestedEvent(context) {
    return createPlatformEvent(exports.PlatformEventType.AGAIN_REQUESTED, context);
}
function createAgainFailedEvent(error) {
    return createPlatformEvent(exports.PlatformEventType.AGAIN_FAILED, undefined, { success: false, error });
}
