// @sharpee/core/events/semantic-event-source.js
"use strict";
/**
 * Semantic event source - specialized event source for story events
 * Builds on the generic event source infrastructure
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SemanticEventSourceImpl = void 0;
exports.createSemanticEventSource = createSemanticEventSource;
exports.createEventSource = createEventSource;
const event_source_1 = require("./event-source");
/**
 * Implementation of semantic event source
 */
class SemanticEventSourceImpl extends event_source_1.SimpleEventSource {
    events = [];
    eventEmitter;
    lastProcessedIndex = 0;
    constructor() {
        super();
        this.eventEmitter = new EventEmitterImpl();
        // Bridge between generic event source and semantic event emitter
        this.subscribe((event) => {
            this.eventEmitter.emit(event);
        });
    }
    addEvent(event) {
        this.events.push(event);
        this.emit(event);
    }
    getAllEvents() {
        return [...this.events];
    }
    getEventsByType(type) {
        return this.events.filter(event => event.type === type);
    }
    getEventsByEntity(entityId) {
        return this.events.filter(event => {
            const entities = event.entities;
            return (entities.actor === entityId ||
                entities.target === entityId ||
                entities.instrument === entityId ||
                entities.location === entityId ||
                (entities.others && entities.others.includes(entityId)));
        });
    }
    getEventsByTag(tag) {
        return this.events.filter(event => event.tags && event.tags.includes(tag));
    }
    filter(predicate) {
        return this.events.filter(predicate);
    }
    clearEvents() {
        this.events = [];
    }
    getEmitter() {
        return this.eventEmitter;
    }
    /**
     * Get events since a specific event ID
     */
    getEventsSince(eventId) {
        if (!eventId) {
            return this.getAllEvents();
        }
        const index = this.events.findIndex(e => e.id === eventId);
        if (index === -1) {
            return this.getAllEvents();
        }
        return this.events.slice(index + 1);
    }
    /**
     * Get unprocessed events and mark them as processed
     */
    getUnprocessedEvents() {
        const unprocessed = this.events.slice(this.lastProcessedIndex);
        this.lastProcessedIndex = this.events.length;
        return unprocessed;
    }
}
exports.SemanticEventSourceImpl = SemanticEventSourceImpl;
/**
 * Implementation of the EventEmitter interface
 */
class EventEmitterImpl {
    listeners = new Map();
    globalListeners = new Set();
    on(type, listener) {
        if (type === '*') {
            this.globalListeners.add(listener);
            return;
        }
        if (!this.listeners.has(type)) {
            this.listeners.set(type, new Set());
        }
        this.listeners.get(type).add(listener);
    }
    off(type, listener) {
        if (type === '*') {
            this.globalListeners.delete(listener);
            return;
        }
        const typeListeners = this.listeners.get(type);
        if (typeListeners) {
            typeListeners.delete(listener);
            if (typeListeners.size === 0) {
                this.listeners.delete(type);
            }
        }
    }
    emit(event) {
        // Emit to specific type listeners
        const typeListeners = this.listeners.get(event.type);
        if (typeListeners) {
            for (const listener of typeListeners) {
                try {
                    listener(event);
                }
                catch (error) {
                    console.error(`Error in event listener for ${event.type}:`, error);
                }
            }
        }
        // Emit to global listeners
        for (const listener of this.globalListeners) {
            try {
                listener(event);
            }
            catch (error) {
                console.error(`Error in global event listener:`, error);
            }
        }
    }
}
/**
 * Create a new semantic event source
 */
function createSemanticEventSource() {
    return new SemanticEventSourceImpl();
}
/**
 * Create event source for backwards compatibility
 * @deprecated Use createSemanticEventSource instead
 */
function createEventSource() {
    return createSemanticEventSource();
}
