// @sharpee/core/events/system-event.js
"use strict";
/**
 * System events are used for debugging, monitoring, and system-level notifications.
 * These are separate from SemanticEvents which represent story-meaningful occurrences.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Subsystems = void 0;
exports.createSystemEvent = createSystemEvent;
exports.isSystemEvent = isSystemEvent;
/**
 * Common subsystem identifiers
 */
exports.Subsystems = {
    PARSER: 'parser',
    VALIDATOR: 'validator',
    EXECUTOR: 'executor',
    WORLD_MODEL: 'world-model',
    TEXT_SERVICE: 'text-service',
    EVENT_PROCESSOR: 'event-processor',
    RULE_ENGINE: 'rule-engine',
};
/**
 * Helper to create a system event with common fields
 */
function createSystemEvent(subsystem, type, data, options) {
    return {
        id: generateEventId(),
        timestamp: Date.now(),
        subsystem,
        type,
        data,
        ...options
    };
}
/**
 * Simple ID generator for events
 * In production, might use UUID or similar
 */
let eventCounter = 0;
function generateEventId() {
    return `sys_${Date.now()}_${++eventCounter}`;
}
/**
 * Type guard to check if an object is a SystemEvent
 */
function isSystemEvent(obj) {
    return (typeof obj === 'object' &&
        obj !== null &&
        'id' in obj &&
        'timestamp' in obj &&
        'subsystem' in obj &&
        'type' in obj &&
        'data' in obj);
}
