// @sharpee/world-model/traits/vehicle/vehicleTrait.js
"use strict";
/**
 * Vehicle Trait - For enterable containers that transport actors
 *
 * Vehicles are enterable containers (bucket, boat, basket, balloon) where:
 * - Actor enters the vehicle (ENTER BUCKET)
 * - Vehicle moves, taking its contents with it
 * - Actor exits the vehicle (EXIT BUCKET)
 *
 * Based on Infocom's vehicle concept from Zork.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.VehicleTrait = void 0;
exports.isVehicleTrait = isVehicleTrait;
exports.createVehicleTrait = createVehicleTrait;
const trait_types_1 = require("../trait-types");
/**
 * Trait for vehicles - enterable containers that transport actors
 */
class VehicleTrait {
    static type = trait_types_1.TraitType.VEHICLE;
    type = trait_types_1.TraitType.VEHICLE;
    /** Type of vehicle determines default behavior */
    vehicleType;
    /**
     * When vehicle moves, all contents (including actors) move with it.
     * This is always true for vehicles.
     */
    movesWithContents = true;
    /**
     * If true, actor cannot use GO/walk commands while in vehicle.
     * Movement must be through vehicle-specific actions.
     * - Bucket: POUR/FILL moves it
     * - Boat: row, current moves it
     * - Balloon: burn/release moves it
     */
    blocksWalkingMovement;
    /**
     * If true, actor must EXIT vehicle before leaving the room on foot.
     * Prevents "GO EAST" while sitting in a bucket.
     */
    requiresExitBeforeLeaving;
    /**
     * Current location/state of the vehicle (e.g., 'top', 'bottom', 'river-1')
     * Used by vehicle-specific logic to track position.
     */
    currentPosition;
    /**
     * Map of position names to room IDs.
     * e.g., { 'top': 'top-of-well', 'bottom': 'well-bottom' }
     */
    positionRooms;
    /**
     * Whether the vehicle is currently operational/usable.
     * Boat might need to be inflated, balloon might need fuel.
     */
    isOperational;
    /**
     * Reason the vehicle is not operational (for error messages).
     */
    notOperationalReason;
    /**
     * If true, visibility passes through to the containing room.
     * - Bucket, raft, boat: transparent (can see the room you're in)
     * - Airplane, submarine: not transparent (see vehicle interior only)
     */
    transparent;
    constructor(options = {}) {
        this.vehicleType = options.vehicleType ?? 'generic';
        this.blocksWalkingMovement = options.blocksWalkingMovement ?? true;
        this.requiresExitBeforeLeaving = options.requiresExitBeforeLeaving ?? true;
        this.currentPosition = options.currentPosition;
        this.positionRooms = options.positionRooms;
        this.isOperational = options.isOperational ?? true;
        this.notOperationalReason = options.notOperationalReason;
        this.transparent = options.transparent ?? true; // Default to transparent for most IF vehicles
    }
}
exports.VehicleTrait = VehicleTrait;
/**
 * Type guard for VehicleTrait
 */
function isVehicleTrait(trait) {
    return trait.type === trait_types_1.TraitType.VEHICLE;
}
/**
 * Factory function for creating VehicleTrait
 */
function createVehicleTrait(options = {}) {
    return new VehicleTrait(options);
}
