// @sharpee/sharpee/cli/build-browser.js
"use strict";
/**
 * CLI: sharpee build-browser
 *
 * Bundles a Sharpee story for web browser deployment.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.runBuildBrowserCommand = runBuildBrowserCommand;
const fs = __importStar(require("fs"));
const path = __importStar(require("path"));
const child_process_1 = require("child_process");
/**
 * Read story info from package.json or index.ts
 */
function getProjectInfo(projectDir) {
    // Try package.json first
    const packagePath = path.join(projectDir, 'package.json');
    if (fs.existsSync(packagePath)) {
        try {
            const pkg = JSON.parse(fs.readFileSync(packagePath, 'utf-8'));
            return {
                storyId: pkg.name || 'my-story',
                storyTitle: pkg.description || pkg.name || 'My Story',
            };
        }
        catch {
            // Fall through
        }
    }
    // Try reading from src/index.ts
    const indexPath = path.join(projectDir, 'src', 'index.ts');
    if (fs.existsSync(indexPath)) {
        const content = fs.readFileSync(indexPath, 'utf-8');
        const idMatch = content.match(/id:\s*['"]([^'"]+)['"]/);
        const titleMatch = content.match(/title:\s*['"]([^'"]+)['"]/);
        if (idMatch || titleMatch) {
            return {
                storyId: idMatch?.[1] || 'my-story',
                storyTitle: titleMatch?.[1] || 'My Story',
            };
        }
    }
    return null;
}
/**
 * Process template placeholders
 */
function processTemplate(content, info) {
    return content
        .replace(/\{\{STORY_ID\}\}/g, info.storyId)
        .replace(/\{\{STORY_TITLE\}\}/g, info.storyTitle);
}
/**
 * Run the build-browser command
 */
async function runBuildBrowserCommand(args) {
    // Check for help
    if (args.includes('--help') || args.includes('-h')) {
        showHelp();
        return;
    }
    const projectDir = process.cwd();
    const minify = !args.includes('--no-minify');
    const sourcemap = !args.includes('--no-sourcemap');
    console.log('\n🔨 Building browser bundle\n');
    // Check if this is a Sharpee project
    const info = getProjectInfo(projectDir);
    if (!info) {
        console.error('Error: This does not appear to be a Sharpee project.');
        console.error('Make sure you have a package.json or src/index.ts with story config.');
        process.exit(1);
    }
    console.log(`  Story: ${info.storyTitle} (${info.storyId})`);
    // Check for browser-entry.ts
    const browserEntryPath = path.join(projectDir, 'src', 'browser-entry.ts');
    if (!fs.existsSync(browserEntryPath)) {
        console.error('\nError: src/browser-entry.ts not found.');
        console.error('Run "sharpee init-browser" first to add browser support.');
        process.exit(1);
    }
    // Create output directory
    const outDir = path.join(projectDir, 'dist', 'web');
    fs.mkdirSync(outDir, { recursive: true });
    // Build with esbuild
    console.log('  Building...');
    const esbuildArgs = [
        browserEntryPath,
        '--bundle',
        '--platform=browser',
        '--target=es2020',
        '--format=iife',
        `--global-name=SharpeeGame`,
        `--outfile=${path.join(outDir, info.storyId + '.js')}`,
        '--define:process.env.NODE_ENV=\\"production\\"',
        '--define:process.env.PARSER_DEBUG=\\"false\\"',
        '--define:process.env.DEBUG_PRONOUNS=\\"\\"',
    ];
    if (minify) {
        esbuildArgs.push('--minify');
    }
    if (sourcemap) {
        esbuildArgs.push('--sourcemap');
    }
    try {
        (0, child_process_1.execSync)(`npx esbuild ${esbuildArgs.join(' ')}`, {
            cwd: projectDir,
            stdio: 'pipe',
        });
        console.log(`  ✓ Built ${info.storyId}.js`);
    }
    catch (error) {
        console.error('  ✗ Build failed');
        if (error.stderr) {
            console.error(error.stderr.toString());
        }
        process.exit(1);
    }
    // Copy HTML template
    const browserHtmlPath = path.join(projectDir, 'browser', 'index.html');
    const defaultHtmlPath = path.join(__dirname, '..', '..', 'templates', 'browser', 'index.html');
    const htmlSource = fs.existsSync(browserHtmlPath) ? browserHtmlPath : defaultHtmlPath;
    if (fs.existsSync(htmlSource)) {
        let htmlContent = fs.readFileSync(htmlSource, 'utf-8');
        htmlContent = processTemplate(htmlContent, info);
        fs.writeFileSync(path.join(outDir, 'index.html'), htmlContent);
        console.log('  ✓ Copied index.html');
    }
    else {
        console.warn('  ⚠ HTML template not found');
    }
    // Copy CSS
    const browserCssPath = path.join(projectDir, 'browser', 'styles.css');
    const defaultCssPath = path.join(__dirname, '..', '..', 'templates', 'browser', 'styles.css');
    const cssSource = fs.existsSync(browserCssPath) ? browserCssPath : defaultCssPath;
    if (fs.existsSync(cssSource)) {
        fs.copyFileSync(cssSource, path.join(outDir, 'styles.css'));
        console.log('  ✓ Copied styles.css');
    }
    else {
        console.warn('  ⚠ CSS file not found');
    }
    // Report bundle size
    const bundlePath = path.join(outDir, info.storyId + '.js');
    if (fs.existsSync(bundlePath)) {
        const stats = fs.statSync(bundlePath);
        const sizeKb = (stats.size / 1024).toFixed(1);
        console.log(`\n  Bundle size: ${sizeKb} KB`);
    }
    console.log('\n✅ Build complete!\n');
    console.log(`Output: ${path.relative(projectDir, outDir)}/`);
    console.log('');
    console.log('To test locally:');
    console.log(`  npx serve ${path.relative(projectDir, outDir)}`);
    console.log('');
}
function showHelp() {
    console.log(`
sharpee build-browser - Build a web browser bundle

Usage: sharpee build-browser [options]

Options:
  --no-minify      Skip minification
  --no-sourcemap   Skip source map generation

Output:
  dist/web/
    index.html     HTML page
    <story-id>.js  JavaScript bundle
    styles.css     Stylesheet

The bundle includes your story, the Sharpee engine, and all dependencies.
`);
}
