// @sharpee/sharpee/cli/init.js
"use strict";
/**
 * CLI: sharpee init
 *
 * Creates a new Sharpee story project with the basic structure.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.runInitCommand = runInitCommand;
const fs = __importStar(require("fs"));
const path = __importStar(require("path"));
const readline = __importStar(require("readline"));
// Template directory relative to this file.
// In source: src/cli/ → ../../templates. In npm publish: cli/ → ../templates.
const TEMPLATES_DIR = fs.existsSync(path.join(__dirname, '..', 'templates', 'story'))
    ? path.join(__dirname, '..', 'templates', 'story')
    : path.join(__dirname, '..', '..', 'templates', 'story');
/**
 * Prompt user for input
 */
async function prompt(question, defaultValue) {
    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout,
    });
    return new Promise((resolve) => {
        const defaultHint = defaultValue ? ` (${defaultValue})` : '';
        rl.question(`${question}${defaultHint}: `, (answer) => {
            rl.close();
            resolve(answer.trim() || defaultValue || '');
        });
    });
}
/**
 * Convert title to kebab-case ID
 */
function toStoryId(title) {
    return title
        .toLowerCase()
        .replace(/[^a-z0-9]+/g, '-')
        .replace(/^-|-$/g, '');
}
/**
 * Read template file and replace placeholders
 */
function processTemplate(templatePath, options) {
    const content = fs.readFileSync(templatePath, 'utf-8');
    return content
        .replace(/\{\{STORY_ID\}\}/g, options.storyId)
        .replace(/\{\{STORY_TITLE\}\}/g, options.storyTitle)
        .replace(/\{\{AUTHOR\}\}/g, options.author)
        .replace(/\{\{DESCRIPTION\}\}/g, options.description);
}
/**
 * Run the init command
 */
async function runInitCommand(args) {
    // Check for help
    if (args.includes('--help') || args.includes('-h')) {
        showHelp();
        return;
    }
    // Check for non-interactive mode
    const useDefaults = args.includes('-y') || args.includes('--yes');
    const filteredArgs = args.filter(a => a !== '-y' && a !== '--yes');
    // Get target directory
    const targetDir = filteredArgs[0] || '.';
    const absoluteTarget = path.resolve(process.cwd(), targetDir);
    console.log('\n📖 Create a new Sharpee story\n');
    // Check if directory exists and is not empty
    if (fs.existsSync(absoluteTarget)) {
        const files = fs.readdirSync(absoluteTarget);
        if (files.length > 0 && !files.every(f => f.startsWith('.'))) {
            console.error(`Error: Directory "${targetDir}" is not empty.`);
            console.error('Please use an empty directory or specify a new one.');
            process.exit(1);
        }
    }
    // Gather project info (use defaults if -y flag)
    const defaultTitle = path.basename(absoluteTarget) || 'My Adventure';
    const storyTitle = useDefaults ? defaultTitle : await prompt('Story title', defaultTitle);
    const storyId = useDefaults ? toStoryId(storyTitle) : await prompt('Story ID (package name)', toStoryId(storyTitle));
    const author = useDefaults ? (process.env.USER || 'Anonymous') : await prompt('Author name', process.env.USER || 'Anonymous');
    const description = useDefaults ? 'An interactive fiction adventure' : await prompt('Description', 'An interactive fiction adventure');
    const options = {
        storyId,
        storyTitle,
        author,
        description,
    };
    console.log('\nCreating project...\n');
    // Create directory structure
    fs.mkdirSync(absoluteTarget, { recursive: true });
    fs.mkdirSync(path.join(absoluteTarget, 'src'), { recursive: true });
    // Copy and process templates
    const templates = [
        { src: 'index.ts.template', dest: 'src/index.ts' },
        { src: 'package.json.template', dest: 'package.json' },
        { src: 'tsconfig.json.template', dest: 'tsconfig.json' },
    ];
    for (const template of templates) {
        const srcPath = path.join(TEMPLATES_DIR, template.src);
        const destPath = path.join(absoluteTarget, template.dest);
        if (fs.existsSync(srcPath)) {
            const content = processTemplate(srcPath, options);
            fs.writeFileSync(destPath, content);
            console.log(`  ✓ Created ${template.dest}`);
        }
        else {
            console.warn(`  ⚠ Template not found: ${template.src}`);
        }
    }
    // Create .gitignore
    const gitignore = `node_modules/
dist/
*.log
.DS_Store
`;
    fs.writeFileSync(path.join(absoluteTarget, '.gitignore'), gitignore);
    console.log('  ✓ Created .gitignore');
    console.log('\n✅ Project created!\n');
    console.log('Next steps:');
    if (targetDir !== '.') {
        console.log(`  cd ${targetDir}`);
    }
    console.log('  npm install');
    console.log('  npm run build');
    console.log('');
    console.log('To add a browser client:');
    console.log('  npx sharpee init-browser');
    console.log('');
}
function showHelp() {
    console.log(`
sharpee init - Create a new Sharpee story project

Usage: sharpee init [directory] [options]

Arguments:
  directory    Target directory (default: current directory)

Options:
  -y, --yes    Use defaults without prompting

Examples:
  sharpee init                    Create in current directory (interactive)
  sharpee init my-adventure       Create in ./my-adventure/ (interactive)
  sharpee init my-adventure -y    Create with defaults (non-interactive)
`);
}
