// @sharpee/stdlib/actions/standard/dropping/dropping-data.js
"use strict";
/**
 * Data builder for dropping action
 *
 * Centralizes all entity snapshot logic for the dropping action.
 * This separates data structure concerns from business logic.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.droppedDataConfig = exports.buildDroppedData = void 0;
exports.determineDroppingMessage = determineDroppingMessage;
const world_model_1 = require("../../../../world-model/index.js");
const snapshot_utils_1 = require("../../base/snapshot-utils");
/**
 * Build dropped event data
 *
 * Creates the complete data structure for dropped events,
 * including entity snapshots and location information.
 */
const buildDroppedData = (context, preState, postState) => {
    const actor = context.player;
    const noun = context.command.directObject?.entity;
    if (!noun) {
        // Shouldn't happen if validation passed
        return {
            item: '',
            itemName: 'nothing',
            toLocation: '',
            toLocationName: 'nowhere'
        };
    }
    // Determine drop location
    const playerLocation = context.world.getLocation(actor.id);
    const dropLocation = playerLocation ? context.world.getEntity(playerLocation) : context.currentLocation;
    if (!dropLocation) {
        throw new Error('No valid drop location found');
    }
    // Capture snapshots
    const itemSnapshot = (0, snapshot_utils_1.captureEntitySnapshot)(noun, context.world, true);
    const actorSnapshot = (0, snapshot_utils_1.captureEntitySnapshot)(actor, context.world, false);
    const locationSnapshot = (0, snapshot_utils_1.captureEntitySnapshot)(dropLocation, context.world, dropLocation.has(world_model_1.TraitType.ROOM));
    // Build base data
    const data = {
        item: noun.id,
        itemName: noun.name,
        toLocation: dropLocation.id,
        toLocationName: dropLocation.name,
        // Add atomic event snapshots
        itemSnapshot: itemSnapshot,
        actorSnapshot: actorSnapshot,
        locationSnapshot: locationSnapshot
    };
    // Add location type flags
    if (!dropLocation.has(world_model_1.TraitType.ROOM)) {
        if (dropLocation.has(world_model_1.TraitType.CONTAINER)) {
            data.toContainer = true;
        }
        else if (dropLocation.has(world_model_1.TraitType.SUPPORTER)) {
            data.toSupporter = true;
        }
    }
    else {
        data.toRoom = true;
    }
    return data;
};
exports.buildDroppedData = buildDroppedData;
/**
 * Determine the success message for dropping
 */
function determineDroppingMessage(droppedData, context) {
    const noun = context.command.directObject?.entity;
    const dropLocation = context.world.getEntity(droppedData.toLocation);
    if (!noun || !dropLocation) {
        return {
            messageId: 'dropped',
            params: { item: droppedData.itemName }
        };
    }
    const params = {
        item: noun.name,
        location: dropLocation.name
    };
    let messageId = 'dropped';
    // Determine message based on drop location
    if (droppedData.toContainer) {
        messageId = 'dropped_in';
        params.container = dropLocation.name;
    }
    else if (droppedData.toSupporter) {
        messageId = 'dropped_on';
        params.supporter = dropLocation.name;
    }
    else if (droppedData.toRoom) {
        // Vary the message based on how the item is dropped
        const verb = context.command.parsed.structure.verb?.text.toLowerCase() || 'drop';
        if (verb === 'discard') {
            messageId = 'dropped_carelessly';
        }
        else if (noun.has(world_model_1.TraitType.IDENTITY)) {
            // Check if item name suggests it's fragile
            const identity = noun.getTrait(world_model_1.IdentityTrait);
            if (identity && identity.name?.toLowerCase().includes('glass')) {
                messageId = 'dropped_quietly';
            }
        }
    }
    return { messageId, params };
}
/**
 * Configuration for dropping data builder
 *
 * Allows stories to extend the data while protecting core fields
 */
exports.droppedDataConfig = {
    builder: exports.buildDroppedData,
    protectedFields: ['item', 'itemName', 'toLocation', 'toLocationName']
};
