// @sharpee/world-model/traits/weapon/weaponTrait.js
"use strict";
/**
 * Weapon trait for entities that can be used to attack
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.WeaponTrait = void 0;
/**
 * Weapon trait indicates an entity can be used to attack.
 *
 * This trait contains only data - all combat logic
 * is in WeaponBehavior and AttackBehavior.
 */
class WeaponTrait {
    static type = 'weapon';
    type = 'weapon';
    // ADR-072 properties
    damage;
    skillBonus;
    isBlessed;
    glowsNearDanger;
    isGlowing;
    requiredTrait;
    // Legacy/additional properties
    minDamage;
    maxDamage;
    weaponType;
    twoHanded;
    attackMessage;
    hitSound;
    breakable;
    durability;
    maxDurability;
    constructor(data = {}) {
        // ADR-072 properties
        this.damage = data.damage ?? data.maxDamage ?? 1;
        this.skillBonus = data.skillBonus ?? 0;
        this.isBlessed = data.isBlessed ?? false;
        this.glowsNearDanger = data.glowsNearDanger ?? false;
        this.isGlowing = data.isGlowing ?? false;
        this.requiredTrait = data.requiredTrait;
        // Legacy properties (for backward compatibility)
        this.minDamage = data.minDamage ?? 1;
        this.maxDamage = data.maxDamage ?? data.minDamage ?? 1;
        this.weaponType = data.weaponType ?? 'blunt';
        this.twoHanded = data.twoHanded ?? false;
        this.attackMessage = data.attackMessage;
        this.hitSound = data.hitSound;
        this.breakable = data.breakable ?? false;
        this.durability = data.durability;
        this.maxDurability = data.maxDurability;
    }
    /**
     * Set the glow state (for elvish sword behavior)
     */
    setGlowing(glowing) {
        this.isGlowing = glowing;
    }
}
exports.WeaponTrait = WeaponTrait;
