// @sharpee/world-model/extensions/base-extension.js
"use strict";
/**
 * Base extension class
 *
 * Provides a convenient base class for creating extensions
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseExtension = void 0;
/**
 * Base class for extensions
 *
 * Provides helper methods for common extension tasks
 */
class BaseExtension {
    registry;
    /**
     * Initialize the extension
     */
    async initialize(registry) {
        this.registry = registry;
        // Register traits
        const traits = this.defineTraits();
        for (const traitDef of traits) {
            registry.registerTrait(this.metadata.namespace, {
                type: traitDef.name,
                implementation: traitDef.trait,
                behavior: traitDef.behavior,
                category: traitDef.category
            });
        }
        // Register events
        const events = this.defineEvents();
        for (const eventDef of events) {
            registry.registerEvent(this.metadata.namespace, {
                type: eventDef.name,
                description: eventDef.description,
                payloadSchema: eventDef.examplePayload
            });
        }
        // Custom initialization
        await this.onInitialize();
    }
    /**
     * Define traits provided by this extension
     * Override in subclasses
     */
    defineTraits() {
        return [];
    }
    /**
     * Define events provided by this extension
     * Override in subclasses
     */
    defineEvents() {
        return [];
    }
    /**
     * Custom initialization logic
     * Override in subclasses
     */
    async onInitialize() {
        // Override in subclasses
    }
    /**
     * Helper to create namespaced type names
     */
    namespaced(name) {
        return `${this.metadata.id}.${name}`;
    }
    /**
     * Helper to create a namespaced event
     */
    createEvent(type, payload) {
        // This would use the actual event creation from core
        return {
            type: this.namespaced(type),
            payload,
            timestamp: Date.now()
        };
    }
}
exports.BaseExtension = BaseExtension;
