// @sharpee/world-model/extensions/index.js
"use strict";
/**
 * Extension system exports
 *
 * This module provides the infrastructure for creating and loading
 * trait extensions that add new functionality to the world model.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseNamespacedId = exports.createNamespacedId = exports.ExtensionLoadError = exports.extensionLoader = exports.ExtensionLoader = exports.extensionRegistry = exports.ExtensionRegistry = void 0;
// Types and interfaces
__exportStar(require("./types"), exports);
// Registry
var registry_1 = require("./registry");
Object.defineProperty(exports, "ExtensionRegistry", { enumerable: true, get: function () { return registry_1.ExtensionRegistry; } });
Object.defineProperty(exports, "extensionRegistry", { enumerable: true, get: function () { return registry_1.extensionRegistry; } });
// Loader
var loader_1 = require("./loader");
Object.defineProperty(exports, "ExtensionLoader", { enumerable: true, get: function () { return loader_1.ExtensionLoader; } });
Object.defineProperty(exports, "extensionLoader", { enumerable: true, get: function () { return loader_1.extensionLoader; } });
Object.defineProperty(exports, "ExtensionLoadError", { enumerable: true, get: function () { return loader_1.ExtensionLoadError; } });
// Re-export commonly used functions
var types_1 = require("./types");
Object.defineProperty(exports, "createNamespacedId", { enumerable: true, get: function () { return types_1.createNamespacedId; } });
Object.defineProperty(exports, "parseNamespacedId", { enumerable: true, get: function () { return types_1.parseNamespacedId; } });
