// @sharpee/world-model/extensions/loader.js
"use strict";
/**
 * Extension loader implementation
 *
 * Manages loading, initialization, and lifecycle of trait extensions
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.extensionLoader = exports.ExtensionLoader = exports.ExtensionLoadError = void 0;
const registry_1 = require("./registry");
/**
 * Extension loading error
 */
class ExtensionLoadError extends Error {
    extensionId;
    cause;
    constructor(message, extensionId, cause) {
        super(message);
        this.extensionId = extensionId;
        this.cause = cause;
        this.name = 'ExtensionLoadError';
    }
}
exports.ExtensionLoadError = ExtensionLoadError;
/**
 * Default implementation of the extension loader
 */
class ExtensionLoader {
    extensions = new Map();
    registry;
    constructor(registry = registry_1.extensionRegistry) {
        this.registry = registry;
    }
    /**
     * Load an extension
     */
    async loadExtension(extension) {
        const { id, namespace, dependencies } = extension.metadata;
        // Check if already loaded
        if (this.extensions.has(id)) {
            throw new ExtensionLoadError(`Extension '${id}' is already loaded`, id);
        }
        // Check dependencies
        if (dependencies) {
            for (const dep of dependencies) {
                if (!dep.optional && !this.extensions.has(dep.id)) {
                    throw new ExtensionLoadError(`Required dependency '${dep.id}' is not loaded`, id);
                }
            }
        }
        try {
            // Initialize the extension
            if (extension.initialize) {
                await extension.initialize();
            }
            // Register traits
            if (extension.traits) {
                for (const trait of extension.traits) {
                    this.registry.registerTrait(namespace, trait);
                }
            }
            // Register events
            if (extension.events) {
                for (const event of extension.events) {
                    this.registry.registerEvent(namespace, event);
                }
            }
            // Register actions
            if (extension.actions) {
                for (const action of extension.actions) {
                    this.registry.registerAction(namespace, action);
                }
            }
            // Store the extension
            this.extensions.set(id, extension);
        }
        catch (error) {
            // Cleanup on failure
            this.registry.clearNamespace(namespace);
            throw new ExtensionLoadError(`Failed to load extension '${id}': ${error}`, id, error instanceof Error ? error : undefined);
        }
    }
    /**
     * Unload an extension
     */
    async unloadExtension(extensionId) {
        const extension = this.extensions.get(extensionId);
        if (!extension) {
            throw new ExtensionLoadError(`Extension '${extensionId}' is not loaded`, extensionId);
        }
        // Check if any loaded extensions depend on this one
        for (const [id, ext] of this.extensions) {
            if (id === extensionId)
                continue;
            const deps = ext.metadata.dependencies || [];
            const dependsOnThis = deps.some(d => d.id === extensionId && !d.optional);
            if (dependsOnThis) {
                throw new ExtensionLoadError(`Cannot unload '${extensionId}': extension '${id}' depends on it`, extensionId);
            }
        }
        try {
            // Cleanup the extension
            if (extension.cleanup) {
                await extension.cleanup();
            }
            // Clear registrations
            this.registry.clearNamespace(extension.metadata.namespace);
            // Remove from loaded extensions
            this.extensions.delete(extensionId);
        }
        catch (error) {
            throw new ExtensionLoadError(`Failed to unload extension '${extensionId}': ${error}`, extensionId, error instanceof Error ? error : undefined);
        }
    }
    /**
     * Get loaded extension by ID
     */
    getExtension(extensionId) {
        return this.extensions.get(extensionId);
    }
    /**
     * Get all loaded extensions
     */
    getLoadedExtensions() {
        return Array.from(this.extensions.values());
    }
    /**
     * Check if an extension is loaded
     */
    isLoaded(extensionId) {
        return this.extensions.has(extensionId);
    }
    /**
     * Get extensions in load order (respecting dependencies)
     */
    getLoadOrder() {
        const visited = new Set();
        const result = [];
        const visit = (id) => {
            if (visited.has(id))
                return;
            visited.add(id);
            const ext = this.extensions.get(id);
            if (!ext)
                return;
            // Visit dependencies first
            const deps = ext.metadata.dependencies || [];
            for (const dep of deps) {
                if (this.extensions.has(dep.id)) {
                    visit(dep.id);
                }
            }
            result.push(id);
        };
        // Visit all extensions
        for (const id of this.extensions.keys()) {
            visit(id);
        }
        return result;
    }
    /**
     * Validate extension dependencies are satisfied
     */
    validateDependencies(extension) {
        const missing = [];
        const deps = extension.metadata.dependencies || [];
        for (const dep of deps) {
            if (!dep.optional && !this.extensions.has(dep.id)) {
                missing.push(dep.id);
            }
        }
        return missing;
    }
}
exports.ExtensionLoader = ExtensionLoader;
// Global loader instance
exports.extensionLoader = new ExtensionLoader();
