// @sharpee/world-model/extensions/manager.js
"use strict";
/**
 * Extension manager implementation
 *
 * Manages loading, unloading, and lifecycle of extensions
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.extensionManager = exports.ExtensionManager = void 0;
const registry_1 = require("./registry");
/**
 * Manages extension lifecycle
 */
class ExtensionManager {
    extensions = new Map();
    extensionAPIs = new Map();
    loadOrder = [];
    /**
     * Load an extension
     */
    async loadExtension(extension) {
        const { id, dependencies = [] } = extension.metadata;
        // Check if already loaded
        if (this.extensions.has(id)) {
            throw new Error(`Extension '${id}' is already loaded`);
        }
        // Validate dependencies
        await this.validateDependencies(dependencies);
        // Get the global extension registry
        const registry = (0, registry_1.getExtensionRegistry)();
        try {
            // Initialize the extension if method is provided
            if (extension.initialize) {
                await extension.initialize(registry);
            }
            // Store the extension
            this.extensions.set(id, extension);
            this.loadOrder.push(id);
            // Get and store API if provided
            const getAPI = extension.getAPI;
            if (getAPI) {
                const api = getAPI.call(extension);
                if (api) {
                    this.extensionAPIs.set(id, api);
                }
            }
            console.log(`Extension '${id}' loaded successfully`);
        }
        catch (error) {
            // Clean up on failure
            registry.clearNamespace(extension.metadata.namespace);
            throw new Error(`Failed to load extension '${id}': ${error}`);
        }
    }
    /**
     * Unload an extension
     */
    async unloadExtension(extensionId) {
        const extension = this.extensions.get(extensionId);
        if (!extension) {
            throw new Error(`Extension '${extensionId}' is not loaded`);
        }
        // Check if other extensions depend on this one
        const dependents = this.findDependents(extensionId);
        if (dependents.length > 0) {
            throw new Error(`Cannot unload '${extensionId}': required by ${dependents.join(', ')}`);
        }
        try {
            // Call shutdown if provided
            if (extension.shutdown) {
                await extension.shutdown();
            }
            // Remove from collections
            this.extensions.delete(extensionId);
            this.extensionAPIs.delete(extensionId);
            this.loadOrder = this.loadOrder.filter(id => id !== extensionId);
            // Note: In a real implementation, we'd also clean up registered traits,
            // events, actions, and services from the global registries
            console.log(`Extension '${extensionId}' unloaded successfully`);
        }
        catch (error) {
            throw new Error(`Failed to unload extension '${extensionId}': ${error}`);
        }
    }
    /**
     * Get loaded extension by ID
     */
    getExtension(extensionId) {
        return this.extensions.get(extensionId);
    }
    /**
     * Get all loaded extensions
     */
    getLoadedExtensions() {
        return this.loadOrder.map(id => this.extensions.get(id));
    }
    /**
     * Check if an extension is loaded
     */
    isLoaded(extensionId) {
        return this.extensions.has(extensionId);
    }
    /**
     * Get extension API
     */
    getExtensionAPI(extensionId) {
        return this.extensionAPIs.get(extensionId);
    }
    /**
     * Validate extension dependencies
     */
    async validateDependencies(dependencies) {
        for (const dep of dependencies) {
            if (!dep.optional && !this.isLoaded(dep.id)) {
                throw new Error(`Required dependency '${dep.id}' is not loaded`);
            }
            // In a real implementation, we'd also check version compatibility
            if (this.isLoaded(dep.id)) {
                const loadedExt = this.getExtension(dep.id);
                if (!this.isVersionCompatible(loadedExt.metadata.version, dep.version)) {
                    throw new Error(`Dependency '${dep.id}' version ${loadedExt.metadata.version} ` +
                        `does not satisfy requirement ${dep.version}`);
                }
            }
        }
    }
    /**
     * Find extensions that depend on a given extension
     */
    findDependents(extensionId) {
        const dependents = [];
        for (const [id, ext] of this.extensions) {
            const deps = ext.metadata.dependencies || [];
            if (deps.some(dep => dep.id === extensionId && !dep.optional)) {
                dependents.push(id);
            }
        }
        return dependents;
    }
    /**
     * Check version compatibility (simplified)
     */
    isVersionCompatible(version, requirement) {
        // This is a simplified version check
        // In a real implementation, use a proper semver library
        if (requirement.startsWith('^')) {
            // Compatible with same major version
            const reqVersion = requirement.slice(1);
            return version.split('.')[0] === reqVersion.split('.')[0];
        }
        if (requirement.startsWith('>=')) {
            // Greater than or equal
            const reqVersion = requirement.slice(2);
            return version >= reqVersion;
        }
        // Exact match
        return version === requirement;
    }
}
exports.ExtensionManager = ExtensionManager;
/**
 * Global extension manager singleton
 */
exports.extensionManager = new ExtensionManager();
