// @sharpee/world-model/extensions/registry.js
"use strict";
/**
 * Extension registry implementation
 *
 * Manages registration and lookup of extension-provided traits,
 * events, actions, and other components.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.extensionRegistry = exports.ExtensionRegistry = void 0;
exports.getExtensionRegistry = getExtensionRegistry;
const types_1 = require("./types");
/**
 * Default implementation of the extension registry
 */
class ExtensionRegistry {
    traits = new Map();
    events = new Map();
    actions = new Map();
    namespaces = new Set();
    /**
     * Register a trait from an extension
     */
    registerTrait(namespace, definition) {
        const fullType = this.getTraitType(namespace, definition.type);
        if (this.traits.has(fullType)) {
            throw new Error(`Trait type '${fullType}' is already registered`);
        }
        this.traits.set(fullType, definition.implementation);
        this.namespaces.add(namespace);
        // TODO: Register behavior if provided
        // This would integrate with a behavior registry
    }
    /**
     * Register an event type from an extension
     */
    registerEvent(namespace, definition) {
        const fullType = this.getEventType(namespace, definition.type);
        if (this.events.has(fullType)) {
            throw new Error(`Event type '${fullType}' is already registered`);
        }
        this.events.set(fullType, definition);
        this.namespaces.add(namespace);
    }
    /**
     * Register an action from an extension
     */
    registerAction(namespace, definition) {
        const fullId = this.getActionId(namespace, definition.id);
        if (this.actions.has(fullId)) {
            throw new Error(`Action '${fullId}' is already registered`);
        }
        this.actions.set(fullId, definition);
        this.namespaces.add(namespace);
    }
    /**
     * Get a namespaced trait type
     */
    getTraitType(namespace, type) {
        return (0, types_1.createNamespacedId)(namespace, type);
    }
    /**
     * Get a namespaced event type
     */
    getEventType(namespace, type) {
        return (0, types_1.createNamespacedId)(namespace, type);
    }
    /**
     * Get a namespaced action ID
     */
    getActionId(namespace, id) {
        return (0, types_1.createNamespacedId)(namespace, id);
    }
    /**
     * Get a trait constructor by its full type
     */
    getTrait(fullType) {
        return this.traits.get(fullType);
    }
    /**
     * Get an event definition by its full type
     */
    getEvent(fullType) {
        return this.events.get(fullType);
    }
    /**
     * Get an action definition by its full ID
     */
    getAction(fullId) {
        return this.actions.get(fullId);
    }
    /**
     * Check if a namespace is registered
     */
    hasNamespace(namespace) {
        return this.namespaces.has(namespace);
    }
    /**
     * Get all registered namespaces
     */
    getNamespaces() {
        return Array.from(this.namespaces);
    }
    /**
     * Get all traits in a namespace
     */
    getTraitsByNamespace(namespace) {
        const prefix = namespace + '.';
        return Array.from(this.traits.keys())
            .filter(type => type.startsWith(prefix));
    }
    /**
     * Get all events in a namespace
     */
    getEventsByNamespace(namespace) {
        const prefix = namespace + '.';
        return Array.from(this.events.keys())
            .filter(type => type.startsWith(prefix));
    }
    /**
     * Get all actions in a namespace
     */
    getActionsByNamespace(namespace) {
        const prefix = namespace + '.';
        return Array.from(this.actions.keys())
            .filter(id => id.startsWith(prefix));
    }
    /**
     * Clear all registrations for a namespace
     */
    clearNamespace(namespace) {
        const prefix = namespace + '.';
        // Remove traits
        for (const type of this.traits.keys()) {
            if (type.startsWith(prefix)) {
                this.traits.delete(type);
            }
        }
        // Remove events
        for (const type of this.events.keys()) {
            if (type.startsWith(prefix)) {
                this.events.delete(type);
            }
        }
        // Remove actions
        for (const id of this.actions.keys()) {
            if (id.startsWith(prefix)) {
                this.actions.delete(id);
            }
        }
        this.namespaces.delete(namespace);
    }
    /**
     * Clear all registrations
     */
    clear() {
        this.traits.clear();
        this.events.clear();
        this.actions.clear();
        this.namespaces.clear();
    }
    /**
     * Get all traits (for compatibility)
     */
    getTraits() {
        return new Map(this.traits);
    }
    /**
     * Get all events (for compatibility)
     */
    getEvents() {
        return new Map(this.events);
    }
    /**
     * Get all actions (for compatibility)
     */
    getActions() {
        return new Map(this.actions);
    }
    /**
     * Get services (placeholder for future implementation)
     */
    getServices() {
        return new Map();
    }
}
exports.ExtensionRegistry = ExtensionRegistry;
// Global registry instance
exports.extensionRegistry = new ExtensionRegistry();
// Function to get the extension registry (for backwards compatibility)
function getExtensionRegistry() {
    return exports.extensionRegistry;
}
