// @sharpee/world-model/traits/scene/sceneTrait.js
"use strict";
/**
 * Scene trait for temporal story phases (ADR-149).
 *
 * Entities with this trait represent named dramatic episodes with
 * begin/end conditions evaluated each turn by the engine. Multiple
 * scenes can be active simultaneously. Scenes can be recurring.
 *
 * Condition closures are NOT serializable. On save/restore, trait
 * data (state, activeTurns, etc.) persists, but stories must
 * re-register conditions after restore.
 *
 * Public interface: SceneTrait, ISceneData, SceneState.
 * Owner context: @sharpee/world-model — traits / temporal
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SceneTrait = void 0;
const trait_types_1 = require("../trait-types");
/**
 * Marks an entity as a temporal scene — a dramatic episode with
 * begin/end conditions polled each turn by the engine.
 */
class SceneTrait {
    static type = trait_types_1.TraitType.SCENE;
    type = trait_types_1.TraitType.SCENE;
    /** Human-readable scene name. */
    name;
    /** Current lifecycle state. */
    state;
    /** Whether the scene can activate more than once. */
    recurring;
    /** Number of turns the scene has been active (0 if not active). */
    activeTurns;
    /** Turn number when the scene last began. */
    beganAtTurn;
    /** Turn number when the scene last ended. */
    endedAtTurn;
    constructor(data) {
        this.name = data.name;
        this.state = data.state ?? 'waiting';
        this.recurring = data.recurring ?? false;
        this.activeTurns = data.activeTurns ?? 0;
        this.beganAtTurn = data.beganAtTurn;
        this.endedAtTurn = data.endedAtTurn;
    }
}
exports.SceneTrait = SceneTrait;
