// @sharpee/stdlib/actions/standard/closing/closing.js
"use strict";
/**
 * Closing action - closes containers and doors
 *
 * This action properly delegates to OpenableBehavior for validation
 * and execution. It follows the validate/execute pattern.
 *
 * Uses four-phase pattern:
 * 1. validate: Check target exists and can be closed
 * 2. execute: Delegate to OpenableBehavior for state changes
 * 3. blocked: Generate error events when validation fails
 * 4. report: Generate success events with closed data
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.closingAction = void 0;
const world_model_1 = require("../../../../world-model/index.js");
const data_builder_types_1 = require("../../data-builder-types");
const constants_1 = require("../../constants");
const closing_data_1 = require("./closing-data");
const closing_messages_1 = require("./closing-messages");
const types_1 = require("../../../scope/types");
function getClosingSharedData(context) {
    return context.sharedData;
}
exports.closingAction = {
    id: constants_1.IFActions.CLOSING,
    // Default scope requirements for this action's slots
    defaultScope: {
        target: types_1.ScopeLevel.REACHABLE
    },
    // ADR-104: Implicit inference requirements
    targetRequirements: {
        trait: world_model_1.TraitType.OPENABLE,
        condition: 'is_open',
        description: 'closable'
    },
    // Closing doesn't require holding the target
    requiresHolding: false,
    requiredMessages: [
        'no_target',
        'not_closable',
        'already_closed',
        'closed',
        'cant_reach',
        'prevents_closing'
    ],
    metadata: {
        requiresDirectObject: true,
        requiresIndirectObject: false,
        directObjectScope: types_1.ScopeLevel.REACHABLE
    },
    group: 'container_manipulation',
    /**
     * Validate whether the close action can be executed
     * Uses behavior validation methods to check preconditions
     */
    validate(context) {
        const noun = context.command.directObject?.entity;
        // Validate we have a target
        if (!noun) {
            return {
                valid: false,
                error: closing_messages_1.ClosingMessages.NO_TARGET
            };
        }
        // Check for interceptor on the target entity (ADR-118)
        const interceptorResult = (0, world_model_1.getInterceptorForAction)(noun, constants_1.IFActions.CLOSING);
        const interceptor = interceptorResult?.interceptor;
        const interceptorData = {};
        // Store for later phases (blocked needs access to onBlocked)
        const sharedData = getClosingSharedData(context);
        sharedData.interceptor = interceptor;
        sharedData.interceptorData = interceptorData;
        // === PRE-VALIDATE HOOK ===
        if (interceptor?.preValidate) {
            const result = interceptor.preValidate(noun, context.world, context.player.id, interceptorData);
            if (result !== null && !result.valid) {
                return {
                    valid: false,
                    error: result.error,
                    params: result.params
                };
            }
        }
        // Check scope - must be able to reach the target
        const scopeCheck = context.requireScope(noun, types_1.ScopeLevel.REACHABLE);
        if (!scopeCheck.ok) {
            return scopeCheck.error;
        }
        // Check if it's openable (things that can open can also close)
        if (!noun.has(world_model_1.TraitType.OPENABLE)) {
            return {
                valid: false,
                error: closing_messages_1.ClosingMessages.NOT_CLOSABLE,
                params: { item: noun.name }
            };
        }
        // Use behavior's canClose method for validation
        if (!world_model_1.OpenableBehavior.canClose(noun)) {
            // Check if it's because it's already closed
            if (!world_model_1.OpenableBehavior.isOpen(noun)) {
                return {
                    valid: false,
                    error: closing_messages_1.ClosingMessages.ALREADY_CLOSED,
                    params: { item: noun.name }
                };
            }
            // Otherwise it can't be closed for some other reason
            return {
                valid: false,
                error: closing_messages_1.ClosingMessages.PREVENTS_CLOSING,
                params: { item: noun.name }
            };
        }
        // Check if closable has special requirements
        const openableTrait = noun.getTrait(world_model_1.OpenableTrait);
        const openableRecord = openableTrait;
        if (openableRecord?.closeRequirements) {
            const requirement = openableRecord.closeRequirements;
            if (requirement.preventedBy) {
                return {
                    valid: false,
                    error: closing_messages_1.ClosingMessages.PREVENTS_CLOSING,
                    params: {
                        item: noun.name,
                        obstacle: requirement.preventedBy
                    }
                };
            }
        }
        return { valid: true };
    },
    /**
     * Execute the close action
     * Assumes validation has already passed - no validation logic here
     * Delegates to OpenableBehavior for actual state changes
     */
    execute(context) {
        const noun = context.command.directObject.entity;
        // Delegate to behavior for closing
        const result = world_model_1.OpenableBehavior.close(noun);
        // Store result for report phase using sharedData
        const sharedData = getClosingSharedData(context);
        sharedData.closeResult = result;
    },
    /**
     * Report events after successful closing
     *
     * Uses simplified event pattern (ADR-097): domain event carries messageId directly.
     * Text-service looks up message from domain event - no separate action.success needed.
     */
    report(context) {
        const noun = context.command.directObject.entity;
        const sharedData = getClosingSharedData(context);
        const result = sharedData.closeResult;
        const events = [];
        // Add the backward compat domain event (closed)
        events.push(context.event('closed', {
            targetId: noun.id,
            targetName: noun.name,
            customMessage: result.closeMessage,
            sound: result.closeSound
        }));
        // Build domain data for the primary event
        const eventData = (0, data_builder_types_1.buildEventData)(closing_data_1.closedDataConfig, context);
        // Add additional fields for backward compatibility
        const isContainer = noun.has(world_model_1.TraitType.CONTAINER);
        const isDoor = noun.has(world_model_1.TraitType.DOOR);
        const isSupporter = noun.has(world_model_1.TraitType.SUPPORTER);
        let hasContents = false;
        let contentsCount = 0;
        let contentsIds = [];
        if (isContainer) {
            const contents = context.world.getContents(noun.id);
            hasContents = contents.length > 0;
            contentsCount = contents.length;
            contentsIds = contents.map(item => item.id);
        }
        // Emit domain event with messageId (simplified pattern - ADR-097)
        events.push(context.event('if.event.closed', {
            // Rendering data (messageId + params for text-service)
            messageId: `${context.action.id}.${closing_messages_1.ClosingMessages.CLOSED}`,
            params: { item: noun.name },
            // Domain data (for event sourcing / handlers)
            ...eventData,
            targetId: noun.id,
            targetName: noun.name,
            actorId: context.player.id,
            containerId: noun.id,
            containerName: noun.name,
            isContainer,
            isDoor,
            isSupporter,
            hasContents,
            contentsCount,
            contentsIds,
            item: noun.name
        }));
        return events;
    },
    /**
     * Generate events when validation fails
     *
     * Uses simplified event pattern (ADR-097): domain event carries messageId directly.
     */
    blocked(context, result) {
        const noun = context.command.directObject?.entity;
        // === ON-BLOCKED HOOK ===
        const sharedData = getClosingSharedData(context);
        const interceptor = sharedData.interceptor;
        const interceptorData = sharedData.interceptorData || {};
        if (interceptor?.onBlocked && noun && result.error) {
            const customEffects = interceptor.onBlocked(noun, context.world, context.player.id, result.error, interceptorData);
            if (customEffects !== null) {
                return customEffects.map(effect => context.event(effect.type, effect.payload));
            }
        }
        const error = result.error || '';
        // If error already contains dots (e.g., story interceptor ID), use as-is; otherwise prefix with action ID
        const messageId = error.includes('.') ? error : `${context.action.id}.${error}`;
        return [context.event('if.event.close_blocked', {
                // Rendering data
                messageId,
                params: {
                    ...result.params,
                    item: noun?.name
                },
                // Domain data
                targetId: noun?.id,
                targetName: noun?.name,
                reason: result.error
            })];
    }
};
