// @sharpee/engine/events/event-emitter.js
"use strict";
/**
 * Event emitter for managing game events
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventEmitter = void 0;
/**
 * Simple event emitter for the game event system
 * Used for story-level event handlers (daemons)
 */
class EventEmitter {
    handlers = new Map();
    /**
     * Register a handler for an event type
     */
    on(eventType, handler) {
        const handlers = this.handlers.get(eventType) || [];
        handlers.push(handler);
        this.handlers.set(eventType, handlers);
    }
    /**
     * Remove a handler for an event type
     */
    off(eventType, handler) {
        const handlers = this.handlers.get(eventType);
        if (handlers) {
            const index = handlers.indexOf(handler);
            if (index !== -1) {
                handlers.splice(index, 1);
            }
        }
    }
    /**
     * Emit an event to all registered handlers
     * Returns any semantic events generated by handlers
     */
    emit(event) {
        const results = [];
        const handlers = this.handlers.get(event.type) || [];
        for (const handler of handlers) {
            const result = handler(event);
            if (result) {
                results.push(...result);
            }
        }
        return results;
    }
    /**
     * Clear all handlers for a specific event type
     */
    clear(eventType) {
        if (eventType) {
            this.handlers.delete(eventType);
        }
        else {
            this.handlers.clear();
        }
    }
    /**
     * Get count of handlers for an event type
     */
    listenerCount(eventType) {
        return this.handlers.get(eventType)?.length || 0;
    }
}
exports.EventEmitter = EventEmitter;
