// @sharpee/plugin-scheduler/scheduler-service.js
"use strict";
/**
 * SchedulerService - Manages Daemons and Fuses (ADR-071)
 *
 * The scheduler runs during the turn cycle, after NPCs act:
 * 1. Player action
 * 2. NPC turns
 * 3. Scheduler tick (daemons run, fuses count down)
 * 4. Turn complete
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SchedulerService = void 0;
exports.createSchedulerService = createSchedulerService;
const seeded_random_1 = require("./seeded-random");
/**
 * Create a unique event ID
 */
function createEventId(prefix) {
    return `${prefix}_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
}
/**
 * SchedulerService implementation
 */
class SchedulerService {
    daemons = new Map();
    daemonStates = new Map();
    fuses = new Map();
    fuseStates = new Map();
    random;
    currentTurn = 0;
    constructor(seed) {
        this.random = (0, seeded_random_1.createSeededRandom)(seed);
    }
    // ==================== Daemon Management ====================
    registerDaemon(daemon) {
        if (this.daemons.has(daemon.id)) {
            throw new Error(`Daemon with id "${daemon.id}" already exists`);
        }
        this.daemons.set(daemon.id, daemon);
        this.daemonStates.set(daemon.id, {
            id: daemon.id,
            isPaused: false,
            runCount: 0,
        });
    }
    removeDaemon(id) {
        this.daemons.delete(id);
        this.daemonStates.delete(id);
    }
    pauseDaemon(id) {
        const state = this.daemonStates.get(id);
        if (state) {
            state.isPaused = true;
        }
    }
    resumeDaemon(id) {
        const state = this.daemonStates.get(id);
        if (state) {
            state.isPaused = false;
        }
    }
    hasDaemon(id) {
        return this.daemons.has(id);
    }
    // ==================== Fuse Management ====================
    setFuse(fuse) {
        // Store original turns for repeating fuses
        const fuseWithOriginal = {
            ...fuse,
            originalTurns: fuse.originalTurns ?? fuse.turns,
        };
        this.fuses.set(fuse.id, fuseWithOriginal);
        this.fuseStates.set(fuse.id, {
            id: fuse.id,
            turnsRemaining: fuse.turns,
            isPaused: false,
            entityId: fuse.entityId,
            skipNextTick: true,
        });
    }
    cancelFuse(id) {
        const fuse = this.fuses.get(id);
        const state = this.fuseStates.get(id);
        if (!fuse || !state) {
            return [];
        }
        // Call onCancel if defined
        let events = [];
        if (fuse.onCancel) {
            // Fuse cancellation may not have a world context available
            const context = this.createContext(undefined, this.currentTurn, '');
            try {
                events = fuse.onCancel(context);
            }
            catch (error) {
                console.error(`Error in fuse "${id}" onCancel:`, error);
            }
        }
        this.fuses.delete(id);
        this.fuseStates.delete(id);
        return events;
    }
    getFuseRemaining(id) {
        return this.fuseStates.get(id)?.turnsRemaining;
    }
    adjustFuse(id, delta) {
        const state = this.fuseStates.get(id);
        if (state) {
            state.turnsRemaining = Math.max(0, state.turnsRemaining + delta);
        }
    }
    pauseFuse(id) {
        const state = this.fuseStates.get(id);
        if (state) {
            state.isPaused = true;
        }
    }
    resumeFuse(id) {
        const state = this.fuseStates.get(id);
        if (state) {
            state.isPaused = false;
        }
    }
    hasFuse(id) {
        return this.fuses.has(id);
    }
    // ==================== Tick (Main Loop) ====================
    tick(world, turn, playerId) {
        this.currentTurn = turn;
        const events = [];
        const fusesTriggered = [];
        const daemonsRun = [];
        // Get player location
        const playerLocation = world.getLocation(playerId) || '';
        // Create context
        const context = this.createContext(world, turn, playerLocation, playerId);
        // 1. Run daemons (sorted by priority, highest first)
        const sortedDaemons = this.getSortedDaemons();
        for (const daemon of sortedDaemons) {
            const state = this.daemonStates.get(daemon.id);
            if (!state || state.isPaused)
                continue;
            // Check condition
            if (daemon.condition && !daemon.condition(context)) {
                continue;
            }
            // Run the daemon
            try {
                const daemonEvents = daemon.run(context);
                if (daemonEvents.length > 0) {
                    events.push(...daemonEvents);
                    daemonsRun.push(daemon.id);
                    state.runCount++;
                    // Remove if runOnce and produced events
                    if (daemon.runOnce) {
                        this.removeDaemon(daemon.id);
                    }
                }
            }
            catch (error) {
                console.error(`Error running daemon "${daemon.id}":`, error);
            }
        }
        // 2. Process fuses
        const fusesToRemove = [];
        const sortedFuses = this.getSortedFuses();
        for (const fuse of sortedFuses) {
            const state = this.fuseStates.get(fuse.id);
            if (!state || state.isPaused)
                continue;
            // Skip fuses set this turn (they start ticking next turn)
            if (state.skipNextTick) {
                state.skipNextTick = false;
                continue;
            }
            // Check tick condition
            if (fuse.tickCondition && !fuse.tickCondition(context)) {
                continue;
            }
            // Decrement turns
            state.turnsRemaining--;
            // Check if triggered
            if (state.turnsRemaining <= 0) {
                try {
                    const fuseEvents = fuse.trigger(context);
                    events.push(...fuseEvents);
                    fusesTriggered.push(fuse.id);
                    if (fuse.repeat && fuse.originalTurns) {
                        // Reset for repeat
                        state.turnsRemaining = fuse.originalTurns;
                    }
                    else {
                        // Mark for removal
                        fusesToRemove.push(fuse.id);
                    }
                }
                catch (error) {
                    console.error(`Error triggering fuse "${fuse.id}":`, error);
                    fusesToRemove.push(fuse.id);
                }
            }
        }
        // Remove triggered fuses
        for (const id of fusesToRemove) {
            this.fuses.delete(id);
            this.fuseStates.delete(id);
        }
        return { events, fusesTriggered, daemonsRun };
    }
    // ==================== Introspection ====================
    getActiveDaemons() {
        const infos = [];
        for (const [id, daemon] of this.daemons) {
            const state = this.daemonStates.get(id);
            if (!state)
                continue;
            infos.push({
                id: daemon.id,
                name: daemon.name,
                isPaused: state.isPaused,
                runCount: state.runCount,
                priority: daemon.priority ?? 0,
            });
        }
        return infos.sort((a, b) => b.priority - a.priority);
    }
    getActiveFuses() {
        const infos = [];
        for (const [id, fuse] of this.fuses) {
            const state = this.fuseStates.get(id);
            if (!state)
                continue;
            infos.push({
                id: fuse.id,
                name: fuse.name,
                turnsRemaining: state.turnsRemaining,
                isPaused: state.isPaused,
                entityId: state.entityId,
                priority: fuse.priority ?? 0,
                repeat: fuse.repeat ?? false,
            });
        }
        return infos.sort((a, b) => b.priority - a.priority);
    }
    // ==================== Serialization ====================
    getState() {
        const daemons = [];
        const fuses = [];
        for (const state of this.daemonStates.values()) {
            const daemon = this.daemons.get(state.id);
            const runnerState = daemon?.getRunnerState?.();
            daemons.push({ ...state, ...(runnerState !== undefined ? { runnerState } : {}) });
        }
        for (const state of this.fuseStates.values()) {
            fuses.push({ ...state });
        }
        return {
            turn: this.currentTurn,
            daemons,
            fuses,
            randomSeed: this.random.getSeed(),
        };
    }
    setState(state) {
        this.currentTurn = state.turn;
        this.random.setSeed(state.randomSeed);
        // Restore daemon states (daemons must be re-registered at game start)
        for (const daemonState of state.daemons) {
            if (this.daemons.has(daemonState.id)) {
                this.daemonStates.set(daemonState.id, { ...daemonState });
                if (daemonState.runnerState) {
                    this.daemons.get(daemonState.id)?.restoreRunnerState?.(daemonState.runnerState);
                }
            }
        }
        // Restore fuse states (fuses must be re-registered at game start)
        for (const fuseState of state.fuses) {
            if (this.fuses.has(fuseState.id)) {
                this.fuseStates.set(fuseState.id, { ...fuseState });
            }
        }
    }
    // ==================== Entity Cleanup ====================
    cleanupEntity(entityId) {
        const events = [];
        // Cancel any fuses bound to this entity
        for (const [id, state] of this.fuseStates) {
            if (state.entityId === entityId) {
                events.push(...this.cancelFuse(id));
            }
        }
        return events;
    }
    // ==================== Random Access ====================
    getRandom() {
        return this.random;
    }
    // ==================== Private Helpers ====================
    createContext(world, turn, playerLocation, playerId) {
        return {
            world,
            turn,
            random: this.random,
            playerLocation,
            playerId: playerId || '',
        };
    }
    getSortedDaemons() {
        return Array.from(this.daemons.values()).sort((a, b) => (b.priority ?? 0) - (a.priority ?? 0));
    }
    getSortedFuses() {
        return Array.from(this.fuses.values()).sort((a, b) => (b.priority ?? 0) - (a.priority ?? 0));
    }
}
exports.SchedulerService = SchedulerService;
/**
 * Create a new SchedulerService instance
 */
function createSchedulerService(seed) {
    return new SchedulerService(seed);
}
