// @sharpee/world-model/traits/identity/identityBehavior.js
"use strict";
// packages/world-model/src/traits/identity/identityBehavior.ts
Object.defineProperty(exports, "__esModule", { value: true });
exports.IdentityBehavior = void 0;
const behavior_1 = require("../../behaviors/behavior");
const trait_types_1 = require("../trait-types");
/**
 * Behavior for entities with identity.
 *
 * Handles name formatting, visibility, and validation of identity data.
 */
class IdentityBehavior extends behavior_1.Behavior {
    static requiredTraits = [trait_types_1.TraitType.IDENTITY];
    /**
     * Check if identity data is valid for game use
     * @throws Error if identity data is invalid
     */
    static validateIdentity(entity) {
        const identity = IdentityBehavior.require(entity, trait_types_1.TraitType.IDENTITY);
        if (!identity.name || !identity.name.trim()) {
            throw new Error(`Entity ${entity.id} has no name`);
        }
        if (identity.properName && identity.article) {
            // Proper names shouldn't have articles
            identity.article = '';
        }
        if (identity.weight !== undefined && identity.weight < 0) {
            throw new Error(`Entity ${entity.id} has negative weight`);
        }
        if (identity.volume !== undefined && identity.volume < 0) {
            throw new Error(`Entity ${entity.id} has negative volume`);
        }
        const validSizes = ['tiny', 'small', 'medium', 'large', 'huge'];
        if (identity.size !== undefined && !validSizes.includes(identity.size)) {
            throw new Error(`Entity ${entity.id} has invalid size: ${identity.size}`);
        }
    }
    /**
     * Format an entity's name for display with proper article
     */
    static formatName(entity, options) {
        const identity = IdentityBehavior.require(entity, trait_types_1.TraitType.IDENTITY);
        let name;
        if (identity.properName || !identity.article) {
            name = identity.name;
        }
        else {
            name = `${identity.article} ${identity.name}`;
        }
        // Handle definite article request
        if (options?.definite && !identity.properName && identity.article !== 'the') {
            name = `the ${identity.name}`;
        }
        // Handle capitalization
        if (options?.capitalize) {
            name = name.charAt(0).toUpperCase() + name.slice(1);
        }
        return name;
    }
    /**
     * Get a possessive form of the entity's name
     */
    static getPossessiveName(entity) {
        const name = IdentityBehavior.formatName(entity);
        // Handle names ending in 's'
        if (name.endsWith('s')) {
            return `${name}'`;
        }
        return `${name}'s`;
    }
    /**
     * Check if a given name matches this entity
     */
    static matchesName(entity, name) {
        const identity = IdentityBehavior.require(entity, trait_types_1.TraitType.IDENTITY);
        const lowerName = name.toLowerCase();
        // Check main name
        if (identity.name.toLowerCase() === lowerName)
            return true;
        // Check aliases
        return identity.aliases.some((alias) => alias.toLowerCase() === lowerName);
    }
    /**
     * Check if an entity should be visible in normal circumstances
     */
    static isConcealed(entity) {
        const identity = IdentityBehavior.require(entity, trait_types_1.TraitType.IDENTITY);
        return identity.concealed;
    }
    /**
     * Set whether an entity is concealed
     */
    static setConcealed(entity, concealed) {
        const identity = IdentityBehavior.require(entity, trait_types_1.TraitType.IDENTITY);
        identity.concealed = concealed;
    }
    /**
     * Reveal a concealed entity (convenience method)
     */
    static reveal(entity) {
        IdentityBehavior.setConcealed(entity, false);
    }
    /**
     * Conceal an entity (convenience method)
     */
    static conceal(entity) {
        IdentityBehavior.setConcealed(entity, true);
    }
    /**
     * Get the weight of the entity (not including contents)
     */
    static getWeight(entity) {
        if (!entity.has(trait_types_1.TraitType.IDENTITY)) {
            return 0; // Default weight for items without identity trait
        }
        const identity = entity.get(trait_types_1.TraitType.IDENTITY);
        return identity?.weight || 0;
    }
    /**
     * Get the volume of the entity
     */
    static getVolume(entity) {
        if (!entity.has(trait_types_1.TraitType.IDENTITY)) {
            return 0; // Default volume for items without identity trait
        }
        const identity = entity.get(trait_types_1.TraitType.IDENTITY);
        return identity?.volume || 0;
    }
    /**
     * Get the size category of the entity
     */
    static getSize(entity) {
        const identity = IdentityBehavior.require(entity, trait_types_1.TraitType.IDENTITY);
        return identity.size;
    }
    /**
     * Get total weight including contents (for containers/supporters)
     * Note: This method needs a world context to access contents
     */
    static getTotalWeight(entity, getContents) {
        const identity = IdentityBehavior.require(entity, trait_types_1.TraitType.IDENTITY);
        let weight = identity.weight || 0;
        // Add weight of contents if this is a container/supporter and we have a way to get contents
        if (getContents && (entity.has(trait_types_1.TraitType.CONTAINER) || entity.has(trait_types_1.TraitType.SUPPORTER))) {
            const contents = getContents(entity.id);
            for (const item of contents) {
                weight += IdentityBehavior.getTotalWeight(item, getContents);
            }
        }
        return weight;
    }
}
exports.IdentityBehavior = IdentityBehavior;
