// @sharpee/core/debug/types.js
"use strict";
/**
 * Debug event types for diagnostic information.
 * These are separate from semantic events and used for development/debugging.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DebugEventTypes = void 0;
/**
 * Common debug event types by subsystem
 */
exports.DebugEventTypes = {
    parser: {
        TOKEN_ANALYSIS: 'token_analysis',
        PATTERN_MATCH: 'pattern_match',
        CANDIDATE_SCORING: 'candidate_scoring',
    },
    validator: {
        ENTITY_RESOLUTION: 'entity_resolution',
        SCOPE_CHECK: 'scope_check',
        AMBIGUITY_RESOLUTION: 'ambiguity_resolution',
        VALIDATION_ERROR: 'validation_error',
    },
    executor: {
        ACTION_START: 'action_start',
        ACTION_COMPLETE: 'action_complete',
        ACTION_ERROR: 'action_error',
    },
    worldModel: {
        ENTITY_CHANGE: 'entity_change',
        RELATION_CHANGE: 'relation_change',
        PROPERTY_CHANGE: 'property_change',
    },
    textService: {
        TEMPLATE_SELECTION: 'template_selection',
        TEXT_GENERATION: 'text_generation',
        CHANNEL_ROUTING: 'channel_routing',
    },
};
