// @sharpee/stdlib/actions/standard/about/about.js
"use strict";
/**
 * About action - displays information about the game
 *
 * This is a meta action that signals the text service to display
 * game information. The text service will query the story config
 * directly for the information it needs.
 *
 * Uses four-phase pattern:
 * 1. validate: Always succeeds (no preconditions)
 * 2. execute: No world mutations
 * 3. blocked: Handle validation failures (n/a - always succeeds)
 * 4. report: Emit about event
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.aboutAction = void 0;
const world_model_1 = require("../../../../world-model/index.js");
const constants_1 = require("../../constants");
exports.aboutAction = {
    id: constants_1.IFActions.ABOUT,
    validate(context) {
        // About action always succeeds
        return {
            valid: true
        };
    },
    execute(context) {
        // No state mutations needed for this meta action
        // The text service will construct output from story config
    },
    blocked(context, result) {
        // About always succeeds, but include blocked for consistency
        return [context.event('if.event.about_displayed', {
                messageId: `if.action.about.${result.error}`,
                params: result.params || {},
                blocked: true,
                reason: result.error
            })];
    },
    report(context) {
        const world = context.world;
        const storyInfoEntities = world.findByTrait(world_model_1.TraitType.STORY_INFO);
        const storyConfig = storyInfoEntities[0]?.getTrait(world_model_1.StoryInfoTrait) || {};
        const eventData = {
            messageId: 'if.action.about.success',
            params: {
                title: storyConfig.title || 'Unknown',
                author: Array.isArray(storyConfig.author) ? storyConfig.author.join(', ') : (storyConfig.author || 'Unknown'),
                version: storyConfig.version || '0.0.0',
                description: storyConfig.description || '',
                engineVersion: storyConfig.engineVersion || '',
                buildDate: storyConfig.buildDate || '',
                clientVersion: storyConfig.clientVersion || '',
                portedBy: storyConfig.portedBy || ''
            }
        };
        return [
            context.event('if.event.about_displayed', eventData)
        ];
    },
    group: "meta",
    metadata: {
        requiresDirectObject: false,
        requiresIndirectObject: false
    }
};
