// @sharpee/world-model/traits/clothing/clothingTrait.js
"use strict";
// packages/world-model/src/traits/clothing/clothingTrait.ts
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClothingTrait = void 0;
const trait_types_1 = require("../trait-types");
/**
 * ClothingTrait is a specialized wearable trait for clothing items.
 * Clothing items (coats, pants, dresses) can have pockets and other special properties.
 *
 * This trait includes all WearableData properties but is a separate trait type
 * to allow for clothing-specific behaviors and queries.
 *
 * Pockets should be created as separate container entities with SceneryTrait
 * and placed inside the clothing item.
 *
 * @example
 * ```typescript
 * const coat = world.createEntity('Winter Coat', 'item');
 * coat.add(new ClothingTrait({ slot: 'torso', material: 'wool' }));
 * coat.add(new ContainerTrait()); // So it can contain pockets
 *
 * const pocket = world.createEntity('inside pocket', 'container');
 * pocket.add(new ContainerTrait({ capacity: 3 }));
 * pocket.add(new SceneryTrait({ cantTakeMessage: "The pocket is sewn into the coat." }));
 * world.moveEntity(pocket.id, coat.id);
 * ```
 */
class ClothingTrait {
    static type = trait_types_1.TraitType.CLOTHING;
    type = trait_types_1.TraitType.CLOTHING;
    // WearableData properties
    worn = false;
    wornBy;
    slot;
    layer;
    wearMessage;
    removeMessage;
    wearableOver;
    blocksSlots;
    weight;
    bulk;
    canRemove;
    bodyPart;
    // ClothingData properties
    material;
    style;
    damageable;
    condition;
    // Public accessor for consistency with WearableTrait
    get isWorn() {
        return this.worn;
    }
    set isWorn(value) {
        this.worn = value;
    }
    constructor(data = {}) {
        // Set wearable defaults and merge with provided data
        this.worn = data.isWorn ?? false;
        this.wornBy = data.wornBy;
        this.slot = data.slot ?? 'clothing';
        this.layer = data.layer ?? 1;
        this.wearMessage = data.wearMessage;
        this.removeMessage = data.removeMessage;
        this.wearableOver = data.wearableOver ?? true;
        this.blocksSlots = data.blocksSlots ?? [];
        this.weight = data.weight ?? 1;
        this.bulk = data.bulk ?? 1;
        this.canRemove = data.canRemove ?? true;
        this.bodyPart = data.bodyPart ?? 'torso';
        // Set clothing-specific defaults
        this.material = data.material ?? 'fabric';
        this.style = data.style ?? 'casual';
        this.damageable = data.damageable ?? true;
        this.condition = data.condition ?? 'good';
    }
}
exports.ClothingTrait = ClothingTrait;
