// @sharpee/text-service/index.js
"use strict";
/**
 * @sharpee/text-service
 *
 * Single text service for Sharpee that:
 * - Receives semantic events
 * - Resolves templates via LanguageProvider
 * - Parses decorations into structured IDecoration tree
 * - Outputs ITextBlock[]
 *
 * Includes CLI renderer for testing and CLI clients.
 *
 * Inspired by FyreVM channel I/O (2009).
 *
 * @see ADR-096 Text Service Architecture
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CORE_DECORATION_TYPES = exports.BLOCK_KEY_PREFIXES = exports.BLOCK_KEYS = exports.extractPlainText = exports.isActionBlock = exports.isRoomBlock = exports.isStatusBlock = exports.isTextBlock = exports.isDecoration = exports.renderStatusLine = exports.renderToString = exports.hasDecorations = exports.parseDecorations = exports.handleGenericEvent = exports.handleGameMessage = exports.handleRevealed = exports.handleRoomDescription = exports.extractValue = exports.createBlock = exports.getChainMetadata = exports.sortEventsForProse = exports.filterEvents = exports.createTextService = exports.TextService = void 0;
var text_service_js_1 = require("./text-service.js");
Object.defineProperty(exports, "TextService", { enumerable: true, get: function () { return text_service_js_1.TextService; } });
Object.defineProperty(exports, "createTextService", { enumerable: true, get: function () { return text_service_js_1.createTextService; } });
// Pipeline stages
var filter_js_1 = require("./stages/filter.js");
Object.defineProperty(exports, "filterEvents", { enumerable: true, get: function () { return filter_js_1.filterEvents; } });
var sort_js_1 = require("./stages/sort.js");
Object.defineProperty(exports, "sortEventsForProse", { enumerable: true, get: function () { return sort_js_1.sortEventsForProse; } });
Object.defineProperty(exports, "getChainMetadata", { enumerable: true, get: function () { return sort_js_1.getChainMetadata; } });
var assemble_js_1 = require("./stages/assemble.js");
Object.defineProperty(exports, "createBlock", { enumerable: true, get: function () { return assemble_js_1.createBlock; } });
Object.defineProperty(exports, "extractValue", { enumerable: true, get: function () { return assemble_js_1.extractValue; } });
var room_js_1 = require("./handlers/room.js");
Object.defineProperty(exports, "handleRoomDescription", { enumerable: true, get: function () { return room_js_1.handleRoomDescription; } });
var revealed_js_1 = require("./handlers/revealed.js");
Object.defineProperty(exports, "handleRevealed", { enumerable: true, get: function () { return revealed_js_1.handleRevealed; } });
var generic_js_1 = require("./handlers/generic.js");
Object.defineProperty(exports, "handleGameMessage", { enumerable: true, get: function () { return generic_js_1.handleGameMessage; } });
Object.defineProperty(exports, "handleGenericEvent", { enumerable: true, get: function () { return generic_js_1.handleGenericEvent; } });
// Decoration parser
var decoration_parser_js_1 = require("./decoration-parser.js");
Object.defineProperty(exports, "parseDecorations", { enumerable: true, get: function () { return decoration_parser_js_1.parseDecorations; } });
Object.defineProperty(exports, "hasDecorations", { enumerable: true, get: function () { return decoration_parser_js_1.hasDecorations; } });
var cli_renderer_js_1 = require("./cli-renderer.js");
Object.defineProperty(exports, "renderToString", { enumerable: true, get: function () { return cli_renderer_js_1.renderToString; } });
Object.defineProperty(exports, "renderStatusLine", { enumerable: true, get: function () { return cli_renderer_js_1.renderStatusLine; } });
var text_blocks_1 = require("../text-blocks/index.js");
Object.defineProperty(exports, "isDecoration", { enumerable: true, get: function () { return text_blocks_1.isDecoration; } });
Object.defineProperty(exports, "isTextBlock", { enumerable: true, get: function () { return text_blocks_1.isTextBlock; } });
Object.defineProperty(exports, "isStatusBlock", { enumerable: true, get: function () { return text_blocks_1.isStatusBlock; } });
Object.defineProperty(exports, "isRoomBlock", { enumerable: true, get: function () { return text_blocks_1.isRoomBlock; } });
Object.defineProperty(exports, "isActionBlock", { enumerable: true, get: function () { return text_blocks_1.isActionBlock; } });
Object.defineProperty(exports, "extractPlainText", { enumerable: true, get: function () { return text_blocks_1.extractPlainText; } });
Object.defineProperty(exports, "BLOCK_KEYS", { enumerable: true, get: function () { return text_blocks_1.BLOCK_KEYS; } });
Object.defineProperty(exports, "BLOCK_KEY_PREFIXES", { enumerable: true, get: function () { return text_blocks_1.BLOCK_KEY_PREFIXES; } });
Object.defineProperty(exports, "CORE_DECORATION_TYPES", { enumerable: true, get: function () { return text_blocks_1.CORE_DECORATION_TYPES; } });
