// @sharpee/text-service/text-service.js
"use strict";
/**
 * Text Service
 *
 * Orchestrates the text output pipeline:
 * 1. Filter - remove system events
 * 2. Sort - order events for prose (ADR-094)
 * 3. Process - route to handlers
 * 4. Assemble - create ITextBlock with decorations
 *
 * Stateless transformer: events in, TextBlocks out.
 * Inspired by FyreVM channel I/O (2009).
 *
 * @see ADR-096 Text Service Architecture
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TextService = void 0;
exports.createTextService = createTextService;
const text_blocks_1 = require("../text-blocks/index.js");
// Pipeline stages
const filter_js_1 = require("./stages/filter.js");
const sort_js_1 = require("./stages/sort.js");
const assemble_js_1 = require("./stages/assemble.js");
const room_js_1 = require("./handlers/room.js");
const revealed_js_1 = require("./handlers/revealed.js");
const generic_js_1 = require("./handlers/generic.js");
const game_js_1 = require("./handlers/game.js");
const help_js_1 = require("./handlers/help.js");
const about_js_1 = require("./handlers/about.js");
/**
 * TextService implementation
 *
 * Orchestrates the pipeline: filter → sort → process → assemble
 */
class TextService {
    languageProvider;
    constructor(languageProvider) {
        this.languageProvider = languageProvider;
    }
    processTurn(events) {
        // Pipeline: filter → sort → process
        const filtered = (0, filter_js_1.filterEvents)(events);
        const sorted = (0, sort_js_1.sortEventsForProse)(filtered);
        // Process each event through handlers
        const handlerContext = {
            languageProvider: this.languageProvider,
        };
        const blocks = [];
        for (const event of sorted) {
            const eventBlocks = this.routeToHandler(event, handlerContext);
            blocks.push(...eventBlocks);
        }
        return blocks;
    }
    /**
     * Route event to appropriate handler
     */
    routeToHandler(event, context) {
        // Domain events with messageId (ADR-097): resolve via language provider directly.
        const result = this.tryProcessDomainEventMessage(event, context);
        if (result) {
            return result;
        }
        switch (event.type) {
            case 'game.started':
                return (0, game_js_1.handleGameStarted)(event, context);
            case 'if.event.room_description':
            case 'if.event.room.description':
                return (0, room_js_1.handleRoomDescription)(event, context);
            case 'game.message':
                return (0, generic_js_1.handleGameMessage)(event, context);
            case 'if.event.revealed':
                return (0, revealed_js_1.handleRevealed)(event, context);
            case 'if.event.help_displayed':
                return (0, help_js_1.handleHelpDisplayed)(event, context);
            case 'if.event.about_displayed':
                return (0, about_js_1.handleAboutDisplayed)(event, context);
            case 'if.event.implicit_take':
                return this.handleImplicitTake(event, context);
            case 'command.failed':
                return this.handleCommandFailed(event, context);
            case 'client.query':
                return this.handleClientQuery(event, context);
            default:
                return (0, generic_js_1.handleGenericEvent)(event, context);
        }
    }
    /**
     * Process domain events that carry messageId directly (ADR-097).
     *
     * All stdlib actions use this pattern. Story actions that emit action.success
     * or action.blocked events fall through to the switch-case handlers below.
     *
     * @returns Text blocks if event has messageId. If the messageId doesn't resolve,
     *   emits a visible error block rather than silently producing nothing.
     */
    tryProcessDomainEventMessage(event, context) {
        const data = event.data;
        // No messageId = fall through to switch-case handlers
        if (!data?.messageId) {
            return null;
        }
        // Skip client.query - needs specialized handling (disambiguation builds options from candidates)
        if (event.type === 'client.query') {
            return null;
        }
        // Look up message via language provider
        if (!context.languageProvider) {
            return null;
        }
        const message = context.languageProvider.getMessage(data.messageId, data.params);
        // If message wasn't found (returns the messageId) or resolved to empty
        // (placeholder registration), fall back to inline text.
        // Many events carry both messageId and pre-rendered text (melee combat, GDT, etc.)
        if (message === data.messageId || !message) {
            const fallback = data.message ?? data.text;
            if (typeof fallback === 'string' && fallback) {
                return [(0, assemble_js_1.createBlock)(text_blocks_1.BLOCK_KEYS.ACTION_RESULT, fallback)];
            }
            // No inline text — skip silently rather than showing an error.
            // Domain events (if.event.*) carry messageId for semantic association,
            // not for text rendering. The actual text comes from game.message events.
            return null;
        }
        // Determine block key based on event type
        const blockKey = event.type.includes('blocked') || event.type.includes('failure')
            ? text_blocks_1.BLOCK_KEYS.ACTION_BLOCKED
            : text_blocks_1.BLOCK_KEYS.ACTION_RESULT;
        return [(0, assemble_js_1.createBlock)(blockKey, message)];
    }
    /**
     * Handle if.event.implicit_take events
     * Produces "(first taking the X)" message
     */
    handleImplicitTake(event, _context) {
        const data = event.data;
        const itemName = data.itemName || 'something';
        return [(0, assemble_js_1.createBlock)(text_blocks_1.BLOCK_KEYS.ACTION_RESULT, `(first taking the ${itemName})`)];
    }
    /**
     * Handle command.failed events
     * These occur when parsing or entity resolution fails
     */
    handleCommandFailed(event, context) {
        const data = event.data;
        // Try to get a user-friendly message based on the reason
        if (data.reason) {
            // Check for specific failure reasons
            if (data.reason.includes('ENTITY_NOT_FOUND') || data.reason.includes('modifiers_not_matched')) {
                // Entity resolution failed - player referred to something that doesn't exist or can't be found
                const message = context.languageProvider?.getMessage('core.entity_not_found')
                    ?? "I don't see that here.";
                return [(0, assemble_js_1.createBlock)(text_blocks_1.BLOCK_KEYS.ERROR, message)];
            }
            // Note: AMBIGUOUS_ENTITY now uses client.query event, not command.failed
            if (data.reason.includes('NO_MATCH') || data.reason.includes('parse')) {
                const message = context.languageProvider?.getMessage('core.command_not_understood')
                    ?? "I don't understand that.";
                return [(0, assemble_js_1.createBlock)(text_blocks_1.BLOCK_KEYS.ERROR, message)];
            }
        }
        // Generic fallback
        const message = context.languageProvider?.getMessage('core.command_failed')
            ?? "I don't understand that.";
        return [(0, assemble_js_1.createBlock)(text_blocks_1.BLOCK_KEYS.ERROR, message)];
    }
    /**
     * Handle client.query events (disambiguation, confirmations, etc.)
     */
    handleClientQuery(event, context) {
        const data = event.data;
        // Only handle disambiguation queries here
        if (data.source !== 'disambiguation') {
            return [];
        }
        // Format candidates as natural list
        const candidateNames = (data.candidates || []).map(c => c.name);
        const options = this.formatCandidateList(candidateNames);
        // Get message template with options
        const message = context.languageProvider?.getMessage('core.disambiguation_prompt', { options })
            ?? `Which do you mean: ${options}?`;
        return [(0, assemble_js_1.createBlock)(text_blocks_1.BLOCK_KEYS.ERROR, message)];
    }
    /**
     * Format a list of candidate names as natural English
     * e.g., "the red ball or the blue ball" or "the sword, the axe, or the knife"
     */
    formatCandidateList(names) {
        if (names.length === 0)
            return '';
        if (names.length === 1)
            return `the ${names[0]}`;
        // Add "the" article to each name
        const withArticles = names.map(n => `the ${n}`);
        if (withArticles.length === 2) {
            return `${withArticles[0]} or ${withArticles[1]}`;
        }
        // Oxford comma style: "the X, the Y, or the Z"
        const last = withArticles.pop();
        return `${withArticles.join(', ')}, or ${last}`;
    }
}
exports.TextService = TextService;
/**
 * Create a TextService with the given LanguageProvider.
 * LanguageProvider supplies templates (standard + story-registered).
 *
 * @param languageProvider - Provider for template resolution
 * @returns Configured TextService instance
 */
function createTextService(languageProvider) {
    return new TextService(languageProvider);
}
