// @sharpee/character/propagation/index.js
"use strict";
/**
 * Information propagation exports (ADR-144)
 *
 * Propagation profiles, evaluation engine, and fact transfer
 * for NPC-to-NPC information flow.
 *
 * Public interface: All re-exported types and functions.
 * Owner context: @sharpee/character / propagation
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PropagationMessages = exports.buildPropagationProfile = exports.resolvePlayerPresence = exports.getVisibilityResults = exports.getVisibilityResult = exports.PROPAGATION_WITNESSED_DEFAULTS = exports.applyTransfers = exports.transferFact = exports.evaluatePropagation = exports.AlreadyToldRecord = void 0;
var propagation_types_1 = require("./propagation-types");
Object.defineProperty(exports, "AlreadyToldRecord", { enumerable: true, get: function () { return propagation_types_1.AlreadyToldRecord; } });
var propagation_evaluator_1 = require("./propagation-evaluator");
Object.defineProperty(exports, "evaluatePropagation", { enumerable: true, get: function () { return propagation_evaluator_1.evaluatePropagation; } });
var fact_transfer_1 = require("./fact-transfer");
Object.defineProperty(exports, "transferFact", { enumerable: true, get: function () { return fact_transfer_1.transferFact; } });
Object.defineProperty(exports, "applyTransfers", { enumerable: true, get: function () { return fact_transfer_1.applyTransfers; } });
var visibility_1 = require("./visibility");
Object.defineProperty(exports, "PROPAGATION_WITNESSED_DEFAULTS", { enumerable: true, get: function () { return visibility_1.PROPAGATION_WITNESSED_DEFAULTS; } });
Object.defineProperty(exports, "getVisibilityResult", { enumerable: true, get: function () { return visibility_1.getVisibilityResult; } });
Object.defineProperty(exports, "getVisibilityResults", { enumerable: true, get: function () { return visibility_1.getVisibilityResults; } });
Object.defineProperty(exports, "resolvePlayerPresence", { enumerable: true, get: function () { return visibility_1.resolvePlayerPresence; } });
var builder_1 = require("./builder");
Object.defineProperty(exports, "buildPropagationProfile", { enumerable: true, get: function () { return builder_1.buildPropagationProfile; } });
var propagation_messages_1 = require("./propagation-messages");
Object.defineProperty(exports, "PropagationMessages", { enumerable: true, get: function () { return propagation_messages_1.PropagationMessages; } });
