// @sharpee/character/propagation/propagation-evaluator.js
"use strict";
/**
 * Propagation evaluation engine (ADR-144)
 *
 * Per-turn evaluator that determines which facts each NPC shares,
 * with whom, and in what order. Pure evaluation logic — does not
 * mutate world state. Returns PropagationTransfer objects that
 * the caller applies via fact-transfer.
 *
 * Public interface: evaluatePropagation, PropagationContext.
 * Owner context: @sharpee/character / propagation
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.evaluatePropagation = evaluatePropagation;
// ---------------------------------------------------------------------------
// Evaluation
// ---------------------------------------------------------------------------
/**
 * Evaluate propagation for a single NPC.
 *
 * Algorithm:
 * 1. Mute check — skip entirely
 * 2. Schedule condition check — skip if not met
 * 3. Find eligible listeners (audience + exclusions)
 * 4. Find eligible facts (tendency whitelist/blacklist + already-told)
 * 5. Apply pace (eager = all, gradual = one, reluctant = wait)
 * 6. Return transfer objects
 *
 * @param ctx - The propagation context
 * @returns Array of transfers to execute
 */
function evaluatePropagation(ctx) {
    const { speaker } = ctx;
    const profile = speaker.profile;
    // No profile or mute → no propagation
    if (!profile || profile.tendency === 'mute') {
        return [];
    }
    // Schedule check
    if (profile.schedule) {
        const scheduleMet = profile.schedule.when.every(cond => speaker.trait.evaluate(cond));
        if (!scheduleMet)
            return [];
    }
    // Find eligible listeners
    const eligibleListeners = findEligibleListeners(ctx);
    if (eligibleListeners.length === 0)
        return [];
    // Find eligible facts for each listener
    const transfers = [];
    for (const listener of eligibleListeners) {
        const facts = findEligibleFacts(ctx, listener);
        if (facts.length === 0)
            continue;
        // Apply pace
        const factsToShare = applyPace(facts, profile, ctx);
        for (const topic of factsToShare) {
            const override = profile.overrides?.[topic];
            const version = override?.spreadsVersion ?? 'truth';
            const coloring = profile.coloring ?? 'neutral';
            transfers.push({
                speakerId: speaker.id,
                listenerId: listener.id,
                topic,
                version,
                coloring,
                witnessedOverride: override?.witnessed,
            });
        }
    }
    return transfers;
}
// ---------------------------------------------------------------------------
// Eligible listeners
// ---------------------------------------------------------------------------
/**
 * Find NPCs in the room that the speaker will share with.
 * Filters by audience type and exclusion list.
 */
function findEligibleListeners(ctx) {
    const { speaker, listeners } = ctx;
    const profile = speaker.profile;
    const audience = profile.audience ?? 'trusted';
    const excludes = new Set(profile.excludes ?? []);
    return listeners.filter(listener => {
        // Excluded by name
        if (excludes.has(listener.id))
            return false;
        // Audience check
        switch (audience) {
            case 'anyone':
                return true;
            case 'trusted':
                // Speaker has positive disposition toward listener
                return speaker.trait.getDispositionValue(listener.id) > 0;
            case 'allied':
                // Both share a loyalty — check if any disposition targets overlap
                // with positive values on both sides
                return hasSharedAlliance(speaker.trait, listener.trait);
            default:
                return false;
        }
    });
}
/**
 * Check if two NPCs share a loyalty (both have positive disposition
 * toward at least one common entity).
 */
function hasSharedAlliance(traitA, traitB) {
    for (const [entityId, valueA] of Object.entries(traitA.dispositions)) {
        if (valueA > 50 && (traitB.dispositions[entityId] ?? 0) > 50) {
            return true;
        }
    }
    return false;
}
// ---------------------------------------------------------------------------
// Eligible facts
// ---------------------------------------------------------------------------
/**
 * Find facts the speaker is willing and able to share with a specific listener.
 * Filters by tendency (chatty/selective), already-told, and player-leverage.
 */
function findEligibleFacts(ctx, listener) {
    const { speaker, alreadyTold } = ctx;
    const profile = speaker.profile;
    const knowledge = speaker.trait.knowledge;
    const withholds = new Set(profile.withholds ?? []);
    const spreadsSet = new Set(profile.spreads ?? []);
    const eligible = [];
    for (const topic of Object.keys(knowledge)) {
        if (alreadyTold.hasTold(speaker.id, listener.id, topic))
            continue;
        if (!meetsAudienceRule(topic, profile, speaker.trait, listener))
            continue;
        if (!meetsTendencyFilter(topic, profile, withholds, spreadsSet))
            continue;
        if (!meetsLeverageRule(topic, profile, knowledge))
            continue;
        eligible.push(topic);
    }
    return eligible;
}
/**
 * Check whether a per-fact audience override allows sharing with this listener.
 * Returns true if there is no override or the override audience matches.
 *
 * @param topic - The knowledge topic to check
 * @param profile - Speaker's propagation profile
 * @param speakerTrait - Speaker's character model trait
 * @param listener - The potential listener
 * @returns Whether the topic passes the audience rule
 */
function meetsAudienceRule(topic, profile, speakerTrait, listener) {
    const override = profile.overrides?.[topic];
    if (!override?.to)
        return true;
    return checkAudienceForListener(override.to, speakerTrait, listener);
}
/**
 * Check whether a topic passes the speaker's tendency filter.
 * Chatty speakers share everything except withheld topics; selective
 * speakers share only explicitly listed topics.
 *
 * @param topic - The knowledge topic to check
 * @param profile - Speaker's propagation profile
 * @param withholds - Pre-computed set of withheld topics
 * @param spreadsSet - Pre-computed set of explicitly shared topics
 * @returns Whether the topic passes the tendency filter
 */
function meetsTendencyFilter(topic, profile, withholds, spreadsSet) {
    switch (profile.tendency) {
        case 'chatty':
            return !withholds.has(topic);
        case 'selective':
            return spreadsSet.has(topic);
        default:
            return true;
    }
}
/**
 * Check whether a topic passes the player-leverage filter.
 * When playerCanLeverage is false, facts sourced from the player are blocked.
 *
 * @param topic - The knowledge topic to check
 * @param profile - Speaker's propagation profile
 * @param knowledge - Speaker's knowledge map
 * @returns Whether the topic passes the leverage rule
 */
function meetsLeverageRule(topic, profile, knowledge) {
    if (profile.playerCanLeverage)
        return true;
    const fact = knowledge[topic];
    if (fact && (fact.source === 'told' || fact.source === 'told by player')) {
        return false;
    }
    return true;
}
/**
 * Check if a listener matches a specific audience type for a per-fact override.
 */
function checkAudienceForListener(audience, speakerTrait, listener) {
    switch (audience) {
        case 'anyone': return true;
        case 'trusted': return speakerTrait.getDispositionValue(listener.id) > 0;
        case 'allied': return hasSharedAlliance(speakerTrait, listener.trait);
    }
}
// ---------------------------------------------------------------------------
// Pace application
// ---------------------------------------------------------------------------
/**
 * Apply pace to limit how many facts are shared per turn.
 */
function applyPace(eligibleFacts, profile, ctx) {
    const pace = profile.pace ?? 'eager';
    switch (pace) {
        case 'eager':
            // Share all eligible facts at once
            return eligibleFacts;
        case 'gradual':
            // One fact per turn
            return eligibleFacts.slice(0, 1);
        case 'reluctant':
            // Requires multiple turns colocated before sharing
            if ((ctx.turnsColocated ?? 0) < 3)
                return [];
            return eligibleFacts.slice(0, 1);
    }
}
