// @sharpee/character/propagation/visibility.js
"use strict";
/**
 * Propagation visibility (ADR-144)
 *
 * Determines what the player sees when NPC-to-NPC information
 * transfer occurs. Three modes: offscreen (absent), witnessed
 * (present), eavesdropped (concealed).
 *
 * Public interface: PropagationVisibility, getVisibilityMessage.
 * Owner context: @sharpee/character / propagation
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PROPAGATION_WITNESSED_DEFAULTS = void 0;
exports.getVisibilityResult = getVisibilityResult;
exports.getVisibilityResults = getVisibilityResults;
exports.resolvePlayerPresence = resolvePlayerPresence;
// ---------------------------------------------------------------------------
// Platform defaults per coloring
// ---------------------------------------------------------------------------
/**
 * Platform default witnessed message IDs per coloring.
 * Authors override per fact via FactOverride.witnessed.
 */
exports.PROPAGATION_WITNESSED_DEFAULTS = {
    neutral: 'character.propagation.witnessed.neutral',
    dramatic: 'character.propagation.witnessed.dramatic',
    vague: 'character.propagation.witnessed.vague',
    fearful: 'character.propagation.witnessed.fearful',
    conspiratorial: 'character.propagation.witnessed.conspiratorial',
};
// ---------------------------------------------------------------------------
// Visibility evaluation
// ---------------------------------------------------------------------------
/**
 * Determine visibility output for a propagation transfer.
 *
 * @param transfer - The propagation transfer
 * @param presence - The player's presence state
 * @returns Visibility result with message ID and player-learns flag
 */
function getVisibilityResult(transfer, presence) {
    switch (presence) {
        case 'absent':
            // Offscreen — state mutation only, no message
            return {
                transfer,
                presence,
                playerLearns: false,
            };
        case 'present':
            // Witnessed — player sees a summary message
            return {
                transfer,
                presence,
                messageId: transfer.witnessedOverride
                    ?? exports.PROPAGATION_WITNESSED_DEFAULTS[transfer.coloring],
                playerLearns: false,
            };
        case 'concealed':
            // Eavesdropped — player sees full dialogue and learns the fact
            return {
                transfer,
                presence,
                messageId: transfer.witnessedOverride
                    ?? exports.PROPAGATION_WITNESSED_DEFAULTS[transfer.coloring],
                playerLearns: true,
            };
    }
}
/**
 * Evaluate visibility for multiple transfers.
 *
 * @param transfers - The transfers to evaluate
 * @param presence - The player's presence state
 * @returns Array of visibility results
 */
function getVisibilityResults(transfers, presence) {
    return transfers.map(t => getVisibilityResult(t, presence));
}
// ---------------------------------------------------------------------------
// Player presence resolution (ADR-148)
// ---------------------------------------------------------------------------
/**
 * Determine the player's presence state relative to an NPC.
 *
 * Used by the propagation evaluator to decide what the player observes:
 * - `absent`: different room — state mutation only, no message
 * - `present`: same room, visible — witnessed summary
 * - `concealed`: same room, hidden — full eavesdropping, player learns the fact
 *
 * NPC-to-player visibility (can the NPC see the player?) is handled separately
 * by ConcealedVisibilityBehavior via the canSee() pipeline.
 *
 * @param world - The world model
 * @param playerId - The player entity ID
 * @param npcId - The NPC entity ID
 * @returns The player's presence state
 */
function resolvePlayerPresence(world, playerId, npcId) {
    const playerRoom = world.getContainingRoom(playerId);
    const npcRoom = world.getContainingRoom(npcId);
    // Different room → absent
    if (!playerRoom || !npcRoom || playerRoom.id !== npcRoom.id) {
        return 'absent';
    }
    // Same room but concealed → eavesdropping
    const player = world.getEntity(playerId);
    if (player?.has('if.trait.concealed_state')) {
        return 'concealed';
    }
    // Same room and visible → witnessed
    return 'present';
}
