// @sharpee/platform-browser/BrowserClient.js
"use strict";
/**
 * BrowserClient - Main orchestrator for browser-based IF games
 *
 * Wires together all managers (save, theme, dialog, menu, input, display)
 * and provides a simple API for story entry points.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.BrowserClient = void 0;
const world_model_1 = require("../world-model/index.js");
const text_service_1 = require("../text-service/index.js");
const ThemeManager_1 = require("./managers/ThemeManager");
const SaveManager_1 = require("./managers/SaveManager");
const DialogManager_1 = require("./managers/DialogManager");
const MenuManager_1 = require("./managers/MenuManager");
const InputManager_1 = require("./managers/InputManager");
const TextDisplay_1 = require("./display/TextDisplay");
const StatusLine_1 = require("./display/StatusLine");
const AudioManager_1 = require("./audio/AudioManager");
class BrowserClient {
    constructor(config) {
        // Game state
        this.currentTurn = 0;
        this.currentScore = 0;
        this.turnOffset = 0; // Offset to add to engine turn after restore
        this.config = {
            autoSave: true,
            ...config,
        };
        this.audioManager = new AudioManager_1.AudioManager();
    }
    /**
     * Initialize the browser client with DOM elements.
     * Call after DOMContentLoaded.
     */
    initialize(elements) {
        this.elements = elements;
        // Set page title and menu title from story config
        document.title = this.config.storyInfo.title;
        const menuTitle = document.getElementById('menu-title');
        if (menuTitle) {
            menuTitle.textContent = this.config.storyInfo.title;
        }
        // Create text display and status line first (no dependencies)
        this.textDisplay = new TextDisplay_1.TextDisplay({
            textContent: elements.textContent,
            mainWindow: elements.mainWindow,
        });
        this.statusLine = new StatusLine_1.StatusLine({
            statusLocation: elements.statusLocation,
            statusScore: elements.statusScore,
        });
    }
    /**
     * Connect to game engine and set up event handlers.
     * Call after creating the engine.
     */
    connectEngine(engine, world) {
        this.engine = engine;
        this.world = world;
        // Now create managers that need world reference
        this.saveManager = new SaveManager_1.SaveManager({
            storagePrefix: this.config.storagePrefix,
            world: this.world,
            onStateChange: () => this.updateStatusLine(),
        });
        // Create dialog manager
        this.dialogManager = new DialogManager_1.DialogManager({
            elements: {
                modalOverlay: this.elements.modalOverlay,
                saveDialog: this.elements.saveDialog,
                restoreDialog: this.elements.restoreDialog,
                startupDialog: this.elements.startupDialog,
                saveNameInput: this.elements.saveNameInput,
                saveSlotsListEl: this.elements.saveSlotsListEl,
                restoreSlotsListEl: this.elements.restoreSlotsListEl,
                noSavesMessage: this.elements.noSavesMessage,
                startupSaveInfo: this.elements.startupSaveInfo,
            },
            saveManager: this.saveManager,
            onDialogOpen: () => this.inputManager?.disable(),
            onDialogClose: () => {
                this.inputManager?.enable();
                this.inputManager?.focus();
            },
            generateSaveName: () => this.saveManager.generateSaveName(this.currentTurn),
            sanitizeSaveName: (name) => this.saveManager.sanitizeSaveName(name),
            performSave: (name) => this.performSave(name),
        });
        // Create theme manager
        this.themeManager = new ThemeManager_1.ThemeManager({
            storageKey: `${this.config.storagePrefix}theme`,
            themes: this.config.themes,
            defaultTheme: this.config.defaultTheme,
        });
        // Create menu manager with handlers
        const menuHandlers = {
            onSave: () => this.handleSave(),
            onRestore: () => this.handleRestore(),
            onRestart: () => this.handleRestart(),
            onQuit: () => this.handleQuit(),
            onThemeSelect: (theme) => this.themeManager.applyTheme(theme),
            onHelp: () => this.engine.executeTurn('help'),
            onAbout: () => this.engine.executeTurn('about'),
        };
        this.menuManager = new MenuManager_1.MenuManager({
            menuBar: this.elements.menuBar,
            handlers: menuHandlers,
        });
        // Create input manager
        this.inputManager = new InputManager_1.InputManager({
            commandInput: this.elements.commandInput,
            onCommand: (cmd) => this.executeCommand(cmd),
            isDialogOpen: () => this.dialogManager.isDialogOpen(),
        });
        // Set up event handlers on engine
        this.setupEngineHandlers();
        // Set up all manager handlers
        this.dialogManager.setupHandlers();
        this.menuManager.setupHandlers();
        this.inputManager.setupHandlers();
        // Apply saved theme
        const savedTheme = this.themeManager.getSavedTheme();
        this.themeManager.applyTheme(savedTheme);
        // Sync existing saves to world
        this.saveManager.syncSavesToWorld();
    }
    /**
     * Set up event handlers on the game engine
     */
    setupEngineHandlers() {
        // Handle text output (ADR-137: extract prompt block for input field)
        this.engine.on('text:output', (blocks, turn) => {
            // Extract prompt block and set on input field
            const promptBlock = blocks.find((b) => b.key === 'prompt');
            if (promptBlock && promptBlock.content.length > 0) {
                const promptText = promptBlock.content[0];
                if (typeof promptText === 'string') {
                    this.inputManager.setPrompt(promptText);
                }
            }
            // Display non-prompt blocks as narrative text
            const displayBlocks = blocks.filter((b) => b.key !== 'prompt');
            const text = (0, text_service_1.renderToString)(displayBlocks);
            console.log('[text:output]', { text, turn, turnOffset: this.turnOffset });
            this.textDisplay.displayText(text);
            // Apply turn offset (set after restore to maintain correct turn count)
            this.currentTurn = turn + this.turnOffset;
            this.updateStatusLine();
            // Auto-save after each turn (if enabled)
            if (this.config.autoSave && turn > 0) {
                this.saveManager.performAutoSave(this.getSaveContext());
            }
        });
        // Handle game events
        this.engine.on('event', (event) => {
            console.log('[event]', event.type, event.data);
            // Forward audio events to the audio manager
            if (event.type.startsWith('audio.')) {
                this.audioManager.handleAudioEvent(event);
                return;
            }
            // Let story handle custom events first
            if (this.config.callbacks?.handleStoryEvent?.(event, this)) {
                return; // Story handled it
            }
            // Beep on errors/failures (classic Infocom style)
            if (event.type === 'command.failed' ||
                event.type === 'action.blocked' ||
                event.type.includes('.blocked') ||
                event.type === 'game.player_death') {
                this.beep();
            }
            // Track score changes
            if (event.type === 'game.score_changed' && event.data) {
                const scoreData = event.data;
                this.currentScore = scoreData.newScore ?? this.currentScore;
                this.updateStatusLine();
                // Celebratory beep on score increase
                if ((scoreData.newScore ?? 0) > (scoreData.oldScore ?? 0)) {
                    this.beep(1000, 50); // Higher, shorter beep for points
                }
            }
            // Handle platform events for save/restore
            if (event.type === 'platform.save_failed') {
                const errorData = event.data;
                this.textDisplay.displayText(`[Save failed: ${errorData?.error || 'Unknown error'}]`);
                this.beep();
            }
            else if (event.type === 'platform.restore_failed') {
                const errorData = event.data;
                this.textDisplay.displayText(`[Restore failed: ${errorData?.error || 'No saved game found'}]`);
                this.beep();
            }
            else if (event.type === 'platform.restore_completed') {
                const restoreData = event.data;
                if (restoreData?.turnCount !== undefined) {
                    this.currentTurn = restoreData.turnCount;
                }
                this.updateStatusLine();
            }
        });
    }
    /**
     * Start the game (check for autosave, show initial look)
     */
    async start() {
        console.log('=== BROWSER CLIENT START ===');
        try {
            // Set client version on StoryInfoTrait for banner display
            if (this.world) {
                const storyInfoEntities = this.world.findByTrait(world_model_1.TraitType.STORY_INFO);
                const trait = storyInfoEntities[0]?.get(world_model_1.TraitType.STORY_INFO);
                if (trait) {
                    trait.clientVersion = this.config.storyInfo.version;
                }
            }
            // Start the engine first (builds the world)
            await this.engine.start();
            // Capture baseline state for delta saves (before any commands or restore)
            this.saveManager.captureBaseline();
            // Sync localStorage saves to world
            this.saveManager.syncSavesToWorld();
            // Check for autosave and restore
            const autosaveData = this.saveManager.loadAutosave();
            if (autosaveData && autosaveData.locations) {
                console.log('[startup] Found autosave, restoring...');
                try {
                    // Restore state without replacing entities (preserves handlers)
                    this.saveManager.restoreWorldState(autosaveData);
                    this.currentTurn = autosaveData.turnCount || 0;
                    this.currentScore = autosaveData.score || 0;
                    // Set offset so engine's turn counter aligns with saved turn
                    this.turnOffset = this.currentTurn - 1;
                    // Restore transcript HTML if available
                    if (autosaveData.transcriptHtml) {
                        const html = this.saveManager.decompressTranscript(autosaveData.transcriptHtml);
                        this.textDisplay.setHTML(html);
                    }
                    console.log('[startup] Restored to turn', this.currentTurn, 'score', this.currentScore, 'offset', this.turnOffset);
                    this.updateStatusLine();
                    // Show restored message
                    this.textDisplay.displayText('[Session restored]');
                    this.textDisplay.scrollToBottom();
                    this.syncScoreFromWorld();
                }
                catch (error) {
                    console.error('[startup] Failed to restore autosave:', error);
                    // Fall through to new game
                    await this.engine.executeTurn('look');
                }
            }
            else {
                // New game - execute initial look
                console.log('Executing initial look command...');
                await this.engine.executeTurn('look');
                console.log('Initial look complete');
            }
            // Focus input
            this.inputManager?.focus();
        }
        catch (error) {
            console.error('=== STARTUP ERROR ===', error);
            this.textDisplay.displayText(`[Startup Error: ${error}]`);
        }
    }
    /**
     * Execute a command
     */
    async executeCommand(command) {
        // Unlock audio on first user command (browsers block autoplay)
        this.audioManager.unlock();
        // Display command echo
        this.textDisplay.displayCommand(command);
        // Execute command
        try {
            await this.engine.executeTurn(command);
            // Sync score from world after each turn
            this.syncScoreFromWorld();
        }
        catch (error) {
            const message = error instanceof Error ? error.message : String(error);
            this.textDisplay.displayText(`[Error: ${message}]`);
        }
    }
    /**
     * Get ISaveRestoreHooks for engine registration
     */
    getSaveRestoreHooks() {
        return {
            onSaveRequested: async (_data) => {
                // Ignore the engine's ISaveData - we'll use our own state capture
                const saved = await this.dialogManager.showSaveDialog();
                if (!saved) {
                    this.textDisplay.displayText('[Save cancelled]');
                }
            },
            onRestoreRequested: async () => {
                const slotName = await this.dialogManager.showRestoreDialog();
                if (!slotName) {
                    this.textDisplay.displayText('[Restore cancelled]');
                    return null;
                }
                const saveData = this.saveManager.loadSaveSlot(slotName);
                if (!saveData) {
                    this.textDisplay.displayText(`[No saved game found: ${slotName}]`);
                    this.beep();
                    return null;
                }
                // Restore world state
                this.saveManager.restoreWorldState(saveData);
                this.currentTurn = saveData.turnCount || 0;
                this.currentScore = saveData.score || 0;
                this.turnOffset = this.currentTurn - 1;
                // Restore transcript
                if (saveData.transcriptHtml) {
                    const html = this.saveManager.decompressTranscript(saveData.transcriptHtml);
                    this.textDisplay.setHTML(html);
                }
                this.updateStatusLine();
                this.textDisplay.displayText(`[Game restored from "${slotName}"]`);
                this.syncScoreFromWorld();
                // Return null since we handled the restore ourselves
                return null;
            },
            onRestartRequested: async (_context) => {
                if (confirm('Are you sure you want to restart? All unsaved progress will be lost.')) {
                    // Clear autosave and reload
                    this.saveManager.clearAutosave();
                    window.location.reload();
                    return true;
                }
                return false;
            },
        };
    }
    // === Menu handlers ===
    async handleSave() {
        await this.dialogManager.showSaveDialog();
    }
    async handleRestore() {
        const hooks = this.getSaveRestoreHooks();
        await hooks.onRestoreRequested();
    }
    async handleRestart() {
        const hooks = this.getSaveRestoreHooks();
        await hooks.onRestartRequested?.({});
    }
    handleQuit() {
        if (confirm('Are you sure you want to quit?')) {
            window.close();
            // If that doesn't work, show message
            this.textDisplay.displayText('[Close this browser tab to quit]');
        }
    }
    // === Public display methods (for story callbacks) ===
    displayText(text) {
        this.textDisplay.displayText(text);
    }
    displayCommand(command) {
        this.textDisplay.displayCommand(command);
    }
    clearScreen() {
        this.textDisplay.clearScreen();
    }
    beep(frequency = 800, duration = 100) {
        this.audioManager.beep(frequency, duration);
    }
    // === Public getters (for story callbacks) ===
    getWorld() {
        return this.world;
    }
    getCurrentTurn() {
        return this.currentTurn;
    }
    getCurrentScore() {
        return this.currentScore;
    }
    // === Private helpers ===
    performSave(slotName) {
        const result = this.saveManager.performSave(slotName, this.getSaveContext());
        if (result.success) {
            this.textDisplay.displayText(`[Game saved as "${slotName}"]`);
        }
        else {
            this.textDisplay.displayText(`[Save failed: ${result.error}]`);
            this.beep();
        }
    }
    getSaveContext() {
        return {
            turnCount: this.currentTurn,
            score: this.currentScore,
            transcriptHtml: this.textDisplay.getHTML(),
        };
    }
    updateStatusLine() {
        const location = this.saveManager.getCurrentLocation();
        this.statusLine.update(location, this.currentScore, this.currentTurn);
    }
    syncScoreFromWorld() {
        const scoring = this.world.getCapability('scoring');
        if (scoring && typeof scoring.scoreValue === 'number') {
            const newScore = scoring.scoreValue;
            if (newScore !== this.currentScore) {
                this.currentScore = newScore;
                this.updateStatusLine();
            }
        }
    }
}
exports.BrowserClient = BrowserClient;
