// @sharpee/platform-browser/index.js
"use strict";
/**
 * @sharpee/platform-browser
 *
 * Browser client infrastructure for Sharpee interactive fiction.
 * Provides reusable managers for save/restore, themes, menus, dialogs, input, and display.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.AudioManager = exports.StatusLine = exports.TextDisplay = exports.InputManager = exports.MenuManager = exports.DialogManager = exports.SaveManager = exports.ThemeManager = exports.AUTOSAVE_SLOT = exports.BrowserClient = void 0;
// Main client
var BrowserClient_1 = require("./BrowserClient");
Object.defineProperty(exports, "BrowserClient", { enumerable: true, get: function () { return BrowserClient_1.BrowserClient; } });
var types_1 = require("./types");
Object.defineProperty(exports, "AUTOSAVE_SLOT", { enumerable: true, get: function () { return types_1.AUTOSAVE_SLOT; } });
// Individual managers (for advanced use)
var ThemeManager_1 = require("./managers/ThemeManager");
Object.defineProperty(exports, "ThemeManager", { enumerable: true, get: function () { return ThemeManager_1.ThemeManager; } });
var SaveManager_1 = require("./managers/SaveManager");
Object.defineProperty(exports, "SaveManager", { enumerable: true, get: function () { return SaveManager_1.SaveManager; } });
var DialogManager_1 = require("./managers/DialogManager");
Object.defineProperty(exports, "DialogManager", { enumerable: true, get: function () { return DialogManager_1.DialogManager; } });
var MenuManager_1 = require("./managers/MenuManager");
Object.defineProperty(exports, "MenuManager", { enumerable: true, get: function () { return MenuManager_1.MenuManager; } });
var InputManager_1 = require("./managers/InputManager");
Object.defineProperty(exports, "InputManager", { enumerable: true, get: function () { return InputManager_1.InputManager; } });
// Display components
var TextDisplay_1 = require("./display/TextDisplay");
Object.defineProperty(exports, "TextDisplay", { enumerable: true, get: function () { return TextDisplay_1.TextDisplay; } });
var StatusLine_1 = require("./display/StatusLine");
Object.defineProperty(exports, "StatusLine", { enumerable: true, get: function () { return StatusLine_1.StatusLine; } });
// Audio
var AudioManager_1 = require("./audio/AudioManager");
Object.defineProperty(exports, "AudioManager", { enumerable: true, get: function () { return AudioManager_1.AudioManager; } });
