// @sharpee/text-blocks/guards.js
"use strict";
/**
 * Type Guards for TextBlock types
 *
 * Utilities for safely working with TextContent, IDecoration, and ITextBlock.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.isDecoration = isDecoration;
exports.isTextBlock = isTextBlock;
exports.hasKeyPrefix = hasKeyPrefix;
exports.isStatusBlock = isStatusBlock;
exports.isRoomBlock = isRoomBlock;
exports.isActionBlock = isActionBlock;
exports.extractPlainText = extractPlainText;
/**
 * Check if content is a decoration (not a plain string).
 *
 * @example
 * const content: TextContent = getContent();
 * if (isDecoration(content)) {
 *   console.log(content.type); // TypeScript knows this is IDecoration
 * }
 */
function isDecoration(content) {
    return typeof content === 'object' && content !== null && 'type' in content && 'content' in content;
}
/**
 * Check if a value is a valid TextBlock.
 *
 * @example
 * if (isTextBlock(value)) {
 *   console.log(value.key);
 * }
 */
function isTextBlock(value) {
    return (typeof value === 'object' &&
        value !== null &&
        'key' in value &&
        'content' in value &&
        typeof value.key === 'string' &&
        Array.isArray(value.content));
}
/**
 * Check if a block key starts with a given prefix.
 *
 * @example
 * if (hasKeyPrefix(block, 'status.')) {
 *   // Render to status bar
 * }
 */
function hasKeyPrefix(block, prefix) {
    return block.key.startsWith(prefix);
}
/**
 * Check if a block is a status block.
 */
function isStatusBlock(block) {
    return hasKeyPrefix(block, 'status.');
}
/**
 * Check if a block is a room-related block.
 */
function isRoomBlock(block) {
    return hasKeyPrefix(block, 'room.');
}
/**
 * Check if a block is an action-related block.
 */
function isActionBlock(block) {
    return hasKeyPrefix(block, 'action.');
}
/**
 * Extract plain text from TextContent, stripping all decorations.
 *
 * @example
 * const text = extractPlainText([
 *   'You take ',
 *   { type: 'item', content: ['the sword'] },
 *   '.'
 * ]);
 * // Returns: "You take the sword."
 */
function extractPlainText(content) {
    return content
        .map((item) => {
        if (typeof item === 'string') {
            return item;
        }
        return extractPlainText(item.content);
    })
        .join('');
}
