// @sharpee/text-blocks/index.js
"use strict";
/**
 * @sharpee/text-blocks
 *
 * Pure interfaces for structured text output in Sharpee IF platform.
 *
 * This package contains only TypeScript interfaces and type guards -
 * no runtime dependencies. It defines the contract between TextService
 * and clients (CLI, React, etc.).
 *
 * @packageDocumentation
 * @see ADR-096: Text Service Architecture
 * @see ADR-091: Text Decorations
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.BLOCK_KEY_PREFIXES = exports.extractPlainText = exports.isActionBlock = exports.isRoomBlock = exports.isStatusBlock = exports.hasKeyPrefix = exports.isTextBlock = exports.isDecoration = exports.BLOCK_KEYS = exports.CORE_BLOCK_KEYS = exports.CORE_DECORATION_TYPES = void 0;
// Constants
var types_js_1 = require("./types.js");
Object.defineProperty(exports, "CORE_DECORATION_TYPES", { enumerable: true, get: function () { return types_js_1.CORE_DECORATION_TYPES; } });
Object.defineProperty(exports, "CORE_BLOCK_KEYS", { enumerable: true, get: function () { return types_js_1.CORE_BLOCK_KEYS; } });
// Alias for text-service compatibility
var types_js_2 = require("./types.js");
Object.defineProperty(exports, "BLOCK_KEYS", { enumerable: true, get: function () { return types_js_2.CORE_BLOCK_KEYS; } });
// Type guards and utilities
var guards_js_1 = require("./guards.js");
Object.defineProperty(exports, "isDecoration", { enumerable: true, get: function () { return guards_js_1.isDecoration; } });
Object.defineProperty(exports, "isTextBlock", { enumerable: true, get: function () { return guards_js_1.isTextBlock; } });
Object.defineProperty(exports, "hasKeyPrefix", { enumerable: true, get: function () { return guards_js_1.hasKeyPrefix; } });
Object.defineProperty(exports, "isStatusBlock", { enumerable: true, get: function () { return guards_js_1.isStatusBlock; } });
Object.defineProperty(exports, "isRoomBlock", { enumerable: true, get: function () { return guards_js_1.isRoomBlock; } });
Object.defineProperty(exports, "isActionBlock", { enumerable: true, get: function () { return guards_js_1.isActionBlock; } });
Object.defineProperty(exports, "extractPlainText", { enumerable: true, get: function () { return guards_js_1.extractPlainText; } });
// Block key prefixes for routing
exports.BLOCK_KEY_PREFIXES = {
    STATUS: 'status.',
    ROOM: 'room.',
    ACTION: 'action.',
};
