// @sharpee/text-blocks/types.js
"use strict";
/**
 * @sharpee/text-blocks
 *
 * Pure interfaces for structured text output.
 * No runtime dependencies - types only.
 *
 * @see ADR-096: Text Service Architecture
 * @see ADR-091: Text Decorations
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CORE_BLOCK_KEYS = exports.CORE_DECORATION_TYPES = void 0;
/**
 * Core decoration types defined by the platform.
 * Stories can extend with custom types.
 */
exports.CORE_DECORATION_TYPES = {
    /** Emphasis (typically italic) */
    EM: 'em',
    /** Strong emphasis (typically bold) */
    STRONG: 'strong',
    /** Item/object name */
    ITEM: 'item',
    /** Room/location name */
    ROOM: 'room',
    /** NPC/character name */
    NPC: 'npc',
    /** Suggested command */
    COMMAND: 'command',
    /** Exit direction */
    DIRECTION: 'direction',
    /** Underlined text */
    UNDERLINE: 'underline',
    /** Struck-through text */
    STRIKETHROUGH: 'strikethrough',
    /** Superscript */
    SUPER: 'super',
    /** Subscript */
    SUB: 'sub',
};
/**
 * Core block keys defined by the platform.
 * Stories can emit custom keys.
 */
exports.CORE_BLOCK_KEYS = {
    /** Room title */
    ROOM_NAME: 'room.name',
    /** Room description */
    ROOM_DESCRIPTION: 'room.description',
    /** Items visible in room */
    ROOM_CONTENTS: 'room.contents',
    /** Action success result */
    ACTION_RESULT: 'action.result',
    /** Action blocked message */
    ACTION_BLOCKED: 'action.blocked',
    /** Current room (status bar) */
    STATUS_ROOM: 'status.room',
    /** Current score (status bar) */
    STATUS_SCORE: 'status.score',
    /** Turn count (status bar) */
    STATUS_TURNS: 'status.turns',
    /** System error */
    ERROR: 'error',
    /** Command prompt */
    PROMPT: 'prompt',
    /** Game/story message */
    GAME_MESSAGE: 'game.message',
    /** Game opening banner */
    GAME_BANNER: 'game.banner',
};
