// @sharpee/parser-en-us/index.js
"use strict";
/**
 * English (US) Parser for Sharpee Interactive Fiction Platform
 *
 * This package provides English-specific parsing functionality
 * for converting natural language commands into structured commands.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.metadata = exports.INANIMATE_THEM = exports.INANIMATE_IT = exports.RECOGNIZED_PRONOUNS = exports.isRecognizedPronoun = exports.PronounContextManager = exports.analyzeBestFailure = exports.Parser = exports.EnglishParser = void 0;
var english_parser_1 = require("./english-parser");
Object.defineProperty(exports, "EnglishParser", { enumerable: true, get: function () { return english_parser_1.EnglishParser; } });
var english_parser_2 = require("./english-parser"); // For convenience
Object.defineProperty(exports, "Parser", { enumerable: true, get: function () { return english_parser_2.EnglishParser; } });
var parse_failure_1 = require("./parse-failure");
Object.defineProperty(exports, "analyzeBestFailure", { enumerable: true, get: function () { return parse_failure_1.analyzeBestFailure; } });
var pronoun_context_1 = require("./pronoun-context");
Object.defineProperty(exports, "PronounContextManager", { enumerable: true, get: function () { return pronoun_context_1.PronounContextManager; } });
Object.defineProperty(exports, "isRecognizedPronoun", { enumerable: true, get: function () { return pronoun_context_1.isRecognizedPronoun; } });
Object.defineProperty(exports, "RECOGNIZED_PRONOUNS", { enumerable: true, get: function () { return pronoun_context_1.RECOGNIZED_PRONOUNS; } });
Object.defineProperty(exports, "INANIMATE_IT", { enumerable: true, get: function () { return pronoun_context_1.INANIMATE_IT; } });
Object.defineProperty(exports, "INANIMATE_THEM", { enumerable: true, get: function () { return pronoun_context_1.INANIMATE_THEM; } });
/**
 * Package metadata
 */
exports.metadata = {
    languageCode: 'en-US',
    languageName: 'English (United States)',
    parserVersion: '1.0.0',
    supportedPatterns: [
        'VERB_ONLY',
        'VERB_NOUN',
        'VERB_PREP_NOUN',
        'VERB_NOUN_PREP_NOUN',
        'DIRECTION_ONLY'
    ],
    features: [
        'compound-verbs',
        'articles',
        'adjectives',
        'abbreviations',
        'multi-word-nouns'
    ]
};
