// @sharpee/parser-en-us/parser-types.js
"use strict";
/**
 * Parser-specific types
 * These were previously in @sharpee/if-domain but are now local to the parser
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.vocabularyRegistry = exports.VocabularyRegistry = exports.PartOfSpeech = exports.ParseErrorType = void 0;
exports.adaptVerbVocabulary = adaptVerbVocabulary;
exports.adaptDirectionVocabulary = adaptDirectionVocabulary;
exports.adaptSpecialVocabulary = adaptSpecialVocabulary;
/**
 * Parse error types
 */
var ParseErrorType;
(function (ParseErrorType) {
    ParseErrorType["UNKNOWN_WORD"] = "UNKNOWN_WORD";
    ParseErrorType["PATTERN_MISMATCH"] = "PATTERN_MISMATCH";
    ParseErrorType["AMBIGUOUS"] = "AMBIGUOUS";
    ParseErrorType["NO_COMMAND"] = "NO_COMMAND";
})(ParseErrorType || (exports.ParseErrorType = ParseErrorType = {}));
/**
 * Vocabulary part of speech
 */
var PartOfSpeech;
(function (PartOfSpeech) {
    PartOfSpeech["VERB"] = "VERB";
    PartOfSpeech["NOUN"] = "NOUN";
    PartOfSpeech["ADJECTIVE"] = "ADJECTIVE";
    PartOfSpeech["ARTICLE"] = "ARTICLE";
    PartOfSpeech["PREPOSITION"] = "PREPOSITION";
    PartOfSpeech["PRONOUN"] = "PRONOUN";
    PartOfSpeech["DIRECTION"] = "DIRECTION";
    PartOfSpeech["SPECIAL"] = "SPECIAL";
})(PartOfSpeech || (exports.PartOfSpeech = PartOfSpeech = {}));
/**
 * Simple vocabulary registry
 */
class VocabularyRegistry {
    entries = new Map();
    clear() {
        this.entries.clear();
    }
    register(entry) {
        const word = entry.word.toLowerCase();
        if (!this.entries.has(word)) {
            this.entries.set(word, []);
        }
        this.entries.get(word).push(entry);
    }
    registerVerbs(verbs) {
        for (const verb of verbs) {
            this.register({
                word: verb.word,
                partOfSpeech: PartOfSpeech.VERB,
                mapsTo: verb.mapsTo,
                priority: verb.priority
            });
        }
    }
    registerDirections(directions) {
        for (const dir of directions) {
            this.register({
                word: dir.word,
                partOfSpeech: PartOfSpeech.DIRECTION,
                mapsTo: dir.mapsTo
            });
        }
    }
    registerSpecial(special) {
        for (const spec of special) {
            this.register({
                word: spec.word,
                partOfSpeech: PartOfSpeech.SPECIAL,
                mapsTo: spec.type
            });
        }
    }
    lookup(word) {
        return this.entries.get(word.toLowerCase()) || [];
    }
}
exports.VocabularyRegistry = VocabularyRegistry;
exports.vocabularyRegistry = new VocabularyRegistry();
/**
 * Adapt verb vocabulary from language provider format
 */
function adaptVerbVocabulary(language) {
    const verbs = [];
    if (language.getVerbs) {
        for (const verbDef of language.getVerbs()) {
            for (const verb of verbDef.verbs) {
                verbs.push({
                    word: verb,
                    mapsTo: verbDef.actionId,
                    priority: 1.0
                });
            }
        }
    }
    return verbs;
}
/**
 * Adapt direction vocabulary from language provider format
 */
function adaptDirectionVocabulary(language) {
    const directions = [];
    if (language.getDirections) {
        for (const dirDef of language.getDirections()) {
            for (const word of dirDef.words) {
                directions.push({
                    word,
                    mapsTo: dirDef.directionId
                });
            }
        }
    }
    return directions;
}
/**
 * Adapt special vocabulary from language provider format
 */
function adaptSpecialVocabulary(language) {
    const special = [];
    if (language.getSpecialWords) {
        const specialWords = language.getSpecialWords();
        for (const [type, words] of Object.entries(specialWords)) {
            for (const word of words) {
                special.push({ word, type });
            }
        }
    }
    return special;
}
