// @sharpee/parser-en-us/parse-failure.js
"use strict";
/**
 * @file Parse Failure Tracking
 * @description Types and utilities for tracking partial match failures
 *              to provide better error messages (Phase 1.2 of parser recommendations)
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.analyzeBestFailure = analyzeBestFailure;
/**
 * Analyze partial match failures to determine the best error to report
 */
function analyzeBestFailure(failures, input, hasVerb) {
    if (failures.length === 0) {
        // No patterns even tried
        if (!input.trim()) {
            return {
                code: 'NO_VERB',
                messageId: 'parser.error.noInput',
                context: {}
            };
        }
        return {
            code: 'UNKNOWN_VERB',
            messageId: 'parser.error.unknownVerb',
            context: { verb: input.split(/\s+/)[0] }
        };
    }
    // Sort by progress descending - most progress = best failure info
    const sorted = [...failures].sort((a, b) => b.progress - a.progress);
    const best = sorted[0];
    // If best failure has verb matched and slot failure, we have good info
    if (best.matchedVerb && best.slotFailure) {
        const sf = best.slotFailure;
        if (sf.reason === 'NO_MATCH') {
            return {
                code: 'ENTITY_NOT_FOUND',
                messageId: 'parser.error.entityNotFound',
                context: {
                    verb: best.matchedVerb,
                    noun: sf.unknownWord || sf.attemptedText,
                    slot: sf.slotName
                }
            };
        }
        if (sf.reason === 'SCOPE_VIOLATION') {
            return {
                code: 'SCOPE_VIOLATION',
                messageId: 'parser.error.scopeViolation',
                context: {
                    verb: best.matchedVerb,
                    noun: sf.attemptedText,
                    slot: sf.slotName
                }
            };
        }
        if (sf.reason === 'AMBIGUOUS') {
            return {
                code: 'AMBIGUOUS_INPUT',
                messageId: 'parser.error.ambiguous',
                context: {
                    verb: best.matchedVerb,
                    noun: sf.attemptedText,
                    candidates: sf.candidates
                }
            };
        }
    }
    // If verb matched but we ran out of tokens (missing object)
    if (best.matchedVerb && best.reason === 'NOT_ENOUGH_TOKENS') {
        // Check if it was looking for a slot
        if (best.expected?.startsWith(':')) {
            const slotName = best.expected.slice(1);
            // Determine if it's direct or indirect object
            if (slotName === 'target' || slotName === 'item' || slotName === 'object') {
                return {
                    code: 'MISSING_OBJECT',
                    messageId: 'parser.error.missingObject',
                    context: { verb: best.matchedVerb }
                };
            }
            else {
                return {
                    code: 'MISSING_INDIRECT',
                    messageId: 'parser.error.missingIndirect',
                    context: {
                        verb: best.matchedVerb,
                        slot: slotName
                    }
                };
            }
        }
        return {
            code: 'MISSING_OBJECT',
            messageId: 'parser.error.missingObject',
            context: { verb: best.matchedVerb }
        };
    }
    // Verb matched but slot failed without details (generic slot failure)
    if (best.matchedVerb && best.reason === 'SLOT_FAILED') {
        return {
            code: 'MISSING_OBJECT',
            messageId: 'parser.error.missingObject',
            context: { verb: best.matchedVerb }
        };
    }
    // If first token matched verb patterns but all failed
    if (best.matchedVerb && best.reason === 'LEFTOVER_TOKENS') {
        return {
            code: 'INVALID_SYNTAX',
            messageId: 'parser.error.invalidSyntax',
            context: {
                verb: best.matchedVerb,
                extraWords: best.failedAtToken
            }
        };
    }
    // No verb matched at all - unknown verb
    if (!hasVerb) {
        return {
            code: 'UNKNOWN_VERB',
            messageId: 'parser.error.unknownVerb',
            context: { verb: input.split(/\s+/)[0] }
        };
    }
    // Fallback
    return {
        code: 'INVALID_SYNTAX',
        messageId: 'parser.error.invalidSyntax',
        context: {}
    };
}
