// @sharpee/parser-en-us/scope-evaluator.js
"use strict";
/**
 * @file Scope Evaluator
 * @description Evaluates scope constraints against the world model
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ScopeEvaluator = void 0;
/**
 * Evaluates scope constraints to find matching entities
 */
class ScopeEvaluator {
    /**
     * Get entities that match a scope constraint
     */
    static getEntitiesInScope(constraint, context) {
        // If no world model, return empty array
        if (!context.world) {
            return [];
        }
        // Start with base scope
        let entities = [];
        switch (constraint.base) {
            case 'all':
                entities = this.getAllEntities(context);
                break;
            case 'visible':
                entities = this.getVisibleEntities(context);
                break;
            case 'touchable':
                entities = this.getTouchableEntities(context);
                break;
            case 'carried':
                entities = this.getCarriedEntities(context);
                break;
            case 'nearby':
                entities = this.getNearbyEntities(context);
                break;
            default:
                entities = [];
        }
        // Apply filters
        for (const filter of constraint.filters) {
            entities = entities.filter(entity => this.matchesFilter(entity, filter, context));
        }
        // Apply trait filters
        if (constraint.traitFilters && constraint.traitFilters.length > 0) {
            entities = entities.filter(entity => constraint.traitFilters.every(traitType => this.entityHasTrait(entity, traitType)));
        }
        // Add explicit entities
        if (constraint.explicitEntities.length > 0) {
            const additionalEntities = constraint.explicitEntities
                .map(id => context.world.getEntity(id))
                .filter(Boolean);
            entities = [...entities, ...additionalEntities];
        }
        // Remove duplicates
        const uniqueIds = new Set(entities.map(e => e.id));
        return entities.filter((e, i, arr) => arr.findIndex(e2 => e2.id === e.id) === i);
    }
    /**
     * Check if a single entity matches a scope constraint
     */
    static entityMatchesScope(entity, constraint, context) {
        const matchingEntities = this.getEntitiesInScope(constraint, context);
        return matchingEntities.some(e => e.id === entity.id);
    }
    /**
     * Get all entities in the world
     */
    static getAllEntities(context) {
        if (!context.world?.getAllEntities) {
            return [];
        }
        return context.world.getAllEntities();
    }
    /**
     * Get visible entities from current location
     */
    static getVisibleEntities(context) {
        if (!context.world?.getVisibleEntities) {
            return [];
        }
        return context.world.getVisibleEntities(context.actorId, context.currentLocation);
    }
    /**
     * Get touchable entities from current location
     */
    static getTouchableEntities(context) {
        if (!context.world?.getTouchableEntities) {
            return [];
        }
        return context.world.getTouchableEntities(context.actorId, context.currentLocation);
    }
    /**
     * Get entities carried by the actor
     */
    static getCarriedEntities(context) {
        if (!context.world?.getCarriedEntities) {
            return [];
        }
        return context.world.getCarriedEntities(context.actorId);
    }
    /**
     * Get nearby entities (visible + adjacent locations)
     */
    static getNearbyEntities(context) {
        if (!context.world?.getNearbyEntities) {
            // Fallback to visible if nearby not implemented
            return this.getVisibleEntities(context);
        }
        return context.world.getNearbyEntities(context.actorId, context.currentLocation);
    }
    /**
     * Check if entity matches a filter
     */
    static matchesFilter(entity, filter, context) {
        if (typeof filter === 'function') {
            // Function constraint
            return filter(entity, context);
        }
        else {
            // Property constraint
            for (const [key, value] of Object.entries(filter)) {
                const entityValue = entity[key];
                if (entityValue !== value) {
                    return false;
                }
            }
            return true;
        }
    }
    /**
     * Check if entity has a specific trait
     * Supports both entity.has() method and entity.get() method patterns
     */
    static entityHasTrait(entity, traitType) {
        // Check for .has() method (trait system standard)
        if (typeof entity.has === 'function') {
            return entity.has(traitType);
        }
        // Check for .get() method returning truthy value (alternate pattern)
        if (typeof entity.get === 'function') {
            const trait = entity.get(traitType);
            return trait !== undefined && trait !== null;
        }
        return false;
    }
    /**
     * Get entity names and aliases for matching
     * Supports both legacy attributes.name and IdentityTrait patterns
     */
    static getEntityNames(entity) {
        const names = [];
        // Check attributes (legacy pattern)
        if (entity.attributes) {
            if (entity.attributes.displayName) {
                names.push(String(entity.attributes.displayName));
            }
            if (entity.attributes.name) {
                names.push(String(entity.attributes.name));
            }
        }
        // Check IdentityTrait (via .get() method)
        if (typeof entity.get === 'function') {
            const identity = entity.get('identity');
            if (identity && typeof identity === 'object') {
                if (identity.name) {
                    names.push(String(identity.name));
                }
                // Also check aliases
                if (Array.isArray(identity.aliases)) {
                    names.push(...identity.aliases.map(String));
                }
            }
        }
        return names;
    }
    /**
     * Find entities by name in a given scope
     */
    static findEntitiesByName(name, constraint, context) {
        const entitiesInScope = this.getEntitiesInScope(constraint, context);
        const searchName = name.toLowerCase();
        // Try exact match first (name or any alias)
        const exactMatches = entitiesInScope.filter(e => {
            if (!e)
                return false;
            const names = this.getEntityNames(e);
            return names.some(n => n.toLowerCase() === searchName);
        });
        if (exactMatches.length > 0) {
            return exactMatches;
        }
        // Try partial match
        const partialMatches = entitiesInScope.filter(e => {
            if (!e)
                return false;
            const names = this.getEntityNames(e);
            return names.some(n => n.toLowerCase().includes(searchName));
        });
        return partialMatches;
    }
}
exports.ScopeEvaluator = ScopeEvaluator;
