// @sharpee/stdlib/query-handlers/quit-handler.js
"use strict";
/**
 * Quit Query Handler
 *
 * Handles responses to quit confirmation queries.
 * Works with the platform events system to emit appropriate events.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.QuitQueryHandler = void 0;
exports.createQuitQueryHandler = createQuitQueryHandler;
const core_1 = require("../../core/index.js");
/**
 * Handler for quit confirmation queries
 */
class QuitQueryHandler {
    eventSource = (0, core_1.createSemanticEventSource)();
    /**
     * Check if this handler can process the query
     */
    canHandle(query) {
        return query.source === core_1.QuerySource.SYSTEM &&
            (query.type === core_1.QueryType.YES_NO || query.type === core_1.QueryType.MULTIPLE_CHOICE) &&
            (query.messageId === 'quit_confirmation' ||
                query.messageId === 'quit_confirm_query' ||
                query.messageId === 'quit_unsaved_query' ||
                query.messageId === 'quit_save_query');
    }
    /**
     * Process the player's response
     */
    handleResponse(response, query) {
        const events = [];
        const context = query.context;
        // Get the selected option
        const selectedOption = query.options?.[response.selectedIndex || 0] ||
            response.response;
        // Handle boolean responses for YES_NO queries
        if (typeof selectedOption === 'boolean') {
            if (selectedOption === true) {
                // Yes response - proceed with quit
                const quitContext = {
                    score: context.score,
                    moves: context.moves,
                    hasUnsavedChanges: context.hasUnsavedProgress || false,
                    force: true, // User confirmed, so treat as forced
                    stats: {
                        maxScore: context.maxScore,
                        nearComplete: context.nearComplete
                    }
                };
                // Emit platform quit requested event
                const quitEvent = (0, core_1.createQuitRequestedEvent)(quitContext);
                events.push(quitEvent);
                // Also emit a query handled event for tracking
                events.push(this.createEvent('query.handled', {
                    queryId: query.id,
                    action: 'quit_confirmed',
                    savedBeforeQuit: false
                }));
            }
            else {
                // No response - cancel quit
                const cancelEvent = (0, core_1.createQuitCancelledEvent)();
                events.push(cancelEvent);
                // Query handled event
                events.push(this.createEvent('query.handled', {
                    queryId: query.id,
                    action: 'quit_cancelled'
                }));
            }
        }
        else {
            // Handle string responses for MULTIPLE_CHOICE queries
            switch (selectedOption) {
                case 'quit':
                case 'quit_without_saving':
                case 'yes':
                    // Build quit context from query context
                    const quitContext = {
                        score: context.score,
                        moves: context.moves,
                        hasUnsavedChanges: context.hasUnsavedProgress || false,
                        force: true, // User confirmed, so treat as forced
                        stats: {
                            maxScore: context.maxScore,
                            nearComplete: context.nearComplete
                        }
                    };
                    // Emit platform quit requested event
                    const quitEvent = (0, core_1.createQuitRequestedEvent)(quitContext);
                    events.push(quitEvent);
                    // Also emit a query handled event for tracking
                    events.push(this.createEvent('query.handled', {
                        queryId: query.id,
                        action: 'quit_confirmed',
                        savedBeforeQuit: false
                    }));
                    break;
                case 'save_and_quit':
                    // First emit save requested event
                    events.push(this.createEvent('platform.save_requested', {
                        saveName: 'autosave',
                        autosave: true,
                        timestamp: Date.now(),
                        metadata: {
                            reason: 'quit',
                            score: context.score,
                            moves: context.moves
                        }
                    }));
                    // Then emit quit requested event
                    const saveAndQuitContext = {
                        score: context.score,
                        moves: context.moves,
                        hasUnsavedChanges: false, // Will be saved first
                        force: true,
                        stats: {
                            maxScore: context.maxScore,
                            savedBeforeQuit: true
                        }
                    };
                    const saveAndQuitEvent = (0, core_1.createQuitRequestedEvent)(saveAndQuitContext);
                    events.push(saveAndQuitEvent);
                    // Query handled event
                    events.push(this.createEvent('query.handled', {
                        queryId: query.id,
                        action: 'save_and_quit',
                        savedBeforeQuit: true
                    }));
                    break;
                case 'cancel':
                case 'no':
                default:
                    // Emit platform quit cancelled event
                    const cancelEvent = (0, core_1.createQuitCancelledEvent)();
                    events.push(cancelEvent);
                    // Query handled event
                    events.push(this.createEvent('query.handled', {
                        queryId: query.id,
                        action: 'quit_cancelled'
                    }));
                    break;
            }
        }
        // Emit all events
        for (const event of events) {
            this.eventSource.emit(event);
        }
    }
    /**
     * Handle query timeout
     */
    handleTimeout(query) {
        // On timeout, treat as cancelled
        const cancelEvent = (0, core_1.createQuitCancelledEvent)();
        this.eventSource.emit(cancelEvent);
        this.eventSource.emit(this.createEvent('query.timeout', {
            queryId: query.id,
            action: 'quit_cancelled',
            reason: 'timeout'
        }));
    }
    /**
     * Handle query cancellation
     */
    handleCancel(query) {
        const cancelEvent = (0, core_1.createQuitCancelledEvent)();
        this.eventSource.emit(cancelEvent);
        this.eventSource.emit(this.createEvent('query.cancelled', {
            queryId: query.id,
            action: 'quit_cancelled',
            reason: 'cancelled'
        }));
    }
    /**
     * Get the event source
     */
    getEventSource() {
        return this.eventSource;
    }
    /**
     * Create a semantic event
     */
    createEvent(type, data) {
        return {
            id: `evt_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`,
            type,
            timestamp: Date.now(),
            entities: {},
            data
        };
    }
}
exports.QuitQueryHandler = QuitQueryHandler;
/**
 * Create a quit query handler instance
 */
function createQuitQueryHandler() {
    return new QuitQueryHandler();
}
