// @sharpee/stdlib/query-handlers/restart-handler.js
"use strict";
/**
 * Restart Query Handler
 *
 * Handles responses to restart confirmation queries.
 * Works with the platform events system to emit appropriate events.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.RestartQueryHandler = void 0;
exports.createRestartQueryHandler = createRestartQueryHandler;
const core_1 = require("../../core/index.js");
/**
 * Handler for restart confirmation queries
 */
class RestartQueryHandler {
    eventSource = (0, core_1.createSemanticEventSource)();
    /**
     * Check if this handler can process the query
     */
    canHandle(query) {
        return query.source === core_1.QuerySource.SYSTEM &&
            (query.type === core_1.QueryType.YES_NO || query.type === core_1.QueryType.MULTIPLE_CHOICE) &&
            (query.messageId === 'restart_confirmation' ||
                query.messageId === 'restart_confirm' ||
                query.messageId === 'restart_unsaved' ||
                query.messageId === 'restart_progress');
    }
    /**
     * Process the player's response
     */
    handleResponse(response, query) {
        const events = [];
        const context = query.context;
        // Get the selected option
        const selectedOption = query.options?.[response.selectedIndex || 0] ||
            response.response;
        // Handle boolean responses for YES_NO queries
        if (typeof selectedOption === 'boolean') {
            if (selectedOption === true) {
                // Yes response - proceed with restart
                const restartContext = {
                    currentProgress: {
                        score: context.score,
                        moves: context.moves,
                        location: context.location
                    },
                    confirmationRequired: false, // Already confirmed
                    hasUnsavedChanges: context.hasUnsavedProgress || false,
                    force: true // User confirmed, so treat as forced
                };
                // Emit platform restart requested event
                const restartEvent = (0, core_1.createRestartRequestedEvent)(restartContext);
                events.push(restartEvent);
                // Also emit a query handled event for tracking
                events.push(this.createEvent('query.handled', {
                    queryId: query.id,
                    action: 'restart_confirmed',
                    savedBeforeRestart: false
                }));
            }
            else {
                // No response - cancel restart
                const cancelEvent = (0, core_1.createRestartCompletedEvent)(false);
                events.push(cancelEvent);
                // Query handled event
                events.push(this.createEvent('query.handled', {
                    queryId: query.id,
                    action: 'restart_cancelled'
                }));
            }
        }
        else {
            // Handle string responses for MULTIPLE_CHOICE queries
            switch (selectedOption) {
                case 'restart':
                case 'restart_without_saving':
                case 'yes':
                    // Build restart context from query context
                    const restartContext = {
                        currentProgress: {
                            score: context.score,
                            moves: context.moves,
                            location: context.location
                        },
                        confirmationRequired: false, // Already confirmed
                        hasUnsavedChanges: context.hasUnsavedProgress || false,
                        force: true // User confirmed, so treat as forced
                    };
                    // Emit platform restart requested event
                    const restartEvent = (0, core_1.createRestartRequestedEvent)(restartContext);
                    events.push(restartEvent);
                    // Also emit a query handled event for tracking
                    events.push(this.createEvent('query.handled', {
                        queryId: query.id,
                        action: 'restart_confirmed',
                        savedBeforeRestart: false
                    }));
                    break;
                case 'save_and_restart':
                    // First emit save requested event
                    events.push(this.createEvent('platform.save_requested', {
                        saveName: 'before_restart',
                        autosave: true,
                        timestamp: Date.now(),
                        metadata: {
                            reason: 'restart',
                            score: context.score,
                            moves: context.moves,
                            location: context.location
                        }
                    }));
                    // Then emit restart requested event
                    const saveAndRestartContext = {
                        currentProgress: {
                            score: context.score,
                            moves: context.moves,
                            location: context.location
                        },
                        confirmationRequired: false,
                        hasUnsavedChanges: false, // Will be saved first
                        force: true
                    };
                    const saveAndRestartEvent = (0, core_1.createRestartRequestedEvent)(saveAndRestartContext);
                    events.push(saveAndRestartEvent);
                    // Query handled event
                    events.push(this.createEvent('query.handled', {
                        queryId: query.id,
                        action: 'save_and_restart',
                        savedBeforeRestart: true
                    }));
                    break;
                case 'cancel':
                case 'no':
                default:
                    // Emit platform restart cancelled event
                    const cancelEvent = (0, core_1.createRestartCompletedEvent)(false);
                    events.push(cancelEvent);
                    // Query handled event
                    events.push(this.createEvent('query.handled', {
                        queryId: query.id,
                        action: 'restart_cancelled'
                    }));
                    break;
            }
        }
        // Emit all events
        for (const event of events) {
            this.eventSource.emit(event);
        }
    }
    /**
     * Handle query timeout
     */
    handleTimeout(query) {
        // On timeout, treat as cancelled
        const cancelEvent = (0, core_1.createRestartCompletedEvent)(false);
        this.eventSource.emit(cancelEvent);
        this.eventSource.emit(this.createEvent('query.timeout', {
            queryId: query.id,
            action: 'restart_cancelled',
            reason: 'timeout'
        }));
    }
    /**
     * Handle query cancellation
     */
    handleCancel(query) {
        const cancelEvent = (0, core_1.createRestartCompletedEvent)(false);
        this.eventSource.emit(cancelEvent);
        this.eventSource.emit(this.createEvent('query.cancelled', {
            queryId: query.id,
            action: 'restart_cancelled',
            reason: 'cancelled'
        }));
    }
    /**
     * Get the event source
     */
    getEventSource() {
        return this.eventSource;
    }
    /**
     * Create a semantic event
     */
    createEvent(type, data) {
        return {
            id: `evt_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`,
            type,
            timestamp: Date.now(),
            entities: {},
            data
        };
    }
}
exports.RestartQueryHandler = RestartQueryHandler;
/**
 * Create a restart query handler instance
 */
function createRestartQueryHandler() {
    return new RestartQueryHandler();
}
