// @sharpee/stdlib/capabilities/debug.js
"use strict";
/**
 * Debug capability for tracking debug/author command settings
 *
 * This capability stores flags for various debug outputs that can be
 * toggled via meta-commands (author commands) without affecting game state.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEBUG_CAPABILITY = exports.DebugCapabilitySchema = void 0;
exports.isAnyDebugEnabled = isAnyDebugEnabled;
exports.createDefaultDebugData = createDefaultDebugData;
/**
 * Debug capability schema
 * Tracks which debug events should be emitted
 */
exports.DebugCapabilitySchema = {
    debugParserEvents: {
        type: 'boolean',
        default: false,
        required: false
    },
    debugValidationEvents: {
        type: 'boolean',
        default: false,
        required: false
    },
    debugSystemEvents: {
        type: 'boolean',
        default: false,
        required: false
    },
    debugEngineEvents: {
        type: 'boolean',
        default: false,
        required: false
    },
    debugTextEvents: {
        type: 'boolean',
        default: false,
        required: false
    }
};
/**
 * Standard capability name for debug settings
 */
exports.DEBUG_CAPABILITY = 'debug';
/**
 * Helper to check if any debug events are enabled
 */
function isAnyDebugEnabled(debugData) {
    if (!debugData)
        return false;
    return !!(debugData.debugParserEvents ||
        debugData.debugValidationEvents ||
        debugData.debugSystemEvents ||
        debugData.debugEngineEvents ||
        debugData.debugTextEvents);
}
/**
 * Helper to create default debug settings
 */
function createDefaultDebugData() {
    return {
        debugParserEvents: false,
        debugValidationEvents: false,
        debugSystemEvents: false,
        debugEngineEvents: false,
        debugTextEvents: false
    };
}
