// @sharpee/stdlib/capabilities/index.js
"use strict";
/**
 * Standard capabilities for the Sharpee stdlib
 *
 * These capabilities provide common game state management patterns
 * that don't naturally fit in the entity-relationship model.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.StandardCapabilitySchemas = exports.createDefaultDebugData = exports.isAnyDebugEnabled = exports.DEBUG_CAPABILITY = exports.DebugCapabilitySchema = exports.CommandHistoryCapabilitySchema = exports.GameMetaCapabilitySchema = exports.ConversationCapabilitySchema = exports.SaveRestoreCapabilitySchema = exports.ScoringCapabilitySchema = void 0;
exports.registerStandardCapabilities = registerStandardCapabilities;
const world_model_1 = require("../../world-model/index.js");
// Import individual capabilities
const scoring_1 = require("./scoring");
Object.defineProperty(exports, "ScoringCapabilitySchema", { enumerable: true, get: function () { return scoring_1.ScoringCapabilitySchema; } });
const save_restore_1 = require("./save-restore");
Object.defineProperty(exports, "SaveRestoreCapabilitySchema", { enumerable: true, get: function () { return save_restore_1.SaveRestoreCapabilitySchema; } });
const conversation_1 = require("./conversation");
Object.defineProperty(exports, "ConversationCapabilitySchema", { enumerable: true, get: function () { return conversation_1.ConversationCapabilitySchema; } });
const game_meta_1 = require("./game-meta");
Object.defineProperty(exports, "GameMetaCapabilitySchema", { enumerable: true, get: function () { return game_meta_1.GameMetaCapabilitySchema; } });
const command_history_1 = require("./command-history");
Object.defineProperty(exports, "CommandHistoryCapabilitySchema", { enumerable: true, get: function () { return command_history_1.CommandHistoryCapabilitySchema; } });
const debug_1 = require("./debug");
Object.defineProperty(exports, "DebugCapabilitySchema", { enumerable: true, get: function () { return debug_1.DebugCapabilitySchema; } });
Object.defineProperty(exports, "DEBUG_CAPABILITY", { enumerable: true, get: function () { return debug_1.DEBUG_CAPABILITY; } });
Object.defineProperty(exports, "isAnyDebugEnabled", { enumerable: true, get: function () { return debug_1.isAnyDebugEnabled; } });
Object.defineProperty(exports, "createDefaultDebugData", { enumerable: true, get: function () { return debug_1.createDefaultDebugData; } });
/**
 * Map of standard capability names to their schemas
 */
exports.StandardCapabilitySchemas = {
    [world_model_1.StandardCapabilities.SCORING]: scoring_1.ScoringCapabilitySchema,
    [world_model_1.StandardCapabilities.SAVE_RESTORE]: save_restore_1.SaveRestoreCapabilitySchema,
    [world_model_1.StandardCapabilities.CONVERSATION]: conversation_1.ConversationCapabilitySchema,
    [world_model_1.StandardCapabilities.GAME_META]: game_meta_1.GameMetaCapabilitySchema,
    [world_model_1.StandardCapabilities.COMMAND_HISTORY]: command_history_1.CommandHistoryCapabilitySchema,
    [world_model_1.StandardCapabilities.DEBUG]: debug_1.DebugCapabilitySchema,
};
/**
 * Helper to register all standard capabilities
 * @param world The world model to register capabilities on
 * @param capabilities Array of capability names to register (defaults to all)
 */
function registerStandardCapabilities(world, capabilities = Object.keys(exports.StandardCapabilitySchemas)) {
    for (const capabilityName of capabilities) {
        const schema = exports.StandardCapabilitySchemas[capabilityName];
        if (schema) {
            world.registerCapability(capabilityName, { schema });
        }
    }
}
