// @sharpee/world-model/traits/combatant/combatantTrait.js
"use strict";
/**
 * Combatant trait for entities that can engage in combat
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CombatantTrait = void 0;
/**
 * Combatant trait indicates an entity can engage in combat.
 *
 * This trait contains only data - all combat logic
 * is in CombatBehavior.
 */
class CombatantTrait {
    static type = 'combatant';
    type = 'combatant';
    // CombatantData properties
    health;
    maxHealth;
    skill;
    baseDamage;
    isConscious;
    recoveryTurns;
    armor;
    attackPower;
    defense;
    isAlive;
    hitMessage;
    deathMessage;
    attackMessage;
    hostile;
    canRetaliate;
    dropsInventory;
    experienceValue;
    isUndead;
    constructor(data = {}) {
        // Set defaults and merge with provided data
        this.health = data.health ?? data.maxHealth ?? 10;
        this.maxHealth = data.maxHealth ?? 10;
        this.skill = data.skill ?? 30; // Default skill level (ADR-072)
        this.baseDamage = data.baseDamage ?? data.attackPower ?? 1; // ADR-072
        this.isConscious = data.isConscious ?? true; // ADR-072
        this.recoveryTurns = data.recoveryTurns; // ADR-072
        this.armor = data.armor ?? 0;
        this.attackPower = data.attackPower ?? 1;
        this.defense = data.defense ?? 0;
        this.isAlive = data.isAlive ?? true;
        this.hitMessage = data.hitMessage;
        this.deathMessage = data.deathMessage;
        this.attackMessage = data.attackMessage;
        this.hostile = data.hostile ?? false;
        this.canRetaliate = data.canRetaliate ?? true;
        this.dropsInventory = data.dropsInventory ?? true;
        this.experienceValue = data.experienceValue ?? 0;
        this.isUndead = data.isUndead ?? false;
    }
    /**
     * Computed property to check if combatant is alive based on health
     */
    get alive() {
        return this.health > 0;
    }
    /**
     * Check if combatant can act (alive and conscious)
     */
    get canAct() {
        return this.isAlive && this.isConscious;
    }
    /**
     * Knock out this combatant (unconscious but alive)
     */
    knockOut(recoveryTurns) {
        this.isConscious = false;
        if (recoveryTurns !== undefined) {
            this.recoveryTurns = recoveryTurns;
        }
    }
    /**
     * Wake up this combatant
     */
    wakeUp() {
        if (this.isAlive) {
            this.isConscious = true;
            this.recoveryTurns = undefined;
        }
    }
    /**
     * Kill this combatant
     */
    kill() {
        this.health = 0;
        this.isAlive = false;
        this.isConscious = false;
    }
    /**
     * Take damage
     * @returns true if combatant was killed
     */
    takeDamage(amount) {
        const effectiveDamage = Math.max(0, amount - this.armor);
        this.health = Math.max(0, this.health - effectiveDamage);
        if (this.health <= 0) {
            this.kill();
            return true;
        }
        // Check for knockout (20% health threshold per ADR-072)
        if (this.health <= this.maxHealth * 0.2) {
            this.knockOut();
        }
        return false;
    }
    /**
     * Heal this combatant
     */
    heal(amount) {
        if (this.isAlive) {
            this.health = Math.min(this.maxHealth, this.health + amount);
            if (this.health > this.maxHealth * 0.2) {
                this.wakeUp();
            }
        }
    }
}
exports.CombatantTrait = CombatantTrait;
