// @sharpee/stdlib/vocabulary/standard-english.js
"use strict";
/**
 * Standard English vocabulary for Sharpee
 *
 * This provides the base vocabulary for standard IF commands
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.commonPrepositions = exports.specialVocabulary = exports.standardDirections = exports.standardVerbs = void 0;
exports.registerStandardVocabulary = registerStandardVocabulary;
const parser_1 = require("../parser");
/**
 * Standard verb vocabulary
 */
exports.standardVerbs = [
    // Movement
    {
        actionId: 'GO',
        verbs: ['go', 'walk', 'move', 'travel', 'head'],
        pattern: 'verb_direction'
    },
    // Taking
    {
        actionId: 'TAKE',
        verbs: ['take', 'get', 'pick', 'grab', 'acquire', 'obtain'],
        pattern: 'verb_noun'
    },
    // Dropping
    {
        actionId: 'DROP',
        verbs: ['drop', 'put', 'place', 'leave', 'discard'],
        pattern: 'verb_noun'
    },
    // Examining
    {
        actionId: 'EXAMINE',
        verbs: ['examine', 'look', 'x', 'ex', 'inspect', 'study', 'observe'],
        pattern: 'verb_noun',
        prepositions: ['at']
    },
    // Opening
    {
        actionId: 'OPEN',
        verbs: ['open', 'unlock'],
        pattern: 'verb_noun'
    },
    // Closing
    {
        actionId: 'CLOSE',
        verbs: ['close', 'shut', 'lock'],
        pattern: 'verb_noun'
    },
    // Inventory
    {
        actionId: 'INVENTORY',
        verbs: ['inventory', 'inv', 'i'],
        pattern: 'verb_only'
    },
    // Looking
    {
        actionId: 'LOOK',
        verbs: ['look', 'l'],
        pattern: 'verb_only'
    },
    // Putting
    {
        actionId: 'PUT',
        verbs: ['put', 'place', 'insert', 'deposit'],
        pattern: 'verb_noun_prep_noun',
        prepositions: ['in', 'into', 'on', 'onto', 'under', 'behind']
    },
    // Using
    {
        actionId: 'USE',
        verbs: ['use', 'apply', 'activate'],
        pattern: 'verb_noun'
    },
    // Talking
    {
        actionId: 'TALK',
        verbs: ['talk', 'speak', 'say', 'tell', 'ask'],
        pattern: 'verb_prep_noun',
        prepositions: ['to', 'with']
    },
    // Giving
    {
        actionId: 'GIVE',
        verbs: ['give', 'offer', 'hand'],
        pattern: 'verb_noun_prep_noun',
        prepositions: ['to']
    },
    // Attacking
    {
        actionId: 'ATTACK',
        verbs: ['attack', 'hit', 'strike', 'fight', 'kill'],
        pattern: 'verb_noun'
    },
    // Waiting
    {
        actionId: 'WAIT',
        verbs: ['wait', 'z'],
        pattern: 'verb_only'
    },
    // Saving
    {
        actionId: 'SAVE',
        verbs: ['save'],
        pattern: 'verb_only'
    },
    // Loading
    {
        actionId: 'LOAD',
        verbs: ['load', 'restore'],
        pattern: 'verb_only'
    },
    // Quitting
    {
        actionId: 'QUIT',
        verbs: ['quit', 'exit', 'q'],
        pattern: 'verb_only'
    },
    // Author/Debug Commands
    {
        actionId: 'author.trace',
        verbs: ['trace'],
        pattern: 'verb_noun' // trace [target] on/off
    }
];
/**
 * Standard direction vocabulary
 */
exports.standardDirections = [
    {
        direction: 'NORTH',
        words: ['north'],
        abbreviations: ['n']
    },
    {
        direction: 'SOUTH',
        words: ['south'],
        abbreviations: ['s']
    },
    {
        direction: 'EAST',
        words: ['east'],
        abbreviations: ['e']
    },
    {
        direction: 'WEST',
        words: ['west'],
        abbreviations: ['w']
    },
    {
        direction: 'NORTHEAST',
        words: ['northeast'],
        abbreviations: ['ne']
    },
    {
        direction: 'NORTHWEST',
        words: ['northwest'],
        abbreviations: ['nw']
    },
    {
        direction: 'SOUTHEAST',
        words: ['southeast'],
        abbreviations: ['se']
    },
    {
        direction: 'SOUTHWEST',
        words: ['southwest'],
        abbreviations: ['sw']
    },
    {
        direction: 'UP',
        words: ['up', 'upward', 'upwards'],
        abbreviations: ['u']
    },
    {
        direction: 'DOWN',
        words: ['down', 'downward', 'downwards'],
        abbreviations: ['d']
    },
    {
        direction: 'IN',
        words: ['in', 'inside', 'enter'],
        abbreviations: []
    },
    {
        direction: 'OUT',
        words: ['out', 'outside', 'exit', 'leave'],
        abbreviations: []
    }
];
/**
 * Special vocabulary
 */
exports.specialVocabulary = {
    pronouns: ['it', 'them', 'that', 'those', 'this', 'these'],
    allWords: ['all', 'everything', 'every'],
    exceptWords: ['except', 'but'],
    articles: ['a', 'an', 'the', 'some']
};
/**
 * Common prepositions
 */
exports.commonPrepositions = [
    'in', 'into', 'on', 'onto', 'at', 'to', 'from',
    'with', 'under', 'over', 'behind', 'beside',
    'through', 'across', 'between', 'among'
];
/**
 * Register all standard vocabulary
 */
function registerStandardVocabulary() {
    // Register verbs
    parser_1.vocabularyRegistry.registerVerbs(exports.standardVerbs);
    // Register directions
    parser_1.vocabularyRegistry.registerDirections(exports.standardDirections);
    // Register special vocabulary
    parser_1.vocabularyRegistry.registerSpecial(exports.specialVocabulary);
    // Register prepositions
    parser_1.vocabularyRegistry.registerProvider({
        id: 'standard-prepositions',
        priority: 100,
        getVocabulary: () => exports.commonPrepositions.map(prep => ({
            word: prep,
            partOfSpeech: 'preposition',
            mapsTo: prep.toUpperCase(),
            source: 'base'
        }))
    });
}
// Auto-register on import
registerStandardVocabulary();
