// @sharpee/stdlib/actions/standard/undoing/undoing.js
"use strict";
/**
 * Undoing action - undo the previous turn
 *
 * This is a meta action that triggers the engine's undo functionality.
 * It emits a platform event that the engine processes to restore
 * the previous game state.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.undoingAction = void 0;
const core_1 = require("../../../../core/index.js");
const constants_1 = require("../../constants");
exports.undoingAction = {
    id: constants_1.IFActions.UNDOING,
    requiredMessages: [
        'undo_success',
        'undo_failed',
        'nothing_to_undo'
    ],
    validate(_context) {
        // Undo is always valid - the engine will determine if there's anything to undo
        return { valid: true };
    },
    execute(_context) {
        // No world mutations - undo is handled by the engine
    },
    blocked(_context, _result) {
        // Undo should never be blocked
        return [];
    },
    report(context) {
        // Emit platform undo requested event
        // The engine will handle this after turn completion
        const platformEvent = (0, core_1.createUndoRequestedEvent)();
        return [platformEvent];
    },
    group: "meta",
    metadata: {
        requiresDirectObject: false,
        requiresIndirectObject: false
    }
};
