// @sharpee/engine/action-context-factory.js
"use strict";
/**
 * Action Context Factory - Creates ActionContext for action execution
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.createActionContext = createActionContext;
const stdlib_1 = require("../stdlib/index.js");
const world_model_1 = require("../world-model/index.js");
const core_1 = require("../core/index.js");
const shared_data_keys_1 = require("./shared-data-keys");
/**
 * Helper to get entity from command slot
 */
function getEntityFromSlot(command, slot) {
    switch (slot) {
        case 'target':
        case 'directObject':
        case 'item':
            return command.directObject?.entity ?? null;
        case 'container':
        case 'recipient':
        case 'instrument':
        case 'indirectObject':
            return command.indirectObject?.entity ?? null;
        default:
            return command.directObject?.entity ?? null;
    }
}
/**
 * Generate appropriate scope error based on requirement and actual scope
 */
function getScopeError(required, actual, entity) {
    const params = { item: entity.name };
    if (actual === stdlib_1.ScopeLevel.UNAWARE) {
        return { valid: false, error: stdlib_1.ScopeErrors.NOT_KNOWN, params };
    }
    if (required >= stdlib_1.ScopeLevel.VISIBLE && actual < stdlib_1.ScopeLevel.VISIBLE) {
        return { valid: false, error: stdlib_1.ScopeErrors.NOT_VISIBLE, params };
    }
    if (required >= stdlib_1.ScopeLevel.REACHABLE && actual < stdlib_1.ScopeLevel.REACHABLE) {
        return { valid: false, error: stdlib_1.ScopeErrors.NOT_REACHABLE, params };
    }
    if (required >= stdlib_1.ScopeLevel.CARRIED && actual < stdlib_1.ScopeLevel.CARRIED) {
        return { valid: false, error: stdlib_1.ScopeErrors.NOT_CARRIED, params };
    }
    return { valid: false, error: stdlib_1.ScopeErrors.OUT_OF_SCOPE, params };
}
/**
 * Create an ActionContext for action execution
 */
function createActionContext(world, gameContext, command, action, scopeResolver) {
    const player = gameContext.player;
    const currentLocation = world.getLocation(player.id)
        ? world.getEntity(world.getLocation(player.id))
        : player;
    // Create sharedData object separately so it can be referenced in closures
    const sharedData = {};
    // Create the event method
    const event = (type, data) => {
        // Add standard entities
        const entities = {
            actor: player.id,
            location: currentLocation?.id || player.id
        };
        // Add entities from command
        if (command.directObject?.entity) {
            entities.target = command.directObject.entity.id;
        }
        if (command.indirectObject?.entity) {
            entities.indirect = command.indirectObject.entity.id;
        }
        // Create the event
        return {
            id: `evt_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`,
            type,
            timestamp: Date.now(),
            data: {
                ...data,
                // Include action context in data instead of metadata
                actionId: action.id,
                turn: gameContext.currentTurn
            },
            entities
        };
    };
    return {
        // World querying
        world,
        player,
        currentLocation: currentLocation,
        command,
        scopeResolver,
        action,
        // Shared data for passing information between phases
        sharedData,
        // Validation result - set by CommandExecutor after validate() phase
        validationResult: undefined,
        // Scope checking methods (delegate to scopeResolver)
        canSee: (entity) => {
            return scopeResolver.canSee(player, entity);
        },
        canReach: (entity) => {
            return scopeResolver.canReach(player, entity);
        },
        canTake: (entity) => {
            // Basic implementation - can be enhanced
            return !entity.has('if.trait.scenery') &&
                !entity.has('if.trait.fixed') &&
                entity.type !== 'room' &&
                entity.type !== 'location';
        },
        isInScope: (entity) => {
            const level = scopeResolver.getScope(player, entity);
            return level >= stdlib_1.ScopeLevel.AWARE;
        },
        getVisible: () => {
            return scopeResolver.getVisible(player);
        },
        getInScope: () => {
            // Get all entities and filter by scope
            const allEntities = world.getAllEntities();
            return allEntities.filter(entity => {
                const level = scopeResolver.getScope(player, entity);
                return level >= stdlib_1.ScopeLevel.AWARE;
            });
        },
        // Scope validation methods (Phase 4 parser refactor)
        getEntityScope: (entity) => {
            return scopeResolver.getScope(player, entity);
        },
        getSlotScope: (slot) => {
            const entity = getEntityFromSlot(command, slot);
            if (!entity) {
                return stdlib_1.ScopeLevel.UNAWARE;
            }
            return scopeResolver.getScope(player, entity);
        },
        requireScope: (entity, required) => {
            const actualScope = scopeResolver.getScope(player, entity);
            if (actualScope >= required) {
                return { ok: true, actualScope };
            }
            const error = getScopeError(required, actualScope, entity);
            return { ok: false, error, actualScope };
        },
        requireSlotScope: (slot, required) => {
            const entity = getEntityFromSlot(command, slot);
            if (!entity) {
                return {
                    ok: false,
                    error: { valid: false, error: 'no_target', params: { slot } }
                };
            }
            const actualScope = scopeResolver.getScope(player, entity);
            if (actualScope >= required) {
                return { ok: true, actualScope };
            }
            const error = getScopeError(required, actualScope, entity);
            return { ok: false, error, actualScope };
        },
        requireCarriedOrImplicitTake: (entity) => {
            const actualScope = scopeResolver.getScope(player, entity);
            // Case 1: Already carried - success, no implicit take needed
            if (actualScope >= stdlib_1.ScopeLevel.CARRIED) {
                return { ok: true };
            }
            // Case 2: Not reachable - can't take it, return scope error
            if (actualScope < stdlib_1.ScopeLevel.REACHABLE) {
                const error = getScopeError(stdlib_1.ScopeLevel.REACHABLE, actualScope, entity);
                return { ok: false, error };
            }
            // Case 3: Reachable but not carried - attempt implicit take
            // Check if implicit take is disabled at story level (ADR-104)
            if (gameContext.implicitActions?.implicitTake === false) {
                return {
                    ok: false,
                    error: { valid: false, error: stdlib_1.ScopeErrors.NOT_CARRIED, params: { item: entity.name } }
                };
            }
            // Check if implicit take is disabled at action level (ADR-104)
            if (action.allowImplicitTake === false) {
                return {
                    ok: false,
                    error: { valid: false, error: stdlib_1.ScopeErrors.NOT_CARRIED, params: { item: entity.name } }
                };
            }
            // Check if entity can be taken (not scenery, room, door)
            if (entity.has(world_model_1.TraitType.SCENERY) || entity.has(world_model_1.TraitType.ROOM) || entity.has(world_model_1.TraitType.DOOR)) {
                return {
                    ok: false,
                    error: { valid: false, error: 'fixed_in_place', params: { item: entity.name } }
                };
            }
            // Create a synthetic command for the taking action
            const takeCommand = {
                parsed: {
                    rawInput: `take ${entity.name}`,
                    action: stdlib_1.takingAction.id,
                    tokens: [],
                    structure: { verb: { tokens: [0], text: 'take', head: 'take' } },
                    pattern: 'VERB_NOUN',
                    confidence: 1.0
                },
                actionId: stdlib_1.takingAction.id,
                directObject: {
                    entity,
                    parsed: {
                        text: entity.name,
                        candidates: [entity.name]
                    }
                }
            };
            // Create a sub-context for the taking action
            const takeContext = createActionContext(world, gameContext, takeCommand, stdlib_1.takingAction, scopeResolver);
            // Run the taking action's validate phase
            const validation = stdlib_1.takingAction.validate(takeContext);
            if (!validation.valid) {
                // Take validation failed - return the error
                return {
                    ok: false,
                    error: {
                        valid: false,
                        error: validation.error || 'cannot_take',
                        params: validation.params
                    }
                };
            }
            // Execute the taking action
            stdlib_1.takingAction.execute(takeContext);
            // Get the report events from the taking action
            const takeEvents = stdlib_1.takingAction.report ? stdlib_1.takingAction.report(takeContext) : [];
            const filteredTakeEvents = takeEvents;
            // Create the implicit take event (for "(first taking the X)" message)
            const implicitTakeEvent = (0, core_1.createEvent)('if.event.implicit_take', {
                item: entity.id,
                itemName: entity.name
            }, {
                actor: player.id,
                target: entity.id,
                location: currentLocation?.id || player.id
            });
            // Combine: implicit take notification + state change events (no success message)
            const implicitTakeEvents = [implicitTakeEvent, ...filteredTakeEvents];
            // Store in sharedData for the report phase to access
            const existingEvents = sharedData[shared_data_keys_1.SharedDataKeys.IMPLICIT_TAKE_EVENTS] ?? [];
            sharedData[shared_data_keys_1.SharedDataKeys.IMPLICIT_TAKE_EVENTS] = [...existingEvents, ...implicitTakeEvents];
            return {
                ok: true,
                implicitTakeEvents
            };
        },
        // Event creation
        event
    };
}
