// @sharpee/engine/command-executor.js
"use strict";
/**
 * Command Executor - Orchestrates command pipeline
 *
 * Responsibilities:
 * - Orchestrate the four-phase pattern (validate → execute → report/blocked)
 * - Handle implicit inference (ADR-104) when validation fails with pronouns
 * - Pass results between phases
 * - Return the final TurnResult
 *
 * All event creation is owned by the action components themselves.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CommandExecutor = void 0;
exports.createCommandExecutor = createCommandExecutor;
const core_1 = require("../core/index.js");
const parser_interface_1 = require("./parser-interface");
const shared_data_keys_1 = require("./shared-data-keys");
const stdlib_1 = require("../stdlib/index.js");
const action_context_factory_1 = require("./action-context-factory");
const capability_dispatch_helper_1 = require("./capability-dispatch-helper");
class CommandExecutor {
    parser;
    validator;
    actionRegistry;
    eventProcessor;
    scopeResolver;
    parsedCommandTransformers = [];
    beforeActionListeners = [];
    constructor(world, actionRegistry, eventProcessor, parser, systemEvents) {
        if (!world)
            throw new Error('World model is required');
        if (!actionRegistry)
            throw new Error('Action registry is required');
        if (!eventProcessor)
            throw new Error('Event processor is required');
        if (!parser)
            throw new Error('Parser is required');
        this.parser = parser;
        this.validator = new stdlib_1.CommandValidator(world, actionRegistry);
        if (systemEvents) {
            this.validator.setSystemEventSource(systemEvents);
        }
        this.actionRegistry = actionRegistry;
        this.eventProcessor = eventProcessor;
    }
    /**
     * Validate a parsed command against the world model.
     *
     * @param command - The parsed command to validate
     * @returns Result with validated command or validation error
     */
    validateCommand(command) {
        return this.validator.validate(command);
    }
    /**
     * Register a transformer that can modify parsed commands before validation.
     * Transformers are called in order of registration.
     *
     * @param transformer - Function to transform parsed commands
     */
    registerParsedCommandTransformer(transformer) {
        this.parsedCommandTransformers.push(transformer);
    }
    /**
     * Unregister a previously registered transformer.
     *
     * @param transformer - The transformer to remove
     * @returns true if the transformer was found and removed
     */
    unregisterParsedCommandTransformer(transformer) {
        const index = this.parsedCommandTransformers.indexOf(transformer);
        if (index !== -1) {
            this.parsedCommandTransformers.splice(index, 1);
            return true;
        }
        return false;
    }
    /**
     * Register a listener for the pre-action hook (ADR-148).
     *
     * Listeners fire after command context creation but before the action's
     * validate phase. They can modify world state (e.g., break concealment).
     *
     * @param listener - The hook listener to register
     */
    onBeforeAction(listener) {
        this.beforeActionListeners.push(listener);
    }
    /**
     * Emit the pre-action hook to all registered listeners.
     */
    emitBeforeAction(data, world) {
        for (const listener of this.beforeActionListeners) {
            listener(data, world);
        }
    }
    async execute(input, world, context, config) {
        const turn = context.currentTurn;
        // Timing tracking
        const startTime = config?.collectTiming ? Date.now() : 0;
        let parseTime = 0;
        let executionTime = 0;
        try {
            // Set world context for parser entity resolution
            const player = world.getPlayer();
            if (player && (0, parser_interface_1.hasWorldContext)(this.parser)) {
                const playerLocation = world.getLocation(player.id) || '';
                this.parser.setWorldContext(world, player.id, playerLocation);
            }
            // Phase 1: Parse
            const parseStart = config?.collectTiming ? Date.now() : 0;
            const parseResult = this.parser.parse(input);
            if (config?.collectTiming) {
                parseTime = Date.now() - parseStart;
            }
            if (!parseResult.success) {
                throw new Error(`Parse failed: ${parseResult.error?.code || 'UNKNOWN'}`);
            }
            // Phase 1.5: Apply parsed command transformers
            // Allows stories to modify parsed commands before validation
            let parsedCommand = parseResult.value;
            for (const transformer of this.parsedCommandTransformers) {
                parsedCommand = transformer(parsedCommand, world);
            }
            // Phase 2: Validate
            const validationResult = this.validator.validate(parsedCommand);
            if (!validationResult.success) {
                // Check for disambiguation - emit client.query event instead of throwing
                if (validationResult.error.code === 'AMBIGUOUS_ENTITY') {
                    const details = validationResult.error.details || {};
                    const candidates = details.ambiguousEntities || [];
                    // Emit client.query event for disambiguation
                    const queryEvent = {
                        id: `query_disambig_${turn}_${Date.now()}`,
                        type: 'client.query',
                        timestamp: Date.now(),
                        entities: {},
                        data: {
                            source: core_1.QuerySource.DISAMBIGUATION,
                            type: core_1.QueryType.DISAMBIGUATION,
                            messageId: 'core.disambiguation_prompt',
                            candidates: candidates,
                            searchText: details.searchText,
                            originalCommand: parsedCommand
                        }
                    };
                    // Return early with query pending
                    return {
                        turn,
                        input,
                        success: false,
                        needsInput: true,
                        events: [queryEvent],
                        error: 'DISAMBIGUATION_NEEDED'
                    };
                }
                // Other validation errors still throw
                throw new Error(`Validation failed: ${validationResult.error.code}`);
            }
            // Phase 3: Execute action's three-phase pattern
            const executionStart = config?.collectTiming ? Date.now() : 0;
            const command = validationResult.value;
            const action = this.actionRegistry.get(command.actionId);
            if (!action) {
                throw new Error(`Action not found: ${command.actionId}`);
            }
            // Create context
            if (!this.scopeResolver) {
                this.scopeResolver = (0, stdlib_1.createScopeResolver)(world);
            }
            const actionContext = (0, action_context_factory_1.createActionContext)(world, context, command, action, this.scopeResolver);
            // Pre-action hook (ADR-148): listeners can modify world state before validation
            this.emitBeforeAction({
                actionId: command.actionId,
                actorId: context.player?.id,
                directObjectId: command.directObject?.entity?.id,
            }, world);
            // Universal Capability Dispatch: Check if any involved entity has a capability for this action
            // If so, the entity's behavior handles the action instead of the stdlib default
            // Check all entities: directObject, indirectObject, and any others
            const involvedEntities = [
                command.directObject?.entity,
                command.indirectObject?.entity
            ];
            const capabilityCheck = (0, capability_dispatch_helper_1.checkCapabilityDispatchMulti)(command.actionId, involvedEntities);
            // Run action's four phases: validate → execute → report (or blocked)
            // If capability dispatch applies, use capability behavior; otherwise use action
            let actionValidation = capabilityCheck.shouldDispatch
                ? (0, capability_dispatch_helper_1.executeCapabilityValidate)(capabilityCheck, actionContext)
                : action.validate(actionContext);
            let currentCommand = command;
            let currentContext = actionContext;
            let useCapabilityDispatch = capabilityCheck.shouldDispatch;
            // ADR-104: Implicit inference - if validation fails and pronoun was used,
            // try to find a valid alternative target
            // Check story-level config first
            const inferenceEnabled = context.implicitActions?.inference !== false;
            if (!actionValidation.valid && action.targetRequirements && inferenceEnabled) {
                const directObject = command.directObject;
                // Check if pronoun was used (INounPhrase has wasPronoun, cast is safe)
                const parsedNounPhrase = directObject?.parsed;
                const wasPronoun = parsedNounPhrase?.wasPronoun === true;
                if (wasPronoun && directObject?.entity) {
                    // Note: directObject.parsed is typed as IParsedObjectReference but at runtime
                    // it's actually an INounPhrase (command-validator sets it from the noun phrase)
                    // Get entities in scope for inference (visible entities)
                    const scopeEntities = this.scopeResolver.getVisible(world.getPlayer());
                    // Try to infer a different target
                    const inferenceResult = (0, stdlib_1.tryInferTarget)(directObject.entity, wasPronoun, action, scopeEntities, world);
                    if (inferenceResult.inferred && inferenceResult.inferredTarget) {
                        // Create a modified command with the inferred target
                        const inferredCommand = {
                            ...command,
                            directObject: {
                                entity: inferenceResult.inferredTarget,
                                parsed: {
                                    ...directObject.parsed,
                                    // Update text to reflect inferred entity
                                    text: inferenceResult.inferredTarget.name
                                }
                            }
                        };
                        // Create new context with inferred command
                        const inferredContext = (0, action_context_factory_1.createActionContext)(world, context, inferredCommand, action, this.scopeResolver);
                        // Mark that inference occurred (for "(the leaflet)" message)
                        const sharedData = inferredContext.sharedData;
                        sharedData[shared_data_keys_1.SharedDataKeys.INFERENCE_PERFORMED] = true;
                        sharedData[shared_data_keys_1.SharedDataKeys.ORIGINAL_TARGET] = directObject.entity;
                        sharedData[shared_data_keys_1.SharedDataKeys.INFERRED_TARGET] = inferenceResult.inferredTarget;
                        // Re-check capability dispatch for inferred target
                        const inferredInvolvedEntities = [
                            inferenceResult.inferredTarget,
                            command.indirectObject?.entity
                        ];
                        const inferredCapabilityCheck = (0, capability_dispatch_helper_1.checkCapabilityDispatchMulti)(command.actionId, inferredInvolvedEntities);
                        // Re-validate with inferred target (using capability dispatch if applicable)
                        const retryValidation = inferredCapabilityCheck.shouldDispatch
                            ? (0, capability_dispatch_helper_1.executeCapabilityValidate)(inferredCapabilityCheck, inferredContext)
                            : action.validate(inferredContext);
                        if (retryValidation.valid) {
                            // Inference succeeded - use the inferred command
                            actionValidation = retryValidation;
                            currentCommand = inferredCommand;
                            currentContext = inferredContext;
                            useCapabilityDispatch = inferredCapabilityCheck.shouldDispatch;
                        }
                    }
                }
            }
            // Thread validation result to later phases via context
            // This allows actions to access data from validate() in execute/report
            currentContext.validationResult = actionValidation;
            let events;
            if (actionValidation.valid) {
                if (useCapabilityDispatch) {
                    // Capability dispatch: use behavior phases
                    (0, capability_dispatch_helper_1.executeCapabilityExecute)(currentContext);
                    events = (0, capability_dispatch_helper_1.executeCapabilityReport)(currentContext);
                }
                else {
                    // Standard action: use action phases
                    const executeResult = action.execute(currentContext);
                    // Check pattern (new vs old)
                    if (executeResult === undefined || executeResult === null) {
                        // New pattern: use report() for success events only
                        if (action.report) {
                            events = action.report(currentContext);
                        }
                        else {
                            throw new Error(`Action ${action.id} uses new pattern but lacks report()`);
                        }
                    }
                    else {
                        // Old pattern: events from execute()
                        events = executeResult;
                    }
                }
            }
            else {
                // Validation failed - use blocked() for error events
                if (useCapabilityDispatch) {
                    // Capability dispatch: use behavior's blocked phase
                    events = (0, capability_dispatch_helper_1.executeCapabilityBlocked)(currentContext, actionValidation, command.actionId);
                }
                else if (action.blocked) {
                    events = action.blocked(currentContext, actionValidation);
                }
                else {
                    // Fallback for unmigrated actions
                    events = [{
                            id: `${turn}-error`,
                            type: 'action.error',
                            timestamp: Date.now(),
                            data: {
                                actionId: command.actionId,
                                messageId: actionValidation.error || 'validation_failed',
                                params: actionValidation.params || {}
                            },
                            entities: {}
                        }];
                }
            }
            if (config?.collectTiming) {
                executionTime = Date.now() - executionStart;
            }
            // Process events and collect reactions (ADR-052 entity handlers)
            let allEvents = events;
            if (events.length > 0) {
                const processed = this.eventProcessor.processEvents(events);
                // Add reaction events from entity handlers
                if (processed.reactions && processed.reactions.length > 0) {
                    allEvents = [...events, ...processed.reactions];
                }
            }
            const result = {
                turn,
                input,
                success: !events.some(e => e.type === 'action.error'),
                events: allEvents,
                actionId: command.actionId,
                parsedCommand: command.parsed,
                validatedCommand: command
            };
            // Add timing data if requested
            if (config?.collectTiming) {
                result.timing = {
                    parsing: parseTime,
                    execution: executionTime,
                    total: Date.now() - startTime
                };
            }
            return result;
        }
        catch (error) {
            // Minimal error handling - just return failure
            const result = {
                turn,
                input,
                success: false,
                events: [{
                        id: `cmd_failed_${turn}_${Date.now()}`,
                        type: 'command.failed',
                        timestamp: Date.now(),
                        entities: {},
                        data: { reason: error.message, input }
                    }],
                error: error.message
            };
            // Add timing data even for errors if requested
            if (config?.collectTiming) {
                result.timing = {
                    parsing: parseTime,
                    execution: executionTime,
                    total: Date.now() - startTime
                };
            }
            return result;
        }
    }
}
exports.CommandExecutor = CommandExecutor;
function createCommandExecutor(world, actionRegistry, eventProcessor, parser, systemEvents) {
    return new CommandExecutor(world, actionRegistry, eventProcessor, parser, systemEvents);
}
