// @sharpee/engine/event-sequencer.js
"use strict";
/**
 * Event Sequencer - Manages event ordering within turns
 *
 * Ensures all events have proper sequence numbers for ordering
 * and grouping within a turn.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.eventSequencer = exports.EventSequenceUtils = void 0;
/**
 * Event sequencer class
 */
class EventSequencer {
    counter = Date.now();
    /**
     * Get next sequence number
     */
    next() {
        return ++this.counter;
    }
    /**
     * Reset turn counter (optional, for testing)
     */
    resetTurn(turn) {
        // No-op in this implementation
    }
    /**
     * Sequence a single event
     */
    sequence(event, turn) {
        return {
            ...event,
            sequence: this.next(),
            timestamp: new Date(),
            turn,
            scope: 'turn'
        };
    }
    /**
     * Sequence multiple events
     */
    sequenceAll(events, turn) {
        return events.map(event => this.sequence(event, turn));
    }
}
/**
 * Event sequence utilities
 */
class EventSequenceUtils {
    /**
     * Sort events by sequence number
     */
    static sort(events) {
        return [...events].sort((a, b) => a.sequence - b.sequence);
    }
    /**
     * Filter events by type
     */
    static filterByType(events, type) {
        return events.filter(e => e.type === type);
    }
    /**
     * Filter events by turn
     */
    static filterByTurn(events, turn) {
        return events.filter(e => e.turn === turn);
    }
    /**
     * Filter events by scope
     */
    static filterByScope(events, scope) {
        return events.filter(e => e.scope === scope);
    }
    /**
     * Group events by type
     */
    static groupByType(events) {
        const groups = {};
        for (const event of events) {
            if (!groups[event.type]) {
                groups[event.type] = [];
            }
            groups[event.type].push(event);
        }
        return groups;
    }
    /**
     * Group events by turn
     */
    static groupByTurn(events) {
        const groups = {};
        for (const event of events) {
            if (!groups[event.turn]) {
                groups[event.turn] = [];
            }
            groups[event.turn].push(event);
        }
        return groups;
    }
    /**
     * Get latest event by type
     */
    static getLatestByType(events, type) {
        const filtered = this.filterByType(events, type);
        if (filtered.length === 0)
            return undefined;
        return filtered.reduce((latest, current) => current.sequence > latest.sequence ? current : latest);
    }
    /**
     * Count events by type
     */
    static countByType(events) {
        const counts = {};
        for (const event of events) {
            counts[event.type] = (counts[event.type] || 0) + 1;
        }
        return counts;
    }
    /**
     * Get events in sequence range
     */
    static getInRange(events, start, end, inclusive = true) {
        if (inclusive) {
            return events.filter(e => e.sequence >= start && e.sequence <= end);
        }
        else {
            return events.filter(e => e.sequence > start && e.sequence < end);
        }
    }
}
exports.EventSequenceUtils = EventSequenceUtils;
// Export singleton instance
exports.eventSequencer = new EventSequencer();
