// @sharpee/engine/platform-operations.js
"use strict";
/**
 * Platform Operations Handler - Handles platform events (save/restore/quit/restart/undo)
 *
 * Extracted from GameEngine as part of Phase 4 remediation.
 * Uses strategy pattern to handle different platform operation types.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PlatformOperationHandler = void 0;
exports.createPlatformOperationHandler = createPlatformOperationHandler;
const core_1 = require("../core/index.js");
const parser_interface_1 = require("./parser-interface");
/**
 * Handler for platform operations
 */
class PlatformOperationHandler {
    saveRestoreHooks;
    saveRestoreService;
    stateProvider;
    vocabularyManager;
    constructor(saveRestoreHooks, saveRestoreService, stateProvider, vocabularyManager) {
        this.saveRestoreHooks = saveRestoreHooks;
        this.saveRestoreService = saveRestoreService;
        this.stateProvider = stateProvider;
        this.vocabularyManager = vocabularyManager;
    }
    /**
     * Process all pending platform operations
     *
     * @param pendingOps - Array of pending platform operations
     * @param context - Platform operation context
     * @param engineCallbacks - Callbacks for engine-level operations
     */
    async processAll(pendingOps, context, engineCallbacks) {
        // Ensure there's an entry for the current turn
        if (!context.turnEvents.has(context.currentTurn)) {
            context.turnEvents.set(context.currentTurn, []);
        }
        for (const platformOp of pendingOps) {
            try {
                const resultEvent = await this.handleOperation(platformOp, context, engineCallbacks);
                if (resultEvent) {
                    // Add to event source and turn events
                    context.eventSource.emit(resultEvent);
                    context.turnEvents.get(context.currentTurn)?.push(resultEvent);
                    context.emitEvent(resultEvent);
                }
            }
            catch (error) {
                console.error(`Error processing platform operation ${platformOp.type}:`, error);
                // Emit appropriate error event
                const errorEvent = this.createErrorEvent(platformOp.type, error instanceof Error ? error.message : 'Unknown error');
                if (errorEvent) {
                    context.eventSource.emit(errorEvent);
                    context.turnEvents.get(context.currentTurn)?.push(errorEvent);
                    context.emitEvent(errorEvent);
                }
            }
        }
    }
    /**
     * Handle a single platform operation
     */
    async handleOperation(platformOp, context, engineCallbacks) {
        switch (platformOp.type) {
            case core_1.PlatformEventType.SAVE_REQUESTED:
                return this.handleSave(platformOp);
            case core_1.PlatformEventType.RESTORE_REQUESTED:
                return this.handleRestore(platformOp, engineCallbacks);
            case core_1.PlatformEventType.QUIT_REQUESTED:
                return this.handleQuit(platformOp, context, engineCallbacks);
            case core_1.PlatformEventType.RESTART_REQUESTED:
                return this.handleRestart(platformOp, context, engineCallbacks);
            case core_1.PlatformEventType.UNDO_REQUESTED:
                return this.handleUndo(engineCallbacks);
            default:
                return null;
        }
    }
    /**
     * Handle save request
     */
    async handleSave(platformOp) {
        if (!this.saveRestoreHooks?.onSaveRequested) {
            return (0, core_1.createSaveCompletedEvent)(false, 'No save handler registered');
        }
        const saveContext = platformOp.payload.context;
        const saveData = this.saveRestoreService.createSaveData(this.stateProvider);
        // Add any additional context from the platform event
        if (saveContext?.saveName) {
            saveData.metadata.description = saveContext.saveName;
        }
        if (saveContext?.metadata) {
            Object.assign(saveData.metadata, saveContext.metadata);
        }
        await this.saveRestoreHooks.onSaveRequested(saveData);
        return (0, core_1.createSaveCompletedEvent)(true);
    }
    /**
     * Handle restore request
     */
    async handleRestore(platformOp, engineCallbacks) {
        if (!this.saveRestoreHooks?.onRestoreRequested) {
            return (0, core_1.createRestoreCompletedEvent)(false, 'No restore handler registered');
        }
        const saveData = await this.saveRestoreHooks.onRestoreRequested();
        if (!saveData) {
            return (0, core_1.createRestoreCompletedEvent)(false, 'No save data available or restore cancelled');
        }
        const result = this.saveRestoreService.loadSaveData(saveData, this.stateProvider);
        // Update context
        engineCallbacks.updateContext({ currentTurn: result.currentTurn });
        // Reset pronoun context
        const parser = engineCallbacks.getParser();
        if (parser && (0, parser_interface_1.hasPronounContext)(parser)) {
            parser.resetPronounContext();
        }
        // Update vocabulary for current scope
        engineCallbacks.updateScopeVocabulary();
        engineCallbacks.emitStateChanged();
        return (0, core_1.createRestoreCompletedEvent)(true);
    }
    /**
     * Handle quit request
     */
    async handleQuit(platformOp, context, engineCallbacks) {
        const quitContext = platformOp.payload.context;
        if (this.saveRestoreHooks?.onQuitRequested) {
            const shouldQuit = await this.saveRestoreHooks.onQuitRequested(quitContext || {});
            if (shouldQuit) {
                engineCallbacks.stopEngine('quit');
                return (0, core_1.createQuitConfirmedEvent)();
            }
            else {
                return (0, core_1.createQuitCancelledEvent)();
            }
        }
        else {
            // No quit hook registered, auto-confirm
            return (0, core_1.createQuitConfirmedEvent)();
        }
    }
    /**
     * Handle restart request
     */
    async handleRestart(platformOp, context, engineCallbacks) {
        const restartContext = platformOp.payload.context;
        if (this.saveRestoreHooks?.onRestartRequested) {
            const shouldRestart = await this.saveRestoreHooks.onRestartRequested(restartContext || {});
            if (shouldRestart) {
                // Reset pronoun context
                const parser = engineCallbacks.getParser();
                if (parser && (0, parser_interface_1.hasPronounContext)(parser)) {
                    parser.resetPronounContext();
                }
                await engineCallbacks.restartStory();
                return (0, core_1.createRestartCompletedEvent)(true);
            }
            else {
                return (0, core_1.createRestartCompletedEvent)(false);
            }
        }
        else {
            // No restart hook registered - default behavior is to restart
            const parser = engineCallbacks.getParser();
            if (parser && (0, parser_interface_1.hasPronounContext)(parser)) {
                parser.resetPronounContext();
            }
            await engineCallbacks.restartStory();
            return (0, core_1.createRestartCompletedEvent)(true);
        }
    }
    /**
     * Handle undo request
     */
    handleUndo(engineCallbacks) {
        const world = this.stateProvider.getWorld();
        const result = this.saveRestoreService.undo(world);
        if (result) {
            // Update context with restored turn
            engineCallbacks.updateContext({ currentTurn: result.turn });
            // Update vocabulary for current scope
            engineCallbacks.updateScopeVocabulary();
            engineCallbacks.emitStateChanged();
            return (0, core_1.createUndoCompletedEvent)(true, result.turn);
        }
        else {
            return (0, core_1.createUndoCompletedEvent)(false, undefined, 'Nothing to undo');
        }
    }
    /**
     * Create an error event for a failed operation
     */
    createErrorEvent(operationType, errorMessage) {
        switch (operationType) {
            case core_1.PlatformEventType.SAVE_REQUESTED:
                return (0, core_1.createSaveCompletedEvent)(false, errorMessage);
            case core_1.PlatformEventType.RESTORE_REQUESTED:
                return (0, core_1.createRestoreCompletedEvent)(false, errorMessage);
            case core_1.PlatformEventType.QUIT_REQUESTED:
                return (0, core_1.createQuitCancelledEvent)();
            case core_1.PlatformEventType.RESTART_REQUESTED:
                return (0, core_1.createRestartCompletedEvent)(false);
            case core_1.PlatformEventType.UNDO_REQUESTED:
                return (0, core_1.createUndoCompletedEvent)(false, undefined, errorMessage);
            default:
                return null;
        }
    }
}
exports.PlatformOperationHandler = PlatformOperationHandler;
/**
 * Create a platform operation handler instance
 */
function createPlatformOperationHandler(saveRestoreHooks, saveRestoreService, stateProvider, vocabularyManager) {
    return new PlatformOperationHandler(saveRestoreHooks, saveRestoreService, stateProvider, vocabularyManager);
}
