// @sharpee/engine/save-restore-service.js
"use strict";
/**
 * Save/Restore Service - Manages game state persistence and undo
 *
 * Extracted from GameEngine as part of Phase 4 remediation.
 * Handles save/restore data creation, undo snapshots, and serialization.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SaveRestoreService = void 0;
exports.createSaveRestoreService = createSaveRestoreService;
const world_model_1 = require("../world-model/index.js");
const core_1 = require("../core/index.js");
/**
 * Service for managing save/restore and undo functionality
 */
class SaveRestoreService {
    // Undo system - circular buffer of world snapshots
    undoSnapshots = [];
    undoSnapshotTurns = [];
    maxUndoSnapshots;
    constructor(config) {
        this.maxUndoSnapshots = config?.maxSnapshots ?? 10;
    }
    /**
     * Create an undo snapshot of the current world state
     */
    createUndoSnapshot(world, currentTurn) {
        if (this.maxUndoSnapshots <= 0)
            return; // Undo disabled
        // Serialize world state
        const snapshot = world.toJSON();
        // Add to circular buffer
        this.undoSnapshots.push(snapshot);
        this.undoSnapshotTurns.push(currentTurn);
        // Trim if over limit
        while (this.undoSnapshots.length > this.maxUndoSnapshots) {
            this.undoSnapshots.shift();
            this.undoSnapshotTurns.shift();
        }
    }
    /**
     * Undo to previous turn
     * @returns The turn number restored to, or null if nothing to undo
     */
    undo(world) {
        if (this.undoSnapshots.length === 0) {
            return null; // Nothing to undo
        }
        // Pop the most recent snapshot
        const snapshot = this.undoSnapshots.pop();
        const turn = this.undoSnapshotTurns.pop();
        // Restore world state
        world.loadJSON(snapshot);
        return { turn };
    }
    /**
     * Check if undo is available
     */
    canUndo() {
        return this.undoSnapshots.length > 0;
    }
    /**
     * Get number of undo levels available
     */
    getUndoLevels() {
        return this.undoSnapshots.length;
    }
    /**
     * Clear all undo snapshots (e.g., after restore)
     */
    clearUndoSnapshots() {
        this.undoSnapshots = [];
        this.undoSnapshotTurns = [];
    }
    /**
     * Create save data from current engine state
     */
    createSaveData(provider) {
        const context = provider.getContext();
        const story = provider.getStory();
        const eventSource = provider.getEventSource();
        const pluginRegistry = provider.getPluginRegistry();
        const world = provider.getWorld();
        const parser = provider.getParser();
        const metadata = {
            storyId: story?.config.id || 'unknown',
            storyVersion: story?.config.version || '0.0.0',
            turnCount: context.currentTurn - 1,
            playTime: Date.now() - context.metadata.started.getTime(),
            description: `Turn ${context.currentTurn - 1}`
        };
        const engineState = {
            eventSource: this.serializeEventSource(eventSource),
            spatialIndex: this.serializeSpatialIndex(world),
            turnHistory: this.serializeTurnHistory(context.history),
            parserState: this.serializeParserState(parser),
            pluginStates: pluginRegistry.getStates()
        };
        return {
            version: '1.0.0',
            timestamp: Date.now(),
            metadata,
            engineState,
            storyConfig: {
                id: story?.config.id || 'unknown',
                version: story?.config.version || '0.0.0',
                title: story?.config.title || 'Unknown',
                author: Array.isArray(story?.config.author)
                    ? story.config.author.join(', ')
                    : (story?.config.author || 'Unknown')
            }
        };
    }
    /**
     * Load save data into engine state
     * @returns New event source with restored events
     */
    loadSaveData(saveData, provider) {
        const story = provider.getStory();
        // Validate save compatibility
        if (saveData.version !== '1.0.0') {
            throw new Error(`Unsupported save version: ${saveData.version}`);
        }
        if (saveData.storyConfig?.id &&
            story?.config.id &&
            saveData.storyConfig.id !== story.config.id) {
            throw new Error(`Save is for different story: ${saveData.storyConfig.id}`);
        }
        // Restore event source
        const eventSource = this.deserializeEventSource(saveData.engineState.eventSource);
        // Restore spatial index (world state)
        const world = provider.getWorld();
        this.deserializeSpatialIndex(saveData.engineState.spatialIndex, world);
        // Restore plugin states if present (ADR-120)
        if (saveData.engineState.pluginStates) {
            provider.getPluginRegistry().setStates(saveData.engineState.pluginStates);
        }
        // Clear undo snapshots after restore
        this.clearUndoSnapshots();
        return {
            eventSource,
            currentTurn: saveData.metadata.turnCount + 1
        };
    }
    /**
     * Serialize event source
     */
    serializeEventSource(eventSource) {
        const events = [];
        for (const event of eventSource.getAllEvents()) {
            events.push({
                id: event.id,
                type: event.type,
                timestamp: event.timestamp || Date.now(),
                data: this.serializeEventData(event.data)
            });
        }
        return events;
    }
    /**
     * Serialize event data, handling functions and special types
     */
    serializeEventData(data) {
        if (!data || typeof data !== 'object') {
            return (data || {});
        }
        const serialized = {};
        for (const [key, value] of Object.entries(data)) {
            if (typeof value === 'function') {
                // Mark functions for special handling during deserialization
                serialized[key] = { __type: 'function', __marker: '[Function]' };
            }
            else if (value && typeof value === 'object') {
                // Recursively serialize nested objects
                if (Array.isArray(value)) {
                    serialized[key] = value.map((item) => typeof item === 'object' ? this.serializeEventData(item) : item);
                }
                else {
                    serialized[key] = this.serializeEventData(value);
                }
            }
            else {
                // Primitive values can be stored directly
                serialized[key] = value;
            }
        }
        return serialized;
    }
    /**
     * Deserialize event source
     */
    deserializeEventSource(events) {
        const eventSource = (0, core_1.createSemanticEventSource)();
        for (const event of events) {
            eventSource.emit({
                id: event.id,
                type: event.type,
                timestamp: event.timestamp,
                data: this.deserializeEventData(event.data),
                entities: {}
            });
        }
        return eventSource;
    }
    /**
     * Deserialize event data, handling function markers
     */
    deserializeEventData(data) {
        if (!data || typeof data !== 'object') {
            return data;
        }
        // Check if this is a function marker
        if (data.__type === 'function') {
            // Return a placeholder function
            return () => '[Serialized Function]';
        }
        if (Array.isArray(data)) {
            return data.map((item) => this.deserializeEventData(item));
        }
        const deserialized = {};
        for (const [key, value] of Object.entries(data)) {
            deserialized[key] = this.deserializeEventData(value);
        }
        return deserialized;
    }
    /**
     * Serialize spatial index (world state)
     */
    serializeSpatialIndex(world) {
        const entities = {};
        const locations = {};
        const relationships = {};
        // Serialize all entities
        for (const entity of world.getAllEntities()) {
            const traits = {};
            // Serialize each trait
            for (const [name, trait] of entity.traits) {
                traits[name] = this.serializeTrait(trait);
            }
            entities[entity.id] = {
                id: entity.id,
                traits,
                entityType: entity.constructor.name
            };
        }
        // Serialize locations and their contents
        const allLocations = world.getAllEntities().filter((e) => e.type === 'room' || e.type === 'location' || e.has('if.trait.room'));
        for (const location of allLocations) {
            const contents = world.getContents(location.id);
            locations[location.id] = {
                id: location.id,
                properties: {
                    name: location.get(world_model_1.TraitType.IDENTITY)?.name || 'Unknown',
                    description: location.get(world_model_1.TraitType.IDENTITY)?.description || ''
                },
                contents: contents.map((e) => e.id),
                connections: this.extractConnections(location, world)
            };
        }
        return { entities, locations, relationships };
    }
    /**
     * Deserialize spatial index
     */
    deserializeSpatialIndex(index, world) {
        // Note: Full deserialization would need to clear and recreate the world
        // For now, this restores entity traits and locations
        // Restore entity traits
        for (const [id, data] of Object.entries(index.entities)) {
            const entity = world.getEntity(id);
            if (entity) {
                // Restore traits
                for (const [name, traitData] of Object.entries(data.traits)) {
                    const trait = this.deserializeTrait(name, traitData);
                    if (trait) {
                        // Remove existing trait if present, then add the restored one
                        if (entity.has(name)) {
                            entity.remove(name);
                        }
                        entity.add(trait);
                    }
                }
            }
        }
        // Restore locations and contents
        for (const [locationId, data] of Object.entries(index.locations)) {
            for (const entityId of data.contents) {
                const entity = world.getEntity(entityId);
                if (entity) {
                    world.moveEntity(entity.id, locationId);
                }
            }
        }
    }
    /**
     * Serialize turn history
     */
    serializeTurnHistory(history) {
        const turns = [];
        for (const [index, result] of history.entries()) {
            turns.push({
                turnNumber: index + 1,
                eventIds: result.events.map((e) => e.id),
                timestamp: result.events[0]?.timestamp || Date.now(),
                command: result.input
            });
        }
        return turns;
    }
    /**
     * Deserialize turn history
     */
    deserializeTurnHistory(turns, eventSource) {
        const history = [];
        for (const turn of turns) {
            // Find the events for this turn
            const events = eventSource.getAllEvents().filter((e) => turn.eventIds.includes(e.id));
            history.push({
                turn: turn.turnNumber,
                input: turn.command || '',
                success: true,
                events
            });
        }
        return history;
    }
    /**
     * Serialize parser state
     */
    serializeParserState(parser) {
        if (!parser) {
            return undefined;
        }
        // Parser state serialization is parser-specific
        // For now, return empty object
        return {};
    }
    /**
     * Serialize a trait
     */
    serializeTrait(trait) {
        if (typeof trait === 'object' && trait !== null) {
            return { ...trait };
        }
        return trait;
    }
    /**
     * Deserialize a trait
     */
    deserializeTrait(name, data) {
        if (data && typeof data === 'object') {
            return { type: name, ...data };
        }
        return null;
    }
    /**
     * Extract connections from a location entity
     */
    extractConnections(location, world) {
        const connections = {};
        // Check for ROOM trait with exits
        const roomTrait = location.get('if.trait.room');
        if (roomTrait?.exits) {
            Object.entries(roomTrait.exits).forEach(([direction, exit]) => {
                if (exit.destination) {
                    connections[direction] = exit.destination;
                }
            });
        }
        // Check for doors in this location
        const contents = world.getContents(location.id);
        contents.forEach((entity) => {
            const doorTrait = entity.get('if.trait.door');
            if (doorTrait) {
                const otherRoom = doorTrait.room1 === location.id ? doorTrait.room2 : doorTrait.room1;
                if (otherRoom) {
                    // Try to determine direction from door name
                    const name = entity.name?.toLowerCase() || '';
                    if (name.includes('north'))
                        connections.north = otherRoom;
                    else if (name.includes('south'))
                        connections.south = otherRoom;
                    else if (name.includes('east'))
                        connections.east = otherRoom;
                    else if (name.includes('west'))
                        connections.west = otherRoom;
                    else
                        connections.door = otherRoom;
                }
            }
        });
        return connections;
    }
}
exports.SaveRestoreService = SaveRestoreService;
/**
 * Create a save/restore service instance
 */
function createSaveRestoreService(config) {
    return new SaveRestoreService(config);
}
