// @sharpee/engine/scene-evaluation-plugin.js
"use strict";
/**
 * Scene evaluation turn plugin (ADR-149).
 *
 * Evaluates scene begin/end conditions each turn. Runs after NPC turns
 * and state machines, before daemons/fuses (priority 60).
 *
 * For each registered scene:
 * - If state='waiting' and begin() returns true → activate, emit scene_began
 * - If state='active' and end() returns true → end (or reset if recurring), emit scene_ended
 * - If state='active' → increment activeTurns
 *
 * Public interface: SceneEvaluationPlugin (TurnPlugin implementation).
 * Owner context: @sharpee/engine — turn cycle
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SceneEvaluationPlugin = void 0;
const world_model_1 = require("../world-model/index.js");
let eventCounter = 0;
/**
 * Creates a minimal ISemanticEvent with the given type and data.
 */
function sceneEvent(type, data) {
    return {
        id: `scene-${++eventCounter}-${Date.now()}`,
        type,
        timestamp: Date.now(),
        entities: {},
        data,
    };
}
class SceneEvaluationPlugin {
    id = 'sharpee.scene-evaluation';
    priority = 60;
    /**
     * Evaluates all registered scene conditions after a successful action.
     */
    onAfterAction(context) {
        const { world, turn } = context;
        const conditions = world.getAllSceneConditions();
        if (conditions.size === 0)
            return [];
        const events = [];
        for (const [sceneId, conds] of conditions) {
            const entity = world.getEntity(sceneId);
            if (!entity)
                continue;
            const trait = entity.get(world_model_1.TraitType.SCENE);
            if (!trait)
                continue;
            if (trait.state === 'waiting') {
                // Evaluate begin condition
                try {
                    if (conds.begin(world)) {
                        trait.state = 'active';
                        trait.activeTurns = 1;
                        trait.beganAtTurn = turn;
                        events.push(sceneEvent('if.event.scene_began', {
                            sceneId,
                            sceneName: trait.name,
                            turn,
                        }));
                    }
                }
                catch {
                    // Condition threw — leave scene in waiting state
                }
            }
            else if (trait.state === 'active') {
                // Evaluate end condition
                let ended = false;
                try {
                    ended = conds.end(world);
                }
                catch {
                    // Condition threw — don't end the scene
                }
                if (ended) {
                    const totalTurns = trait.activeTurns;
                    trait.endedAtTurn = turn;
                    if (trait.recurring) {
                        trait.state = 'waiting';
                    }
                    else {
                        trait.state = 'ended';
                    }
                    trait.activeTurns = 0;
                    events.push(sceneEvent('if.event.scene_ended', {
                        sceneId,
                        sceneName: trait.name,
                        turn,
                        totalTurns,
                    }));
                }
                else {
                    // Scene still active — increment turn counter
                    trait.activeTurns++;
                }
            }
            // state === 'ended' and not recurring: skip
        }
        return events;
    }
}
exports.SceneEvaluationPlugin = SceneEvaluationPlugin;
