// @sharpee/engine/shared-data-keys.js
"use strict";
/**
 * Typed constants for sharedData keys used by the engine orchestration.
 *
 * SharedData is passed between action phases (validate/execute/report/blocked)
 * and allows phases to communicate without modifying the context directly.
 *
 * This file defines only engine-level keys. Actions define their own
 * action-specific keys as needed.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SharedDataKeys = void 0;
/**
 * Constants for engine-level sharedData keys.
 * Using constants prevents typos and enables IDE autocomplete.
 */
exports.SharedDataKeys = {
    /** Whether pronoun inference was performed (ADR-104) */
    INFERENCE_PERFORMED: 'inferencePerformed',
    /** The original target entity before inference */
    ORIGINAL_TARGET: 'originalTarget',
    /** The inferred target entity after inference */
    INFERRED_TARGET: 'inferredTarget',
    /** Events from implicit take actions (ADR-104) */
    IMPLICIT_TAKE_EVENTS: 'implicitTakeEvents',
    /** Validation result from the validate phase */
    VALIDATION_RESULT: 'validationResult',
};
