// @sharpee/engine/story.js
"use strict";
/**
 * Story configuration and interfaces
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.StoryWithEvents = void 0;
exports.validateStoryConfig = validateStoryConfig;
const event_emitter_1 = require("./events/event-emitter");
/**
 * Extended story class with event handling capabilities
 */
class StoryWithEvents {
    config;
    eventEmitter;
    constructor(config) {
        this.config = config;
        this.eventEmitter = new event_emitter_1.EventEmitter();
    }
    /**
     * Register a story-level event handler (daemon)
     */
    on(eventType, handler) {
        this.eventEmitter.on(eventType, handler);
    }
    /**
     * Remove a story-level event handler
     */
    off(eventType, handler) {
        this.eventEmitter.off(eventType, handler);
    }
    /**
     * Emit an event to story-level handlers
     * Returns any semantic events generated by handlers
     */
    emit(event) {
        return this.eventEmitter.emit(event);
    }
    // Abstract methods that must be implemented by subclasses
    initializeWorld(world) {
        throw new Error('initializeWorld must be implemented');
    }
    createPlayer(world) {
        throw new Error('createPlayer must be implemented');
    }
}
exports.StoryWithEvents = StoryWithEvents;
/**
 * Validate story configuration
 */
function validateStoryConfig(config) {
    if (!config.id || !config.title || !config.author || !config.version) {
        throw new Error('Missing required story configuration fields');
    }
    // Validate semantic version (allows prerelease tags like -beta, -alpha.1, etc.)
    const versionRegex = /^\d+\.\d+\.\d+(-[a-zA-Z0-9.-]+)?$/;
    if (!versionRegex.test(config.version)) {
        throw new Error('Invalid version format. Must be semantic version (e.g., 1.0.0 or 1.0.0-beta)');
    }
}
