// @sharpee/engine/turn-event-processor.js
"use strict";
/**
 * Turn Event Processor - Processes events during turn execution
 *
 * Extracted from GameEngine as part of Phase 4 remediation.
 * Handles event enrichment, perception filtering, and event emission.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TurnEventProcessor = void 0;
exports.processEvent = processEvent;
exports.createTurnEventProcessor = createTurnEventProcessor;
const core_1 = require("../core/index.js");
/**
 * Generate a unique event ID
 */
function generateEventId() {
    return `event-${Date.now()}-${Math.random().toString(36).substr(2, 9)}`;
}
/**
 * Event normalization - ensures consistent event structure
 */
function normalizeEvent(event) {
    return {
        ...event,
        id: event.id || generateEventId(),
        type: event.type.toLowerCase(),
        timestamp: event.timestamp || Date.now(),
        entities: event.entities || {},
        data: event.data,
        tags: event.tags,
        priority: event.priority,
        narrate: event.narrate
    };
}
/**
 * Event enrichment - adds turn, actor, and location context
 */
function enrichEvent(event, context) {
    const enriched = { ...event };
    if (context) {
        if (context.turn !== undefined && enriched.data && typeof enriched.data === 'object') {
            enriched.data = { ...enriched.data, turn: context.turn };
        }
        if (context.playerId && !enriched.entities.actor) {
            enriched.entities = { ...enriched.entities, actor: context.playerId };
        }
        if (context.locationId && !enriched.entities.location) {
            enriched.entities = { ...enriched.entities, location: context.locationId };
        }
    }
    if (!enriched.tags) {
        enriched.tags = [];
    }
    if (enriched.type.startsWith('action.') && !enriched.tags.includes('action')) {
        enriched.tags = [...enriched.tags, 'action'];
    }
    else if (enriched.type.startsWith('system.') && !enriched.tags.includes('system')) {
        enriched.tags = [...enriched.tags, 'system'];
    }
    else if (enriched.type.startsWith('game.') && !enriched.tags.includes('game')) {
        enriched.tags = [...enriched.tags, 'game'];
    }
    return enriched;
}
/**
 * Process an event through normalization and enrichment
 */
function processEvent(event, context) {
    return enrichEvent(normalizeEvent(event), context);
}
/**
 * Service for processing turn events
 */
class TurnEventProcessor {
    perceptionService;
    constructor(perceptionService) {
        this.perceptionService = perceptionService;
    }
    /**
     * Process action events from command execution
     *
     * @param events - Raw events from command executor
     * @param enrichmentContext - Context for event enrichment
     * @param player - Player entity for perception filtering
     * @param world - World model for perception filtering
     * @returns Processed events and platform events
     */
    processActionEvents(events, enrichmentContext, player, world) {
        const context = {
            turn: enrichmentContext.turn,
            playerId: enrichmentContext.playerId,
            locationId: enrichmentContext.locationId
        };
        let semanticEvents = events.map((e) => processEvent(e, context));
        // Apply perception filtering if service is configured
        if (this.perceptionService) {
            semanticEvents = this.perceptionService.filterEvents(semanticEvents, player, world);
        }
        // Check for platform request events
        const platformEvents = [];
        for (const event of semanticEvents) {
            if ((0, core_1.isPlatformRequestEvent)(event)) {
                platformEvents.push(event);
            }
        }
        return { semanticEvents, platformEvents };
    }
    /**
     * Process semantic events (e.g., from NPC or scheduler ticks)
     *
     * @param events - Semantic events to process
     * @param enrichmentContext - Context for event enrichment
     * @param player - Player entity for perception filtering
     * @param world - World model for perception filtering
     * @returns Processed events and platform events
     */
    processSemanticEvents(events, enrichmentContext, player, world) {
        const context = {
            turn: enrichmentContext.turn,
            playerId: enrichmentContext.playerId,
            locationId: enrichmentContext.locationId
        };
        // Process events through the pipeline
        let semanticEvents = events.map((e) => processEvent(e, context));
        // Apply perception filtering if service is configured
        if (this.perceptionService) {
            semanticEvents = this.perceptionService.filterEvents(semanticEvents, player, world);
        }
        // Check for platform request events
        const platformEvents = [];
        for (const event of semanticEvents) {
            if ((0, core_1.isPlatformRequestEvent)(event)) {
                platformEvents.push(event);
            }
        }
        return { semanticEvents, platformEvents };
    }
    /**
     * Emit events through all configured channels
     *
     * @param semanticEvents - Events to emit
     * @param eventSource - Event source for tracking
     * @param turnEvents - Turn events map to update
     * @param turn - Current turn number
     * @param config - Engine config with event callback
     * @param eventEmitter - Callback for engine event emission
     * @param entityDispatcher - Optional callback for entity handler dispatch
     */
    emitEvents(semanticEvents, eventSource, turnEvents, turn, config, eventEmitter, entityDispatcher) {
        // Store events for this turn
        const existingEvents = turnEvents.get(turn) || [];
        turnEvents.set(turn, [...existingEvents, ...semanticEvents]);
        // Track in event source for save/restore
        for (const event of semanticEvents) {
            eventSource.emit(event);
        }
        // Emit events if configured
        if (config.onEvent) {
            for (const event of semanticEvents) {
                config.onEvent(event);
            }
        }
        // Emit through engine's event system
        for (const event of semanticEvents) {
            eventEmitter(event);
            // Dispatch to entity handlers if provided
            if (entityDispatcher) {
                entityDispatcher(event);
            }
        }
    }
    /**
     * Check for victory events in the processed events
     *
     * @param events - Events to check
     * @returns Victory details if found, null otherwise
     */
    checkForVictory(events) {
        for (const event of events) {
            if (event.type === 'story.victory') {
                const data = event.data;
                return {
                    reason: data?.reason || 'Story completed',
                    score: data?.score || 0
                };
            }
        }
        return null;
    }
}
exports.TurnEventProcessor = TurnEventProcessor;
/**
 * Create a turn event processor instance
 */
function createTurnEventProcessor(perceptionService) {
    return new TurnEventProcessor(perceptionService);
}
