// @sharpee/engine/vocabulary-manager.js
"use strict";
/**
 * Vocabulary Manager - Manages entity vocabulary for parser scope resolution
 *
 * Extracted from GameEngine as part of Phase 4 remediation.
 * Handles registering entity nouns/adjectives with the vocabulary registry
 * to enable parser noun resolution within the current scope.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.VocabularyManager = void 0;
exports.createVocabularyManager = createVocabularyManager;
const stdlib_1 = require("../stdlib/index.js");
/**
 * Manages vocabulary registration for entities in scope
 */
class VocabularyManager {
    /**
     * Update vocabulary for a single entity
     *
     * @param entity - The entity to register
     * @param inScope - Whether the entity is currently in scope
     */
    updateEntityVocabulary(entity, inScope) {
        const identityTrait = entity.get('IDENTITY');
        if (identityTrait && typeof identityTrait === 'object') {
            const identity = identityTrait;
            // Build nouns from name and aliases
            const nouns = [];
            if (identity.name) {
                nouns.push(identity.name.toLowerCase());
            }
            if (identity.aliases && Array.isArray(identity.aliases)) {
                nouns.push(...identity.aliases);
            }
            stdlib_1.vocabularyRegistry.registerEntity({
                entityId: entity.id,
                nouns: nouns,
                adjectives: identity.adjectives || [],
                inScope
            });
        }
    }
    /**
     * Update vocabulary for all entities based on current scope
     *
     * Marks all entities as out of scope first, then marks
     * entities visible to the player as in scope.
     *
     * @param world - The world model
     * @param playerId - The player entity ID
     */
    updateScopeVocabulary(world, playerId) {
        const player = world.getEntity(playerId);
        if (!player)
            return;
        const inScope = world.getInScope(playerId);
        // Mark all entities as out of scope first
        for (const entity of world.getAllEntities()) {
            this.updateEntityVocabulary(entity, false);
        }
        // Mark in-scope entities
        for (const entity of inScope) {
            this.updateEntityVocabulary(entity, true);
        }
    }
}
exports.VocabularyManager = VocabularyManager;
/**
 * Create a vocabulary manager instance
 */
function createVocabularyManager() {
    return new VocabularyManager();
}
