// @sharpee/ext-testing/checkpoints/store.js
"use strict";
/**
 * Checkpoint Store
 *
 * File-based storage for checkpoints.
 * Supports both Node.js (filesystem) and browser (localStorage) environments.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.createFileStore = createFileStore;
exports.createMemoryStore = createMemoryStore;
exports.createLocalStorageStore = createLocalStorageStore;
const serializer_js_1 = require("./serializer.js");
/**
 * Create a file-based checkpoint store (Node.js)
 */
function createFileStore(directory) {
    // Dynamic import for Node.js fs module
    let fs;
    let path;
    const ensureFs = async () => {
        if (!fs) {
            fs = await Promise.resolve().then(() => __importStar(require('fs')));
            path = await Promise.resolve().then(() => __importStar(require('path')));
        }
        return { fs, path: path };
    };
    const getFilePath = (pathModule, name) => {
        return pathModule.join(directory, `${name}.json`);
    };
    const store = {
        async save(name, data) {
            const { fs, path } = await ensureFs();
            // Ensure directory exists
            if (!fs.existsSync(directory)) {
                fs.mkdirSync(directory, { recursive: true });
            }
            const filePath = getFilePath(path, name);
            const json = JSON.stringify(data, null, 2);
            fs.writeFileSync(filePath, json, 'utf-8');
        },
        async load(name) {
            const { fs, path } = await ensureFs();
            const filePath = getFilePath(path, name);
            if (!fs.existsSync(filePath)) {
                return undefined;
            }
            try {
                const json = fs.readFileSync(filePath, 'utf-8');
                const data = JSON.parse(json);
                if (!(0, serializer_js_1.validateCheckpoint)(data)) {
                    console.warn(`Invalid checkpoint data in ${filePath}`);
                    return undefined;
                }
                return data;
            }
            catch (error) {
                console.error(`Failed to load checkpoint ${name}:`, error);
                return undefined;
            }
        },
        async list() {
            const { fs } = await ensureFs();
            if (!fs.existsSync(directory)) {
                return [];
            }
            const files = fs.readdirSync(directory);
            return files
                .filter((f) => f.endsWith('.json'))
                .map((f) => f.replace(/\.json$/, ''));
        },
        async delete(name) {
            const { fs, path } = await ensureFs();
            const filePath = getFilePath(path, name);
            if (!fs.existsSync(filePath)) {
                return false;
            }
            fs.unlinkSync(filePath);
            return true;
        },
        async exists(name) {
            const { fs, path } = await ensureFs();
            const filePath = getFilePath(path, name);
            return fs.existsSync(filePath);
        },
    };
    return store;
}
/**
 * Create a memory-based checkpoint store (for testing or browser)
 */
function createMemoryStore() {
    const checkpoints = new Map();
    return {
        async save(name, data) {
            checkpoints.set(name, data);
        },
        async load(name) {
            return checkpoints.get(name);
        },
        async list() {
            return Array.from(checkpoints.keys());
        },
        async delete(name) {
            return checkpoints.delete(name);
        },
        async exists(name) {
            return checkpoints.has(name);
        },
    };
}
/**
 * Create a localStorage-based checkpoint store (browser)
 */
function createLocalStorageStore(prefix = 'sharpee-checkpoint-') {
    const getKey = (name) => `${prefix}${name}`;
    return {
        async save(name, data) {
            const json = JSON.stringify(data);
            localStorage.setItem(getKey(name), json);
        },
        async load(name) {
            const json = localStorage.getItem(getKey(name));
            if (!json)
                return undefined;
            try {
                const data = JSON.parse(json);
                if (!(0, serializer_js_1.validateCheckpoint)(data)) {
                    console.warn(`Invalid checkpoint data for ${name}`);
                    return undefined;
                }
                return data;
            }
            catch {
                return undefined;
            }
        },
        async list() {
            const names = [];
            for (let i = 0; i < localStorage.length; i++) {
                const key = localStorage.key(i);
                if (key?.startsWith(prefix)) {
                    names.push(key.slice(prefix.length));
                }
            }
            return names;
        },
        async delete(name) {
            const key = getKey(name);
            if (localStorage.getItem(key) === null) {
                return false;
            }
            localStorage.removeItem(key);
            return true;
        },
        async exists(name) {
            return localStorage.getItem(getKey(name)) !== null;
        },
    };
}
