// @sharpee/world-model/behaviors/attack.js
"use strict";
/**
 * Attack behavior coordinator
 *
 * Coordinates the various combat behaviors to handle attacks
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.AttackBehavior = void 0;
const trait_types_1 = require("../traits/trait-types");
const equippedTrait_1 = require("../traits/equipped/equippedTrait");
const weaponTrait_1 = require("../traits/weapon/weaponTrait");
const weaponBehavior_1 = require("../traits/weapon/weaponBehavior");
const breakableBehavior_1 = require("../traits/breakable/breakableBehavior");
const destructibleBehavior_1 = require("../traits/destructible/destructibleBehavior");
const combatantBehavior_1 = require("../traits/combatant/combatantBehavior");
/**
 * Attack behavior coordinator.
 *
 * This behavior coordinates attacks using weapon, breakable, destructible, and combat behaviors.
 * It determines which behavior to use based on the target's traits.
 */
class AttackBehavior {
    /**
     * Perform an attack
     * @param target The entity being attacked
     * @param weapon Optional weapon being used
     * @param world The world model
     * @returns Combined attack result
     */
    static attack(target, weapon, world) {
        // Calculate weapon damage if using a weapon
        let weaponDamage = 1; // Default unarmed damage
        let weaponType;
        let weaponResult;
        if (weapon && weapon.has(trait_types_1.TraitType.WEAPON)) {
            weaponResult = weaponBehavior_1.WeaponBehavior.calculateDamage(weapon);
            weaponDamage = weaponResult.damage;
            weaponType = weaponResult.weaponType;
        }
        // Try behaviors in order: Breakable, Destructible, Combatant
        // 1. Check if target is breakable (one-hit destruction)
        if (target.has(trait_types_1.TraitType.BREAKABLE)) {
            const breakResult = breakableBehavior_1.BreakableBehavior.break(target, world);
            if (breakResult.success) {
                return {
                    success: true,
                    type: 'broke',
                    damage: weaponDamage,
                    targetDestroyed: true,
                    debrisCreated: breakResult.debrisCreated,
                    weaponBroke: weaponResult?.weaponBroke,
                    message: breakResult.message
                };
            }
        }
        // 2. Check if target is destructible (multi-hit with HP)
        if (target.has(trait_types_1.TraitType.DESTRUCTIBLE)) {
            const damageResult = destructibleBehavior_1.DestructibleBehavior.damage(target, weaponDamage, weaponType, world);
            if (damageResult.success) {
                return {
                    success: true,
                    type: damageResult.destroyed ? 'destroyed' : 'damaged',
                    damage: damageResult.damage,
                    remainingHitPoints: damageResult.remainingHitPoints,
                    targetDestroyed: damageResult.destroyed,
                    transformedTo: damageResult.transformedTo,
                    exitRevealed: damageResult.exitRevealed,
                    weaponBroke: weaponResult?.weaponBroke,
                    message: damageResult.message
                };
            }
            else {
                // Attack failed due to requirements
                return {
                    success: false,
                    type: 'ineffective',
                    damage: 0,
                    message: damageResult.requiresWeapon
                        ? 'You need a weapon to damage that.'
                        : damageResult.wrongWeaponType
                            ? 'That weapon won\'t work on this target.'
                            : damageResult.invulnerable
                                ? 'That cannot be damaged.'
                                : 'Your attack has no effect.'
                };
            }
        }
        // 3. Check if target is a combatant (has health)
        if (target.has(trait_types_1.TraitType.COMBATANT)) {
            const combatResult = combatantBehavior_1.CombatBehavior.attack(target, weaponDamage, world);
            if (combatResult.success) {
                return {
                    success: true,
                    type: combatResult.killed ? 'killed' : 'hit',
                    damage: combatResult.damage,
                    remainingHitPoints: combatResult.remainingHealth,
                    targetKilled: combatResult.killed,
                    itemsDropped: combatResult.droppedItems,
                    weaponBroke: weaponResult?.weaponBroke,
                    message: combatResult.killed ? combatResult.deathMessage : combatResult.message
                };
            }
        }
        // 4. Target has no combat-related traits - attack is ineffective
        return {
            success: false,
            type: 'ineffective',
            damage: 0,
            message: 'Your attack has no effect on that.'
        };
    }
    /**
     * Check if a target can be attacked
     */
    static canAttack(target, weapon) {
        // Can attack if target has any combat-related trait
        return target.has(trait_types_1.TraitType.BREAKABLE) ||
            target.has(trait_types_1.TraitType.DESTRUCTIBLE) ||
            target.has(trait_types_1.TraitType.COMBATANT);
    }
    /**
     * Infer the best weapon from inventory
     * @param inventory Array of entities in player's inventory
     * @returns The best weapon found, or undefined
     */
    static inferWeapon(inventory) {
        // Find all weapons
        const weapons = inventory.filter(item => item.has(trait_types_1.TraitType.WEAPON));
        if (weapons.length === 0)
            return undefined;
        // Prefer equipped weapons
        const equipped = weapons.filter(w => {
            if (!w.has(trait_types_1.TraitType.EQUIPPED))
                return false;
            const equippedTrait = w.get(equippedTrait_1.EquippedTrait);
            return equippedTrait && equippedTrait.isEquipped;
        });
        if (equipped.length > 0) {
            // Return equipped weapon with highest max damage
            return equipped.reduce((best, current) => {
                const bestWeapon = best.get(weaponTrait_1.WeaponTrait);
                const currentWeapon = current.get(weaponTrait_1.WeaponTrait);
                if (!bestWeapon || !currentWeapon)
                    return best;
                const bestDamage = bestWeapon.maxDamage || 0;
                const currentDamage = currentWeapon.maxDamage || 0;
                return currentDamage > bestDamage ? current : best;
            });
        }
        // No equipped weapons, return weapon with highest max damage
        return weapons.reduce((best, current) => {
            const bestWeapon = best.get(weaponTrait_1.WeaponTrait);
            const currentWeapon = current.get(weaponTrait_1.WeaponTrait);
            if (!bestWeapon || !currentWeapon)
                return best;
            const bestDamage = bestWeapon.maxDamage || 0;
            const currentDamage = currentWeapon.maxDamage || 0;
            return currentDamage > bestDamage ? current : best;
        });
    }
}
exports.AttackBehavior = AttackBehavior;
