// @sharpee/stdlib/actions/standard/locking/locking.js
"use strict";
/**
 * Locking action - locks containers and doors
 *
 * This action properly delegates to LockableBehavior for validation
 * and execution. It follows the three-phase pattern.
 *
 * Uses three-phase pattern:
 * 1. validate: Check if target is lockable and key requirements
 * 2. execute: Call LockableBehavior.lock(), store result in sharedData
 * 3. report: Generate events from sharedData
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.lockingAction = void 0;
const world_model_1 = require("../../../../world-model/index.js");
const constants_1 = require("../../constants");
const scope_1 = require("../../../scope");
const lock_shared_1 = require("../lock-shared");
const locking_messages_1 = require("./locking-messages");
function getLockingSharedData(context) {
    return context.sharedData;
}
exports.lockingAction = {
    id: constants_1.IFActions.LOCKING,
    // Default scope requirements for this action's slots
    defaultScope: {
        target: scope_1.ScopeLevel.REACHABLE,
        key: scope_1.ScopeLevel.CARRIED
    },
    requiredMessages: [
        'no_target',
        'not_lockable',
        'no_key',
        'wrong_key',
        'already_locked',
        'not_closed',
        'locked',
        'locked_with',
        'cant_reach',
        'key_not_held'
    ],
    group: 'lock_manipulation',
    /**
     * Validate whether the lock action can be executed
     * Uses behavior validation methods to check preconditions
     */
    validate(context) {
        const noun = context.command.directObject?.entity;
        // ADR-080: Prefer instrument field (from .instrument() patterns), fall back to indirectObject
        const withKey = context.command.instrument?.entity ?? context.command.indirectObject?.entity;
        // Validate we have a target
        if (!noun) {
            return {
                valid: false,
                error: locking_messages_1.MESSAGES.NO_TARGET
            };
        }
        // Check scope - must be able to reach the target
        const scopeCheck = context.requireScope(noun, scope_1.ScopeLevel.REACHABLE);
        if (!scopeCheck.ok) {
            return scopeCheck.error;
        }
        // Check if it's lockable
        if (!noun.has(world_model_1.TraitType.LOCKABLE)) {
            return {
                valid: false,
                error: locking_messages_1.MESSAGES.NOT_LOCKABLE,
                params: { item: noun.name }
            };
        }
        // Use behavior's canLock method for validation
        if (!world_model_1.LockableBehavior.canLock(noun)) {
            // Check specific reason why it can't be locked
            if (world_model_1.LockableBehavior.isLocked(noun)) {
                return {
                    valid: false,
                    error: locking_messages_1.MESSAGES.ALREADY_LOCKED,
                    params: { item: noun.name }
                };
            }
            if (noun.has(world_model_1.TraitType.OPENABLE) && world_model_1.OpenableBehavior.isOpen(noun)) {
                return {
                    valid: false,
                    error: locking_messages_1.MESSAGES.NOT_CLOSED,
                    params: { item: noun.name }
                };
            }
        }
        // Validate key requirements using shared helper
        const keyValidation = (0, lock_shared_1.validateKeyRequirements)(context, noun, withKey, true);
        if (keyValidation) {
            return keyValidation;
        }
        return { valid: true };
    },
    /**
     * Execute the lock action - performs mutations only
     * Assumes validation has already passed
     * Delegates to LockableBehavior for actual state changes
     */
    execute(context) {
        const noun = context.command.directObject.entity;
        // ADR-080: Prefer instrument field, fall back to indirectObject
        const withKey = context.command.instrument?.entity ?? context.command.indirectObject?.entity;
        const sharedData = getLockingSharedData(context);
        // Analyze the locking context
        const analysis = (0, lock_shared_1.analyzeLockContext)(context, noun, withKey);
        // Delegate to behavior for locking (world mutation)
        const result = world_model_1.LockableBehavior.lock(noun, withKey);
        // Store basic info
        sharedData.targetId = noun.id;
        sharedData.targetName = noun.name;
        // Check if the behavior reported failure
        if (!result.success) {
            sharedData.failed = true;
            // Map behavior failure to error message
            if (result.wrongKey) {
                sharedData.errorMessageId = 'wrong_key';
            }
            else if (result.noKey) {
                sharedData.errorMessageId = 'no_key';
            }
            else if (result.alreadyLocked) {
                sharedData.errorMessageId = 'already_locked';
            }
            else {
                sharedData.errorMessageId = 'cant_lock';
            }
            return;
        }
        // Locking succeeded
        sharedData.failed = false;
        // Add type information from analysis
        if (analysis.isContainer) {
            sharedData.isContainer = true;
        }
        if (analysis.isDoor) {
            sharedData.isDoor = true;
        }
        // Add key information
        if (withKey) {
            sharedData.keyId = withKey.id;
            sharedData.keyName = withKey.name;
        }
        // Add sound information from result
        if (result.lockSound) {
            sharedData.sound = result.lockSound;
        }
        // Determine success message
        sharedData.messageId = (0, lock_shared_1.determineLockMessage)(true, !!withKey);
        sharedData.params = {
            item: noun.name
        };
        // Add container/door info
        if (analysis.isContainer) {
            sharedData.params.isContainer = true;
        }
        if (analysis.isDoor) {
            sharedData.params.isDoor = true;
        }
        if (withKey) {
            sharedData.params.key = withKey.name;
        }
    },
    /**
     * Report phase - generates events after successful execution
     *
     * Uses simplified event pattern (ADR-097): domain event carries messageId directly.
     */
    report(context) {
        const sharedData = getLockingSharedData(context);
        // Check if behavior failed (safety net for edge cases)
        if (sharedData.failed) {
            return [context.event('if.event.lock_blocked', {
                    messageId: `${context.action.id}.${sharedData.errorMessageId}`,
                    params: { item: sharedData.targetName },
                    targetId: sharedData.targetId,
                    targetName: sharedData.targetName,
                    reason: sharedData.errorMessageId
                })];
        }
        // Emit domain event with messageId (simplified pattern - ADR-097)
        return [
            context.event('if.event.locked', {
                // Rendering data (messageId + params for text-service)
                messageId: `${context.action.id}.${sharedData.messageId}`,
                params: sharedData.params,
                // Domain data (for event sourcing / handlers)
                targetId: sharedData.targetId,
                targetName: sharedData.targetName,
                actorId: context.player.id,
                isContainer: sharedData.isContainer,
                isDoor: sharedData.isDoor,
                keyId: sharedData.keyId,
                keyName: sharedData.keyName,
                sound: sharedData.sound
            })
        ];
    },
    /**
     * Generate events when validation fails
     *
     * Uses simplified event pattern (ADR-097): domain event carries messageId directly.
     */
    blocked(context, result) {
        const noun = context.command.directObject?.entity;
        return [context.event('if.event.lock_blocked', {
                // Rendering data
                messageId: `${context.action.id}.${result.error}`,
                params: result.params || {},
                // Domain data
                targetId: noun?.id,
                targetName: noun?.name,
                reason: result.error
            })];
    },
    metadata: {
        requiresDirectObject: true,
        requiresIndirectObject: false,
        directObjectScope: scope_1.ScopeLevel.REACHABLE
    }
};
