// @sharpee/stdlib/combat/combat-service.js
"use strict";
/**
 * Combat Service (ADR-072)
 *
 * Handles combat resolution using a skill-based probability system.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CombatService = void 0;
exports.createCombatService = createCombatService;
exports.applyCombatResult = applyCombatResult;
exports.findWieldedWeapon = findWieldedWeapon;
const world_model_1 = require("../../world-model/index.js");
const combat_messages_1 = require("./combat-messages");
/**
 * Combat Service implementation
 */
class CombatService {
    /**
     * Calculate hit chance based on skills
     *
     * Base chance starts at 50%.
     * Each point of skill advantage = +1%.
     * Clamped to 10%-95% (always some chance either way).
     */
    calculateHitChance(attackerSkill, defenderSkill, weaponBonus) {
        const baseChance = 50;
        const skillDiff = attackerSkill + weaponBonus - defenderSkill;
        return Math.max(10, Math.min(95, baseChance + skillDiff));
    }
    /**
     * Resolve an attack
     */
    resolveAttack(context) {
        const { attacker, target, weapon, random } = context;
        // Get combatant traits
        const attackerCombat = attacker.get(world_model_1.TraitType.COMBATANT);
        const targetCombat = target.get(world_model_1.TraitType.COMBATANT);
        const weaponTrait = weapon?.get(world_model_1.TraitType.WEAPON);
        // Get skill values
        const attackerSkill = attackerCombat?.skill ?? 30;
        const defenderSkill = targetCombat?.skill ?? 30;
        const weaponBonus = weaponTrait?.skillBonus ?? 0;
        // Calculate hit chance
        const hitChance = this.calculateHitChance(attackerSkill, defenderSkill, weaponBonus);
        // Roll to hit
        const roll = random.int(1, 100);
        const hit = roll <= hitChance;
        // Base message data
        const messageData = {
            attackerName: attacker.name,
            targetName: target.name,
            weaponName: weapon?.name,
        };
        if (!hit) {
            return {
                hit: false,
                damage: 0,
                targetNewHealth: targetCombat?.health ?? 0,
                targetKnockedOut: false,
                targetKilled: false,
                messageId: combat_messages_1.CombatMessages.ATTACK_MISSED,
                messageData,
            };
        }
        // Calculate damage
        const baseDamage = attackerCombat?.baseDamage ?? 1;
        const weaponDamage = weaponTrait?.damage ?? 0;
        let totalDamage = baseDamage + weaponDamage;
        // Blessed weapon bonus against undead (if applicable)
        if (weaponTrait?.isBlessed) {
            // Check if target is undead (story would define this trait)
            const targetIdentity = target.get(world_model_1.TraitType.IDENTITY);
            if (targetIdentity?.isUndead) {
                totalDamage = Math.floor(totalDamage * 1.5);
            }
        }
        // Apply armor reduction
        const armor = targetCombat?.armor ?? 0;
        const effectiveDamage = Math.max(1, totalDamage - armor); // Always do at least 1 damage
        // Calculate new health
        const currentHealth = targetCombat?.health ?? 10;
        const newHealth = Math.max(0, currentHealth - effectiveDamage);
        // Determine outcome
        const killed = newHealth <= 0;
        const maxHealth = targetCombat?.maxHealth ?? 10;
        const knockedOut = !killed && newHealth <= maxHealth * 0.2;
        // Add damage to message data
        messageData.damage = effectiveDamage;
        // Determine message based on outcome
        let messageId;
        if (killed) {
            messageId = combat_messages_1.CombatMessages.ATTACK_KILLED;
        }
        else if (knockedOut) {
            messageId = combat_messages_1.CombatMessages.ATTACK_KNOCKED_OUT;
        }
        else if (effectiveDamage >= maxHealth * 0.3) {
            messageId = combat_messages_1.CombatMessages.ATTACK_HIT_HEAVY;
        }
        else if (effectiveDamage <= maxHealth * 0.1) {
            messageId = combat_messages_1.CombatMessages.ATTACK_HIT_LIGHT;
        }
        else {
            messageId = combat_messages_1.CombatMessages.ATTACK_HIT;
        }
        return {
            hit: true,
            damage: effectiveDamage,
            targetNewHealth: newHealth,
            targetKnockedOut: knockedOut,
            targetKilled: killed,
            messageId,
            messageData,
        };
    }
    /**
     * Check if an entity can attack another
     */
    canAttack(attacker, target) {
        // Check if target has combatant trait
        const targetCombat = target.get(world_model_1.TraitType.COMBATANT);
        if (!targetCombat) {
            return {
                valid: false,
                messageId: combat_messages_1.CombatMessages.CANNOT_ATTACK,
                messageData: { targetName: target.name },
            };
        }
        // Check if target is alive
        if (!targetCombat.isAlive) {
            return {
                valid: false,
                messageId: combat_messages_1.CombatMessages.ALREADY_DEAD,
                messageData: { targetName: target.name },
            };
        }
        // Check if target is unconscious (optional - some games allow attacking unconscious targets)
        // For now, we allow it but could add a flag to control this
        return { valid: true };
    }
    /**
     * Get health status for an entity
     */
    getHealthStatus(entity) {
        const combatant = entity.get(world_model_1.TraitType.COMBATANT);
        if (!combatant) {
            return 'healthy'; // Non-combatants are "healthy"
        }
        if (!combatant.isAlive) {
            return 'dead';
        }
        if (!combatant.isConscious) {
            return 'unconscious';
        }
        const healthPercent = combatant.health / combatant.maxHealth;
        if (healthPercent <= 0.1) {
            return 'near_death';
        }
        else if (healthPercent <= 0.3) {
            return 'badly_wounded';
        }
        else if (healthPercent <= 0.7) {
            return 'wounded';
        }
        else {
            return 'healthy';
        }
    }
}
exports.CombatService = CombatService;
/**
 * Create a new Combat Service instance
 */
function createCombatService() {
    return new CombatService();
}
/**
 * Apply combat result to the target entity
 * Handles health updates, consciousness changes, death, and inventory dropping
 */
function applyCombatResult(target, result, world) {
    const info = { droppedItems: [] };
    if (!result.hit)
        return info;
    const combatant = target.get(world_model_1.TraitType.COMBATANT);
    if (!combatant)
        return info;
    // Update health
    combatant.health = result.targetNewHealth;
    // Update consciousness
    if (result.targetKnockedOut) {
        combatant.knockOut();
    }
    // Update alive status and handle death
    if (result.targetKilled) {
        combatant.kill();
        // Drop inventory if configured
        if (combatant.dropsInventory) {
            const location = world.getLocation(target.id);
            if (location) {
                const inventory = world.getContents(target.id);
                for (const item of inventory) {
                    world.moveEntity(item.id, location);
                    info.droppedItems.push(item.id);
                }
            }
        }
    }
    return info;
}
/**
 * Find the wielded weapon for an entity
 */
function findWieldedWeapon(entity, world) {
    // Get entity's inventory
    const inventory = world.getContents(entity.id);
    // Find weapons
    const weapons = inventory.filter((item) => item.has(world_model_1.TraitType.WEAPON));
    if (weapons.length === 0) {
        return undefined;
    }
    // Prefer equipped weapons
    const equipped = weapons.filter((w) => {
        if (!w.has(world_model_1.TraitType.EQUIPPED))
            return false;
        const equippedTrait = w.get(world_model_1.TraitType.EQUIPPED);
        return equippedTrait?.isEquipped;
    });
    if (equipped.length > 0) {
        // Return weapon with highest damage
        return equipped.reduce((best, current) => {
            const bestWeapon = best.get(world_model_1.TraitType.WEAPON);
            const currentWeapon = current.get(world_model_1.TraitType.WEAPON);
            return currentWeapon.damage > bestWeapon.damage ? current : best;
        });
    }
    // Return first weapon found (or best by damage)
    return weapons.reduce((best, current) => {
        const bestWeapon = best.get(world_model_1.TraitType.WEAPON);
        const currentWeapon = current.get(world_model_1.TraitType.WEAPON);
        return currentWeapon.damage > bestWeapon.damage ? current : best;
    });
}
