// @sharpee/platform-browser/managers/DialogManager.js
"use strict";
/**
 * DialogManager - handles modal dialogs (save, restore, startup)
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DialogManager = void 0;
const types_1 = require("../types");
class DialogManager {
    constructor(config) {
        // Dialog state
        this._isDialogOpen = false;
        this.selectedSaveSlot = null;
        // Pending promise resolvers
        this.pendingSaveResolve = null;
        this.pendingRestoreResolve = null;
        this.pendingStartupResolve = null;
        this.elements = config.elements;
        this.saveManager = config.saveManager;
        this.onDialogOpen = config.onDialogOpen;
        this.onDialogClose = config.onDialogClose;
        this.generateSaveName = config.generateSaveName;
        this.sanitizeSaveName = config.sanitizeSaveName;
        this.performSave = config.performSave;
    }
    /**
     * Check if any dialog is currently open
     */
    isDialogOpen() {
        return this._isDialogOpen;
    }
    /**
     * Show the save dialog
     */
    showSaveDialog() {
        return new Promise((resolve) => {
            console.log('[save-dialog] Opening save dialog...');
            const { modalOverlay, saveDialog, saveNameInput, saveSlotsListEl } = this.elements;
            if (!modalOverlay || !saveDialog) {
                console.error('[save-dialog] Modal elements not found!');
                resolve(false);
                return;
            }
            this.pendingSaveResolve = resolve;
            this._isDialogOpen = true;
            this.selectedSaveSlot = null;
            this.onDialogOpen?.();
            // Populate save slots list
            this.populateSaveSlotsList(saveSlotsListEl, true);
            // Set suggested save name
            if (saveNameInput) {
                saveNameInput.value = this.generateSaveName();
                saveNameInput.select();
            }
            // Show dialog
            modalOverlay.classList.remove('modal-hidden');
            saveDialog.classList.remove('modal-hidden');
            // Focus name input
            saveNameInput?.focus();
        });
    }
    /**
     * Hide the save dialog
     */
    hideSaveDialog(saved) {
        const { modalOverlay, saveDialog } = this.elements;
        modalOverlay?.classList.add('modal-hidden');
        saveDialog?.classList.add('modal-hidden');
        this._isDialogOpen = false;
        this.onDialogClose?.();
        if (this.pendingSaveResolve) {
            this.pendingSaveResolve(saved);
            this.pendingSaveResolve = null;
        }
    }
    /**
     * Show the restore dialog
     */
    showRestoreDialog() {
        return new Promise((resolve) => {
            console.log('[restore-dialog] Opening restore dialog...');
            const { modalOverlay, restoreDialog, restoreSlotsListEl, noSavesMessage } = this.elements;
            if (!modalOverlay || !restoreDialog) {
                console.error('[restore-dialog] Modal elements not found!');
                resolve(null);
                return;
            }
            this.pendingRestoreResolve = resolve;
            this._isDialogOpen = true;
            this.selectedSaveSlot = null;
            this.onDialogOpen?.();
            // Get user saves (excluding autosave)
            const userSaves = this.saveManager.getUserSaves();
            console.log('[restore-dialog] Available saves:', userSaves);
            // Populate restore slots list
            this.populateSaveSlotsList(restoreSlotsListEl, false);
            // Show/hide no saves message
            if (noSavesMessage) {
                if (userSaves.length === 0) {
                    noSavesMessage.classList.remove('modal-hidden');
                    restoreSlotsListEl?.classList.add('modal-hidden');
                }
                else {
                    noSavesMessage.classList.add('modal-hidden');
                    restoreSlotsListEl?.classList.remove('modal-hidden');
                }
            }
            // Restore button disabled until a slot is selected
            const restoreBtn = document.getElementById('restore-confirm-btn');
            if (restoreBtn) {
                restoreBtn.disabled = true;
            }
            // Show dialog
            modalOverlay.classList.remove('modal-hidden');
            restoreDialog.classList.remove('modal-hidden');
            // Focus first save slot or cancel button
            if (userSaves.length > 0) {
                const firstSlot = restoreSlotsListEl?.querySelector('.save-slot');
                firstSlot?.focus();
            }
        });
    }
    /**
     * Hide the restore dialog
     */
    hideRestoreDialog(slotName) {
        const { modalOverlay, restoreDialog } = this.elements;
        modalOverlay?.classList.add('modal-hidden');
        restoreDialog?.classList.add('modal-hidden');
        this._isDialogOpen = false;
        this.onDialogClose?.();
        if (this.pendingRestoreResolve) {
            this.pendingRestoreResolve(slotName);
            this.pendingRestoreResolve = null;
        }
    }
    /**
     * Show the startup dialog (continue saved game or start new?)
     */
    showStartupDialog(meta) {
        return new Promise((resolve) => {
            const { modalOverlay, startupDialog, startupSaveInfo } = this.elements;
            if (!modalOverlay || !startupDialog) {
                // Fall back to starting new game if dialog not available
                resolve(false);
                return;
            }
            this.pendingStartupResolve = resolve;
            this._isDialogOpen = true;
            // Populate save info
            if (startupSaveInfo) {
                const date = new Date(meta.timestamp);
                const dateStr = date.toLocaleDateString();
                const timeStr = date.toLocaleTimeString([], { hour: '2-digit', minute: '2-digit' });
                startupSaveInfo.textContent = `Found "${meta.name}" from ${dateStr} ${timeStr} (Turn ${meta.turnCount}, ${meta.location})`;
            }
            // Show dialog
            modalOverlay.classList.remove('modal-hidden');
            startupDialog.classList.remove('modal-hidden');
            // Focus continue button
            const continueBtn = document.getElementById('startup-continue-btn');
            continueBtn?.focus();
        });
    }
    /**
     * Hide the startup dialog
     */
    hideStartupDialog(shouldContinue) {
        const { modalOverlay, startupDialog } = this.elements;
        modalOverlay?.classList.add('modal-hidden');
        startupDialog?.classList.add('modal-hidden');
        this._isDialogOpen = false;
        if (this.pendingStartupResolve) {
            this.pendingStartupResolve(shouldContinue);
            this.pendingStartupResolve = null;
        }
    }
    /**
     * Populate a save slots list element
     */
    populateSaveSlotsList(listEl, forSave) {
        if (!listEl)
            return;
        // Get saves, filtering out autosave for restore dialog
        const allSaves = this.saveManager.getSaveIndex();
        const saves = forSave ? allSaves : allSaves.filter(s => s.name !== types_1.AUTOSAVE_SLOT);
        listEl.innerHTML = '';
        if (saves.length === 0 && forSave) {
            const emptyMsg = document.createElement('div');
            emptyMsg.className = 'no-saves-message';
            emptyMsg.textContent = 'No existing saves';
            listEl.appendChild(emptyMsg);
            return;
        }
        for (const save of saves) {
            const slot = document.createElement('div');
            slot.className = 'save-slot';
            slot.tabIndex = 0;
            slot.dataset.slotName = save.name;
            const nameSpan = document.createElement('span');
            nameSpan.className = 'save-slot-name';
            nameSpan.textContent = save.name;
            const infoSpan = document.createElement('span');
            infoSpan.className = 'save-slot-info';
            const date = new Date(save.timestamp);
            const dateStr = date.toLocaleDateString();
            const timeStr = date.toLocaleTimeString([], { hour: '2-digit', minute: '2-digit' });
            infoSpan.textContent = `Turn ${save.turnCount} | ${save.location} | ${dateStr} ${timeStr}`;
            slot.appendChild(nameSpan);
            slot.appendChild(infoSpan);
            // Click handler
            slot.addEventListener('click', () => {
                this.selectSaveSlot(listEl, save.name, forSave);
            });
            // Double-click for quick restore
            if (!forSave) {
                slot.addEventListener('dblclick', () => {
                    this.hideRestoreDialog(save.name);
                });
            }
            // Keyboard handler
            slot.addEventListener('keydown', (e) => {
                if (e.key === 'Enter' || e.key === ' ') {
                    e.preventDefault();
                    this.selectSaveSlot(listEl, save.name, forSave);
                    if (!forSave) {
                        // For restore, Enter on selected slot confirms
                        if (this.selectedSaveSlot === save.name) {
                            this.hideRestoreDialog(save.name);
                        }
                    }
                }
                else if (e.key === 'ArrowDown') {
                    e.preventDefault();
                    const next = slot.nextElementSibling;
                    next?.focus();
                }
                else if (e.key === 'ArrowUp') {
                    e.preventDefault();
                    const prev = slot.previousElementSibling;
                    prev?.focus();
                }
            });
            listEl.appendChild(slot);
        }
    }
    /**
     * Select a save slot
     */
    selectSaveSlot(listEl, slotName, forSave) {
        // Remove selection from all slots
        listEl.querySelectorAll('.save-slot').forEach(el => el.classList.remove('selected'));
        // Select clicked slot
        const slot = listEl.querySelector(`[data-slot-name="${slotName}"]`);
        slot?.classList.add('selected');
        this.selectedSaveSlot = slotName;
        // For save dialog, also update the name input
        if (forSave && this.elements.saveNameInput) {
            this.elements.saveNameInput.value = slotName;
        }
        // For restore dialog, enable the restore button
        if (!forSave) {
            const restoreBtn = document.getElementById('restore-confirm-btn');
            if (restoreBtn) {
                restoreBtn.disabled = false;
            }
        }
    }
    /**
     * Set up modal dialog event handlers
     */
    setupHandlers() {
        const { saveNameInput, saveDialog, restoreDialog, startupDialog, modalOverlay } = this.elements;
        // Save dialog buttons
        const saveConfirmBtn = document.getElementById('save-confirm-btn');
        const saveCancelBtn = document.getElementById('save-cancel-btn');
        saveConfirmBtn?.addEventListener('click', () => {
            const name = this.sanitizeSaveName(saveNameInput?.value || this.generateSaveName());
            if (name) {
                this.performSave(name);
                this.hideSaveDialog(true);
            }
        });
        saveCancelBtn?.addEventListener('click', () => {
            this.hideSaveDialog(false);
        });
        // Restore dialog buttons
        const restoreConfirmBtn = document.getElementById('restore-confirm-btn');
        const restoreCancelBtn = document.getElementById('restore-cancel-btn');
        restoreConfirmBtn?.addEventListener('click', () => {
            if (this.selectedSaveSlot) {
                this.hideRestoreDialog(this.selectedSaveSlot);
            }
        });
        restoreCancelBtn?.addEventListener('click', () => {
            this.hideRestoreDialog(null);
        });
        // Startup dialog buttons
        const startupContinueBtn = document.getElementById('startup-continue-btn');
        const startupNewBtn = document.getElementById('startup-new-btn');
        startupContinueBtn?.addEventListener('click', () => {
            this.hideStartupDialog(true);
        });
        startupNewBtn?.addEventListener('click', () => {
            this.hideStartupDialog(false);
        });
        // Enter key in save name input
        saveNameInput?.addEventListener('keydown', (e) => {
            if (e.key === 'Enter') {
                e.preventDefault();
                const name = this.sanitizeSaveName(saveNameInput?.value || this.generateSaveName());
                if (name) {
                    this.performSave(name);
                    this.hideSaveDialog(true);
                }
            }
            else if (e.key === 'Escape') {
                this.hideSaveDialog(false);
            }
        });
        // Escape key closes dialogs
        document.addEventListener('keydown', (e) => {
            if (e.key === 'Escape' && this._isDialogOpen) {
                if (saveDialog && !saveDialog.classList.contains('modal-hidden')) {
                    this.hideSaveDialog(false);
                }
                else if (restoreDialog && !restoreDialog.classList.contains('modal-hidden')) {
                    this.hideRestoreDialog(null);
                }
                else if (startupDialog && !startupDialog.classList.contains('modal-hidden')) {
                    this.hideStartupDialog(false);
                }
            }
        });
        // Click overlay to close
        modalOverlay?.addEventListener('click', (e) => {
            if (e.target === modalOverlay) {
                if (saveDialog && !saveDialog.classList.contains('modal-hidden')) {
                    this.hideSaveDialog(false);
                }
                else if (restoreDialog && !restoreDialog.classList.contains('modal-hidden')) {
                    this.hideRestoreDialog(null);
                }
                else if (startupDialog && !startupDialog.classList.contains('modal-hidden')) {
                    this.hideStartupDialog(false);
                }
            }
        });
    }
}
exports.DialogManager = DialogManager;
