// @sharpee/platform-browser/managers/InputManager.js
"use strict";
/**
 * InputManager - handles command input and history navigation
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.InputManager = void 0;
class InputManager {
    constructor(config) {
        // Command history
        this.commandHistory = [];
        this.historyIndex = -1;
        this.commandInput = config.commandInput;
        this.onCommand = config.onCommand;
        this.isDialogOpen = config.isDialogOpen;
    }
    /**
     * Set up input event handlers
     */
    setupHandlers() {
        if (!this.commandInput)
            return;
        // Handle keyboard input
        this.commandInput.addEventListener('keydown', (e) => {
            if (e.key === 'Enter') {
                this.handleSubmit();
            }
            else if (e.key === 'ArrowUp') {
                e.preventDefault();
                this.navigateHistory(-1);
            }
            else if (e.key === 'ArrowDown') {
                e.preventDefault();
                this.navigateHistory(1);
            }
        });
        // Keep focus on input (but not when dialog is open)
        document.addEventListener('click', (e) => {
            if (this.commandInput && !this.commandInput.disabled && !this.isDialogOpen()) {
                // Don't steal focus from dialog elements
                const target = e.target;
                if (!target.closest('.modal-dialog')) {
                    this.commandInput.focus();
                }
            }
        });
    }
    /**
     * Handle command submission
     */
    async handleSubmit() {
        if (!this.commandInput)
            return;
        const command = this.commandInput.value.trim();
        if (!command)
            return;
        // Add to history
        this.commandHistory.push(command);
        this.historyIndex = this.commandHistory.length;
        // Clear input
        this.commandInput.value = '';
        // Execute command
        await this.onCommand(command);
    }
    /**
     * Navigate command history
     */
    navigateHistory(direction) {
        if (!this.commandInput)
            return;
        const newIndex = this.historyIndex + direction;
        if (newIndex < 0)
            return;
        if (newIndex >= this.commandHistory.length) {
            this.historyIndex = this.commandHistory.length;
            this.commandInput.value = '';
            return;
        }
        this.historyIndex = newIndex;
        this.commandInput.value = this.commandHistory[this.historyIndex];
        // Move cursor to end
        this.commandInput.setSelectionRange(this.commandInput.value.length, this.commandInput.value.length);
    }
    /**
     * Clear the input field
     */
    clearInput() {
        if (this.commandInput) {
            this.commandInput.value = '';
        }
    }
    /**
     * Focus the input field
     */
    focus() {
        this.commandInput?.focus();
    }
    /**
     * Disable the input field
     */
    disable() {
        if (this.commandInput) {
            this.commandInput.disabled = true;
        }
    }
    /**
     * Enable the input field
     */
    enable() {
        if (this.commandInput) {
            this.commandInput.disabled = false;
        }
    }
    /**
     * Set the command prompt text displayed before the input field
     */
    setPrompt(text) {
        if (!this.commandInput)
            return;
        const promptSpan = this.commandInput.parentElement?.querySelector('.prompt');
        if (promptSpan) {
            promptSpan.textContent = text;
        }
    }
    /**
     * Get the command history
     */
    getHistory() {
        return [...this.commandHistory];
    }
}
exports.InputManager = InputManager;
