// @sharpee/platform-browser/managers/SaveManager.js
"use strict";
/**
 * SaveManager - handles save/restore operations using localStorage
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SaveManager = void 0;
const lz_string_1 = require("lz-string");
const types_1 = require("../types");
class SaveManager {
    constructor(config) {
        this.baseline = null;
        this.storagePrefix = config.storagePrefix;
        this.indexKey = `${config.storagePrefix}saves-index`;
        this.savePrefix = `${config.storagePrefix}save-`;
        this.world = config.world;
        this.onStateChange = config.onStateChange;
    }
    /**
     * Capture the initial world state as baseline for delta saves.
     * Call after initializeWorld() completes but before the first turn.
     */
    captureBaseline() {
        this.baseline = this.captureWorldState();
        console.log('[save] Baseline captured:', Object.keys(this.baseline.locations).length, 'entities');
    }
    /**
     * Get list of all saved games from index
     */
    getSaveIndex() {
        try {
            const json = localStorage.getItem(this.indexKey);
            if (!json)
                return [];
            return JSON.parse(json);
        }
        catch {
            return [];
        }
    }
    /**
     * Update save index with new/updated slot
     */
    updateSaveIndex(meta) {
        const index = this.getSaveIndex();
        const existing = index.findIndex(s => s.name === meta.name);
        if (existing >= 0) {
            index[existing] = meta;
        }
        else {
            index.push(meta);
        }
        // Sort by timestamp descending (newest first)
        index.sort((a, b) => b.timestamp - a.timestamp);
        localStorage.setItem(this.indexKey, JSON.stringify(index));
    }
    /**
     * Get current player location name
     */
    getCurrentLocation() {
        try {
            const player = this.world.getPlayer();
            if (player) {
                const locationId = this.world.getLocation(player.id);
                if (locationId) {
                    const room = this.world.getEntity(locationId);
                    if (room) {
                        return room.name || 'Unknown';
                    }
                }
            }
        }
        catch {
            // Ignore errors
        }
        return 'Unknown';
    }
    /**
     * Generate a suggested save name based on current state
     */
    generateSaveName(turnCount) {
        const location = this.getCurrentLocation()
            .toLowerCase()
            .replace(/[^a-z0-9]+/g, '-')
            .replace(/^-|-$/g, '')
            .substring(0, 15);
        return `turn-${turnCount}-${location}`;
    }
    /**
     * Sanitize save name for storage key
     */
    sanitizeSaveName(name) {
        return name
            .toLowerCase()
            .replace(/[^a-z0-9-_]/g, '-')
            .replace(/-+/g, '-')
            .replace(/^-|-$/g, '')
            .substring(0, 30) || 'save';
    }
    /**
     * Capture current world state (locations and traits) without full serialization
     */
    captureWorldState() {
        const locations = {};
        const traits = {};
        for (const entity of this.world.getAllEntities()) {
            // Capture location
            const locationId = this.world.getLocation(entity.id);
            locations[entity.id] = locationId || null;
            // Capture all trait data
            const entityTraits = {};
            for (const [traitName, trait] of entity.traits) {
                // Serialize trait by spreading its properties (excluding methods)
                const traitData = {};
                const traitAsAny = trait;
                for (const key of Object.keys(trait)) {
                    const value = traitAsAny[key];
                    if (typeof value !== 'function') {
                        traitData[key] = value;
                    }
                }
                entityTraits[traitName] = traitData;
            }
            traits[entity.id] = entityTraits;
        }
        return { locations, traits };
    }
    /**
     * Capture only the state that changed from the baseline.
     * Falls back to full capture if no baseline is set.
     */
    captureDelta() {
        const current = this.captureWorldState();
        if (!this.baseline) {
            return current;
        }
        const deltaLocations = {};
        const deltaTraits = {};
        // Diff locations
        for (const [id, loc] of Object.entries(current.locations)) {
            if (loc !== this.baseline.locations[id]) {
                deltaLocations[id] = loc;
            }
        }
        // Diff traits
        for (const [id, entityTraits] of Object.entries(current.traits)) {
            const baseEntityTraits = this.baseline.traits[id] || {};
            for (const [traitName, traitData] of Object.entries(entityTraits)) {
                const baseTrait = baseEntityTraits[traitName];
                if (!baseTrait || !this.shallowEqual(traitData, baseTrait)) {
                    if (!deltaTraits[id])
                        deltaTraits[id] = {};
                    deltaTraits[id][traitName] = traitData;
                }
            }
        }
        return { locations: deltaLocations, traits: deltaTraits };
    }
    /**
     * Shallow equality check for trait data objects.
     */
    shallowEqual(a, b) {
        const keysA = Object.keys(a);
        const keysB = Object.keys(b);
        if (keysA.length !== keysB.length)
            return false;
        for (const key of keysA) {
            if (a[key] !== b[key])
                return false;
        }
        return true;
    }
    /**
     * Restore world state (locations and traits) to existing entities.
     * This preserves entity handlers unlike world.loadJSON().
     */
    restoreWorldState(state) {
        // Restore locations
        for (const [entityId, locationId] of Object.entries(state.locations)) {
            if (locationId) {
                const entity = this.world.getEntity(entityId);
                if (entity) {
                    this.world.moveEntity(entityId, locationId);
                }
            }
        }
        // Restore trait data
        for (const [entityId, entityTraits] of Object.entries(state.traits)) {
            const entity = this.world.getEntity(entityId);
            if (!entity)
                continue;
            for (const [traitName, traitData] of Object.entries(entityTraits)) {
                const trait = entity.get(traitName);
                if (trait) {
                    // Update trait properties
                    const traitAsAny = trait;
                    for (const [key, value] of Object.entries(traitData)) {
                        if (key !== 'type' && typeof value !== 'function') {
                            traitAsAny[key] = value;
                        }
                    }
                }
            }
        }
    }
    /**
     * Perform the actual save to localStorage
     */
    performSave(slotName, context, silent = false) {
        try {
            // Capture only changed state (delta from baseline)
            const { locations, traits } = this.captureDelta();
            // Compress the transcript HTML for storage
            const compressedHtml = (0, lz_string_1.compressToUTF16)(context.transcriptHtml);
            const saveData = {
                version: '3.0.0-delta',
                timestamp: Date.now(),
                turnCount: context.turnCount,
                score: context.score,
                locations,
                traits,
                transcriptHtml: compressedHtml,
            };
            // Compress entire payload to obscure contents and reduce size
            const key = this.savePrefix + slotName;
            const compressed = (0, lz_string_1.compressToUTF16)(JSON.stringify(saveData));
            localStorage.setItem(key, compressed);
            // Update index
            const meta = {
                name: slotName,
                timestamp: saveData.timestamp,
                turnCount: context.turnCount,
                location: this.getCurrentLocation(),
            };
            this.updateSaveIndex(meta);
            console.log('[save] Game saved to', key, meta);
            // Sync to world so stdlib actions know saves exist
            this.syncSavesToWorld();
            this.onStateChange?.();
            return { success: true };
        }
        catch (error) {
            const message = error instanceof Error ? error.message : String(error);
            console.error('[save] Failed:', message);
            return { success: false, error: message };
        }
    }
    /**
     * Perform auto-save (silent, no dialog)
     */
    performAutoSave(context) {
        try {
            const { locations, traits } = this.captureDelta();
            const compressedHtml = (0, lz_string_1.compressToUTF16)(context.transcriptHtml);
            const saveData = {
                version: '3.0.0-delta',
                timestamp: Date.now(),
                turnCount: context.turnCount,
                score: context.score,
                locations,
                traits,
                transcriptHtml: compressedHtml,
            };
            const key = this.savePrefix + types_1.AUTOSAVE_SLOT;
            const compressed = (0, lz_string_1.compressToUTF16)(JSON.stringify(saveData));
            localStorage.setItem(key, compressed);
            const meta = {
                name: types_1.AUTOSAVE_SLOT,
                timestamp: Date.now(),
                turnCount: context.turnCount,
                location: this.getCurrentLocation(),
            };
            this.updateSaveIndex(meta);
            console.log('[autosave] Saved at turn', context.turnCount);
            this.syncSavesToWorld();
        }
        catch (error) {
            console.error('[autosave] Failed:', error);
        }
    }
    /**
     * Load autosave data from localStorage
     */
    loadAutosave() {
        return this.loadSaveSlot(types_1.AUTOSAVE_SLOT);
    }
    /**
     * Load any save slot from localStorage
     */
    loadSaveSlot(slotName) {
        try {
            const key = this.savePrefix + slotName;
            const raw = localStorage.getItem(key);
            if (!raw)
                return null;
            // v3 saves are lz-string compressed; v2 saves are raw JSON
            let json;
            if (raw.startsWith('{')) {
                // Legacy v2 uncompressed JSON
                json = raw;
            }
            else {
                const decompressed = (0, lz_string_1.decompressFromUTF16)(raw);
                if (!decompressed)
                    return null;
                json = decompressed;
            }
            return JSON.parse(json);
        }
        catch (error) {
            console.error('[load] Failed to load slot:', slotName, error);
            return null;
        }
    }
    /**
     * Decompress transcript HTML from save data
     */
    decompressTranscript(compressedHtml) {
        try {
            return (0, lz_string_1.decompressFromUTF16)(compressedHtml) || '';
        }
        catch {
            return '';
        }
    }
    /**
     * Sync localStorage saves to world sharedData so stdlib actions know saves exist
     */
    syncSavesToWorld() {
        const saves = this.getSaveIndex();
        if (saves.length === 0)
            return;
        // Build saves object that the stdlib restoring action expects
        const savesObj = {};
        for (const save of saves) {
            savesObj[save.name] = {
                name: save.name,
                timestamp: save.timestamp,
                moves: save.turnCount,
            };
        }
        // Register sharedData capability if not exists, then update with saves
        if (!this.world.hasCapability('sharedData')) {
            this.world.registerCapability('sharedData', {
                initialData: { saves: savesObj }
            });
        }
        else {
            this.world.updateCapability('sharedData', { saves: savesObj });
        }
        console.log('[sync] Synced', saves.length, 'saves to world sharedData');
    }
    /**
     * Clear autosave (for restart)
     */
    clearAutosave() {
        try {
            const key = this.savePrefix + types_1.AUTOSAVE_SLOT;
            localStorage.removeItem(key);
            // Also update index
            const index = this.getSaveIndex().filter(s => s.name !== types_1.AUTOSAVE_SLOT);
            localStorage.setItem(this.indexKey, JSON.stringify(index));
            console.log('[autosave] Cleared');
        }
        catch {
            // Ignore errors
        }
    }
    /**
     * Check if there's a recent save to continue from
     * Returns the most recent save or null
     */
    checkForSavedGame() {
        const saves = this.getSaveIndex();
        if (saves.length === 0)
            return null;
        // Return most recent save (already sorted by timestamp desc)
        return saves[0];
    }
    /**
     * Get user saves (excluding autosave)
     */
    getUserSaves() {
        return this.getSaveIndex().filter(s => s.name !== types_1.AUTOSAVE_SLOT);
    }
}
exports.SaveManager = SaveManager;
