// @sharpee/platform-browser/managers/ThemeManager.js
"use strict";
/**
 * ThemeManager - handles theme switching and persistence
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ThemeManager = void 0;
class ThemeManager {
    constructor(config) {
        this.storageKey = config.storageKey;
        this.themes = config.themes;
        this.defaultTheme = config.defaultTheme;
    }
    /**
     * Apply saved theme immediately (call before DOM ready).
     * This is a static method to be called as IIFE at module load time
     * to prevent flash of default theme.
     *
     * @example
     * // In browser-entry.ts, call immediately:
     * ThemeManager.applyEarlyTheme('dungeo-theme');
     */
    static applyEarlyTheme(storageKey) {
        try {
            const savedTheme = localStorage.getItem(storageKey);
            if (savedTheme) {
                document.documentElement.setAttribute('data-theme', savedTheme);
            }
        }
        catch {
            // localStorage not available, use default
        }
    }
    /**
     * Get saved theme from localStorage
     */
    getSavedTheme() {
        try {
            return localStorage.getItem(this.storageKey) || this.defaultTheme;
        }
        catch {
            return this.defaultTheme;
        }
    }
    /**
     * Save theme to localStorage
     */
    saveTheme(theme) {
        try {
            localStorage.setItem(this.storageKey, theme);
        }
        catch {
            // localStorage not available, ignore
        }
    }
    /**
     * Apply a theme to the document and update menu checkmarks
     */
    applyTheme(theme) {
        document.documentElement.setAttribute('data-theme', theme);
        this.updateMenuCheckmarks(theme);
        this.saveTheme(theme);
        console.log('[theme] Applied:', theme);
    }
    /**
     * Update theme option checkmarks in the menu
     */
    updateMenuCheckmarks(activeTheme) {
        document.querySelectorAll('.theme-option').forEach(opt => {
            const optTheme = opt.dataset.theme;
            if (optTheme === activeTheme) {
                opt.classList.add('active');
            }
            else {
                opt.classList.remove('active');
            }
        });
    }
    /**
     * Get the available themes
     */
    getThemes() {
        return this.themes;
    }
    /**
     * Get the default theme
     */
    getDefaultTheme() {
        return this.defaultTheme;
    }
}
exports.ThemeManager = ThemeManager;
