// @sharpee/world-model/traits/door/doorBehavior.js
"use strict";
// packages/world-model/src/traits/door/doorBehavior.ts
Object.defineProperty(exports, "__esModule", { value: true });
exports.DoorBehavior = void 0;
const behavior_1 = require("../../behaviors/behavior");
const trait_types_1 = require("../trait-types");
/**
 * Behavior for door entities.
 *
 * Handles the logic for doors that connect two rooms.
 */
class DoorBehavior extends behavior_1.Behavior {
    static requiredTraits = [trait_types_1.TraitType.DOOR];
    /**
     * Get the rooms this door connects
     */
    static getRooms(door) {
        const trait = DoorBehavior.require(door, trait_types_1.TraitType.DOOR);
        return [trait.room1, trait.room2];
    }
    /**
     * Get the other room when coming from a specific room
     * @returns The other room ID, or undefined if the door doesn't connect to the current room
     */
    static getOtherRoom(door, currentRoom) {
        const trait = DoorBehavior.require(door, trait_types_1.TraitType.DOOR);
        if (trait.room1 === currentRoom) {
            return trait.room2;
        }
        else if (trait.room2 === currentRoom) {
            // Check if traversal is allowed in this direction
            return trait.bidirectional ? trait.room1 : undefined;
        }
        return undefined;
    }
    /**
     * Check if the door can be traversed in both directions
     */
    static isBidirectional(door) {
        const trait = DoorBehavior.require(door, trait_types_1.TraitType.DOOR);
        return trait.bidirectional;
    }
    /**
     * Check if the door connects two specific rooms (in any order)
     */
    static connects(door, room1, room2) {
        const trait = DoorBehavior.require(door, trait_types_1.TraitType.DOOR);
        return (trait.room1 === room1 && trait.room2 === room2) ||
            (trait.room1 === room2 && trait.room2 === room1);
    }
    /**
     * Check if the door connects to a specific room
     */
    static connectsTo(door, roomId) {
        const trait = DoorBehavior.require(door, trait_types_1.TraitType.DOOR);
        return trait.room1 === roomId || trait.room2 === roomId;
    }
    /**
     * Get the entry room (for one-way doors)
     * This is the room you can enter from
     */
    static getEntryRoom(door) {
        const trait = DoorBehavior.require(door, trait_types_1.TraitType.DOOR);
        return trait.room1;
    }
    /**
     * Get the exit room (for one-way doors)
     * This is the room you exit to
     */
    static getExitRoom(door) {
        const trait = DoorBehavior.require(door, trait_types_1.TraitType.DOOR);
        return trait.room2;
    }
}
exports.DoorBehavior = DoorBehavior;
