// @sharpee/stdlib/actions/context-adapter.js
"use strict";
/**
 * Context adapter to bridge between stdlib's rich context and if-domain's contracts
 *
 * This adapter converts between:
 * - ValidatedCommand (with parser internals) → CommandInput (simple contract)
 * - ActionContext (stdlib implementation) → IActionContext (if-domain contract)
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ActionContextAdapter = void 0;
exports.toCommandInput = toCommandInput;
exports.createDomainContext = createDomainContext;
/**
 * Convert a ValidatedCommand to the simpler CommandInput
 */
function toCommandInput(validated) {
    const input = {
        actionId: validated.actionId,
        inputText: validated.parsed.rawInput
    };
    // Convert direct object if present
    if (validated.directObject) {
        input.directObject = {
            entity: validated.directObject.entity,
            matchedText: validated.directObject.parsed.text,
            referenceType: 'definite' // TODO: extract from parsed tokens
        };
    }
    // Convert indirect object if present
    if (validated.indirectObject) {
        input.indirectObject = {
            entity: validated.indirectObject.entity,
            matchedText: validated.indirectObject.parsed.text,
            referenceType: 'definite' // TODO: extract from parsed tokens
        };
    }
    // Extract preposition if present
    if (validated.parsed.structure?.preposition) {
        input.preposition = validated.parsed.structure.preposition.text;
    }
    return input;
}
/**
 * Adapter that makes stdlib's ActionContext implement if-domain's IActionContext
 */
class ActionContextAdapter {
    context;
    command;
    action;
    constructor(context, action) {
        this.context = context;
        this.command = toCommandInput(context.command);
        this.action = new ActionAdapter(action);
    }
    get player() {
        return this.context.player;
    }
    get currentLocation() {
        return this.context.currentLocation;
    }
    canSee(entity) {
        return this.context.canSee(entity);
    }
    canReach(entity) {
        return this.context.canReach(entity);
    }
    canTake(entity) {
        return this.context.canTake(entity);
    }
    isInScope(entity) {
        return this.context.isInScope(entity);
    }
    getVisible() {
        return this.context.getVisible();
    }
    getInScope() {
        return this.context.getInScope();
    }
    getEntity(entityId) {
        return this.context.world.getEntity(entityId);
    }
    getEntityLocation(entityId) {
        return this.context.world.getLocation(entityId);
    }
    getEntityContents(entityId) {
        return this.context.world.getContents(entityId);
    }
    getContainingRoom(entityId) {
        return this.context.world.getContainingRoom(entityId);
    }
    getWorldCapability(name) {
        return this.context.world.getCapability(name);
    }
    moveEntity(entityId, newLocationId) {
        // This would need to emit an event, but for now return false
        // as actions shouldn't directly mutate state
        return false;
    }
    event(type, data) {
        return this.context.event(type, data);
    }
}
exports.ActionContextAdapter = ActionContextAdapter;
/**
 * Adapter that makes stdlib's Action implement if-domain's IAction
 */
class ActionAdapter {
    action;
    constructor(action) {
        this.action = action;
    }
    get id() {
        return this.action.id;
    }
    validate(context) {
        // Convert IActionContext back to ActionContext for the stdlib action
        // This is a bit circular but necessary for the adapter pattern
        // In practice, we'd pass through the original context
        return this.action.validate(context);
    }
    execute(context) {
        // Similar conversion for execute
        const result = this.action.execute(context);
        // Handle both old pattern (returns events) and new pattern (returns void)
        if (result === undefined || result === null) {
            // New pattern: execute returned void, call report if available
            if ('report' in this.action && typeof this.action.report === 'function') {
                return this.action.report(context);
            }
            // If no report method, return empty array
            return [];
        }
        // Old pattern: execute returned events
        return result;
    }
    get requiredMessages() {
        return this.action.requiredMessages;
    }
    get descriptionMessageId() {
        return this.action.descriptionMessageId;
    }
    get examplesMessageId() {
        return this.action.examplesMessageId;
    }
}
/**
 * Create an if-domain compliant context from stdlib context
 */
function createDomainContext(context, action) {
    return new ActionContextAdapter(context, action);
}
