// @sharpee/stdlib/actions/context.js
"use strict";
/**
 * Read-only action context implementation
 *
 * Provides query methods for actions without allowing state mutations
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReadOnlyActionContext = void 0;
const world_model_1 = require("../../world-model/index.js");
const types_1 = require("../scope/types");
/**
 * @deprecated Use createActionContext from enhanced-context.ts instead
 * This class is kept for backward compatibility but will be removed in Phase 4
 */
class ReadOnlyActionContext {
    world;
    player;
    currentLocation;
    command;
    action;
    scopeResolver;
    sharedData = {};
    constructor(world, player, currentLocation, command, action, scopeResolver) {
        this.world = world;
        this.player = player;
        this.currentLocation = currentLocation;
        this.command = command;
        // Temporary compatibility - action is required by interface but this class is deprecated
        this.action = action || {
            id: 'unknown',
            validate: () => ({ valid: true }),
            execute: () => []
        };
        this.scopeResolver = scopeResolver || {}; // Temporary fix for deprecated class
    }
    /**
     * Check if an entity is visible to the player
     */
    canSee(entity) {
        return this.world.canSee(this.player.id, entity.id);
    }
    /**
     * Check if an entity is physically reachable by the player
     *
     * An entity is reachable if:
     * 1. It's in the same location as the player
     * 2. It's held by the player
     * 3. It's in an open container in the same location
     * 4. It's on a supporter in the same location
     */
    canReach(entity) {
        const entityLocation = this.world.getLocation(entity.id);
        // Held by player
        if (entityLocation === this.player.id) {
            return true;
        }
        // In same location as player
        if (entityLocation === this.currentLocation.id) {
            return true;
        }
        // In an accessible container/supporter
        if (entityLocation) {
            const container = this.world.getEntity(entityLocation);
            if (container) {
                const containerLocation = this.world.getLocation(container.id);
                // Container is in same location as player
                if (containerLocation === this.currentLocation.id) {
                    // Check if it's an open container
                    if (container.has('container')) {
                        const openableTrait = container.getTrait(world_model_1.OpenableTrait);
                        if (!openableTrait || openableTrait.isOpen) {
                            return true;
                        }
                    }
                    // Or a supporter (always accessible)
                    if (container.has('supporter')) {
                        return true;
                    }
                }
                // Container is held by player
                if (containerLocation === this.player.id) {
                    // Check if it's an open container
                    if (container.has('container')) {
                        const openableTrait = container.getTrait(world_model_1.OpenableTrait);
                        if (!openableTrait || openableTrait.isOpen) {
                            return true;
                        }
                    }
                }
            }
        }
        return false;
    }
    /**
     * Check if an entity can be taken by the player
     *
     * An entity can be taken if:
     * 1. It's visible
     * 2. It's reachable
     * 3. It's not a room
     * 4. It's not scenery (fixed in place)
     * 5. It's not already held by the player
     */
    canTake(entity) {
        // Must be visible
        if (!this.canSee(entity)) {
            return false;
        }
        // Must be reachable
        if (!this.canReach(entity)) {
            return false;
        }
        // Can't take rooms
        if (entity.has('room')) {
            return false;
        }
        // Can't take scenery
        if (entity.has('scenery')) {
            return false;
        }
        // Already held
        const location = this.world.getLocation(entity.id);
        if (location === this.player.id) {
            return false;
        }
        return true;
    }
    /**
     * Check if an entity is in scope for the player
     */
    isInScope(entity) {
        const inScopeEntities = this.world.getInScope(this.player.id);
        return inScopeEntities.some(e => e.id === entity.id);
    }
    /**
     * Get all entities visible to the player
     */
    getVisible() {
        return this.world.getVisible(this.player.id);
    }
    /**
     * Get all entities in scope for the player
     */
    getInScope() {
        return this.world.getInScope(this.player.id);
    }
    /**
     * Get capability data (convenience method)
     */
    getCapability(name) {
        return this.world.getCapability(name);
    }
    /**
     * Event creation method (required by ActionContext interface)
     * @deprecated This implementation throws an error - use createActionContext instead
     */
    event(type, data) {
        throw new Error('ReadOnlyActionContext does not support event creation. Use createActionContext from enhanced-context.ts instead.');
    }
    // =========================================================================
    // Scope validation methods (Phase 4 - stub implementations for deprecated class)
    // =========================================================================
    /**
     * Get the scope level for an entity.
     * @deprecated This implementation returns basic scope levels - use createActionContext instead
     */
    getEntityScope(entity) {
        // Basic implementation for deprecated class
        const entityLocation = this.world.getLocation(entity.id);
        // Carried - in player's inventory
        if (entityLocation === this.player.id) {
            return types_1.ScopeLevel.CARRIED;
        }
        // Reachable - in same location
        if (entityLocation === this.currentLocation.id || this.canReach(entity)) {
            return types_1.ScopeLevel.REACHABLE;
        }
        // Visible - can see but not reach
        if (this.canSee(entity)) {
            return types_1.ScopeLevel.VISIBLE;
        }
        // Not in scope
        return types_1.ScopeLevel.UNAWARE;
    }
    /**
     * Get the scope level for an entity in a command slot.
     * @deprecated This implementation returns basic scope levels - use createActionContext instead
     */
    getSlotScope(slot) {
        const slotMap = {
            target: 'directObject',
            item: 'directObject',
            container: 'indirectObject',
            recipient: 'indirectObject'
        };
        const commandSlot = slotMap[slot] || slot;
        const ref = this.command[commandSlot];
        const entity = ref?.entity;
        if (!entity) {
            return types_1.ScopeLevel.UNAWARE;
        }
        return this.getEntityScope(entity);
    }
    /**
     * Check if an entity meets a required scope level.
     * @deprecated This implementation is basic - use createActionContext instead
     */
    requireScope(entity, required) {
        const actualScope = this.getEntityScope(entity);
        if (actualScope >= required) {
            return { ok: true, actualScope };
        }
        // Return appropriate error based on what's missing
        let errorCode;
        if (required === types_1.ScopeLevel.CARRIED) {
            errorCode = 'scope.not_carried';
        }
        else if (required === types_1.ScopeLevel.REACHABLE) {
            errorCode = 'scope.not_reachable';
        }
        else if (required === types_1.ScopeLevel.VISIBLE) {
            errorCode = 'scope.not_visible';
        }
        else {
            errorCode = 'scope.out_of_scope';
        }
        return {
            ok: false,
            actualScope,
            error: {
                valid: false,
                error: errorCode,
                params: { entityId: entity.id, entityName: entity.name, required, actual: actualScope }
            }
        };
    }
    /**
     * Check if a command slot entity meets a required scope level.
     * @deprecated This implementation is basic - use createActionContext instead
     */
    requireSlotScope(slot, required) {
        const slotMap = {
            target: 'directObject',
            item: 'directObject',
            container: 'indirectObject',
            recipient: 'indirectObject'
        };
        const commandSlot = slotMap[slot] || slot;
        const ref = this.command[commandSlot];
        const entity = ref?.entity;
        if (!entity) {
            return {
                ok: false,
                error: {
                    valid: false,
                    error: 'scope.not_found',
                    params: { slot }
                }
            };
        }
        return this.requireScope(entity, required);
    }
    /**
     * Check if an entity is carried, attempting implicit take if needed.
     * @deprecated This implementation throws an error - use createActionContext instead
     */
    requireCarriedOrImplicitTake(_entity) {
        throw new Error('ReadOnlyActionContext does not support implicit takes. Use createActionContext from enhanced-context.ts instead.');
    }
}
exports.ReadOnlyActionContext = ReadOnlyActionContext;
