// @sharpee/stdlib/actions/data-builder-types.js
"use strict";
/**
 * Action Data Builder Types
 *
 * Provides base types and utilities for building action event data.
 * This addresses the snapshot code smell (ADR-061) by centralizing
 * data structure logic outside of action business logic.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.actionDataRegistry = exports.ActionDataRegistry = void 0;
exports.buildEventData = buildEventData;
exports.buildValidationErrorData = buildValidationErrorData;
/**
 * Build event data using a configuration
 *
 * Combines base builder with optional extensions while protecting core fields
 */
function buildEventData(config, context, preState, postState) {
    // Build base data
    let data = config.builder(context, preState, postState);
    // Apply extension if present
    if (config.extension) {
        const extended = config.extension(data, context, preState, postState);
        // Restore protected fields if specified
        if (config.protectedFields) {
            for (const field of config.protectedFields) {
                if (field in data) {
                    extended[field] = data[field];
                }
            }
        }
        data = extended;
    }
    return data;
}
/**
 * Build validation error data with entity snapshots
 *
 * Centralizes the logic for building error events with entity data
 */
function buildValidationErrorData(context, error, params) {
    const errorData = {
        actionId: context.action.id,
        error: error,
        messageId: error,
        params: params || {}
    };
    // Add entity snapshots if entities are available
    // This will be expanded to use captureEntitySnapshot once we import it
    if (context.command.directObject?.entity) {
        errorData.params.target = context.command.directObject.entity.name;
        errorData.params.targetId = context.command.directObject.entity.id;
    }
    if (context.command.indirectObject?.entity) {
        errorData.params.indirectTarget = context.command.indirectObject.entity.name;
        errorData.params.indirectTargetId = context.command.indirectObject.entity.id;
    }
    return errorData;
}
/**
 * Registry for action data builders
 *
 * Allows registration and retrieval of data builders by action ID
 */
class ActionDataRegistry {
    builders = new Map();
    /**
     * Register a data builder for an action
     */
    register(actionId, config) {
        this.builders.set(actionId, config);
    }
    /**
     * Get a data builder for an action
     */
    get(actionId) {
        return this.builders.get(actionId);
    }
    /**
     * Check if a data builder is registered
     */
    has(actionId) {
        return this.builders.has(actionId);
    }
    /**
     * Build data for an action using its registered builder
     */
    buildData(actionId, context, preState, postState) {
        const config = this.get(actionId);
        if (!config) {
            return undefined;
        }
        return buildEventData(config, context, preState, postState);
    }
}
exports.ActionDataRegistry = ActionDataRegistry;
// Global registry instance
exports.actionDataRegistry = new ActionDataRegistry();
