// @sharpee/stdlib/actions/enhanced-types.js
"use strict";
/**
 * Enhanced action system types
 *
 * This provides a cleaner, more author-friendly action system while
 * maintaining the event-driven architecture. Actions return events,
 * but the enhanced context makes it easy to create those events.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventTypes = exports.MessageNamespaces = exports.ScopeErrors = void 0;
/**
 * Standard error codes for scope failures.
 * These map to message IDs in the language layer.
 */
exports.ScopeErrors = {
    /** Entity is completely unknown to the player */
    NOT_KNOWN: 'scope.not_known',
    /** Entity is known but not currently visible */
    NOT_VISIBLE: 'scope.not_visible',
    /** Entity is visible but not physically reachable */
    NOT_REACHABLE: 'scope.not_reachable',
    /** Entity must be carried but is not in inventory */
    NOT_CARRIED: 'scope.not_carried',
    /** Generic scope failure */
    OUT_OF_SCOPE: 'scope.out_of_scope'
};
/**
 * Standard message namespaces
 */
exports.MessageNamespaces = {
    SYSTEM: 'if.system',
    ACTION: 'if.action',
    ERROR: 'if.error',
    GAME: 'if.game',
    STORY: 'story',
    EXTENSION: 'ext'
};
/**
 * Standard event types
 */
exports.EventTypes = {
    TEXT: 'text',
    ACTION_SUCCESS: 'action.success',
    ACTION_ERROR: 'action.error',
    ACTION_FAILED: 'action.failed',
    GAME_MESSAGE: 'game.message',
    SYSTEM_MESSAGE: 'system.message'
};
