// @sharpee/stdlib/actions/registry.js
"use strict";
/**
 * Action registry implementation
 *
 * Manages registration and lookup of actions.
 * Actions are pure logic - patterns come from the language provider.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.StandardActionRegistry = void 0;
class StandardActionRegistry {
    actions = new Map();
    actionsByPattern = new Map(); // pattern -> actions
    actionsByGroup = new Map(); // group -> actions
    languageProvider = null;
    /**
     * Set the language provider for pattern resolution
     * @param provider Language provider instance (required for pattern resolution)
     */
    setLanguageProvider(provider) {
        this.languageProvider = provider;
        // Rebuild pattern mappings when language provider changes
        this.rebuildPatternMappings();
    }
    /**
     * Register an action
     */
    register(action) {
        this.actions.set(action.id, action);
        // Add to group mapping if specified
        if (action.group) {
            const groupActions = this.actionsByGroup.get(action.group) || [];
            groupActions.push(action);
            this.actionsByGroup.set(action.group, groupActions);
        }
        // Handle direct aliases (backward compatibility)
        const actionRecord = action;
        if ('aliases' in action && Array.isArray(actionRecord.aliases)) {
            const aliases = actionRecord.aliases;
            for (const alias of aliases) {
                const normalizedAlias = alias.toLowerCase();
                // For direct aliases, only keep the latest action (override previous)
                this.actionsByPattern.set(normalizedAlias, [action]);
            }
        }
        // Update pattern mappings if we have a language provider
        if (this.languageProvider) {
            this.updatePatternMappingsForAction(action);
        }
    }
    /**
     * Register multiple actions
     */
    registerMany(actions) {
        for (const action of actions) {
            this.register(action);
        }
    }
    /**
     * Get an action by ID
     */
    get(actionId) {
        return this.actions.get(actionId);
    }
    /**
     * Get all registered actions
     */
    getAll() {
        return Array.from(this.actions.values());
    }
    /**
     * Check if an action is registered
     */
    has(actionId) {
        return this.actions.has(actionId);
    }
    /**
     * Find actions by pattern
     */
    findByPattern(pattern) {
        return this.actionsByPattern.get(pattern.toLowerCase()) || [];
    }
    /**
     * Get actions by group
     */
    getByGroup(group) {
        return this.actionsByGroup.get(group) || [];
    }
    /**
     * Find an action by ID or pattern (backward compatibility)
     * @param idOrPattern Action ID or pattern to search for
     * @returns First matching action or undefined
     */
    find(idOrPattern) {
        // First try direct ID lookup
        const directLookup = this.get(idOrPattern);
        if (directLookup)
            return directLookup;
        // Then try pattern lookup
        const patternMatches = this.findByPattern(idOrPattern);
        return patternMatches[0]; // Return first match
    }
    /**
     * Register messages for an action (placeholder for future implementation)
     */
    registerMessages(actionId, messages) {
        // This would be implemented when we have a message registry
        // For now, messages are handled by the language provider
    }
    /**
     * Update pattern mappings for a single action
     */
    updatePatternMappingsForAction(action) {
        if (!this.languageProvider) {
            // Language provider not yet set - patterns will be mapped when it's provided
            return;
        }
        const patterns = this.languageProvider.getActionPatterns(action.id);
        if (patterns) {
            for (const pattern of patterns) {
                const normalizedPattern = pattern.toLowerCase();
                const actions = this.actionsByPattern.get(normalizedPattern) || [];
                if (!actions.includes(action)) {
                    actions.push(action);
                    // Sort by priority (higher priority first)
                    actions.sort((a, b) => (b.priority || 0) - (a.priority || 0));
                    this.actionsByPattern.set(normalizedPattern, actions);
                }
            }
        }
    }
    /**
     * Rebuild all pattern mappings from language provider
     */
    rebuildPatternMappings() {
        this.actionsByPattern.clear();
        for (const action of this.actions.values()) {
            this.updatePatternMappingsForAction(action);
        }
    }
}
exports.StandardActionRegistry = StandardActionRegistry;
