// @sharpee/world-model/entities/entity-store.js
"use strict";
// packages/world-model/src/entities/entity-store.ts
Object.defineProperty(exports, "__esModule", { value: true });
exports.EntityStore = void 0;
const if_entity_1 = require("./if-entity");
/**
 * Entity store that works with IFEntity instances.
 * Provides trait-aware entity management.
 */
class EntityStore {
    ifEntities;
    constructor() {
        this.ifEntities = new Map();
    }
    /**
     * Add an IF entity to the store
     */
    add(entity) {
        this.ifEntities.set(entity.id, entity);
    }
    /**
     * Get an IF entity by ID
     */
    get(id) {
        return this.ifEntities.get(id);
    }
    /**
     * Check if an entity exists
     */
    has(id) {
        return this.ifEntities.has(id);
    }
    /**
     * Remove an entity from the store
     */
    remove(id) {
        const entity = this.ifEntities.get(id);
        if (entity) {
            entity.clearTraits(); // Cleanup traits before removal
        }
        return this.ifEntities.delete(id);
    }
    /**
     * Get all entities
     */
    getAll() {
        return Array.from(this.ifEntities.values());
    }
    /**
     * Get entities by type
     */
    getByType(type) {
        return this.getAll().filter(entity => entity.type === type);
    }
    /**
     * Find entities with a specific trait
     */
    findWithTrait(traitType) {
        return this.getAll().filter(entity => entity.has(traitType));
    }
    /**
     * Find entities with all specified traits
     */
    findWithAllTraits(...traitTypes) {
        return this.getAll().filter(entity => entity.hasAll(...traitTypes));
    }
    /**
     * Find entities with any of the specified traits
     */
    findWithAnyTraits(...traitTypes) {
        return this.getAll().filter(entity => entity.hasAny(...traitTypes));
    }
    /**
     * Clear all entities from the store
     */
    clear() {
        // Cleanup all entities
        for (const entity of this.ifEntities.values()) {
            entity.clearTraits();
        }
        this.ifEntities.clear();
    }
    /**
     * Get the number of entities in the store
     */
    get size() {
        return this.ifEntities.size;
    }
    /**
     * Iterate over all entities
     */
    [Symbol.iterator]() {
        return this.ifEntities.values();
    }
    /**
     * Serialize all entities to JSON
     */
    toJSON() {
        return this.getAll().map(entity => entity.toJSON());
    }
    /**
     * Load entities from JSON data
     */
    static fromJSON(json) {
        const store = new EntityStore();
        if (Array.isArray(json)) {
            for (const entityData of json) {
                const entity = if_entity_1.IFEntity.fromJSON(entityData);
                store.add(entity);
            }
        }
        return store;
    }
}
exports.EntityStore = EntityStore;
