// @sharpee/world-model/entities/entity-types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EntityType = void 0;
exports.isEntityType = isEntityType;
exports.getEntityTypePrefix = getEntityTypePrefix;
/**
 * Standard Interactive Fiction entity types
 *
 * These represent the fundamental object types in IF games.
 * The values match the type strings used throughout the system.
 */
exports.EntityType = {
    /** A location in the game world */
    ROOM: 'room',
    /** A doorway or portal between rooms */
    DOOR: 'door',
    /** A generic takeable object */
    ITEM: 'item',
    /** A character (NPC or player) */
    ACTOR: 'actor',
    /** An object that can contain other objects */
    CONTAINER: 'container',
    /** An object that can support other objects on top */
    SUPPORTER: 'supporter',
    /** Fixed decorative objects that can't be taken */
    SCENERY: 'scenery',
    /** A directional exit (rarely used as entity) */
    EXIT: 'exit',
    /** A geographic grouping of rooms (ADR-149) */
    REGION: 'region',
    /** A temporal phase of the story (ADR-149) */
    SCENE: 'scene',
    /** Generic object type (default) */
    OBJECT: 'object'
};
/**
 * Type guard to check if a string is a valid IF entity type
 */
function isEntityType(type) {
    return Object.values(exports.EntityType).includes(type);
}
/**
 * Get the ID prefix for an entity type
 * Used for generating consistent entity IDs
 */
function getEntityTypePrefix(type) {
    const prefixMap = {
        [exports.EntityType.ROOM]: 'r',
        [exports.EntityType.DOOR]: 'd',
        [exports.EntityType.ITEM]: 'i',
        [exports.EntityType.ACTOR]: 'a',
        [exports.EntityType.CONTAINER]: 'c',
        [exports.EntityType.SUPPORTER]: 's',
        [exports.EntityType.SCENERY]: 'y',
        [exports.EntityType.EXIT]: 'e',
        [exports.EntityType.REGION]: 'rg',
        [exports.EntityType.SCENE]: 'sc',
        [exports.EntityType.OBJECT]: 'o'
    };
    return prefixMap[type] || 'o';
}
