// @sharpee/stdlib/actions/standard/taking/taking.js
"use strict";
/**
 * Taking action - picks up objects
 *
 * Uses four-phase pattern:
 * 1. validate: Check object can be taken (not scenery, not self, etc.)
 * 2. execute: Transfer the item to actor's inventory
 * 3. report: Generate success events with item and container snapshots
 * 4. blocked: Generate error events when validation fails
 *
 * Supports multi-object commands:
 * - "take all" - takes all portable reachable items
 * - "take all but X" - takes all except specified items
 * - "take X and Y" - takes multiple specified items
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.takingAction = void 0;
const world_model_1 = require("../../../../world-model/index.js");
const constants_1 = require("../../constants");
const types_1 = require("../../../scope/types");
const taking_messages_1 = require("./taking-messages");
// Import type guards and typed interfaces
const taking_types_1 = require("./taking-types");
// Import multi-object helpers
const multi_object_handler_1 = require("../../../helpers/multi-object-handler");
// ============================================================================
// Helper Functions (standalone to avoid `this` issues in object literal)
// ============================================================================
/**
 * Validate taking a single entity
 */
function validateSingleEntity(context, noun) {
    const actor = context.player;
    // Check for interceptor on the target entity (ADR-118)
    // This runs first so entity-specific blocks (e.g., white-hot axe) take priority
    const interceptorResult = (0, world_model_1.getInterceptorForAction)(noun, constants_1.IFActions.TAKING);
    if (interceptorResult) {
        const { interceptor } = interceptorResult;
        const interceptorData = {};
        // Store interceptor in sharedData for later phases
        const sharedData = (0, taking_types_1.getTakingSharedData)(context);
        sharedData._interceptor = interceptor;
        sharedData._interceptorData = interceptorData;
        // === PRE-VALIDATE HOOK ===
        if (interceptor.preValidate) {
            const result = interceptor.preValidate(noun, context.world, actor.id, interceptorData);
            if (result !== null && !result.valid) {
                return {
                    valid: false,
                    error: result.error,
                    params: result.params
                };
            }
        }
    }
    // Check scope - must be able to reach the item
    const scopeCheck = context.requireScope(noun, types_1.ScopeLevel.REACHABLE);
    if (!scopeCheck.ok) {
        return scopeCheck.error;
    }
    // Can't take yourself
    if (noun.id === actor.id) {
        return { valid: false, error: taking_messages_1.TakingMessages.CANT_TAKE_SELF };
    }
    // Check if already holding the item
    const currentLocation = context.world.getLocation(noun.id);
    if (currentLocation === actor.id) {
        return {
            valid: false,
            error: taking_messages_1.TakingMessages.ALREADY_HAVE,
            params: { item: noun.name }
        };
    }
    // Can't take rooms (business rule)
    if (noun.has(world_model_1.TraitType.ROOM)) {
        return {
            valid: false,
            error: taking_messages_1.TakingMessages.CANT_TAKE_ROOM,
            params: { item: noun.name }
        };
    }
    // Can't take scenery (fixed in place)
    if (noun.has(world_model_1.TraitType.SCENERY)) {
        const customMessage = world_model_1.SceneryBehavior.getCantTakeMessage(noun);
        return {
            valid: false,
            error: customMessage || taking_messages_1.TakingMessages.FIXED_IN_PLACE,
            params: { item: noun.name }
        };
    }
    // === POST-VALIDATE HOOK ===
    if (interceptorResult) {
        const interceptor = (0, taking_types_1.getTakingSharedData)(context)._interceptor;
        const interceptorData = (0, taking_types_1.getTakingSharedData)(context)._interceptorData ?? {};
        if (interceptor?.postValidate) {
            const result = interceptor.postValidate(noun, context.world, actor.id, interceptorData);
            if (result !== null && !result.valid) {
                return {
                    valid: false,
                    error: result.error,
                    params: result.params
                };
            }
        }
    }
    // Use ActorBehavior to validate capacity constraints
    if (!world_model_1.ActorBehavior.canTakeItem(actor, noun, context.world)) {
        // Check if it's a container capacity issue
        if (actor.has(world_model_1.TraitType.CONTAINER)) {
            const containerTrait = actor.get(world_model_1.TraitType.CONTAINER);
            if ((0, taking_types_1.hasCapacityLimit)(containerTrait)) {
                // Check item count limit
                if (containerTrait.capacity.maxItems !== undefined) {
                    const contents = context.world.getContents(actor.id);
                    const currentCount = contents.filter((item) => {
                        if (!item.has || !item.has(world_model_1.TraitType.WEARABLE)) {
                            return true;
                        }
                        const wearableTrait = item.get(world_model_1.TraitType.WEARABLE);
                        return !(0, taking_types_1.isWearableTrait)(wearableTrait) || !(wearableTrait.isWorn ?? wearableTrait.worn);
                    }).length;
                    if (currentCount >= containerTrait.capacity.maxItems) {
                        return { valid: false, error: taking_messages_1.TakingMessages.CONTAINER_FULL };
                    }
                }
                // Check weight limit
                if (containerTrait.capacity.maxWeight !== undefined) {
                    const currentWeight = world_model_1.ContainerBehavior.getTotalWeight(actor, context.world);
                    const itemWeight = world_model_1.IdentityBehavior.getWeight(noun);
                    if (currentWeight + itemWeight > containerTrait.capacity.maxWeight) {
                        return { valid: false, error: taking_messages_1.TakingMessages.TOO_HEAVY };
                    }
                }
            }
        }
        // Otherwise generic failure
        return {
            valid: false,
            error: taking_messages_1.TakingMessages.CANNOT_TAKE,
            params: { item: noun.name }
        };
    }
    return { valid: true };
}
/**
 * Validate a multi-object command (take all, take X and Y)
 */
function validateMultiObject(context) {
    const playerId = context.player.id;
    const items = (0, multi_object_handler_1.expandMultiObject)(context, {
        scope: 'reachable',
        // Filter out items already carried by the player
        filter: (entity, world) => world.getLocation(entity.id) !== playerId
    });
    if (items.length === 0) {
        return { valid: false, error: taking_messages_1.TakingMessages.NOTHING_TO_TAKE };
    }
    // Validate each entity, store results
    const results = items.map(item => {
        const validation = validateSingleEntity(context, item.entity);
        return {
            entity: item.entity,
            success: validation.valid,
            error: validation.valid ? undefined : validation.error
        };
    });
    // Store results for execute/report phases
    const sharedData = (0, taking_types_1.getTakingSharedData)(context);
    sharedData.multiObjectResults = results;
    // Valid if at least one can be taken
    const anySuccess = results.some(r => r.success);
    if (!anySuccess) {
        // All failed - return the first error
        return { valid: false, error: results[0].error };
    }
    return { valid: true };
}
/**
 * Execute taking a single entity
 */
function executeSingleEntity(context, noun, result) {
    const actor = context.player;
    // Capture context BEFORE any mutations
    const previousLocation = context.world.getLocation(noun.id);
    result.previousLocation = previousLocation;
    // Check if item is worn and needs to be removed first
    if (noun.has(world_model_1.TraitType.WEARABLE)) {
        const wearableTrait = noun.get(world_model_1.TraitType.WEARABLE);
        if ((0, taking_types_1.isWearableTrait)(wearableTrait) && (wearableTrait.isWorn ?? wearableTrait.worn)) {
            // Mark that we implicitly removed a worn item
            result.implicitlyRemoved = true;
            result.wasWorn = true;
            // Get the wearer (the one who has the item currently)
            const wearer = previousLocation ? context.world.getEntity(previousLocation) : null;
            // Remove the worn status
            if (wearer) {
                world_model_1.WearableBehavior.remove(noun, wearer);
            }
        }
    }
    // Perform the actual move
    context.world.moveEntity(noun.id, actor.id);
    // Score points for taking items with points value (ADR-129)
    const identity = noun.getTrait(world_model_1.IdentityTrait);
    if (identity?.points) {
        context.world.awardScore(noun.id, identity.points, identity.pointsDescription ?? identity.name ?? noun.id);
    }
}
/**
 * Generate success events for taking a single entity
 *
 * Uses simplified event pattern (ADR-097): domain event carries messageId directly.
 * Text-service looks up message from domain event - no separate action.success needed.
 */
function reportSingleSuccess(context, noun, result, events, isMultiObject = false) {
    const actor = context.player;
    // Check if we implicitly removed a worn item
    if (result.implicitlyRemoved) {
        const previousLocation = result.previousLocation;
        const container = previousLocation ? context.world.getEntity(previousLocation) : null;
        events.push(context.event('if.event.removed', {
            implicit: true,
            item: noun.name,
            container: container?.name
        }));
    }
    // Determine if taken from a container/supporter
    const previousLocation = result.previousLocation;
    const isFromContainerOrSupporter = previousLocation &&
        previousLocation !== context.world.getLocation(actor.id);
    // Get container name for message
    const containerEntity = previousLocation ? context.world.getEntity(previousLocation) : null;
    // Determine message ID based on context
    let messageKey;
    if (isMultiObject) {
        messageKey = taking_messages_1.TakingMessages.TAKEN_MULTI;
    }
    else if (isFromContainerOrSupporter) {
        messageKey = taking_messages_1.TakingMessages.TAKEN_FROM;
    }
    else {
        messageKey = taking_messages_1.TakingMessages.TAKEN;
    }
    // Build params for message lookup
    const params = {
        item: noun.name,
        itemId: noun.id,
        actor: actor.name,
        actorId: actor.id,
        previousLocation: result.previousLocation,
        container: containerEntity?.name || ''
    };
    // Emit domain event with messageId (simplified pattern - ADR-097)
    // Text-service will look up message directly from this event
    events.push(context.event('if.event.taken', {
        // Rendering data (messageId + params for text-service)
        messageId: `${context.action.id}.${messageKey}`,
        params,
        // Domain data (for event sourcing / handlers)
        ...params
    }));
}
/**
 * Generate blocked event for a single entity that couldn't be taken
 *
 * Uses simplified event pattern (ADR-097): domain event carries messageId directly.
 */
function reportSingleBlocked(context, noun, error, events) {
    events.push(context.event('if.event.take_blocked', {
        // Rendering data
        messageId: `${context.action.id}.${error}`,
        params: { item: noun.name },
        // Domain data
        item: noun.name,
        itemId: noun.id,
        reason: error
    }));
}
// ============================================================================
// Action Definition
// ============================================================================
exports.takingAction = {
    id: constants_1.IFActions.TAKING,
    // Default scope requirements for this action's slots
    defaultScope: {
        target: types_1.ScopeLevel.REACHABLE
    },
    requiredMessages: [
        'no_target',
        'cant_take_self',
        'already_have',
        'cant_take_room',
        'fixed_in_place',
        'container_full',
        'too_heavy',
        'taken',
        'taken_from',
        'nothing_to_take'
    ],
    metadata: {
        requiresDirectObject: true,
        requiresIndirectObject: false,
        directObjectScope: types_1.ScopeLevel.REACHABLE
    },
    validate(context) {
        // Check for multi-object command
        if ((0, multi_object_handler_1.isMultiObjectCommand)(context)) {
            return validateMultiObject(context);
        }
        // Single object validation
        const noun = context.command.directObject?.entity;
        if (!noun) {
            return { valid: false, error: taking_messages_1.TakingMessages.NO_TARGET };
        }
        return validateSingleEntity(context, noun);
    },
    execute(context) {
        const sharedData = (0, taking_types_1.getTakingSharedData)(context);
        // Check for multi-object command
        if (sharedData.multiObjectResults) {
            // Execute for each successful validation
            for (const result of sharedData.multiObjectResults) {
                if (result.success) {
                    executeSingleEntity(context, result.entity, result);
                }
            }
            return;
        }
        // Single object execution
        const noun = context.command.directObject.entity;
        executeSingleEntity(context, noun, sharedData);
    },
    report(context) {
        const sharedData = (0, taking_types_1.getTakingSharedData)(context);
        const events = [];
        // Check for multi-object command
        if (sharedData.multiObjectResults) {
            // Generate events for each item (success and failure)
            // Use compact format when multiple items
            const isMultiObject = sharedData.multiObjectResults.length > 1;
            for (const result of sharedData.multiObjectResults) {
                if (result.success) {
                    reportSingleSuccess(context, result.entity, result, events, isMultiObject);
                }
                else {
                    reportSingleBlocked(context, result.entity, result.error, events);
                }
            }
            return events;
        }
        // Single object report
        const noun = context.command.directObject.entity;
        reportSingleSuccess(context, noun, sharedData, events, false);
        return events;
    },
    blocked(context, result) {
        // blocked() is called when validation fails
        // Uses simplified event pattern (ADR-097)
        const noun = context.command.directObject?.entity;
        // Fully-qualified message IDs (containing dots) are used as-is;
        // short keys (e.g., 'fixed_in_place') get the action ID prefix
        const error = result.error || '';
        const messageId = error.includes('.') ? error : `${context.action.id}.${error}`;
        return [context.event('if.event.take_blocked', {
                // Rendering data
                messageId,
                params: {
                    ...result.params,
                    item: noun?.name
                },
                // Domain data
                item: noun?.name,
                itemId: noun?.id,
                reason: result.error
            })];
    },
    group: "object_manipulation"
};
