// @sharpee/stdlib/actions/standard/taking/taking-data.js
"use strict";
/**
 * Data builder for taking action
 *
 * Centralizes all entity snapshot logic for the taking action.
 * This separates data structure concerns from business logic.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.takenDataConfig = exports.buildTakenData = void 0;
const snapshot_utils_1 = require("../../base/snapshot-utils");
const taking_types_1 = require("./taking-types");
/**
 * Build taken event data
 *
 * Creates the complete data structure for taken events,
 * including entity snapshots.
 */
const buildTakenData = (context, preState, postState) => {
    const actor = context.player;
    const noun = context.command.directObject?.entity;
    if (!noun) {
        // Shouldn't happen if validation passed
        return {
            item: 'nothing'
        };
    }
    // Get the previous location from typed sharedData
    const sharedData = (0, taking_types_1.getTakingSharedData)(context);
    const previousLocation = sharedData.previousLocation;
    // Capture snapshots after the mutation
    const itemSnapshot = (0, snapshot_utils_1.captureEntitySnapshot)(noun, context.world, true);
    const actorSnapshot = (0, snapshot_utils_1.captureEntitySnapshot)(actor, context.world, false);
    const data = {
        // New atomic structure
        itemSnapshot: itemSnapshot,
        actorSnapshot: actorSnapshot,
        // Backward compatibility
        item: noun.name
    };
    // Add container/supporter info if item was taken from one
    if (previousLocation && previousLocation !== context.world.getLocation(actor.id)) {
        const container = context.world.getEntity(previousLocation);
        if (container) {
            data.fromLocation = previousLocation;
            if (container.has && container.has('container')) {
                data.fromContainer = true;
                data.container = container.name;
            }
            else if (container.has && container.has('supporter')) {
                data.fromSupporter = true;
                data.container = container.name; // Use container field for backward compatibility
                data.supporter = container.name;
            }
            else if (container.has && container.has('actor')) {
                // When taking from another actor (e.g., worn items)
                data.container = container.name;
            }
        }
    }
    return data;
};
exports.buildTakenData = buildTakenData;
/**
 * Configuration for taking data builder
 *
 * Allows stories to extend the data while protecting core fields
 */
exports.takenDataConfig = {
    builder: exports.buildTakenData,
    protectedFields: ['item', 'itemSnapshot', 'actorSnapshot']
};
