// @sharpee/stdlib/actions/standard/taking/taking-types.js
"use strict";
/**
 * Type definitions and type guards for the taking action
 *
 * Provides type-safe access to traits and shared data
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.isWearableTrait = isWearableTrait;
exports.isContainerTrait = isContainerTrait;
exports.hasCapacityLimit = hasCapacityLimit;
exports.isTakingSharedData = isTakingSharedData;
exports.getTakingSharedData = getTakingSharedData;
exports.setTakingSharedData = setTakingSharedData;
/**
 * Type guard for wearable trait data
 * Checks for both 'worn' (internal) and 'isWorn' (accessor) properties
 */
function isWearableTrait(trait) {
    if (!trait || typeof trait !== 'object') {
        return false;
    }
    const t = trait;
    // Check for either worn (internal) or isWorn (public accessor)
    return (('isWorn' in t && typeof t.isWorn === 'boolean') ||
        ('worn' in t && typeof t.worn === 'boolean'));
}
function isContainerTrait(trait) {
    return (trait !== null &&
        typeof trait === 'object' &&
        ('capacity' in trait || 'isTransparent' in trait || 'enterable' in trait));
}
function hasCapacityLimit(trait) {
    return (isContainerTrait(trait) &&
        trait.capacity !== undefined &&
        typeof trait.capacity === 'object');
}
/**
 * Type guard to check if shared data has taking-specific fields
 */
function isTakingSharedData(data) {
    // We don't require all fields, just check it's an object
    // that could contain our fields
    return typeof data === 'object' && data !== null;
}
/**
 * Safely get typed shared data from context
 */
function getTakingSharedData(context) {
    if (!isTakingSharedData(context.sharedData)) {
        // Initialize if needed
        context.sharedData = {};
    }
    return context.sharedData;
}
/**
 * Helper to set taking shared data with type safety
 */
function setTakingSharedData(context, data) {
    Object.assign(context.sharedData, data);
}
