// @sharpee/world-model/traits/character-model/character-vocabulary.js
"use strict";
/**
 * Character model vocabulary types (ADR-141)
 *
 * String literal union types for all word-based authoring inputs.
 * Authors interact with these words; internal numeric values are
 * implementation details managed by CharacterModelTrait.
 *
 * Public interface: All exported types and maps.
 * Owner context: world-model / character-model trait
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CONFIDENCE_VALUES = exports.STABLE_COGNITIVE_PROFILE = exports.THREAT_VALUES = exports.MOOD_AXES = exports.DISPOSITION_RANGES = exports.INTENSITY_VALUES = void 0;
exports.parsePersonalityExpr = parsePersonalityExpr;
exports.dispositionToValue = dispositionToValue;
exports.valueToDisposition = valueToDisposition;
exports.nearestMood = nearestMood;
exports.valueToThreat = valueToThreat;
/** Maps intensity words to internal 0-1 values. */
exports.INTENSITY_VALUES = {
    slightly: 0.2,
    somewhat: 0.4,
    bare: 0.6,
    very: 0.8,
    extremely: 0.95,
};
/**
 * Parse a PersonalityExpr into trait name and numeric value.
 *
 * @param expr - A personality expression like 'very honest' or 'loyal'
 * @returns Tuple of [trait name, intensity value]
 */
function parsePersonalityExpr(expr) {
    const parts = expr.split(' ');
    if (parts.length === 2) {
        const intensity = parts[0];
        const trait = parts[1];
        return [trait, exports.INTENSITY_VALUES[intensity]];
    }
    return [expr, exports.INTENSITY_VALUES.bare];
}
/** Internal numeric ranges for each disposition word. */
exports.DISPOSITION_RANGES = {
    'despises': { min: -100, max: -90, midpoint: -95 },
    'hates': { min: -90, max: -70, midpoint: -80 },
    'dislikes': { min: -70, max: -50, midpoint: -60 },
    'wary of': { min: -40, max: -20, midpoint: -30 },
    'neutral': { min: -10, max: 10, midpoint: 0 },
    'likes': { min: 30, max: 50, midpoint: 40 },
    'trusts': { min: 50, max: 70, midpoint: 60 },
    'devoted to': { min: 80, max: 100, midpoint: 90 },
};
/**
 * Resolve a disposition word to its midpoint numeric value.
 *
 * @param word - A disposition word
 * @returns The midpoint of the disposition range
 */
function dispositionToValue(word) {
    return exports.DISPOSITION_RANGES[word].midpoint;
}
/**
 * Resolve a numeric disposition value back to the nearest word.
 *
 * @param value - A numeric disposition value (-100 to 100)
 * @returns The disposition word whose range contains the value
 */
function valueToDisposition(value) {
    if (value <= -90)
        return 'despises';
    if (value <= -70)
        return 'hates';
    if (value <= -50)
        return 'dislikes';
    if (value <= -20)
        return 'wary of';
    if (value <= 10)
        return 'neutral';
    if (value <= 50)
        return 'likes';
    if (value <= 70)
        return 'trusts';
    return 'devoted to';
}
/**
 * Internal valence-arousal coordinates for each mood.
 * Valence: -1 (negative) to +1 (positive).
 * Arousal: 0 (low energy) to 1 (high energy).
 */
exports.MOOD_AXES = {
    'calm': { valence: 0.3, arousal: 0.1 },
    'content': { valence: 0.5, arousal: 0.2 },
    'cheerful': { valence: 0.8, arousal: 0.5 },
    'nervous': { valence: -0.3, arousal: 0.5 },
    'anxious': { valence: -0.4, arousal: 0.7 },
    'panicked': { valence: -0.7, arousal: 0.95 },
    'angry': { valence: -0.6, arousal: 0.7 },
    'furious': { valence: -0.9, arousal: 0.95 },
    'sad': { valence: -0.5, arousal: 0.2 },
    'grieving': { valence: -0.8, arousal: 0.3 },
    'suspicious': { valence: -0.3, arousal: 0.4 },
    'confused': { valence: -0.2, arousal: 0.3 },
    'resigned': { valence: -0.4, arousal: 0.1 },
};
/**
 * Find the closest mood word to a valence-arousal coordinate.
 *
 * @param valence - Valence value (-1 to 1)
 * @param arousal - Arousal value (0 to 1)
 * @returns The mood word with the smallest Euclidean distance
 */
function nearestMood(valence, arousal) {
    let best = 'calm';
    let bestDist = Infinity;
    for (const [mood, axes] of Object.entries(exports.MOOD_AXES)) {
        const dist = (axes.valence - valence) ** 2 + (axes.arousal - arousal) ** 2;
        if (dist < bestDist) {
            bestDist = dist;
            best = mood;
        }
    }
    return best;
}
/** Maps threat words to internal 0-100 values. */
exports.THREAT_VALUES = {
    'safe': 0,
    'uneasy': 20,
    'wary': 40,
    'threatened': 60,
    'cornered': 80,
    'desperate': 95,
};
/**
 * Resolve a numeric threat value back to the nearest threat level.
 *
 * @param value - A numeric threat value (0-100)
 * @returns The threat level word
 */
function valueToThreat(value) {
    if (value <= 10)
        return 'safe';
    if (value <= 30)
        return 'uneasy';
    if (value <= 50)
        return 'wary';
    if (value <= 70)
        return 'threatened';
    if (value <= 85)
        return 'cornered';
    return 'desperate';
}
/** Default stable cognitive profile. */
exports.STABLE_COGNITIVE_PROFILE = {
    perception: 'accurate',
    beliefFormation: 'flexible',
    coherence: 'focused',
    lucidity: 'stable',
    selfModel: 'intact',
};
/** Maps confidence words to internal 0-1 values. */
exports.CONFIDENCE_VALUES = {
    'uncertain': 0.2,
    'suspects': 0.4,
    'believes': 0.7,
    'certain': 0.95,
};
