// @sharpee/stdlib/actions/standard/index.js
"use strict";
/**
 * Standard Interactive Fiction actions
 *
 * These are the core actions that most IF games will use.
 * Each action is a pure function that validates conditions and returns events.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.standardActions = exports.takingOffAction = exports.wearingAction = exports.throwingAction = exports.removingAction = exports.takingAction = void 0;
// Export actions and their unique event types
var taking_1 = require("./taking");
Object.defineProperty(exports, "takingAction", { enumerable: true, get: function () { return taking_1.takingAction; } });
__exportStar(require("./dropping"), exports);
__exportStar(require("./examining"), exports);
__exportStar(require("./opening"), exports);
__exportStar(require("./closing/closing"), exports);
__exportStar(require("./going"), exports);
__exportStar(require("./looking"), exports);
__exportStar(require("./inventory"), exports);
__exportStar(require("./waiting"), exports);
__exportStar(require("./sleeping"), exports);
__exportStar(require("./scoring"), exports);
__exportStar(require("./help"), exports);
__exportStar(require("./about"), exports);
__exportStar(require("./version"), exports);
__exportStar(require("./locking"), exports);
__exportStar(require("./unlocking"), exports);
__exportStar(require("./switching_on"), exports);
__exportStar(require("./switching_off"), exports);
__exportStar(require("./entering"), exports);
__exportStar(require("./exiting"), exports);
__exportStar(require("./climbing"), exports);
__exportStar(require("./searching"), exports);
__exportStar(require("./listening"), exports);
__exportStar(require("./smelling"), exports);
__exportStar(require("./touching"), exports);
__exportStar(require("./putting"), exports);
__exportStar(require("./inserting"), exports);
__exportStar(require("./reading"), exports);
// Removing exports its own events but reuses TakenEventData
var removing_1 = require("./removing");
Object.defineProperty(exports, "removingAction", { enumerable: true, get: function () { return removing_1.removingAction; } });
__exportStar(require("./giving"), exports);
__exportStar(require("./showing"), exports);
// Throwing action with its unique events
var throwing_1 = require("./throwing");
Object.defineProperty(exports, "throwingAction", { enumerable: true, get: function () { return throwing_1.throwingAction; } });
__exportStar(require("./pushing"), exports);
__exportStar(require("./pulling"), exports);
// export * from './turning'; // Removed - TURNABLE trait doesn't exist
// export * from './using'; // Removed - USE is not idiomatic IF
// Capability-dispatch actions (ADR-090)
__exportStar(require("./lowering"), exports);
__exportStar(require("./raising"), exports);
// Wearing and taking_off share some event types
var wearing_1 = require("./wearing");
Object.defineProperty(exports, "wearingAction", { enumerable: true, get: function () { return wearing_1.wearingAction; } });
var taking_off_1 = require("./taking_off");
Object.defineProperty(exports, "takingOffAction", { enumerable: true, get: function () { return taking_off_1.takingOffAction; } });
__exportStar(require("./eating"), exports);
__exportStar(require("./drinking"), exports);
__exportStar(require("./talking"), exports);
// export * from './asking'; // Moved to conversation extension
// export * from './telling'; // Moved to conversation extension
// export * from './answering'; // Moved to conversation extension
__exportStar(require("./attacking"), exports);
__exportStar(require("./saving"), exports);
__exportStar(require("./restoring"), exports);
__exportStar(require("./quitting"), exports);
__exportStar(require("./restarting"), exports);
__exportStar(require("./undoing"), exports);
__exportStar(require("./again"), exports);
// Concealment actions (ADR-148)
__exportStar(require("./hiding"), exports);
// Import all actions for easy registration
const taking_2 = require("./taking"); // Now from folder
const dropping_1 = require("./dropping"); // Now from folder
const examining_1 = require("./examining"); // Now from folder
const opening_1 = require("./opening"); // Now from folder
const closing_1 = require("./closing/closing"); // Now from folder
const going_1 = require("./going"); // Now from folder
const looking_1 = require("./looking"); // Now from folder
const inventory_1 = require("./inventory"); // Now from folder
const waiting_1 = require("./waiting"); // Now from folder
const sleeping_1 = require("./sleeping"); // Now from folder
const scoring_1 = require("./scoring"); // Now from folder
const help_1 = require("./help"); // Now from folder
const about_1 = require("./about"); // Now from folder
const version_1 = require("./version"); // Now from folder
const locking_1 = require("./locking"); // Now from folder
const unlocking_1 = require("./unlocking"); // Now from folder
const switching_on_1 = require("./switching_on"); // Now from folder
const switching_off_1 = require("./switching_off"); // Now from folder
const entering_1 = require("./entering"); // Now from folder
const exiting_1 = require("./exiting"); // Now from folder
const climbing_1 = require("./climbing"); // Now from folder
const searching_1 = require("./searching"); // Now from folder
const listening_1 = require("./listening");
const smelling_1 = require("./smelling");
const touching_1 = require("./touching");
const putting_1 = require("./putting"); // Now from folder
const inserting_1 = require("./inserting"); // Now from folder
const reading_1 = require("./reading"); // Now from folder
const removing_2 = require("./removing"); // Now from folder
const giving_1 = require("./giving"); // Now from folder
const showing_1 = require("./showing"); // Now from folder
const throwing_2 = require("./throwing"); // Now from folder
const pushing_1 = require("./pushing");
const pulling_1 = require("./pulling");
// import { turningAction } from './turning'; // Removed - TURNABLE trait doesn't exist
// import { usingAction } from './using'; // Removed - USE is not idiomatic IF
// Capability-dispatch actions (ADR-090)
const lowering_1 = require("./lowering");
const raising_1 = require("./raising");
const wearing_2 = require("./wearing");
const taking_off_2 = require("./taking_off");
const eating_1 = require("./eating");
const drinking_1 = require("./drinking");
const talking_1 = require("./talking"); // Now from folder
// import { askingAction } from './asking'; // Moved to conversation extension
// import { tellingAction } from './telling'; // Moved to conversation extension
// import { answeringAction } from './answering'; // Moved to conversation extension
const attacking_1 = require("./attacking"); // Now from folder
const saving_1 = require("./saving"); // Now from folder
const restoring_1 = require("./restoring"); // Now from folder
const quitting_1 = require("./quitting"); // Now from folder
const restarting_1 = require("./restarting"); // Now from folder
const undoing_1 = require("./undoing"); // Now from folder
const again_1 = require("./again"); // Now from folder
// Concealment actions (ADR-148)
const hiding_1 = require("./hiding");
const hiding_2 = require("./hiding");
// Import author/debug actions
const author_1 = require("../author");
const traceAction = new author_1.TraceAction();
exports.standardActions = [
    taking_2.takingAction,
    dropping_1.droppingAction,
    examining_1.examiningAction,
    opening_1.openingAction,
    closing_1.closingAction,
    going_1.goingAction,
    looking_1.lookingAction,
    inventory_1.inventoryAction,
    waiting_1.waitingAction,
    sleeping_1.sleepingAction,
    scoring_1.scoringAction,
    help_1.helpAction,
    about_1.aboutAction,
    version_1.versionAction,
    locking_1.lockingAction,
    unlocking_1.unlockingAction,
    switching_on_1.switchingOnAction,
    switching_off_1.switchingOffAction,
    entering_1.enteringAction,
    exiting_1.exitingAction,
    climbing_1.climbingAction,
    searching_1.searchingAction,
    listening_1.listeningAction,
    smelling_1.smellingAction,
    touching_1.touchingAction,
    putting_1.puttingAction,
    inserting_1.insertingAction,
    reading_1.reading,
    removing_2.removingAction,
    giving_1.givingAction,
    showing_1.showingAction,
    throwing_2.throwingAction,
    pushing_1.pushingAction,
    pulling_1.pullingAction,
    // turningAction, // Removed - TURNABLE trait doesn't exist
    // usingAction, // Removed - USE is not idiomatic IF
    // Capability-dispatch actions (ADR-090)
    lowering_1.loweringAction,
    raising_1.raisingAction,
    wearing_2.wearingAction,
    taking_off_2.takingOffAction,
    eating_1.eatingAction,
    drinking_1.drinkingAction,
    talking_1.talkingAction,
    // askingAction, // Moved to conversation extension
    // tellingAction, // Moved to conversation extension
    // answeringAction, // Moved to conversation extension
    attacking_1.attackingAction,
    saving_1.savingAction,
    restoring_1.restoringAction,
    quitting_1.quittingAction,
    restarting_1.restartingAction,
    undoing_1.undoingAction,
    again_1.againAction,
    // Concealment actions (ADR-148)
    hiding_1.hidingAction,
    hiding_2.revealingAction,
    // Author/debug actions
    traceAction
];
