// @sharpee/stdlib/actions/standard/searching-helpers.js
"use strict";
/**
 * Shared helpers for searching and examining actions
 *
 * These utilities help with content discovery, concealed item handling,
 * and message determination for sensory actions.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.analyzeSearchTarget = analyzeSearchTarget;
exports.revealConcealedItems = revealConcealedItems;
exports.determineSearchMessage = determineSearchMessage;
exports.buildSearchEventData = buildSearchEventData;
const world_model_1 = require("../../../world-model/index.js");
/**
 * Analyze search target and find all contents
 */
function analyzeSearchTarget(context, target) {
    // If no target, search the current location
    const searchTarget = target || context.currentLocation;
    const isLocation = !target;
    // Get all contents
    const contents = context.world.getContents(searchTarget.id);
    // Find concealed items
    const concealedItems = contents.filter(item => {
        if (item.has(world_model_1.TraitType.IDENTITY)) {
            return world_model_1.IdentityBehavior.isConcealed(item);
        }
        return false;
    });
    // Determine target type
    let targetType;
    if (isLocation) {
        targetType = 'location';
    }
    else if (searchTarget.has(world_model_1.TraitType.CONTAINER)) {
        targetType = 'container';
    }
    else if (searchTarget.has(world_model_1.TraitType.SUPPORTER)) {
        targetType = 'supporter';
    }
    else {
        targetType = 'object';
    }
    return {
        target: searchTarget,
        contents,
        concealedItems,
        isLocation,
        targetType
    };
}
/**
 * Reveal concealed items found during search
 */
function revealConcealedItems(items) {
    items.forEach(item => {
        if (item.has(world_model_1.TraitType.IDENTITY)) {
            world_model_1.IdentityBehavior.reveal(item);
        }
    });
}
/**
 * Determine the appropriate search message based on what was found
 */
function determineSearchMessage(searchContext) {
    const { target, contents, concealedItems, targetType } = searchContext;
    const params = {
        target: target.name
    };
    // Found concealed items - most important result
    if (concealedItems.length > 0) {
        params.items = concealedItems.map(item => item.name).join(', ');
        params.where = getLocationPreposition(targetType);
        return { messageId: 'found_concealed', params };
    }
    // Target-specific messages when nothing concealed
    switch (targetType) {
        case 'container':
            if (contents.length === 0) {
                return { messageId: 'empty_container', params };
            }
            else {
                params.items = contents.map(item => item.name).join(', ');
                return { messageId: 'container_contents', params };
            }
        case 'supporter':
            if (contents.length > 0) {
                params.items = contents.map(item => item.name).join(', ');
                return { messageId: 'supporter_contents', params };
            }
            else {
                return { messageId: 'nothing_special', params };
            }
        case 'location':
            return { messageId: 'searched_location', params };
        case 'object':
        default:
            if (contents.length > 0) {
                return { messageId: 'searched_object', params };
            }
            else {
                return { messageId: 'nothing_special', params };
            }
    }
}
/**
 * Get the appropriate preposition for where items were found
 */
function getLocationPreposition(targetType) {
    switch (targetType) {
        case 'container':
            return 'inside';
        case 'supporter':
            return 'on';
        case 'location':
        case 'object':
        default:
            return 'here';
    }
}
/**
 * Build search event data
 */
function buildSearchEventData(searchContext) {
    return {
        target: searchContext.target.id,
        targetName: searchContext.target.name,
        foundItems: searchContext.concealedItems.map(item => item.id),
        foundItemNames: searchContext.concealedItems.map(item => item.name),
        searchingLocation: searchContext.isLocation
    };
}
