// @sharpee/stdlib/actions/standard/wearable-shared.js
"use strict";
/**
 * Shared helpers for wearing and taking_off actions
 *
 * Centralizes common logic for wearable item manipulation including:
 * - Layering conflict detection
 * - Event data building
 * - Wearable validation helpers
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.analyzeWearableContext = analyzeWearableContext;
exports.checkWearingConflicts = checkWearingConflicts;
exports.checkRemovalBlockers = checkRemovalBlockers;
exports.buildWearableEventParams = buildWearableEventParams;
exports.createWearableBlockedEvent = createWearableBlockedEvent;
exports.createWearableSuccessEvent = createWearableSuccessEvent;
exports.isWornByActor = isWornByActor;
exports.hasRemovalRestrictions = hasRemovalRestrictions;
const world_model_1 = require("../../../world-model/index.js");
/**
 * Analyzes the wearable context for an action
 */
function analyzeWearableContext(context, item, actor) {
    const wearableTrait = item.get(world_model_1.TraitType.WEARABLE);
    const inventory = context.world.getContents(actor.id);
    return {
        item,
        actor,
        wearableTrait,
        inventory
    };
}
/**
 * Checks for layering conflicts when wearing an item
 * Returns conflicting item if found, null otherwise
 */
function checkWearingConflicts(wearableContext) {
    const { item, wearableTrait, inventory } = wearableContext;
    // Check for body part conflicts (non-layered items)
    if (wearableTrait.bodyPart && wearableTrait.layer === undefined) {
        const conflictingItem = inventory.find(invItem => {
            if (invItem.id === item.id)
                return false;
            if (!invItem.has(world_model_1.TraitType.WEARABLE))
                return false;
            const otherWearable = invItem.get(world_model_1.TraitType.WEARABLE);
            return otherWearable.worn &&
                otherWearable.bodyPart === wearableTrait.bodyPart &&
                otherWearable.layer === undefined;
        });
        if (conflictingItem) {
            return conflictingItem;
        }
    }
    // Check for layering rule violations
    if (wearableTrait.layer !== undefined && wearableTrait.bodyPart) {
        const wornItems = inventory.filter(invItem => {
            if (!invItem.has(world_model_1.TraitType.WEARABLE))
                return false;
            const otherWearable = invItem.get(world_model_1.TraitType.WEARABLE);
            return otherWearable.worn;
        });
        for (const wornItem of wornItems) {
            const otherWearable = wornItem.get(world_model_1.TraitType.WEARABLE);
            if (otherWearable.layer !== undefined &&
                otherWearable.layer > wearableTrait.layer &&
                otherWearable.bodyPart === wearableTrait.bodyPart) {
                return wornItem; // Can't wear under something already worn
            }
        }
    }
    return null;
}
/**
 * Checks for items blocking removal due to layering
 * Returns blocking item if found, null otherwise
 */
function checkRemovalBlockers(wearableContext) {
    const { item, wearableTrait, inventory } = wearableContext;
    if (wearableTrait.layer !== undefined && wearableTrait.bodyPart) {
        const blockingItem = inventory.find(invItem => {
            if (invItem.id === item.id)
                return false;
            if (!invItem.has(world_model_1.TraitType.WEARABLE))
                return false;
            const otherWearable = invItem.get(world_model_1.TraitType.WEARABLE);
            return otherWearable.worn &&
                otherWearable.layer !== undefined &&
                otherWearable.layer > wearableTrait.layer &&
                otherWearable.bodyPart === wearableTrait.bodyPart;
        });
        return blockingItem || null;
    }
    return null;
}
/**
 * Builds event parameters including body part and layer info
 */
function buildWearableEventParams(item, wearableTrait) {
    const params = {
        item: item.name
    };
    if (wearableTrait.bodyPart) {
        params.bodyPart = wearableTrait.bodyPart;
    }
    if (wearableTrait.layer !== undefined) {
        params.layer = wearableTrait.layer;
    }
    return params;
}
/**
 * Creates a blocked event for wearable actions
 *
 * Uses domain event pattern (ADR-097): event carries messageId directly.
 */
function createWearableBlockedEvent(context, messageId, reason, params) {
    const eventType = context.action.id === 'if.action.wearing'
        ? 'if.event.wear_blocked'
        : 'if.event.take_off_blocked';
    return context.event(eventType, {
        messageId: `${context.action.id}.${messageId}`,
        params: params || {},
        reason
    });
}
/**
 * Creates a success event for wearable actions
 *
 * Uses domain event pattern (ADR-097): event carries messageId directly.
 */
function createWearableSuccessEvent(context, messageId, params) {
    const eventType = context.action.id === 'if.action.wearing'
        ? 'if.event.worn'
        : 'if.event.removed';
    return context.event(eventType, {
        messageId: `${context.action.id}.${messageId}`,
        params
    });
}
/**
 * Checks if an item is currently worn by the actor
 */
function isWornByActor(item, actor) {
    if (!item.has(world_model_1.TraitType.WEARABLE))
        return false;
    const wearable = item.get(world_model_1.TraitType.WEARABLE);
    return wearable.worn && wearable.wornBy === actor.id;
}
/**
 * Checks if an item has special removal restrictions
 */
function hasRemovalRestrictions(wearableTrait) {
    // Check for cursed or otherwise unremovable items
    return !!wearableTrait.cursed;
}
