// @sharpee/ext-testing/index.js
"use strict";
/**
 * @sharpee/ext-testing
 *
 * Debug and testing tools extension for Sharpee IF engine.
 *
 * Provides:
 * - Interactive debug mode (GDT-style) with short codes
 * - Test commands ($teleport, $take, $assert, etc.) for transcripts
 * - Checkpoint save/restore system
 * - Playtester annotations (planned)
 *
 * @example
 * ```typescript
 * import { TestingExtension } from "./index.js";
 *
 * const testing = new TestingExtension({
 *   debugMode: { enabled: true, prefix: 'gdt' },
 *   testMode: { enabled: true, deterministicRandom: true },
 *   checkpoints: { directory: './saves' }
 * });
 *
 * // Execute GDT command
 * const result = testing.executeGdtCommand('AH west-of-house', world);
 *
 * // Execute test command
 * const result = testing.executeTestCommand('$teleport west-of-house', world);
 *
 * // Save/restore checkpoints
 * await testing.saveCheckpoint('before-troll', world);
 * await testing.restoreCheckpoint('before-troll', world);
 * ```
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.createEmptyContext = exports.captureContext = exports.createAnnotationStore = exports.createLocalStorageStore = exports.createMemoryStore = exports.createFileStore = exports.validateCheckpoint = exports.deserializeCheckpoint = exports.serializeCheckpoint = exports.parseTestInput = exports.parseGdtInput = exports.createCommandRegistry = exports.formatLocationChain = exports.formatEntity = exports.createDebugContext = exports.TestingExtension = void 0;
// Main extension class
var extension_js_1 = require("./extension.js");
Object.defineProperty(exports, "TestingExtension", { enumerable: true, get: function () { return extension_js_1.TestingExtension; } });
// Context utilities
var debug_context_js_1 = require("./context/debug-context.js");
Object.defineProperty(exports, "createDebugContext", { enumerable: true, get: function () { return debug_context_js_1.createDebugContext; } });
Object.defineProperty(exports, "formatEntity", { enumerable: true, get: function () { return debug_context_js_1.formatEntity; } });
Object.defineProperty(exports, "formatLocationChain", { enumerable: true, get: function () { return debug_context_js_1.formatLocationChain; } });
// Command registry utilities
var registry_js_1 = require("./commands/registry.js");
Object.defineProperty(exports, "createCommandRegistry", { enumerable: true, get: function () { return registry_js_1.createCommandRegistry; } });
Object.defineProperty(exports, "parseGdtInput", { enumerable: true, get: function () { return registry_js_1.parseGdtInput; } });
Object.defineProperty(exports, "parseTestInput", { enumerable: true, get: function () { return registry_js_1.parseTestInput; } });
// Checkpoint utilities
var serializer_js_1 = require("./checkpoints/serializer.js");
Object.defineProperty(exports, "serializeCheckpoint", { enumerable: true, get: function () { return serializer_js_1.serializeCheckpoint; } });
Object.defineProperty(exports, "deserializeCheckpoint", { enumerable: true, get: function () { return serializer_js_1.deserializeCheckpoint; } });
Object.defineProperty(exports, "validateCheckpoint", { enumerable: true, get: function () { return serializer_js_1.validateCheckpoint; } });
var store_js_1 = require("./checkpoints/store.js");
Object.defineProperty(exports, "createFileStore", { enumerable: true, get: function () { return store_js_1.createFileStore; } });
Object.defineProperty(exports, "createMemoryStore", { enumerable: true, get: function () { return store_js_1.createMemoryStore; } });
Object.defineProperty(exports, "createLocalStorageStore", { enumerable: true, get: function () { return store_js_1.createLocalStorageStore; } });
// Annotation utilities (ADR-109)
var index_js_1 = require("./annotations/index.js");
Object.defineProperty(exports, "createAnnotationStore", { enumerable: true, get: function () { return index_js_1.createAnnotationStore; } });
Object.defineProperty(exports, "captureContext", { enumerable: true, get: function () { return index_js_1.captureContext; } });
Object.defineProperty(exports, "createEmptyContext", { enumerable: true, get: function () { return index_js_1.createEmptyContext; } });
