// @sharpee/lang-en-us/index.js
"use strict";
/**
 * English (US) Language Package for Sharpee IF Platform
 *
 * Provides English language support including:
 * - Vocabulary (verbs, nouns, adjectives, etc.)
 * - Grammar patterns
 * - Text formatting and lemmatization
 * - Message templates
 * - Action patterns and messages
 * - NPC messages (ADR-070)
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getParserErrorMessage = exports.parserErrors = exports.irregularPlurals = exports.abbreviations = exports.englishWords = exports.directionMap = exports.ordinalNumbers = exports.cardinalNumbers = exports.LanguageProvider = exports.default = exports.EnglishLanguageProvider = void 0;
var language_provider_1 = require("./language-provider");
Object.defineProperty(exports, "EnglishLanguageProvider", { enumerable: true, get: function () { return language_provider_1.EnglishLanguageProvider; } });
Object.defineProperty(exports, "default", { enumerable: true, get: function () { return __importDefault(language_provider_1).default; } });
var language_provider_2 = require("./language-provider"); // For convenience
Object.defineProperty(exports, "LanguageProvider", { enumerable: true, get: function () { return language_provider_2.EnglishLanguageProvider; } });
// Export grammar types and constants
__exportStar(require("./grammar"), exports);
// Export action language definitions
__exportStar(require("./actions"), exports);
// Export NPC language definitions (ADR-070)
__exportStar(require("./npc"), exports);
// Export perspective placeholder resolution (ADR-089)
__exportStar(require("./perspective"), exports);
// Export formatter system (ADR-095)
__exportStar(require("./formatters"), exports);
// Export vocabulary data for parser use (ADR-082)
var words_1 = require("./data/words");
Object.defineProperty(exports, "cardinalNumbers", { enumerable: true, get: function () { return words_1.cardinalNumbers; } });
Object.defineProperty(exports, "ordinalNumbers", { enumerable: true, get: function () { return words_1.ordinalNumbers; } });
Object.defineProperty(exports, "directionMap", { enumerable: true, get: function () { return words_1.directionMap; } });
Object.defineProperty(exports, "englishWords", { enumerable: true, get: function () { return words_1.englishWords; } });
Object.defineProperty(exports, "abbreviations", { enumerable: true, get: function () { return words_1.abbreviations; } });
Object.defineProperty(exports, "irregularPlurals", { enumerable: true, get: function () { return words_1.irregularPlurals; } });
// Export parser error messages (Phase 1.2 recommendations)
var messages_1 = require("./data/messages");
Object.defineProperty(exports, "parserErrors", { enumerable: true, get: function () { return messages_1.parserErrors; } });
Object.defineProperty(exports, "getParserErrorMessage", { enumerable: true, get: function () { return messages_1.getParserErrorMessage; } });
