// @sharpee/event-processor/effects/world-query.js
"use strict";
/**
 * WorldQuery: Read-only view of WorldModel for event handlers
 *
 * Per ADR-075, handlers receive a read-only query interface instead of
 * full world access. This preserves stdlib as the gatekeeper for mutations.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.createWorldQuery = createWorldQuery;
/**
 * Create a WorldQuery wrapper around a WorldModel
 *
 * This provides read-only access to the world model for event handlers.
 */
function createWorldQuery(world) {
    return {
        getEntity: (id) => world.getEntity(id),
        hasEntity: (id) => world.hasEntity(id),
        getPlayer: () => world.getPlayer(),
        getCurrentRoom: () => {
            const player = world.getPlayer();
            return player ? world.getContainingRoom(player.id) : undefined;
        },
        getContainingRoom: (entityId) => world.getContainingRoom(entityId),
        getLocation: (entityId) => world.getLocation(entityId),
        getContents: (containerId) => world.getContents(containerId),
        findByTrait: (traitType) => world.findByTrait(traitType),
        findWhere: (predicate) => world.findWhere(predicate),
        getStateValue: (key) => world.getStateValue(key),
        getCapability: (name) => world.getCapability(name),
        hasCapability: (name) => world.hasCapability(name),
    };
}
