// @sharpee/core/types/result.js
"use strict";
// packages/core/src/types/result.ts
Object.defineProperty(exports, "__esModule", { value: true });
exports.Result = void 0;
/**
 * Convenience functions for working with Result types
 */
exports.Result = {
    /**
     * Create a successful result
     */
    ok(value) {
        return { success: true, value };
    },
    /**
     * Create a failed result
     */
    fail(error) {
        return { success: false, error };
    },
    /**
     * Check if a result is successful
     */
    isOk(result) {
        return result.success;
    },
    /**
     * Check if a result is a failure
     */
    isFail(result) {
        return !result.success;
    },
    /**
     * Map a successful result to a new value
     */
    map(result, fn) {
        if (result.success) {
            return exports.Result.ok(fn(result.value));
        }
        return result;
    },
    /**
     * Map a failed result to a new error
     */
    mapError(result, fn) {
        if (!result.success) {
            return exports.Result.fail(fn(result.error));
        }
        return result;
    },
    /**
     * Chain results together
     */
    flatMap(result, fn) {
        if (result.success) {
            return fn(result.value);
        }
        return result;
    },
    /**
     * Get the value or throw an error
     */
    unwrap(result) {
        if (result.success) {
            return result.value;
        }
        throw result.error;
    },
    /**
     * Get the value or return a default
     */
    unwrapOr(result, defaultValue) {
        if (result.success) {
            return result.value;
        }
        return defaultValue;
    }
};
