// @sharpee/stdlib/actions/standard/going/going-data.js
"use strict";
/**
 * Data builder for going action
 *
 * Centralizes all room and movement snapshot logic for the going action.
 * This separates data structure concerns from business logic.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.actorEnteredDataConfig = exports.actorExitedDataConfig = exports.actorMovedDataConfig = exports.buildActorEnteredData = exports.buildActorExitedData = exports.buildActorMovedData = void 0;
exports.determineGoingMessage = determineGoingMessage;
const world_model_1 = require("../../../../world-model/index.js");
const snapshot_utils_1 = require("../../base/snapshot-utils");
const going_1 = require("./going");
/**
 * Find the source room (where we came from) and the exit configuration used
 * Since we've already moved, we need to find which room has an exit to our current location
 */
function findSourceRoomAndExit(currentRoom, direction, world) {
    const allEntities = world.getAllEntities();
    const allRooms = allEntities.filter(e => e.has(world_model_1.TraitType.ROOM));
    // Find the room that has an exit leading to our current location
    for (const room of allRooms) {
        const exitConfig = world_model_1.RoomBehavior.getExit(room, direction);
        if (exitConfig && exitConfig.destination === currentRoom.id) {
            return { room, mapHint: exitConfig.mapHint };
        }
    }
    // If we can't find source room, use the current room as a fallback
    return { room: currentRoom };
}
/**
 * Build actor moved event data
 */
const buildActorMovedData = (context, preState, postState) => {
    const actor = context.player;
    // Get direction from context (should already be a Direction constant)
    // Direction can come from extras or from directObject name
    let direction = context.command.parsed.extras?.direction;
    // If no direction in extras, check if directObject has a name that could be a direction
    if (!direction && context.command.directObject?.entity) {
        const entityName = context.command.directObject.entity.name ||
            context.command.directObject.entity.attributes?.name;
        if (entityName) {
            direction = entityName;
        }
    }
    const oppositeDir = (0, world_model_1.getOppositeDirection)(direction);
    // Get actual current location after movement (not the cached one from context)
    const currentLocationId = context.world.getLocation(actor.id);
    const currentRoom = context.world.getEntity(currentLocationId);
    // Find source room and exit configuration (includes mapHint)
    const { room: sourceRoom, mapHint } = findSourceRoomAndExit(currentRoom, direction, context.world);
    // Capture room snapshots for atomic events
    const sourceSnapshot = (0, snapshot_utils_1.captureRoomSnapshot)(sourceRoom, context.world, false);
    const destinationSnapshot = (0, snapshot_utils_1.captureRoomSnapshot)(currentRoom, context.world, false);
    const actorSnapshot = (0, snapshot_utils_1.captureEntitySnapshot)(actor, context.world, true); // Include inventory for darkness checks
    // Check if this was the first visit (stored during execute phase via sharedData)
    const sharedData = (0, going_1.getGoingSharedData)(context);
    const firstVisit = sharedData.isFirstVisit === true;
    const result = {
        // New atomic structure
        actor: actorSnapshot,
        sourceRoom: sourceSnapshot,
        destinationRoom: destinationSnapshot,
        // Direction as constant
        direction: direction,
        fromRoom: sourceRoom.id,
        toRoom: currentRoom.id,
        oppositeDirection: oppositeDir,
        firstVisit: firstVisit
    };
    // Include mapHint if present on the exit (ADR-113)
    if (mapHint) {
        result.mapHint = mapHint;
    }
    return result;
};
exports.buildActorMovedData = buildActorMovedData;
/**
 * Build actor exited event data
 */
const buildActorExitedData = (context, preState, postState) => {
    const actor = context.player;
    // Get direction from context (should already be a Direction constant)
    // Direction can come from extras or from directObject name
    let direction = context.command.parsed.extras?.direction;
    // If no direction in extras, check if directObject has a name that could be a direction
    if (!direction && context.command.directObject?.entity) {
        const entityName = context.command.directObject.entity.name ||
            context.command.directObject.entity.attributes?.name;
        if (entityName) {
            direction = entityName;
        }
    }
    // Get actual current location after movement (not the cached one from context)
    const currentLocationId = context.world.getLocation(actor.id);
    const currentRoom = context.world.getEntity(currentLocationId);
    return {
        actorId: actor.id,
        direction: direction,
        toRoom: currentRoom.id
    };
};
exports.buildActorExitedData = buildActorExitedData;
/**
 * Build actor entered event data
 */
const buildActorEnteredData = (context, preState, postState) => {
    const actor = context.player;
    // Get direction from context (should already be a Direction constant)
    // Direction can come from extras or from directObject name
    let direction = context.command.parsed.extras?.direction;
    // If no direction in extras, check if directObject has a name that could be a direction
    if (!direction && context.command.directObject?.entity) {
        const entityName = context.command.directObject.entity.name ||
            context.command.directObject.entity.attributes?.name;
        if (entityName) {
            direction = entityName;
        }
    }
    const oppositeDir = (0, world_model_1.getOppositeDirection)(direction);
    // Get actual current location after movement (not the cached one from context)
    const currentLocationId = context.world.getLocation(actor.id);
    const currentRoom = context.world.getEntity(currentLocationId);
    // Find source room
    const { room: sourceRoom } = findSourceRoomAndExit(currentRoom, direction, context.world);
    return {
        actorId: actor.id,
        direction: oppositeDir,
        fromRoom: sourceRoom.id
    };
};
exports.buildActorEnteredData = buildActorEnteredData;
/**
 * Determine the success message for going action
 */
function determineGoingMessage(movedData) {
    const messageId = movedData.firstVisit ? 'first_visit' : 'moved_to';
    const destinationRoom = movedData.destinationRoom;
    const destinationName = destinationRoom?.name ||
        destinationRoom?.attributes?.name ||
        destinationRoom?.id ||
        'unknown';
    const rawDirection = movedData.direction;
    const direction = rawDirection.toLowerCase();
    return {
        messageId,
        params: {
            direction,
            destination: destinationName
        }
    };
}
/**
 * Configuration for going data builders
 */
exports.actorMovedDataConfig = {
    builder: exports.buildActorMovedData,
    protectedFields: ['direction', 'fromRoom', 'toRoom']
};
exports.actorExitedDataConfig = {
    builder: exports.buildActorExitedData,
    protectedFields: ['actorId', 'direction']
};
exports.actorEnteredDataConfig = {
    builder: exports.buildActorEnteredData,
    protectedFields: ['actorId', 'direction']
};
