// @sharpee/if-domain/parser-contracts/parser-factory.js
"use strict";
/**
 * Parser Factory
 *
 * Creates language-specific parser instances.
 * The actual parser implementations are in separate packages
 * and must be registered before use.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ParserFactory = void 0;
const parserRegistry = new Map();
/**
 * Parser factory for creating language-specific parsers
 */
class ParserFactory {
    /**
     * Register a parser implementation for a language
     *
     * @example
     * import { EnglishParser } from "../../parser-en-us/index.js";
     * ParserFactory.registerParser('en-US', EnglishParser);
     */
    static registerParser(languageCode, parserClass) {
        const normalized = languageCode.toLowerCase();
        parserRegistry.set(normalized, parserClass);
        // Also register without region code
        const langOnly = normalized.split('-')[0];
        if (langOnly !== normalized && !parserRegistry.has(langOnly)) {
            parserRegistry.set(langOnly, parserClass);
        }
    }
    /**
     * Create a parser for the specified language
     *
     * @throws Error if no parser is registered for the language
     */
    static createParser(languageCode, languageProvider) {
        const normalized = languageCode.toLowerCase();
        // Try exact match first
        let ParserClass = parserRegistry.get(normalized);
        // Try language without region
        if (!ParserClass) {
            const langOnly = normalized.split('-')[0];
            ParserClass = parserRegistry.get(langOnly);
        }
        if (!ParserClass) {
            const available = Array.from(parserRegistry.keys()).sort();
            throw new Error(`No parser registered for language: ${languageCode}. ` +
                `Available languages: ${available.join(', ')}. ` +
                `Register a parser using ParserFactory.registerParser().`);
        }
        return new ParserClass(languageProvider);
    }
    /**
     * Get list of registered language codes
     */
    static getRegisteredLanguages() {
        return Array.from(parserRegistry.keys()).sort();
    }
    /**
     * Check if a parser is registered for a language
     */
    static isLanguageRegistered(languageCode) {
        const normalized = languageCode.toLowerCase();
        if (parserRegistry.has(normalized))
            return true;
        const langOnly = normalized.split('-')[0];
        return parserRegistry.has(langOnly);
    }
    /**
     * Clear all registered parsers (mainly for testing)
     */
    static clearRegistry() {
        parserRegistry.clear();
    }
}
exports.ParserFactory = ParserFactory;
