// @sharpee/world-model/constants/directions.js
"use strict";
/**
 * Language-agnostic direction constants for Interactive Fiction
 *
 * These constants represent spatial relationships, not English words.
 * The parser owns the mapping from English words to these constants.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DirectionOpposites = exports.Direction = void 0;
exports.getOppositeDirection = getOppositeDirection;
exports.isDirection = isDirection;
exports.Direction = {
    NORTH: 'NORTH',
    SOUTH: 'SOUTH',
    EAST: 'EAST',
    WEST: 'WEST',
    NORTHEAST: 'NORTHEAST',
    NORTHWEST: 'NORTHWEST',
    SOUTHEAST: 'SOUTHEAST',
    SOUTHWEST: 'SOUTHWEST',
    UP: 'UP',
    DOWN: 'DOWN',
    IN: 'IN',
    OUT: 'OUT'
};
/**
 * Map of opposite directions using constants
 */
exports.DirectionOpposites = {
    [exports.Direction.NORTH]: exports.Direction.SOUTH,
    [exports.Direction.SOUTH]: exports.Direction.NORTH,
    [exports.Direction.EAST]: exports.Direction.WEST,
    [exports.Direction.WEST]: exports.Direction.EAST,
    [exports.Direction.NORTHEAST]: exports.Direction.SOUTHWEST,
    [exports.Direction.NORTHWEST]: exports.Direction.SOUTHEAST,
    [exports.Direction.SOUTHEAST]: exports.Direction.NORTHWEST,
    [exports.Direction.SOUTHWEST]: exports.Direction.NORTHEAST,
    [exports.Direction.UP]: exports.Direction.DOWN,
    [exports.Direction.DOWN]: exports.Direction.UP,
    [exports.Direction.IN]: exports.Direction.OUT,
    [exports.Direction.OUT]: exports.Direction.IN
};
/**
 * Get the opposite direction
 */
function getOppositeDirection(direction) {
    return exports.DirectionOpposites[direction];
}
/**
 * Check if a value is a valid Direction constant
 */
function isDirection(value) {
    return typeof value === 'string' && Object.values(exports.Direction).includes(value);
}
