// @sharpee/world-model/traits/region/regionTrait.js
"use strict";
/**
 * Region trait for geographic groupings of rooms (ADR-149).
 *
 * Entities with this trait represent named spatial regions. Rooms declare
 * membership via RoomTrait.regionId. Regions can be nested via parentRegionId.
 *
 * Public interface: RegionTrait, IRegionData.
 * Owner context: @sharpee/world-model — traits / spatial
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.RegionTrait = void 0;
const trait_types_1 = require("../trait-types");
/**
 * Marks an entity as a spatial region that groups rooms.
 *
 * Rooms reference their region via `RoomTrait.regionId`. Regions can form
 * a hierarchy through `parentRegionId` — a room in a child region is
 * implicitly in all ancestor regions.
 */
class RegionTrait {
    static type = trait_types_1.TraitType.REGION;
    type = trait_types_1.TraitType.REGION;
    /** Human-readable region name. */
    name;
    /** Parent region entity ID for nesting (optional). */
    parentRegionId;
    /** Region-wide ambient sound propagated to contained rooms. */
    ambientSound;
    /** Region-wide ambient smell propagated to contained rooms. */
    ambientSmell;
    /** Whether rooms in this region default to dark. */
    defaultDark;
    constructor(data) {
        this.name = data.name;
        this.parentRegionId = data.parentRegionId;
        this.ambientSound = data.ambientSound;
        this.ambientSmell = data.ambientSmell;
        this.defaultDark = data.defaultDark ?? false;
    }
}
exports.RegionTrait = RegionTrait;
