// @sharpee/world-model/traits/edible/edibleTrait.js
"use strict";
// packages/world-model/src/traits/edible/edibleTrait.ts
Object.defineProperty(exports, "__esModule", { value: true });
exports.EdibleTrait = void 0;
const trait_types_1 = require("../trait-types");
/**
 * Edible trait indicates an entity can be eaten or drunk.
 *
 * This trait contains only data - all consumption logic
 * is in EdibleBehavior.
 */
class EdibleTrait {
    static type = trait_types_1.TraitType.EDIBLE;
    type = trait_types_1.TraitType.EDIBLE;
    // Core EdibleData properties
    nutrition;
    servings;
    liquid;
    consumeMessage;
    remainsType;
    hasEffect;
    effectDescription;
    weight;
    bulk;
    // Extended properties
    taste;
    effects;
    satisfiesHunger;
    satisfiesThirst;
    constructor(data = {}) {
        this.servings = data.servings ?? 1;
        this.nutrition = data.nutrition ?? 1;
        this.liquid = data.liquid ?? false;
        this.consumeMessage = data.consumeMessage;
        this.remainsType = data.remainsType;
        this.hasEffect = data.hasEffect ?? (data.effects && data.effects.length > 0) ?? false;
        this.effectDescription = data.effectDescription;
        this.weight = data.weight ?? 1;
        this.bulk = data.bulk ?? 1;
        // Extended properties
        this.taste = data.taste;
        this.effects = data.effects;
        this.satisfiesHunger = data.satisfiesHunger;
        this.satisfiesThirst = data.satisfiesThirst;
    }
}
exports.EdibleTrait = EdibleTrait;
