// @sharpee/world-model/traits/edible/edibleBehavior.js
"use strict";
// packages/world-model/src/traits/edible/edibleBehavior.ts
Object.defineProperty(exports, "__esModule", { value: true });
exports.EdibleBehavior = void 0;
const behavior_1 = require("../../behaviors/behavior");
const trait_types_1 = require("../trait-types");
const core_1 = require("../../../core/index.js");
const if_events_1 = require("../../constants/if-events");
const action_failures_1 = require("../../constants/action-failures");
/**
 * Behavior for edible entities.
 *
 * Handles the logic for consuming food and drink.
 */
class EdibleBehavior extends behavior_1.Behavior {
    static requiredTraits = [trait_types_1.TraitType.EDIBLE];
    /**
     * Check if an item can be consumed
     */
    static canConsume(item) {
        const edible = EdibleBehavior.require(item, trait_types_1.TraitType.EDIBLE);
        return (edible.servings ?? 1) > 0;
    }
    /**
     * Consume the item
     * @returns Events describing what happened
     */
    static consume(item, actor) {
        const edible = EdibleBehavior.require(item, trait_types_1.TraitType.EDIBLE);
        let currentServings = edible.servings ?? 1;
        const isLiquid = edible.liquid ?? false;
        if (currentServings <= 0) {
            return [
                (0, core_1.createEvent)(if_events_1.IFEvents.ACTION_FAILED, {
                    action: isLiquid ? 'drink' : 'eat',
                    reason: action_failures_1.ActionFailureReason.NOTHING_LEFT,
                    customMessage: "There's nothing left."
                }, { target: item.id, actor: actor.id })
            ];
        }
        currentServings--;
        edible.servings = currentServings;
        const events = [
            (0, core_1.createEvent)(isLiquid ? if_events_1.IFEvents.ITEM_DRUNK : if_events_1.IFEvents.ITEM_EATEN, {
                nutrition: edible.nutrition,
                servingsLeft: currentServings,
                customMessage: edible.consumeMessage,
                hasEffect: edible.hasEffect,
                effectDescription: edible.effectDescription
            }, { target: item.id, actor: actor.id })
        ];
        // If fully consumed, note that
        if (currentServings <= 0) {
            events.push((0, core_1.createEvent)(if_events_1.IFEvents.ITEM_DESTROYED, {
                reason: 'consumed',
                remainsType: edible.remainsType
            }, { target: item.id, actor: actor.id }));
        }
        return events;
    }
    /**
     * Check if item is fully consumed
     */
    static isEmpty(item) {
        const edible = EdibleBehavior.require(item, trait_types_1.TraitType.EDIBLE);
        return (edible.servings ?? 1) <= 0;
    }
    /**
     * Check if this is a liquid
     */
    static isLiquid(item) {
        const edible = EdibleBehavior.require(item, trait_types_1.TraitType.EDIBLE);
        return edible.liquid ?? false;
    }
    /**
     * Get nutrition value
     */
    static getNutrition(item) {
        const edible = EdibleBehavior.require(item, trait_types_1.TraitType.EDIBLE);
        return edible.nutrition;
    }
    /**
     * Get remaining servings
     */
    static getServings(item) {
        const edible = EdibleBehavior.require(item, trait_types_1.TraitType.EDIBLE);
        return edible.servings ?? 1;
    }
    /**
     * Check if consuming has special effects
     */
    static hasEffect(item) {
        const edible = EdibleBehavior.require(item, trait_types_1.TraitType.EDIBLE);
        return edible.hasEffect;
    }
    /**
     * Get taste quality
     */
    static getTaste(item) {
        const edible = EdibleBehavior.require(item, trait_types_1.TraitType.EDIBLE);
        return edible.taste;
    }
    /**
     * Get effects array
     */
    static getEffects(item) {
        const edible = EdibleBehavior.require(item, trait_types_1.TraitType.EDIBLE);
        return edible.effects;
    }
    /**
     * Check if item has a specific effect
     */
    static hasSpecificEffect(item, effect) {
        const effects = EdibleBehavior.getEffects(item);
        return effects?.includes(effect) ?? false;
    }
    /**
     * Check if eating satisfies hunger
     */
    static satisfiesHunger(item) {
        const edible = EdibleBehavior.require(item, trait_types_1.TraitType.EDIBLE);
        return edible.satisfiesHunger;
    }
}
exports.EdibleBehavior = EdibleBehavior;
