// @sharpee/stdlib/actions/standard/waiting/waiting.js
"use strict";
/**
 * Waiting action - passes time without doing anything
 *
 * This is a signal action that emits an event to indicate the player
 * chose to wait. The engine handles turn advancement and daemon processing.
 *
 * Uses four-phase pattern:
 * 1. validate: Always succeeds (no preconditions for waiting)
 * 2. execute: No world mutations (stores location in sharedData)
 * 3. blocked: Handle validation failures (n/a - always succeeds)
 * 4. report: Emits if.event.waited signal for engine/daemons
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.waitingAction = void 0;
const constants_1 = require("../../constants");
function getWaitingSharedData(context) {
    return context.sharedData;
}
exports.waitingAction = {
    id: constants_1.IFActions.WAITING,
    requiredMessages: [
        'time_passes'
    ],
    group: "meta",
    metadata: {
        requiresDirectObject: false,
        requiresIndirectObject: false
    },
    validate(context) {
        // Waiting always succeeds - no preconditions
        return { valid: true };
    },
    execute(context) {
        // Waiting has NO world mutations
        // Just store location info in sharedData for report phase
        const location = context.currentLocation;
        const sharedData = getWaitingSharedData(context);
        sharedData.locationId = location?.id;
        sharedData.locationName = location?.name;
    },
    blocked(context, result) {
        // Waiting always succeeds, but include blocked for consistency
        return [context.event('if.event.wait_blocked', {
                blocked: true,
                messageId: `${context.action.id}.${result.error}`,
                params: result.params,
                reason: result.error
            })];
    },
    report(context) {
        const events = [];
        const sharedData = getWaitingSharedData(context);
        // Emit waited event with messageId for text rendering
        // Engine/daemons listen to this to advance turn counter and process scheduled events
        events.push(context.event('if.event.waited', {
            messageId: `${context.action.id}.time_passes`,
            turnsPassed: 1,
            location: sharedData.locationId,
            locationName: sharedData.locationName
        }));
        return events;
    }
};
