// @sharpee/world-model/traits/lockable/lockableTrait.js
"use strict";
// packages/world-model/src/traits/lockable/lockableTrait.ts
Object.defineProperty(exports, "__esModule", { value: true });
exports.LockableTrait = void 0;
const trait_types_1 = require("../trait-types");
/**
 * Lockable trait for entities that can be locked and unlocked.
 * Usually combined with OpenableTrait.
 *
 * This trait contains only data - all locking/unlocking logic
 * is in LockableBehavior.
 */
class LockableTrait {
    static type = trait_types_1.TraitType.LOCKABLE;
    type = trait_types_1.TraitType.LOCKABLE;
    // LockableData properties
    isLocked;
    startsLocked;
    keyId;
    keyIds;
    acceptsMasterKey;
    lockMessage;
    unlockMessage;
    alreadyLockedMessage;
    alreadyUnlockedMessage;
    lockedMessage;
    wrongKeyMessage;
    autoLock;
    lockSound;
    unlockSound;
    constructor(data = {}) {
        // Set defaults and merge with provided data
        this.startsLocked = data.startsLocked ?? false;
        this.isLocked = data.isLocked ?? this.startsLocked;
        this.keyId = data.keyId;
        this.keyIds = data.keyIds;
        this.acceptsMasterKey = data.acceptsMasterKey ?? true;
        this.lockMessage = data.lockMessage;
        this.unlockMessage = data.unlockMessage;
        this.alreadyLockedMessage = data.alreadyLockedMessage;
        this.alreadyUnlockedMessage = data.alreadyUnlockedMessage;
        this.lockedMessage = data.lockedMessage;
        this.wrongKeyMessage = data.wrongKeyMessage;
        this.autoLock = data.autoLock ?? false;
        this.lockSound = data.lockSound;
        this.unlockSound = data.unlockSound;
    }
}
exports.LockableTrait = LockableTrait;
