// @sharpee/world-model/traits/lockable/lockableBehavior.js
"use strict";
// packages/world-model/src/traits/lockable/lockableBehavior.ts
Object.defineProperty(exports, "__esModule", { value: true });
exports.LockableBehavior = void 0;
const behavior_1 = require("../../behaviors/behavior");
const trait_types_1 = require("../trait-types");
/**
 * Behavior for lockable entities.
 *
 * Handles the logic for locking and unlocking entities.
 * Requires LOCKABLE trait, OPENABLE is optional.
 */
class LockableBehavior extends behavior_1.Behavior {
    static requiredTraits = [trait_types_1.TraitType.LOCKABLE];
    /**
     * Check if a key can unlock this entity
     */
    static canUnlockWith(entity, keyId) {
        const lockable = LockableBehavior.require(entity, trait_types_1.TraitType.LOCKABLE);
        // Check specific key
        if (lockable.keyId === keyId) {
            return true;
        }
        // Check multiple keys
        if (lockable.keyIds && lockable.keyIds.includes(keyId)) {
            return true;
        }
        return false;
    }
    /**
     * Check if a key can lock this entity (usually same as unlock)
     */
    static canLockWith(entity, keyId) {
        // By default, same keys that unlock can also lock
        return this.canUnlockWith(entity, keyId);
    }
    /**
     * Check if this entity requires a key to unlock
     */
    static requiresKey(entity) {
        const lockable = LockableBehavior.require(entity, trait_types_1.TraitType.LOCKABLE);
        return !!(lockable.keyId || lockable.keyIds?.length);
    }
    /**
     * Check if an entity can be locked
     */
    static canLock(entity) {
        const lockable = LockableBehavior.require(entity, trait_types_1.TraitType.LOCKABLE);
        // Check if open (only if openable)
        if (entity.has(trait_types_1.TraitType.OPENABLE)) {
            const openable = entity.get(trait_types_1.TraitType.OPENABLE);
            if (openable.isOpen) {
                return false; // Can't lock something that's open
            }
        }
        return !lockable.isLocked;
    }
    /**
     * Check if an entity can be unlocked
     */
    static canUnlock(entity) {
        const lockable = LockableBehavior.require(entity, trait_types_1.TraitType.LOCKABLE);
        return lockable.isLocked;
    }
    /**
     * Lock the entity
     * @returns Result describing what happened
     */
    static lock(entity, keyEntity) {
        const lockable = LockableBehavior.require(entity, trait_types_1.TraitType.LOCKABLE);
        if (lockable.isLocked) {
            return {
                success: false,
                alreadyLocked: true,
                stateChanged: false
            };
        }
        // Can't lock something that's open (only check if openable)
        if (entity.has(trait_types_1.TraitType.OPENABLE)) {
            const openable = entity.get(trait_types_1.TraitType.OPENABLE);
            if (openable.isOpen) {
                return {
                    success: false,
                    notClosed: true,
                    stateChanged: false
                };
            }
        }
        // Check if key is required
        if (LockableBehavior.requiresKey(entity)) {
            if (!keyEntity) {
                return {
                    success: false,
                    noKey: true,
                    stateChanged: false
                };
            }
            // Check if key is valid
            if (!this.canLockWith(entity, keyEntity.id)) {
                return {
                    success: false,
                    wrongKey: true,
                    stateChanged: false
                };
            }
        }
        // Lock it
        lockable.isLocked = true;
        return {
            success: true,
            stateChanged: true,
            lockMessage: lockable.lockMessage,
            lockSound: lockable.lockSound
        };
    }
    /**
     * Unlock the entity with a key
     * @returns Result describing what happened
     */
    static unlock(entity, keyEntity) {
        const lockable = LockableBehavior.require(entity, trait_types_1.TraitType.LOCKABLE);
        if (!lockable.isLocked) {
            return {
                success: false,
                alreadyUnlocked: true,
                stateChanged: false
            };
        }
        // Check if key is required
        if (LockableBehavior.requiresKey(entity)) {
            if (!keyEntity) {
                return {
                    success: false,
                    noKey: true,
                    stateChanged: false
                };
            }
            // Check if key is valid
            if (!this.canUnlockWith(entity, keyEntity.id)) {
                return {
                    success: false,
                    wrongKey: true,
                    stateChanged: false
                };
            }
        }
        // Unlock it
        lockable.isLocked = false;
        return {
            success: true,
            stateChanged: true,
            unlockMessage: lockable.unlockMessage,
            unlockSound: lockable.unlockSound
        };
    }
    /**
     * Force unlock without a key (for admin/debug)
     */
    static forceUnlock(entity) {
        const lockable = LockableBehavior.require(entity, trait_types_1.TraitType.LOCKABLE);
        lockable.isLocked = false;
    }
    /**
     * Check if entity is currently locked
     */
    static isLocked(entity) {
        const lockable = LockableBehavior.require(entity, trait_types_1.TraitType.LOCKABLE);
        return lockable.isLocked;
    }
    /**
     * Handle auto-lock when closing
     */
    static handleClose(entity) {
        const lockable = LockableBehavior.require(entity, trait_types_1.TraitType.LOCKABLE);
        if (lockable.autoLock && !lockable.isLocked) {
            return this.lock(entity);
        }
        return null;
    }
    /**
     * Check if we can open this entity (not locked)
     */
    static canOpen(entity) {
        const lockable = LockableBehavior.require(entity, trait_types_1.TraitType.LOCKABLE);
        return !lockable.isLocked;
    }
    /**
     * Get the locked message
     */
    static getLockedMessage(entity) {
        const lockable = LockableBehavior.require(entity, trait_types_1.TraitType.LOCKABLE);
        return lockable.lockedMessage;
    }
}
exports.LockableBehavior = LockableBehavior;
