// @sharpee/stdlib/actions/standard/searching/searching.js
"use strict";
/**
 * Searching action - search objects or locations for hidden items
 *
 * This action allows players to search containers, supporters, or locations
 * to find concealed items or discover additional details.
 *
 * Uses four-phase pattern:
 * 1. validate: Check target is searchable (container open if applicable)
 * 2. execute: Analyze contents, reveal concealed items (mutation)
 * 3. blocked: Generate events when validation fails
 * 4. report: Generate success events
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.searchingAction = void 0;
const world_model_1 = require("../../../../world-model/index.js");
const constants_1 = require("../../constants");
const types_1 = require("../../../scope/types");
const searching_helpers_1 = require("../searching-helpers");
function getSearchingSharedData(context) {
    return context.sharedData;
}
exports.searchingAction = {
    id: constants_1.IFActions.SEARCHING,
    // Default scope requirements for this action's slots
    defaultScope: {
        target: types_1.ScopeLevel.REACHABLE
    },
    requiredMessages: [
        'not_visible',
        'not_reachable',
        'container_closed',
        'nothing_special',
        'found_items',
        'empty_container',
        'container_contents',
        'supporter_contents',
        'searched_location',
        'searched_object',
        'found_concealed'
    ],
    metadata: {
        requiresDirectObject: true,
        requiresIndirectObject: false,
        directObjectScope: types_1.ScopeLevel.REACHABLE
    },
    validate(context) {
        const target = context.command.directObject?.entity;
        // If no target, we'll search the current location (always valid)
        if (!target) {
            return { valid: true };
        }
        // Check if it's a container that needs to be open
        if (target.has(world_model_1.TraitType.CONTAINER) && target.has(world_model_1.TraitType.OPENABLE)) {
            if (!world_model_1.OpenableBehavior.isOpen(target)) {
                return {
                    valid: false,
                    error: 'container_closed',
                    params: { target: target.name }
                };
            }
        }
        return { valid: true };
    },
    execute(context) {
        const target = context.command.directObject?.entity;
        const sharedData = getSearchingSharedData(context);
        // Analyze what we're searching
        const searchContext = (0, searching_helpers_1.analyzeSearchTarget)(context, target);
        // Reveal any concealed items found (world mutation)
        if (searchContext.concealedItems.length > 0) {
            (0, searching_helpers_1.revealConcealedItems)(searchContext.concealedItems);
        }
        // Build event data and message
        const eventData = (0, searching_helpers_1.buildSearchEventData)(searchContext);
        const { messageId, params } = (0, searching_helpers_1.determineSearchMessage)(searchContext);
        // Store in sharedData for report phase
        sharedData.eventData = eventData;
        sharedData.messageId = messageId;
        sharedData.params = params;
    },
    blocked(context, result) {
        const target = context.command.directObject?.entity;
        return [context.event('if.event.searched', {
                blocked: true,
                messageId: `${context.action.id}.${result.error}`,
                params: { target: target?.name, ...result.params },
                reason: result.error,
                targetId: target?.id,
                targetName: target?.name
            })];
    },
    report(context) {
        const events = [];
        const sharedData = getSearchingSharedData(context);
        // Emit searched event with messageId for text rendering
        events.push(context.event('if.event.searched', {
            messageId: `${context.action.id}.${sharedData.messageId || 'nothing_special'}`,
            params: sharedData.params || {},
            ...sharedData.eventData
        }));
        return events;
    },
    group: "sensory"
};
