// @sharpee/ext-testing/commands/registry.js
"use strict";
/**
 * Command Registry
 *
 * Central registry for debug and test commands.
 * Commands can be looked up by GDT code or test syntax.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.createCommandRegistry = createCommandRegistry;
exports.parseGdtInput = parseGdtInput;
exports.parseTestInput = parseTestInput;
/**
 * Create a new command registry
 */
function createCommandRegistry() {
    const byCode = new Map();
    const byTestSyntax = new Map();
    const allCommands = [];
    const registry = {
        register(command) {
            // Register by code (uppercase for case-insensitive lookup)
            byCode.set(command.code.toUpperCase(), command);
            // Register by test syntax if provided
            if (command.testSyntax) {
                byTestSyntax.set(command.testSyntax.toLowerCase(), command);
            }
            allCommands.push(command);
        },
        getByCode(code) {
            return byCode.get(code.toUpperCase());
        },
        getByTestSyntax(syntax) {
            return byTestSyntax.get(syntax.toLowerCase());
        },
        getAll() {
            return [...allCommands];
        },
        getByCategory(category) {
            return allCommands.filter((cmd) => cmd.category === category);
        },
    };
    return registry;
}
/**
 * Parse a GDT-style command input
 * Format: "CODE arg1 arg2" or just "CODE"
 */
function parseGdtInput(input) {
    const trimmed = input.trim();
    const parts = trimmed.split(/\s+/);
    const code = parts[0] || '';
    const args = parts.slice(1);
    return { code: code.toUpperCase(), args };
}
/**
 * Parse a test command input
 * Format: "$command arg1 arg2" or just "$command"
 */
function parseTestInput(input) {
    const trimmed = input.trim();
    if (!trimmed.startsWith('$')) {
        return undefined;
    }
    const withoutPrefix = trimmed.slice(1);
    const parts = withoutPrefix.split(/\s+/);
    const syntax = parts[0] || '';
    const args = parts.slice(1);
    return { syntax: syntax.toLowerCase(), args };
}
