// @sharpee/ext-testing/context/debug-context.js
"use strict";
/**
 * Debug Context Implementation
 *
 * Provides a unified context for debug and test commands to interact with the game world.
 * Generalizes the pattern from Dungeo's GDTContext.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.createDebugContext = createDebugContext;
exports.formatEntity = formatEntity;
exports.formatLocationChain = formatLocationChain;
const world_model_1 = require("../../world-model/index.js");
/**
 * Create a debug context for the given world
 */
function createDebugContext(world) {
    // Create AuthorModel for bypassing game rules
    const author = new world_model_1.AuthorModel(world.getDataStore(), world);
    // Debug flags storage
    const flags = new Map();
    // Find player entity
    const player = world.getPlayer();
    if (!player) {
        throw new Error('Cannot create debug context: no player entity found');
    }
    const context = {
        world,
        author,
        player,
        flags,
        // ========================================================================
        // Entity Lookup
        // ========================================================================
        findEntity(idOrName) {
            // Try exact ID match first
            const byId = world.getEntity(idOrName);
            if (byId)
                return byId;
            // Normalize search term: lowercase, convert hyphens to spaces
            const searchLower = idOrName.toLowerCase();
            const searchNormalized = searchLower.replace(/-/g, ' ');
            const allEntities = world.getAllEntities();
            // Exact name match (with normalized search)
            const exactMatch = allEntities.find((e) => e.name?.toLowerCase() === searchNormalized ||
                e.name?.toLowerCase() === searchLower);
            if (exactMatch)
                return exactMatch;
            // Partial name match (with normalized search)
            const partialMatch = allEntities.find((e) => e.name?.toLowerCase().includes(searchNormalized) ||
                e.name?.toLowerCase().includes(searchLower));
            if (partialMatch)
                return partialMatch;
            // ID contains search term
            const idMatch = allEntities.find((e) => e.id.toLowerCase().includes(searchLower));
            return idMatch;
        },
        findRoom(idOrName) {
            const entity = context.findEntity(idOrName);
            if (entity && entity.type === 'room') {
                return entity;
            }
            // Search only rooms
            const searchLower = idOrName.toLowerCase();
            const rooms = world.getAllEntities().filter((e) => e.type === 'room');
            const exactMatch = rooms.find((r) => r.name?.toLowerCase() === searchLower);
            if (exactMatch)
                return exactMatch;
            // Check aliases (from IdentityTrait) for exact match
            const aliasMatch = rooms.find((r) => {
                const identity = r.get(world_model_1.IdentityTrait);
                return identity?.aliases?.some((a) => a.toLowerCase() === searchLower);
            });
            if (aliasMatch)
                return aliasMatch;
            const partialMatch = rooms.find((r) => r.name?.toLowerCase().includes(searchLower) ||
                r.id.toLowerCase().includes(searchLower));
            return partialMatch;
        },
        findEntities(predicate) {
            return world.getAllEntities().filter(predicate);
        },
        // ========================================================================
        // Location Queries
        // ========================================================================
        getPlayerLocation() {
            const locationId = world.getLocation(player.id);
            if (!locationId)
                return undefined;
            return world.getEntity(locationId);
        },
        getInventory() {
            return world.getContents(player.id);
        },
        getContents(locationId) {
            return world.getContents(locationId);
        },
        // ========================================================================
        // Mutations (bypass game rules via AuthorModel)
        // ========================================================================
        teleportPlayer(roomId) {
            const room = context.findRoom(roomId);
            if (!room)
                return false;
            author.moveEntity(player.id, room.id);
            return true;
        },
        moveObject(objectId, locationId) {
            const object = context.findEntity(objectId);
            const location = context.findEntity(locationId);
            if (!object || !location)
                return false;
            author.moveEntity(object.id, location.id);
            return true;
        },
        takeObject(objectId) {
            const object = context.findEntity(objectId);
            if (!object)
                return false;
            author.moveEntity(object.id, player.id);
            return true;
        },
        removeObject(objectId) {
            const object = context.findEntity(objectId);
            if (!object)
                return false;
            // Move to limbo (null location)
            author.moveEntity(object.id, 'limbo');
            return true;
        },
        spawnObject(objectId, locationId) {
            const object = context.findEntity(objectId);
            const location = context.findEntity(locationId);
            if (!object || !location)
                return false;
            author.moveEntity(object.id, location.id);
            return true;
        },
        // ========================================================================
        // Flag Operations
        // ========================================================================
        setFlag(name, value) {
            flags.set(name, value);
        },
        getFlag(name) {
            return flags.get(name) ?? false;
        },
    };
    return context;
}
/**
 * Format entity for debug display
 */
function formatEntity(entity) {
    const traitTypes = entity.traits
        ? Array.from(entity.traits.values()).map((t) => t.type.split('.').pop()).join(', ')
        : 'none';
    return `${entity.id} (${entity.name || 'unnamed'}) [${traitTypes}]`;
}
/**
 * Format location chain for an entity
 */
function formatLocationChain(entity, world) {
    const chain = [entity.id];
    let current = entity;
    let depth = 0;
    while (depth < 10) {
        const locationId = world.getLocation(current.id);
        if (!locationId)
            break;
        chain.push(locationId);
        const location = world.getEntity(locationId);
        if (!location)
            break;
        current = location;
        depth++;
    }
    return chain.join(' -> ');
}
