// @sharpee/stdlib/actions/standard/eating/eating.js
"use strict";
/**
 * Eating action - consume edible items
 *
 * This action handles eating items that have the EDIBLE trait.
 * It validates that the item is edible and not a drink.
 *
 * Uses four-phase pattern:
 * 1. validate: Check if item is reachable, edible, not a drink, and not consumed
 * 2. execute: Delegate to EdibleBehavior for consumption
 * 3. blocked: Generate events when validation fails
 * 4. report: Generate success events
 *
 * Supports implicit take - "eat apple" when apple is on ground will
 * automatically take it first.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.eatingAction = void 0;
const world_model_1 = require("../../../../world-model/index.js");
const constants_1 = require("../../constants");
const types_1 = require("../../../scope/types");
function getEatingSharedData(context) {
    return context.sharedData;
}
exports.eatingAction = {
    id: constants_1.IFActions.EATING,
    // Default scope requirements for this action's slots
    defaultScope: {
        item: types_1.ScopeLevel.REACHABLE
    },
    // ADR-104: Implicit inference requirements
    targetRequirements: {
        trait: world_model_1.TraitType.EDIBLE,
        description: 'edible'
    },
    // Eating requires holding the food
    requiresHolding: true,
    requiredMessages: [
        'no_item',
        'not_visible',
        'not_reachable',
        'not_edible',
        'is_drink',
        'already_consumed',
        'eaten',
        'eaten_all',
        'eaten_some',
        'delicious',
        'tasty',
        'bland',
        'awful',
        'filling',
        'still_hungry',
        'poisonous'
    ],
    metadata: {
        requiresDirectObject: true,
        requiresIndirectObject: false,
        directObjectScope: types_1.ScopeLevel.REACHABLE
    },
    validate(context) {
        const item = context.command.directObject?.entity;
        // Must have an item to eat
        if (!item) {
            return {
                valid: false,
                error: 'no_item'
            };
        }
        // Check scope - must be able to reach the item
        const scopeCheck = context.requireScope(item, types_1.ScopeLevel.REACHABLE);
        if (!scopeCheck.ok) {
            return scopeCheck.error;
        }
        // Check if item is edible
        if (!item.has(world_model_1.TraitType.EDIBLE)) {
            return {
                valid: false,
                error: 'not_edible',
                params: { item: item.name }
            };
        }
        // Use behavior for liquid check (it's a drink, not food)
        if (world_model_1.EdibleBehavior.isLiquid(item)) {
            return {
                valid: false,
                error: 'is_drink',
                params: { item: item.name }
            };
        }
        // Use behavior for consumption check (servings > 0)
        if (!world_model_1.EdibleBehavior.canConsume(item)) {
            return {
                valid: false,
                error: 'already_consumed',
                params: { item: item.name }
            };
        }
        // Implicit take - "eat apple" when apple is on ground
        const carryCheck = context.requireCarriedOrImplicitTake(item);
        if (!carryCheck.ok) {
            return carryCheck.error;
        }
        return { valid: true };
    },
    /**
     * Execute the eat action - performs mutations only
     * Assumes validation has already passed
     */
    execute(context) {
        const item = context.command.directObject.entity;
        const actor = context.player;
        // Capture state before mutation for event data
        const servingsBefore = world_model_1.EdibleBehavior.getServings(item);
        const nutrition = world_model_1.EdibleBehavior.getNutrition(item);
        const taste = world_model_1.EdibleBehavior.getTaste(item);
        const effects = world_model_1.EdibleBehavior.getEffects(item);
        const satisfiesHunger = world_model_1.EdibleBehavior.satisfiesHunger(item);
        // Delegate to behavior for consumption (decrements servings)
        world_model_1.EdibleBehavior.consume(item, actor);
        // Calculate servings remaining after consumption
        const servingsRemaining = world_model_1.EdibleBehavior.getServings(item);
        // Build event data
        const eventData = {
            item: item.id,
            itemName: item.name
        };
        // Add nutritional information if available
        if (nutrition !== undefined && nutrition !== 1) {
            eventData.nutrition = nutrition;
        }
        // Include servings info for multi-serving food (servingsBefore > 1)
        // Note: We can't detect if single-serving food was originally multi-serving
        // after all but one serving was consumed. This is a known limitation.
        if (servingsBefore > 1) {
            eventData.servings = servingsBefore;
            eventData.servingsRemaining = servingsRemaining;
        }
        // Add effects if present
        if (effects && effects.length > 0) {
            eventData.effects = effects;
        }
        // Add hunger satisfaction if set
        if (satisfiesHunger !== undefined) {
            eventData.satisfiesHunger = satisfiesHunger;
        }
        // Determine message based on properties
        let messageId = 'eaten';
        // Message priority: effects > taste > servings > hunger > default
        if (effects && effects.includes('poison')) {
            messageId = 'poisonous';
        }
        else if (taste) {
            switch (taste) {
                case 'delicious':
                    messageId = 'delicious';
                    break;
                case 'tasty':
                case 'good':
                    messageId = 'tasty';
                    break;
                case 'bland':
                case 'plain':
                    messageId = 'bland';
                    break;
                case 'awful':
                case 'terrible':
                    messageId = 'awful';
                    break;
            }
        }
        else if (servingsBefore > 1 && servingsRemaining > 0) {
            // Multi-serving food with servings remaining
            messageId = 'eaten_some';
        }
        else if (servingsBefore > 1 && servingsRemaining === 0) {
            // Finished all servings of multi-serving food (e.g., ate 2-serving food going to 0)
            messageId = 'eaten_all';
        }
        else if (satisfiesHunger === true) {
            messageId = 'filling';
        }
        else if (satisfiesHunger === false) {
            messageId = 'still_hungry';
        }
        // Store in sharedData for report phase
        const sharedData = getEatingSharedData(context);
        sharedData.itemId = item.id;
        sharedData.itemName = item.name;
        sharedData.messageId = messageId;
        sharedData.eventData = eventData;
    },
    /**
     * Generate events when validation fails
     */
    blocked(context, result) {
        const item = context.command.directObject?.entity;
        return [context.event('if.event.eaten', {
                blocked: true,
                messageId: `${context.action.id}.${result.error}`,
                params: { item: item?.name, ...result.params },
                reason: result.error,
                itemId: item?.id,
                itemName: item?.name
            })];
    },
    /**
     * Report phase - generates all events after successful execution
     */
    report(context) {
        const events = [];
        const sharedData = getEatingSharedData(context);
        // Prepend any implicit take events (from requireCarriedOrImplicitTake)
        if (context.sharedData.implicitTakeEvents) {
            events.push(...context.sharedData.implicitTakeEvents);
        }
        // Emit the EATEN event with messageId for text rendering
        // eventData already contains item (id) and itemName (string)
        events.push(context.event('if.event.eaten', {
            messageId: `${context.action.id}.${sharedData.messageId}`,
            params: { item: sharedData.eventData?.itemName },
            ...sharedData.eventData
        }));
        return events;
    },
    group: "interaction"
};
