// @sharpee/stdlib/actions/standard/hiding/hiding.js
"use strict";
/**
 * Hiding action — enter concealment behind/under/on/inside a hiding spot (ADR-148)
 *
 * Adds ConcealedStateTrait to the player entity, making them invisible
 * to NPCs via the visibility capability system.
 *
 * Uses four-phase pattern:
 * 1. validate: Target must have ConcealmentTrait with matching position
 * 2. execute: Add ConcealedStateTrait to player
 * 3. report: Emit if.event.player_concealed
 * 4. blocked: Emit error message (nothing_to_hide, cant_hide_there, already_hidden)
 *
 * Public interface: hidingAction.
 * Owner context: @sharpee/stdlib / actions / hiding
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.hidingAction = void 0;
const constants_1 = require("../../constants");
const world_model_1 = require("../../../../world-model/index.js");
function getHidingSharedData(context) {
    return context.sharedData;
}
exports.hidingAction = {
    id: constants_1.IFActions.HIDING,
    requiredMessages: [
        'behind',
        'under',
        'on',
        'inside',
        'nothing_to_hide',
        'cant_hide_there',
        'already_hidden',
    ],
    group: 'interaction',
    metadata: {
        requiresDirectObject: true,
        requiresIndirectObject: false,
    },
    validate(context) {
        const player = context.player;
        // Already concealed?
        if ((0, world_model_1.isConcealed)(player)) {
            return { valid: false, error: 'already_hidden' };
        }
        // Get the target entity
        const target = context.command.directObject?.entity;
        if (!target) {
            return { valid: false, error: 'nothing_to_hide' };
        }
        // Target must have ConcealmentTrait
        const concealmentTrait = target.get(world_model_1.ConcealmentTrait.type);
        if (!concealmentTrait) {
            return {
                valid: false,
                error: 'nothing_to_hide',
                params: { target: target.name },
            };
        }
        // Get the requested position from grammar default semantics (via extras)
        const position = context.command.parsed.extras?.position;
        if (!position) {
            return { valid: false, error: 'nothing_to_hide' };
        }
        // Check position is supported
        if (!concealmentTrait.supportsPosition(position)) {
            return {
                valid: false,
                error: 'cant_hide_there',
                params: { target: target.name, position },
            };
        }
        // Store data for execute/report phases
        const sharedData = getHidingSharedData(context);
        sharedData.targetId = target.id;
        sharedData.targetName = target.name;
        sharedData.position = position;
        sharedData.quality = concealmentTrait.quality;
        return { valid: true };
    },
    execute(context) {
        const sharedData = getHidingSharedData(context);
        const player = context.player;
        // Add ConcealedStateTrait — this IS the concealment state
        player.add(new world_model_1.ConcealedStateTrait({
            targetId: sharedData.targetId,
            position: sharedData.position,
            quality: sharedData.quality,
        }));
    },
    blocked(context, result) {
        return [context.event('if.event.hide_blocked', {
                blocked: true,
                messageId: `${context.action.id}.${result.error}`,
                params: result.params || {},
                reason: result.error,
            })];
    },
    report(context) {
        const sharedData = getHidingSharedData(context);
        return [context.event('if.event.player_concealed', {
                messageId: `${context.action.id}.${sharedData.position}`,
                targetId: sharedData.targetId,
                targetName: sharedData.targetName,
                position: sharedData.position,
                quality: sharedData.quality,
            })];
    },
};
