// @sharpee/stdlib/actions/standard/hiding/revealing.js
"use strict";
/**
 * Revealing action — voluntarily end concealment (ADR-148)
 *
 * Removes ConcealedStateTrait from the player entity.
 * Grammar: "stand up", "come out", "reveal myself", "unhide", "stop hiding"
 *
 * Uses four-phase pattern:
 * 1. validate: Player must be concealed
 * 2. execute: Remove ConcealedStateTrait
 * 3. report: Emit if.event.player_revealed with reason 'explicit'
 * 4. blocked: Emit error (not_hidden)
 *
 * Public interface: revealingAction.
 * Owner context: @sharpee/stdlib / actions / hiding
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.revealingAction = void 0;
const constants_1 = require("../../constants");
const world_model_1 = require("../../../../world-model/index.js");
function getRevealingSharedData(context) {
    return context.sharedData;
}
exports.revealingAction = {
    id: constants_1.IFActions.REVEALING,
    requiredMessages: [
        'revealed',
        'not_hidden',
    ],
    group: 'interaction',
    metadata: {
        requiresDirectObject: false,
        requiresIndirectObject: false,
    },
    validate(context) {
        const player = context.player;
        if (!(0, world_model_1.isConcealed)(player)) {
            return { valid: false, error: 'not_hidden' };
        }
        // Store current concealment info for report
        const concealment = (0, world_model_1.getConcealmentState)(player);
        const sharedData = getRevealingSharedData(context);
        sharedData.wasHidingBehind = concealment?.targetId;
        sharedData.position = concealment?.position;
        return { valid: true };
    },
    execute(context) {
        const player = context.player;
        player.remove(world_model_1.ConcealedStateTrait.type);
    },
    blocked(context, result) {
        return [context.event('if.event.reveal_blocked', {
                blocked: true,
                messageId: `${context.action.id}.${result.error}`,
                params: result.params || {},
                reason: result.error,
            })];
    },
    report(context) {
        return [context.event('if.event.player_revealed', {
                messageId: `${context.action.id}.revealed`,
                reason: 'explicit',
            })];
    },
};
