// @sharpee/plugin-npc/npc-plugin.js
"use strict";
/**
 * NpcPlugin - Wraps NPC service as a TurnPlugin (ADR-070, ADR-120)
 *
 * Priority 100: Runs before scheduler (50) and state machines (75).
 * NPCs act immediately after the player's action.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.NpcPlugin = void 0;
const stdlib_1 = require("../stdlib/index.js");
class NpcPlugin {
    id = 'sharpee.plugin.npc';
    priority = 100;
    service;
    constructor() {
        this.service = (0, stdlib_1.createNpcService)();
        // Register standard NPC behaviors (ADR-070)
        this.service.registerBehavior(stdlib_1.guardBehavior);
        this.service.registerBehavior(stdlib_1.passiveBehavior);
    }
    onAfterAction(ctx) {
        return this.service.tick({
            world: ctx.world,
            turn: ctx.turn,
            random: ctx.random,
            playerLocation: ctx.playerLocation,
            playerId: ctx.playerId
        });
    }
    /** NPC state lives in entity traits/metadata - no plugin-level state needed */
    getState() {
        return {};
    }
    setState(_state) {
        // No-op: NPC state is stored in world model entities
    }
    /** Public access for stories that need to register NPC behaviors */
    getNpcService() {
        return this.service;
    }
}
exports.NpcPlugin = NpcPlugin;
