// @sharpee/world-model/traits/all-traits.js
"use strict";
/**
 * All trait classes in one convenient import
 *
 * Usage:
 * import { Traits } from "../index.js";
 *
 * const openable = entity.get(TraitType.OPENABLE) as Traits.OpenableTrait;
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Traits = void 0;
exports.isActorTrait = isActorTrait;
exports.isContainerTrait = isContainerTrait;
exports.isDoorTrait = isDoorTrait;
exports.isEdibleTrait = isEdibleTrait;
exports.isExitTrait = isExitTrait;
exports.isIdentityTrait = isIdentityTrait;
exports.isLightSourceTrait = isLightSourceTrait;
exports.isLockableTrait = isLockableTrait;
exports.isOpenableTrait = isOpenableTrait;
exports.isReadableTrait = isReadableTrait;
exports.isRoomTrait = isRoomTrait;
exports.isSceneryTrait = isSceneryTrait;
exports.isSupporterTrait = isSupporterTrait;
exports.isSwitchableTrait = isSwitchableTrait;
exports.isWearableTrait = isWearableTrait;
exports.isClothingTrait = isClothingTrait;
exports.isAttachedTrait = isAttachedTrait;
exports.isPullableTrait = isPullableTrait;
exports.isPushableTrait = isPushableTrait;
exports.isButtonTrait = isButtonTrait;
exports.isMoveableSceneryTrait = isMoveableSceneryTrait;
exports.isWeaponTrait = isWeaponTrait;
exports.isBreakableTrait = isBreakableTrait;
exports.isDestructibleTrait = isDestructibleTrait;
exports.isCombatantTrait = isCombatantTrait;
exports.isEquippedTrait = isEquippedTrait;
exports.isCharacterModelTrait = isCharacterModelTrait;
// Import all trait classes
const actorTrait_1 = require("./actor/actorTrait");
const attachedTrait_1 = require("./attached/attachedTrait");
const buttonTrait_1 = require("./button/buttonTrait");
const clothingTrait_1 = require("./clothing/clothingTrait");
const containerTrait_1 = require("./container/containerTrait");
const doorTrait_1 = require("./door/doorTrait");
const climbableTrait_1 = require("./climbable/climbableTrait");
const edibleTrait_1 = require("./edible/edibleTrait");
const exitTrait_1 = require("./exit/exitTrait");
const identityTrait_1 = require("./identity/identityTrait");
const lightSourceTrait_1 = require("./light-source/lightSourceTrait");
const lockableTrait_1 = require("./lockable/lockableTrait");
const moveableSceneryTrait_1 = require("./moveable-scenery/moveableSceneryTrait");
const openableTrait_1 = require("./openable/openableTrait");
const pullableTrait_1 = require("./pullable/pullableTrait");
const pushableTrait_1 = require("./pushable/pushableTrait");
const readableTrait_1 = require("./readable/readableTrait");
const roomTrait_1 = require("./room/roomTrait");
const sceneryTrait_1 = require("./scenery/sceneryTrait");
const supporterTrait_1 = require("./supporter/supporterTrait");
const switchableTrait_1 = require("./switchable/switchableTrait");
const wearableTrait_1 = require("./wearable/wearableTrait");
const weaponTrait_1 = require("./weapon/weaponTrait");
const breakableTrait_1 = require("./breakable/breakableTrait");
const destructibleTrait_1 = require("./destructible/destructibleTrait");
const combatantTrait_1 = require("./combatant/combatantTrait");
const equippedTrait_1 = require("./equipped/equippedTrait");
const enterableTrait_1 = require("./enterable/enterableTrait");
const characterModelTrait_1 = require("./character-model/characterModelTrait");
// Export as namespace for easy access
exports.Traits = {
    ActorTrait: actorTrait_1.ActorTrait,
    AttachedTrait: attachedTrait_1.AttachedTrait,
    ButtonTrait: buttonTrait_1.ButtonTrait,
    ClimbableTrait: climbableTrait_1.ClimbableTrait,
    ClothingTrait: clothingTrait_1.ClothingTrait,
    ContainerTrait: containerTrait_1.ContainerTrait,
    DoorTrait: doorTrait_1.DoorTrait,
    EdibleTrait: edibleTrait_1.EdibleTrait,
    ExitTrait: exitTrait_1.ExitTrait,
    IdentityTrait: identityTrait_1.IdentityTrait,
    LightSourceTrait: lightSourceTrait_1.LightSourceTrait,
    LockableTrait: lockableTrait_1.LockableTrait,
    MoveableSceneryTrait: moveableSceneryTrait_1.MoveableSceneryTrait,
    OpenableTrait: openableTrait_1.OpenableTrait,
    PullableTrait: pullableTrait_1.PullableTrait,
    PushableTrait: pushableTrait_1.PushableTrait,
    ReadableTrait: readableTrait_1.ReadableTrait,
    RoomTrait: roomTrait_1.RoomTrait,
    SceneryTrait: sceneryTrait_1.SceneryTrait,
    SupporterTrait: supporterTrait_1.SupporterTrait,
    SwitchableTrait: switchableTrait_1.SwitchableTrait,
    WearableTrait: wearableTrait_1.WearableTrait,
    WeaponTrait: weaponTrait_1.WeaponTrait,
    BreakableTrait: breakableTrait_1.BreakableTrait,
    DestructibleTrait: destructibleTrait_1.DestructibleTrait,
    CombatantTrait: combatantTrait_1.CombatantTrait,
    EquippedTrait: equippedTrait_1.EquippedTrait,
    EnterableTrait: enterableTrait_1.EnterableTrait,
    CharacterModelTrait: characterModelTrait_1.CharacterModelTrait,
};
// Type guards for each trait
function isActorTrait(trait) {
    return trait?.type === 'actor';
}
function isContainerTrait(trait) {
    return trait?.type === 'container';
}
function isDoorTrait(trait) {
    return trait?.type === 'door';
}
function isEdibleTrait(trait) {
    return trait?.type === 'edible';
}
function isExitTrait(trait) {
    return trait?.type === 'exit';
}
function isIdentityTrait(trait) {
    return trait?.type === 'identity';
}
function isLightSourceTrait(trait) {
    return trait?.type === 'lightSource';
}
function isLockableTrait(trait) {
    return trait?.type === 'lockable';
}
function isOpenableTrait(trait) {
    return trait?.type === 'openable';
}
function isReadableTrait(trait) {
    return trait?.type === 'readable';
}
function isRoomTrait(trait) {
    return trait?.type === 'room';
}
function isSceneryTrait(trait) {
    return trait?.type === 'scenery';
}
function isSupporterTrait(trait) {
    return trait?.type === 'supporter';
}
function isSwitchableTrait(trait) {
    return trait?.type === 'switchable';
}
function isWearableTrait(trait) {
    return trait?.type === 'wearable';
}
function isClothingTrait(trait) {
    return trait?.type === 'clothing';
}
function isAttachedTrait(trait) {
    return trait?.type === 'attached';
}
function isPullableTrait(trait) {
    return trait?.type === 'pullable';
}
function isPushableTrait(trait) {
    return trait?.type === 'pushable';
}
function isButtonTrait(trait) {
    return trait?.type === 'button';
}
function isMoveableSceneryTrait(trait) {
    return trait?.type === 'moveableScenery';
}
function isWeaponTrait(trait) {
    return trait?.type === 'weapon';
}
function isBreakableTrait(trait) {
    return trait?.type === 'breakable';
}
function isDestructibleTrait(trait) {
    return trait?.type === 'destructible';
}
function isCombatantTrait(trait) {
    return trait?.type === 'combatant';
}
function isEquippedTrait(trait) {
    return trait?.type === 'equipped';
}
function isCharacterModelTrait(trait) {
    return trait?.type === 'characterModel';
}
// Note: isEnterableTrait is exported from ./enterable/enterableTrait.ts
