// @sharpee/world-model/traits/implementations.js
"use strict";
/**
 * Trait implementations map
 *
 * Maps trait types to their implementation classes
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.StoryInfoTrait = exports.EnterableTrait = exports.VehicleTrait = exports.CharacterModelTrait = exports.OpenInventoryTrait = exports.NpcTrait = exports.EquippedTrait = exports.CombatantTrait = exports.DestructibleTrait = exports.BreakableTrait = exports.WeaponTrait = exports.MoveableSceneryTrait = exports.ButtonTrait = exports.PushableTrait = exports.AttachedTrait = exports.PullableTrait = exports.ClimbableTrait = exports.ExitTrait = exports.ActorTrait = exports.SceneTrait = exports.RegionTrait = exports.DoorTrait = exports.LightSourceTrait = exports.ReadableTrait = exports.SwitchableTrait = exports.LockableTrait = exports.OpenableTrait = exports.SceneryTrait = exports.EdibleTrait = exports.ClothingTrait = exports.WearableTrait = exports.RoomTrait = exports.SupporterTrait = exports.ContainerTrait = exports.IdentityTrait = exports.TRAIT_IMPLEMENTATIONS = void 0;
exports.getTraitImplementation = getTraitImplementation;
exports.createTrait = createTrait;
const trait_types_1 = require("./trait-types");
// Import all trait implementations from their new locations
const identityTrait_1 = require("./identity/identityTrait");
Object.defineProperty(exports, "IdentityTrait", { enumerable: true, get: function () { return identityTrait_1.IdentityTrait; } });
const containerTrait_1 = require("./container/containerTrait");
Object.defineProperty(exports, "ContainerTrait", { enumerable: true, get: function () { return containerTrait_1.ContainerTrait; } });
const supporterTrait_1 = require("./supporter/supporterTrait");
Object.defineProperty(exports, "SupporterTrait", { enumerable: true, get: function () { return supporterTrait_1.SupporterTrait; } });
const roomTrait_1 = require("./room/roomTrait");
Object.defineProperty(exports, "RoomTrait", { enumerable: true, get: function () { return roomTrait_1.RoomTrait; } });
const wearableTrait_1 = require("./wearable/wearableTrait");
Object.defineProperty(exports, "WearableTrait", { enumerable: true, get: function () { return wearableTrait_1.WearableTrait; } });
const clothingTrait_1 = require("./clothing/clothingTrait");
Object.defineProperty(exports, "ClothingTrait", { enumerable: true, get: function () { return clothingTrait_1.ClothingTrait; } });
const edibleTrait_1 = require("./edible/edibleTrait");
Object.defineProperty(exports, "EdibleTrait", { enumerable: true, get: function () { return edibleTrait_1.EdibleTrait; } });
const sceneryTrait_1 = require("./scenery/sceneryTrait");
Object.defineProperty(exports, "SceneryTrait", { enumerable: true, get: function () { return sceneryTrait_1.SceneryTrait; } });
const openableTrait_1 = require("./openable/openableTrait");
Object.defineProperty(exports, "OpenableTrait", { enumerable: true, get: function () { return openableTrait_1.OpenableTrait; } });
const lockableTrait_1 = require("./lockable/lockableTrait");
Object.defineProperty(exports, "LockableTrait", { enumerable: true, get: function () { return lockableTrait_1.LockableTrait; } });
const switchableTrait_1 = require("./switchable/switchableTrait");
Object.defineProperty(exports, "SwitchableTrait", { enumerable: true, get: function () { return switchableTrait_1.SwitchableTrait; } });
const readableTrait_1 = require("./readable/readableTrait");
Object.defineProperty(exports, "ReadableTrait", { enumerable: true, get: function () { return readableTrait_1.ReadableTrait; } });
const lightSourceTrait_1 = require("./light-source/lightSourceTrait");
Object.defineProperty(exports, "LightSourceTrait", { enumerable: true, get: function () { return lightSourceTrait_1.LightSourceTrait; } });
// Spatial traits
const doorTrait_1 = require("./door/doorTrait");
Object.defineProperty(exports, "DoorTrait", { enumerable: true, get: function () { return doorTrait_1.DoorTrait; } });
const climbableTrait_1 = require("./climbable/climbableTrait");
Object.defineProperty(exports, "ClimbableTrait", { enumerable: true, get: function () { return climbableTrait_1.ClimbableTrait; } });
const regionTrait_1 = require("./region/regionTrait");
Object.defineProperty(exports, "RegionTrait", { enumerable: true, get: function () { return regionTrait_1.RegionTrait; } });
const sceneTrait_1 = require("./scene/sceneTrait");
Object.defineProperty(exports, "SceneTrait", { enumerable: true, get: function () { return sceneTrait_1.SceneTrait; } });
// Basic traits
const actorTrait_1 = require("./actor/actorTrait");
Object.defineProperty(exports, "ActorTrait", { enumerable: true, get: function () { return actorTrait_1.ActorTrait; } });
// New traits
const exitTrait_1 = require("./exit/exitTrait");
Object.defineProperty(exports, "ExitTrait", { enumerable: true, get: function () { return exitTrait_1.ExitTrait; } });
// Manipulation traits
const pullableTrait_1 = require("./pullable/pullableTrait");
Object.defineProperty(exports, "PullableTrait", { enumerable: true, get: function () { return pullableTrait_1.PullableTrait; } });
const attachedTrait_1 = require("./attached/attachedTrait");
Object.defineProperty(exports, "AttachedTrait", { enumerable: true, get: function () { return attachedTrait_1.AttachedTrait; } });
const pushableTrait_1 = require("./pushable/pushableTrait");
Object.defineProperty(exports, "PushableTrait", { enumerable: true, get: function () { return pushableTrait_1.PushableTrait; } });
const buttonTrait_1 = require("./button/buttonTrait");
Object.defineProperty(exports, "ButtonTrait", { enumerable: true, get: function () { return buttonTrait_1.ButtonTrait; } });
const moveableSceneryTrait_1 = require("./moveable-scenery/moveableSceneryTrait");
Object.defineProperty(exports, "MoveableSceneryTrait", { enumerable: true, get: function () { return moveableSceneryTrait_1.MoveableSceneryTrait; } });
// Combat traits
const weaponTrait_1 = require("./weapon/weaponTrait");
Object.defineProperty(exports, "WeaponTrait", { enumerable: true, get: function () { return weaponTrait_1.WeaponTrait; } });
const breakableTrait_1 = require("./breakable/breakableTrait");
Object.defineProperty(exports, "BreakableTrait", { enumerable: true, get: function () { return breakableTrait_1.BreakableTrait; } });
const destructibleTrait_1 = require("./destructible/destructibleTrait");
Object.defineProperty(exports, "DestructibleTrait", { enumerable: true, get: function () { return destructibleTrait_1.DestructibleTrait; } });
const combatantTrait_1 = require("./combatant/combatantTrait");
Object.defineProperty(exports, "CombatantTrait", { enumerable: true, get: function () { return combatantTrait_1.CombatantTrait; } });
const equippedTrait_1 = require("./equipped/equippedTrait");
Object.defineProperty(exports, "EquippedTrait", { enumerable: true, get: function () { return equippedTrait_1.EquippedTrait; } });
// NPC traits (ADR-070)
const npcTrait_1 = require("./npc/npcTrait");
Object.defineProperty(exports, "NpcTrait", { enumerable: true, get: function () { return npcTrait_1.NpcTrait; } });
const openInventoryTrait_1 = require("./open-inventory/openInventoryTrait");
Object.defineProperty(exports, "OpenInventoryTrait", { enumerable: true, get: function () { return openInventoryTrait_1.OpenInventoryTrait; } });
// Character model (ADR-141)
const characterModelTrait_1 = require("./character-model/characterModelTrait");
Object.defineProperty(exports, "CharacterModelTrait", { enumerable: true, get: function () { return characterModelTrait_1.CharacterModelTrait; } });
// Transport traits
const vehicleTrait_1 = require("./vehicle/vehicleTrait");
Object.defineProperty(exports, "VehicleTrait", { enumerable: true, get: function () { return vehicleTrait_1.VehicleTrait; } });
const enterableTrait_1 = require("./enterable/enterableTrait");
Object.defineProperty(exports, "EnterableTrait", { enumerable: true, get: function () { return enterableTrait_1.EnterableTrait; } });
// Concealment traits (ADR-148)
const concealmentTrait_1 = require("./concealment/concealmentTrait");
const concealedStateTrait_1 = require("./concealment/concealedStateTrait");
// System traits
const storyInfoTrait_1 = require("./story-info/storyInfoTrait");
Object.defineProperty(exports, "StoryInfoTrait", { enumerable: true, get: function () { return storyInfoTrait_1.StoryInfoTrait; } });
/**
 * Map of trait types to their constructors
 */
exports.TRAIT_IMPLEMENTATIONS = {
    // Core traits (organized in their own folders)
    [trait_types_1.TraitType.IDENTITY]: identityTrait_1.IdentityTrait,
    [trait_types_1.TraitType.CONTAINER]: containerTrait_1.ContainerTrait,
    [trait_types_1.TraitType.SUPPORTER]: supporterTrait_1.SupporterTrait,
    [trait_types_1.TraitType.ROOM]: roomTrait_1.RoomTrait,
    [trait_types_1.TraitType.WEARABLE]: wearableTrait_1.WearableTrait,
    [trait_types_1.TraitType.CLOTHING]: clothingTrait_1.ClothingTrait,
    [trait_types_1.TraitType.EDIBLE]: edibleTrait_1.EdibleTrait,
    [trait_types_1.TraitType.SCENERY]: sceneryTrait_1.SceneryTrait,
    // Interactive traits
    [trait_types_1.TraitType.OPENABLE]: openableTrait_1.OpenableTrait,
    [trait_types_1.TraitType.LOCKABLE]: lockableTrait_1.LockableTrait,
    [trait_types_1.TraitType.SWITCHABLE]: switchableTrait_1.SwitchableTrait,
    [trait_types_1.TraitType.READABLE]: readableTrait_1.ReadableTrait,
    [trait_types_1.TraitType.LIGHT_SOURCE]: lightSourceTrait_1.LightSourceTrait,
    // Spatial traits
    [trait_types_1.TraitType.DOOR]: doorTrait_1.DoorTrait,
    [trait_types_1.TraitType.CLIMBABLE]: climbableTrait_1.ClimbableTrait,
    [trait_types_1.TraitType.REGION]: regionTrait_1.RegionTrait,
    [trait_types_1.TraitType.SCENE]: sceneTrait_1.SceneTrait,
    // Basic traits
    [trait_types_1.TraitType.ACTOR]: actorTrait_1.ActorTrait,
    // New traits
    [trait_types_1.TraitType.EXIT]: exitTrait_1.ExitTrait,
    // Manipulation traits
    [trait_types_1.TraitType.PULLABLE]: pullableTrait_1.PullableTrait,
    [trait_types_1.TraitType.ATTACHED]: attachedTrait_1.AttachedTrait,
    [trait_types_1.TraitType.PUSHABLE]: pushableTrait_1.PushableTrait,
    [trait_types_1.TraitType.BUTTON]: buttonTrait_1.ButtonTrait,
    [trait_types_1.TraitType.MOVEABLE_SCENERY]: moveableSceneryTrait_1.MoveableSceneryTrait,
    // Combat traits
    [trait_types_1.TraitType.WEAPON]: weaponTrait_1.WeaponTrait,
    [trait_types_1.TraitType.BREAKABLE]: breakableTrait_1.BreakableTrait,
    [trait_types_1.TraitType.DESTRUCTIBLE]: destructibleTrait_1.DestructibleTrait,
    [trait_types_1.TraitType.COMBATANT]: combatantTrait_1.CombatantTrait,
    [trait_types_1.TraitType.EQUIPPED]: equippedTrait_1.EquippedTrait,
    // NPC traits (ADR-070)
    [trait_types_1.TraitType.NPC]: npcTrait_1.NpcTrait,
    [trait_types_1.TraitType.OPEN_INVENTORY]: openInventoryTrait_1.OpenInventoryTrait,
    // Character model (ADR-141)
    [trait_types_1.TraitType.CHARACTER_MODEL]: characterModelTrait_1.CharacterModelTrait,
    // Transport traits
    [trait_types_1.TraitType.VEHICLE]: vehicleTrait_1.VehicleTrait,
    [trait_types_1.TraitType.ENTERABLE]: enterableTrait_1.EnterableTrait,
    // Concealment traits (ADR-148)
    [trait_types_1.TraitType.CONCEALMENT]: concealmentTrait_1.ConcealmentTrait,
    [trait_types_1.TraitType.CONCEALED_STATE]: concealedStateTrait_1.ConcealedStateTrait,
    // System traits
    [trait_types_1.TraitType.STORY_INFO]: storyInfoTrait_1.StoryInfoTrait,
    // Object property traits
    // Deprecated traits - these are handled differently now:
    // LOCATION - tracked by world model internally
    // PORTABLE - objects are takeable by default unless they have SceneryTrait
    // FIXED - replaced by SCENERY
};
/**
 * Get trait implementation by type
 */
function getTraitImplementation(type) {
    return exports.TRAIT_IMPLEMENTATIONS[type];
}
/**
 * Create trait instance by type
 */
function createTrait(type, data) {
    const TraitClass = exports.TRAIT_IMPLEMENTATIONS[type];
    if (!TraitClass) {
        throw new Error(`Unknown trait type: ${type}`);
    }
    return new TraitClass(data);
}
