// @sharpee/world-model/traits/trait-types.js
"use strict";
/**
 * Centralized trait type definitions
 *
 * All trait types used in the world model system
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TRAIT_CATEGORIES = exports.TraitCategory = exports.TraitType = void 0;
exports.isValidTraitType = isValidTraitType;
exports.getTraitCategory = getTraitCategory;
exports.getTraitsByCategory = getTraitsByCategory;
exports.getAllTraitTypes = getAllTraitTypes;
exports.registerTraitType = registerTraitType;
/**
 * Trait types as a const object for extensibility
 */
exports.TraitType = {
    // Standard traits
    IDENTITY: 'identity',
    CONTAINER: 'container',
    SUPPORTER: 'supporter',
    ROOM: 'room',
    WEARABLE: 'wearable',
    CLOTHING: 'clothing',
    EDIBLE: 'edible',
    SCENERY: 'scenery',
    // Interactive traits
    OPENABLE: 'openable',
    LOCKABLE: 'lockable',
    SWITCHABLE: 'switchable',
    READABLE: 'readable',
    LIGHT_SOURCE: 'lightSource',
    // Manipulation traits
    PULLABLE: 'pullable',
    ATTACHED: 'attached',
    PUSHABLE: 'pushable',
    BUTTON: 'button',
    MOVEABLE_SCENERY: 'moveableScenery',
    // Spatial traits
    DOOR: 'door',
    CLIMBABLE: 'climbable',
    REGION: 'region',
    SCENE: 'scene',
    // Object property traits
    // Basic traits
    ACTOR: 'actor',
    // New traits
    EXIT: 'exit',
    // Combat traits
    WEAPON: 'weapon',
    BREAKABLE: 'breakable',
    DESTRUCTIBLE: 'destructible',
    COMBATANT: 'combatant',
    EQUIPPED: 'equipped',
    // NPC traits (ADR-070)
    NPC: 'npc',
    OPEN_INVENTORY: 'openInventory',
    // Character model (ADR-141)
    CHARACTER_MODEL: 'characterModel',
    // Transport traits
    VEHICLE: 'vehicle',
    ENTERABLE: 'enterable',
    // Concealment traits (ADR-148)
    CONCEALMENT: 'if.trait.concealment',
    CONCEALED_STATE: 'if.trait.concealed_state',
    // System traits
    STORY_INFO: 'storyInfo'
};
/**
 * Trait categories for organization
 */
exports.TraitCategory = {
    STANDARD: 'standard',
    INTERACTIVE: 'interactive',
    ADVANCED: 'advanced'
};
/**
 * Map trait types to categories
 */
exports.TRAIT_CATEGORIES = {
    // Standard
    [exports.TraitType.IDENTITY]: exports.TraitCategory.STANDARD,
    [exports.TraitType.CONTAINER]: exports.TraitCategory.STANDARD,
    [exports.TraitType.SUPPORTER]: exports.TraitCategory.STANDARD,
    [exports.TraitType.ROOM]: exports.TraitCategory.STANDARD,
    [exports.TraitType.WEARABLE]: exports.TraitCategory.STANDARD,
    [exports.TraitType.CLOTHING]: exports.TraitCategory.STANDARD,
    [exports.TraitType.EDIBLE]: exports.TraitCategory.STANDARD,
    [exports.TraitType.SCENERY]: exports.TraitCategory.STANDARD,
    // Interactive
    [exports.TraitType.OPENABLE]: exports.TraitCategory.INTERACTIVE,
    [exports.TraitType.LOCKABLE]: exports.TraitCategory.INTERACTIVE,
    [exports.TraitType.SWITCHABLE]: exports.TraitCategory.INTERACTIVE,
    [exports.TraitType.READABLE]: exports.TraitCategory.INTERACTIVE,
    [exports.TraitType.LIGHT_SOURCE]: exports.TraitCategory.INTERACTIVE,
    // Manipulation
    [exports.TraitType.PULLABLE]: exports.TraitCategory.INTERACTIVE,
    [exports.TraitType.ATTACHED]: exports.TraitCategory.INTERACTIVE,
    [exports.TraitType.PUSHABLE]: exports.TraitCategory.INTERACTIVE,
    [exports.TraitType.BUTTON]: exports.TraitCategory.INTERACTIVE,
    [exports.TraitType.MOVEABLE_SCENERY]: exports.TraitCategory.INTERACTIVE,
    // Spatial
    [exports.TraitType.DOOR]: exports.TraitCategory.STANDARD,
    [exports.TraitType.CLIMBABLE]: exports.TraitCategory.INTERACTIVE,
    [exports.TraitType.EXIT]: exports.TraitCategory.STANDARD,
    [exports.TraitType.REGION]: exports.TraitCategory.STANDARD,
    [exports.TraitType.SCENE]: exports.TraitCategory.STANDARD,
    // Object property
    // Basic
    [exports.TraitType.ACTOR]: exports.TraitCategory.STANDARD,
    // Combat
    [exports.TraitType.WEAPON]: exports.TraitCategory.INTERACTIVE,
    [exports.TraitType.BREAKABLE]: exports.TraitCategory.INTERACTIVE,
    [exports.TraitType.DESTRUCTIBLE]: exports.TraitCategory.INTERACTIVE,
    [exports.TraitType.COMBATANT]: exports.TraitCategory.INTERACTIVE,
    [exports.TraitType.EQUIPPED]: exports.TraitCategory.INTERACTIVE,
    // NPC traits
    [exports.TraitType.NPC]: exports.TraitCategory.STANDARD,
    [exports.TraitType.OPEN_INVENTORY]: exports.TraitCategory.STANDARD,
    // Character model (ADR-141)
    [exports.TraitType.CHARACTER_MODEL]: exports.TraitCategory.ADVANCED,
    // Transport traits
    [exports.TraitType.VEHICLE]: exports.TraitCategory.INTERACTIVE,
    [exports.TraitType.ENTERABLE]: exports.TraitCategory.INTERACTIVE,
    // Concealment traits (ADR-148)
    [exports.TraitType.CONCEALMENT]: exports.TraitCategory.INTERACTIVE,
    [exports.TraitType.CONCEALED_STATE]: exports.TraitCategory.INTERACTIVE,
    // System traits
    [exports.TraitType.STORY_INFO]: exports.TraitCategory.STANDARD
};
/**
 * Helper to check if a trait type exists
 */
function isValidTraitType(type) {
    return Object.values(exports.TraitType).includes(type);
}
/**
 * Get trait category
 */
function getTraitCategory(type) {
    return exports.TRAIT_CATEGORIES[type];
}
/**
 * Get all trait types in a category
 */
function getTraitsByCategory(category) {
    return Object.entries(exports.TRAIT_CATEGORIES)
        .filter(([_, cat]) => cat === category)
        .map(([type, _]) => type);
}
/**
 * Get all trait types
 */
function getAllTraitTypes() {
    return Object.values(exports.TraitType);
}
/**
 * Add a new trait type at runtime (for extensions)
 */
function registerTraitType(name, value, category = exports.TraitCategory.STANDARD) {
    // as any: justified — runtime extension of const enum object
    exports.TraitType[name] = value;
    // Add to categories
    exports.TRAIT_CATEGORIES[value] = category;
}
