// @sharpee/world-model/traits/trait.js
"use strict";
/**
 * Base trait types and utilities
 *
 * Traits are pure data structures with no behavior.
 * All logic belongs in behaviors.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TraitRegistry = void 0;
exports.isTrait = isTrait;
/**
 * Type guard to check if an object is a trait
 */
function isTrait(obj) {
    return obj && typeof obj === 'object' && 'type' in obj && typeof obj.type === 'string';
}
/**
 * Registry for trait constructors (optional use)
 */
class TraitRegistry {
    static traits = new Map();
    static register(trait) {
        if (this.traits.has(trait.type)) {
            throw new Error(`Trait type '${trait.type}' is already registered`);
        }
        this.traits.set(trait.type, trait);
    }
    static get(type) {
        return this.traits.get(type);
    }
    static has(type) {
        return this.traits.has(type);
    }
    static clear() {
        this.traits.clear();
    }
    static getAll() {
        return new Map(this.traits);
    }
}
exports.TraitRegistry = TraitRegistry;
