// @sharpee/stdlib/actions/standard/listening/listening.js
"use strict";
/**
 * Listening action - listen for sounds in the environment
 *
 * This action allows players to listen for sounds in their current location
 * or from specific objects.
 *
 * Uses four-phase pattern:
 * 1. validate: Always succeeds (no preconditions for listening)
 * 2. execute: Analyze sounds (no world mutations)
 * 3. blocked: Handle validation failures (n/a - always succeeds)
 * 4. report: Emit listened event and success message
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.listeningAction = void 0;
const world_model_1 = require("../../../../world-model/index.js");
const constants_1 = require("../../constants");
const scope_1 = require("../../../scope");
function getListeningSharedData(context) {
    return context.sharedData;
}
/**
 * Analyzes what sounds can be heard from a target or environment
 * Shared logic between validate and execute phases
 */
function analyzeListening(context) {
    const target = context.command.directObject?.entity;
    const eventData = {};
    const params = {};
    let messageId;
    if (target) {
        eventData.target = target.id;
        params.target = target.name;
        // Check if the target has any sound-related properties
        let hasSound = false;
        if (target.has(world_model_1.TraitType.SWITCHABLE)) {
            const switchableTrait = target.get(world_model_1.TraitType.SWITCHABLE);
            if (switchableTrait.isOn) {
                hasSound = true;
                eventData.hasSound = true;
                eventData.soundType = 'device';
                messageId = 'device_running';
            }
            else {
                messageId = 'device_off';
            }
        }
        else if (target.has(world_model_1.TraitType.CONTAINER)) {
            const contents = context.world.getContents(target.id);
            if (contents.length > 0) {
                eventData.hasContents = true;
                eventData.contentCount = contents.length;
                // Check if any contents are liquid
                const hasLiquid = contents.some(item => {
                    if (item.has(world_model_1.TraitType.EDIBLE)) {
                        const edibleTrait = item.get(world_model_1.TraitType.EDIBLE);
                        return edibleTrait.liquid;
                    }
                    return false;
                });
                messageId = hasLiquid ? 'liquid_sounds' : 'container_sounds';
                hasSound = true;
            }
            else {
                messageId = 'no_sound';
            }
        }
        else {
            messageId = 'no_sound';
        }
        // Override with generic message if no specific sound
        if (!hasSound && !messageId) {
            messageId = 'listened_to';
        }
    }
    else {
        // Listening to the general environment
        eventData.listeningToEnvironment = true;
        // Check for any active sound sources in the location
        const location = context.currentLocation;
        const contents = context.world.getContents(location.id);
        const soundSources = contents.filter(item => {
            if (item.has(world_model_1.TraitType.SWITCHABLE)) {
                const switchableTrait = item.get(world_model_1.TraitType.SWITCHABLE);
                return switchableTrait.isOn;
            }
            return false;
        });
        if (soundSources.length > 0) {
            eventData.soundSources = soundSources.map(s => s.id);
            params.devices = soundSources.map(s => s.name).join(', ');
            messageId = 'active_devices';
        }
        else {
            messageId = 'silence';
        }
        eventData.roomId = location.id;
    }
    return {
        messageId: messageId,
        params,
        eventData
    };
}
exports.listeningAction = {
    id: constants_1.IFActions.LISTENING,
    requiredMessages: [
        'not_visible',
        'silence',
        'ambient_sounds',
        'active_devices',
        'no_sound',
        'device_running',
        'device_off',
        'container_sounds',
        'liquid_sounds',
        'listened_to',
        'listened_environment'
    ],
    group: "sensory",
    metadata: {
        requiresDirectObject: false,
        requiresIndirectObject: false,
        directObjectScope: scope_1.ScopeLevel.AWARE
    },
    validate(context) {
        // Listening always succeeds - no preconditions
        return { valid: true };
    },
    execute(context) {
        // Listening has NO world mutations - it's a sensory action
        // Analyze sounds and store in sharedData for report phase
        const analysis = analyzeListening(context);
        const sharedData = getListeningSharedData(context);
        sharedData.messageId = analysis.messageId;
        sharedData.params = analysis.params;
        sharedData.eventData = analysis.eventData;
    },
    blocked(context, result) {
        // Listening always succeeds, but include blocked for consistency
        const target = context.command.directObject?.entity;
        return [context.event('if.event.listen_blocked', {
                blocked: true,
                messageId: `${context.action.id}.${result.error}`,
                params: {
                    ...result.params,
                    target: target?.name
                },
                reason: result.error,
                targetId: target?.id,
                targetName: target?.name
            })];
    },
    report(context) {
        const events = [];
        const sharedData = getListeningSharedData(context);
        // Emit listened event with messageId for text rendering
        events.push(context.event('if.event.listened', {
            messageId: `${context.action.id}.${sharedData.messageId || 'silence'}`,
            params: sharedData.params,
            ...sharedData.eventData
        }));
        return events;
    }
};
