// @sharpee/world-model/capabilities/capability-registry.js
"use strict";
/**
 * Capability Registry (ADR-090)
 *
 * Registry for trait-behavior bindings. Stories register behaviors
 * for their traits, and the dispatch system looks them up at runtime.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.registerCapabilityBehavior = registerCapabilityBehavior;
exports.getBehaviorForCapability = getBehaviorForCapability;
exports.getBehaviorBinding = getBehaviorBinding;
exports.hasCapabilityBehavior = hasCapabilityBehavior;
exports.unregisterCapabilityBehavior = unregisterCapabilityBehavior;
exports.clearCapabilityRegistry = clearCapabilityRegistry;
exports.getAllCapabilityBindings = getAllCapabilityBindings;
/**
 * Registry storage: key = `${traitType}:${capability}`
 * Uses globalThis to ensure the registry is shared across module boundaries
 * (e.g., when story modules import from packages but actions run from bundle)
 */
const CAPABILITY_REGISTRY_KEY = '__sharpee_capability_behaviors__';
function getCapabilityRegistry() {
    const global = globalThis;
    if (!global[CAPABILITY_REGISTRY_KEY]) {
        global[CAPABILITY_REGISTRY_KEY] = new Map();
    }
    return global[CAPABILITY_REGISTRY_KEY];
}
const behaviorRegistry = {
    get size() { return getCapabilityRegistry().size; },
    has(key) { return getCapabilityRegistry().has(key); },
    get(key) { return getCapabilityRegistry().get(key); },
    set(key, value) { return getCapabilityRegistry().set(key, value); },
    delete(key) { return getCapabilityRegistry().delete(key); },
    clear() { return getCapabilityRegistry().clear(); },
    keys() { return getCapabilityRegistry().keys(); },
    values() { return getCapabilityRegistry().values(); },
    entries() { return getCapabilityRegistry().entries(); },
    [Symbol.iterator]() { return getCapabilityRegistry()[Symbol.iterator](); }
};
/**
 * Generate registry key for trait+capability combination.
 */
function registryKey(traitType, capability) {
    return `${traitType}:${capability}`;
}
/**
 * Register a behavior for a trait+capability combination.
 *
 * Stories call this to register their behaviors during initialization.
 * Each trait+capability pair can only have one behavior registered.
 *
 * @param traitType - The trait type identifier (e.g., 'dungeo.trait.basket_elevator')
 * @param capability - The action ID (e.g., 'if.action.lowering')
 * @param behavior - The behavior implementation
 * @param options - Optional configuration (priority, resolution override, mode override)
 *
 * @example
 * ```typescript
 * // Basic registration (uses global defaults)
 * registerCapabilityBehavior(
 *   BasketElevatorTrait.type,
 *   'if.action.lowering',
 *   BasketLoweringBehavior
 * );
 *
 * // With priority and overrides
 * registerCapabilityBehavior(
 *   TrollAxeTrait.type,
 *   'if.action.taking',
 *   TrollAxeTakingBehavior,
 *   { priority: 100, resolution: 'any-blocks' }
 * );
 * ```
 */
function registerCapabilityBehavior(traitType, capability, behavior, options) {
    const key = registryKey(traitType, capability);
    if (behaviorRegistry.has(key)) {
        throw new Error(`Behavior already registered for trait "${traitType}" and capability "${capability}"`);
    }
    behaviorRegistry.set(key, {
        traitType,
        capability,
        behavior,
        priority: options?.priority ?? 0,
        resolution: options?.resolution,
        mode: options?.mode,
        validateBinding: options?.validateBinding
    });
}
/**
 * Get behavior for a trait instance and capability.
 *
 * Called by capability-dispatch actions to find the right behavior.
 *
 * @param trait - The trait instance
 * @param capability - The action ID
 * @returns The behavior, or undefined if not registered
 */
function getBehaviorForCapability(trait, capability) {
    const binding = getBehaviorBinding(trait, capability);
    return binding?.behavior;
}
/**
 * Get the full binding for a trait instance and capability.
 *
 * Includes priority and configuration overrides. Used by dispatch helper
 * for resolution logic.
 *
 * @param trait - The trait instance
 * @param capability - The action ID
 * @returns The binding, or undefined if not registered
 */
function getBehaviorBinding(trait, capability) {
    const traitType = trait.constructor.type;
    const key = registryKey(traitType, capability);
    const binding = behaviorRegistry.get(key);
    if (!binding) {
        return undefined;
    }
    // Runtime validation if provided
    if (binding.validateBinding && !binding.validateBinding(trait)) {
        throw new Error(`Behavior validation failed for trait "${traitType}", capability "${capability}"`);
    }
    return binding;
}
/**
 * Check if a behavior is registered for a trait+capability.
 *
 * @param traitType - The trait type identifier
 * @param capability - The action ID
 */
function hasCapabilityBehavior(traitType, capability) {
    return behaviorRegistry.has(registryKey(traitType, capability));
}
/**
 * Unregister a behavior (primarily for testing).
 *
 * @param traitType - The trait type identifier
 * @param capability - The action ID
 */
function unregisterCapabilityBehavior(traitType, capability) {
    behaviorRegistry.delete(registryKey(traitType, capability));
}
/**
 * Clear all registered behaviors (for testing).
 */
function clearCapabilityRegistry() {
    behaviorRegistry.clear();
}
/**
 * Get all registered bindings (for debugging/introspection).
 */
function getAllCapabilityBindings() {
    return new Map(getCapabilityRegistry());
}
