// @sharpee/world-model/capabilities/capability-defaults.js
"use strict";
/**
 * Capability Defaults (ADR-090 Extension)
 *
 * Global configuration for how capabilities are resolved when multiple
 * entities claim the same capability. Stories can override per-registration.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.defineCapabilityDefaults = defineCapabilityDefaults;
exports.getCapabilityConfig = getCapabilityConfig;
exports.hasCapabilityDefaults = hasCapabilityDefaults;
exports.clearCapabilityDefaults = clearCapabilityDefaults;
exports.getAllCapabilityDefaults = getAllCapabilityDefaults;
/**
 * Default configuration used when no specific config is defined.
 */
const DEFAULT_CONFIG = {
    resolution: 'first-wins',
    mode: 'blocking'
};
/**
 * Registry of capability configurations.
 */
const capabilityDefaults = new Map();
/**
 * Define default configuration for a capability.
 *
 * Called during platform or story initialization to set how a capability
 * should behave when multiple entities claim it.
 *
 * @param capabilityId - The capability ID (e.g., 'if.scope.visible', 'if.action.taking')
 * @param config - Partial config, missing values use defaults
 *
 * @example
 * ```typescript
 * // Visibility should block if ANY entity says it's hidden
 * defineCapabilityDefaults('if.scope.visible', {
 *   resolution: 'any-blocks',
 *   mode: 'blocking'
 * });
 * ```
 */
function defineCapabilityDefaults(capabilityId, config) {
    capabilityDefaults.set(capabilityId, {
        resolution: config.resolution ?? DEFAULT_CONFIG.resolution,
        mode: config.mode ?? DEFAULT_CONFIG.mode
    });
}
/**
 * Get configuration for a capability.
 *
 * Returns the registered config or defaults if not registered.
 *
 * @param capabilityId - The capability ID
 * @returns The capability configuration
 */
function getCapabilityConfig(capabilityId) {
    return capabilityDefaults.get(capabilityId) ?? { ...DEFAULT_CONFIG };
}
/**
 * Check if a capability has explicit defaults defined.
 *
 * @param capabilityId - The capability ID
 * @returns true if defaults have been defined
 */
function hasCapabilityDefaults(capabilityId) {
    return capabilityDefaults.has(capabilityId);
}
/**
 * Clear all capability defaults (for testing).
 */
function clearCapabilityDefaults() {
    capabilityDefaults.clear();
}
/**
 * Get all defined capability defaults (for debugging/introspection).
 */
function getAllCapabilityDefaults() {
    return new Map(capabilityDefaults);
}
