// @sharpee/world-model/capabilities/entity-builder.js
"use strict";
/**
 * Type-Safe Entity Builder (ADR-090)
 *
 * Builder pattern for creating entities with compile-time capability
 * conflict detection. When two traits claim the same capability,
 * TypeScript will error (and runtime will throw as backup).
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.EntityBuilder = void 0;
exports.buildEntity = buildEntity;
/**
 * Type-safe entity builder that tracks claimed capabilities.
 *
 * Provides runtime checks for capability conflicts. TypeScript's
 * type system can catch conflicts at compile time when traits
 * declare capabilities with `as const`.
 *
 * @example
 * ```typescript
 * const basket = new EntityBuilder(entity)
 *   .add(new IdentityTrait({ name: 'basket' }))
 *   .add(new ContainerTrait())
 *   .add(new BasketElevatorTrait({ ... }))  // claims: lowering, raising
 *   .build();
 *
 * // Runtime error - duplicate capability:
 * const bad = new EntityBuilder(entity)
 *   .add(new BasketElevatorTrait({ ... }))  // claims: lowering
 *   .add(new MirrorPoleTrait({ ... }))      // also claims: lowering
 *   .build();  // throws!
 * ```
 */
class EntityBuilder {
    entity;
    claimedCaps = new Set();
    constructor(entity) {
        this.entity = entity;
    }
    /**
     * Add a trait to the entity.
     *
     * Checks for capability conflicts at runtime. If two traits
     * claim the same capability, throws an error.
     *
     * @param trait - The trait to add
     * @returns This builder for chaining
     * @throws Error if capability conflict detected
     */
    add(trait) {
        const traitClass = trait.constructor;
        const newCaps = traitClass.capabilities ?? [];
        // Check for capability conflicts
        for (const cap of newCaps) {
            if (this.claimedCaps.has(cap)) {
                // Find which trait already claimed it
                const existingTrait = this.findTraitWithCapability(cap);
                const existingType = existingTrait
                    ? existingTrait.constructor.type
                    : 'unknown';
                throw new Error(`Entity "${this.entity.id}": capability "${cap}" already claimed by trait "${existingType}". ` +
                    `New trait "${traitClass.type}" cannot also claim it.`);
            }
            this.claimedCaps.add(cap);
        }
        this.entity.add(trait);
        return this;
    }
    /**
     * Build and return the entity.
     */
    build() {
        return this.entity;
    }
    /**
     * Get all capabilities claimed so far.
     */
    getClaimedCapabilities() {
        return Array.from(this.claimedCaps);
    }
    /**
     * Find which trait claimed a capability.
     */
    findTraitWithCapability(capability) {
        for (const trait of this.entity.traits.values()) {
            const traitClass = trait.constructor;
            if (traitClass.capabilities?.includes(capability)) {
                return trait;
            }
        }
        return undefined;
    }
}
exports.EntityBuilder = EntityBuilder;
/**
 * Create an entity builder for type-safe trait composition.
 *
 * This is a convenience function that wraps an existing entity
 * in a builder. Use when you want capability conflict checking.
 *
 * @param entity - The entity to wrap
 * @returns A new EntityBuilder
 *
 * @example
 * ```typescript
 * const basket = buildEntity(world.createEntity('basket', EntityType.CONTAINER))
 *   .add(new IdentityTrait({ name: 'rusty iron basket' }))
 *   .add(new ContainerTrait({ capacity: { maxItems: 10 } }))
 *   .add(new BasketElevatorTrait({ ... }))
 *   .build();
 * ```
 */
function buildEntity(entity) {
    return new EntityBuilder(entity);
}
