// @sharpee/world-model/capabilities/interceptor-helpers.js
"use strict";
/**
 * Interceptor Helpers (ADR-118)
 *
 * Helper functions for working with trait interceptors.
 * Mirrors capability-helpers.ts pattern but for interceptors.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.findTraitWithInterceptor = findTraitWithInterceptor;
exports.hasInterceptor = hasInterceptor;
exports.getEntityInterceptors = getEntityInterceptors;
exports.traitHasInterceptor = traitHasInterceptor;
exports.getInterceptorTraits = getInterceptorTraits;
/**
 * Find a trait on the entity that has an interceptor registered for the given action.
 *
 * Note: This checks both the static `interceptors` declaration on the trait class
 * AND the registry. Use this when you need the trait instance.
 *
 * @param entity - The entity to search
 * @param actionId - The action ID to find
 * @returns The trait that has an interceptor registered, or undefined
 *
 * @example
 * ```typescript
 * const trait = findTraitWithInterceptor(boat, 'if.action.entering');
 * if (trait) {
 *   // Entity has an interceptor for entering
 * }
 * ```
 */
function findTraitWithInterceptor(entity, actionId) {
    for (const trait of entity.traits.values()) {
        // First check static declaration (if available)
        const traitClass = trait.constructor;
        if (traitClass.interceptors?.includes(actionId)) {
            return trait;
        }
    }
    return undefined;
}
/**
 * Check if an entity has any trait declaring an interceptor for an action.
 *
 * @param entity - The entity to check
 * @param actionId - The action ID to check for
 * @returns True if entity has a trait with this interceptor
 *
 * @example
 * ```typescript
 * if (hasInterceptor(entity, 'if.action.entering')) {
 *   // Entity can intercept entering actions
 * }
 * ```
 */
function hasInterceptor(entity, actionId) {
    return findTraitWithInterceptor(entity, actionId) !== undefined;
}
/**
 * Get all interceptor action IDs declared by an entity's traits.
 *
 * @param entity - The entity to inspect
 * @returns Array of action IDs the entity intercepts
 */
function getEntityInterceptors(entity) {
    const interceptors = [];
    for (const trait of entity.traits.values()) {
        const traitClass = trait.constructor;
        if (traitClass.interceptors) {
            interceptors.push(...traitClass.interceptors);
        }
    }
    return interceptors;
}
/**
 * Type guard to check if a trait declares a specific interceptor.
 *
 * @param trait - The trait to check
 * @param actionId - The action ID to check for
 * @param traitType - Optional: specific trait class to narrow to
 * @returns True if trait has the interceptor (and is of specified type)
 *
 * @example
 * ```typescript
 * if (traitHasInterceptor(trait, 'if.action.entering', InflatableTrait)) {
 *   // trait is narrowed to InflatableTrait
 *   console.log(trait.isInflated);
 * }
 * ```
 */
function traitHasInterceptor(trait, actionId, traitType) {
    const traitClass = trait.constructor;
    const hasAction = traitClass.interceptors?.includes(actionId) ?? false;
    if (traitType) {
        return hasAction && trait instanceof traitType;
    }
    return hasAction;
}
/**
 * Find all traits on an entity that have interceptors.
 *
 * @param entity - The entity to inspect
 * @returns Array of traits that declare at least one interceptor
 */
function getInterceptorTraits(entity) {
    return Array.from(entity.traits.values()).filter(trait => {
        const traitClass = trait.constructor;
        return traitClass.interceptors && traitClass.interceptors.length > 0;
    });
}
