// @sharpee/world-model/capabilities/interceptor-registry.js
"use strict";
/**
 * Interceptor Registry (ADR-118)
 *
 * Registry for trait-interceptor bindings. Stories register interceptors
 * for their traits, and stdlib actions look them up at runtime.
 *
 * Mirrors capability-registry.ts pattern but for interceptors instead
 * of full-delegation behaviors.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.registerActionInterceptor = registerActionInterceptor;
exports.getInterceptorForAction = getInterceptorForAction;
exports.getInterceptorBinding = getInterceptorBinding;
exports.hasActionInterceptor = hasActionInterceptor;
exports.unregisterActionInterceptor = unregisterActionInterceptor;
exports.clearInterceptorRegistry = clearInterceptorRegistry;
exports.getAllInterceptorBindings = getAllInterceptorBindings;
/**
 * Registry storage: key = `${traitType}:${actionId}`
 * Uses globalThis to ensure the registry is shared across module boundaries
 * (e.g., when story modules import from packages but actions run from bundle)
 */
const INTERCEPTOR_REGISTRY_KEY = '__sharpee_interceptor_registry__';
// Use globalThis to share registry across module instances
function getInterceptorRegistry() {
    const global = globalThis;
    if (!global[INTERCEPTOR_REGISTRY_KEY]) {
        global[INTERCEPTOR_REGISTRY_KEY] = new Map();
    }
    return global[INTERCEPTOR_REGISTRY_KEY];
}
// For backwards compatibility, expose as const (but delegates to globalThis)
const interceptorRegistry = {
    get size() { return getInterceptorRegistry().size; },
    has(key) { return getInterceptorRegistry().has(key); },
    get(key) { return getInterceptorRegistry().get(key); },
    set(key, value) { return getInterceptorRegistry().set(key, value); },
    delete(key) { return getInterceptorRegistry().delete(key); },
    clear() { return getInterceptorRegistry().clear(); },
    keys() { return getInterceptorRegistry().keys(); },
    values() { return getInterceptorRegistry().values(); },
    entries() { return getInterceptorRegistry().entries(); },
    [Symbol.iterator]() { return getInterceptorRegistry()[Symbol.iterator](); }
};
/**
 * Generate registry key for trait+action combination.
 */
function registryKey(traitType, actionId) {
    return `${traitType}:${actionId}`;
}
/**
 * Register an interceptor for a trait+action combination.
 *
 * Stories call this to register their interceptors during initialization.
 * Each trait+action pair can only have one interceptor registered.
 *
 * @param traitType - The trait type identifier (e.g., 'dungeo.trait.inflatable')
 * @param actionId - The action ID (e.g., 'if.action.entering')
 * @param interceptor - The interceptor implementation
 * @param options - Optional configuration (priority)
 *
 * @example
 * ```typescript
 * // Register boat puncture interceptor
 * registerActionInterceptor(
 *   InflatableTrait.type,
 *   'if.action.entering',
 *   InflatableEnteringInterceptor
 * );
 *
 * // With priority
 * registerActionInterceptor(
 *   TrollGuardianTrait.type,
 *   'if.action.going',
 *   TrollGoingInterceptor,
 *   { priority: 100 }
 * );
 * ```
 */
function registerActionInterceptor(traitType, actionId, interceptor, options) {
    const key = registryKey(traitType, actionId);
    if (interceptorRegistry.has(key)) {
        throw new Error(`Interceptor already registered for trait "${traitType}" and action "${actionId}"`);
    }
    interceptorRegistry.set(key, {
        traitType,
        actionId,
        interceptor,
        priority: options?.priority ?? 0
    });
}
/**
 * Get interceptor for an entity and action.
 *
 * Finds a trait on the entity that has an interceptor registered for
 * the given action. Looks up by trait type string in the registry.
 *
 * @param entity - The entity to check
 * @param actionId - The action ID
 * @returns The interceptor and trait, or undefined if not found
 *
 * @example
 * ```typescript
 * const result = getInterceptorForAction(boat, 'if.action.entering');
 * if (result) {
 *   const { interceptor, trait, binding } = result;
 *   // Call interceptor hooks during action execution
 * }
 * ```
 */
function getInterceptorForAction(entity, actionId) {
    // Find all traits that have an interceptor registered for this action
    const candidates = [];
    for (const trait of entity.traits.values()) {
        // Look up interceptor by trait type string (more reliable than static property)
        const traitType = trait.type;
        const key = registryKey(traitType, actionId);
        const binding = interceptorRegistry.get(key);
        if (binding) {
            candidates.push({ trait, binding });
        }
    }
    if (candidates.length === 0) {
        return undefined;
    }
    // Sort by priority (highest first) and return the first one
    candidates.sort((a, b) => b.binding.priority - a.binding.priority);
    const { trait, binding } = candidates[0];
    return {
        interceptor: binding.interceptor,
        trait,
        binding
    };
}
/**
 * Get the interceptor binding for a trait type and action.
 *
 * @param traitType - The trait type identifier
 * @param actionId - The action ID
 * @returns The binding, or undefined if not registered
 */
function getInterceptorBinding(traitType, actionId) {
    return interceptorRegistry.get(registryKey(traitType, actionId));
}
/**
 * Check if an interceptor is registered for a trait+action.
 *
 * @param traitType - The trait type identifier
 * @param actionId - The action ID
 */
function hasActionInterceptor(traitType, actionId) {
    return interceptorRegistry.has(registryKey(traitType, actionId));
}
/**
 * Unregister an interceptor (primarily for testing).
 *
 * @param traitType - The trait type identifier
 * @param actionId - The action ID
 */
function unregisterActionInterceptor(traitType, actionId) {
    interceptorRegistry.delete(registryKey(traitType, actionId));
}
/**
 * Clear all registered interceptors (for testing).
 */
function clearInterceptorRegistry() {
    interceptorRegistry.clear();
}
/**
 * Get all registered interceptor bindings (for debugging/introspection).
 */
function getAllInterceptorBindings() {
    return new Map(interceptorRegistry);
}
