// @sharpee/world-model/traits/actor/actorTrait.js
"use strict";
// packages/world-model/src/traits/actor/actorTrait.ts
Object.defineProperty(exports, "__esModule", { value: true });
exports.ActorTrait = exports.HONORIFICS = exports.PRONOUNS = void 0;
const trait_types_1 = require("../trait-types");
/**
 * Standard pronoun sets (ADR-089)
 */
exports.PRONOUNS = {
    HE_HIM: {
        subject: 'he', object: 'him', possessive: 'his',
        possessiveAdj: 'his', reflexive: 'himself', verbForm: 'singular'
    },
    SHE_HER: {
        subject: 'she', object: 'her', possessive: 'hers',
        possessiveAdj: 'her', reflexive: 'herself', verbForm: 'singular'
    },
    THEY_THEM: {
        subject: 'they', object: 'them', possessive: 'theirs',
        possessiveAdj: 'their', reflexive: 'themselves', verbForm: 'plural'
    },
    // Neopronouns
    XE_XEM: {
        subject: 'xe', object: 'xem', possessive: 'xyrs',
        possessiveAdj: 'xyr', reflexive: 'xemself', verbForm: 'singular'
    },
    ZE_ZIR: {
        subject: 'ze', object: 'zir', possessive: 'zirs',
        possessiveAdj: 'zir', reflexive: 'zirself', verbForm: 'singular'
    },
    ZE_HIR: {
        subject: 'ze', object: 'hir', possessive: 'hirs',
        possessiveAdj: 'hir', reflexive: 'hirself', verbForm: 'singular'
    },
    EY_EM: {
        subject: 'ey', object: 'em', possessive: 'eirs',
        possessiveAdj: 'eir', reflexive: 'emself', verbForm: 'singular'
    },
    FAE_FAER: {
        subject: 'fae', object: 'faer', possessive: 'faers',
        possessiveAdj: 'faer', reflexive: 'faerself', verbForm: 'singular'
    },
};
/**
 * Standard honorifics/titles (ADR-089)
 */
exports.HONORIFICS = {
    MR: 'Mr.',
    MRS: 'Mrs.',
    MS: 'Ms.',
    MX: 'Mx.', // Gender-neutral
    MISS: 'Miss',
    DR: 'Dr.',
    PROF: 'Prof.',
};
/**
 * Trait for entities that can act in the world (player, NPCs, etc.)
 *
 * Actors can:
 * - Perform actions
 * - Hold inventory (actors inherently have container functionality)
 * - Move between locations
 * - Interact with objects
 *
 * Like rooms, actors have built-in container functionality for their inventory.
 * The actual containment relationships are stored in the SpatialIndex.
 */
class ActorTrait {
    static type = trait_types_1.TraitType.ACTOR;
    type = trait_types_1.TraitType.ACTOR;
    isPlayer = false;
    isPlayable = true;
    state;
    /** Pronouns - defaults to they/them (ADR-089) */
    pronouns = exports.PRONOUNS.THEY_THEM;
    /** Optional honorific/title */
    honorific;
    /** Grammatical gender for localization */
    grammaticalGender;
    /** Brief description for disambiguation */
    briefDescription;
    // Container functionality for inventory
    capacity;
    allowedTypes;
    excludedTypes;
    // Actors are not transparent (can't see inside inventory) and not enterable
    isTransparent = false;
    enterable = false;
    customProperties;
    constructor(data) {
        if (data) {
            // Handle basic properties
            this.isPlayer = data.isPlayer ?? this.isPlayer;
            this.isPlayable = data.isPlayable ?? this.isPlayable;
            this.state = data.state;
            this.customProperties = data.customProperties;
            // Handle pronouns (ADR-089)
            if (data.pronouns) {
                this.pronouns = data.pronouns;
            }
            this.honorific = data.honorific;
            this.grammaticalGender = data.grammaticalGender;
            this.briefDescription = data.briefDescription;
            // Handle container properties
            this.capacity = data.capacity;
            this.allowedTypes = data.allowedTypes;
            this.excludedTypes = data.excludedTypes;
        }
    }
    /**
     * Get the primary pronoun set (first if array, or the single set)
     */
    getPrimaryPronouns() {
        return Array.isArray(this.pronouns) ? this.pronouns[0] : this.pronouns;
    }
    /**
     * Set pronouns for this actor
     */
    setPronouns(pronouns) {
        this.pronouns = pronouns;
    }
    /**
     * Set inventory limits
     */
    setInventoryLimit(limit) {
        this.capacity = { ...this.capacity, ...limit };
    }
    /**
     * Mark as player character
     */
    makePlayer() {
        this.isPlayer = true;
        this.isPlayable = true;
    }
    /**
     * Set custom property
     */
    setCustomProperty(key, value) {
        if (!this.customProperties) {
            this.customProperties = {};
        }
        this.customProperties[key] = value;
    }
    /**
     * Get custom property
     */
    getCustomProperty(key) {
        return this.customProperties?.[key];
    }
}
exports.ActorTrait = ActorTrait;
