// @sharpee/world-model/traits/actor/actorBehavior.js
"use strict";
// packages/world-model/src/traits/actor/actorBehavior.ts
Object.defineProperty(exports, "__esModule", { value: true });
exports.ActorBehavior = void 0;
const behavior_1 = require("../../behaviors/behavior");
const trait_types_1 = require("../trait-types");
const containerBehavior_1 = require("../container/containerBehavior");
const identityBehavior_1 = require("../identity/identityBehavior");
const wearableBehavior_1 = require("../wearable/wearableBehavior");
/**
 * Behavior for actors in the world.
 *
 * Handles inventory management, movement validation, and actor-specific logic.
 */
class ActorBehavior extends behavior_1.Behavior {
    static requiredTraits = [trait_types_1.TraitType.ACTOR];
    /**
     * Check if this is the player character
     */
    static isPlayer(entity) {
        const actor = ActorBehavior.require(entity, trait_types_1.TraitType.ACTOR);
        return actor.isPlayer || false;
    }
    /**
     * Check if this actor can be controlled by the player
     */
    static isPlayable(entity) {
        const actor = ActorBehavior.require(entity, trait_types_1.TraitType.ACTOR);
        return actor.isPlayable || false;
    }
    /**
     * Get actor's pronouns
     */
    static getPronouns(entity) {
        const actor = ActorBehavior.require(entity, trait_types_1.TraitType.ACTOR);
        return actor.pronouns;
    }
    /**
     * Get a specific pronoun
     */
    static getPronoun(entity, type) {
        const pronouns = ActorBehavior.getPronouns(entity);
        return pronouns[type];
    }
    /**
     * Check if actor can carry an item
     */
    static canCarry(actor, item, world) {
        const actorTrait = ActorBehavior.require(actor, trait_types_1.TraitType.ACTOR);
        // Check inventory limits
        if (actorTrait.capacity) {
            const limit = actorTrait.capacity;
            // Check item count
            if (limit.maxItems !== undefined) {
                const currentCount = world.getContents(actor.id).length;
                if (currentCount >= limit.maxItems) {
                    return false;
                }
            }
            // Check weight limit
            if (limit.maxWeight !== undefined) {
                const currentWeight = ActorBehavior.getCarriedWeight(actor, world);
                const itemWeight = identityBehavior_1.IdentityBehavior.getWeight(item);
                if (currentWeight + itemWeight > limit.maxWeight) {
                    return false;
                }
            }
            // Check volume limit
            if (limit.maxVolume !== undefined) {
                const currentVolume = ActorBehavior.getCarriedVolume(actor, world);
                const itemVolume = identityBehavior_1.IdentityBehavior.getVolume(item);
                if (currentVolume + itemVolume > limit.maxVolume) {
                    return false;
                }
            }
        }
        return true;
    }
    /**
     * Get total weight of carried items
     */
    static getCarriedWeight(actor, world) {
        const items = world.getContents(actor.id);
        return items.reduce((total, item) => {
            return total + identityBehavior_1.IdentityBehavior.getWeight(item);
        }, 0);
    }
    /**
     * Get total volume of carried items
     */
    static getCarriedVolume(actor, world) {
        const items = world.getContents(actor.id);
        return items.reduce((total, item) => {
            return total + identityBehavior_1.IdentityBehavior.getVolume(item);
        }, 0);
    }
    /**
     * Get remaining carrying capacity
     */
    static getRemainingCapacity(actor, world) {
        const actorTrait = ActorBehavior.require(actor, trait_types_1.TraitType.ACTOR);
        const result = {};
        if (actorTrait.capacity) {
            const limit = actorTrait.capacity;
            if (limit.maxItems !== undefined) {
                const current = world.getContents(actor.id).length;
                result.items = limit.maxItems - current;
            }
            if (limit.maxWeight !== undefined) {
                result.weight = limit.maxWeight - ActorBehavior.getCarriedWeight(actor, world);
            }
            if (limit.maxVolume !== undefined) {
                result.volume = limit.maxVolume - ActorBehavior.getCarriedVolume(actor, world);
            }
        }
        return result;
    }
    /**
     * Check if actor is holding a specific item
     */
    static isHolding(actor, itemId, world) {
        const itemLocation = world.getLocation(itemId);
        return itemLocation === actor.id;
    }
    /**
     * Get actor's current state
     */
    static getState(entity) {
        const actor = ActorBehavior.require(entity, trait_types_1.TraitType.ACTOR);
        return actor.state;
    }
    /**
     * Set actor's state
     */
    static setState(entity, state) {
        const actor = ActorBehavior.require(entity, trait_types_1.TraitType.ACTOR);
        actor.state = state;
    }
    /**
     * Get custom property value
     */
    static getCustomProperty(entity, key) {
        const actor = ActorBehavior.require(entity, trait_types_1.TraitType.ACTOR);
        return actor.customProperties?.[key];
    }
    /**
     * Set custom property value
     */
    static setCustomProperty(entity, key, value) {
        const actor = ActorBehavior.require(entity, trait_types_1.TraitType.ACTOR);
        if (!actor.customProperties) {
            actor.customProperties = {};
        }
        actor.customProperties[key] = value;
    }
    /**
     * Find the player actor in a collection of entities
     */
    static findPlayer(entities) {
        return entities.find(e => e.has(trait_types_1.TraitType.ACTOR) && ActorBehavior.isPlayer(e));
    }
    /**
     * Check if an actor can take an item (capacity validation)
     * @param actor The actor trying to take the item
     * @param item The item being taken
     * @param world World query interface
     * @returns True if the actor can take the item
     */
    static canTakeItem(actor, item, world) {
        const actorTrait = actor.get(trait_types_1.TraitType.ACTOR);
        if (!actorTrait)
            return false;
        // If actor has CONTAINER trait, delegate to ContainerBehavior
        if (actor.has(trait_types_1.TraitType.CONTAINER)) {
            return containerBehavior_1.ContainerBehavior.canAccept(actor, item, world);
        }
        // Check weight capacity
        if (actorTrait?.capacity?.maxWeight !== undefined) {
            const maxWeight = actorTrait.capacity.maxWeight;
            const currentWeight = ActorBehavior.getCarriedWeight(actor, world);
            // Note: ItemWeight calculation would need world interface extension
            // For now, assume weight checking is done at higher level
            const itemWeight = 0; // TODO: Add weight support to IWorldQuery
            if (currentWeight + itemWeight > maxWeight) {
                return false;
            }
        }
        // Check item count if defined on ACTOR trait (and not already checked via CONTAINER)
        if (!actor.has(trait_types_1.TraitType.CONTAINER) && actorTrait?.capacity?.maxItems !== undefined) {
            const currentCount = world.getContents(actor.id)
                .filter(item => !wearableBehavior_1.WearableBehavior.isWorn(item)).length;
            if (currentCount >= actorTrait.capacity.maxItems) {
                return false;
            }
        }
        return true;
    }
    /**
     * Have an actor take an item
     * @param actor The actor taking the item
     * @param item The item being taken
     * @param world World query interface
     * @returns Result describing what happened
     */
    static takeItem(actor, item, world) {
        // Check if actor already has the item
        const currentLocation = world.getLocation(item.id);
        if (currentLocation === actor.id) {
            return {
                success: false,
                alreadyHeld: true,
                stateChanged: false
            };
        }
        // Check capacity constraints
        if (!ActorBehavior.canTakeItem(actor, item, world)) {
            const actorTrait = actor.get(trait_types_1.TraitType.ACTOR);
            // Determine specific failure reason
            if (actorTrait?.capacity?.maxWeight !== undefined) {
                const maxWeight = actorTrait.capacity.maxWeight;
                const currentWeight = ActorBehavior.getCarriedWeight(actor, world);
                // Note: ItemWeight calculation would need world interface extension
                const itemWeight = 0; // TODO: Add weight support to IWorldQuery
                if (currentWeight + itemWeight > maxWeight) {
                    return {
                        success: false,
                        tooHeavy: true,
                        stateChanged: false
                    };
                }
            }
            return {
                success: false,
                inventoryFull: true,
                stateChanged: false
            };
        }
        // Check if item needs removal from container/supporter
        let needsRemoval = false;
        let fromContainer;
        if (currentLocation) {
            // Note: getEntity would need to be added to IWorldQuery interface
            // For now, assume container detection is done at higher level
            needsRemoval = true; // TODO: Add entity lookup support to IWorldQuery
            fromContainer = currentLocation;
        }
        // All checks passed - taking is valid
        return {
            success: true,
            stateChanged: true,
            needsRemoval,
            fromContainer
        };
    }
    /**
     * Have an actor drop an item
     * @param actor The actor dropping the item
     * @param item The item being dropped
     * @param world World query interface
     * @returns Result describing what happened
     */
    static dropItem(actor, item, world) {
        // Check if actor actually has the item
        const currentLocation = world.getLocation(item.id);
        if (currentLocation !== actor.id) {
            return {
                success: false,
                notHeld: true,
                stateChanged: false
            };
        }
        // Check if item is still worn (can't drop worn items)
        if (wearableBehavior_1.WearableBehavior.isWorn(item)) {
            return {
                success: false,
                stillWorn: true,
                stateChanged: false
            };
        }
        // All checks passed - dropping is valid
        return {
            success: true,
            stateChanged: true
        };
    }
}
exports.ActorBehavior = ActorBehavior;
