// @sharpee/media/audio/audio-registry.js
"use strict";
/**
 * AudioRegistry — central registration system for all audio in a story.
 *
 * Public interface: AudioCue, VariationPool, DuckingConfig, RoomAtmosphere,
 * AtmosphereBuilder, and AudioRegistry class. Stories create one AudioRegistry
 * in initializeWorld() and populate it with cues, pools, and atmospheres.
 * Actions and handlers reference registered names only — never raw audio
 * parameters.
 *
 * Owner context: @sharpee/media (ADR-138)
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.AtmosphereBuilder = exports.AudioRegistry = void 0;
const core_1 = require("../../core/index.js");
// Ensure declaration merging is active
require("./registry-merge");
/**
 * Central registry for all audio in a story.
 * Stories populate this during initializeWorld(). Actions and handlers
 * reference registered names — never raw audio parameters.
 *
 * @example
 * ```typescript
 * const audio = new AudioRegistry();
 *
 * // Register a simple cue
 * audio.registerCue('skin.activate', () =>
 *   createTypedEvent('audio.sfx', { src: 'sfx/skin-activate.mp3', volume: 0.8 })
 * );
 *
 * // Register a variation pool (multiple files, random selection + jitter)
 * audio.registerPool('footstep.metal', {
 *   sources: ['sfx/step-metal-1.mp3', 'sfx/step-metal-2.mp3', 'sfx/step-metal-3.mp3'],
 *   volume: 0.6,
 *   volumeJitter: 0.1,
 *   pitchJitter: 0.05,
 * });
 *
 * // Register a room atmosphere using the fluent builder
 * audio.atmosphere('entropy.room.bridge')
 *   .ambient('ambient/ship-hum.mp3', 'environment', 0.3)
 *   .ambient('ambient/console-beeps.mp3', 'machinery', 0.15)
 *   .music('music/bridge-theme.mp3', 0.4)
 *   .effect('lowpass', 'ambient:environment', { frequency: 2000, q: 1 })
 *   .build();
 *
 * // In an action or handler — just use the name
 * const events = audio.cue('skin.activate');
 * ```
 */
class AudioRegistry {
    cues = new Map();
    pools = new Map();
    atmospheres = new Map();
    fadeDefaultValues = {
        ambientIn: 2000,
        ambientOut: 2000,
        musicIn: 1000,
        effectTransition: 2000,
    };
    duckingConfig = {
        duckVolume: 0.3,
        attackMs: 100,
        releaseMs: 500,
        targets: ['music', 'ambient'],
    };
    // ── Cues ────────────────────────────────────────────────────────────
    /**
     * Register a named audio cue (single sound, full control).
     *
     * @param name - Unique cue identifier (e.g., 'skin.activate', 'door.open')
     * @param cue - Factory function that returns a fresh ISemanticEvent
     */
    registerCue(name, cue) {
        this.cues.set(name, cue);
    }
    /**
     * Register a variation pool — multiple files for one logical sound.
     * When fired, the registry picks a random source and applies jitter.
     *
     * @param name - Unique pool identifier (e.g., 'footstep.metal')
     * @param pool - Pool configuration with sources and jitter parameters
     */
    registerPool(name, pool) {
        this.pools.set(name, pool);
    }
    /**
     * Fire a registered cue or pool by name. Returns empty array if not registered.
     *
     * Resolution order:
     * 1. Check cues (exact factory)
     * 2. Check pools (random selection + jitter -> AudioSfxEvent)
     * 3. Return [] (silent degradation)
     *
     * @param name - Registered cue or pool name
     * @returns Array of events to emit (empty if name is not registered)
     */
    cue(name) {
        const factory = this.cues.get(name);
        if (factory)
            return [factory()];
        const pool = this.pools.get(name);
        if (pool)
            return [this.resolvePool(pool)];
        return [];
    }
    resolvePool(pool) {
        const src = pool.sources[Math.floor(Math.random() * pool.sources.length)];
        const baseVolume = pool.volume ?? 1.0;
        const volumeJitter = pool.volumeJitter ?? 0;
        const pitchJitter = pool.pitchJitter ?? 0;
        const volume = Math.max(0, Math.min(1, baseVolume + (Math.random() * 2 - 1) * volumeJitter));
        const rate = 1.0 + (Math.random() * 2 - 1) * pitchJitter;
        return (0, core_1.createTypedEvent)('audio.sfx', { src, volume, rate, duck: pool.duck });
    }
    // ── Ducking ─────────────────────────────────────────────────────────
    /**
     * Override the default ducking behavior.
     *
     * @param config - Partial ducking configuration to merge with defaults
     */
    setDucking(config) {
        this.duckingConfig = { ...this.duckingConfig, ...config };
    }
    /**
     * Get the current ducking configuration.
     *
     * @returns Read-only ducking config
     */
    getDucking() {
        return this.duckingConfig;
    }
    // ── Atmospheres ─────────────────────────────────────────────────────
    /**
     * Register a room atmosphere by room ID (raw object).
     *
     * @param roomId - Entity ID of the room
     * @param atmosphere - Complete atmosphere definition
     */
    registerAtmosphere(roomId, atmosphere) {
        this.atmospheres.set(roomId, atmosphere);
    }
    /**
     * Start a fluent atmosphere builder for a room.
     *
     * @param roomId - Entity ID of the room
     * @returns AtmosphereBuilder instance — call .build() to register
     */
    atmosphere(roomId) {
        return new AtmosphereBuilder(this, roomId);
    }
    /**
     * Get the registered atmosphere for a room.
     *
     * @param roomId - Entity ID of the room
     * @returns The atmosphere definition, or undefined if none registered
     */
    getAtmosphere(roomId) {
        return this.atmospheres.get(roomId);
    }
    // ── Fade defaults ───────────────────────────────────────────────────
    /**
     * Override default fade durations (ms).
     *
     * @param defaults - Partial fade defaults to merge
     */
    setFadeDefaults(defaults) {
        this.fadeDefaultValues = { ...this.fadeDefaultValues, ...defaults };
    }
    /**
     * Get the current fade defaults.
     *
     * @returns Read-only fade defaults
     */
    getFadeDefaults() {
        return this.fadeDefaultValues;
    }
}
exports.AudioRegistry = AudioRegistry;
/**
 * Fluent builder for room atmospheres.
 * Avoids verbose JSON object literals in registration code.
 *
 * @example
 * ```typescript
 * registry.atmosphere('entropy.room.bridge')
 *   .ambient('ambient/ship-hum.mp3', 'environment', 0.3)
 *   .music('music/bridge-theme.mp3', 0.4)
 *   .build();
 * ```
 */
class AtmosphereBuilder {
    registry;
    roomId;
    _ambient = [];
    _music;
    _effect;
    constructor(registry, roomId) {
        this.registry = registry;
        this.roomId = roomId;
    }
    /**
     * Add an ambient channel to the atmosphere.
     *
     * @param src - Audio file path
     * @param channel - Named channel identifier
     * @param volume - Playback volume (0.0-1.0)
     * @returns this (for chaining)
     */
    ambient(src, channel, volume) {
        this._ambient = [...this._ambient, { src, channel, volume }];
        return this;
    }
    /**
     * Set the music track for the atmosphere.
     *
     * @param src - Audio file path
     * @param volume - Playback volume (0.0-1.0)
     * @returns this (for chaining)
     */
    music(src, volume) {
        this._music = { src, volume };
        return this;
    }
    /**
     * Set an audio effect for the atmosphere.
     *
     * @param effect - Effect type (e.g., 'reverb', 'lowpass')
     * @param target - Mix target to apply the effect to
     * @param params - Effect-specific parameters
     * @returns this (for chaining)
     */
    effect(effect, target, params) {
        this._effect = { effect, target, params };
        return this;
    }
    /**
     * Register the atmosphere with the registry.
     */
    build() {
        this.registry.registerAtmosphere(this.roomId, {
            ambient: this._ambient,
            music: this._music,
            effect: this._effect,
        });
    }
}
exports.AtmosphereBuilder = AtmosphereBuilder;
