// @sharpee/character/goals/builder.js
"use strict";
/**
 * Goal builder API (ADR-145 layer 5)
 *
 * Fluent builder for defining NPC goals with activation conditions,
 * pursuit modes, behavior sequences, and interruption rules.
 * Returns from CharacterBuilder.goal(id) and compiles to GoalDef
 * stored in CompiledCharacter.goalDefs.
 *
 * Public interface: GoalBuilder.
 * Owner context: @sharpee/character / goals
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.GoalBuilder = void 0;
/**
 * Fluent builder for a single goal definition.
 *
 * Usage:
 * ```
 * builder.goal('eliminate-player')
 *   .activatesWhen('knows murder discovered', 'has weapon')
 *   .priority('critical')
 *   .mode('prepared')
 *   .pursues([
 *     { type: 'moveTo', target: 'study' },
 *     { type: 'acquire', target: 'knife' },
 *   ])
 *   .actsWhen('alone with player')
 *   .act('colonel-attacks-player')
 *   .onInterrupt('colonel-retreats')
 *   .resumeOnClear(true)
 *   .done()
 * ```
 */
class GoalBuilder {
    _id;
    _parent;
    _finalize;
    _activatesWhen = [];
    _interruptedBy = [];
    _priority = 'medium';
    _mode = 'sequential';
    _steps = [];
    _actsWhen = [];
    _actMessageId;
    _onInterrupt;
    _resumeOnClear = false;
    /**
     * Create a new goal builder.
     *
     * @param id - Unique goal identifier
     * @param parent - Parent builder to return to on .done()
     * @param finalize - Callback to register the compiled GoalDef
     */
    constructor(id, parent, finalize) {
        this._id = id;
        this._parent = parent;
        this._finalize = finalize;
    }
    /**
     * Set predicate conditions that activate this goal.
     *
     * @param predicates - Predicate names (all must be true)
     * @returns this for chaining
     */
    activatesWhen(...predicates) {
        this._activatesWhen.push(...predicates);
        return this;
    }
    /**
     * Set goal priority.
     *
     * @param priority - Priority word
     * @returns this for chaining
     */
    priority(priority) {
        this._priority = priority;
        return this;
    }
    /**
     * Set pursuit mode.
     *
     * @param mode - How the NPC pursues the goal
     * @returns this for chaining
     */
    mode(mode) {
        this._mode = mode;
        return this;
    }
    /**
     * Set the behavior sequence (for sequential and prepared modes).
     *
     * @param steps - Array of goal steps
     * @returns this for chaining
     */
    pursues(steps) {
        this._steps = [...steps];
        return this;
    }
    /**
     * Set conditions for the final act (opportunistic/prepared modes).
     *
     * @param predicates - Predicate names (all must be true)
     * @returns this for chaining
     */
    actsWhen(...predicates) {
        this._actsWhen.push(...predicates);
        return this;
    }
    /**
     * Set the message ID for the final act.
     *
     * @param messageId - Message ID
     * @returns this for chaining
     */
    act(messageId) {
        this._actMessageId = messageId;
        return this;
    }
    /**
     * Set predicate conditions that interrupt (suspend) this goal.
     *
     * @param predicates - Predicate names
     * @returns this for chaining
     */
    interruptedBy(...predicates) {
        this._interruptedBy.push(...predicates);
        return this;
    }
    /**
     * Set the message ID when the goal is interrupted.
     *
     * @param messageId - Message ID
     * @returns this for chaining
     */
    onInterrupt(messageId) {
        this._onInterrupt = messageId;
        return this;
    }
    /**
     * Set whether the goal resumes from where it left off after interruption.
     *
     * @param resume - Whether to resume
     * @returns this for chaining
     */
    resumeOnClear(resume) {
        this._resumeOnClear = resume;
        return this;
    }
    /**
     * Finalize this goal definition and return the parent builder.
     *
     * @returns The parent builder
     */
    done() {
        this._finalize(this._buildDef());
        return this._parent;
    }
    /**
     * Compile the parent builder, auto-finalizing this goal.
     * Allows calling .compile() directly from a goal chain without .done().
     *
     * @returns Compiled character data (delegates to parent's compile())
     */
    compile() {
        const parent = this.done();
        return parent.compile();
    }
    /** @internal Build the GoalDef without finalizing. */
    _buildDef() {
        const def = {
            id: this._id,
            activatesWhen: [...this._activatesWhen],
            priority: this._priority,
            mode: this._mode,
        };
        if (this._steps.length > 0)
            def.steps = [...this._steps];
        if (this._interruptedBy.length > 0)
            def.interruptedBy = [...this._interruptedBy];
        if (this._actsWhen.length > 0)
            def.actsWhen = [...this._actsWhen];
        if (this._actMessageId !== undefined)
            def.actMessageId = this._actMessageId;
        if (this._onInterrupt !== undefined)
            def.onInterrupt = this._onInterrupt;
        if (this._resumeOnClear)
            def.resumeOnClear = true;
        return def;
    }
}
exports.GoalBuilder = GoalBuilder;
