// @sharpee/character/goals/pathfinding.js
"use strict";
/**
 * NPC pathfinding (ADR-145)
 *
 * BFS over room connection graph filtered by NPC movement profile.
 * NPCs can only pathfind through rooms they know and passages they
 * have access to.
 *
 * Public interface: findNextRoom, RoomGraph, RoomConnection.
 * Owner context: @sharpee/character / goals
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SimpleRoomGraph = void 0;
exports.findNextRoom = findNextRoom;
/**
 * Simple room graph implementation for testing.
 * Production code can implement the RoomGraph interface directly.
 */
class SimpleRoomGraph {
    connections = new Map();
    /**
     * Add a bidirectional connection between two rooms.
     *
     * @param from - Source room ID
     * @param to - Destination room ID
     * @param passageId - Optional passage ID
     */
    addConnection(from, to, passageId) {
        this.addDirected(from, to, passageId);
        this.addDirected(to, from, passageId); // Reverse direction for bidirectional edge
    }
    /**
     * Add a one-way connection.
     *
     * @param from - Source room ID
     * @param to - Destination room ID
     * @param passageId - Optional passage ID
     */
    addDirected(from, to, passageId) {
        const existing = this.connections.get(from) ?? [];
        existing.push({ from, to, passageId });
        this.connections.set(from, existing);
    }
    getConnections(roomId) {
        return this.connections.get(roomId) ?? [];
    }
}
exports.SimpleRoomGraph = SimpleRoomGraph;
// ---------------------------------------------------------------------------
// Pathfinding
// ---------------------------------------------------------------------------
/**
 * Find the next room the NPC should move to on the shortest path
 * toward a target room, filtered by movement profile.
 *
 * Uses BFS (breadth-first search) over the room graph. Only traverses
 * rooms the NPC knows about and passages the NPC has access to.
 *
 * @param currentRoom - The NPC's current room ID
 * @param targetRoom - The destination room ID
 * @param graph - The room connection graph
 * @param movement - The NPC's movement profile
 * @returns The next room ID to move to, or null if unreachable
 */
function findNextRoom(currentRoom, targetRoom, graph, movement) {
    if (currentRoom === targetRoom)
        return null; // Already there
    // BFS
    const visited = new Set();
    // Queue entries: [roomId, firstStepFromCurrent]
    const queue = [];
    visited.add(currentRoom);
    // Seed with adjacent rooms
    for (const conn of graph.getConnections(currentRoom)) {
        if (!isAccessible(conn, movement))
            continue;
        if (conn.to === targetRoom)
            return conn.to;
        visited.add(conn.to);
        queue.push([conn.to, conn.to]); // firstStep is this neighbor
    }
    // BFS loop
    while (queue.length > 0) {
        const [room, firstStep] = queue.shift();
        for (const conn of graph.getConnections(room)) {
            if (visited.has(conn.to))
                continue;
            if (!isAccessible(conn, movement))
                continue;
            if (conn.to === targetRoom)
                return firstStep;
            visited.add(conn.to);
            queue.push([conn.to, firstStep]);
        }
    }
    return null; // Unreachable
}
// ---------------------------------------------------------------------------
// Movement profile checks
// ---------------------------------------------------------------------------
/**
 * Check if a connection is accessible: the NPC can traverse the passage
 * and knows the destination room.
 *
 * Combines passage access and room knowledge into a single check,
 * used by both the BFS seed loop and the main traversal loop.
 *
 * @param conn - The room connection to check
 * @param movement - The NPC's movement profile
 * @returns true if the NPC can use this connection
 */
function isAccessible(conn, movement) {
    return canTraverse(conn, movement) && knowsRoom(conn.to, movement);
}
/**
 * Check if the NPC can traverse a connection.
 * If the connection has a passageId, the NPC must have access to it.
 */
function canTraverse(conn, movement) {
    if (!conn.passageId)
        return true; // No passage restriction
    if (movement.access === 'all')
        return true;
    return movement.access.includes(conn.passageId);
}
/**
 * Check if the NPC knows about a room.
 */
function knowsRoom(roomId, movement) {
    if (movement.knows === 'all')
        return true;
    return movement.knows.includes(roomId);
}
