// @sharpee/character/goals/step-evaluator.js
"use strict";
/**
 * Goal step evaluator (ADR-145)
 *
 * Evaluates the current step of an active goal each NPC turn.
 * Pure evaluation logic — returns step results that the caller applies.
 *
 * Public interface: evaluateGoalStep, GoalStepContext.
 * Owner context: @sharpee/character / goals
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.evaluateGoalStep = evaluateGoalStep;
const pathfinding_1 = require("./pathfinding");
// ---------------------------------------------------------------------------
// Step evaluation
// ---------------------------------------------------------------------------
/**
 * Evaluate a single goal step for an NPC.
 *
 * Each step type produces a StepResult:
 * - completed: step is done, advance to next
 * - in-progress: step partially done (e.g., moving toward target)
 * - waiting: conditions not met, hold this turn
 * - blocked: cannot proceed (e.g., unreachable target)
 *
 * @param goal - The active goal
 * @param ctx - The evaluation context
 * @returns The step evaluation result
 */
function evaluateGoalStep(goal, ctx) {
    // Opportunistic mode — no steps, just check act conditions
    if (goal.def.mode === 'opportunistic') {
        return evaluateOpportunistic(goal, ctx);
    }
    // Prepared mode with all prep steps done — switch to opportunistic
    if (goal.def.mode === 'prepared' && goal.prepared) {
        return evaluateOpportunistic(goal, ctx);
    }
    // Sequential/prepared mode — execute current step
    const steps = goal.def.steps;
    if (!steps || goal.currentStep >= steps.length) {
        return { status: 'completed' };
    }
    const step = steps[goal.currentStep];
    return evaluateStep(step, ctx);
}
// ---------------------------------------------------------------------------
// Individual step type evaluators
// ---------------------------------------------------------------------------
function evaluateStep(step, ctx) {
    switch (step.type) {
        case 'seek':
            return evaluateSeek(step.target, step.from, step.witnessed, ctx);
        case 'moveTo':
            return evaluateSeek(step.target, undefined, step.witnessed, ctx);
        case 'acquire':
            return evaluateAcquire(step.target, step.witnessed, ctx);
        case 'waitFor':
            return evaluateWaitFor(step.conditions, ctx);
        case 'act':
        case 'say':
            return {
                status: 'completed',
                witnessed: ctx.playerPresent ? step.witnessed ?? step.messageId : undefined,
            };
        case 'give':
            return evaluateGive(step.item, step.target, step.witnessed, ctx);
        case 'drop':
            return evaluateDrop(step.item, step.witnessed, ctx);
    }
}
function evaluateSeek(target, from, witnessed, ctx) {
    // Resolve target room — could be a room ID or an entity ID
    let targetRoom = target;
    if (ctx.getEntityRoom) {
        const entityRoom = ctx.getEntityRoom(target);
        if (entityRoom)
            targetRoom = entityRoom;
    }
    // Use 'from' hint if provided and NPC is not already moving
    if (from && ctx.currentRoom !== from) {
        targetRoom = from;
    }
    // Already at target
    if (ctx.currentRoom === targetRoom) {
        return { status: 'completed' };
    }
    // Find next room toward target
    const nextRoom = (0, pathfinding_1.findNextRoom)(ctx.currentRoom, targetRoom, ctx.roomGraph, ctx.movement);
    if (nextRoom === null) {
        return { status: 'blocked', reason: `Cannot reach ${targetRoom}` };
    }
    return {
        status: 'in-progress',
        witnessed: ctx.playerPresent ? witnessed : undefined,
    };
}
/**
 * Shared logic for steps that complete when a target entity is in the NPC's room.
 * Used by both 'acquire' and 'give' step types.
 *
 * @param target - Entity ID to check for room co-location
 * @param witnessed - Optional message ID when player observes
 * @param ctx - The goal step evaluation context
 * @returns Completed if target is in room, waiting otherwise
 */
function evaluateTargetInRoom(target, witnessed, ctx) {
    if (ctx.isInRoom(target, ctx.currentRoom)) {
        return {
            status: 'completed',
            witnessed: ctx.playerPresent ? witnessed : undefined,
        };
    }
    return { status: 'waiting' };
}
function evaluateAcquire(target, witnessed, ctx) {
    return evaluateTargetInRoom(target, witnessed, ctx);
}
function evaluateWaitFor(conditions, ctx) {
    const allMet = conditions.every(cond => ctx.trait.evaluate(cond));
    if (allMet) {
        return { status: 'completed' };
    }
    return { status: 'waiting' };
}
function evaluateGive(_item, target, witnessed, ctx) {
    return evaluateTargetInRoom(target, witnessed, ctx);
}
function evaluateDrop(_item, witnessed, ctx) {
    return {
        status: 'completed',
        witnessed: ctx.playerPresent ? witnessed : undefined,
    };
}
// ---------------------------------------------------------------------------
// Opportunistic evaluation
// ---------------------------------------------------------------------------
function evaluateOpportunistic(goal, ctx) {
    if (!goal.def.actsWhen) {
        return { status: 'waiting' };
    }
    const allMet = goal.def.actsWhen.every(cond => ctx.trait.evaluate(cond));
    if (allMet) {
        return {
            status: 'completed',
            witnessed: ctx.playerPresent
                ? goal.def.actMessageId
                : undefined,
        };
    }
    return { status: 'waiting' };
}
