// @sharpee/world-model/traits/room/roomTrait.js
"use strict";
// packages/world-model/src/traits/room/roomTrait.ts
Object.defineProperty(exports, "__esModule", { value: true });
exports.RoomTrait = void 0;
const trait_types_1 = require("../trait-types");
/**
 * Room trait marks an entity as a location/room in the game world.
 * Rooms are special entities that represent locations and can contain other entities.
 *
 * Rooms inherently have container functionality - they can hold items, actors, and other entities.
 * The actual containment relationships are stored in the SpatialIndex, not in the trait itself.
 */
class RoomTrait {
    static type = trait_types_1.TraitType.ROOM;
    type = trait_types_1.TraitType.ROOM;
    // RoomData properties
    visited;
    exits;
    blockedExits;
    outdoor;
    isDark;
    isOutdoors;
    isUnderground;
    initialDescription;
    initialDescriptionId;
    ambientSound;
    ambientSmell;
    regionId;
    tags;
    // Container functionality
    capacity;
    allowedTypes;
    excludedTypes;
    // Rooms are always transparent (contents visible) and enterable
    isTransparent = true;
    enterable = true;
    constructor(data = {}) {
        // Set defaults and merge with provided data
        this.visited = data.visited ?? false;
        this.exits = data.exits ?? {};
        this.blockedExits = data.blockedExits;
        this.outdoor = data.outdoor ?? false;
        this.isDark = data.isDark ?? false; // Default to lit
        this.isOutdoors = data.isOutdoors ?? false;
        this.isUnderground = data.isUnderground ?? false;
        this.initialDescription = data.initialDescription;
        this.initialDescriptionId = data.initialDescriptionId;
        this.ambientSound = data.ambientSound;
        this.ambientSmell = data.ambientSmell;
        this.regionId = data.regionId;
        this.tags = data.tags ?? [];
        // Container properties
        this.capacity = data.capacity;
        this.allowedTypes = data.allowedTypes;
        this.excludedTypes = data.excludedTypes;
    }
}
exports.RoomTrait = RoomTrait;
