// @sharpee/world-model/traits/room/roomBehavior.js
"use strict";
// packages/world-model/src/traits/room/roomBehavior.ts
Object.defineProperty(exports, "__esModule", { value: true });
exports.RoomBehavior = void 0;
const behavior_1 = require("../../behaviors/behavior");
const trait_types_1 = require("../trait-types");
const if_events_1 = require("../../constants/if-events");
/**
 * Behavior for room entities.
 *
 * Handles the logic for rooms including exits, lighting, and visits.
 * All methods are pure and only operate on the given room entity.
 */
class RoomBehavior extends behavior_1.Behavior {
    static requiredTraits = [trait_types_1.TraitType.ROOM];
    /**
     * Get the exit in a given direction
     */
    static getExit(room, direction) {
        const roomTrait = RoomBehavior.require(room, trait_types_1.TraitType.ROOM);
        if (!roomTrait.exits) {
            return null;
        }
        const exitInfo = roomTrait.exits[direction];
        return exitInfo || null;
    }
    /**
     * Check if an exit exists in a direction
     */
    static hasExit(room, direction) {
        return this.getExit(room, direction) !== undefined;
    }
    /**
     * Check if an exit is blocked
     */
    static isExitBlocked(room, direction) {
        const roomTrait = RoomBehavior.require(room, trait_types_1.TraitType.ROOM);
        return roomTrait.blockedExits?.hasOwnProperty(direction) ?? false;
    }
    /**
     * Get blocked exit message
     */
    static getBlockedMessage(room, direction) {
        const roomTrait = RoomBehavior.require(room, trait_types_1.TraitType.ROOM);
        return roomTrait.blockedExits?.[direction];
    }
    /**
     * Add or update an exit in this room
     */
    static setExit(room, direction, destination, via) {
        const roomTrait = RoomBehavior.require(room, trait_types_1.TraitType.ROOM);
        // Initialize exits if needed
        if (!roomTrait.exits) {
            roomTrait.exits = {};
        }
        roomTrait.exits[direction] = {
            destination,
            via
        };
        // Remove any blocked message for this direction
        if (roomTrait.blockedExits) {
            delete roomTrait.blockedExits[direction];
        }
    }
    /**
     * Block an exit with a message
     */
    static blockExit(room, direction, message) {
        const roomTrait = RoomBehavior.require(room, trait_types_1.TraitType.ROOM);
        if (!roomTrait.blockedExits) {
            roomTrait.blockedExits = {};
        }
        roomTrait.blockedExits[direction] = message;
        return [{
                id: `${Date.now()}-${Math.random()}`,
                timestamp: Date.now(),
                type: if_events_1.IFEvents.MOVEMENT_BLOCKED,
                entities: {
                    location: room.id
                },
                data: {
                    direction: direction,
                    message
                }
            }];
    }
    /**
     * Unblock an exit
     */
    static unblockExit(room, direction) {
        const roomTrait = RoomBehavior.require(room, trait_types_1.TraitType.ROOM);
        if (!roomTrait.blockedExits) {
            return [];
        }
        if (!roomTrait.blockedExits[direction]) {
            return [];
        }
        delete roomTrait.blockedExits[direction];
        return [{
                id: `${Date.now()}-${Math.random()}`,
                timestamp: Date.now(),
                type: if_events_1.IFEvents.NEW_EXIT_REVEALED,
                entities: {
                    location: room.id
                },
                data: {
                    direction: direction,
                    unblocked: true
                }
            }];
    }
    /**
     * Remove an exit from this room
     */
    static removeExit(room, direction) {
        const roomTrait = RoomBehavior.require(room, trait_types_1.TraitType.ROOM);
        // Remove the exit if exits exist
        if (roomTrait.exits) {
            delete roomTrait.exits[direction];
        }
        // Remove any blocked message
        if (roomTrait.blockedExits) {
            delete roomTrait.blockedExits[direction];
        }
    }
    /**
     * Mark room as visited
     */
    static markVisited(room, actor) {
        const roomTrait = RoomBehavior.require(room, trait_types_1.TraitType.ROOM);
        if (roomTrait.visited) {
            return [];
        }
        roomTrait.visited = true;
        return [{
                id: `${Date.now()}-${Math.random()}`,
                timestamp: Date.now(),
                type: if_events_1.IFEvents.ROOM_FIRST_ENTERED,
                entities: {
                    location: room.id,
                    actor: actor.id
                },
                data: {
                    hasInitialDescription: !!roomTrait.initialDescription
                }
            }];
    }
    /**
     * Check if room has been visited
     */
    static hasBeenVisited(room) {
        const roomTrait = RoomBehavior.require(room, trait_types_1.TraitType.ROOM);
        return roomTrait.visited;
    }
    /**
     * Get all exits from the room
     */
    static getAllExits(room) {
        const roomTrait = RoomBehavior.require(room, trait_types_1.TraitType.ROOM);
        if (!roomTrait.exits) {
            return new Map();
        }
        return new Map(Object.entries(roomTrait.exits));
    }
    /**
     * Get available (non-blocked) exits
     */
    static getAvailableExits(room) {
        const roomTrait = RoomBehavior.require(room, trait_types_1.TraitType.ROOM);
        const available = new Map();
        if (!roomTrait.exits) {
            return available;
        }
        for (const [direction, exitInfo] of Object.entries(roomTrait.exits)) {
            if (!this.isExitBlocked(room, direction)) {
                available.set(direction, exitInfo);
            }
        }
        return available;
    }
    /**
     * Check if room is outdoors
     */
    static isOutdoors(room) {
        const roomTrait = RoomBehavior.require(room, trait_types_1.TraitType.ROOM);
        return roomTrait.isOutdoors || false;
    }
    /**
     * Check if room is underground
     */
    static isUnderground(room) {
        const roomTrait = RoomBehavior.require(room, trait_types_1.TraitType.ROOM);
        return roomTrait.isUnderground || false;
    }
    /**
     * Get room region
     */
    static getRegion(room) {
        const roomTrait = RoomBehavior.require(room, trait_types_1.TraitType.ROOM);
        return roomTrait.regionId;
    }
    /**
     * Check if room has a specific tag
     */
    static hasTag(room, tag) {
        const roomTrait = RoomBehavior.require(room, trait_types_1.TraitType.ROOM);
        return roomTrait.tags.includes(tag.toLowerCase());
    }
    /**
     * Add a tag to the room
     */
    static addTag(room, tag) {
        const roomTrait = RoomBehavior.require(room, trait_types_1.TraitType.ROOM);
        const normalizedTag = tag.toLowerCase();
        if (!roomTrait.tags.includes(normalizedTag)) {
            roomTrait.tags.push(normalizedTag);
        }
    }
    /**
     * Remove a tag from the room
     */
    static removeTag(room, tag) {
        const roomTrait = RoomBehavior.require(room, trait_types_1.TraitType.ROOM);
        const normalizedTag = tag.toLowerCase();
        const index = roomTrait.tags.indexOf(normalizedTag);
        if (index >= 0) {
            roomTrait.tags.splice(index, 1);
        }
    }
}
exports.RoomBehavior = RoomBehavior;
