// @sharpee/stdlib/actions/standard/looking/looking.js
"use strict";
/**
 * Looking action - Provides description of current location and visible items
 *
 * Uses four-phase pattern:
 * 1. validate: Always valid (basic sensory action)
 * 2. execute: Mark room as visited (only mutation)
 * 3. report: Generate success events with complete state snapshots
 * 4. blocked: Generate error events (never called since looking always valid)
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.lookingAction = void 0;
const world_model_1 = require("../../../../world-model/index.js");
const constants_1 = require("../../constants");
const emit_illustrations_1 = require("../../helpers/emit-illustrations");
const data_builder_types_1 = require("../../data-builder-types");
const looking_data_1 = require("./looking-data");
exports.lookingAction = {
    id: constants_1.IFActions.LOOKING,
    requiredMessages: [
        'room_description',
        'room_description_brief',
        'room_dark',
        'contents_list',
        'nothing_special',
        'in_container',
        'on_supporter',
        'cant_see_in_dark',
        'look_around',
        'examine_surroundings'
    ],
    validate(context) {
        // Looking is always valid - it's a basic sensory action
        return { valid: true };
    },
    execute(context) {
        // Only mutation: mark room as visited
        const location = context.currentLocation;
        const room = context.world.getContainingRoom(context.player.id);
        if (room && room.hasTrait(world_model_1.TraitType.ROOM)) {
            const roomTrait = room.getTrait(world_model_1.RoomTrait);
            if (roomTrait && !roomTrait.visited) {
                roomTrait.visited = true;
            }
        }
        // No events returned - they're generated in report()
    },
    report(context) {
        // report() is only called on success - looking always succeeds
        const events = [];
        // Build looked event data
        const lookedEventData = (0, data_builder_types_1.buildEventData)(looking_data_1.lookingEventDataConfig, context);
        // Determine message and params
        const isDark = lookedEventData.isDark;
        const { messageId, params } = (0, looking_data_1.determineLookingMessage)(context, isDark);
        // If dark, emit looked event with dark messageId and return early
        if (isDark) {
            events.push(context.event('if.event.looked', {
                messageId: `${context.action.id}.${messageId}`,
                params,
                ...lookedEventData
            }));
            return events;
        }
        // Emit looked event as domain event (no messageId - specialized handler handles room description)
        const lookedEvent = context.event('if.event.looked', lookedEventData);
        events.push(lookedEvent);
        // Build and emit room description event (specialized handler renders this)
        const roomDescData = (0, data_builder_types_1.buildEventData)(looking_data_1.roomDescriptionDataConfig, context);
        events.push(context.event('if.event.room.description', roomDescData));
        // Emit illustration events for the room (ADR-124)
        const room = context.world.getContainingRoom(context.player.id);
        if (room) {
            events.push(...(0, emit_illustrations_1.emitIllustrations)(room, 'on-enter', lookedEvent.id, context));
        }
        // Build list contents data
        const listData = (0, data_builder_types_1.buildEventData)(looking_data_1.listContentsDataConfig, context);
        // Emit contents_list if there are direct items in the room
        if (params.hasItems) {
            events.push(context.event('if.event.list.contents', {
                messageId: `${context.action.id}.contents_list`,
                params,
                ...listData
            }));
        }
        // Emit messages for container/supporter contents
        const openContainerContents = listData.openContainerContents;
        if (openContainerContents && openContainerContents.length > 0) {
            for (const containerInfo of openContainerContents) {
                const contentsMessageId = containerInfo.preposition === 'in'
                    ? 'container_contents'
                    : 'surface_contents';
                const containerKey = containerInfo.preposition === 'in' ? 'container' : 'surface';
                events.push(context.event('if.event.list.contents', {
                    messageId: `${context.action.id}.${contentsMessageId}`,
                    params: {
                        [containerKey]: containerInfo.containerName,
                        items: containerInfo.itemNames.join(', ')
                    },
                    containerId: containerInfo.containerId,
                    containerName: containerInfo.containerName,
                    preposition: containerInfo.preposition,
                    itemIds: containerInfo.itemIds,
                    itemNames: containerInfo.itemNames
                }));
            }
        }
        return events;
    },
    blocked(context, result) {
        // blocked() is called when validation fails
        // Looking always succeeds, so this should never be called
        return [context.event('if.event.looked', {
                blocked: true,
                reason: result.error,
                messageId: `${context.action.id}.${result.error}`,
                params: result.params,
                actorId: context.player.id
            })];
    },
    group: "observation",
    metadata: {
        requiresDirectObject: false,
        requiresIndirectObject: false
    }
};
