// @sharpee/world-model/traits/climbable/climbableBehavior.js
"use strict";
/**
 * Behavior functions for climbable objects
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClimbableBehavior = void 0;
const trait_types_1 = require("../trait-types");
/**
 * Behavior functions for climbable objects
 */
class ClimbableBehavior {
    /**
     * Check if an entity is climbable
     */
    static isClimbable(entity) {
        if (!entity.has(trait_types_1.TraitType.CLIMBABLE)) {
            return false;
        }
        const trait = entity.get(trait_types_1.TraitType.CLIMBABLE);
        return trait.canClimb;
    }
    /**
     * Attempt to climb an object
     */
    static climb(entity) {
        if (!entity.has(trait_types_1.TraitType.CLIMBABLE)) {
            return {
                success: false,
                blocked: true,
                blockedReason: 'not_climbable'
            };
        }
        const trait = entity.get(trait_types_1.TraitType.CLIMBABLE);
        if (!trait.canClimb) {
            return {
                success: false,
                blocked: true,
                blockedReason: trait.blockedMessage || 'cant_climb'
            };
        }
        return {
            success: true,
            destination: trait.destination,
            direction: trait.direction,
            successMessage: trait.successMessage
        };
    }
    /**
     * Get climb direction for an object
     */
    static getDirection(entity) {
        if (!entity.has(trait_types_1.TraitType.CLIMBABLE)) {
            return undefined;
        }
        const trait = entity.get(trait_types_1.TraitType.CLIMBABLE);
        return trait.direction;
    }
    /**
     * Get climb destination for an object
     */
    static getDestination(entity) {
        if (!entity.has(trait_types_1.TraitType.CLIMBABLE)) {
            return undefined;
        }
        const trait = entity.get(trait_types_1.TraitType.CLIMBABLE);
        return trait.destination;
    }
}
exports.ClimbableBehavior = ClimbableBehavior;
