// @sharpee/world-model/traits/climbable/climbableTrait.js
"use strict";
/**
 * Trait for objects that can be climbed (ladders, trees, mountains, etc.)
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClimbableTrait = void 0;
exports.isClimbableTrait = isClimbableTrait;
exports.createClimbableTrait = createClimbableTrait;
const trait_types_1 = require("../trait-types");
/**
 * Trait for climbable objects
 */
class ClimbableTrait {
    static type = trait_types_1.TraitType.CLIMBABLE;
    type = trait_types_1.TraitType.CLIMBABLE;
    /** Whether the object can currently be climbed */
    canClimb;
    /** Optional message when climbing is blocked */
    blockedMessage;
    /** Direction of climbing (up, down, or both) */
    direction;
    /** Destination entity ID when climbing (e.g., top of tree, other side of fence) */
    destination;
    /** Message shown when successfully climbing */
    successMessage;
    constructor(options = {}) {
        this.canClimb = options.canClimb ?? true;
        this.blockedMessage = options.blockedMessage;
        this.direction = options.direction ?? 'both';
        this.destination = options.destination;
        this.successMessage = options.successMessage;
    }
}
exports.ClimbableTrait = ClimbableTrait;
/**
 * Type guard for ClimbableTrait
 */
function isClimbableTrait(trait) {
    return trait.type === trait_types_1.TraitType.CLIMBABLE;
}
/**
 * Factory function for creating ClimbableTrait
 */
function createClimbableTrait(options = {}) {
    return new ClimbableTrait(options);
}
