// @sharpee/world-model/traits/exit/exitBehavior.js
"use strict";
// packages/world-model/src/traits/exit/exitBehavior.ts
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExitBehavior = void 0;
const if_entity_1 = require("../../entities/if-entity");
const trait_types_1 = require("../trait-types");
const exitTrait_1 = require("./exitTrait");
const core_1 = require("../../../core/index.js");
const if_events_1 = require("../../constants/if-events");
/**
 * Behavior for exit entities.
 * Handles all logic related to movement through exits.
 */
class ExitBehavior {
    static requiredTraits = [trait_types_1.TraitType.EXIT];
    /**
     * Check if an exit matches a command
     */
    static matchesCommand(exit, command) {
        const trait = exit.get(trait_types_1.TraitType.EXIT);
        if (!trait)
            return false;
        // Check main command
        if (trait.command.toLowerCase() === command.toLowerCase()) {
            return true;
        }
        // Check direction
        if (trait.direction && trait.direction.toLowerCase() === command.toLowerCase()) {
            return true;
        }
        // Check aliases
        if (trait.aliases) {
            return trait.aliases.some(alias => alias.toLowerCase() === command.toLowerCase());
        }
        return false;
    }
    /**
     * Check if an exit can be used
     */
    static canUse(exit, actor, world) {
        const trait = exit.get(trait_types_1.TraitType.EXIT);
        if (!trait)
            return false;
        // Check if exit is visible
        if (!trait.visible) {
            return false;
        }
        // Check if exit is conditional
        if (trait.conditional && trait.conditionId) {
            // TODO: Check condition with world model
            // This would involve checking game state, items held, etc.
            return true; // For now, allow all conditional exits
        }
        // Check if exit is blocked (door closed, etc.)
        if (exit.has(trait_types_1.TraitType.OPENABLE)) {
            const openable = exit.get(trait_types_1.TraitType.OPENABLE);
            if (openable && !openable.isOpen) {
                return false;
            }
        }
        if (exit.has(trait_types_1.TraitType.LOCKABLE)) {
            const lockable = exit.get(trait_types_1.TraitType.LOCKABLE);
            if (lockable && lockable.isLocked) {
                return false;
            }
        }
        return true;
    }
    /**
     * Use an exit to move between locations
     */
    static use(exit, actor, world) {
        const trait = exit.get(trait_types_1.TraitType.EXIT);
        if (!trait) {
            return [
                (0, core_1.createEvent)(if_events_1.IFEvents.ACTION_FAILED, {
                    action: 'use_exit',
                    reason: 'not_an_exit'
                }, { target: exit.id, actor: actor.id })
            ];
        }
        // Check if can use
        if (!this.canUse(exit, actor, world)) {
            const reason = this.getBlockedReason(exit);
            return [
                (0, core_1.createEvent)(if_events_1.IFEvents.ACTION_FAILED, {
                    action: 'use_exit',
                    reason,
                    message: trait.blockedMessage
                }, { target: exit.id, actor: actor.id })
            ];
        }
        // Move the actor
        const events = [];
        // Custom use message event
        if (trait.useMessage) {
            events.push((0, core_1.createEvent)(if_events_1.IFEvents.CUSTOM_MESSAGE, { message: trait.useMessage }, { actor: actor.id }));
        }
        // Movement event
        events.push((0, core_1.createEvent)(if_events_1.IFEvents.MOVED, {
            from: trait.from,
            to: trait.to,
            via: exit.id,
            direction: trait.direction,
            command: trait.command
        }, { actor: actor.id }));
        return events;
    }
    /**
     * Get reason why exit is blocked
     */
    static getBlockedReason(exit) {
        const trait = exit.get(trait_types_1.TraitType.EXIT);
        if (!trait)
            return 'not_an_exit';
        if (!trait.visible) {
            return 'exit_not_visible';
        }
        if (exit.has(trait_types_1.TraitType.LOCKABLE)) {
            const lockable = exit.get(trait_types_1.TraitType.LOCKABLE);
            if (lockable && lockable.isLocked) {
                return 'exit_locked';
            }
        }
        if (exit.has(trait_types_1.TraitType.OPENABLE)) {
            const openable = exit.get(trait_types_1.TraitType.OPENABLE);
            if (openable && !openable.isOpen) {
                return 'exit_closed';
            }
        }
        if (trait.conditional) {
            return 'condition_not_met';
        }
        return 'exit_blocked';
    }
    /**
     * Get all exits from a location
     */
    static getExitsFrom(locationId, world) {
        // TODO: This should query the world model
        // return world.find({
        //   type: 'exit',
        //   where: e => e.get(TraitType.EXIT)?.from === locationId
        // });
        return [];
    }
    /**
     * Get visible exits from a location
     */
    static getVisibleExitsFrom(locationId, world) {
        return this.getExitsFrom(locationId, world).filter(exit => {
            const trait = exit.get(trait_types_1.TraitType.EXIT);
            return trait && trait.visible;
        });
    }
    /**
     * Get listed exits from a location (for room descriptions)
     */
    static getListedExitsFrom(locationId, world) {
        return this.getExitsFrom(locationId, world).filter(exit => {
            const trait = exit.get(trait_types_1.TraitType.EXIT);
            return trait && trait.visible && trait.listed;
        });
    }
    /**
     * Create a bidirectional exit (creates reverse exit entity)
     */
    static createBidirectional(exitData, world) {
        if (!exitData.from || !exitData.to || !exitData.command) {
            throw new Error('Bidirectional exit requires from, to, and command');
        }
        // Create forward exit
        const forwardId = `${exitData.from}-to-${exitData.to}`;
        const forward = new if_entity_1.IFEntity(forwardId, 'exit');
        forward.add(new exitTrait_1.ExitTrait({
            ...exitData,
            bidirectional: true
        }));
        // Create reverse exit
        const reverseId = `${exitData.to}-to-${exitData.from}`;
        const reverse = new if_entity_1.IFEntity(reverseId, 'exit');
        const reverseData = {
            from: exitData.to,
            to: exitData.from,
            command: exitData.reverseCommand || this.getReverseCommand(exitData.command),
            direction: exitData.reverseDirection || this.getReverseDirection(exitData.direction),
            visible: exitData.visible,
            listed: exitData.listed,
            bidirectional: true
        };
        reverse.add(new exitTrait_1.ExitTrait(reverseData));
        return [forward, reverse];
    }
    /**
     * Get reverse direction for standard directions
     */
    static getReverseDirection(direction) {
        if (!direction)
            return undefined;
        const reverseMap = {
            'north': 'south',
            'south': 'north',
            'east': 'west',
            'west': 'east',
            'northeast': 'southwest',
            'northwest': 'southeast',
            'southeast': 'northwest',
            'southwest': 'northeast',
            'up': 'down',
            'down': 'up',
            'in': 'out',
            'out': 'in'
        };
        return reverseMap[direction.toLowerCase()];
    }
    /**
     * Get reverse command for standard commands
     */
    static getReverseCommand(command) {
        // If it's a direction command, reverse it
        const direction = command.replace(/^go\s+/, '');
        const reverse = this.getReverseDirection(direction);
        if (reverse) {
            return command.startsWith('go ') ? `go ${reverse}` : reverse;
        }
        // For custom commands, append "back"
        return `go back`;
    }
}
exports.ExitBehavior = ExitBehavior;
