// @sharpee/world-model/traits/exit/exitTrait.js
"use strict";
// packages/world-model/src/traits/exit/exitTrait.ts
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExitTrait = void 0;
const trait_types_1 = require("../trait-types");
/**
 * Exit trait for entities that represent passages between locations.
 * Used for doors, passages, portals, and any custom exits like "xyzzy".
 *
 * This trait contains only data - all behavior is in ExitBehavior.
 */
class ExitTrait {
    static type = trait_types_1.TraitType.EXIT;
    type = trait_types_1.TraitType.EXIT;
    /** Source location ID (must be an entity ID, not a name) */
    from;
    /** Destination location ID (must be an entity ID, not a name) */
    to;
    /** Standard direction (north, south, up, etc.) - optional */
    direction;
    /** Command to use this exit (e.g., "go north", "enter portal", "xyzzy") */
    command;
    /** Alternative commands that work for this exit */
    aliases;
    /** Whether this exit is visible to players */
    visible = true;
    /** Whether this exit appears in room descriptions */
    listed = true;
    /** Whether this exit works in reverse (bidirectional) */
    bidirectional = false;
    /** The reverse command if bidirectional */
    reverseCommand;
    /** The reverse direction if bidirectional */
    reverseDirection;
    /** Custom message when using this exit */
    useMessage;
    /** Custom message when this exit is blocked */
    blockedMessage;
    /** Whether this exit requires special conditions */
    conditional = false;
    /** Condition identifier (checked by behaviors) */
    conditionId;
    constructor(data) {
        if (!data.from || !data.to || !data.command) {
            throw new Error('ExitTrait requires from, to, and command');
        }
        this.from = data.from;
        this.to = data.to;
        this.command = data.command;
        if (data.direction !== undefined)
            this.direction = data.direction;
        if (data.aliases !== undefined)
            this.aliases = data.aliases;
        if (data.visible !== undefined)
            this.visible = data.visible;
        if (data.listed !== undefined)
            this.listed = data.listed;
        if (data.bidirectional !== undefined)
            this.bidirectional = data.bidirectional;
        if (data.reverseCommand !== undefined)
            this.reverseCommand = data.reverseCommand;
        if (data.reverseDirection !== undefined)
            this.reverseDirection = data.reverseDirection;
        if (data.useMessage !== undefined)
            this.useMessage = data.useMessage;
        if (data.blockedMessage !== undefined)
            this.blockedMessage = data.blockedMessage;
        if (data.conditional !== undefined)
            this.conditional = data.conditional;
        if (data.conditionId !== undefined)
            this.conditionId = data.conditionId;
    }
}
exports.ExitTrait = ExitTrait;
