// @sharpee/ext-testing/annotations/context.js
"use strict";
/**
 * Annotation Context Capture
 *
 * Utility to capture current game state for annotations.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.captureContext = captureContext;
exports.createEmptyContext = createEmptyContext;
/**
 * Capture the current game state for annotation context
 */
function captureContext(world, lastCommand, lastResponse) {
    const player = world.getPlayer();
    const locationId = player ? world.getLocation(player.id) : undefined;
    const location = locationId ? world.getEntity(locationId) : undefined;
    const inventory = player ? world.getContents(player.id) : [];
    return {
        roomId: locationId ?? 'unknown',
        roomName: location?.name ?? locationId ?? 'unknown',
        turn: world.getStateValue('turnCount') ?? world.getStateValue('moves') ?? 0,
        score: world.getStateValue('score') ?? 0,
        lastCommand,
        lastResponse,
        inventory: inventory.map((e) => e.name ?? e.id),
    };
}
/**
 * Create an empty/default context (for when no command has been executed yet)
 */
function createEmptyContext() {
    return {
        roomId: 'unknown',
        roomName: 'unknown',
        turn: 0,
        score: 0,
        lastCommand: '',
        lastResponse: '',
        inventory: [],
    };
}
