// @sharpee/character/influence/index.js
"use strict";
/**
 * NPC influence exports (ADR-146)
 *
 * Influence definitions, evaluation engine, duration tracking,
 * and PC influence handling.
 *
 * Public interface: All re-exported types, classes, and functions.
 * Owner context: @sharpee/character / influence
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.InfluenceMessages = exports.InfluenceBuilder = exports.evaluatePcInfluence = exports.InfluenceTracker = exports.evaluateActiveInfluence = exports.evaluatePassiveInfluences = exports.checkResistance = void 0;
var influence_evaluator_1 = require("./influence-evaluator");
Object.defineProperty(exports, "checkResistance", { enumerable: true, get: function () { return influence_evaluator_1.checkResistance; } });
Object.defineProperty(exports, "evaluatePassiveInfluences", { enumerable: true, get: function () { return influence_evaluator_1.evaluatePassiveInfluences; } });
Object.defineProperty(exports, "evaluateActiveInfluence", { enumerable: true, get: function () { return influence_evaluator_1.evaluateActiveInfluence; } });
var influence_duration_1 = require("./influence-duration");
Object.defineProperty(exports, "InfluenceTracker", { enumerable: true, get: function () { return influence_duration_1.InfluenceTracker; } });
var pc_influence_1 = require("./pc-influence");
Object.defineProperty(exports, "evaluatePcInfluence", { enumerable: true, get: function () { return pc_influence_1.evaluatePcInfluence; } });
var builder_1 = require("./builder");
Object.defineProperty(exports, "InfluenceBuilder", { enumerable: true, get: function () { return builder_1.InfluenceBuilder; } });
var influence_messages_1 = require("./influence-messages");
Object.defineProperty(exports, "InfluenceMessages", { enumerable: true, get: function () { return influence_messages_1.InfluenceMessages; } });
