// @sharpee/transcript-tester/fast-cli.js
#!/usr/bin/env node
"use strict";
/**
 * Fast Transcript Tester CLI
 *
 * Uses pre-bundled platform (dist/cli/sharpee.js) and story for instant loading.
 * Supports --chain flag for walkthrough testing.
 *
 * Usage:
 *   node packages/transcript-tester/dist/fast-cli.js [transcript-files...] [options]
 *   node packages/transcript-tester/dist/fast-cli.js --chain wt-*.transcript
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
const path = __importStar(require("path"));
const fs = __importStar(require("fs"));
const readline = __importStar(require("readline"));
const parser_1 = require("./parser");
const runner_1 = require("./runner");
const reporter_1 = require("./reporter");
// Load the pre-bundled platform
// __dirname is packages/transcript-tester/dist, so go up 3 levels to repo root
const bundlePath = path.resolve(__dirname, '..', '..', '..', 'dist', 'sharpee.js');
const platform = require(bundlePath);
// Extract what we need from the bundle
const { GameEngine, WorldModel, EntityType, Parser, LanguageProvider, PerceptionService, TestingExtension, renderToString } = platform;
/**
 * Parse command line arguments
 */
function parseArgs(args) {
    const options = {
        transcriptPaths: [],
        verbose: false,
        emitTraits: false,
        stopOnFailure: false,
        chain: false,
        play: false,
        worldJson: false,
        restore: null,
        storyPath: 'stories/dungeo'
    };
    let i = 0;
    while (i < args.length) {
        const arg = args[i];
        if (arg === '--verbose' || arg === '-v') {
            options.verbose = true;
        }
        else if (arg === '--emit-traits') {
            options.emitTraits = true;
            options.verbose = true; // --emit-traits implies --verbose
        }
        else if (arg === '--stop-on-failure' || arg === '-s') {
            options.stopOnFailure = true;
        }
        else if (arg === '--chain' || arg === '-c') {
            options.chain = true;
        }
        else if (arg === '--play' || arg === '-p') {
            options.play = true;
        }
        else if (arg === '--world-json') {
            options.worldJson = true;
        }
        else if (arg.startsWith('--restore=')) {
            options.restore = arg.split('=')[1];
            options.play = true; // --restore implies --play
        }
        else if (arg === '--restore') {
            i++;
            if (i < args.length) {
                options.restore = args[i];
                options.play = true;
            }
        }
        else if (arg === '--story') {
            i++;
            if (i < args.length) {
                options.storyPath = args[i];
            }
        }
        else if (arg === '--help' || arg === '-h') {
            printHelp();
            process.exit(0);
        }
        else if (!arg.startsWith('-')) {
            options.transcriptPaths.push(arg);
        }
        i++;
    }
    return options;
}
/**
 * Print help message
 */
function printHelp() {
    console.log(`
Fast Transcript Tester - Uses pre-bundled platform for instant loading

Usage:
  fast-transcript-test [transcript-files...] [options]

Arguments:
  transcript-files   One or more .transcript files to run

Options:
  -c, --chain            Chain transcripts (don't reset game state between them)
  -v, --verbose          Show detailed output for each command
  --emit-traits          Show entity traits for objects referenced in events (implies --verbose)
  -s, --stop-on-failure  Stop on first failure
  -p, --play             Interactive play mode (REPL)
  --world-json           Dump initialized world model as JSON to stdout and exit
  --restore <name>       Restore from save file and enter play mode
  --story <path>         Story path (default: stories/dungeo)
  -h, --help             Show this help message

Examples:
  fast-transcript-test stories/dungeo/walkthroughs/wt-01-get-torch-early.transcript
  fast-transcript-test --chain stories/dungeo/walkthroughs/wt-*.transcript
  fast-transcript-test --play
  fast-transcript-test --restore wt-11
`);
}
/**
 * Load the story and create a testable game
 */
function loadStoryAndCreateGame(storyPath) {
    // Resolve story path
    const resolvedPath = path.isAbsolute(storyPath)
        ? storyPath
        : path.resolve(process.cwd(), storyPath);
    // Load the story module
    const distPath = path.join(resolvedPath, 'dist', 'index.js');
    const storyModule = require(distPath);
    const story = storyModule.story || storyModule.default;
    if (!story) {
        throw new Error(`Story module at ${storyPath} does not export 'story' or 'default'`);
    }
    // Create world and player
    const world = new WorldModel();
    const player = world.createEntity('player', EntityType.ACTOR);
    world.setPlayer(player.id);
    // Create parser and language
    const language = new LanguageProvider();
    const parser = new Parser(language);
    // Extend parser and language with story-specific vocabulary
    if (story.extendParser) {
        story.extendParser(parser);
    }
    if (story.extendLanguage) {
        story.extendLanguage(language);
    }
    // Create perception service
    const perceptionService = new PerceptionService();
    // Create engine
    const engine = new GameEngine({
        world,
        player,
        parser,
        language,
        perceptionService,
    });
    // Set the story and start
    engine.setStory(story);
    engine.start();
    // Create testing extension
    const testingExtension = TestingExtension ? new TestingExtension() : null;
    // Capture text output and events
    let lastOutput = '';
    let outputBuffer = [];
    let lastEvents = [];
    let lastTurnResult = null;
    engine.on('text:output', (blocks) => {
        outputBuffer.push(renderToString(blocks));
    });
    let eventBuffer = [];
    engine.on('event', (event) => {
        eventBuffer.push(event);
    });
    const testableGame = {
        engine,
        world,
        testingExtension,
        lastOutput: '',
        lastEvents: [],
        lastTurnResult: null,
        async executeCommand(input) {
            outputBuffer = [];
            eventBuffer = [];
            lastEvents = [];
            lastTurnResult = null;
            try {
                const result = await engine.executeTurn(input);
                if (result) {
                    lastTurnResult = result;
                    lastEvents = eventBuffer;
                }
            }
            catch (error) {
                const errorMessage = error instanceof Error ? error.message : String(error);
                outputBuffer.push(`Error: ${errorMessage}`);
            }
            lastOutput = outputBuffer.join('\n');
            testableGame.lastOutput = lastOutput;
            testableGame.lastEvents = lastEvents;
            testableGame.lastTurnResult = lastTurnResult;
            return lastOutput;
        },
    };
    return testableGame;
}
/**
 * Run interactive play mode (REPL)
 */
async function runInteractiveMode(game) {
    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout
    });
    let debugMode = false;
    let traceMode = false;
    console.log('\n--- Interactive Mode (Fast) ---');
    console.log('Type commands to play. Special commands:');
    console.log('  /quit, /q    - Exit the game');
    console.log('  /debug       - Toggle debug mode (show events + parsed/validated JSON)');
    console.log('  /trace       - Toggle parser trace mode (PARSER_DEBUG env)');
    console.log('  /events      - Show events from last command');
    console.log('  /parsed      - Show parsed command from last turn');
    console.log('  /validated   - Show validated command from last turn');
    console.log('  /look, /l    - Shortcut for "look"');
    console.log('  /inv, /i     - Shortcut for "inventory"');
    console.log('');
    const initialOutput = await game.executeCommand('look');
    console.log(initialOutput);
    const prompt = () => {
        rl.question('\n> ', async (input) => {
            const trimmed = input.trim();
            if (!trimmed) {
                prompt();
                return;
            }
            if (trimmed === '/quit' || trimmed === '/q') {
                console.log('Goodbye!');
                rl.close();
                process.exit(0);
                return;
            }
            if (trimmed === '/debug') {
                debugMode = !debugMode;
                console.log(`Debug mode: ${debugMode ? 'ON' : 'OFF'}`);
                prompt();
                return;
            }
            if (trimmed === '/trace') {
                traceMode = !traceMode;
                process.env.PARSER_DEBUG = traceMode ? 'true' : '';
                console.log(`Parser trace: ${traceMode ? 'ON' : 'OFF'}`);
                prompt();
                return;
            }
            if (trimmed === '/events') {
                if (game.lastEvents && game.lastEvents.length > 0) {
                    console.log('\nEvents from last command:');
                    for (const event of game.lastEvents) {
                        console.log(`  ${event.type}`);
                        if (event.data && Object.keys(event.data).length > 0) {
                            console.log(`    ${JSON.stringify(event.data, null, 2).split('\n').join('\n    ')}`);
                        }
                    }
                }
                else {
                    console.log('(No events from last command)');
                }
                prompt();
                return;
            }
            if (trimmed === '/parsed') {
                if (game.lastTurnResult && game.lastTurnResult.parsedCommand) {
                    console.log('\nParsed command:');
                    console.log(JSON.stringify(game.lastTurnResult.parsedCommand, null, 2));
                }
                else {
                    console.log('(No parsed command from last turn)');
                }
                prompt();
                return;
            }
            if (trimmed === '/validated') {
                if (game.lastTurnResult && game.lastTurnResult.validatedCommand) {
                    console.log('\nValidated command:');
                    console.log(JSON.stringify(game.lastTurnResult.validatedCommand, null, 2));
                }
                else {
                    console.log('(No validated command from last turn)');
                }
                prompt();
                return;
            }
            let command = trimmed;
            if (trimmed === '/look' || trimmed === '/l') {
                command = 'look';
            }
            else if (trimmed === '/inv' || trimmed === '/i') {
                command = 'inventory';
            }
            try {
                const output = await game.executeCommand(command);
                console.log(output);
                if (debugMode) {
                    if (game.lastTurnResult) {
                        if (game.lastTurnResult.parsedCommand) {
                            console.log('\n[Parsed]');
                            console.log(JSON.stringify(game.lastTurnResult.parsedCommand, null, 2));
                        }
                        if (game.lastTurnResult.validatedCommand) {
                            console.log('\n[Validated]');
                            console.log(JSON.stringify(game.lastTurnResult.validatedCommand, null, 2));
                        }
                    }
                    if (game.lastEvents && game.lastEvents.length > 0) {
                        console.log('\n[Events]');
                        for (const event of game.lastEvents) {
                            const data = event.data && Object.keys(event.data).length > 0
                                ? ` ${JSON.stringify(event.data)}`
                                : '';
                            console.log(`  ${event.type}${data}`);
                        }
                    }
                }
            }
            catch (error) {
                console.error(`Error: ${error instanceof Error ? error.message : error}`);
            }
            prompt();
        });
    };
    prompt();
}
/**
 * Main entry point
 */
async function main() {
    const args = process.argv.slice(2);
    if (args.length === 0) {
        printHelp();
        process.exit(1);
    }
    const options = parseArgs(args);
    // World JSON dump mode
    if (options.worldJson) {
        const game = loadStoryAndCreateGame(options.storyPath);
        const world = game.world;
        const allEntities = world.getAllEntities();
        const rooms = [];
        const entities = [];
        const npcs = [];
        for (const entity of allEntities) {
            const identity = entity.get('identity');
            const name = identity?.name || entity.id;
            const location = world.getLocation(entity.id) || null;
            const traitTypes = entity.getTraitTypes();
            if (traitTypes.includes('room')) {
                const roomTrait = entity.get('room');
                const exits = {};
                if (roomTrait?.exits) {
                    for (const [dir, exitData] of Object.entries(roomTrait.exits)) {
                        const dest = exitData?.destination;
                        if (dest) {
                            // Resolve destination name
                            const destEntity = world.getEntity(dest);
                            const destIdentity = destEntity?.get('identity');
                            exits[dir] = dest;
                            if (destIdentity?.name) {
                                exits[dir + '_name'] = destIdentity.name;
                            }
                        }
                    }
                }
                rooms.push({
                    id: entity.id,
                    name,
                    aliases: identity?.aliases || [],
                    isDark: roomTrait?.isDark || false,
                    regionId: roomTrait?.regionId || null,
                    exits,
                });
            }
            else if (traitTypes.includes('actor') && !traitTypes.includes('player')) {
                const npcTrait = entity.get('npc');
                npcs.push({
                    id: entity.id,
                    name,
                    location,
                    traits: traitTypes,
                    behaviorId: npcTrait?.behaviorId || null,
                });
            }
            else if (entity.type !== 'player'
                && !traitTypes.includes('region')
                && !traitTypes.includes('scene')) {
                entities.push({
                    id: entity.id,
                    name,
                    location,
                    traits: traitTypes,
                });
            }
        }
        // Collect regions (ADR-149)
        const regions = [];
        for (const entity of allEntities) {
            const regionTrait = entity.get('region');
            if (regionTrait) {
                regions.push({
                    id: entity.id,
                    name: regionTrait.name,
                    parentRegionId: regionTrait.parentRegionId || null,
                });
            }
        }
        // Collect scenes (ADR-149)
        const scenes = [];
        for (const entity of allEntities) {
            const sceneTrait = entity.get('scene');
            if (sceneTrait) {
                scenes.push({
                    id: entity.id,
                    name: sceneTrait.name,
                    state: sceneTrait.state,
                    recurring: sceneTrait.recurring,
                });
            }
        }
        // Engine introspection — actions with patterns and metadata
        const introspection = game.engine.introspect();
        const output = {
            storyPath: options.storyPath,
            rooms,
            entities,
            npcs,
            actions: introspection.actions,
            traits: introspection.traits,
            behaviors: introspection.behaviors,
            regions,
            scenes,
        };
        process.stdout.write(JSON.stringify(output, null, 2));
        process.exit(0);
    }
    // Interactive play mode
    if (options.play) {
        console.log(`Loading story from: ${options.storyPath} (using bundle)`);
        const game = loadStoryAndCreateGame(options.storyPath);
        // Restore from save file if requested
        if (options.restore) {
            const savesDir = path.join(options.storyPath, 'saves');
            const savePath = path.join(savesDir, `${options.restore}.json`);
            if (!fs.existsSync(savePath)) {
                console.error(`Save file not found: ${savePath}`);
                console.error(`Available saves:`);
                if (fs.existsSync(savesDir)) {
                    const files = fs.readdirSync(savesDir).filter(f => f.endsWith('.json'));
                    for (const f of files) {
                        console.error(`  ${f.replace('.json', '')}`);
                    }
                }
                else {
                    console.error(`  (no saves directory — run --chain walkthroughs first)`);
                }
                process.exit(1);
            }
            console.log(`Restoring from: ${savePath}`);
            const worldState = fs.readFileSync(savePath, 'utf-8');
            game.world.loadJSON(worldState);
            console.log(`Restored save: ${options.restore}`);
        }
        await runInteractiveMode(game);
        return;
    }
    if (options.transcriptPaths.length === 0) {
        console.error('Error: No transcript files specified');
        printHelp();
        process.exit(1);
    }
    console.log(`Loading story from: ${options.storyPath} (using bundle)`);
    console.log(`Found ${options.transcriptPaths.length} transcript(s) to run`);
    if (options.chain) {
        console.log(`Chain mode: Game state will persist between transcripts`);
    }
    // Load the game once (will reload for each transcript unless chaining)
    let game = loadStoryAndCreateGame(options.storyPath);
    // Run all transcripts
    const results = [];
    for (const transcriptPath of options.transcriptPaths) {
        // Parse the transcript
        const transcript = (0, parser_1.parseTranscriptFile)(transcriptPath);
        // Validate
        const errors = (0, parser_1.validateTranscript)(transcript);
        if (errors.length > 0) {
            console.error(`\nErrors in ${transcriptPath}:`);
            for (const err of errors) {
                console.error(`  - ${err}`);
            }
            continue;
        }
        // Reload story for each transcript to reset state (unless chaining)
        if (!options.chain) {
            game = loadStoryAndCreateGame(options.storyPath);
        }
        // Run the transcript with saves directory based on story path
        const savesDirectory = path.join(options.storyPath, 'saves');
        const result = await (0, runner_1.runTranscript)(transcript, game, {
            verbose: options.verbose,
            emitTraits: options.emitTraits,
            stopOnFailure: options.stopOnFailure,
            savesDirectory,
            testingExtension: game.testingExtension
        });
        results.push(result);
        // Report individual transcript results
        (0, reporter_1.reportTranscript)(result, { verbose: options.verbose, emitTraits: options.emitTraits });
        // Stop if requested and there was a failure
        if (options.stopOnFailure && result.failed > 0) {
            break;
        }
    }
    // Aggregate results
    const runResult = {
        transcripts: results,
        totalPassed: results.reduce((sum, r) => sum + r.passed, 0),
        totalFailed: results.reduce((sum, r) => sum + r.failed, 0),
        totalExpectedFailures: results.reduce((sum, r) => sum + r.expectedFailures, 0),
        totalSkipped: results.reduce((sum, r) => sum + r.skipped, 0),
        totalDuration: results.reduce((sum, r) => sum + r.duration, 0)
    };
    // Final report if multiple transcripts
    if (results.length > 1) {
        (0, reporter_1.reportTestRun)(runResult, { verbose: options.verbose });
    }
    // Exit with appropriate code
    process.exit((0, reporter_1.getExitCode)(runResult));
}
// Run
main().catch(error => {
    console.error('Fatal error:', error);
    process.exit(1);
});
