// @sharpee/transcript-tester/index.js
"use strict";
/**
 * @sharpee/transcript-tester
 *
 * Transcript-based testing for Sharpee interactive fiction stories.
 *
 * Usage:
 *   npx transcript-test <story-path> [transcripts...]
 *
 * See ADR-073 for format specification.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.findTranscripts = exports.createTestableGame = exports.loadStory = exports.formatTraitProse = exports.formatEntityTraitLines = exports.getExitCode = exports.reportTestRun = exports.reportTranscript = exports.runTranscript = exports.validateTranscript = exports.parseTranscriptFile = exports.parseTranscript = void 0;
// Types
__exportStar(require("./types"), exports);
// Parser
var parser_1 = require("./parser");
Object.defineProperty(exports, "parseTranscript", { enumerable: true, get: function () { return parser_1.parseTranscript; } });
Object.defineProperty(exports, "parseTranscriptFile", { enumerable: true, get: function () { return parser_1.parseTranscriptFile; } });
Object.defineProperty(exports, "validateTranscript", { enumerable: true, get: function () { return parser_1.validateTranscript; } });
// Runner
var runner_1 = require("./runner");
Object.defineProperty(exports, "runTranscript", { enumerable: true, get: function () { return runner_1.runTranscript; } });
// Reporter
var reporter_1 = require("./reporter");
Object.defineProperty(exports, "reportTranscript", { enumerable: true, get: function () { return reporter_1.reportTranscript; } });
Object.defineProperty(exports, "reportTestRun", { enumerable: true, get: function () { return reporter_1.reportTestRun; } });
Object.defineProperty(exports, "getExitCode", { enumerable: true, get: function () { return reporter_1.getExitCode; } });
// Trait Formatter
var trait_formatter_1 = require("./trait-formatter");
Object.defineProperty(exports, "formatEntityTraitLines", { enumerable: true, get: function () { return trait_formatter_1.formatEntityTraitLines; } });
Object.defineProperty(exports, "formatTraitProse", { enumerable: true, get: function () { return trait_formatter_1.formatTraitProse; } });
// Story Loader
var story_loader_1 = require("./story-loader");
Object.defineProperty(exports, "loadStory", { enumerable: true, get: function () { return story_loader_1.loadStory; } });
Object.defineProperty(exports, "createTestableGame", { enumerable: true, get: function () { return story_loader_1.createTestableGame; } });
Object.defineProperty(exports, "findTranscripts", { enumerable: true, get: function () { return story_loader_1.findTranscripts; } });
