// @sharpee/transcript-tester/story-loader.js
"use strict";
/**
 * Story Loader
 *
 * Dynamically loads and initializes a story for testing.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.loadStory = loadStory;
exports.createTestableGame = createTestableGame;
exports.findTranscripts = findTranscripts;
const path = __importStar(require("path"));
const engine_1 = require("../engine/index.js");
const text_service_1 = require("../text-service/index.js");
const world_model_1 = require("../world-model/index.js");
const parser_en_us_1 = require("../parser-en-us/index.js");
const stdlib_1 = require("../stdlib/index.js");
// @ts-ignore
const lang_en_us_1 = require("../lang-en-us/index.js");
/**
 * Load a story from a path and create a testable game instance
 */
async function loadStory(storyPath) {
    // Resolve the story path
    const resolvedPath = path.isAbsolute(storyPath)
        ? storyPath
        : path.resolve(process.cwd(), storyPath);
    // Try to load the story module
    let storyModule;
    try {
        // Try loading the compiled dist version first
        const distPath = path.join(resolvedPath, 'dist', 'index.js');
        storyModule = require(distPath);
    }
    catch (e) {
        try {
            // Fall back to src (for ts-node environments)
            const srcPath = path.join(resolvedPath, 'src', 'index.ts');
            storyModule = require(srcPath);
        }
        catch (e2) {
            throw new Error(`Could not load story from ${storyPath}: ${e}`);
        }
    }
    const story = storyModule.story || storyModule.default;
    if (!story) {
        throw new Error(`Story module at ${storyPath} does not export 'story' or 'default'`);
    }
    // Create the game instance
    return createTestableGame(story);
}
/**
 * Create a testable game from a story instance
 */
function createTestableGame(story) {
    // Create world and player
    const world = new world_model_1.WorldModel();
    const player = world.createEntity('player', world_model_1.EntityType.ACTOR);
    world.setPlayer(player.id);
    // Create parser and language
    const language = new lang_en_us_1.LanguageProvider();
    const parser = new parser_en_us_1.Parser(language);
    // Extend parser and language with story-specific vocabulary
    if (story.extendParser) {
        story.extendParser(parser);
    }
    if (story.extendLanguage) {
        story.extendLanguage(language);
    }
    // Create perception service
    const perceptionService = new stdlib_1.PerceptionService();
    // Create engine (TextService created internally from language provider)
    const engine = new engine_1.GameEngine({
        world,
        player,
        parser,
        language,
        perceptionService,
    });
    // Set the story and start
    engine.setStory(story);
    engine.start();
    // Capture text output and events
    let lastOutput = '';
    let outputBuffer = [];
    let lastEvents = [];
    let lastTurnResult = null;
    engine.on('text:output', (blocks) => {
        outputBuffer.push((0, text_service_1.renderToString)(blocks));
    });
    // Capture ALL events through the event emitter (includes scheduler/NPC events)
    let eventBuffer = [];
    engine.on('event', (event) => {
        eventBuffer.push(event);
    });
    // Create the testable game interface
    const testableGame = {
        engine,
        world,
        lastOutput: '',
        lastEvents: [],
        lastTurnResult: null,
        async executeCommand(input) {
            outputBuffer = [];
            eventBuffer = []; // Reset event buffer for this command
            lastEvents = [];
            lastTurnResult = null;
            try {
                const result = await engine.executeTurn(input);
                if (result) {
                    lastTurnResult = result;
                    // Use eventBuffer which captures ALL events (action + NPC + scheduler)
                    lastEvents = eventBuffer;
                }
            }
            catch (error) {
                const errorMessage = error instanceof Error ? error.message : String(error);
                outputBuffer.push(`Error: ${errorMessage}`);
            }
            lastOutput = outputBuffer.join('\n');
            testableGame.lastOutput = lastOutput;
            testableGame.lastEvents = lastEvents;
            testableGame.lastTurnResult = lastTurnResult;
            return lastOutput;
        },
    };
    return testableGame;
}
/**
 * Find all transcript files in a directory
 */
function findTranscripts(dir, pattern = '*.transcript') {
    const glob = require('glob');
    const resolvedDir = path.isAbsolute(dir) ? dir : path.resolve(process.cwd(), dir);
    const files = glob.sync(path.join(resolvedDir, '**', pattern));
    return files;
}
