// @sharpee/core/query/types.js
"use strict";
/**
 * Core types for the PC communication query system
 *
 * These types define the structure of queries, responses, and validation
 * for the client query system (ADR-018).
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.StandardValidators = exports.QueryType = exports.QuerySource = void 0;
/**
 * Sources that can initiate queries
 */
var QuerySource;
(function (QuerySource) {
    /** System-level queries (save, quit, restart) */
    QuerySource["SYSTEM"] = "system";
    /** Parser disambiguation queries */
    QuerySource["DISAMBIGUATION"] = "disambiguation";
    /** NPC dialogue queries */
    QuerySource["NPC"] = "npc";
    /** Game mechanic queries (passwords, combinations) */
    QuerySource["GAME_MECHANIC"] = "mechanic";
    /** Narrative-driven queries */
    QuerySource["NARRATIVE"] = "narrative";
})(QuerySource || (exports.QuerySource = QuerySource = {}));
/**
 * Types of queries that can be presented
 */
var QueryType;
(function (QueryType) {
    /** Simple yes/no question */
    QueryType["YES_NO"] = "yes_no";
    /** Multiple choice with predefined options */
    QueryType["MULTIPLE_CHOICE"] = "multiple_choice";
    /** Open-ended text input */
    QueryType["FREE_TEXT"] = "free_text";
    /** Numeric input only */
    QueryType["NUMERIC"] = "numeric";
    /** Special type for disambiguation */
    QueryType["DISAMBIGUATION"] = "disambiguation";
})(QueryType || (exports.QueryType = QueryType = {}));
/**
 * Standard validators
 */
exports.StandardValidators = {
    /** Validate yes/no responses */
    yesNo: (response) => {
        const normalized = response.toLowerCase().trim();
        const yesVariants = ['yes', 'y', 'yeah', 'yep', 'sure', 'ok', 'okay'];
        const noVariants = ['no', 'n', 'nope', 'nah', 'cancel'];
        if (yesVariants.includes(normalized)) {
            return { valid: true, normalized: true };
        }
        if (noVariants.includes(normalized)) {
            return { valid: true, normalized: false };
        }
        return {
            valid: false,
            message: 'Please answer yes or no.',
            hint: 'You can also use y/n'
        };
    },
    /** Validate numeric responses */
    numeric: (response, min, max) => {
        const num = parseInt(response.trim(), 10);
        if (isNaN(num)) {
            return {
                valid: false,
                message: 'Please enter a number.'
            };
        }
        if (min !== undefined && num < min) {
            return {
                valid: false,
                message: `Please enter a number at least ${min}.`
            };
        }
        if (max !== undefined && num > max) {
            return {
                valid: false,
                message: `Please enter a number no more than ${max}.`
            };
        }
        return { valid: true, normalized: num };
    },
    /** Validate multiple choice responses */
    multipleChoice: (response, options) => {
        const trimmed = response.trim();
        // Check if it's a number
        const num = parseInt(trimmed, 10);
        if (!isNaN(num) && num >= 1 && num <= options.length) {
            return { valid: true, normalized: num - 1 };
        }
        // Check if it matches an option (case insensitive)
        const index = options.findIndex(opt => opt.toLowerCase().includes(trimmed.toLowerCase()) ||
            trimmed.toLowerCase().includes(opt.toLowerCase()));
        if (index >= 0) {
            return { valid: true, normalized: index };
        }
        return {
            valid: false,
            message: `Please choose a number 1-${options.length} or type part of an option.`
        };
    }
};
