// @sharpee/event-processor/handlers/movement.js
"use strict";
/**
 * Movement event handlers
 *
 * These handlers apply movement-related events to the world model
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.handleClimbed = exports.handleExited = exports.handleEntered = exports.handleRemovedFrom = exports.handlePutOn = exports.handlePutIn = exports.handleActorMoved = exports.handleRemoved = exports.handleDropped = exports.handleTaken = void 0;
exports.registerMovementHandlers = registerMovementHandlers;
const world_model_1 = require("../../world-model/index.js");
const if_domain_1 = require("../../if-domain/index.js");
/**
 * Handle TAKEN event - move item to actor
 */
const handleTaken = (event, world) => {
    const { actor, target } = event.entities;
    if (actor && target) {
        // Remove worn status if applicable
        const targetEntity = world.getEntity(target);
        if (targetEntity && targetEntity.has(world_model_1.TraitType.WEARABLE)) {
            // Remove worn status if wearable
            world.updateEntity(target, (entity) => {
                const wearableTrait = entity.get(world_model_1.TraitType.WEARABLE);
                if (wearableTrait?.worn) {
                    wearableTrait.worn = false;
                }
            });
        }
        // Move the item to the actor
        world.moveEntity(target, actor);
    }
};
exports.handleTaken = handleTaken;
/**
 * Handle DROPPED event - move item to location
 */
const handleDropped = (event, world) => {
    const { actor, target, location } = event.entities;
    const dropLocation = location || (actor ? world.getLocation(actor) : undefined);
    if (target && dropLocation) {
        world.moveEntity(target, dropLocation);
    }
};
exports.handleDropped = handleDropped;
/**
 * Handle REMOVED event - remove worn status
 */
const handleRemoved = (event, world) => {
    const { target } = event.entities;
    if (target) {
        const entity = world.getEntity(target);
        if (entity && entity.has(world_model_1.TraitType.WEARABLE)) {
            // Remove worn status
            world.updateEntity(target, (entity) => {
                const wearableTrait = entity.get(world_model_1.TraitType.WEARABLE);
                if (wearableTrait?.worn) {
                    wearableTrait.worn = false;
                }
            });
        }
    }
};
exports.handleRemoved = handleRemoved;
/**
 * Handle ACTOR_MOVED event - move actor to new location
 */
const handleActorMoved = (event, world) => {
    const { actor, location } = event.entities;
    if (actor && location) {
        // Mark previous location as visited if it's a room
        const currentLocation = world.getLocation(actor);
        if (currentLocation) {
            const currentRoom = world.getEntity(currentLocation);
            if (currentRoom && currentRoom.has(world_model_1.TraitType.ROOM)) {
                // Mark as visited
                world.updateEntity(currentLocation, (room) => {
                    const roomTrait = room.get(world_model_1.TraitType.ROOM);
                    if (roomTrait && !roomTrait.visited) {
                        roomTrait.visited = true;
                    }
                });
            }
        }
        // Move the actor
        world.moveEntity(actor, location);
        // Mark new location as visited if it's a room
        const newRoom = world.getEntity(location);
        if (newRoom && newRoom.has(world_model_1.TraitType.ROOM)) {
            // Mark new location as visited
            world.updateEntity(location, (room) => {
                const roomTrait = room.get(world_model_1.TraitType.ROOM);
                if (roomTrait && !roomTrait.visited) {
                    roomTrait.visited = true;
                }
            });
        }
    }
};
exports.handleActorMoved = handleActorMoved;
/**
 * Handle PUT_IN event - move item into container
 */
const handlePutIn = (event, world) => {
    const { target, location } = event.entities;
    if (target && location) {
        world.moveEntity(target, location);
    }
};
exports.handlePutIn = handlePutIn;
/**
 * Handle PUT_ON event - move item onto supporter
 */
const handlePutOn = (event, world) => {
    const { target, location } = event.entities;
    if (target && location) {
        world.moveEntity(target, location);
    }
};
exports.handlePutOn = handlePutOn;
/**
 * Handle REMOVED_FROM event - move item from container to actor
 */
const handleRemovedFrom = (event, world) => {
    const { actor, target } = event.entities;
    if (actor && target) {
        world.moveEntity(target, actor);
    }
};
exports.handleRemovedFrom = handleRemovedFrom;
/**
 * Handle ENTERED event - actor enters an object or vehicle
 */
const handleEntered = (event, world) => {
    const { actor, location } = event.entities;
    if (actor && location) {
        // Move the actor to the new location
        world.moveEntity(actor, location);
    }
};
exports.handleEntered = handleEntered;
/**
 * Handle EXITED event - actor exits an object or vehicle
 */
const handleExited = (event, world) => {
    const { actor, location } = event.entities;
    if (actor && location) {
        // Move the actor to the new location
        world.moveEntity(actor, location);
    }
};
exports.handleExited = handleExited;
/**
 * Handle CLIMBED event - record climbing action (movement handled separately)
 */
const handleClimbed = (event, world) => {
    // This event is mainly for tracking/logging purposes
    // The actual movement is handled by ENTERED or ACTOR_MOVED events
    // Could be used for achievements, scoring, or special effects
};
exports.handleClimbed = handleClimbed;
/**
 * Register all movement handlers
 */
function registerMovementHandlers(world) {
    world.registerEventHandler(if_domain_1.IFEvents.TAKEN, exports.handleTaken);
    world.registerEventHandler(if_domain_1.IFEvents.DROPPED, exports.handleDropped);
    world.registerEventHandler(if_domain_1.IFEvents.REMOVED, exports.handleRemoved);
    world.registerEventHandler(if_domain_1.IFEvents.ACTOR_MOVED, exports.handleActorMoved);
    world.registerEventHandler(if_domain_1.IFEvents.PUT_IN, exports.handlePutIn);
    world.registerEventHandler(if_domain_1.IFEvents.PUT_ON, exports.handlePutOn);
    world.registerEventHandler(if_domain_1.IFEvents.REMOVED_FROM, exports.handleRemovedFrom);
    world.registerEventHandler(if_domain_1.IFEvents.ENTERED, exports.handleEntered);
    world.registerEventHandler(if_domain_1.IFEvents.EXITED, exports.handleExited);
    world.registerEventHandler(if_domain_1.IFEvents.CLIMBED, exports.handleClimbed);
}
