// @sharpee/event-processor/handlers/state-change.js
"use strict";
/**
 * State change event handlers
 *
 * These handlers apply state-changing events to the world model
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.handleDrunk = exports.handleEaten = exports.handleWorn = exports.handleSwitchedOff = exports.handleSwitchedOn = exports.handleUnlocked = exports.handleLocked = exports.handleClosed = exports.handleOpened = void 0;
exports.registerStateChangeHandlers = registerStateChangeHandlers;
const world_model_1 = require("../../world-model/index.js");
const if_domain_1 = require("../../if-domain/index.js");
/**
 * Handle OPENED event - set openable trait to open
 */
const handleOpened = (event, world) => {
    const { target } = event.entities;
    if (target) {
        world.updateEntity(target, (entity) => {
            const openableTrait = entity.get(world_model_1.TraitType.OPENABLE);
            if (openableTrait) {
                openableTrait.isOpen = true;
            }
        });
        // TODO: Handle connected doors - DoorTrait doesn't have connectedDoor property
        // Need to determine the door connection logic
    }
};
exports.handleOpened = handleOpened;
/**
 * Handle CLOSED event - set openable trait to closed
 */
const handleClosed = (event, world) => {
    const { target } = event.entities;
    if (target) {
        world.updateEntity(target, (entity) => {
            const openableTrait = entity.get(world_model_1.TraitType.OPENABLE);
            if (openableTrait) {
                openableTrait.isOpen = false;
            }
        });
        // TODO: Handle connected doors
    }
};
exports.handleClosed = handleClosed;
/**
 * Handle LOCKED event - set lockable trait to locked
 */
const handleLocked = (event, world) => {
    const { target } = event.entities;
    if (target) {
        world.updateEntity(target, (entity) => {
            const lockableTrait = entity.get(world_model_1.TraitType.LOCKABLE);
            if (lockableTrait) {
                lockableTrait.isLocked = true;
            }
        });
        // TODO: Handle connected doors
    }
};
exports.handleLocked = handleLocked;
/**
 * Handle UNLOCKED event - set lockable trait to unlocked
 */
const handleUnlocked = (event, world) => {
    const { target } = event.entities;
    if (target) {
        world.updateEntity(target, (entity) => {
            const lockableTrait = entity.get(world_model_1.TraitType.LOCKABLE);
            if (lockableTrait) {
                lockableTrait.isLocked = false;
            }
        });
        // TODO: Handle connected doors
    }
};
exports.handleUnlocked = handleUnlocked;
/**
 * Handle SWITCHED_ON event - set switchable trait to on
 */
const handleSwitchedOn = (event, world) => {
    const { target } = event.entities;
    if (target) {
        world.updateEntity(target, (entity) => {
            const switchableTrait = entity.get(world_model_1.TraitType.SWITCHABLE);
            if (switchableTrait) {
                switchableTrait.isOn = true;
            }
            // If it's a light source, set it to lit
            const lightTrait = entity.get(world_model_1.TraitType.LIGHT_SOURCE);
            if (lightTrait) {
                lightTrait.isLit = true;
            }
        });
    }
};
exports.handleSwitchedOn = handleSwitchedOn;
/**
 * Handle SWITCHED_OFF event - set switchable trait to off
 */
const handleSwitchedOff = (event, world) => {
    const { target } = event.entities;
    if (target) {
        world.updateEntity(target, (entity) => {
            const switchableTrait = entity.get(world_model_1.TraitType.SWITCHABLE);
            if (switchableTrait) {
                switchableTrait.isOn = false;
            }
            // If it's a light source, set it to unlit
            const lightTrait = entity.get(world_model_1.TraitType.LIGHT_SOURCE);
            if (lightTrait) {
                lightTrait.isLit = false;
            }
        });
    }
};
exports.handleSwitchedOff = handleSwitchedOff;
/**
 * Handle WORN event - set wearable trait to worn
 */
const handleWorn = (event, world) => {
    const { target } = event.entities;
    if (target) {
        world.updateEntity(target, (entity) => {
            const wearableTrait = entity.get(world_model_1.TraitType.WEARABLE);
            if (wearableTrait) {
                wearableTrait.worn = true;
            }
        });
    }
};
exports.handleWorn = handleWorn;
/**
 * Handle EATEN event - remove the eaten item
 */
const handleEaten = (event, world) => {
    const { target } = event.entities;
    if (target) {
        world.removeEntity(target);
    }
};
exports.handleEaten = handleEaten;
/**
 * Handle DRUNK event - remove the drunk item
 */
const handleDrunk = (event, world) => {
    const { target } = event.entities;
    if (target) {
        world.removeEntity(target);
    }
};
exports.handleDrunk = handleDrunk;
/**
 * Register all state change handlers
 */
function registerStateChangeHandlers(world) {
    world.registerEventHandler(if_domain_1.IFEvents.OPENED, exports.handleOpened);
    world.registerEventHandler(if_domain_1.IFEvents.CLOSED, exports.handleClosed);
    world.registerEventHandler(if_domain_1.IFEvents.LOCKED, exports.handleLocked);
    world.registerEventHandler(if_domain_1.IFEvents.UNLOCKED, exports.handleUnlocked);
    world.registerEventHandler(if_domain_1.IFEvents.SWITCHED_ON, exports.handleSwitchedOn);
    world.registerEventHandler(if_domain_1.IFEvents.SWITCHED_OFF, exports.handleSwitchedOff);
    world.registerEventHandler(if_domain_1.IFEvents.WORN, exports.handleWorn);
    world.registerEventHandler(if_domain_1.IFEvents.EATEN, exports.handleEaten);
    world.registerEventHandler(if_domain_1.IFEvents.DRUNK, exports.handleDrunk);
}
