// @sharpee/plugin-state-machine/effect-executor.js
"use strict";
/**
 * Effect Executor - applies state machine effects to the world model.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.executeEffects = executeEffects;
const guard_evaluator_1 = require("./guard-evaluator");
let eventCounter = 0;
function executeEffects(effects, world, bindings, playerId, machineId) {
    const events = [];
    for (const effect of effects) {
        const result = executeEffect(effect, world, bindings, playerId, machineId);
        events.push(...result);
    }
    return events;
}
function executeEffect(effect, world, bindings, playerId, machineId) {
    switch (effect.type) {
        case 'move': {
            const entityId = (0, guard_evaluator_1.resolveRef)(effect.entityRef, bindings);
            const destId = (0, guard_evaluator_1.resolveRef)(effect.destinationRef, bindings);
            world.moveEntity(entityId, destId);
            return [createEvent(`sm.entity_moved`, machineId, {
                    entityId,
                    destination: destId,
                })];
        }
        case 'remove': {
            const entityId = (0, guard_evaluator_1.resolveRef)(effect.entityRef, bindings);
            world.removeEntity(entityId);
            return [createEvent(`sm.entity_removed`, machineId, { entityId })];
        }
        case 'set_trait': {
            const entityId = (0, guard_evaluator_1.resolveRef)(effect.entityRef, bindings);
            const entity = world.getEntity(entityId);
            if (entity) {
                const trait = entity.getTrait(effect.trait);
                if (trait) {
                    trait[effect.property] = effect.value;
                }
            }
            return [];
        }
        case 'set_state': {
            world.setStateValue(effect.key, effect.value);
            return [];
        }
        case 'message': {
            return [createEvent(effect.messageId, machineId, {
                    ...(effect.params || {}),
                    source: 'state-machine',
                })];
        }
        case 'emit_event': {
            const entities = {};
            if (effect.entities) {
                for (const [role, ref] of Object.entries(effect.entities)) {
                    if (ref)
                        entities[role] = (0, guard_evaluator_1.resolveRef)(ref, bindings);
                }
            }
            return [{
                    id: `sm-event-${++eventCounter}`,
                    type: effect.eventType,
                    timestamp: Date.now(),
                    entities,
                    data: effect.data,
                }];
        }
        case 'custom': {
            const result = effect.execute(world, bindings, playerId);
            const events = [];
            if (result.events) {
                for (const e of result.events) {
                    events.push({
                        id: `sm-custom-${++eventCounter}`,
                        type: e.type,
                        timestamp: Date.now(),
                        entities: (e.entities || {}),
                        data: e.data,
                    });
                }
            }
            if (result.messages) {
                for (const m of result.messages) {
                    events.push(createEvent(m.messageId, machineId, {
                        ...(m.params || {}),
                        source: 'state-machine',
                    }));
                }
            }
            return events;
        }
    }
}
function createEvent(type, machineId, data) {
    return {
        id: `sm-${++eventCounter}`,
        type,
        timestamp: Date.now(),
        entities: {},
        data,
        tags: ['state-machine', machineId],
    };
}
