// @sharpee/plugin-state-machine/state-machine-plugin.js
"use strict";
/**
 * StateMachinePlugin - Wraps state machine registry as a TurnPlugin (ADR-119, ADR-120)
 *
 * Priority 75: Runs after NPCs (100) but before scheduler (50).
 * Evaluates state machine transitions after each successful player action.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.StateMachinePlugin = void 0;
const state_machine_runtime_1 = require("./state-machine-runtime");
class StateMachinePlugin {
    id = 'sharpee.plugin.state-machine';
    priority = 75;
    registry;
    constructor() {
        this.registry = new state_machine_runtime_1.StateMachineRegistry();
    }
    onAfterAction(ctx) {
        const evalCtx = {
            world: ctx.world,
            playerId: ctx.playerId,
            playerLocation: ctx.playerLocation,
            turn: ctx.turn,
            actionId: ctx.actionResult?.actionId,
            actionTargetId: ctx.actionResult?.targetId,
            actionEvents: ctx.actionEvents?.map(e => ({
                type: e.type,
                data: e.data,
                entities: e.entities,
            })),
        };
        return this.registry.evaluate(evalCtx);
    }
    getState() {
        return this.registry.getState();
    }
    setState(state) {
        this.registry.setState(state);
    }
    getRegistry() {
        return this.registry;
    }
}
exports.StateMachinePlugin = StateMachinePlugin;
