// @sharpee/plugin-state-machine/state-machine-runtime.js
"use strict";
/**
 * State Machine Runtime (ADR-119)
 *
 * Manages registered state machines and evaluates transitions each turn.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.StateMachineRegistry = void 0;
const guard_evaluator_1 = require("./guard-evaluator");
const effect_executor_1 = require("./effect-executor");
class StateMachineRegistry {
    machines = new Map();
    register(definition, bindings = {}) {
        if (this.machines.has(definition.id)) {
            throw new Error(`State machine already registered: ${definition.id}`);
        }
        // Validate initial state exists
        if (!definition.states[definition.initialState]) {
            throw new Error(`Initial state '${definition.initialState}' not found in machine '${definition.id}'`);
        }
        this.machines.set(definition.id, {
            definition,
            bindings: { ...bindings },
            currentState: definition.initialState,
            history: [definition.initialState],
        });
    }
    unregister(id) {
        this.machines.delete(id);
    }
    getMachineState(id) {
        return this.machines.get(id)?.currentState;
    }
    getMachineHistory(id) {
        return this.machines.get(id)?.history;
    }
    evaluate(ctx) {
        const allEvents = [];
        for (const machine of this.machines.values()) {
            const stateDef = machine.definition.states[machine.currentState];
            if (!stateDef || stateDef.terminal || !stateDef.transitions?.length)
                continue;
            const transition = this.findMatchingTransition(stateDef.transitions, machine, ctx);
            if (transition) {
                const events = this.executeTransition(machine, transition, ctx);
                allEvents.push(...events);
            }
        }
        return allEvents;
    }
    findMatchingTransition(transitions, machine, ctx) {
        // Sort by priority (higher first), stable sort
        const sorted = [...transitions].sort((a, b) => (b.priority ?? 0) - (a.priority ?? 0));
        for (const transition of sorted) {
            if (!this.triggerMatches(transition.trigger, machine, ctx))
                continue;
            if (transition.guard && !(0, guard_evaluator_1.evaluateGuard)(transition.guard, ctx.world, machine.bindings, ctx.playerId))
                continue;
            return transition;
        }
        return undefined;
    }
    triggerMatches(trigger, machine, ctx) {
        switch (trigger.type) {
            case 'action': {
                if (ctx.actionId !== trigger.actionId)
                    return false;
                if (trigger.targetEntity) {
                    const expectedId = trigger.targetEntity.startsWith('$')
                        ? machine.bindings[trigger.targetEntity]
                        : trigger.targetEntity;
                    if (ctx.actionTargetId !== expectedId)
                        return false;
                }
                return true;
            }
            case 'event': {
                if (!ctx.actionEvents)
                    return false;
                return ctx.actionEvents.some(event => {
                    if (event.type !== trigger.eventId)
                        return false;
                    if (trigger.filter && event.data) {
                        const data = event.data;
                        for (const [key, value] of Object.entries(trigger.filter)) {
                            // Resolve binding references in filter values
                            const expected = typeof value === 'string' && value.startsWith('$')
                                ? machine.bindings[value]
                                : value;
                            if (data[key] !== expected)
                                return false;
                        }
                    }
                    return true;
                });
            }
            case 'condition': {
                return (0, guard_evaluator_1.evaluateGuard)(trigger.condition, ctx.world, machine.bindings, ctx.playerId);
            }
        }
    }
    executeTransition(machine, transition, ctx) {
        const events = [];
        const currentStateDef = machine.definition.states[machine.currentState];
        const targetStateDef = machine.definition.states[transition.target];
        if (!targetStateDef) {
            throw new Error(`Target state '${transition.target}' not found in machine '${machine.definition.id}'`);
        }
        // 1. Execute onExit effects of current state
        if (currentStateDef?.onExit) {
            events.push(...(0, effect_executor_1.executeEffects)(currentStateDef.onExit, ctx.world, machine.bindings, ctx.playerId, machine.definition.id));
        }
        // 2. Execute transition effects
        if (transition.effects) {
            events.push(...(0, effect_executor_1.executeEffects)(transition.effects, ctx.world, machine.bindings, ctx.playerId, machine.definition.id));
        }
        // 3. Move to target state
        machine.currentState = transition.target;
        machine.history.push(transition.target);
        // 4. Execute onEnter effects of target state
        if (targetStateDef.onEnter) {
            events.push(...(0, effect_executor_1.executeEffects)(targetStateDef.onEnter, ctx.world, machine.bindings, ctx.playerId, machine.definition.id));
        }
        // 5. Emit transition event
        events.push({
            id: `sm-transition-${Date.now()}`,
            type: 'sm.transition',
            timestamp: Date.now(),
            entities: {},
            data: {
                machineId: machine.definition.id,
                from: machine.history[machine.history.length - 2],
                to: transition.target,
            },
            tags: ['state-machine', machine.definition.id],
        });
        return events;
    }
    // ─── Serialization ──────────────────────────────────────────────────────
    getState() {
        const instances = [];
        for (const machine of this.machines.values()) {
            instances.push({
                id: machine.definition.id,
                currentState: machine.currentState,
                history: [...machine.history],
            });
        }
        return { instances };
    }
    setState(state) {
        for (const instanceState of state.instances) {
            const machine = this.machines.get(instanceState.id);
            if (machine) {
                machine.currentState = instanceState.currentState;
                machine.history = [...instanceState.history];
            }
        }
    }
}
exports.StateMachineRegistry = StateMachineRegistry;
