// @sharpee/stdlib/actions/standard/touching/touching.js
"use strict";
/**
 * Touching action - touch or feel objects
 *
 * This action allows players to touch objects to discover their
 * texture, temperature, or other tactile properties.
 *
 * Uses four-phase pattern:
 * 1. validate: Check target exists
 * 2. execute: Compute tactile properties (no world mutations)
 * 3. blocked: Generate events when validation fails
 * 4. report: Generate success events
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.touchingAction = void 0;
const world_model_1 = require("../../../../world-model/index.js");
const constants_1 = require("../../constants");
const types_1 = require("../../../scope/types");
function getTouchingSharedData(context) {
    return context.sharedData;
}
exports.touchingAction = {
    id: constants_1.IFActions.TOUCHING,
    // Default scope requirements for this action's slots
    defaultScope: {
        target: types_1.ScopeLevel.REACHABLE
    },
    requiredMessages: [
        'no_target',
        'not_visible',
        'not_reachable',
        'feels_normal',
        'feels_warm',
        'feels_hot',
        'feels_cold',
        'feels_soft',
        'feels_hard',
        'feels_smooth',
        'feels_rough',
        'feels_wet',
        'device_vibrating',
        'immovable_object',
        'liquid_container',
        'touched',
        'touched_gently',
        'poked',
        'prodded',
        'patted',
        'stroked'
    ],
    metadata: {
        requiresDirectObject: true,
        requiresIndirectObject: false,
        directObjectScope: types_1.ScopeLevel.REACHABLE
    },
    validate(context) {
        const target = context.command.directObject?.entity;
        // Must have a target to touch
        if (!target) {
            return {
                valid: false,
                error: 'no_target',
                params: {}
            };
        }
        // Check scope - must be able to reach the target
        const scopeCheck = context.requireScope(target, types_1.ScopeLevel.REACHABLE);
        if (!scopeCheck.ok) {
            return scopeCheck.error;
        }
        // Tactile property computation happens in execute phase
        return { valid: true };
    },
    execute(context) {
        // Touching has NO world mutations - it's a sensory action
        // Just compute tactile properties and store in sharedData for report phase
        const target = context.command.directObject?.entity;
        const sharedData = getTouchingSharedData(context);
        // Build event data with tactile properties
        const eventData = {
            target: target.id,
            targetName: target.name
        };
        let messageId = 'feels_normal';
        // Check various traits for tactile properties
        // Check if it's a light source (might be hot if lit)
        if (target.has(world_model_1.TraitType.LIGHT_SOURCE)) {
            const lightTrait = target.get(world_model_1.TraitType.LIGHT_SOURCE);
            if (lightTrait.isLit) {
                eventData.temperature = 'hot';
                eventData.isLit = true;
                messageId = 'feels_hot';
            }
        }
        // Check if it's a device (might be warm if on)
        if (target.has(world_model_1.TraitType.SWITCHABLE) && messageId === 'feels_normal') {
            const switchableTrait = target.get(world_model_1.TraitType.SWITCHABLE);
            if (switchableTrait.isOn) {
                eventData.temperature = 'warm';
                eventData.isActive = true;
                // Check if it's vibrating (special case for some devices)
                const identity = target.get(world_model_1.TraitType.IDENTITY);
                if (identity?.description?.toLowerCase().includes('vibrat')) {
                    messageId = 'device_vibrating';
                }
                else {
                    messageId = 'feels_warm';
                }
            }
        }
        // Determine texture based on traits
        if (target.has(world_model_1.TraitType.WEARABLE)) {
            eventData.texture = 'soft';
            if (messageId === 'feels_normal') {
                messageId = 'feels_soft';
            }
        }
        else if (target.has(world_model_1.TraitType.DOOR)) {
            eventData.texture = 'smooth';
            eventData.material = 'hard';
            if (messageId === 'feels_normal') {
                messageId = 'feels_smooth';
            }
        }
        else if (target.has(world_model_1.TraitType.CONTAINER) || target.has(world_model_1.TraitType.SUPPORTER)) {
            eventData.texture = 'solid';
            // Check for liquid contents
            if (target.has(world_model_1.TraitType.CONTAINER)) {
                const contents = context.world.getContents(target.id);
                const hasLiquid = contents.some(item => {
                    if (item.has(world_model_1.TraitType.EDIBLE)) {
                        const edibleTrait = item.get(world_model_1.TraitType.EDIBLE);
                        return edibleTrait.isDrink;
                    }
                    return false;
                });
                if (hasLiquid) {
                    messageId = 'liquid_container';
                }
            }
            if (messageId === 'feels_normal') {
                messageId = 'feels_hard';
            }
        }
        else if (target.has(world_model_1.TraitType.EDIBLE)) {
            const edibleTrait = target.get(world_model_1.TraitType.EDIBLE);
            if (edibleTrait.isDrink) {
                eventData.texture = 'liquid';
                if (messageId === 'feels_normal') {
                    messageId = 'feels_wet';
                }
            }
        }
        // Check if it's scenery (usually immovable)
        if (target.has(world_model_1.TraitType.SCENERY)) {
            eventData.immovable = true;
            if (messageId === 'feels_normal') {
                messageId = 'immovable_object';
            }
        }
        // Determine touch verb from command
        const verb = context.command.parsed.structure.verb?.text.toLowerCase() || 'touch';
        if (messageId === 'feels_normal') {
            // Use verb-specific messages for normal touches
            switch (verb) {
                case 'poke':
                    messageId = 'poked';
                    break;
                case 'prod':
                    messageId = 'prodded';
                    break;
                case 'pat':
                    messageId = 'patted';
                    break;
                case 'stroke':
                    messageId = 'stroked';
                    break;
                case 'feel':
                    messageId = 'touched_gently';
                    break;
                default:
                    messageId = 'touched';
            }
        }
        // Store computed data in sharedData for report phase
        sharedData.targetId = target.id;
        sharedData.targetName = target.name;
        sharedData.messageId = messageId;
        sharedData.eventData = eventData;
    },
    /**
     * Generate events when validation fails
     */
    blocked(context, result) {
        const target = context.command.directObject?.entity;
        return [context.event('if.event.touch_blocked', {
                blocked: true,
                messageId: `${context.action.id}.${result.error}`,
                params: {
                    ...result.params,
                    target: target?.name
                },
                reason: result.error,
                targetId: target?.id,
                targetName: target?.name
            })];
    },
    report(context) {
        const events = [];
        const sharedData = getTouchingSharedData(context);
        // Emit touched event with messageId for text rendering
        events.push(context.event('if.event.touched', {
            messageId: `${context.action.id}.${sharedData.messageId || 'touched'}`,
            params: { target: sharedData.targetName },
            ...sharedData.eventData
        }));
        return events;
    },
    group: "sensory"
};
