// @sharpee/stdlib/actions/base/migration-shim.js
"use strict";
/**
 * Migration shim for transitioning actions to the three-phase pattern
 *
 * This provides utilities to help migrate existing actions from the
 * two-phase pattern (validate + execute) to the three-phase pattern
 * (validate + execute + report).
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ThreePhaseAction = void 0;
exports.createThreePhaseAction = createThreePhaseAction;
exports.adaptOldAction = adaptOldAction;
exports.splitExecuteMethod = splitExecuteMethod;
exports.createMigrationLogger = createMigrationLogger;
/**
 * Base class for migrating actions to three-phase pattern
 *
 * Subclasses can override executeWithMutation() and generateReport()
 * to implement the new pattern while maintaining compatibility.
 */
class ThreePhaseAction {
    requiredMessages;
    descriptionMessageId;
    examplesMessageId;
    group;
    priority;
    /**
     * Execute phase - handles both old and new patterns
     *
     * This method checks if the subclass has implemented the new pattern.
     * If so, it calls executeWithMutation() and returns void.
     * Otherwise, it falls back to the old pattern.
     */
    execute(context) {
        // Check if subclass implements the new pattern
        if (this.executeWithMutation && this.generateReport) {
            // New pattern: perform mutations and return void
            this.executeWithMutation(context);
            return;
        }
        // Old pattern: must be overridden in subclass
        throw new Error(`Action ${this.id} must implement either executeWithMutation/generateReport ` +
            `or override execute() to return events`);
    }
    /**
     * Report phase - generate events after mutations
     *
     * This is automatically used when executeWithMutation is implemented.
     */
    report(context) {
        if (this.generateReport) {
            return this.generateReport(context);
        }
        // Should never reach here if execute() is properly implemented
        throw new Error(`Action ${this.id} must implement generateReport()`);
    }
}
exports.ThreePhaseAction = ThreePhaseAction;
/**
 * Helper to create a three-phase action from functions
 *
 * This is useful for quickly converting existing actions without
 * creating a full class.
 */
function createThreePhaseAction(config) {
    return {
        id: config.id,
        requiredMessages: config.requiredMessages,
        descriptionMessageId: config.descriptionMessageId,
        examplesMessageId: config.examplesMessageId,
        group: config.group,
        priority: config.priority,
        validate: config.validate,
        execute: (context) => {
            config.execute(context);
            return; // Return void for new pattern
        },
        report: config.report
    };
}
/**
 * Adapter to wrap old-style actions in the new interface
 *
 * This allows old actions to work with the new CommandExecutor
 * without modification.
 */
function adaptOldAction(oldAction) {
    return {
        ...oldAction,
        // Old pattern: execute returns events directly
        execute: (context) => oldAction.execute(context)
        // No report method needed for old pattern
    };
}
/**
 * Helper to split an old execute method into execute + report
 *
 * This helps when migrating actions that mix mutations and event generation.
 *
 * @param executeOld The old execute method that does both
 * @returns Object with separate execute and report methods
 */
function splitExecuteMethod(executeOld) {
    // Store events generated during execute
    let capturedEvents = null;
    return {
        execute: (context) => {
            // Run the old method but capture its events
            capturedEvents = executeOld(context);
            // Don't return anything (void)
        },
        report: (context) => {
            // Return the captured events
            if (!capturedEvents) {
                throw new Error('Report called before execute');
            }
            const events = capturedEvents;
            capturedEvents = null; // Reset for next execution
            return events;
        }
    };
}
/**
 * Create a migration wrapper that logs when old patterns are used
 *
 * This helps identify which actions still need migration.
 */
function createMigrationLogger(action, logger = console.warn) {
    const wrapped = {
        ...action,
        execute: (context) => {
            const result = action.execute(context);
            // Check if this is the old pattern
            if (result !== undefined && result !== null) {
                logger(`Action ${action.id} is using old execute pattern (returns events)`);
            }
            return result;
        }
    };
    // Add report if the original has it
    if (action.report) {
        wrapped.report = action.report;
    }
    return wrapped;
}
