// @sharpee/stdlib/actions/base/report-helpers.js
"use strict";
/**
 * Report phase helpers for reducing boilerplate across actions
 *
 * These helpers standardize error event generation in the report phase,
 * ensuring consistent error handling across all stdlib actions.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.handleValidationError = handleValidationError;
exports.handleExecutionError = handleExecutionError;
exports.handleReportErrors = handleReportErrors;
const snapshot_utils_1 = require("./snapshot-utils");
/**
 * Handle validation errors in report phase
 *
 * Creates standardized error events when validation fails.
 * Returns null if validation passed (valid === true).
 *
 * @param context The action context
 * @param validationResult The result from the validate phase
 * @param options Optional configuration for snapshot capture
 * @returns Array of error events if validation failed, null otherwise
 *
 * @example
 * report(context, validationResult, executionError) {
 *   const errorEvents = handleValidationError(context, validationResult);
 *   if (errorEvents) return errorEvents;
 *
 *   // ... success logic
 * }
 */
function handleValidationError(context, validationResult, options = {}) {
    if (!validationResult || validationResult.valid) {
        return null;
    }
    const errorParams = { ...(validationResult.params || {}) };
    // Optionally capture entity snapshots for richer error context
    if (options.includeTargetSnapshot !== false && context.command.directObject?.entity) {
        errorParams.targetSnapshot = (0, snapshot_utils_1.captureEntitySnapshot)(context.command.directObject.entity, context.world, false);
    }
    if (options.includeIndirectSnapshot !== false && context.command.indirectObject?.entity) {
        errorParams.indirectTargetSnapshot = (0, snapshot_utils_1.captureEntitySnapshot)(context.command.indirectObject.entity, context.world, false);
    }
    return [
        context.event('action.error', {
            actionId: context.action.id,
            error: validationResult.error || 'validation_failed',
            reason: validationResult.error || 'validation_failed',
            messageId: validationResult.messageId || validationResult.error || 'action_failed',
            params: errorParams
        })
    ];
}
/**
 * Handle execution errors in report phase
 *
 * Creates standardized error events when execution throws an error.
 * Returns null if no execution error occurred.
 *
 * @param context The action context
 * @param executionError Error thrown during execute phase (if any)
 * @returns Array of error events if execution failed, null otherwise
 *
 * @example
 * report(context, validationResult, executionError) {
 *   const errorEvents = handleExecutionError(context, executionError);
 *   if (errorEvents) return errorEvents;
 *
 *   // ... success logic
 * }
 */
function handleExecutionError(context, executionError) {
    if (!executionError) {
        return null;
    }
    return [
        context.event('action.error', {
            actionId: context.action.id,
            error: 'execution_failed',
            messageId: 'action_failed',
            params: { error: executionError.message }
        })
    ];
}
/**
 * Combined helper - handles both validation and execution errors
 *
 * This is the recommended helper for most actions. It checks validation
 * first, then execution errors, returning error events for whichever
 * failed first (or null if both succeeded).
 *
 * @param context The action context
 * @param validationResult The result from the validate phase
 * @param executionError Error thrown during execute phase (if any)
 * @param options Optional configuration for snapshot capture
 * @returns Array of error events if any error occurred, null otherwise
 *
 * @example
 * report(context, validationResult, executionError) {
 *   const errorEvents = handleReportErrors(context, validationResult, executionError);
 *   if (errorEvents) return errorEvents;
 *
 *   // All clear - generate success events
 *   const noun = context.command.directObject!.entity!;
 *   return [
 *     context.event('if.event.taken', { item: noun.name, messageId: 'taken' })
 *   ];
 * }
 */
function handleReportErrors(context, validationResult, executionError, options = {}) {
    // Check validation errors first
    const validationEvents = handleValidationError(context, validationResult, options);
    if (validationEvents)
        return validationEvents;
    // Then check execution errors
    const executionEvents = handleExecutionError(context, executionError);
    if (executionEvents)
        return executionEvents;
    // No errors
    return null;
}
