// @sharpee/world-model/traits/openable/openableTrait.js
"use strict";
// packages/world-model/src/traits/openable/openableTrait.ts
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpenableTrait = void 0;
const trait_types_1 = require("../trait-types");
/**
 * Openable trait for entities that can be opened and closed.
 * Used for doors, containers, books, etc.
 *
 * This trait contains only data - all logic for opening/closing
 * is in OpenableBehavior.
 */
class OpenableTrait {
    static type = trait_types_1.TraitType.OPENABLE;
    type = trait_types_1.TraitType.OPENABLE;
    // OpenableData properties
    isOpen;
    startsOpen;
    openMessage;
    closeMessage;
    alreadyOpenMessage;
    alreadyClosedMessage;
    revealsContents;
    canClose;
    openSound;
    closeSound;
    openDescription;
    closedDescription;
    constructor(data = {}) {
        // Set defaults and merge with provided data
        this.startsOpen = data.startsOpen ?? false;
        this.isOpen = data.isOpen ?? this.startsOpen;
        this.openMessage = data.openMessage;
        this.closeMessage = data.closeMessage;
        this.alreadyOpenMessage = data.alreadyOpenMessage;
        this.alreadyClosedMessage = data.alreadyClosedMessage;
        this.revealsContents = data.revealsContents ?? true;
        this.canClose = data.canClose ?? true;
        this.openSound = data.openSound;
        this.closeSound = data.closeSound;
        this.openDescription = data.openDescription;
        this.closedDescription = data.closedDescription;
    }
}
exports.OpenableTrait = OpenableTrait;
