// @sharpee/world-model/traits/openable/openableBehavior.js
"use strict";
// packages/world-model/src/traits/openable/openableBehavior.ts
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpenableBehavior = void 0;
const behavior_1 = require("../../behaviors/behavior");
const trait_types_1 = require("../trait-types");
/**
 * Behavior for openable entities.
 *
 * Handles the logic for opening and closing entities.
 */
class OpenableBehavior extends behavior_1.Behavior {
    static requiredTraits = [trait_types_1.TraitType.OPENABLE];
    /**
     * Check if an entity can be opened
     */
    static canOpen(entity) {
        const openable = OpenableBehavior.require(entity, trait_types_1.TraitType.OPENABLE);
        return !openable.isOpen;
    }
    /**
     * Check if an entity can be closed
     */
    static canClose(entity) {
        const openable = OpenableBehavior.require(entity, trait_types_1.TraitType.OPENABLE);
        return openable.isOpen && openable.canClose;
    }
    /**
     * Open the entity
     * @returns Result describing what happened
     */
    static open(entity) {
        const openable = OpenableBehavior.require(entity, trait_types_1.TraitType.OPENABLE);
        if (openable.isOpen) {
            return {
                success: false,
                alreadyOpen: true,
                stateChanged: false
            };
        }
        // Open it
        openable.isOpen = true;
        return {
            success: true,
            stateChanged: true,
            openMessage: openable.openMessage,
            openSound: openable.openSound,
            revealsContents: openable.revealsContents
        };
    }
    /**
     * Close the entity
     * @returns Result describing what happened
     */
    static close(entity) {
        const openable = OpenableBehavior.require(entity, trait_types_1.TraitType.OPENABLE);
        if (!openable.isOpen) {
            return {
                success: false,
                alreadyClosed: true,
                stateChanged: false
            };
        }
        if (!openable.canClose) {
            return {
                success: false,
                cantClose: true,
                stateChanged: false
            };
        }
        // Close it
        openable.isOpen = false;
        return {
            success: true,
            stateChanged: true,
            closeMessage: openable.closeMessage,
            closeSound: openable.closeSound
        };
    }
    /**
     * Toggle open/closed state
     * @returns Result from either open or close
     */
    static toggle(entity) {
        const openable = OpenableBehavior.require(entity, trait_types_1.TraitType.OPENABLE);
        if (openable.isOpen) {
            return OpenableBehavior.close(entity);
        }
        else {
            return OpenableBehavior.open(entity);
        }
    }
    /**
     * Check if entity is currently open
     */
    static isOpen(entity) {
        const openable = OpenableBehavior.require(entity, trait_types_1.TraitType.OPENABLE);
        return openable.isOpen;
    }
    /**
     * Check if opening reveals contents
     */
    static revealsContents(entity) {
        const openable = OpenableBehavior.require(entity, trait_types_1.TraitType.OPENABLE);
        return openable.revealsContents;
    }
}
exports.OpenableBehavior = OpenableBehavior;
