// @sharpee/world-model/traits/breakable/breakableBehavior.js
"use strict";
/**
 * Behavior for breakable entities
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.BreakableBehavior = void 0;
const behavior_1 = require("../../behaviors/behavior");
const trait_types_1 = require("../trait-types");
/**
 * Behavior for breakable entities.
 *
 * Handles breaking entities and creating debris.
 * This is a world-aware behavior because it needs to create/remove entities.
 */
class BreakableBehavior extends behavior_1.Behavior {
    static requiredTraits = [trait_types_1.TraitType.BREAKABLE];
    /**
     * Check if an entity can be broken
     */
    static canBreak(entity) {
        if (!entity.has(trait_types_1.TraitType.BREAKABLE)) {
            return false;
        }
        const breakable = entity.get(trait_types_1.TraitType.BREAKABLE);
        return !breakable.broken;
    }
    /**
     * Break the entity
     * @param entity The entity to break
     * @param world The world model (needed to create debris and remove entity)
     * @returns Result describing what happened
     */
    static break(entity, world) {
        if (!entity.has(trait_types_1.TraitType.BREAKABLE)) {
            return {
                success: false,
                message: 'Entity is not breakable'
            };
        }
        const breakable = entity.get(trait_types_1.TraitType.BREAKABLE);
        if (breakable.broken) {
            return {
                success: false,
                alreadyBroken: true
            };
        }
        // Mark as broken
        breakable.broken = true;
        const result = {
            success: true
        };
        // If broken item should be removed (common for fragile items)
        // The action can decide whether to remove based on game logic
        // For now we'll mark that removal is recommended
        result.itemRemoved = false; // Let the action decide
        return result;
    }
    /**
     * Check if entity is broken
     */
    static isBroken(entity) {
        if (!entity.has(trait_types_1.TraitType.BREAKABLE)) {
            return false;
        }
        const breakable = entity.get(trait_types_1.TraitType.BREAKABLE);
        return breakable.broken;
    }
}
exports.BreakableBehavior = BreakableBehavior;
