// @sharpee/world-model/traits/concealment/concealmentTrait.js
"use strict";
/**
 * Concealment trait for hiding spots (ADR-148)
 *
 * Entities with this trait can serve as hiding spots for actors.
 * The trait declares which positions are available (behind, under, on, inside)
 * and the quality of concealment the spot provides.
 *
 * Public interface: ConcealmentTrait, ConcealmentPosition, isConcealmentTrait.
 * Owner context: @sharpee/world-model / traits
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConcealmentTrait = void 0;
exports.isConcealmentTrait = isConcealmentTrait;
/**
 * Trait for entities that serve as hiding spots.
 *
 * @example
 * ```typescript
 * const curtain = world.createEntity('curtain', 'object');
 * curtain.add(new ConcealmentTrait({
 *   positions: ['behind'],
 *   quality: 'good',
 * }));
 * ```
 */
class ConcealmentTrait {
    static type = 'if.trait.concealment';
    type = 'if.trait.concealment';
    /** Which positions this entity supports for hiding */
    positions;
    /** How many actors can hide here simultaneously (default: 1) */
    capacity;
    /** Quality of concealment — used by story-level NPC detection logic */
    quality;
    constructor(options) {
        this.positions = options.positions;
        this.quality = options.quality;
        this.capacity = options.capacity ?? 1;
    }
    /**
     * Check if this entity supports a given hiding position.
     *
     * @param position - The position to check
     * @returns True if the position is available
     */
    supportsPosition(position) {
        return this.positions.includes(position);
    }
}
exports.ConcealmentTrait = ConcealmentTrait;
/**
 * Type guard for ConcealmentTrait
 */
function isConcealmentTrait(trait) {
    return trait.type === ConcealmentTrait.type;
}
