// @sharpee/world-model/traits/concealment/concealedStateTrait.js
"use strict";
/**
 * Concealed state trait — dynamic trait applied to actors when hiding (ADR-148)
 *
 * Added by the hiding action, removed by the revealing action or implicit reveal.
 * Presence of this trait IS the concealed state — no boolean flag needed.
 *
 * Registers the if.scope.visible capability to block VisibilityBehavior.canSee().
 * When an NPC calls canSee() on a concealed actor, the capability behavior
 * returns { valid: false }, making the actor invisible.
 *
 * Public interface: ConcealedStateTrait, IConcealedStateTrait, isConcealed, getConcealmentState.
 * Owner context: @sharpee/world-model / traits
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConcealedStateTrait = void 0;
exports.isConcealed = isConcealed;
exports.getConcealmentState = getConcealmentState;
const VisibilityBehavior_1 = require("../../world/VisibilityBehavior");
/**
 * Dynamic trait applied to an actor to mark them as concealed.
 *
 * The trait registers the `if.scope.visible` capability so that
 * VisibilityBehavior.canSee() automatically returns false for
 * concealed actors.
 */
class ConcealedStateTrait {
    static type = 'if.trait.concealed_state';
    type = 'if.trait.concealed_state';
    /** Registers the visibility capability — concealed actors block canSee() */
    static capabilities = [VisibilityBehavior_1.VISIBILITY_CAPABILITY];
    targetId;
    position;
    quality;
    constructor(data) {
        this.targetId = data.targetId;
        this.position = data.position;
        this.quality = data.quality;
    }
}
exports.ConcealedStateTrait = ConcealedStateTrait;
/**
 * Check if an entity is concealed.
 *
 * @param entity - The entity to check
 * @returns True if the entity has ConcealedStateTrait
 */
function isConcealed(entity) {
    return entity.has(ConcealedStateTrait.type);
}
/**
 * Get the concealment details for an entity, or undefined if not concealed.
 *
 * @param entity - The entity to check
 * @returns The concealment state, or undefined
 */
function getConcealmentState(entity) {
    const trait = entity.get(ConcealedStateTrait.type);
    if (!trait)
        return undefined;
    return trait;
}
