// @sharpee/stdlib/actions/standard/showing/showing.js
"use strict";
/**
 * Showing action - show objects to NPCs or other actors
 *
 * This action makes NPCs aware of objects without transferring ownership.
 * Useful for puzzles where NPCs react to seeing specific items.
 *
 * Uses four-phase pattern:
 * 1. validate: Check item and viewer exist and are valid
 * 2. execute: Analyze show reaction (no world mutations)
 * 3. blocked: Generate events when validation fails
 * 4. report: Generate success events
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.showingAction = void 0;
const types_1 = require("../../../scope/types");
const world_model_1 = require("../../../../world-model/index.js");
const constants_1 = require("../../constants");
function getShowingSharedData(context) {
    return context.sharedData;
}
function analyzeShowAction(context) {
    const item = context.command.directObject?.entity;
    const viewer = context.command.indirectObject?.entity;
    if (!item || !viewer) {
        return null;
    }
    // Check if wearing the item
    const wearableTrait = item.has(world_model_1.TraitType.WEARABLE) ?
        item.get(world_model_1.TraitType.WEARABLE) : null;
    const isWearing = wearableTrait?.worn === true;
    // Build event data
    const eventData = {
        item: item.id,
        itemName: item.name,
        viewer: viewer.id,
        viewerName: viewer.name,
        isWorn: isWearing
    };
    const params = {
        item: item.name,
        viewer: viewer.name
    };
    // Determine viewer reaction
    let messageId = 'shown';
    // Add item properties that might trigger reactions
    if (item.has(world_model_1.TraitType.IDENTITY)) {
        const identity = item.get(world_model_1.TraitType.IDENTITY);
        if (identity.properName && identity.name) {
            eventData.itemProperName = identity.name;
        }
    }
    // Check if viewer has any special reactions to items
    const viewerActor = viewer.getTrait(world_model_1.ActorTrait);
    const reactions = viewerActor?.customProperties?.['reactions'];
    if (viewerActor && reactions) {
        const itemName = item.name.toLowerCase();
        // Check for specific item reactions
        if (reactions.recognizes && reactions.recognizes.some((r) => itemName.includes(r))) {
            messageId = 'viewer_recognizes';
            eventData.recognized = true;
        }
        else if (reactions.impressed && reactions.impressed.some((i) => itemName.includes(i))) {
            messageId = 'viewer_impressed';
            eventData.impressed = true;
        }
        else if (reactions.unimpressed && reactions.unimpressed.some((u) => itemName.includes(u))) {
            messageId = 'viewer_unimpressed';
        }
        else if (reactions.examines && reactions.examines.some((e) => itemName.includes(e))) {
            messageId = 'viewer_examines';
        }
        else {
            messageId = 'viewer_nods';
        }
    }
    // Special message if showing worn item
    if (isWearing && messageId === 'shown') {
        messageId = 'wearing_shown';
    }
    return {
        item,
        viewer,
        isWearing,
        messageId,
        eventData,
        params
    };
}
exports.showingAction = {
    id: constants_1.IFActions.SHOWING,
    // Default scope requirements for this action's slots
    defaultScope: {
        item: types_1.ScopeLevel.REACHABLE, // REACHABLE allows implicit take
        viewer: types_1.ScopeLevel.VISIBLE
    },
    requiredMessages: [
        'no_item',
        'no_viewer',
        'not_carrying',
        'viewer_not_visible',
        'viewer_too_far',
        'not_actor',
        'self',
        'shown',
        'viewer_examines',
        'viewer_nods',
        'viewer_impressed',
        'viewer_unimpressed',
        'viewer_recognizes',
        'wearing_shown'
    ],
    metadata: {
        requiresDirectObject: true,
        requiresIndirectObject: true,
        directObjectScope: types_1.ScopeLevel.REACHABLE, // REACHABLE allows implicit take
        indirectObjectScope: types_1.ScopeLevel.VISIBLE
    },
    validate(context) {
        const actor = context.player;
        const item = context.command.directObject?.entity;
        const viewer = context.command.indirectObject?.entity;
        // Validate we have both item and viewer
        if (!item) {
            return {
                valid: false,
                error: 'no_item',
                params: {}
            };
        }
        if (!viewer) {
            return {
                valid: false,
                error: 'no_viewer',
                params: {}
            };
        }
        // Item must be carried (or implicitly takeable)
        // This enables "show apple to bob" when apple is on the ground
        const carryCheck = context.requireCarriedOrImplicitTake(item);
        if (!carryCheck.ok) {
            return carryCheck.error;
        }
        // Check if viewer is close enough to see
        const viewerLocation = context.world.getLocation?.(viewer.id);
        const actorLocation = context.world.getLocation?.(actor.id);
        if (viewerLocation !== actorLocation) {
            return {
                valid: false,
                error: 'viewer_too_far',
                params: { viewer: viewer.name }
            };
        }
        // Check if viewer is an actor (has eyes to see)
        if (!viewer.has(world_model_1.TraitType.ACTOR)) {
            return {
                valid: false,
                error: 'not_actor'
            };
        }
        // Prevent showing to self
        if (viewer.id === actor.id) {
            return {
                valid: false,
                error: 'self',
                params: { item: item.name }
            };
        }
        return { valid: true };
    },
    execute(context) {
        // Showing has NO world mutations - it's a social interaction
        // Analyze show reaction and store in sharedData for report phase
        const analysis = analyzeShowAction(context);
        const sharedData = getShowingSharedData(context);
        if (analysis) {
            sharedData.messageId = analysis.messageId;
            sharedData.eventData = analysis.eventData;
            sharedData.params = analysis.params;
        }
    },
    blocked(context, result) {
        const item = context.command.directObject?.entity;
        const viewer = context.command.indirectObject?.entity;
        return [context.event('if.event.show_blocked', {
                blocked: true,
                messageId: `${context.action.id}.${result.error}`,
                params: {
                    ...result.params,
                    item: item?.name,
                    viewer: viewer?.name
                },
                reason: result.error,
                itemId: item?.id,
                itemName: item?.name,
                viewerId: viewer?.id,
                viewerName: viewer?.name
            })];
    },
    report(context) {
        const events = [];
        const sharedData = getShowingSharedData(context);
        // Prepend any implicit take events (from requireCarriedOrImplicitTake)
        if (context.sharedData.implicitTakeEvents) {
            events.push(...context.sharedData.implicitTakeEvents);
        }
        // Emit shown event with messageId for text rendering
        events.push(context.event('if.event.shown', {
            messageId: `${context.action.id}.${sharedData.messageId || 'shown'}`,
            params: sharedData.params,
            ...sharedData.eventData
        }));
        return events;
    },
    group: "social"
};
