// @sharpee/stdlib/services/PerceptionService.js
"use strict";
/**
 * PerceptionService - Filters events based on what the player can perceive
 *
 * This service sits between action execution and the text service, transforming
 * events that describe things the player cannot perceive (due to darkness,
 * blindness, etc.) into appropriate alternative events.
 *
 * @see ADR-069 Perception-Based Event Filtering
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PerceptionService = void 0;
const world_model_1 = require("../../world-model/index.js");
/**
 * Default implementation of IPerceptionService
 *
 * Filters events based on environmental and actor state:
 * - Darkness (room is dark, no light source)
 * - Blindness (future: actor has blind trait)
 * - Blindfold (future: actor wearing something over eyes)
 */
class PerceptionService {
    /**
     * Filter events based on what the actor can perceive.
     *
     * Visual events (room descriptions, contents lists) are transformed
     * into perception-blocked events when the player cannot see.
     */
    filterEvents(events, actor, world) {
        const location = this.getActorRoom(actor, world);
        if (!location) {
            return events; // Can't determine location, pass through
        }
        const canSee = this.canPerceive(actor, location, world, 'sight');
        return events.map((event) => {
            // Only filter visual events
            if (!this.isVisualEvent(event)) {
                return event;
            }
            // If player can see, pass through unchanged
            if (canSee) {
                return event;
            }
            // Transform to perception-blocked event
            return this.createPerceptionBlockedEvent(event, actor, location, world);
        });
    }
    /**
     * Check if an actor can perceive using a specific sense.
     */
    canPerceive(actor, location, world, sense) {
        switch (sense) {
            case 'sight':
                return this.canSeeVisually(actor, location, world);
            case 'hearing':
                return this.canHear(actor, location, world);
            case 'smell':
                return this.canSmell(actor, location, world);
            case 'touch':
                return this.canTouch(actor, location, world);
            default:
                return true;
        }
    }
    /**
     * Check if actor can see in the given location.
     *
     * Checks (in order):
     * 1. Actor blindness trait
     * 2. Actor wearing blindfold
     * 3. Location darkness (via VisibilityBehavior)
     */
    canSeeVisually(actor, location, world) {
        // Check actor state first
        if (this.isBlind(actor)) {
            return false;
        }
        if (this.isWearingBlindfold(actor, world)) {
            return false;
        }
        // Check environmental darkness
        // Note: VisibilityBehavior.isDark expects a concrete WorldModel
        // We cast here since IWorldModel is compatible
        if (world_model_1.VisibilityBehavior.isDark(location, world)) {
            return false;
        }
        return true;
    }
    /**
     * Check if actor can hear. Currently always true (future extension point).
     */
    canHear(_actor, _location, _world) {
        // Future: check deafness trait, earplugs, etc.
        return true;
    }
    /**
     * Check if actor can smell. Currently always true (future extension point).
     */
    canSmell(_actor, _location, _world) {
        return true;
    }
    /**
     * Check if actor can touch. Currently always true (future extension point).
     */
    canTouch(_actor, _location, _world) {
        return true;
    }
    /**
     * Check if an actor has a blindness trait.
     */
    isBlind(_actor) {
        // Future: check for BLIND trait
        // For now, no actors are blind
        return false;
    }
    /**
     * Check if an actor is wearing something that blocks vision.
     */
    isWearingBlindfold(_actor, _world) {
        // Future: check worn items for blindfold trait
        // For now, no items are blindfolds
        return false;
    }
    /**
     * Get the room an actor is currently in.
     */
    getActorRoom(actor, world) {
        return world.getContainingRoom(actor.id) || null;
    }
    /**
     * Check if an event requires visual perception.
     */
    isVisualEvent(event) {
        // Room descriptions always require sight
        if (event.type === 'if.event.room.description') {
            return true;
        }
        // Contents lists require sight
        if (event.type === 'if.event.contents.listed') {
            return true;
        }
        // action.success with contents_list messageId requires sight
        if (event.type === 'action.success') {
            const data = event.data;
            if (data?.messageId === 'contents_list') {
                return true;
            }
        }
        return false;
    }
    /**
     * Create a perception-blocked event to replace a filtered event.
     */
    createPerceptionBlockedEvent(originalEvent, actor, location, world) {
        const reason = this.getBlockReason(actor, location, world);
        const blockedData = {
            originalType: originalEvent.type,
            reason,
            sense: 'sight',
            originalData: originalEvent.data,
        };
        return {
            ...originalEvent,
            type: 'if.event.perception.blocked',
            data: blockedData,
        };
    }
    /**
     * Determine why perception is blocked.
     */
    getBlockReason(actor, location, world) {
        if (this.isBlind(actor)) {
            return 'blindness';
        }
        if (this.isWearingBlindfold(actor, world)) {
            return 'blindfolded';
        }
        if (world_model_1.VisibilityBehavior.isDark(location, world)) {
            return 'darkness';
        }
        return 'unknown';
    }
}
exports.PerceptionService = PerceptionService;
