// @sharpee/plugins/plugin-registry.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PluginRegistry = void 0;
class PluginRegistry {
    plugins = new Map();
    register(plugin) {
        if (this.plugins.has(plugin.id)) {
            throw new Error(`Plugin already registered: ${plugin.id}`);
        }
        this.plugins.set(plugin.id, plugin);
    }
    unregister(id) {
        this.plugins.delete(id);
    }
    clear() {
        this.plugins.clear();
    }
    getAll() {
        return Array.from(this.plugins.values()).sort((a, b) => b.priority - a.priority);
    }
    getById(id) {
        return this.plugins.get(id);
    }
    getStates() {
        const states = {};
        for (const plugin of this.plugins.values()) {
            if (plugin.getState) {
                states[plugin.id] = plugin.getState();
            }
        }
        return states;
    }
    setStates(states) {
        for (const [id, state] of Object.entries(states)) {
            const plugin = this.plugins.get(id);
            if (plugin?.setState) {
                plugin.setState(state);
            }
        }
    }
}
exports.PluginRegistry = PluginRegistry;
