// @sharpee/stdlib/parser/index.js
"use strict";
/**
 * Parser module exports
 *
 * Parser contracts have been moved to @sharpee/if-domain.
 * This module now only re-exports from if-domain for backward compatibility.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.adaptSpecialVocabulary = exports.adaptDirectionVocabulary = exports.adaptVerbVocabulary = exports.GrammarPatterns = exports.vocabularyRegistry = exports.VocabularyRegistry = exports.PartOfSpeech = exports.ParseErrorType = exports.ParserFactory = void 0;
// Re-export parser and vocabulary contracts from if-domain
var if_domain_1 = require("../../if-domain/index.js");
Object.defineProperty(exports, "ParserFactory", { enumerable: true, get: function () { return if_domain_1.ParserFactory; } });
Object.defineProperty(exports, "ParseErrorType", { enumerable: true, get: function () { return if_domain_1.ParseErrorType; } });
Object.defineProperty(exports, "PartOfSpeech", { enumerable: true, get: function () { return if_domain_1.PartOfSpeech; } });
Object.defineProperty(exports, "VocabularyRegistry", { enumerable: true, get: function () { return if_domain_1.VocabularyRegistry; } });
Object.defineProperty(exports, "vocabularyRegistry", { enumerable: true, get: function () { return if_domain_1.vocabularyRegistry; } });
Object.defineProperty(exports, "GrammarPatterns", { enumerable: true, get: function () { return if_domain_1.GrammarPatterns; } });
// Vocabulary adapters
Object.defineProperty(exports, "adaptVerbVocabulary", { enumerable: true, get: function () { return if_domain_1.adaptVerbVocabulary; } });
Object.defineProperty(exports, "adaptDirectionVocabulary", { enumerable: true, get: function () { return if_domain_1.adaptDirectionVocabulary; } });
Object.defineProperty(exports, "adaptSpecialVocabulary", { enumerable: true, get: function () { return if_domain_1.adaptSpecialVocabulary; } });
