// @sharpee/stdlib/actions/standard/exiting/exiting.js
"use strict";
/**
 * Exiting action - exit from containers, supporters, or other enterable objects
 *
 * This action handles exiting objects that the actor is currently inside/on.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.exitingAction = void 0;
const world_model_1 = require("../../../../world-model/index.js");
const constants_1 = require("../../constants");
const exiting_messages_1 = require("./exiting-messages");
function getExitingSharedData(context) {
    return context.sharedData;
}
const types_1 = require("../../../scope/types");
exports.exitingAction = {
    id: constants_1.IFActions.EXITING,
    requiredMessages: [
        'already_outside',
        'container_closed',
        'cant_exit',
        'exited',
        'exited_from',
        'nowhere_to_go'
    ],
    group: 'movement',
    validate(context) {
        const actor = context.player;
        const currentLocation = context.world.getLocation(actor.id);
        if (!currentLocation) {
            return {
                valid: false,
                error: exiting_messages_1.ExitingMessages.NOWHERE_TO_GO
            };
        }
        const currentContainer = context.world.getEntity(currentLocation);
        if (!currentContainer) {
            return {
                valid: false,
                error: exiting_messages_1.ExitingMessages.NOWHERE_TO_GO
            };
        }
        // Check if we're in something we can exit from
        // Rooms cannot be exited (use GO to move between rooms)
        const isRoom = currentContainer.has(world_model_1.TraitType.ROOM);
        if (isRoom) {
            return {
                valid: false,
                error: exiting_messages_1.ExitingMessages.ALREADY_OUTSIDE
            };
        }
        // Find the parent location (where we'll exit to)
        const parentLocation = context.world.getLocation(currentLocation);
        if (!parentLocation) {
            return {
                valid: false,
                error: exiting_messages_1.ExitingMessages.NOWHERE_TO_GO
            };
        }
        // Check if container needs to be open to exit using behavior
        if (currentContainer.has(world_model_1.TraitType.CONTAINER) && currentContainer.has(world_model_1.TraitType.OPENABLE)) {
            if (!world_model_1.OpenableBehavior.isOpen(currentContainer)) {
                return {
                    valid: false,
                    error: exiting_messages_1.ExitingMessages.CONTAINER_CLOSED,
                    params: { container: currentContainer.name }
                };
            }
        }
        return { valid: true };
    },
    /**
     * Execute the exit action - performs mutations only
     * Assumes validation has already passed
     */
    execute(context) {
        const actor = context.player;
        const currentLocation = context.world.getLocation(actor.id); // Safe because validate ensures it exists
        const currentContainer = context.world.getEntity(currentLocation); // Safe because validate ensures it exists
        const parentLocation = context.world.getLocation(currentLocation); // Safe because validate ensures it exists
        // Determine preposition based on what we're exiting from
        let preposition = 'from';
        if (currentContainer.has(world_model_1.TraitType.CONTAINER)) {
            preposition = 'out of';
        }
        else if (currentContainer.has(world_model_1.TraitType.SUPPORTER)) {
            preposition = 'off';
        }
        // Simply move the actor to the parent location - that's all!
        context.world.moveEntity(actor.id, parentLocation);
        // Store state for report phase using sharedData
        const state = {
            fromLocation: currentLocation,
            fromLocationName: currentContainer.name,
            toLocation: parentLocation,
            preposition
        };
        const sharedData = getExitingSharedData(context);
        sharedData.exitingState = state;
    },
    /**
     * Report phase - generates events after successful execution
     * Only called on success path - validation has already passed
     */
    report(context) {
        // Get stored state from execute phase
        const sharedData = getExitingSharedData(context);
        const state = sharedData.exitingState;
        if (!state) {
            // This shouldn't happen, but handle gracefully
            return [
                context.event('if.event.exited', {
                    messageId: `${context.action.id}.action_failed`,
                    error: 'Missing state from execute phase'
                })
            ];
        }
        // Create the EXITED event with messageId for text rendering
        return [context.event('if.event.exited', {
                messageId: `${context.action.id}.exited`,
                params: { place: state.fromLocationName },
                fromLocation: state.fromLocation,
                fromLocationName: state.fromLocationName,
                toLocation: state.toLocation,
                preposition: state.preposition
            })];
    },
    /**
     * Generate events when validation fails
     * Called instead of execute/report when validate returns invalid
     */
    blocked(context, result) {
        return [context.event('if.event.exited', {
                blocked: true,
                messageId: `${context.action.id}.${result.error}`,
                params: result.params || {},
                reason: result.error
            })];
    },
    metadata: {
        requiresDirectObject: false,
        requiresIndirectObject: false,
        directObjectScope: types_1.ScopeLevel.REACHABLE
    }
};
