// @sharpee/lang-en-us/formatters/list.js
"use strict";
/**
 * List Formatters
 *
 * Formatters for joining arrays of items into prose.
 *
 * @see ADR-095 Message Templates with Formatters
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.countFormatter = exports.commaListFormatter = exports.orListFormatter = exports.listFormatter = void 0;
/**
 * Extract name from value (string or EntityInfo)
 */
function getName(value) {
    return typeof value === 'string' ? value : value.name;
}
/**
 * Ensure value is an array
 */
function ensureArray(value) {
    if (Array.isArray(value)) {
        return value.map((v) => getName(v));
    }
    return [getName(value)];
}
/**
 * "list" formatter - join with commas and "and"
 *
 * Single: "a sword"
 * Two: "a sword and a key"
 * Three+: "a sword, a key, and a coin"
 */
const listFormatter = (value, _context) => {
    const items = ensureArray(value);
    if (items.length === 0)
        return '';
    if (items.length === 1)
        return items[0];
    if (items.length === 2)
        return `${items[0]} and ${items[1]}`;
    const allButLast = items.slice(0, -1);
    const last = items[items.length - 1];
    return `${allButLast.join(', ')}, and ${last}`;
};
exports.listFormatter = listFormatter;
/**
 * "or-list" formatter - join with commas and "or"
 *
 * Single: "north"
 * Two: "north or south"
 * Three+: "north, south, or east"
 */
const orListFormatter = (value, _context) => {
    const items = ensureArray(value);
    if (items.length === 0)
        return '';
    if (items.length === 1)
        return items[0];
    if (items.length === 2)
        return `${items[0]} or ${items[1]}`;
    const allButLast = items.slice(0, -1);
    const last = items[items.length - 1];
    return `${allButLast.join(', ')}, or ${last}`;
};
exports.orListFormatter = orListFormatter;
/**
 * "comma-list" formatter - join with commas only (no conjunction)
 *
 * "a sword, a key, a coin"
 */
const commaListFormatter = (value, _context) => {
    const items = ensureArray(value);
    return items.join(', ');
};
exports.commaListFormatter = commaListFormatter;
/**
 * "count" formatter - number + noun (with pluralization)
 *
 * Example: {items:count} with 3 swords → "3 swords"
 * Example: {items:count} with 1 sword → "1 sword"
 */
const countFormatter = (value, _context) => {
    const items = ensureArray(value);
    const count = items.length;
    if (count === 0)
        return 'nothing';
    if (count === 1)
        return `1 ${items[0]}`;
    // For count > 1, assume the items are already formatted
    // The item name should be pluralized by caller or be inherently plural
    return `${count} items`;
};
exports.countFormatter = countFormatter;
