// @sharpee/core/random/seeded-random.js
"use strict";
/**
 * Seeded Random Number Generator Interface
 *
 * Provides deterministic randomness for reproducible game behavior.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.createSeededRandom = createSeededRandom;
/**
 * Create a seeded random number generator
 *
 * Uses a Linear Congruential Generator (LCG) for deterministic randomness.
 * This ensures reproducible behavior for testing and save/load.
 */
function createSeededRandom(seed) {
    // Use current time as default seed
    let currentSeed = seed ?? Date.now();
    // LCG constants (same as glibc)
    const a = 1103515245;
    const c = 12345;
    const m = 2 ** 31;
    function next() {
        currentSeed = (a * currentSeed + c) % m;
        return currentSeed / m;
    }
    function int(min, max) {
        return Math.floor(next() * (max - min + 1)) + min;
    }
    function chance(probability) {
        return next() < probability;
    }
    function pick(array) {
        if (array.length === 0) {
            throw new Error('Cannot pick from empty array');
        }
        return array[int(0, array.length - 1)];
    }
    function shuffle(array) {
        const result = [...array];
        for (let i = result.length - 1; i > 0; i--) {
            const j = int(0, i);
            [result[i], result[j]] = [result[j], result[i]];
        }
        return result;
    }
    function getSeed() {
        return currentSeed;
    }
    function setSeed(newSeed) {
        currentSeed = newSeed;
    }
    return {
        next,
        int,
        chance,
        pick,
        shuffle,
        getSeed,
        setSeed,
    };
}
