// @sharpee/world-model/world/AuthorModel.js
"use strict";
/**
 * AuthorModel — unrestricted world model access for authoring and setup.
 *
 * Public interface: Implements IWorldModel. Entity creation and movement
 * bypass validation. All other methods delegate to the backing WorldModel.
 *
 * Owner context: packages/world-model. Used during initializeWorld() for
 * setup that requires bypassing game rules (placing items in closed
 * containers, etc.).
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.AuthorModel = void 0;
const if_entity_1 = require("../entities/if-entity");
/**
 * AuthorModel provides unrestricted access to the world state for authoring,
 * testing, and world setup. It bypasses validation rules for entity creation
 * and movement. All other IWorldModel methods delegate to the backing WorldModel.
 *
 * @example
 * ```typescript
 * const author = new AuthorModel(world.getDataStore(), world);
 * const medicine = author.createEntity('Aspirin', 'item');
 * author.moveEntity(medicine.id, closedCabinet.id); // Works even though closed
 * ```
 */
class AuthorModel {
    dataStore;
    worldModel;
    constructor(dataStore, worldModel) {
        this.dataStore = dataStore;
        this.worldModel = worldModel;
    }
    /**
     * Get the shared data store.
     */
    getDataStore() {
        return this.dataStore;
    }
    // ========== Overridden Methods (bypass validation) ==========
    /**
     * Create a new entity without validation.
     *
     * @param name - Display name for the entity
     * @param type - Entity type (room, item, actor, etc.)
     * @returns The created entity
     */
    createEntity(name, type = 'object') {
        const id = this.generateId(type);
        const entity = new if_entity_1.IFEntity(id, type, {
            attributes: {
                displayName: name,
                name: name,
                entityType: type
            }
        });
        this.dataStore.entities.set(id, entity);
        return entity;
    }
    /**
     * Move an entity without validation. Can move into closed/locked containers.
     *
     * @param entityId - ID of entity to move
     * @param targetId - ID of target location (null to remove from world)
     * @returns Always true (no validation to fail)
     */
    moveEntity(entityId, targetId) {
        const currentLocation = this.dataStore.spatialIndex.getParent(entityId);
        if (currentLocation) {
            this.dataStore.spatialIndex.removeChild(currentLocation, entityId);
        }
        if (targetId !== null) {
            this.dataStore.spatialIndex.addChild(targetId, entityId);
        }
        return true;
    }
    // ========== Delegated Methods ==========
    // Entity Management
    getEntity(id) {
        return this.dataStore.entities.get(id);
    }
    hasEntity(id) {
        return this.dataStore.entities.has(id);
    }
    removeEntity(id) {
        return this.worldModel.removeEntity(id);
    }
    getAllEntities() {
        return Array.from(this.dataStore.entities.values());
    }
    updateEntity(entityId, updater) {
        this.worldModel.updateEntity(entityId, updater);
    }
    // Spatial Management
    getLocation(entityId) {
        return this.dataStore.spatialIndex.getParent(entityId);
    }
    getContents(containerId, options) {
        return this.worldModel.getContents(containerId, options);
    }
    canMoveEntity(entityId, targetId) {
        return true; // AuthorModel always allows moves
    }
    getContainingRoom(entityId) {
        return this.worldModel.getContainingRoom(entityId);
    }
    getAllContents(entityId, options) {
        return this.worldModel.getAllContents(entityId, options);
    }
    // World State Management
    getState() {
        return { ...this.dataStore.state };
    }
    setState(state) {
        Object.assign(this.dataStore.state, state);
    }
    getStateValue(key) {
        return this.dataStore.state[key];
    }
    setStateValue(key, value) {
        this.dataStore.state[key] = value;
    }
    // Prompt
    getPrompt() {
        return this.worldModel.getPrompt();
    }
    setPrompt(prompt) {
        this.worldModel.setPrompt(prompt);
    }
    // Query Operations
    findByTrait(traitType) {
        return this.worldModel.findByTrait(traitType);
    }
    findByType(entityType) {
        return this.worldModel.findByType(entityType);
    }
    findWhere(predicate) {
        return this.worldModel.findWhere(predicate);
    }
    getVisible(observerId) {
        return this.worldModel.getVisible(observerId);
    }
    getInScope(observerId) {
        return this.worldModel.getInScope(observerId);
    }
    canSee(observerId, targetId) {
        return this.worldModel.canSee(observerId, targetId);
    }
    // Relationship Queries
    getRelated(entityId, relationshipType) {
        return this.worldModel.getRelated(entityId, relationshipType);
    }
    areRelated(entity1Id, entity2Id, relationshipType) {
        return this.worldModel.areRelated(entity1Id, entity2Id, relationshipType);
    }
    addRelationship(entity1Id, entity2Id, relationshipType) {
        this.worldModel.addRelationship(entity1Id, entity2Id, relationshipType);
    }
    removeRelationship(entity1Id, entity2Id, relationshipType) {
        this.worldModel.removeRelationship(entity1Id, entity2Id, relationshipType);
    }
    // Utility Methods
    getTotalWeight(entityId) {
        return this.worldModel.getTotalWeight(entityId);
    }
    wouldCreateLoop(entityId, targetId) {
        return this.worldModel.wouldCreateLoop(entityId, targetId);
    }
    findPath(fromRoomId, toRoomId) {
        return this.worldModel.findPath(fromRoomId, toRoomId);
    }
    getPlayer() {
        return this.worldModel.getPlayer();
    }
    setPlayer(entityId) {
        this.worldModel.setPlayer(entityId);
    }
    // Convenience Creators
    connectRooms(room1Id, room2Id, direction) {
        this.worldModel.connectRooms(room1Id, room2Id, direction);
    }
    createDoor(displayName, opts) {
        return this.worldModel.createDoor(displayName, opts);
    }
    // Region Management (ADR-149)
    createRegion(id, options) {
        return this.worldModel.createRegion(id, options);
    }
    assignRoom(roomId, regionId) {
        this.worldModel.assignRoom(roomId, regionId);
    }
    isInRegion(entityId, regionId) {
        return this.worldModel.isInRegion(entityId, regionId);
    }
    getRegionCrossings(fromRoomId, toRoomId) {
        return this.worldModel.getRegionCrossings(fromRoomId, toRoomId);
    }
    // Scene Management (ADR-149)
    createScene(id, options) {
        return this.worldModel.createScene(id, options);
    }
    getSceneConditions(sceneId) {
        return this.worldModel.getSceneConditions(sceneId);
    }
    getAllSceneConditions() {
        return this.worldModel.getAllSceneConditions();
    }
    isSceneActive(sceneId) {
        return this.worldModel.isSceneActive(sceneId);
    }
    hasSceneEnded(sceneId) {
        return this.worldModel.hasSceneEnded(sceneId);
    }
    hasSceneHappened(sceneId) {
        return this.worldModel.hasSceneHappened(sceneId);
    }
    // Capability Management
    registerCapability(name, registration) {
        this.worldModel.registerCapability(name, registration);
    }
    updateCapability(name, data) {
        this.worldModel.updateCapability(name, data);
    }
    getCapability(name) {
        return this.worldModel.getCapability(name);
    }
    hasCapability(name) {
        return this.worldModel.hasCapability(name);
    }
    // Score Ledger
    awardScore(id, points, description) {
        return this.worldModel.awardScore(id, points, description);
    }
    revokeScore(id) {
        return this.worldModel.revokeScore(id);
    }
    hasScore(id) {
        return this.worldModel.hasScore(id);
    }
    getScore() {
        return this.worldModel.getScore();
    }
    getScoreEntries() {
        return this.worldModel.getScoreEntries();
    }
    setMaxScore(max) {
        this.worldModel.setMaxScore(max);
    }
    getMaxScore() {
        return this.worldModel.getMaxScore();
    }
    // Persistence
    toJSON() {
        return this.worldModel.toJSON();
    }
    loadJSON(json) {
        this.worldModel.loadJSON(json);
    }
    clear() {
        this.worldModel.clear();
    }
    // Event System
    registerEventHandler(eventType, handler) {
        this.worldModel.registerEventHandler(eventType, handler);
    }
    unregisterEventHandler(eventType) {
        this.worldModel.unregisterEventHandler(eventType);
    }
    registerEventValidator(eventType, validator) {
        this.worldModel.registerEventValidator(eventType, validator);
    }
    registerEventPreviewer(eventType, previewer) {
        this.worldModel.registerEventPreviewer(eventType, previewer);
    }
    connectEventProcessor(wiring) {
        this.worldModel.connectEventProcessor(wiring);
    }
    chainEvent(triggerType, handler, options) {
        this.worldModel.chainEvent(triggerType, handler, options);
    }
    applyEvent(event) {
        this.worldModel.applyEvent(event);
    }
    canApplyEvent(event) {
        return this.worldModel.canApplyEvent(event);
    }
    previewEvent(event) {
        return this.worldModel.previewEvent(event);
    }
    getAppliedEvents() {
        return this.worldModel.getAppliedEvents();
    }
    getEventsSince(timestamp) {
        return this.worldModel.getEventsSince(timestamp);
    }
    clearEventHistory() {
        this.worldModel.clearEventHistory();
    }
    // Scope Management
    getScopeRegistry() {
        return this.worldModel.getScopeRegistry();
    }
    addScopeRule(rule) {
        this.worldModel.addScopeRule(rule);
    }
    removeScopeRule(ruleId) {
        return this.worldModel.removeScopeRule(ruleId);
    }
    evaluateScope(actorId, actionId) {
        return this.worldModel.evaluateScope(actorId, actionId);
    }
    // Vocabulary Management
    getGrammarVocabularyProvider() {
        return this.worldModel.getGrammarVocabularyProvider();
    }
    // ========== Author-Only Convenience Methods ==========
    /**
     * Move multiple entities to a container in one operation.
     */
    populate(containerId, entityIds) {
        for (const entityId of entityIds) {
            this.moveEntity(entityId, containerId);
        }
    }
    /**
     * Add a trait to an entity.
     */
    addTrait(entityId, trait) {
        const entity = this.dataStore.entities.get(entityId);
        if (entity) {
            entity.add(trait);
        }
    }
    /**
     * Remove a trait from an entity.
     */
    removeTrait(entityId, traitType) {
        const entity = this.dataStore.entities.get(entityId);
        if (entity) {
            entity.remove(traitType);
        }
    }
    // ========== Private Helpers ==========
    generateId(type) {
        const TYPE_PREFIXES = {
            'room': 'r',
            'door': 'd',
            'item': 'i',
            'actor': 'a',
            'container': 'c',
            'supporter': 's',
            'scenery': 'y',
            'exit': 'e',
            'object': 'o'
        };
        const prefix = TYPE_PREFIXES[type] || TYPE_PREFIXES['object'];
        const counter = this.dataStore.idCounters.get(prefix) || 0;
        const nextCounter = counter + 1;
        if (nextCounter > 1295) {
            throw new Error(`ID overflow for type '${type}' (prefix '${prefix}')`);
        }
        this.dataStore.idCounters.set(prefix, nextCounter);
        const base36 = nextCounter.toString(36).padStart(2, '0');
        return `${prefix}${base36}`;
    }
}
exports.AuthorModel = AuthorModel;
