// @sharpee/world-model/world/index.js
"use strict";
// index.ts - World model exports
Object.defineProperty(exports, "__esModule", { value: true });
exports.StandardCapabilities = exports.AuthorModel = exports.VisibilityBehavior = exports.SpatialIndex = exports.GrammarVocabularyProvider = exports.WorldModel = void 0;
// Export all types and the WorldModel interface/class from the same file
var WorldModel_1 = require("./WorldModel");
Object.defineProperty(exports, "WorldModel", { enumerable: true, get: function () { return WorldModel_1.WorldModel; } });
// Re-export grammar vocabulary types from if-domain for convenience
var if_domain_1 = require("../../if-domain/index.js");
Object.defineProperty(exports, "GrammarVocabularyProvider", { enumerable: true, get: function () { return if_domain_1.GrammarVocabularyProvider; } });
// Export other classes
var SpatialIndex_1 = require("./SpatialIndex");
Object.defineProperty(exports, "SpatialIndex", { enumerable: true, get: function () { return SpatialIndex_1.SpatialIndex; } });
var VisibilityBehavior_1 = require("./VisibilityBehavior");
Object.defineProperty(exports, "VisibilityBehavior", { enumerable: true, get: function () { return VisibilityBehavior_1.VisibilityBehavior; } });
var AuthorModel_1 = require("./AuthorModel");
Object.defineProperty(exports, "AuthorModel", { enumerable: true, get: function () { return AuthorModel_1.AuthorModel; } });
// Export capability types
var capabilities_1 = require("./capabilities");
Object.defineProperty(exports, "StandardCapabilities", { enumerable: true, get: function () { return capabilities_1.StandardCapabilities; } });
