// @sharpee/world-model/world/ScoreLedger.js
"use strict";
/**
 * ScoreLedger — score tracking subsystem for WorldModel.
 *
 * Public interface: award, revoke, has, getTotal, getEntries, setMax, getMax,
 * toJSON, fromJSON, clear.
 *
 * Owner context: packages/world-model (ADR-129). Extracted from WorldModel
 * to isolate scoring concerns and simplify serialization.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ScoreLedger = void 0;
/**
 * Tracks score entries awarded during gameplay.
 * Each entry has a unique ID to prevent double-awards and support revocation.
 */
class ScoreLedger {
    ledger = [];
    maxScore = 0;
    /**
     * Award points for an achievement. Returns false if already awarded.
     *
     * @param id - Unique identifier for this score entry
     * @param points - Points to award
     * @param description - Human-readable description
     * @returns true if awarded, false if id was already in the ledger
     */
    award(id, points, description) {
        if (this.ledger.some(e => e.id === id)) {
            return false;
        }
        this.ledger.push({ id, points, description });
        return true;
    }
    /**
     * Revoke a previously awarded score entry.
     *
     * @param id - The entry ID to revoke
     * @returns true if revoked, false if not found
     */
    revoke(id) {
        const idx = this.ledger.findIndex(e => e.id === id);
        if (idx < 0)
            return false;
        this.ledger.splice(idx, 1);
        return true;
    }
    /**
     * Check if a score entry has been awarded.
     *
     * @param id - The entry ID to check
     * @returns true if the entry exists in the ledger
     */
    has(id) {
        return this.ledger.some(e => e.id === id);
    }
    /**
     * Get the total score (sum of all entry points).
     *
     * @returns Total points across all entries
     */
    getTotal() {
        let total = 0;
        for (const entry of this.ledger) {
            total += entry.points;
        }
        return total;
    }
    /**
     * Get a copy of all score entries.
     *
     * @returns Array of score entries (defensive copy)
     */
    getEntries() {
        return [...this.ledger];
    }
    /**
     * Set the maximum possible score.
     *
     * @param max - Maximum score value
     */
    setMax(max) {
        this.maxScore = max;
    }
    /**
     * Get the maximum possible score.
     *
     * @returns Maximum score value
     */
    getMax() {
        return this.maxScore;
    }
    /**
     * Serialize the ledger for persistence.
     *
     * @returns Serializable data object
     */
    toJSON() {
        return {
            scoreLedger: this.ledger,
            scoreMaxScore: this.maxScore,
        };
    }
    /**
     * Restore the ledger from serialized data.
     *
     * @param data - Previously serialized ledger data
     */
    fromJSON(data) {
        if (data.scoreLedger) {
            this.ledger = data.scoreLedger;
        }
        if (data.scoreMaxScore !== undefined) {
            this.maxScore = data.scoreMaxScore;
        }
    }
    /**
     * Reset the ledger to empty state.
     */
    clear() {
        this.ledger = [];
        this.maxScore = 0;
    }
}
exports.ScoreLedger = ScoreLedger;
