// @sharpee/world-model/world/SharedDataStore.js
"use strict";
// SharedDataStore.ts - Shared state wrapper for WorldModel and AuthorModel
Object.defineProperty(exports, "__esModule", { value: true });
exports.SharedDataStore = void 0;
const SpatialIndex_1 = require("./SpatialIndex");
/**
 * Wrapper class that holds references to shared state between WorldModel and AuthorModel.
 * This ensures that changes made by one model are visible to the other.
 */
class SharedDataStore {
    entities;
    spatialIndex;
    state;
    _playerId;
    relationships;
    idCounters;
    nameToId;
    idToName;
    constructor() {
        this.entities = new Map();
        this.spatialIndex = new SpatialIndex_1.SpatialIndex();
        this.state = {};
        this._playerId = undefined;
        this.relationships = new Map();
        this.idCounters = new Map();
        this.nameToId = new Map();
        this.idToName = new Map();
    }
    get playerId() {
        return this._playerId;
    }
    set playerId(value) {
        this._playerId = value;
    }
    /**
     * Clear all data in the store
     */
    clear() {
        this.entities.clear();
        this.spatialIndex.clear();
        this.state = {};
        this._playerId = undefined;
        this.relationships.clear();
        this.idCounters.clear();
        this.nameToId.clear();
        this.idToName.clear();
    }
}
exports.SharedDataStore = SharedDataStore;
