// @sharpee/world-model/world/SpatialIndex.js
"use strict";
// SpatialIndex.ts - Efficient spatial relationship tracking
Object.defineProperty(exports, "__esModule", { value: true });
exports.SpatialIndex = void 0;
class SpatialIndex {
    // Map from parent ID to set of child IDs
    parentToChildren = new Map();
    // Map from child ID to parent ID
    childToParent = new Map();
    addChild(parentId, childId) {
        // Remove from any existing parent (but preserve this entity's children)
        const currentParent = this.childToParent.get(childId);
        if (currentParent) {
            this.removeChild(currentParent, childId);
        }
        // Add to new parent
        if (!this.parentToChildren.has(parentId)) {
            this.parentToChildren.set(parentId, new Set());
        }
        this.parentToChildren.get(parentId).add(childId);
        this.childToParent.set(childId, parentId);
    }
    removeChild(parentId, childId) {
        // Always clean the actual parent's children set, not just the specified one.
        // This prevents stale data when removeChild is called with the wrong parentId.
        const actualParent = this.childToParent.get(childId);
        const targetParent = actualParent ?? parentId;
        const children = this.parentToChildren.get(targetParent);
        if (children) {
            children.delete(childId);
            if (children.size === 0) {
                this.parentToChildren.delete(targetParent);
            }
        }
        this.childToParent.delete(childId);
    }
    remove(entityId) {
        // Remove as child from its parent
        const parent = this.childToParent.get(entityId);
        if (parent) {
            this.removeChild(parent, entityId);
        }
        // Remove all children relationships (this entity is being deleted)
        const children = this.parentToChildren.get(entityId);
        if (children) {
            // Copy to avoid mutation during iteration
            const childArray = Array.from(children);
            childArray.forEach(childId => {
                this.childToParent.delete(childId);
            });
            this.parentToChildren.delete(entityId);
        }
    }
    getParent(childId) {
        return this.childToParent.get(childId);
    }
    getChildren(parentId) {
        const children = this.parentToChildren.get(parentId);
        return children ? Array.from(children) : [];
    }
    hasChildren(parentId) {
        const children = this.parentToChildren.get(parentId);
        return children ? children.size > 0 : false;
    }
    getAllDescendants(parentId, maxDepth = 10) {
        const result = [];
        const visited = new Set();
        const traverse = (id, depth) => {
            if (depth > maxDepth || visited.has(id))
                return;
            visited.add(id);
            const children = this.getChildren(id);
            result.push(...children);
            children.forEach(childId => traverse(childId, depth + 1));
        };
        traverse(parentId, 0);
        return result;
    }
    getAncestors(childId, maxDepth = 10) {
        const result = [];
        let current = childId;
        let depth = 0;
        while (depth < maxDepth) {
            const parent = this.getParent(current);
            if (!parent)
                break;
            result.push(parent);
            current = parent;
            depth++;
        }
        return result;
    }
    clear() {
        this.parentToChildren.clear();
        this.childToParent.clear();
    }
    toJSON() {
        return {
            parentToChildren: Array.from(this.parentToChildren.entries()).map(([parent, children]) => ({
                parent,
                children: Array.from(children)
            })),
            childToParent: Array.from(this.childToParent.entries())
        };
    }
    loadJSON(data) {
        this.clear();
        if (data.parentToChildren) {
            for (const { parent, children } of data.parentToChildren) {
                this.parentToChildren.set(parent, new Set(children));
            }
        }
        if (data.childToParent) {
            for (const [child, parent] of data.childToParent) {
                this.childToParent.set(child, parent);
            }
        }
    }
}
exports.SpatialIndex = SpatialIndex;
