// @sharpee/world-model/world/VisibilityBehavior.js
"use strict";
// VisibilityBehavior.ts - Visibility system for IF
Object.defineProperty(exports, "__esModule", { value: true });
exports.VisibilityBehavior = exports.VISIBILITY_CAPABILITY = void 0;
const behavior_1 = require("../behaviors/behavior");
const trait_types_1 = require("../traits/trait-types");
const switchableBehavior_1 = require("../traits/switchable/switchableBehavior");
const vehicleTrait_1 = require("../traits/vehicle/vehicleTrait");
const roomTrait_1 = require("../traits/room/roomTrait");
const containerTrait_1 = require("../traits/container/containerTrait");
const openableTrait_1 = require("../traits/openable/openableTrait");
const lightSourceTrait_1 = require("../traits/light-source/lightSourceTrait");
const sceneryTrait_1 = require("../traits/scenery/sceneryTrait");
const identityTrait_1 = require("../traits/identity/identityTrait");
const capabilities_1 = require("../capabilities");
/**
 * Standard capability ID for visibility control.
 * Entities can claim this capability to control their own visibility.
 */
exports.VISIBILITY_CAPABILITY = 'if.scope.visible';
class VisibilityBehavior extends behavior_1.Behavior {
    static requiredTraits = [];
    /**
     * Determines if a room is effectively dark (no usable light sources).
     * This is the single source of truth for darkness checking.
     *
     * A room is dark if:
     * 1. It has RoomTrait with isDark = true
     * 2. There are no accessible, active light sources
     *
     * @param room - The room entity to check
     * @param world - The world model
     * @returns true if the room is dark and has no accessible light sources
     */
    static isDark(room, world) {
        const roomTrait = room.getTrait(roomTrait_1.RoomTrait);
        if (!roomTrait || !roomTrait.isDark) {
            return false; // Room isn't marked as dark
        }
        return !this.hasLightSource(room, world);
    }
    /**
     * Determines if a light source entity is currently providing light.
     * Checks isLit property first, falls back to switchable state, defaults to lit.
     *
     * Priority order:
     * 1. Explicit isLit property (e.g., lit torch, extinguished candle)
     * 2. Switchable trait state (e.g., flashlight isOn: true)
     * 3. Default to lit (e.g., glowing gems, phosphorescent moss)
     */
    static isLightActive(entity) {
        const lightTrait = entity.getTrait(lightSourceTrait_1.LightSourceTrait);
        // Explicit isLit property takes precedence
        if (lightTrait?.isLit !== undefined) {
            return lightTrait.isLit === true;
        }
        // Fall back to switchable state
        if (entity.hasTrait(trait_types_1.TraitType.SWITCHABLE)) {
            return switchableBehavior_1.SwitchableBehavior.isOn(entity);
        }
        // Default: light sources without explicit state are lit
        return true;
    }
    /**
     * Determines if an observer can see a target entity
     */
    static canSee(observer, target, world) {
        // Can always see self
        if (observer.id === target.id)
            return true;
        // Check if target is invisible via SceneryTrait
        const targetScenery = target.getTrait(sceneryTrait_1.SceneryTrait);
        if (targetScenery && targetScenery.visible === false) {
            return false;
        }
        // Check if target has visibility capability that blocks being seen
        const visibilityTrait = (0, capabilities_1.findTraitWithCapability)(target, exports.VISIBILITY_CAPABILITY);
        if (visibilityTrait) {
            const behavior = (0, capabilities_1.getBehaviorForCapability)(visibilityTrait, exports.VISIBILITY_CAPABILITY);
            if (behavior) {
                const result = behavior.validate(target, world, observer.id, {});
                if (!result.valid) {
                    return false; // Entity blocks visibility via capability
                }
            }
        }
        // Get rooms
        const observerRoom = world.getContainingRoom(observer.id);
        const targetRoom = world.getContainingRoom(target.id);
        // Must be in same room (or target is the room)
        if (target.hasTrait(trait_types_1.TraitType.ROOM)) {
            // Check if we're in the target room
            if (observerRoom?.id === target.id) {
                // Always know what room you're in, even in darkness
                return true;
            }
            return false;
        }
        if (!observerRoom || observerRoom.id !== targetRoom?.id) {
            return false;
        }
        // Check if room is dark
        const roomTrait = observerRoom.getTrait(roomTrait_1.RoomTrait);
        if (roomTrait && roomTrait.isDark) {
            // In a dark room, need light to see
            if (!this.hasLightSource(observerRoom, world)) {
                // Special cases in darkness:
                // 1. Can see lit light sources
                if (target.hasTrait(trait_types_1.TraitType.LIGHT_SOURCE) &&
                    target.getTrait(lightSourceTrait_1.LightSourceTrait)?.isLit === true) {
                    return true;
                }
                // 2. Can see items you're carrying (by feel)
                const targetLocation = world.getLocation(target.id);
                if (targetLocation === observer.id) {
                    return true;
                }
                // Otherwise can't see in darkness
                return false;
            }
        }
        // Check line of sight through containers
        const hasLineOfSight = this.hasLineOfSight(observer.id, target.id, world);
        return hasLineOfSight;
    }
    /**
     * Gets all entities visible to an observer
     */
    static getVisible(observer, world) {
        const visible = [];
        const seen = new Set(); // Track what we've already added
        const observerRoom = world.getContainingRoom(observer.id);
        if (!observerRoom)
            return visible;
        // Add the room itself if visible
        if (this.canSee(observer, observerRoom, world)) {
            visible.push(observerRoom);
            seen.add(observerRoom.id);
        }
        // Check if room is dark
        const roomTrait = observerRoom.getTrait(roomTrait_1.RoomTrait);
        const isDark = roomTrait && roomTrait.isDark;
        const hasLight = this.hasLightSource(observerRoom, world);
        // If it's dark and no light, only see specific things
        if (isDark && !hasLight) {
            // Can always see what you're carrying (by feel)
            const carried = world.getContents(observer.id);
            for (const entity of carried) {
                if (!seen.has(entity.id)) {
                    visible.push(entity);
                    seen.add(entity.id);
                }
            }
            // Add any lit light sources in the room
            const roomContents = world.getAllContents(observerRoom.id, { recursive: true });
            for (const entity of roomContents) {
                if (entity.hasTrait(trait_types_1.TraitType.LIGHT_SOURCE)) {
                    const light = entity.getTrait(lightSourceTrait_1.LightSourceTrait);
                    if (light && light.isLit === true && !seen.has(entity.id)) {
                        visible.push(entity);
                        seen.add(entity.id);
                    }
                }
            }
            return visible;
        }
        // Get contents of the room
        const roomContents = world.getContents(observerRoom.id);
        // Add visible entities in the room
        for (const entity of roomContents) {
            if (entity.id !== observer.id && !seen.has(entity.id)) {
                // Check if entity is concealed (hidden until revealed via SEARCH or game event)
                const identity = entity.getTrait(identityTrait_1.IdentityTrait);
                if (identity && identity.concealed === true) {
                    continue;
                }
                // Check if entity is visible via SceneryTrait
                const scenery = entity.getTrait(sceneryTrait_1.SceneryTrait);
                if (scenery && scenery.visible === false) {
                    continue;
                }
                // Check if entity has visibility capability that blocks being seen
                const visibilityTrait = (0, capabilities_1.findTraitWithCapability)(entity, exports.VISIBILITY_CAPABILITY);
                if (visibilityTrait) {
                    const behavior = (0, capabilities_1.getBehaviorForCapability)(visibilityTrait, exports.VISIBILITY_CAPABILITY);
                    if (behavior) {
                        const result = behavior.validate(entity, world, observer.id, {});
                        if (!result.valid) {
                            continue; // Entity blocks visibility via capability
                        }
                    }
                }
                visible.push(entity);
                seen.add(entity.id);
                // Check contents of this entity if it's a container/supporter/actor
                if (entity.hasTrait(trait_types_1.TraitType.CONTAINER) ||
                    entity.hasTrait(trait_types_1.TraitType.SUPPORTER) ||
                    entity.hasTrait(trait_types_1.TraitType.ACTOR)) {
                    this.addVisibleContents(entity, visible, seen, world);
                }
            }
        }
        // Add carried items
        const carried = world.getContents(observer.id);
        for (const entity of carried) {
            if (!seen.has(entity.id)) {
                visible.push(entity);
                seen.add(entity.id);
                // Check contents of carried containers
                if (entity.hasTrait(trait_types_1.TraitType.CONTAINER) || entity.hasTrait(trait_types_1.TraitType.SUPPORTER)) {
                    this.addVisibleContents(entity, visible, seen, world);
                }
            }
        }
        return visible;
    }
    /**
     * Recursively adds visible contents of a container/supporter/actor
     */
    static addVisibleContents(container, visible, seen, world) {
        // Check if we can see inside this container
        if (container.hasTrait(trait_types_1.TraitType.CONTAINER)) {
            const containerTrait = container.getTrait(containerTrait_1.ContainerTrait);
            const isTransparent = containerTrait?.isTransparent ?? false;
            if (!isTransparent && container.hasTrait(trait_types_1.TraitType.OPENABLE)) {
                const openable = container.getTrait(openableTrait_1.OpenableTrait);
                const isOpen = openable?.isOpen ?? false;
                if (!isOpen) {
                    return; // Can't see inside closed opaque container
                }
            }
        }
        // Get contents (including worn items for actors)
        const contents = world.getContents(container.id, { includeWorn: true });
        for (const entity of contents) {
            if (!seen.has(entity.id)) {
                // Check if entity is visible via SceneryTrait
                const scenery = entity.getTrait(sceneryTrait_1.SceneryTrait);
                if (scenery && scenery.visible === false) {
                    continue;
                }
                // Check if entity has visibility capability that blocks being seen
                const visibilityTrait = (0, capabilities_1.findTraitWithCapability)(entity, exports.VISIBILITY_CAPABILITY);
                if (visibilityTrait) {
                    const behavior = (0, capabilities_1.getBehaviorForCapability)(visibilityTrait, exports.VISIBILITY_CAPABILITY);
                    if (behavior) {
                        // Note: We don't have a direct observer here, use container's id as proxy
                        const result = behavior.validate(entity, world, container.id, {});
                        if (!result.valid) {
                            continue; // Entity blocks visibility via capability
                        }
                    }
                }
                visible.push(entity);
                seen.add(entity.id);
                // Recurse into nested containers
                if (entity.hasTrait(trait_types_1.TraitType.CONTAINER) ||
                    entity.hasTrait(trait_types_1.TraitType.SUPPORTER) ||
                    entity.hasTrait(trait_types_1.TraitType.ACTOR)) {
                    this.addVisibleContents(entity, visible, seen, world);
                }
            }
        }
    }
    /**
     * Checks if a room has any accessible, active light sources.
     * Handles nested containers, worn items, and various light source types.
     */
    static hasLightSource(room, world) {
        // Check all entities in the room, including worn items
        const contents = world.getAllContents(room.id, { recursive: true, includeWorn: true });
        for (const entity of contents) {
            if (!entity.hasTrait(trait_types_1.TraitType.LIGHT_SOURCE))
                continue;
            // Check if light is active (isLit, switchable, or default)
            if (!this.isLightActive(entity))
                continue;
            // Check if the light is accessible (not in a closed container)
            if (!this.isAccessible(entity.id, room.id, world))
                continue;
            return true;
        }
        return false;
    }
    /**
     * Walks the containment chain from an entity upward, checking whether any
     * closed opaque container blocks the path.
     *
     * This is the single implementation of the container-walk algorithm used by
     * `isAccessible`, `hasLineOfSight`, and `isVisible`.
     *
     * At each hop:
     * - Actors are transparent (carried/worn items are always reachable)
     * - Opaque closed containers block
     * - Transparent containers, open containers, and non-openable opaque containers pass
     *
     * @param entityId - Starting entity to walk upward from
     * @param world - The world model
     * @param stopAtId - Stop when this ancestor is reached (e.g., the room).
     *                   If omitted, walks until reaching a room or the top of the tree.
     * @returns true if no closed opaque container blocks the path
     */
    static isContainmentPathClear(entityId, world, stopAtId) {
        let current = entityId;
        while (true) {
            if (stopAtId !== undefined && current === stopAtId)
                return true;
            const parentId = world.getLocation(current);
            if (!parentId)
                return true;
            const container = world.getEntity(parentId);
            if (!container)
                return true;
            // Actors don't block access/visibility of their contents
            if (container.hasTrait(trait_types_1.TraitType.ACTOR)) {
                current = parentId;
                continue;
            }
            // Opaque + closed containers block the path
            if (container.hasTrait(trait_types_1.TraitType.CONTAINER)) {
                const containerTrait = container.getTrait(containerTrait_1.ContainerTrait);
                const isTransparent = containerTrait?.isTransparent ?? false;
                if (!isTransparent && container.hasTrait(trait_types_1.TraitType.OPENABLE)) {
                    const openable = container.getTrait(openableTrait_1.OpenableTrait);
                    const isOpen = openable?.isOpen ?? false;
                    if (!isOpen) {
                        return false;
                    }
                }
            }
            // Stop at rooms (natural top of containment)
            if (container.hasTrait(trait_types_1.TraitType.ROOM))
                return true;
            current = parentId;
        }
    }
    /**
     * Checks if an entity is accessible from a room (not blocked by closed containers)
     */
    static isAccessible(entityId, roomId, world) {
        return this.isContainmentPathClear(entityId, world, roomId);
    }
    /**
     * Checks if there's a line of sight between observer and target.
     * Walks the target's containment chain to verify no closed opaque container
     * blocks visibility.
     */
    static hasLineOfSight(observerId, targetId, world) {
        return this.isContainmentPathClear(targetId, world);
    }
    /**
     * Checks if an entity is visible in its current context
     * (used for filtering queries)
     */
    static isVisible(entity, world) {
        // Check if explicitly invisible via SceneryTrait
        const scenery = entity.getTrait(sceneryTrait_1.SceneryTrait);
        if (scenery && scenery.visible === false) {
            return false;
        }
        // Check if entity has visibility capability that blocks being seen
        // Note: For isVisible we don't have an observer, so we pass empty string
        const visibilityTrait = (0, capabilities_1.findTraitWithCapability)(entity, exports.VISIBILITY_CAPABILITY);
        if (visibilityTrait) {
            const behavior = (0, capabilities_1.getBehaviorForCapability)(visibilityTrait, exports.VISIBILITY_CAPABILITY);
            if (behavior) {
                const result = behavior.validate(entity, world, '', {});
                if (!result.valid) {
                    return false; // Entity blocks visibility via capability
                }
            }
        }
        // Check containment path for closed opaque containers
        return this.isContainmentPathClear(entity.id, world);
    }
    /**
     * Gets the location that should be described when an observer looks around.
     * Accounts for vehicle/container transparency:
     * - In a transparent vehicle → describe the room
     * - In an open container → describe the room
     * - In a closed container → describe the container
     * - In a room → describe the room
     *
     * @returns The entity to describe and optionally the immediate container
     */
    static getDescribableLocation(observer, world) {
        const immediateLocationId = world.getLocation(observer.id);
        if (!immediateLocationId) {
            // Observer is not in anything - unusual but handle it
            return { location: observer, immediateContainer: null };
        }
        const immediateLocation = world.getEntity(immediateLocationId);
        if (!immediateLocation) {
            return { location: observer, immediateContainer: null };
        }
        // If immediate location is a room, we're done
        if (immediateLocation.hasTrait(trait_types_1.TraitType.ROOM)) {
            return { location: immediateLocation, immediateContainer: null };
        }
        // Check if we're in a vehicle
        if (immediateLocation.hasTrait(trait_types_1.TraitType.VEHICLE)) {
            const vehicleTrait = immediateLocation.getTrait(vehicleTrait_1.VehicleTrait);
            if (vehicleTrait?.transparent) {
                // Transparent vehicle - describe the room
                const room = world.getContainingRoom(observer.id);
                if (room) {
                    return { location: room, immediateContainer: immediateLocation };
                }
            }
            // Opaque vehicle - describe the vehicle interior
            return { location: immediateLocation, immediateContainer: null };
        }
        // Check if we're in a container
        if (immediateLocation.hasTrait(trait_types_1.TraitType.CONTAINER)) {
            // Check if container is open (visibility passes through)
            if (immediateLocation.hasTrait(trait_types_1.TraitType.OPENABLE)) {
                const openable = immediateLocation.getTrait(openableTrait_1.OpenableTrait);
                if (openable?.isOpen) {
                    // Open container - describe the room
                    const room = world.getContainingRoom(observer.id);
                    if (room) {
                        return { location: room, immediateContainer: immediateLocation };
                    }
                }
            }
            // Closed container - describe the container
            return { location: immediateLocation, immediateContainer: null };
        }
        // Default: describe what we're in
        return { location: immediateLocation, immediateContainer: null };
    }
}
exports.VisibilityBehavior = VisibilityBehavior;
