// @sharpee/world-model/world/WorldModel.js
"use strict";
// WorldModel.ts - Core world model interface and implementation for Sharpee IF Platform
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorldModel = void 0;
const if_entity_1 = require("../entities/if-entity");
const entity_types_1 = require("../entities/entity-types");
const trait_types_1 = require("../traits/trait-types");
const room_1 = require("../traits/room");
const roomBehavior_1 = require("../traits/room/roomBehavior");
const regionTrait_1 = require("../traits/region/regionTrait");
const sceneTrait_1 = require("../traits/scene/sceneTrait");
const door_1 = require("../traits/door");
const scenery_1 = require("../traits/scenery");
const identityTrait_1 = require("../traits/identity/identityTrait");
const openableTrait_1 = require("../traits/openable/openableTrait");
const lockableTrait_1 = require("../traits/lockable/lockableTrait");
const wearableTrait_1 = require("../traits/wearable/wearableTrait");
const clothingTrait_1 = require("../traits/clothing/clothingTrait");
const exitTrait_1 = require("../traits/exit/exitTrait");
const directions_1 = require("../constants/directions");
const SpatialIndex_1 = require("./SpatialIndex");
const VisibilityBehavior_1 = require("./VisibilityBehavior");
const WorldSerializer_1 = require("./WorldSerializer");
const container_utils_1 = require("../traits/container/container-utils");
const if_domain_1 = require("../../if-domain/index.js");
const scope_registry_1 = require("../scope/scope-registry");
const scope_evaluator_1 = require("../scope/scope-evaluator");
// Event system — extracted to WorldEventSystem.ts, re-exported for backward compat
const WorldEventSystem_1 = require("./WorldEventSystem");
// Score Ledger (ADR-129)
const ScoreLedger_1 = require("./ScoreLedger");
// Type prefixes for entity ID generation
const TYPE_PREFIXES = {
    'room': 'r',
    'door': 'd',
    'item': 'i',
    'actor': 'a',
    'container': 'c',
    'supporter': 's',
    'scenery': 'y',
    'exit': 'e',
    'object': 'o' // default
};
class WorldModel {
    entities = new Map();
    state = {};
    playerId;
    spatialIndex;
    config;
    capabilities = {};
    // Score Ledger (ADR-129)
    scoreLedger = new ScoreLedger_1.ScoreLedger();
    // Scene condition closures (ADR-149) — not serialized
    sceneConditions = new Map();
    // ID generation
    idCounters = new Map();
    // Event system (ADR-086, ADR-094) — delegates to WorldEventSystem
    eventSystem = new WorldEventSystem_1.WorldEventSystem();
    // Persistence — delegates to WorldSerializer
    serializer = new WorldSerializer_1.WorldSerializer();
    platformEvents;
    // Scope system
    scopeRegistry;
    scopeEvaluator;
    // Vocabulary system (ADR-082)
    grammarVocabularyProvider;
    constructor(config = {}, platformEvents) {
        this.config = {
            enableSpatialIndex: true,
            maxDepth: 10,
            strictMode: false,
            ...config
        };
        this.spatialIndex = new SpatialIndex_1.SpatialIndex();
        this.platformEvents = platformEvents;
        this.eventSystem.setWorldRef(this);
        // Initialize scope system
        this.scopeRegistry = new scope_registry_1.ScopeRegistry();
        this.scopeEvaluator = new scope_evaluator_1.RuleScopeEvaluator(this.scopeRegistry);
        // Initialize vocabulary system (ADR-082)
        this.grammarVocabularyProvider = new if_domain_1.GrammarVocabularyProvider();
        // Register default scope rules
        this.registerDefaultScopeRules();
    }
    // Platform event emission helper
    emitPlatformEvent(type, data) {
        if (this.platformEvents) {
            this.platformEvents.addEvent({
                id: `world_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`,
                timestamp: Date.now(),
                type: `platform.world.${type}`,
                entities: {},
                data: {
                    subsystem: 'world',
                    ...data
                },
                tags: ['platform', 'world', 'debug'],
                priority: 0,
                narrate: false
            });
        }
    }
    // Capability Management
    registerCapability(name, registration = {}) {
        if (this.capabilities[name]) {
            if (this.config.strictMode) {
                throw new Error(`Capability '${name}' is already registered`);
            }
            return;
        }
        // Initialize capability with schema and initial data
        const initialData = {};
        // Apply schema defaults
        if (registration.schema) {
            for (const [field, fieldDef] of Object.entries(registration.schema)) {
                if (fieldDef.default !== undefined) {
                    initialData[field] = fieldDef.default;
                }
            }
        }
        // Override with provided initial data
        if (registration.initialData) {
            Object.assign(initialData, registration.initialData);
        }
        this.capabilities[name] = {
            data: initialData,
            schema: registration.schema
        };
    }
    updateCapability(name, updates) {
        const capability = this.capabilities[name];
        if (!capability) {
            if (this.config.strictMode) {
                throw new Error(`Capability '${name}' is not registered`);
            }
            return;
        }
        // Validate updates against schema if present
        if (capability.schema) {
            for (const [field, value] of Object.entries(updates)) {
                const fieldDef = capability.schema[field];
                if (fieldDef) {
                    // Basic type validation
                    const actualType = Array.isArray(value) ? 'array' : typeof value;
                    if (actualType !== fieldDef.type && value !== null && value !== undefined) {
                        if (this.config.strictMode) {
                            throw new Error(`Invalid type for capability '${name}' field '${field}': expected ${fieldDef.type}, got ${actualType}`);
                        }
                    }
                }
            }
        }
        // Apply updates
        Object.assign(capability.data, updates);
    }
    getCapability(name) {
        return this.capabilities[name]?.data;
    }
    hasCapability(name) {
        return name in this.capabilities;
    }
    // ID Generation
    generateId(type) {
        const prefix = TYPE_PREFIXES[type] || TYPE_PREFIXES['object'];
        const counter = this.idCounters.get(prefix) || 0;
        // Increment counter
        const nextCounter = counter + 1;
        // Check for overflow (base36 with 2 chars = 1296 max)
        if (nextCounter > 1295) {
            throw new Error(`ID overflow for type '${type}' (prefix '${prefix}'). Maximum 1296 entities per type.`);
        }
        this.idCounters.set(prefix, nextCounter);
        // Convert to base36 and pad to 2 characters
        const base36 = nextCounter.toString(36).padStart(2, '0');
        return `${prefix}${base36}`;
    }
    // Entity Management
    createEntity(displayName, type = 'object') {
        // Validate entity type
        if (!(0, entity_types_1.isEntityType)(type)) {
            throw new Error(`Unknown entity type: '${type}'. Valid types are: ${Object.values(entity_types_1.EntityType).join(', ')}`);
        }
        // Generate ID based on type
        const id = this.generateId(type);
        const entity = new if_entity_1.IFEntity(id, type, {
            attributes: {
                displayName: displayName,
                name: displayName, // For compatibility
                entityType: type
            }
        });
        // Add to entity map
        this.entities.set(id, entity);
        return entity;
    }
    getEntity(id) {
        return this.entities.get(id);
    }
    hasEntity(id) {
        return this.entities.has(id);
    }
    removeEntity(id) {
        const entity = this.entities.get(id);
        if (!entity)
            return false;
        // Remove from spatial index
        this.spatialIndex.remove(id);
        // Remove from any containers
        const location = this.getLocation(id);
        if (location) {
            this.moveEntity(id, null);
        }
        // Remove entity
        return this.entities.delete(id);
    }
    getAllEntities() {
        return Array.from(this.entities.values());
    }
    updateEntity(entityId, updater) {
        const entity = this.getEntity(entityId);
        if (!entity) {
            if (this.config.strictMode) {
                throw new Error(`Cannot update non-existent entity: ${entityId}`);
            }
            return;
        }
        // Call the updater - entity is mutable so changes happen in place
        updater(entity);
        // Future: Could emit change events here for reactive systems
        // this.emitChange({ type: 'entity-updated', entityId });
    }
    // Spatial Management
    getLocation(entityId) {
        return this.spatialIndex.getParent(entityId);
    }
    getContents(containerId, options = {}) {
        const contents = this.spatialIndex.getChildren(containerId);
        let entities = contents
            .map(id => this.getEntity(id))
            .filter((e) => e !== undefined);
        if (options.visibleOnly) {
            // Filter to only visible entities
            entities = entities.filter(e => {
                const scenery = e.getTrait(scenery_1.SceneryTrait);
                return !scenery || scenery.visible !== false;
            });
        }
        if (!options.includeWorn) {
            // Filter out worn items (check both WEARABLE and CLOTHING traits)
            entities = entities.filter(e => {
                const wearable = e.getTrait(wearableTrait_1.WearableTrait);
                const clothing = e.getTrait(clothingTrait_1.ClothingTrait);
                const wornFromWearable = wearable && wearable.isWorn;
                const wornFromClothing = clothing && clothing.isWorn;
                return !(wornFromWearable || wornFromClothing);
            });
        }
        return entities;
    }
    moveEntity(entityId, targetId) {
        if (!this.canMoveEntity(entityId, targetId)) {
            this.emitPlatformEvent('move_entity_failed', {
                entityId,
                targetId,
                reason: 'cannot_move'
            });
            return false;
        }
        // Remove from current location
        const currentLocation = this.getLocation(entityId);
        if (currentLocation) {
            this.spatialIndex.removeChild(currentLocation, entityId);
        }
        // Add to new location
        if (targetId) {
            this.spatialIndex.addChild(targetId, entityId);
        }
        // Emit platform event
        this.emitPlatformEvent('entity_moved', {
            entityId,
            fromLocation: currentLocation,
            toLocation: targetId
        });
        return true;
    }
    canMoveEntity(entityId, targetId) {
        const entity = this.getEntity(entityId);
        if (!entity)
            return false;
        // Can always remove from world
        if (targetId === null)
            return true;
        const target = this.getEntity(targetId);
        if (!target)
            return false;
        // Check for containment loops
        if (this.wouldCreateLoop(entityId, targetId)) {
            return false;
        }
        // Check if target can contain - using our new utility
        if (!(0, container_utils_1.canContain)(target)) {
            return false;
        }
        // Check container constraints
        if (target.hasTrait(trait_types_1.TraitType.CONTAINER) && target.hasTrait(trait_types_1.TraitType.OPENABLE)) {
            const openable = target.getTrait(openableTrait_1.OpenableTrait);
            if (openable && !openable.isOpen) {
                return false;
            }
        }
        return true;
    }
    getContainingRoom(entityId) {
        let current = entityId;
        let depth = 0;
        while (current && depth < this.config.maxDepth) {
            const parent = this.getLocation(current);
            if (!parent)
                return undefined;
            const parentEntity = this.getEntity(parent);
            if (parentEntity?.hasTrait(trait_types_1.TraitType.ROOM)) {
                return parentEntity;
            }
            current = parent;
            depth++;
        }
        return undefined;
    }
    getAllContents(entityId, options = {}) {
        const result = [];
        const visited = new Set();
        const traverse = (id, depth = 0, isRoot = false) => {
            if (visited.has(id) || depth > this.config.maxDepth)
                return;
            visited.add(id);
            // For the root entity, use the provided options
            // For recursive calls, always include worn items to get complete contents
            const contentsOptions = isRoot ? {
                visibleOnly: options.visibleOnly,
                includeWorn: options.includeWorn
            } : {
                visibleOnly: options.visibleOnly,
                includeWorn: true // Always include worn items when recursing
            };
            const contents = this.getContents(id, contentsOptions);
            result.push(...contents);
            if (options.recursive) {
                contents.forEach(item => traverse(item.id, depth + 1, false));
            }
        };
        traverse(entityId, 0, true);
        return result;
    }
    // World State Management
    getState() {
        return { ...this.state };
    }
    setState(state) {
        this.state = { ...state };
    }
    getStateValue(key) {
        return this.state[key];
    }
    setStateValue(key, value) {
        this.state[key] = value;
    }
    // Prompt (ADR-137)
    getPrompt() {
        return this.state[if_domain_1.PROMPT_STATE_KEY] ?? if_domain_1.DefaultPrompt;
    }
    setPrompt(prompt) {
        this.state[if_domain_1.PROMPT_STATE_KEY] = prompt;
    }
    // Query Operations
    findByTrait(traitType) {
        return this.findWhere(e => e.hasTrait(traitType));
    }
    findByType(entityType) {
        return this.findWhere(e => e.type === entityType);
    }
    findWhere(predicate) {
        return Array.from(this.entities.values()).filter(predicate);
    }
    // Scope methods moved to end of class to use new scope system
    // Relationship Queries
    relationships = new Map();
    getRelated(entityId, relationshipType) {
        const entityRels = this.relationships.get(entityId);
        if (!entityRels)
            return [];
        const related = entityRels.get(relationshipType);
        return related ? Array.from(related) : [];
    }
    areRelated(entity1Id, entity2Id, relationshipType) {
        const related = this.getRelated(entity1Id, relationshipType);
        return related.includes(entity2Id);
    }
    addRelationship(entity1Id, entity2Id, relationshipType) {
        // Ensure entities exist
        if (!this.hasEntity(entity1Id) || !this.hasEntity(entity2Id)) {
            if (this.config.strictMode) {
                throw new Error('Cannot add relationship between non-existent entities');
            }
            return;
        }
        // Add forward relationship
        if (!this.relationships.has(entity1Id)) {
            this.relationships.set(entity1Id, new Map());
        }
        const entityRels = this.relationships.get(entity1Id);
        if (!entityRels.has(relationshipType)) {
            entityRels.set(relationshipType, new Set());
        }
        entityRels.get(relationshipType).add(entity2Id);
    }
    removeRelationship(entity1Id, entity2Id, relationshipType) {
        const entityRels = this.relationships.get(entity1Id);
        if (!entityRels)
            return;
        const related = entityRels.get(relationshipType);
        if (related) {
            related.delete(entity2Id);
            if (related.size === 0) {
                entityRels.delete(relationshipType);
            }
        }
        if (entityRels.size === 0) {
            this.relationships.delete(entity1Id);
        }
    }
    // Utility Methods
    getTotalWeight(entityId) {
        const entity = this.getEntity(entityId);
        if (!entity)
            return 0;
        let weight = entity.weight || 0;
        // Add weight of all contents
        const contents = this.getAllContents(entityId, { recursive: true });
        for (const item of contents) {
            weight += item.weight || 0;
        }
        return weight;
    }
    wouldCreateLoop(entityId, targetId) {
        if (entityId === targetId)
            return true;
        let current = targetId;
        const visited = new Set();
        while (current) {
            if (visited.has(current))
                return false; // Already checked this path
            if (current === entityId)
                return true; // Found a loop
            visited.add(current);
            current = this.getLocation(current) || '';
        }
        return false;
    }
    findPath(fromRoomId, toRoomId) {
        if (fromRoomId === toRoomId)
            return [];
        const fromRoom = this.getEntity(fromRoomId);
        const toRoom = this.getEntity(toRoomId);
        if (!fromRoom?.hasTrait(trait_types_1.TraitType.ROOM) || !toRoom?.hasTrait(trait_types_1.TraitType.ROOM)) {
            return null;
        }
        // Simple BFS pathfinding
        const queue = [{ roomId: fromRoomId, path: [] }];
        const visited = new Set();
        while (queue.length > 0) {
            const { roomId, path } = queue.shift();
            if (visited.has(roomId))
                continue;
            visited.add(roomId);
            // Get exits from this room
            const room = this.getEntity(roomId);
            if (!room)
                continue;
            const roomTrait = room.getTrait(room_1.RoomTrait);
            const exits = roomTrait?.exits || {};
            for (const [direction, exitInfo] of Object.entries(exits)) {
                let targetRoom;
                let pathElement;
                if (typeof exitInfo === 'string') {
                    // Simple string - treat as direct room connection
                    targetRoom = exitInfo;
                    pathElement = exitInfo; // For now, use room ID as path element
                }
                else if (typeof exitInfo === 'object') {
                    // ExitInfo object
                    if (exitInfo.via) {
                        // Has a door/exit entity
                        const exitEntity = this.getEntity(exitInfo.via);
                        if (!exitEntity)
                            continue;
                        pathElement = exitInfo.via;
                        // Get destination based on entity type
                        if (exitEntity.hasTrait(trait_types_1.TraitType.DOOR)) {
                            const door = exitEntity.getTrait(door_1.DoorTrait);
                            targetRoom = door?.room1 === roomId ? door?.room2 : door?.room1;
                        }
                        else if (exitEntity.hasTrait(trait_types_1.TraitType.EXIT)) {
                            targetRoom = exitEntity.getTrait(exitTrait_1.ExitTrait)?.to;
                        }
                        else {
                            // Not a valid door/exit entity
                            continue;
                        }
                    }
                    else if (exitInfo.destination) {
                        // Direct room connection
                        targetRoom = exitInfo.destination;
                        pathElement = targetRoom; // Use room ID as path element
                    }
                }
                if (!targetRoom || !pathElement)
                    continue;
                if (targetRoom === toRoomId) {
                    // Found the destination - return path
                    // For direct connections, return empty array (no doors to pass through)
                    if (pathElement === targetRoom && path.length === 0) {
                        return [];
                    }
                    return [...path, pathElement];
                }
                if (!visited.has(targetRoom)) {
                    // For direct connections, don't add room ID to path
                    const newPath = pathElement === targetRoom ? path : [...path, pathElement];
                    queue.push({ roomId: targetRoom, path: newPath });
                }
            }
        }
        return null;
    }
    getPlayer() {
        return this.playerId ? this.getEntity(this.playerId) : undefined;
    }
    setPlayer(entityId) {
        if (!this.hasEntity(entityId)) {
            throw new Error(`Cannot set player to non-existent entity: ${entityId}`);
        }
        this.playerId = entityId;
    }
    // Score Ledger (ADR-129) — delegates to ScoreLedger
    awardScore(id, points, description) {
        return this.scoreLedger.award(id, points, description);
    }
    revokeScore(id) {
        return this.scoreLedger.revoke(id);
    }
    hasScore(id) {
        return this.scoreLedger.has(id);
    }
    getScore() {
        return this.scoreLedger.getTotal();
    }
    getScoreEntries() {
        return this.scoreLedger.getEntries();
    }
    setMaxScore(max) {
        this.scoreLedger.setMax(max);
    }
    getMaxScore() {
        return this.scoreLedger.getMax();
    }
    // Persistence — delegates to WorldSerializer
    toJSON() {
        return this.serializer.serialize(this.getSerializableState(), this.scoreLedger);
    }
    loadJSON(json) {
        // Preserve code registrations that aren't part of serialized state
        const savedEventChains = this.eventSystem.preserveChains();
        const savedCapabilities = { ...this.capabilities };
        // Clear current state
        this.clear();
        // Restore code registrations
        this.eventSystem.restoreChains(savedEventChains);
        this.capabilities = savedCapabilities;
        // Delegate deserialization
        const serializableState = this.getSerializableState();
        this.serializer.deserialize(json, serializableState, this.scoreLedger);
        // Sync primitives back (objects are shared by reference)
        this.playerId = serializableState.playerId;
    }
    getSerializableState() {
        return {
            entities: this.entities,
            spatialIndex: this.spatialIndex,
            state: this.state,
            playerId: this.playerId,
            relationships: this.relationships,
            idCounters: this.idCounters,
            capabilities: this.capabilities,
        };
    }
    clear() {
        this.entities.clear();
        this.state = {};
        this.playerId = undefined;
        this.spatialIndex.clear();
        this.relationships.clear();
        this.idCounters.clear();
        this.capabilities = {};
        this.scoreLedger.clear();
        this.grammarVocabularyProvider.clear();
        this.eventSystem.clear();
    }
    // Event system — delegates to WorldEventSystem
    registerEventHandler(eventType, handler) {
        this.eventSystem.registerEventHandler(eventType, handler);
    }
    unregisterEventHandler(eventType) {
        this.eventSystem.unregisterEventHandler(eventType);
    }
    registerEventValidator(eventType, validator) {
        this.eventSystem.registerEventValidator(eventType, validator);
    }
    registerEventPreviewer(eventType, previewer) {
        this.eventSystem.registerEventPreviewer(eventType, previewer);
    }
    connectEventProcessor(wiring) {
        this.eventSystem.connectEventProcessor(wiring);
    }
    chainEvent(triggerType, handler, options = {}) {
        this.eventSystem.chainEvent(triggerType, handler, options);
    }
    applyEvent(event) {
        this.eventSystem.applyEvent(event);
    }
    canApplyEvent(event) {
        return this.eventSystem.canApplyEvent(event);
    }
    previewEvent(event) {
        return this.eventSystem.previewEvent(event);
    }
    getAppliedEvents() {
        return this.eventSystem.getAppliedEvents();
    }
    getEventsSince(timestamp) {
        return this.eventSystem.getEventsSince(timestamp);
    }
    clearEventHistory() {
        this.eventSystem.clearEventHistory();
    }
    // Get the data store for sharing with AuthorModel
    getDataStore() {
        // Return a reference to the WorldModel's internal state
        // AuthorModel will share these same objects
        return {
            entities: this.entities,
            spatialIndex: this.spatialIndex,
            state: this.state,
            playerId: this.playerId,
            relationships: this.relationships,
            idCounters: this.idCounters,
            capabilities: this.capabilities
        };
    }
    // Scope Management Methods
    getScopeRegistry() {
        return this.scopeRegistry;
    }
    // Vocabulary Management (ADR-082)
    getGrammarVocabularyProvider() {
        return this.grammarVocabularyProvider;
    }
    addScopeRule(rule) {
        this.scopeRegistry.addRule(rule);
        this.emitPlatformEvent('scope_rule_added', {
            ruleId: rule.id,
            fromLocations: rule.fromLocations,
            forActions: rule.forActions
        });
    }
    removeScopeRule(ruleId) {
        const removed = this.scopeRegistry.removeRule(ruleId);
        if (removed) {
            this.emitPlatformEvent('scope_rule_removed', { ruleId });
        }
        return removed;
    }
    evaluateScope(actorId, actionId) {
        const actor = this.getEntity(actorId);
        if (!actor) {
            console.warn('evaluateScope: No actor found for ID:', actorId);
            return [];
        }
        const currentLocation = this.getLocation(actorId);
        if (!currentLocation) {
            console.warn('evaluateScope: No location found for actor:', actorId);
            return [];
        }
        const context = {
            world: this,
            actorId,
            actionId,
            currentLocation
        };
        const result = this.scopeEvaluator.evaluate(context);
        return Array.from(result.entityIds);
    }
    /**
     * Register default scope rules for standard visibility
     */
    registerDefaultScopeRules() {
        // Basic visibility rule - entities in same room
        this.addScopeRule({
            id: 'default_room_visibility',
            fromLocations: '*',
            includeEntities: (context) => {
                const room = this.getEntity(context.currentLocation);
                if (!room) {
                    console.warn('No room found for location:', context.currentLocation);
                    return [];
                }
                // Get all entities in the room
                const contents = this.getContents(context.currentLocation);
                const entityIds = contents.map(e => e.id);
                // Add the room itself
                entityIds.push(context.currentLocation);
                // Add nested contents (in containers, on supporters)
                for (const entity of contents) {
                    const nested = this.getAllContents(entity.id);
                    entityIds.push(...nested.map(e => e.id));
                }
                return entityIds;
            },
            priority: 50,
            source: 'core'
        });
        // Carried items are always in scope
        this.addScopeRule({
            id: 'default_inventory_visibility',
            fromLocations: '*',
            includeEntities: (context) => {
                const carried = this.getContents(context.actorId);
                const entityIds = carried.map(e => e.id);
                // Add nested contents of carried items
                for (const entity of carried) {
                    const nested = this.getAllContents(entity.id);
                    entityIds.push(...nested.map(e => e.id));
                }
                return entityIds;
            },
            priority: 100,
            source: 'core'
        });
    }
    /**
     * Get physically visible entities using VisibilityBehavior (line-of-sight)
     *
     * Returns entities that can actually be seen by the observer based on:
     * - Physical location (must be in same room)
     * - Container state (can't see inside closed opaque containers)
     * - Lighting conditions (limited visibility in darkness)
     * - Scenery visibility flags
     *
     * This is for perception and display purposes. For command resolution,
     * use getInScope() instead which includes entities that can be referenced
     * even if not visible.
     *
     * @param observerId - The entity doing the observing
     * @returns Array of entities that are physically visible
     */
    getVisible(observerId) {
        const observer = this.getEntity(observerId);
        if (!observer)
            return [];
        return VisibilityBehavior_1.VisibilityBehavior.getVisible(observer, this);
    }
    /**
     * Get entities in scope for command resolution using the scope system
     *
     * Returns entities that can be referenced in commands, including:
     * - All entities in the current room
     * - Items in containers (even closed ones)
     * - Carried items and their contents
     * - Entities made available by scope rules (e.g., through windows)
     *
     * This is for parser/command resolution. For physical visibility,
     * use getVisible() instead.
     *
     * @param observerId - The entity whose scope to evaluate
     * @returns Array of entities that can be referenced in commands
     */
    getInScope(observerId) {
        const entityIds = this.evaluateScope(observerId);
        return entityIds
            .map(id => this.getEntity(id))
            .filter((e) => e !== undefined);
    }
    /**
     * Check if observer can physically see target (line-of-sight)
     *
     * Uses VisibilityBehavior to determine if target is visible based on:
     * - Physical location
     * - Container state
     * - Lighting conditions
     *
     * This is for perception checks. To check if an entity can be
     * referenced in commands, check if it's in getInScope() instead.
     *
     * @param observerId - The entity doing the observing
     * @param targetId - The entity being observed
     * @returns true if observer can see target
     */
    canSee(observerId, targetId) {
        const observer = this.getEntity(observerId);
        const target = this.getEntity(targetId);
        if (!observer || !target)
            return false;
        return VisibilityBehavior_1.VisibilityBehavior.canSee(observer, target, this);
    }
    // ========== Convenience Creators ==========
    /**
     * Create a bidirectional connection between two rooms.
     * Sets exits in both directions (e.g. NORTH on room1, SOUTH on room2).
     */
    connectRooms(room1Id, room2Id, direction) {
        const room1 = this.getEntity(room1Id);
        const room2 = this.getEntity(room2Id);
        if (!room1 || !room2) {
            throw new Error(`connectRooms: both rooms must exist (${room1Id}, ${room2Id})`);
        }
        const opposite = (0, directions_1.getOppositeDirection)(direction);
        roomBehavior_1.RoomBehavior.setExit(room1, direction, room2Id);
        roomBehavior_1.RoomBehavior.setExit(room2, opposite, room1Id);
    }
    /**
     * Create a door entity and wire it into both rooms' exit data.
     * The door is placed in room1 spatially (for scope resolution).
     */
    createDoor(displayName, opts) {
        const room1 = this.getEntity(opts.room1Id);
        const room2 = this.getEntity(opts.room2Id);
        if (!room1 || !room2) {
            throw new Error(`createDoor: both rooms must exist (${opts.room1Id}, ${opts.room2Id})`);
        }
        const door = this.createEntity(displayName, entity_types_1.EntityType.DOOR);
        door.add(new identityTrait_1.IdentityTrait({
            name: displayName,
            aliases: opts.aliases,
            description: opts.description,
        }));
        door.add(new door_1.DoorTrait({
            room1: opts.room1Id,
            room2: opts.room2Id,
        }));
        door.add(new scenery_1.SceneryTrait());
        door.add(new openableTrait_1.OpenableTrait({ isOpen: opts.isOpen ?? false }));
        if (opts.isLocked !== undefined || opts.keyId) {
            door.add(new lockableTrait_1.LockableTrait({
                isLocked: opts.isLocked ?? true,
                ...(opts.keyId ? { requiredKey: opts.keyId } : {}),
            }));
        }
        // Wire exits through the door
        const opposite = (0, directions_1.getOppositeDirection)(opts.direction);
        roomBehavior_1.RoomBehavior.setExit(room1, opts.direction, opts.room2Id, door.id);
        roomBehavior_1.RoomBehavior.setExit(room2, opposite, opts.room1Id, door.id);
        // Place door in room1 for scope resolution
        this.moveEntity(door.id, opts.room1Id);
        return door;
    }
    // ── Region Management (ADR-149) ──────────────────────────────────
    /**
     * Creates a region entity with RegionTrait atomically.
     *
     * @param id - Explicit entity ID for the region (e.g., 'reg-underground').
     *             Must be unique; throws if already exists.
     * @param options - Region configuration.
     * @returns The created region entity.
     */
    createRegion(id, options) {
        if (this.hasEntity(id)) {
            throw new Error(`createRegion: entity '${id}' already exists`);
        }
        if (options.parentRegionId && !this.hasEntity(options.parentRegionId)) {
            throw new Error(`createRegion: parent region '${options.parentRegionId}' not found`);
        }
        const entity = new if_entity_1.IFEntity(id, entity_types_1.EntityType.REGION, {
            attributes: {
                displayName: options.name,
                name: options.name,
                entityType: entity_types_1.EntityType.REGION,
            },
        });
        entity.add(new regionTrait_1.RegionTrait(options));
        this.entities.set(id, entity);
        return entity;
    }
    /**
     * Assigns a room to a region by setting RoomTrait.regionId.
     *
     * @param roomId - ID of the room entity.
     * @param regionId - ID of the region entity (must exist and have RegionTrait).
     * @throws If room not found, region not found, or types are wrong.
     */
    assignRoom(roomId, regionId) {
        const room = this.getEntity(roomId);
        if (!room) {
            throw new Error(`assignRoom: room '${roomId}' not found`);
        }
        const roomTrait = room.get(trait_types_1.TraitType.ROOM);
        if (!roomTrait) {
            throw new Error(`assignRoom: entity '${roomId}' does not have RoomTrait`);
        }
        const region = this.getEntity(regionId);
        if (!region) {
            throw new Error(`assignRoom: region '${regionId}' not found`);
        }
        if (!region.hasTrait(trait_types_1.TraitType.REGION)) {
            throw new Error(`assignRoom: entity '${regionId}' is not a region`);
        }
        roomTrait.regionId = regionId;
    }
    /**
     * Tests whether an entity (or its containing room) is in a region,
     * traversing the parent region hierarchy.
     *
     * @param entityId - The entity to test. If it's a room, checks its regionId.
     *                   Otherwise, resolves the entity's containing room first.
     * @param regionId - The region to test membership against.
     * @returns true if the entity is in the region or any child of it.
     */
    isInRegion(entityId, regionId) {
        const entity = this.getEntity(entityId);
        if (!entity)
            return false;
        // If the entity is a room, use its regionId directly.
        // Otherwise, resolve the containing room.
        let roomRegionId;
        if (entity.type === entity_types_1.EntityType.ROOM) {
            roomRegionId = entity.get(trait_types_1.TraitType.ROOM)?.regionId;
        }
        else {
            const room = this.getContainingRoom(entityId);
            if (!room)
                return false;
            roomRegionId = room.get(trait_types_1.TraitType.ROOM)?.regionId;
        }
        if (!roomRegionId)
            return false;
        // Walk up the parent chain from the room's region
        return this.regionAncestryIncludes(roomRegionId, regionId);
    }
    /**
     * Computes which regions are exited and entered when moving between rooms.
     * Exit list is innermost-first; entry list is outermost-first.
     *
     * @param fromRoomId - The source room entity ID.
     * @param toRoomId - The destination room entity ID.
     * @returns Region IDs exited and entered. Both empty if same region or no regions.
     */
    getRegionCrossings(fromRoomId, toRoomId) {
        const fromRoom = this.getEntity(fromRoomId);
        const toRoom = this.getEntity(toRoomId);
        const fromChain = fromRoom
            ? this.getRegionAncestry(fromRoom.get(trait_types_1.TraitType.ROOM)?.regionId)
            : [];
        const toChain = toRoom
            ? this.getRegionAncestry(toRoom.get(trait_types_1.TraitType.ROOM)?.regionId)
            : [];
        // Convert to sets for fast lookup
        const fromSet = new Set(fromChain);
        const toSet = new Set(toChain);
        // Exited: regions in fromChain but not in toChain (innermost first — natural order)
        const exited = fromChain.filter(id => !toSet.has(id));
        // Entered: regions in toChain but not in fromChain (outermost first — reverse natural order)
        const entered = toChain.filter(id => !fromSet.has(id)).reverse();
        return { exited, entered };
    }
    // ── Private region helpers ───────────────────────────────────────────
    /**
     * Builds the ancestry chain for a region: [self, parent, grandparent, ...].
     * Returns empty array if regionId is undefined or not found.
     */
    getRegionAncestry(regionId) {
        const chain = [];
        let currentId = regionId;
        const visited = new Set(); // guard against cycles
        while (currentId) {
            if (visited.has(currentId))
                break;
            visited.add(currentId);
            const region = this.getEntity(currentId);
            if (!region)
                break;
            const trait = region.get(trait_types_1.TraitType.REGION);
            if (!trait)
                break;
            chain.push(currentId);
            currentId = trait.parentRegionId;
        }
        return chain;
    }
    /**
     * Checks whether a region ancestry chain includes a target region.
     */
    regionAncestryIncludes(startRegionId, targetRegionId) {
        let currentId = startRegionId;
        const visited = new Set();
        while (currentId) {
            if (currentId === targetRegionId)
                return true;
            if (visited.has(currentId))
                return false;
            visited.add(currentId);
            const region = this.getEntity(currentId);
            if (!region)
                return false;
            const trait = region.get(trait_types_1.TraitType.REGION);
            if (!trait)
                return false;
            currentId = trait.parentRegionId;
        }
        return false;
    }
    // ── Scene Management (ADR-149) ───────────────────────────────────
    /**
     * Creates a scene entity with SceneTrait and registers condition closures.
     *
     * @param id - Explicit entity ID (e.g., 'scene-flood'). Must be unique.
     * @param options - Scene name, begin/end conditions, recurring flag.
     * @returns The created scene entity.
     */
    createScene(id, options) {
        if (this.hasEntity(id)) {
            throw new Error(`createScene: entity '${id}' already exists`);
        }
        const entity = new if_entity_1.IFEntity(id, entity_types_1.EntityType.SCENE, {
            attributes: {
                displayName: options.name,
                name: options.name,
                entityType: entity_types_1.EntityType.SCENE,
            },
        });
        entity.add(new sceneTrait_1.SceneTrait({
            name: options.name,
            recurring: options.recurring,
        }));
        this.entities.set(id, entity);
        // Store condition closures separately (not serializable)
        this.sceneConditions.set(id, {
            begin: options.begin,
            end: options.end,
        });
        return entity;
    }
    /**
     * Retrieves the condition closures for a scene.
     * Returns undefined if the scene has no registered conditions
     * (e.g., after save/restore before re-registration).
     *
     * @param sceneId - The scene entity ID.
     */
    getSceneConditions(sceneId) {
        return this.sceneConditions.get(sceneId);
    }
    /**
     * Returns all registered scene conditions. Used by the engine's
     * scene evaluation phase to iterate over scenes each turn.
     */
    getAllSceneConditions() {
        return this.sceneConditions;
    }
    /**
     * Returns true if the scene is currently in the 'active' state.
     *
     * @param sceneId - The scene entity ID.
     */
    isSceneActive(sceneId) {
        const entity = this.getEntity(sceneId);
        if (!entity)
            return false;
        return entity.get(trait_types_1.TraitType.SCENE)?.state === 'active';
    }
    /**
     * Returns true if the scene has reached the 'ended' state.
     * For recurring scenes, this is true only while in the 'ended' state
     * (resets to 'waiting' when the scene re-activates).
     *
     * @param sceneId - The scene entity ID.
     */
    hasSceneEnded(sceneId) {
        const entity = this.getEntity(sceneId);
        if (!entity)
            return false;
        return entity.get(trait_types_1.TraitType.SCENE)?.state === 'ended';
    }
    /**
     * Returns true if the scene has ever been active (beganAtTurn is set).
     *
     * @param sceneId - The scene entity ID.
     */
    hasSceneHappened(sceneId) {
        const entity = this.getEntity(sceneId);
        if (!entity)
            return false;
        return entity.get(trait_types_1.TraitType.SCENE)?.beganAtTurn !== undefined;
    }
}
exports.WorldModel = WorldModel;
